---
# title: "PAR-CLIP_MD"
output: html_document
---
# Identification of CLIP peaks: Novoalign  
 
```{r include=F ,echo=F,warning=F,message=FALSE}
rm(list=ls())
# example
# chr1:59,692,533-59,692,616

# if (!requireNamespace("BiocManager", quietly = TRUE))
#     install.packages("BiocManager")

# 
# BiocManager::install("TxDb.Mmusculus.UCSC.mm10.knownGene")
# BiocManager::install('VariantAnnotation')
# BiocManager::install("org.Mm.eg.db")
# BiocManager::install("trackViewer")
# install.packages("WhopGenome")
# BiocManager::install("BayesPeak")
# BiocManager::install("biomaRt")
# BiocManager::install("ComplexHeatmap")
# install.packages("ggpubr")
# install.packages("car")
install.packages("installr")
library("installr")
remove.packages("ComplexHeatmap","ComplexHeatmap")

install.packages("devtools")
library(devtools)

Sys.setenv(R_REMOTES_NO_ERRORS_FROM_WARNINGS="true")
install_github("jokergoo/ComplexHeatmap")
    library("ComplexHeatmap")
BiocManager::install("ComplexHeatmap_2.5.1")
    install.packages("ComplexHeatmap_2.5.1")
    library("ComplexHeatmap_2.5.1")
    packageVersion("ComplexHeatmap")
source("https://bioconductor.org/biocLite.R")
biocLite("ComplexHeatmap")
    library("ComplexHeatmap")



library("car")
    library(VariantAnnotation,quietly = T,verbose = F,warn.conflicts = F,logical.return = F)
    library(GenomicRanges,quietly = T,verbose = F,warn.conflicts = F,logical.return = F)
    #library(WhopGenome)
    library(trackViewer)
    library("pheatmap")

    library(vcfR)
    library(seqinr,quietly = T,verbose = F,warn.conflicts = F,logical.return = F)
    library(ggplot2,quietly = T,verbose = F,warn.conflicts = F,logical.return = F)
    library("viridis",quietly = T,verbose = F,warn.conflicts = F,logical.return = F)
    library(edgeR,quietly = T,verbose = F)
    library('GenomicFeatures',quietly = T,verbose = F,warn.conflicts = F,logical.return = F)
    library('rtracklayer',quietly = T,verbose = F,warn.conflicts = F,logical.return = F)
    #library(GeneStructureTools)
    library(matrixStats,quietly = T,verbose = F,warn.conflicts = F,logical.return = F)
    library(plyr,quietly = T,verbose = F,warn.conflicts = F,logical.return = F)
    library(tidyr,quietly = T,verbose = F,warn.conflicts = F,logical.return = F)
    library(fitdistrplus,quietly = T,verbose = F,warn.conflicts = F,logical.return = F)
    library(stringr,quietly = T,verbose = F,warn.conflicts = F,logical.return = F)
    library(data.table)
    library(reshape)
    library(knitr)
library(stringi)
library(GeneStructureTools)
library(biomaRt)
library(plotly)
library(tidyr)
library(GenomicRanges)
library(RColorBrewer)
library('gplots')
library(ggpubr)
     library(circlize)
library('regioneR')

source("PARCLIP_Annotation_MD10.R")


blank_theme <- theme_minimal()+
  theme(
  axis.title.x = element_blank(),
  axis.title.y = element_blank(),
  panel.border = element_blank(),
  panel.grid=element_blank(),
  axis.ticks = element_blank(),
  plot.title=element_text(size=14, face="bold")
  )

incd_rRNA=T
```

  
## Processing data   

**1. Align reads Novoalign**  

novoalign -c 32 -t 15,3 -l 20 -x 4 -g 20 -s 1 -o SAM -R 0 -r All  
-d mm10_gencode_prerRNA   
-f Ro_Clip_iCountcutadpt.fastq  
<br>  

**2. UMI-tools remove duplicates**  
umi_tools dedup  
--method unique  
--multimapping-detection-method=NH  
--umi-separator rbc:   
--log2stderr   
-I Ro_Clip_iCountcutadpt_all.unique.NH.mm.s.bam    
-S Ro_Clip_iCountcutadpt_all.unique.NH.mm.ddup.bam   
-L Ro_Clip_iCountcutadpt_all.unique.NH.mm.s.log  
<br>  

**3. Select CLIP peaks**  

  Identify peaks using bedtools merge and count number of reads that overlap with the identified peaks in both the treated and Control samples.  


```{r echo=F,warning=F,warning=T,eval=T,include=F }

peaks=fread(paste0("./NovoAlign_umi/All/recomb_NHtag/Peaks_postUniq_postMM_inc45S/Ro_Clip_iCountcutadpt_all.unique.NH.mm.ddup.s.unique.bam.peaks.bedtools.bed"), header=F, sep="\t",stringsAsFactors = F,data.table=F)
peaks$V2=peaks$V2+1

peaks$ID=paste0(peaks$V1,":",peaks$V2,"-",peaks$V3)
peaks$ID2=paste0(peaks$V1,":",peaks$V2,"-",peaks$V3,'_',peaks$V6)


WT_counts2=fread(paste0("./rbl.wolin.parclip/inst/extdata/peaks/clip_Control_novoalign_random.peakscount.txt"), header=F, sep="\t",stringsAsFactors = F,data.table=F)
# WT_counts2=WT_counts2[WT_counts2$V1%in%chrm,]
WT_counts2$V2=WT_counts2$V2+1

WT_counts2$ID=paste0(WT_counts2$V1,":",(WT_counts2$V2),"-",WT_counts2$V3)
WT_counts2$ID2=paste0(WT_counts2$V1,":",(WT_counts2$V2),"-",WT_counts2$V3,'_',WT_counts2$V6)
WT_counts2$CPM=WT_counts2$V7/(24625/1000000)
colnames(WT_counts2)[colnames(WT_counts2)%in%'V7']='V7_WT'

peaks=merge(peaks,WT_counts2[,c('ID2','V7_WT',"CPM")],by='ID2',all.x=T)



#############################################
########## samtools
#############################################

########## Unique
Ro_countsUniq=fread(paste0("./NovoAlign_umi/All/recomb_NHtag/Peaks_postUniq_postMM_inc45S/Ro_Clip_iCountcutadpt_all.unique.NH.mm.ddup.s.unique.peakscount.txt"), header=F, sep="\t",stringsAsFactors = F,data.table=F)
Ro_countsUniq=Ro_countsUniq[,c('V1','V2','V3','V6','V7')]
colnames(Ro_countsUniq)= c('V1','V2','V3','V6','Counts_Unique')
Ro_countsUniq$V2=Ro_countsUniq$V2+1

Ro_countsUniq$ID=paste0(Ro_countsUniq$V1,":",(Ro_countsUniq$V2),"-",Ro_countsUniq$V3)
Ro_countsUniq$ID2=paste0(Ro_countsUniq$V1,":",(Ro_countsUniq$V2),"-",Ro_countsUniq$V3,'_',Ro_countsUniq$V6)
# Ro_countsUniq=Ro_countsUniq[Ro_countsUniq$V4>=5,]

#### Unique No strand

Ro_countsNS=fread(paste0("./NovoAlign_umi/All/recomb_NHtag/Peaks_postUniq_postMM_inc45S/Ro_Clip_iCountcutadpt_all.unique.NH.mm.ddup.s.unique.peakscountNoStrand.txt"), header=F, sep="\t",stringsAsFactors = F,data.table=F)
Ro_countsNS=Ro_countsNS[,c('V1','V2','V3','V6','V7')]
colnames(Ro_countsNS)= c('V1','V2','V3','V6','Counts_Unique_NStrd')
Ro_countsNS$V2=Ro_countsNS$V2+1

Ro_countsNS$ID=paste0(Ro_countsNS$V1,":",(Ro_countsNS$V2),"-",Ro_countsNS$V3)
Ro_countsNS$ID2=paste0(Ro_countsNS$V1,":",(Ro_countsNS$V2),"-",Ro_countsNS$V3,'_',Ro_countsNS$V6)
# Ro_countsNS=Ro_countsNS[Ro_countsNS$V4>=5,]



########## multimap
Ro_countsMM=fread(paste0("./NovoAlign_umi/All/recomb_NHtag/Peaks_postUniq_postMM_inc45S/Ro_Clip_iCountcutadpt_all.unique.NH.mm.ddup.s.mm.peakscount.txt"), header=F, sep="\t",stringsAsFactors = F,data.table=F)
Ro_countsMM=Ro_countsMM[,c('V1','V2','V3','V6','V7')]
colnames(Ro_countsMM)= c('V1','V2','V3','V6','Counts_MM')
Ro_countsMM$V2=Ro_countsMM$V2+1

Ro_countsMM$ID=paste0(Ro_countsMM$V1,":",(Ro_countsMM$V2),"-",Ro_countsMM$V3)
Ro_countsMM$ID2=paste0(Ro_countsMM$V1,":",(Ro_countsMM$V2),"-",Ro_countsMM$V3,'_',Ro_countsMM$V6)


########## recomb_NHtag
Ro_countsComb=fread(paste0("./NovoAlign_umi/All/recomb_NHtag/Peaks_postUniq_postMM_inc45S/Ro_Clip_iCountcutadpt_all.unique.NH.mm.ddup.s.peakscount.txt"), header=F, sep="\t",stringsAsFactors = F,data.table=F)
Ro_countsComb=Ro_countsComb[,c('V1','V2','V3','V6','V7')]
colnames(Ro_countsComb)= c('V1','V2','V3','V6','Counts_All')
Ro_countsComb$V2=Ro_countsComb$V2+1


Ro_countsComb$ID=paste0(Ro_countsComb$V1,":",(Ro_countsComb$V2),"-",Ro_countsComb$V3)
Ro_countsComb$ID2=paste0(Ro_countsComb$V1,":",(Ro_countsComb$V2),"-",Ro_countsComb$V3,'_',Ro_countsComb$V6)

#######################################################################################################################################
##################################################################################################################################


# #########
Ro_counts=merge(Ro_countsUniq,Ro_countsMM[,c("ID2","Counts_MM")],by='ID2',all.x=T)
Ro_counts=merge(Ro_counts,Ro_countsComb[,c("ID2","Counts_All")],by='ID2',all.x=T)
Ro_counts=merge(Ro_counts,WT_counts2[,c('ID2','V7_WT')],by='ID2',all.x=T)
colnames(Ro_counts)[which(colnames(Ro_counts)%in%c('V1','V2','V3','V6'))]=c('chr','start','end','strand')
# colnames(Ro_counts)[which(colnames(Ro_counts)%in%c('Unique_reads','MultiMap_reads','Counts_All'))]=c('Counts_Unique','Counts_MM','Counts_All')


peaksST=merge(peaks,Ro_counts[,c('ID2','Counts_Unique','Counts_MM','Counts_All')],by='ID2',all.x=T,suffix=c('_WT','_Ro'))
colnames(peaksST)[which(colnames(peaksST)%in%c('V1','V2','V3','V6'))]=c('chr','start','end','strand')


peaksST$perc_mm=(peaksST$Counts_MM/peaksST$Counts_All)*100
peaksST$Perc_Control=0
peaksST=peaksST[peaksST$Counts_Unique>=5,]
peaksST=peaksST[!peaksST$chr%in%c('chrM',"GL456211.1","GL456216.1","GL456233.1","JH584299.1","KK082441.1","KZ289077.1","KZ289079.1","KZ289081.1"),]
# unique(peaksST$V1)

# xcomp=merge(Ro_counts,Ro_countsx,by='ID',suffixes=c('bamtools','Custom'))
# xcomp$diff=(xcomp$V7bamtools-xcomp$V7Custom)
# plot(xcomp$V7bamtools,xcomp$V7Custom)

#######################################################################################################################################
##################################################################################################################################

# peaksST[is.na(peaksST$V7_WT),'V7_WT']=0
# peaksST$Perc_Control=(peaksST$V7_WT/peaksST$Counts_All)*100
# 
# peaksST[peaksST$Perc_Control>100,"Perc_Control"]=100

og=fread(paste0("CLIP_Peaks_MD6.txt"), header=T, sep="\t",stringsAsFactors = F,data.table=F)
# og=merge(og[,c('ID','strand','Counts_Unique')],peaksST[,c('ID','Counts_Unique')],by="ID",all.x=T,suffixes=c('.og','.new'))
# og=merge(og,peaks[,c('ID','V4')],by="ID",all.x=T);colnames(og)[colnames(og)%in%'V4']='V4.newBed.PreCut'
# 
# peaksbed=fread("./NovoAlign_umi/All/recomb_NHtag/Peaks_postUniq_postMM_inc45S/Ro_Clip_iCountcutadpt_all.unique.NH.mm.ddup.s.unique.bam.peaks.bed",header=F,sep="\t",check.names=FALSE,data.table=F)
# peaksbed$name=paste0(peaksbed$V1,":",(peaksbed$V2),"-",peaksbed$V3)
# og=merge(og,peaksbed[,c('name','V5')],by.y='name',by.x="ID",all.x=T);colnames(og)[colnames(og)%in%'V5']='V7.newPeaksbed.Cut'
# 
# og=merge(og,w3peaks_newtrim[,c('ID','count')],by="ID",all.x=T);colnames(og)[colnames(og)%in%'count']='count.ModBed_newtrim'

# og=merge(og,Ro_counts_comp[,c('ID','V7')],by="ID",all.x=T);colnames(og)[colnames(og)%in%'V7']='V7.newBedtools.postCut'
# og=merge(og[,c('ID','Counts_Unique')],Ro_counts_comp[,c('ID','Counts_Unique','Unique_reads_NStrd','diff')],by="ID",all.x=T)
# og$diffog=og$Counts_Unique.x-og$Counts_Unique.y

# 
# og$width=abs(og$end-og$start)

```


```{r fig.asp=.5, fig.align='center',fig.height=6, fig.width=12,echo=F,warning=F,eval=T,include=F}

# mm=read.delim(paste0("./NovoAlign_umi/All/recomb_NHtag/peaks_Dist50nt_postUniq_postMM_2primary/Ro_Clip_iCountcutadpt_all.unique.NH.mm.ddup.s.mm.names.txt"), header=F, sep="\t",stringsAsFactors = F)
# mm=separate(mm,V1,into = c('name','umi'),sep = ':rbc:',remove = F)
# 
# primary=read.delim(paste0("./NovoAlign_umi/All/recomb_NHtag/peaks_Dist50nt_postUniq_postMM_2primary/Ro_Clip_iCountcutadpt_all.unique.NH.mm.ddup.s.mmPrimary.names.txt"), header=F, sep="\t",stringsAsFactors = F)
# primary=separate(primary,V1,into = c('name','umi'),sep = ':rbc:',remove = F)

#############################################
########## Feature Counts
#############################################

FtrCount_uniq=read.delim(paste0("./NovoAlign_umi/All/recomb_NHtag/peaks_Dist50nt_postUniq_postMM_2primary/Ro_Clip_iCountcutadpt_all.unique.NH.mm.ddup.s.FCountUnique.txt"), header=T, sep="\t",stringsAsFactors = F,comment.char = '#');colnames(FtrCount_uniq)[7]='Counts';
FtrCount_uniq$Start=as.numeric(FtrCount_uniq$Start)-1;FtrCount_uniq$End=as.numeric(FtrCount_uniq$End)
FtrCount_uniq=FtrCount_uniq[!is.na(FtrCount_uniq$Start),]
FtrCount_uniq$Start=FtrCount_uniq$Start+1

FtrCount_uniq$ID=paste0(FtrCount_uniq$Chr,":",FtrCount_uniq$Start,"-",FtrCount_uniq$End)
FtrCount_uniq$ID2=paste0(FtrCount_uniq$Chr,":",FtrCount_uniq$Start,"-",FtrCount_uniq$End,"_",FtrCount_uniq$Strand)

######################################################################
FtrCount_uniq=FtrCount_uniq[FtrCount_uniq$Counts>=5,]
######################################################################


FtrCount_all=read.delim(paste0("./NovoAlign_umi/All/recomb_NHtag/peaks_Dist50nt_postUniq_postMM_2primary/Ro_Clip_iCountcutadpt_all.unique.NH.mm.ddup.s.FCountAll.txt"), header=T, sep="\t",stringsAsFactors = F,comment.char = '#');colnames(FtrCount_all)[7]='Counts';
FtrCount_all$Start=as.numeric(FtrCount_all$Start)-1;FtrCount_all$End=as.numeric(FtrCount_all$End)
FtrCount_all=FtrCount_all[!is.na(FtrCount_all$Start),]
FtrCount_all$Start=FtrCount_all$Start+1

FtrCount_all$ID=paste0(FtrCount_all$Chr,":",FtrCount_all$Start,"-",FtrCount_all$End)
FtrCount_all$ID2=paste0(FtrCount_all$Chr,":",FtrCount_all$Start,"-",FtrCount_all$End,"_",FtrCount_all$Strand)


FtrCount_Primary=read.delim(paste0("./NovoAlign_umi/All/recomb_NHtag/peaks_Dist50nt_postUniq_postMM_2primary/Ro_Clip_iCountcutadpt_all.unique.NH.mm.ddup.s.FCountAll_primary.txt"), header=T, sep="\t",stringsAsFactors = F,comment.char = '#');colnames(FtrCount_Primary)[7]='Counts';
FtrCount_Primary$Start=as.numeric(FtrCount_Primary$Start)-1;FtrCount_Primary$End=as.numeric(FtrCount_Primary$End)
FtrCount_Primary=FtrCount_Primary[!is.na(FtrCount_Primary$Start),]
FtrCount_Primary$Start=FtrCount_Primary$Start+1

FtrCount_Primary$ID=paste0(FtrCount_Primary$Chr,":",FtrCount_Primary$Start,"-",FtrCount_Primary$End)

FtrCount_frac=read.delim(paste0("./NovoAlign_umi/All/recomb_NHtag/peaks_Dist50nt_postUniq_postMM_2primary/Ro_Clip_iCountcutadpt_all.unique.NH.mm.ddup.s.FCountAll_frac.txt"), header=T, sep="\t",stringsAsFactors = F,comment.char = '#');colnames(FtrCount_frac)[7]='Counts';
FtrCount_frac$Start=as.numeric(FtrCount_frac$Start)-1;FtrCount_frac$End=as.numeric(FtrCount_frac$End)
FtrCount_frac=FtrCount_frac[!is.na(FtrCount_frac$Start),]
FtrCount_frac$Start=FtrCount_frac$Start+1

FtrCount_frac$ID=paste0(FtrCount_frac$Chr,":",FtrCount_frac$Start,"-",FtrCount_frac$End)


FtrCount_frac_prim=read.delim(paste0("./NovoAlign_umi/All/recomb_NHtag/peaks_Dist50nt_postUniq_postMM_2primary/Ro_Clip_iCountcutadpt_all.unique.NH.mm.ddup.s.FCountAll_FracPrime.txt"), header=T, sep="\t",stringsAsFactors = F,comment.char = '#');colnames(FtrCount_frac_prim)[7]='Counts'
FtrCount_frac_prim$Start=as.numeric(FtrCount_frac_prim$Start)-1;FtrCount_frac_prim$End=as.numeric(FtrCount_frac_prim$End)
FtrCount_frac_prim=FtrCount_frac_prim[!is.na(FtrCount_frac_prim$Start),]
FtrCount_frac_prim$Start=FtrCount_frac_prim$Start+1

FtrCount_frac_prim$ID=paste0(FtrCount_frac_prim$Chr,":",FtrCount_frac_prim$Start,"-",FtrCount_frac_prim$End)


FtrCount=merge(FtrCount_uniq,FtrCount_all[,c('ID','Counts')],by='ID',suffixes=c("_Unique","_All"),all.x=T)
FtrCount=merge(FtrCount,FtrCount_frac[,c('ID','Counts')],by='ID',all.x=T)
FtrCount=merge(FtrCount,FtrCount_Primary[,c('ID','Counts')],by='ID',suffixes=c("_fracMM","_primary"),all.x=T)
FtrCount=merge(FtrCount,FtrCount_frac_prim[,c('ID','Counts')],by='ID',all.x=T)
colnames(FtrCount)[which(colnames(FtrCount)%in%'Counts')]='Counts_Primary_fracMM'
colnames(FtrCount)[which(colnames(FtrCount)%in%c('Chr','Start','End','Strand'))]=c('chr','start','end','strand')
FtrCount$Counts_MM=FtrCount$Counts_All-FtrCount$Counts_Unique

FtrCount=FtrCount[duplicated(FtrCount)==F,]


#######

FtrCount_compare=merge(Ro_counts[,c('ID','strand','Counts_Unique','Counts_MM',"Counts_All")],FtrCount_uniq[,c('ID','Counts')],by='ID',all=T)
FtrCount_compare=FtrCount_compare[duplicated(FtrCount_compare)==F,]
FtrCount_compare=merge(FtrCount_compare,FtrCount_all[,c('ID','Counts')],by='ID',suffixes=c("_Unique_ftrCount","_All_ftrCount"),all.x=T)
FtrCount_compare=merge(FtrCount_compare,FtrCount_frac[,c('ID','Counts')],by='ID',all.x=T)
FtrCount_compare=merge(FtrCount_compare,FtrCount_Primary[,c('ID','Counts')],by='ID',suffixes=c("_frac_ftrCount","_primary_ftrCount"),all.x=T)
FtrCount_compare=merge(FtrCount_compare,FtrCount_frac_prim[,c('ID','Counts')],by='ID',all.x=T)
colnames(FtrCount_compare)[which(colnames(FtrCount_compare)%in%c('Counts'))]=c('Counts_Primary_fracMM')

FtrCount_compare=FtrCount_compare[duplicated(FtrCount_compare)==F,]
FtrCount_compare$diff_Pimary=FtrCount_compare$Counts_primary-FtrCount_compare$Counts_Primary_fracMM
 
FtrCount_compare$diff_unique=FtrCount_compare$Counts_Unique-FtrCount_compare$Counts_Unique
FtrCount_compare$diff_All=FtrCount_compare$Counts_All-FtrCount_compare$Counts_All
FtrCount_compare$diff=FtrCount_compare$Counts_Unique-FtrCount_compare$Counts_All



#############################################
########## Feature Counts Control
#############################################

FtrCountCntrl_uniq=read.delim(paste0("./NovoAlign_umi/All/recomb_NHtag/peaks_Dist50nt_postUniq_postMM_2primary/Control_Clip_iCountcutadpt_all.unique.NH.mm.ddup.s.FCountUnique.txt"), header=T, sep="\t",stringsAsFactors = F,comment.char = '#');colnames(FtrCountCntrl_uniq)[7]='CountsCntrl_Unique';
FtrCountCntrl_uniq$Start=as.numeric(FtrCountCntrl_uniq$Start)-1;FtrCountCntrl_uniq$End=as.numeric(FtrCountCntrl_uniq$End)
FtrCountCntrl_uniq=FtrCountCntrl_uniq[!is.na(FtrCountCntrl_uniq$Start),]
FtrCountCntrl_uniq$Start=FtrCountCntrl_uniq$Start+1

FtrCountCntrl_uniq$ID=paste0(FtrCountCntrl_uniq$Chr,":",FtrCountCntrl_uniq$Start,"-",FtrCountCntrl_uniq$End)
FtrCountCntrl_uniq$ID2=paste0(FtrCountCntrl_uniq$Chr,":",FtrCountCntrl_uniq$Start,"-",FtrCountCntrl_uniq$End,"_",FtrCountCntrl_uniq$Strand)


FtrCountCntrl_all=read.delim(paste0("./NovoAlign_umi/All/recomb_NHtag/peaks_Dist50nt_postUniq_postMM_2primary/Control_Clip_iCountcutadpt_all.unique.NH.mm.ddup.s.FCountAll.txt"), header=T, sep="\t",stringsAsFactors = F,comment.char = '#');colnames(FtrCountCntrl_all)[7]='CountsCntrl_All';
FtrCountCntrl_all$Start=as.numeric(FtrCountCntrl_all$Start)-1;FtrCountCntrl_all$End=as.numeric(FtrCountCntrl_all$End)
FtrCountCntrl_all=FtrCountCntrl_all[!is.na(FtrCountCntrl_all$Start),]
FtrCountCntrl_all$Start=FtrCountCntrl_all$Start+1

FtrCountCntrl_all$ID=paste0(FtrCountCntrl_all$Chr,":",FtrCountCntrl_all$Start,"-",FtrCountCntrl_all$End)
FtrCountCntrl_all$ID2=paste0(FtrCountCntrl_all$Chr,":",FtrCountCntrl_all$Start,"-",FtrCountCntrl_all$End,"_",FtrCountCntrl_all$Strand)


FtrCountCntrl_Primary=read.delim(paste0("./NovoAlign_umi/All/recomb_NHtag/peaks_Dist50nt_postUniq_postMM_2primary/Control_Clip_iCountcutadpt_all.unique.NH.mm.ddup.s.FCountAll_primary.txt"), header=T, sep="\t",stringsAsFactors = F,comment.char = '#');colnames(FtrCountCntrl_Primary)[7]='CountsCntrl_Primary';
FtrCountCntrl_Primary$Start=as.numeric(FtrCountCntrl_Primary$Start)-1;FtrCountCntrl_Primary$End=as.numeric(FtrCountCntrl_Primary$End)
FtrCountCntrl_Primary=FtrCountCntrl_Primary[!is.na(FtrCountCntrl_Primary$Start),]
FtrCountCntrl_Primary$Start=FtrCountCntrl_Primary$Start+1

FtrCountCntrl_Primary$ID=paste0(FtrCountCntrl_Primary$Chr,":",FtrCountCntrl_Primary$Start,"-",FtrCountCntrl_Primary$End)

FtrCountCntrl_frac=read.delim(paste0("./NovoAlign_umi/All/recomb_NHtag/peaks_Dist50nt_postUniq_postMM_2primary/Control_Clip_iCountcutadpt_all.unique.NH.mm.ddup.s.FCountAll_frac.txt"), header=T, sep="\t",stringsAsFactors = F,comment.char = '#');colnames(FtrCountCntrl_frac)[7]='CountsCntrl_frac_ftrCount';
FtrCountCntrl_frac$Start=as.numeric(FtrCountCntrl_frac$Start)-1;FtrCountCntrl_frac$End=as.numeric(FtrCountCntrl_frac$End)
FtrCountCntrl_frac=FtrCountCntrl_frac[!is.na(FtrCountCntrl_frac$Start),]
FtrCountCntrl_frac$Start=FtrCountCntrl_frac$Start+1

FtrCountCntrl_frac$ID=paste0(FtrCountCntrl_frac$Chr,":",FtrCountCntrl_frac$Start,"-",FtrCountCntrl_frac$End)


FtrCountCntrl_frac_prim=read.delim(paste0("./NovoAlign_umi/All/recomb_NHtag/peaks_Dist50nt_postUniq_postMM_2primary/Control_Clip_iCountcutadpt_all.unique.NH.mm.ddup.s.FCountAll_FracPrime.txt"), header=T, sep="\t",stringsAsFactors = F,comment.char = '#');colnames(FtrCountCntrl_frac_prim)[7]='CountsCntrl_Primary_fracMM'
FtrCountCntrl_frac_prim$Start=as.numeric(FtrCountCntrl_frac_prim$Start)-1;FtrCountCntrl_frac_prim$End=as.numeric(FtrCountCntrl_frac_prim$End)
FtrCountCntrl_frac_prim=FtrCountCntrl_frac_prim[!is.na(FtrCountCntrl_frac_prim$Start),]
FtrCountCntrl_frac_prim$Start=FtrCountCntrl_frac_prim$Start+1

FtrCountCntrl_frac_prim$ID=paste0(FtrCountCntrl_frac_prim$Chr,":",FtrCountCntrl_frac_prim$Start,"-",FtrCountCntrl_frac_prim$End)


FtrCount=FtrCount[duplicated(FtrCount)==F,]

FtrCount=merge(FtrCount,FtrCountCntrl_uniq[,c('ID','CountsCntrl_Unique')],by='ID',all.x=T)
FtrCount=merge(FtrCount,FtrCountCntrl_all[,c('ID','CountsCntrl_All')],by='ID',all.x=T)
FtrCount=merge(FtrCount,FtrCountCntrl_Primary[,c('ID','CountsCntrl_Primary')],by='ID',all.x=T)
FtrCount=merge(FtrCount,FtrCountCntrl_frac[,c('ID','CountsCntrl_frac_ftrCount')],by='ID',all.x=T)
FtrCount=merge(FtrCount,FtrCountCntrl_frac_prim[,c('ID','CountsCntrl_Primary_fracMM')],by='ID',all.x=T)

FtrCount=FtrCount[duplicated(FtrCount)==F,]

FtrCount$perc_mm=(FtrCount$Counts_MM/FtrCount$Counts_All)*100
FtrCount$Perc_Control=0

FtrCount=FtrCount[!FtrCount$chr%in%c('chrM',"GL456211.1","GL456216.1","GL456233.1","JH584299.1","KK082441.1","KZ289077.1","KZ289079.1","KZ289081.1"),]

FtrCount=FtrCount[duplicated(FtrCount)==F,]

# 'Counts_All','Counts_Unique','Counts_MM','Counts_primary','Counts_fracMM','Counts_Primary_fracMM'
```

```{r ,fig.align='center',fig.height=6, fig.width=12,echo=F,warning=F,eval=T,include=T}
##############################################
#### Select Samtools or Feature Counts
##############################################
 
peaks2=FtrCount
# peaks2=peaksST ### rRNA looks alot bigger because double count large cDNA regions for both neg and positive reads
 
# peaks2=peaks2[!peaks2$chr%in%'BK000964.3',]
```

```{r ,fig.align='center',fig.height=6, fig.width=12,echo=F,warning=F,eval=T,include=T}


 
x=ggplot(peaks2,aes(x=Counts_Unique))+geom_histogram(bins = 30)+scale_x_continuous(trans = "log2") + theme_classic()+ylab("# of peaks")+xlab("Peak Counts (unique) for Ro-Clip")+ggtitle("Distribution of peak counts for Ro-Clip")
#+geom_vline(xintercept = (100),color="red")+geom_vline(xintercept = (10),color="blue"))
ggplotly(x,tooltip = c("x", "y"))
```

```{r ,fig.align='center',fig.height=3, fig.width=6,echo=F,warning=F,eval=T,include=F}

p=peaks2[,c("CountsCntrl_Unique","CountsCntrl_All","CountsCntrl_Primary","CountsCntrl_frac_ftrCount"),drop=F]
colnames(p)=c("Unique","Unique + MM","Unique + Best map MM","Unique + scaled MM")
# p=peaks2[,c("CountsCntrl_Unique","CountsCntrl_All","CountsCntrl_Primary","CountsCntrl_frac_ftrCount","Counts_Unique","Counts_All","Counts_primary","Counts_fracMM"),drop=F]
p=melt(p);colnames(p)=c('Counting_method','Counts')
p$Counts=p$Counts+1
```
```{r ,fig.align='center',fig.height=4, fig.width=8,echo=F,warning=F,eval=T,include=T}
x=ggplot(p,aes(x=Counts,line=Counting_method,color=Counting_method))+geom_density(size=1)+scale_x_continuous(trans = "log2") + theme_classic()+ylab("Density")+xlab("Peak Counts for Control sample")+ggtitle("Distribution of peak counts for Control samples")+scale_color_brewer(palette = "Dark2")
#+geom_vline(xintercept = (100),color="red")+geom_vline(xintercept = (10),color="blue"))
# ggplotly(x,tooltip = c("x", "y"))
ggplotly(x)

# p$Counts=log2(p$Counts)
# x=ggplot(p,aes(x=Counts,line=Counting_method,color=Counting_method))+geom_density() + theme_classic()+ylab("# of peaks")+xlab("Peak Counts (unique) for Ro-Clip")+ggtitle("Distribution of peak counts for Ro-Clip")
# #+geom_vline(xintercept = (100),color="red")+geom_vline(xintercept = (10),color="blue"))
# # ggplotly(x,tooltip = c("x", "y"))
# x

# x=ggplot(peaks2,aes(x=CountsCntrl_Unique))+geom_histogram(bins = 30)+scale_x_continuous(trans = "log2") + theme_classic()+ylab("# of peaks")+xlab("Peak Counts (unique) for Ro-Clip")+ggtitle("Distribution of peak counts for Ro-Clip")
# #+geom_vline(xintercept = (100),color="red")+geom_vline(xintercept = (10),color="blue"))
# ggplotly(x,tooltip = c("x", "y"))
```

```{r ,fig.align='center',fig.height=6, fig.width=12,echo=F,warning=F,eval=F,include=F}
  x=ggplot(peaks2,aes(x=Counts_All,y=Counts_MM))+geom_point(aes(text=ID))+scale_y_continuous(trans='log2')+scale_x_continuous(trans='log2')+
  theme_classic()+
  xlab("Total Read Count")+
  ylab("Multimapped Read Count")+ 
  ggtitle("Comparison of unique Reads to Total Reads")
ggplotly(x,tooltip = c("x", "y",'text'))

  x=ggplot(peaks2,aes(x=Counts_Unique,y=Counts_MM))+geom_point(aes(text=ID))+scale_y_continuous(trans='log2')+scale_x_continuous(trans='log2')+
  theme_classic()+
  xlab("Unique Read Count")+
  ylab("Multimapped Read Count")+ 
  ggtitle("Comparison of unique Reads to Total Reads")
ggplotly(x,tooltip = c("x", "y",'text'))

  x=ggplot(peaks2,aes(x=Counts_All,y=perc_mm))+geom_point(aes(text=ID))+scale_x_continuous(trans='log2')+
  theme_classic()+
  xlab("Total Read Count")+
  ylab("% Multimapped Reads")+ 
  ggtitle("Comparison of unique Reads to Total Reads")
ggplotly(x,tooltip = c("x", "y",'text'))

```

```{r ,fig.align='center',fig.height=6, fig.width=12,echo=F,warning=F,eval=F,include=F}
  x=ggplot(peaks2,aes(x=Counts_All,y=Perc_Control))+geom_point()+scale_x_continuous(trans='log2')+geom_vline(xintercept = (100),color="red")+geom_vline(xintercept = (100),color="red")+geom_hline(yintercept = (10),color="blue")+theme_classic()+xlab("Peak Counts for Ro-Clip")+ylab("Ro-Clip counts / Control counts")+ ggtitle("Comparison of Peak counts between Ro-Clip and Control")
ggplotly(x,tooltip = c("x", "y"))

   
# Red line indicates a hard cut off of 100 counts in the Ro-Clip sample.  
# Blue line indicates a hard cut off with a 10% ratio of wild type reads compared to the Ro-Clip. 
# <br>  

```
```{r fig.align='center',fig.height=6, fig.width=12,echo=F,warning=F,eval=T}

# peaks2=peaks2[,c("V1","V2","V3","V6","ID","V7_WT",'Counts_Unique','Counts_MM','Counts_All','perc_mm',"Perc_Control")]

colnames(peaks2)[colnames(peaks2)%in%c("V7_WT")]=c('counts_Cntrl')


peaks.GR = GRanges(seqnames = as.character(peaks2$chr), ranges=IRanges(start = as.numeric(peaks2$start), end = as.numeric(peaks2$end)),strand = peaks2$strand,
                   ID=peaks2$ID,
                   Counts_All=peaks2$Counts_All,
                   Counts_Unique=peaks2$Counts_Unique,
                   Counts_MM=peaks2$Counts_MM,
                   Perc_Control=peaks2$Perc_Control )


```



```{r include=F ,echo=F,warning=F,eval=F}

################################################
### pileup to get read depth 
# 
bamEdit=BamFile("./PAR-iCLIP/clip_Ro_2_unq.bam")
# pupx=PileupParam(distinguish_strands=T, distinguish_nucleotides=T,include_deletions=T,include_insertions =T)
pupx=PileupParam(distinguish_strands=T,distinguish_nucleotides=F,include_deletions=T,include_insertions =T,min_base_quality=0,max_depth=10000)

pu=pileup(bamEdit,pileupParam=pupx)
# unique(pu$seqnames)
pu$pu_ID=paste0(pu$seqnames,":",pu$pos,"_",pu$strand)
```

```{r echo=F,warning=F,include=F ,eval=F}
# **2. Identify CLIP peak Mutations**   
#     
#   mpileup | bcfctools call variants  
```


```{r include=F ,echo=F,warning=F,eval=F}
mpile_Var <- read.vcfR( "./rbl.wolin.parclip/inst/extdata/Variants/clip_Ro_2_novoalign_unq_mpileonly3.vcf", verbose = FALSE )
  mpile_Var <- cbind(as.data.frame(getFIX(mpile_Var)), INFO2df(mpile_Var))
mpile_Var=separate(mpile_Var,DP4,into=c('ref_F','ref_R','alt_F','alt_R'),sep = ",")
mpile_Var$ref_F=as.numeric(mpile_Var$ref_F);mpile_Var$ref_R=as.numeric(mpile_Var$ref_R)
mpile_Var$alt_F=as.numeric(mpile_Var$alt_F);mpile_Var$alt_R=as.numeric(mpile_Var$alt_R)

mpile_Var$HQref=rowSums(mpile_Var[,c('ref_F','ref_R')])
mpile_Var$HQalt=rowSums(mpile_Var[,c('alt_F','alt_R')])

mpile_Var$ALT =(stri_split_fixed(mpile_Var$ALT, ',', simplify=T))[,1]
mpile_Var$TYPE=NA
mpile_Var[(stri_length(mpile_Var$REF)+stri_length(mpile_Var$ALT))==2,'TYPE']="snp"
mpile_Var[(stri_length(mpile_Var$REF)+stri_length(mpile_Var$ALT))>2,'TYPE']="ins/del"
mpile_Var[(stri_length(mpile_Var$REF)+stri_length(mpile_Var$ALT))>2,c('REF','ALT')]=NA

mpile_Var$Variant=paste0(mpile_Var$REF,"-",mpile_Var$ALT)


mpile_Var$freq=mpile_Var$HQalt/(mpile_Var$HQref+mpile_Var$HQalt)
mpile_Var=mpile_Var[mpile_Var$freq>.5,]
mpile_Var=mpile_Var[(mpile_Var$HQref+mpile_Var$HQalt)>10,]


mpile_Var$IDvar=paste0(mpile_Var$CHROM,":",mpile_Var$POS)



mpile_Var.GR = GRanges(seqnames = as.character(mpile_Var$CHROM), ranges=IRanges(start = as.numeric(as.character(mpile_Var$POS)), end = as.numeric(as.character(mpile_Var$POS))),strand = "*",IDvar=mpile_Var$IDvar,REF=mpile_Var$REF,ALT=mpile_Var$ALT,TYPE=mpile_Var$TYPE,Variant=mpile_Var$Variant,DP=mpile_Var$DP,HQref=mpile_Var$HQref,HQalt=mpile_Var$HQalt,freq=mpile_Var$freq )

mpile_Var2=as.data.table(mpile_Var.GR)


        gr = peaks.GR
        gr2 = mpile_Var.GR
        xo=as.data.frame(GenomicRanges::findOverlaps(gr,gr2,type = "any",ignore.strand=F))
          
        qh=as.data.frame(gr[xo$queryHits],row.names = NULL)
          colnames(qh)=paste0(colnames(qh),"_peaks")
        sh=as.data.frame(gr2[xo$subjectHits],row.names = NULL)
        
        
        peaks_Anno_mpile=cbind(qh,sh)

        peaks_Anno_mpile=merge(peaks2,peaks_Anno_mpile,by.x="ID",by.y="ID_peaks",all.x=T)


peaks_Anno_mpile$IDvar2=paste0(peaks_Anno_mpile$IDvar,"_",peaks_Anno_mpile$V5)
peaks_Anno_mpile=peaks_Anno_mpile[duplicated(peaks_Anno_mpile[,colnames(peaks_Anno_mpile)%in%c('pu_ID',"count")==F,])==F,]

        peaks_Anno_mpile=merge(peaks_Anno_mpile,pu[,c('pu_ID',"count")],by.x="IDvar2",by.y="pu_ID",all.x=T)
        peaks_Anno_mpile=peaks_Anno_mpile[duplicated(peaks_Anno_mpile[,colnames(peaks_Anno_mpile)%in%c('pu_ID',"count")==F,])==F,]
                
                major=(peaks_Anno_mpile$count/peaks_Anno_mpile$V4)
                peaks_Anno_mpile[which((major>.25|is.na(major))==F),colnames(peaks_Anno_mpile)%in%colnames(peaks2)==F]=NA
                
peaks_Anno_mpile=peaks_Anno_mpile[is.na(peaks_Anno_mpile$IDvar)==F,]

```

```{r include=F ,echo=F,warning=F,eval=F}

# FB_Var <- read.vcfR( "./rbl.wolin.parclip/inst/extdata/Variants/clip_Ro_2_novoalign_random_FB.vcf", verbose = FALSE )
#   FB_Var <- cbind(as.data.frame(getFIX(FB_Var)), INFO2df(FB_Var))
FB_Var2 <- read.vcfR( "./rbl.wolin.parclip/inst/extdata/Variants/clip_Ro_2_novoalign_random_FB2.vcf", verbose = FALSE )
  FB_Var2 <- cbind(as.data.frame(getFIX(FB_Var2)), INFO2df(FB_Var2))
# FB_Var3 <- read.vcfR( "./rbl.wolin.parclip/inst/extdata/Variants/clip_Ro_2_novoalign_random_FB3.vcf", verbose = FALSE )
#   FB_Var3 <- cbind(as.data.frame(getFIX(FB_Var3)), INFO2df(FB_Var3))
# xxx  
# FB_Var$IDvar=paste0(FB_Var$CHROM,":",FB_Var$POS)
FB_Var2$IDvar=paste0(FB_Var2$CHROM,":",FB_Var2$POS)
# FB_Var3$IDvar=paste0(FB_Var3$CHROM,":",FB_Var3$POS)

################
#### Get Allele count
################
mat <- stri_split_fixed(FB_Var2$AO, ',', simplify=T)
mat <- `dim<-`(as.numeric(mat), dim(mat))  # convert to numeric and save dims
FB_Var2$AOcount=rowSums(mat, na.rm=T)
# View(FB_Var2[,c("AO","AOcount")])
FB_Var2$freq=FB_Var2$AOcount/FB_Var2$DP
FB_Var2cut=FB_Var2[FB_Var2$freq>.5,]
FB_Var2cut=FB_Var2cut[(FB_Var2cut$DP)>50,]
        

#### Get Allele TYPE
FB_Var2cut$Variant=paste0(FB_Var2cut$REF,"-",FB_Var2cut$ALT)
FB_Var2cut$Variant=NA


#### Get variant type
FB_Var2cut=FB_Var2cut[grepl("snp",FB_Var2cut$TYPE)==F,]
FB_Var2cut[grepl("del",FB_Var2cut$TYPE),"TYPE"]="ins/del"
FB_Var2cut[grepl("ins",FB_Var2cut$TYPE),"TYPE"]="ins/del"
FB_Var2cut[grepl("complex",FB_Var2cut$TYPE),"TYPE"]="ins/del"
FB_Var2cut[grepl("mnp",FB_Var2cut$TYPE),"TYPE"]="ins/del"
# unique(FB_Var2cut$TYPE)

mat <- stri_split_fixed((FB_Var2cut$TYPE), ',', simplify=T)
for (n in 1:nrow(mat)) {
  v=unique(mat[n,])
  v=v[(v>0)]
  if(length(v)>1){v=paste(v,collapse = ",")}
  FB_Var2cut[n,"TYPE"]=v
}

```
```{r include=F,eval=F ,echo=F,warning=F}
################
#### overlap with CLIP peaks
################

FB_Varcut=FB_Var2cut
FB_Var.GR = GRanges(seqnames = as.character(FB_Varcut$CHROM), ranges=IRanges(start = as.numeric(as.character(FB_Varcut$POS)), end = as.numeric(as.character(FB_Varcut$POS))),strand = "*",IDvar=FB_Varcut$IDvar,REF=FB_Varcut$REF,ALT=FB_Varcut$ALT,Variant=FB_Varcut$Variant,type=FB_Varcut$TYPE,DP=FB_Varcut$DP,AF=FB_Varcut$AF,AOcount=FB_Varcut$AOcount,AO=FB_Varcut$AO,freq=FB_Varcut$freq,PAO=FB_Varcut$PAO,SRF=FB_Varcut$SRF,SRR=FB_Varcut$SRR,SAF=FB_Varcut$SAF,SAR=FB_Varcut$SAR )

        gr = peaks.GR
        gr2 = FB_Var.GR
        xo=as.data.frame(GenomicRanges::findOverlaps(gr,gr2,type = "any",ignore.strand=F))
          
        qh=as.data.frame(gr[xo$queryHits],row.names = NULL)
          colnames(qh)=paste0(colnames(qh),"_peaks")
        sh=as.data.frame(gr2[xo$subjectHits],row.names = NULL)
        
        
        peaks_Anno_FB=cbind(qh,sh[,c("seqnames","start","end",'width','IDvar','REF','ALT','Variant','DP','AO','AOcount','freq','type')])

        peaks_Anno_FB=merge(peaks2,peaks_Anno_FB[,c("ID_peaks","seqnames","start","end",'width','IDvar','REF','ALT','Variant','DP','AO','AOcount','freq','type')],by.x="ID",by.y="ID_peaks",all.x=T)
        
                        
######### only select variants that have # reads are >25% of total reads for peak
####  DP does not match IGV so use pileup for reads that map to mutation
        
peaks_Anno_FB$IDvar2=paste0(peaks_Anno_FB$IDvar,"_",peaks_Anno_FB$V5)
peaks_Anno_FB=peaks_Anno_FB[duplicated(peaks_Anno_FB[,colnames(peaks_Anno_FB)%in%c('pu_ID',"count")==F,])==F,]

        peaks_Anno_FB=merge(peaks_Anno_FB,pu[,c('pu_ID',"count")],by.x="IDvar2",by.y="pu_ID",all.x=T)
        peaks_Anno_FB=peaks_Anno_FB[duplicated(peaks_Anno_FB[,colnames(peaks_Anno_FB)%in%c('pu_ID',"count")==F,])==F,]
                
                major=(peaks_Anno_FB$count/peaks_Anno_FB$V4)
                peaks_Anno_FB[which((major>.25|is.na(major))==F),colnames(peaks_Anno_FB)%in%colnames(peaks2)==F]=NA

                
#### Select variants that are not in mpile
peaks_Anno_FB=peaks_Anno_FB[peaks_Anno_FB$IDvar%in%intersect(peaks_Anno_FB$IDvar,peaks_Anno_mpile$IDvar)==F,]

```

```{r include=F,eval=T ,echo=F,warning=F,eval=F}

################
#### Combine Variant information
################

Peaksdata=as.data.frame(matrix(nrow=nrow(peaks2),ncol =25))
colnames(Peaksdata)=c("ID","strand",'Counts_All','Counts_Unique','Counts_MM','Counts_primary','Counts_fracMM','Counts_Primary_fracMM','perc_mm','counts_Cntrl',"Perc_Control",'nVar','nVar_mp','IDvar_mp',"type_mp",'freq_mp','REF_mp','ALT_mp','Variant_mp','IDvar',"Var_type",'freq','REF','ALT','Variant')
for(x in 1:nrow(peaks2)){
  p=peaks2[x,'ID']
  # pd=as.numeric(peaks2[x,'V4'])
  
  pam=peaks_Anno_mpile[peaks_Anno_mpile$ID%in%p,]
  Peaksdata[x,'ID']=p
  Peaksdata[x,'strand']=peaks2[x,'V6']
  Peaksdata[x,'Counts_All']=peaks2[x,'Counts_All']
  Peaksdata[x,'Counts_Unique']=peaks2[x,'Counts_Unique']
  Peaksdata[x,'Counts_MM']=peaks2[x,'Counts_MM']
  Peaksdata[x,'Counts_primary']=peaks2[x,'Counts_primary']
  Peaksdata[x,'Counts_fracMM']=peaks2[x,'Counts_fracMM']
  Peaksdata[x,'Counts_Primary_fracMM']=peaks2[x,'Counts_Primary_fracMM']
  Peaksdata[x,'perc_mm']=peaks2[x,'perc_mm']
  Peaksdata[x,'counts_Cntrl']=peaks2[x,'counts_Cntrl']
  Peaksdata[x,'Perc_Control']=peaks2[x,'Perc_Control']
  Peaksdata[x,'nVar_mp']=sum(is.na((pam$seqnames))==F)
  Peaksdata[x,'IDvar_mp']=paste(unlist(pam$IDvar), collapse =" | ")
    Peaksdata[x,'type_mp']=paste(unlist(pam$TYPE), collapse =" | ")
  # if(sum(is.na((pam$seqnames))==F)>0){Peaksdata[x,'type_mp']='snp'}
  Peaksdata[x,'freq_mp']=paste(unlist(pam$freq), collapse =" | ")
  Peaksdata[x,'REF_mp']=paste(unlist(pam$REF), collapse =" | ")
    Peaksdata[x,'ALT_mp']=paste(unlist(pam$ALT), collapse =" | ")
        Peaksdata[x,'Variant_mp']=paste(unlist(pam$Variant), collapse =" | ")

  pa=peaks_Anno_FB[peaks_Anno_FB$ID%in%p,]
  pa=pa[grepl("snp",pa$Var_type)==F,]
  
  Peaksdata[x,'nVar']=sum(is.na((pa$seqnames))==F)
  Peaksdata[x,'Var_type']=paste(unlist(pa$Var_type), collapse =" | ")
  Peaksdata[x,'IDvar']=paste(unlist(pa$IDvar), collapse =" | ")
  Peaksdata[x,'freq']=paste(unlist(pa$freq), collapse =" | ")
  Peaksdata[x,'REF']=paste(unlist(pa$REF), collapse =" | ")
    # Peaksdata[x,'ALT']=paste(unlist(pa$ALT), collapse =" | ")
      Peaksdata[x,'ALT']=NA

     Peaksdata[x,'Variant']=paste(unlist(pa$Variant), collapse =" | ")


}
Peaksdata[Peaksdata=="NA"]=NA


Peaksdata$Variant_mp=gsub("NA-NA","NA",Peaksdata$Variant_mp)


Peaksdata2=as.data.frame(matrix(nrow=nrow(Peaksdata),ncol =12))
colnames(Peaksdata2)=c("ID","strand",'Counts_All','Counts_Unique','Counts_MM','Counts_primary','Counts_fracMM','Counts_Primary_fracMM','perc_mm',"counts_Cntrl","Perc_Control",'Variant_Count',"Var_type",'Variant Location','Variant Mutation')

Peaksdata2[,c("ID","strand",'Counts_All','Counts_Unique','Counts_MM','Counts_primary','Counts_fracMM','Counts_Primary_fracMM','perc_mm',"counts_Cntrl","Perc_Control")]=Peaksdata[,c("ID","strand",'Counts_All','Counts_Unique','Counts_MM','Counts_primary','Counts_fracMM','Counts_Primary_fracMM','perc_mm',"counts_Cntrl","Perc_Control")]
Peaksdata2[,c('Variant_Count')]=rowSums(Peaksdata[,c('nVar','nVar_mp')])

t=Peaksdata$nVar_mp>0&Peaksdata$nVar>0
Peaksdata2[t,c('Var_type')]=paste0(Peaksdata[t,'type_mp'],' || ',Peaksdata[t,'Var_type'])
Peaksdata2[t,c('Variant Location')]=paste0(Peaksdata[t,'IDvar_mp'],' || ',Peaksdata[t,'IDvar'])
Peaksdata2[t,c('Variant Mutation')]=paste0(Peaksdata[t,'Variant_mp'],' || ',Peaksdata[t,'Variant'])

t=Peaksdata$nVar_mp>0&Peaksdata$nVar==0
Peaksdata2[t,c('Var_type')]=Peaksdata[t,'type_mp']
Peaksdata2[t,c('Variant Location')]=Peaksdata[t,'IDvar_mp']
Peaksdata2[t,c('Variant Mutation')]=Peaksdata[t,'Variant_mp']

t=Peaksdata$nVar_mp==0&Peaksdata$nVar>0
Peaksdata2[t,c('Var_type')]=Peaksdata[t,'Var_type']
Peaksdata2[t,c('Variant Location')]=Peaksdata[t,'IDvar']
Peaksdata2[t,c('Variant Mutation')]=Peaksdata[t,'Variant']

# write.table(Peaksdata2,file=("Clip_Peaks.txt"), sep = "\t", row.names = FALSE, col.names = T, append = F, quote= FALSE)

# chr6:47781646
# chrY:90741488
# chr17:39845130-39845292
# Peaksdata=as.data.frame(Peaksdata)
# Peaksdata=Peaksdata[grepl(pattern = "chrY",x=Peaksdata[,"ID"])==F,]
# Peaksdata=Peaksdata[grepl(pattern = "chrX",x=Peaksdata[,"ID"])==F,]

```


<br>  

```{r fig.align='center',fig.height=6, fig.width=9,echo=F,warning=F,include=F,eval=F}
p=Peaksdata2


p[is.na(p$Var_type),'Var_type']="No Mutation"
p[grepl("_",p$Var_type),'Var_type']="Multiple"
p[grepl("\\|",p$Var_type),'Var_type']="Multiple"

p$Var_type=as.factor(p$Var_type)
p$Var_type=factor(p$Var_type,levels=c('snp','ins/del','Multiple','No Mutation'))

p=p[p$Variant_Count>0,]
gg=ggplot(p,aes(fill=Var_type,x=.5) )+geom_bar(position="stack")+theme_classic()+xlim(0,2)+ggtitle("Type of Mutation")+
theme(plot.title = element_text(hjust = 0.5,size = 25),axis.title=element_text(size=20),axis.text=element_text(size=15),legend.text = element_text( size=30))+
  annotate('text',label=paste0(sum(Peaksdata2$Variant_Count==0),' Peaks'),x=1.5,y=500,size=10)+
annotate('text',label=paste0('have no mutation'),x=1.5,y=430,size=10)
gg


# pp=p;pp$Variant_Count=as.character(pp$Variant_Count)
pp=Peaksdata2
pp=pp[pp$Variant_Count>0,]
pp$Variant_Count=as.character(pp$Variant_Count)
gg=ggplot(pp,aes(fill=Variant_Count,x=.5) )+geom_bar(position="stack")+theme_classic()+xlim(0,2)+ggtitle("Number of mutations per CLIP peak")+
theme(plot.title = element_text(hjust = 0.5,size = 25),axis.title=element_text(size=20),axis.text=element_text(size=15),legend.text = element_text( size=30))+
  annotate('text',label=paste0(sum(Peaksdata2$Variant_Count==0),' Peaks'),x=1.5,y=500,size=10)+
annotate('text',label=paste0('have no mutation'),x=1.5,y=430,size=10)
gg

# <br>  <br>  
# We can look at the distribution of read counts for each Mutation type to see if there is a bias for lower expressed reads
# <br>  <br>  
```


```{r fig.align='center',fig.height=6, fig.width=9,echo=F,warning=F,include=F,eval=F}
p=Peaksdata2

p[is.na(p$Var_type),'Var_type']="No Mutation"
p[grepl("_",p$Var_type),'Var_type']="Multiple"
p[grepl("\\|",p$Var_type),'Var_type']="Multiple"

p$Var_type=as.factor(p$Var_type)
p$Var_type=factor(p$Var_type,levels=c('snp','ins/del','Multiple','No Mutation'))

gg=ggplot(p,aes(x=Var_type,y=Counts_All,fill=Var_type) )+geom_violin()+theme_classic()+geom_jitter(height = 0, width = 0.1,size=1)+
  scale_y_continuous(trans = "log2")+
  ggtitle('Peak Counts for Ro-Clip\nby Type of Mutation')+ylab("Counts - Ro-Clip")+
theme(plot.title = element_text(hjust = 0.5,size = 25),axis.title=element_text(size=20),axis.text=element_text(size=15),legend.text = element_text( size=30))

gg

# 
# gg=ggplot(p,aes(x=Var_type,y=counts_Cntrl,fill=Var_type) )+geom_violin()+theme_classic()+geom_jitter(height = 0, width = 0.1,size=1)+
#   scale_y_continuous(trans = "log2")+
#   ggtitle('Peaks with high background in\nControl sample')+ylab("Counts - Control")+
# theme(plot.title = element_text(hjust = 0.5,size = 25),axis.title=element_text(size=20),axis.text=element_text(size=15),legend.text = element_text( size=30))
# 
# gg

pp=p
# pp[pp$Perc_Control>50,"Perc_Control"]=50
gg=ggplot(pp,aes(x=Var_type,y=Perc_Control,fill=Var_type) )+geom_violin()+theme_classic()+geom_jitter(height = 0, width = 0.1,size=1)+
    scale_y_continuous(trans = "log2")+geom_hline(yintercept = 10,color ='red')+
  ggtitle('Peaks with high background in\nControl sample')+ylab("Ratio of Control counts / Ro counts\n(%- Log2 scale)")+
theme(plot.title = element_text(hjust = 0.5,size = 25),axis.title=element_text(size=20),axis.text=element_text(size=15),legend.text = element_text( size=30))+
    annotate('text',label=paste0('10%'),x=.7,y=12,size=5)


gg

```


```{r fig.align='center',fig.height=6, fig.width=12,echo=F,warning=F,include=F,eval=F}

# <br>  
# 
# If Variant is SNP A-G/T-C are most abundant mutations  
# 
# <br>  

# p=Peaksdata2[Peaksdata2$strand=="+",]
# p=Peaksdata2[Peaksdata2$strand=="-",]
p=Peaksdata2
pp=p[grep(pattern = "snp",x = p$Var_type),]
pp=unlist(stri_split_fixed(pp$`Variant Mutation`,"|"))
pp=gsub(" ","",pp);pp=pp=gsub("_NA","",pp);pp=pp=gsub("NA","",pp)
pp=as.data.frame(pp[pp!=""])
pp=count(pp)
colnames(pp)[1]="pp"
ggplot(pp )+geom_bar(aes(x=pp,y=freq),stat="identity")+theme_classic()+
  ggtitle("SNP Mutations")+xlab("")+ylab("Count")+
theme(plot.title = element_text(hjust = 0.5,size = 25),axis.title=element_text(size=20),axis.text=element_text(size=15),legend.text = element_text( size=30))

```


```{r Annotations, fig.asp=.5, fig.align='center',fig.height=6, fig.width=12,echo=F,warning=F,echo=F}

#### if no Variant identification....

colnames(peaks2)[which(colnames(peaks2)%in%c("V1","V2","V3","V6"))]=c('chr','start','end','strand')
Peaksdata2=peaks2

#### uncoment for variant identification....  
  
# Peaksdata2=separate(Peaksdata2,col=ID,into = c('chr','start'),sep = ":",remove = F)
# Peaksdata2=separate(Peaksdata2,col=start,into = c('start','end'),sep = "-",remove = T)\
#################################################################
  
  
  
Peaksdata2_anno=CLIPannotation(Peaksdata2[,c('chr','start','end','strand')],T)
Peaksdata2_annox=Peaksdata2_anno
# Peaksdata2_anno=Peaksdata2_annox

Peaksdata2_anno=merge(Peaksdata2,Peaksdata2_anno[,!colnames(Peaksdata2_anno)%in%c('chr','start','end','strand')],by='ID')
```



## Identify CLIP peak Gene Location  

### Identify CLIP peak Host gene   

Using GTF file from GENCODE v23  
Peaks were annotated with overlapping gene  


### Identify CLIP peak In RNA specific Genes  

```{r echo=F,results='asis'}
Classification_ncRNA=read.delim(paste0("./annotation/ncRNA_Annotations.txt"), header=T, sep="\t",stringsAsFactors = F)

Classification_ncRNA['rRNA_DNA','contents']=gsub(c(', TermrRNA, repeat_RI-1, repeat_RII-1, APS1_X52413_1, APS1_X52413_2, Block_1, repeat_RI-2, repeat_RII-2, APS1_X52413_3, APS2_X52412_1, Block_2, repeat_RI-3, repeat_RII-3, APS1_X52413_4, repeat_RII-4, APS2_52412_2, Sal_box_1, Sal_box_2, Sal_box_3, Sal_box_4, Sal_box_5, Sal_box_6, Sal_box_7, Sal_box_8, Sal_box_9, spacer_promoter_X06186, enhancer_V00847, Sal_box_0, promoter_M18062, UPE, core_rRNA_promoter'),"",Classification_ncRNA['rRNA_DNA','contents'])

kable(Classification_ncRNA[,1:3])

```

<br>  

### Identify if CLIP peak overlaps with Intron or Exonic region   

<br>  

Peaks were annotated by whether they overlap with Host gene intron/exon region
Intron coordinates were calculated from GTF file.

<!-- A second column was added to idenify if the peak also overlapped with the 5'UTR 3'UTR or CDS (Column: Featrue 2) -->


### Identify peaks in repeat regions   

   Annotate all repeat regions/Classes identified in Repeatmasker Annotation file (UCSC Table browser)  
   Data was not filtered based on any of the identified Repeats.  
    1) LINE/SINE   
    2) LTR   
    3) DNA   
    4) Satellites   
    5) Simple Repeats   
    6) Low Complexity   
    7) Other   
    8) Unknown   
    <br>  




```{r echo=F,warning=F, include=F,eval=T}

p=as.data.frame(((colSums(is.na(Peaksdata2_anno[,grep('Repeat_',colnames(Peaksdata2_anno))])==F)/nrow(Peaksdata2_anno))*100))
p$Repeat_Type=rownames(p);colnames(p)=c('Frequency','Repeat_Type')
p$Repeat_Type=gsub("Repeat_","",p$Repeat_Type)

ggplot(p )+geom_bar(aes(x=Repeat_Type,y=Frequency),stat="identity")+theme_classic()+#ggtitle("SNP Mutations")+
  xlab("")+ylab("Frequency of peaks in \nrepeat regions (%)")+
theme(plot.title = element_text(hjust = 0.5,size = 25),axis.title=element_text(size=20),axis.text=element_text(size=15),legend.text = element_text( size=30))+theme(axis.text.x = element_text(angle = -45))


```

```{r echo=F,warning=F, include=T,eval=T}
p=as.data.frame(((colSums(is.na(Peaksdata2_anno[,grep('Same_Repeat_',colnames(Peaksdata2_anno))])==F)/nrow(Peaksdata2_anno))*100))
p$Repeat_Type=rownames(p);colnames(p)=c('Frequency','Repeat_Type')
p$Repeat_Type=gsub("Same_Repeat_","",p$Repeat_Type)

ggplot(p )+geom_bar(aes(x=Repeat_Type,y=Frequency),stat="identity")+theme_classic()+#ggtitle("SNP Mutations")+
  xlab("")+ylab("Frequency of peaks in \nrepeat regions (%)")+ggtitle("Same Strand")+
theme(plot.title = element_text(hjust = 0.5,size = 15),axis.title=element_text(size=15),axis.text=element_text(size=10),legend.text = element_text( size=10))+theme(axis.text.x = element_text(angle = -45))
```
   
```{r echo=F,warning=F, include=F,eval=T}

p=as.data.frame(((colSums(is.na(Peaksdata2_anno[,grep('Oppo_Repeat_',colnames(Peaksdata2_anno))])==F)/nrow(Peaksdata2_anno))*100))
p$Repeat_Type=rownames(p);colnames(p)=c('Frequency','Repeat_Type')
p$Repeat_Type=gsub("Oppo_Repeat_","",p$Repeat_Type)

ggplot(p )+geom_bar(aes(x=Repeat_Type,y=Frequency),stat="identity")+theme_classic()+#ggtitle("SNP Mutations")+
  xlab("")+ylab("Frequency of peaks in \nrepeat regions (%)")+ggtitle("Opposite Strand")+
theme(plot.title = element_text(hjust = 0.5,size = 15),axis.title=element_text(size=15),axis.text=element_text(size=10),legend.text = element_text( size=10))+theme(axis.text.x = element_text(angle = -45))

```



```{r fig.asp=.5, fig.align='center',fig.height=6, fig.width=12,echo=F,warning=F,eval=T,include=F}

## Calculate average MAPQ for each peak   
calcMAPQ=F

library(foreach)
library(GenomicAlignments)
library(parallel)
library(doParallel)
# install.packages('doSNOW')
library(doSNOW)

param <- ScanBamParam(what=c("mapq", "flag"))
Bam=readGAlignments('./NovoAlign_umi/All/recomb_NHtag/bam/Ro_Clip_iCountcutadpt_all.unique.NH.mm.ddup.s.bam',use.names=T,index='./NovoAlign_umi/All/recomb_NHtag/bam/Ro_Clip_iCountcutadpt_all.unique.NH.mm.ddup.s.bam.bai', param=param)
# colnames(as.data.frame(Bam))
mcols(Bam)$qname=names(Bam)


# p=Peaksdata2_anno[,c('ID',colnames(Peaksdata2_anno)[grep('Repeat_',colnames(Peaksdata2_anno))])]
# p=Peaksdata2_anno
# p2=p[,colnames(p)[grep('Repeat_',colnames(p))]]
# p2=rowSums((is.na(p2)==F))
# 
# Peaksdata2_RP=Peaksdata2_anno[p2>0,]

Peaksdata2_RP=Peaksdata2_anno
       
# Peaksdata2_MM=Peaksdata2_anno[Peaksdata2_anno$Counts_MM>1,]
# Peaksdata2_MM=Peaksdata2_MM[Peaksdata2_MM$Same_Comb_biotype_exon%in%'ncRNA',]



#############################################
######### for loop

# out=as.data.frame(matrix(nrow=nrow(Peaksdata2_RP),ncol=2));
# colnames(out)=c("ID","Avg_mapq")
# 
#  system.time(
# for (x in 1:nrow(Peaksdata2_RP)) {
# # for (x in 1:50) {
# 
#   PL=Peaksdata2_RP[x,]
# 
#   PL.GR=GRanges(seqnames = as.character(PL$chr), ranges=IRanges(start = as.numeric(PL$start), end = as.numeric(PL$end)),strand = PL$strand,
#                    ID=PL$ID)
#         q = PL.GR
#         s = Bam
# 
#         xo=as.data.frame(GenomicRanges::findOverlaps(q,s,type = "any",ignore.strand=T))
#         qh=as.data.frame(q[xo$queryHits],row.names = NULL)
#           colnames(qh)=paste0(colnames(qh),"_peaks")
#         sh=as.data.frame(s[xo$subjectHits],row.names = NULL)
#           colnames(sh)=paste0(colnames(sh),"_bam")
#         bam_PL=cbind(qh,sh)
# 
#        out[x,'ID']=PL$ID
#        out[x,'Avg_mapq']=mean(bam_PL$mapq_bam)
# 
#        # remove(PL,q,s,xo,qh,sh,bam_PL,PL.GR)
# })
# 
# out=out[!is.na(out$ID),]

##############################################
#### setting up parallization ( cant get it to go faster) 
# mapq=function(i){  
#   PL=i
#   PL.GR=GRanges(seqnames = as.character(PL$chr), ranges=IRanges(start = as.numeric(PL$start), end = as.numeric(PL$end)),strand = PL$strand,
#                    ID=PL$ID)
#         q = PL.GR
#         s = Bam
# 
#         xo=as.data.frame(GenomicRanges::findOverlaps(q,s,type = "any",ignore.strand=T))
#         qh=as.data.frame(q[xo$queryHits],row.names = NULL)
#           colnames(qh)=paste0(colnames(qh),"_peaks")
#         sh=as.data.frame(s[xo$subjectHits],row.names = NULL)
#           colnames(sh)=paste0(colnames(sh),"_bam")
#         bam_PL=cbind(qh,sh)
#         
#        out1[,'ID']=PL$ID
#        out1[,'Avg_mapq']=mean(bam_PL$mapq_bam)
#        
#        # remove(PL,q,s,xo,qh,sh,bam_PL,PL.GR)
#       # return(out1)
#        }
# 
# system.time( {
#   cl=makeCluster(detectCores()[1]-4)
# registerDoParallel(cl)
#   
#    out1=as.data.frame(matrix(nrow=1,ncol=2));
# colnames(out1)=c("ID","Avg_mapq")
# 
# o=foreach(x =1:100,.packages = 'GenomicRanges',.combine = rbind) %do% {
#   # if (x%in%seq(from = 0, to = 10000, by = 50)) {print(x/nrow(Peaksdata2_RP))       }
#   mapq(Peaksdata2_RP[x,])
# 
# }
# stopCluster(cl)
# })

##############################################
#### setting up mclapply ~3x-4x faster

if (calcMAPQ==T) {
  
glist=function(i){
  PL=i
  GRanges(seqnames = as.character(PL[1]), ranges=IRanges(start = as.numeric(PL[2]), end = as.numeric(PL[3])),strand = PL[4])
  }
# remove(gout)
gout=apply(Peaksdata2_RP[,c('chr','start','end','strand','ID')],1,glist)
# class(gout$`1`)

mapq=function(i){  
  # i=gout$`1`
     # out1=as.data.frame(matrix(nrow=1,ncol=2));
          q=i
          s = Bam
  # 
        xo=as.data.frame(GenomicRanges::findOverlaps(q,s,type = "any",ignore.strand=T))
        # qh=as.data.frame(q[xo$queryHits],row.names = NULL)
          # colnames(qh)=paste0(colnames(qh),"_peaks")
        sh=as.data.frame(s[xo$subjectHits],row.names = NULL)
          # colnames(sh)=paste0(colnames(sh),"_bam")
        # bam_PL=cbind(qh,sh)

       # out1[,'ID']=pl[5]
       # out1[,'Avg_mapq']=mean(bam_PL$mapq_bam)
       mean(sh$mapq)
       # remove(PL,q,s,xo,qh,sh,bam_PL,PL.GR)
      # return(out1)
       }
 # system.time( 

 ### normal lapply is same as for loop
system.time({outl=mclapply(gout,mapq,mc.cores=8)}  )

Peaksdata2_RP$Avg_mapq=unlist(outl )
Peaksdata2_anno=Peaksdata2_RP

write.table(Peaksdata2_RP[,c('ID','Avg_mapq')],file=("CLIP_Peaks_mapq.txt"), sep = "\t", row.names = FALSE, col.names = T, append = F, quote= FALSE,na = "")
# out$parallel=unlist(outl )
# out$comp=out$Avg_mapq==out$parallel
}

if (calcMAPQ==F) {
  CLIP_Peaks_mapq=fread(paste0("CLIP_Peaks_mapq.txt"), header=T, sep="\t",stringsAsFactors = F,data.table=F)
  
 Peaksdata2_anno= merge(Peaksdata2_anno,CLIP_Peaks_mapq,by="ID",all.x = T)
}
```




### Asigning Clip peak attributes   

Not all Peaks overlap with a single feature so peak assignments were assigned by priority:  

**ncRNA > Protein coding : Exonic > repeats > Pseudogene > Antisense Feature > Protein Coding : Intronic > lncRNA > no Feature**   

rDNA transcript (BK000964.3) was included in reference genome but not included in peak counts. rDNA will affect multimapping count and will show up in alternative mapping location.   	

All annotations from RNA type, Repeat regions, and Intronic/exonic regions are annotated in the Table.   

<br>  


## Overview of Ro targets by CLIP peak catagory   


### Peak Attributes   

Peak counts do not include rRNA peaks   


```{r fig.asp=.5, fig.align='center',fig.height=4, fig.width=8,echo=F,warning=F,eval=F,include=F}
p=Peaksdata2_anno
# p$Same_Comb_biotype_exon=p$Comb_biotype_exon_Oppo


p[is.na(p$Comb_biotype_exon),'Same_Comb_biotype_exon']='no Feature'
colnames(p)[colnames(p)%in%'Same_Comb_biotype_exon']='RNA_Type'
p$RNA=''
gg=ggplot(p,aes(fill=RNA_Type,x=RNA) )+geom_bar(position="stack",width=.5)+theme_classic()+ggtitle("Number of peaks by catagory\n(Determined by unique reads only)")+
# theme(plot.title = element_text(hjust = 0.5,size = 25),axis.title=element_text(size=20),axis.text=element_text(size=15),legend.text = element_text( size=30))+
 scale_fill_brewer(palette = "Dark2") 
ggplotly(gg)


# View(Peaksdata2_anno[,c('ID','Same_Comb_biotype_exon','Same_Comb_biotype_ncRNA','Oppo_Comb_biotype_exon','Oppo_Comb_biotype_ncRNA','Comb_biotype_exon_Oppo')])

# ncRNA > exonic > repeats > pseudogene > antisense > intronic > lncRNA > noFeature 


```


 
```{r fig.asp=.5, fig.align='center',fig.height=4, fig.width=8,echo=F,warning=F,eval=T,include=T}
p=Peaksdata2_anno
p$Same_Comb_biotype_exon=p$Comb_biotype_exon_Oppo

  p=p[!p$Same_Comb_biotype_ncRNA%in%'rRNA',]


p[is.na(p$Comb_biotype_exon),'Same_Comb_biotype_exon']='no Feature'
colnames(p)[colnames(p)%in%'Same_Comb_biotype_exon']='RNA_Type'
p$RNA=''
gg=ggplot(p,aes(fill=RNA_Type,x=RNA) )+geom_bar(position="stack",width=.5)+theme_classic()+ggtitle("Number of peaks by catagory\n(Determined by unique reads only)")+
# theme(plot.title = element_text(hjust = 0.5,size = 25),axis.title=element_text(size=20),axis.text=element_text(size=15),legend.text = element_text( size=30))+
 scale_fill_brewer(palette = "Dark2") 
ggplotly(gg)


# View(Peaksdata2_anno[,c('ID','Same_Comb_biotype_exon','Same_Comb_biotype_ncRNA','Oppo_Comb_biotype_exon','Oppo_Comb_biotype_ncRNA','Comb_biotype_exon_Oppo')])

# ncRNA > exonic > repeats > pseudogene > antisense > intronic > lncRNA > noFeature 


```



<!-- <font size="1">Columns: RNA_type_Classification </font>      -->
<!-- <br/>      -->

```{r fig.asp=.5, fig.align='center',fig.height=4, fig.width=8,echo=F,warning=F,eval=T,include=T}

# p=Peaksdata2_anno
# p=p[p$Same_Comb_biotype_exon%in%'ncRNA',]
# p=paste0(p$type_comb,collapse = ",")
# p=as.data.frame(strsplit(p,','));colnames(p)='RNA_Type'
# p=p[!p$RNA_Type%in%c('pseudogene','protein_coding'),,drop=F]
# p$RNA=''
# gg=ggplot(p,aes(fill=RNA_Type,x=RNA) )+geom_bar(position="stack",width=.5)+theme_classic()+ggtitle("ncRNA only: Counted by catagory\n(Peak Can be counted multiple times)")+
# theme(plot.title = element_text(hjust = 0.5,size = 25),axis.title=element_text(size=20),axis.text=element_text(size=15),legend.text = element_text( size=30))+
 # scale_color_brewer(palette = "Paired")+scale_fill_brewer(palette = "Paired") 
# ggplotly(gg)

p=Peaksdata2_anno
p$Same_Comb_biotype_exon=p$Comb_biotype_exon_Oppo
# if (incd_rRNA==F) {
  p=p[!p$Same_Comb_biotype_ncRNA%in%'rRNA',]
# }

p=p[p$Same_Comb_biotype_exon%in%'ncRNA','Same_Comb_biotype_ncRNA',drop=F]
p$RNA=""
gg=ggplot(p,aes(fill=Same_Comb_biotype_ncRNA,x=RNA) )+geom_bar(width=.5)+theme_classic()+ggtitle("Number of peaks by catagory : ncRNA only\n(Determined by unique reads only)")+
 scale_color_brewer(palette = "Paired")+scale_fill_brewer(palette = "Paired") +  blank_theme+ 
  coord_polar("y",start=0)
(gg)

 
# colnames(Peaksdata3)
```       
<br>  
```{r fig.asp=.5, fig.align='center',fig.height=4, fig.width=8,echo=F,warning=F,eval=F,include=F}
p=Peaksdata2_anno
p$Same_Comb_biotype_exon=p$Comb_biotype_exon_Oppo
if (incd_rRNA==F) {
  p=p[!p$Same_Comb_biotype_ncRNA%in%'rRNA',]
}

p=p[p$Same_Comb_biotype_exon%in%'ncRNA',]

u=unique(p$Same_Comb_biotype_ncRNA)
pcount=data.frame(u,1:length(u));colnames(pcount)=c('RNA_Subtype','Counts_All');pcount$Counts_Unique=NA;pcount$Counts_MM=NA;pcount$Peak_count=NA
for (x in 1:length(u)) {
  pam=p[p$Same_Comb_biotype_ncRNA%in%u[x],]
  pcount[x,1]=u[x]
  pcount[x,'Counts_All']=colSums(pam[,'Counts_All',drop=F],na.rm = T)
  pcount[x,'Counts_Unique']=colSums(pam[,'Counts_Unique',drop=F],na.rm = T)
  pcount[x,'Counts_MM']=colSums(pam[,'Counts_MM',drop=F],na.rm = T)
  pcount[x,'Counts_fracMM']=colSums(pam[,'Counts_fracMM',drop=F],na.rm = T)
  pcount[x,'Peak_count']=nrow(pam)
}
pcount$RNA=''

gg=ggplot(pcount,aes(fill=RNA_Subtype,x=RNA,y=Counts_Unique) )+geom_bar(stat='identity',width=.5)+theme_classic()+ggtitle("ncRNA only: # of reads (unique) by catagory")+
 scale_color_brewer(palette = "Paired")+scale_fill_brewer(palette = "Paired") +  blank_theme+ 
  coord_polar("y",start=0)
(gg)


gg=ggplot(pcount,aes(fill=RNA_Subtype,x=RNA,y=Counts_fracMM) )+geom_bar(stat='identity',width=.5)+theme_classic()+ggtitle("ncRNA only: # of reads (unique + Scaled MM) by catagory")+
 scale_color_brewer(palette = "Paired")+scale_fill_brewer(palette = "Paired") +  blank_theme+ 
  coord_polar("y",start=0)
(gg)

gg=ggplot(pcount,aes(fill=RNA_Subtype,x=RNA,y=Counts_All) )+geom_bar(stat='identity',width=.5)+theme_classic()+ggtitle("ncRNA only: # of reads (unique+MM) by catagory")+
 scale_color_brewer(palette = "Paired")+scale_fill_brewer(palette = "Paired") +  blank_theme+ 
  coord_polar("y",start=0)
(gg)


# colnames(Peaksdata3)
```      

<br>  
<br>   
<!-- ## Overview of Ro targets by CLIP peak attributs (no rRNA peaks) -->

```{r fig.asp=.5, fig.align='center',fig.height=4, fig.width=8,echo=F,warning=F,eval=T,include=F}
p=Peaksdata2_anno
p$Same_Comb_biotype_exon=p$Comb_biotype_exon_Oppo

if (incd_rRNA==F) {
  p=p[!p$Same_Comb_biotype_ncRNA%in%'rRNA',]
}

p[is.na(p$Same_Comb_biotype_exon),'Same_Comb_biotype_exon']='no Feature'
colnames(p)[colnames(p)%in%'Same_Comb_biotype_exon']='RNA_Type'
p$RNA=''
gg=ggplot(p,aes(fill=RNA_Type,x=RNA) )+geom_bar(position="stack",width=.5)+theme_classic()+ggtitle("BioType - Counted by catagory\n(Each Clip Peak only Counted Once)")+
# theme(plot.title = element_text(hjust = 0.5,size = 25),axis.title=element_text(size=20),axis.text=element_text(size=15),legend.text = element_text( size=30))+
 scale_fill_brewer(palette = "Dark2") 
ggplotly(gg)
```

<!-- <font size="1">Columns: RNA_type_Classification </font>      -->
<!-- <br/>      -->

```{r fig.asp=.5, fig.align='center',fig.height=4, fig.width=8,echo=F,warning=F,eval=T,include=F}

# p=Peaksdata2_anno
# # p=p[(p$Same_Comb_biotype_ncRNA%in%'rRNA')==F,]
# 
# p=p[p$Comb_biotype_exon%in%'ncRNA',]
# p=paste0(p$type_comb,collapse = ",")
# p=as.data.frame(strsplit(p,','));colnames(p)='RNA_Type'
# p=p[!p$RNA_Type%in%c('pseudogene','protein_coding'),,drop=F]
# p$RNA=''
# gg=ggplot(p,aes(fill=RNA_Type,x=RNA) )+geom_bar(position="stack",width=.5)+theme_classic()+ggtitle("ncRNA only: Counted by catagory\n(Peak Can be counted multiple times)")+
# theme(plot.title = element_text(hjust = 0.5,size = 25),axis.title=element_text(size=20),axis.text=element_text(size=15),legend.text = element_text( size=30))+
 # scale_color_brewer(palette = "Paired")+scale_fill_brewer(palette = "Paired") 
# ggplotly(gg)

p=Peaksdata2_anno
p$Same_Comb_biotype_exon=p$Comb_biotype_exon_Oppo
if (incd_rRNA==F) {
  p=p[!p$Same_Comb_biotype_ncRNA%in%'rRNA',]
}

p=p[p$Same_Comb_biotype_exon%in%'ncRNA','Same_Comb_biotype_ncRNA',drop=F]
p$RNA=""
gg=ggplot(p,aes(fill=Same_Comb_biotype_ncRNA,x=RNA) )+geom_bar(width=.5)+theme_classic()+ggtitle("ncRNA only: # of peaks per catagory")+
 scale_color_brewer(palette = "Paired")+scale_fill_brewer(palette = "Paired") +  blank_theme+ 
  coord_polar("y",start=0)
(gg)

 
```    


## Opposite strand CLIP peak attributs  


```{r , fig.align='center',fig.height=6, fig.width=12,echo=F,warning=F}

p=Peaksdata2_anno
p=p[p$Comb_biotype_exon_Oppo%in%'Antisense Feature',]
if (incd_rRNA==F) {
  p=p[!p$Same_Comb_biotype_ncRNA%in%'rRNA',]
}

colnames(p)[colnames(p)%in%'Oppo_Comb_biotype_exon']='RNA_Type'
p$RNA=''

ggplot(p,aes(fill=RNA_Type,x=RNA_Type) )+geom_bar(stat="count")+theme_classic()+
  ggtitle("Opposite Strand Feature - Number of peaks by catagory\n(Determined by unique reads only")+xlab("")+
theme(plot.title = element_text(hjust = 0.5),axis.title=element_text(size=15),axis.text=element_text(size=10),legend.position = "none",axis.text.x = element_text(angle = -45,vjust = .5))+
 scale_fill_brewer(palette = "Dark2") 
# ggplotly(ggg)


# gg=ggplot(p,aes(fill=RNA_Type,x=RNA) )+geom_bar(position="stack",width=.5)+theme_classic()+ggtitle("Opposite Strand Feature - Number of peaks by catagory\n(Determined by unique reads only")+
# # theme(plot.title = element_text(hjust = 0.5,size = 25),axis.title=element_text(size=20),axis.text=element_text(size=15),legend.text = element_text( size=30))+
#  scale_fill_brewer(palette = "Dark2") 
# ggplotly(gg)
```

```{r fig.asp=.5, fig.align='center',fig.height=6, fig.width=12,echo=F,warning=F,eval=T,incude=T}

p=Peaksdata2_anno
if (incd_rRNA==F) {
  p=p[!p$Same_Comb_biotype_ncRNA%in%'rRNA',]
}

p$transition=paste0(p$Same_Comb_biotype_exon,'->\n',p$Oppo_Comb_biotype_exon)
p=p[p$Comb_biotype_exon_Oppo%in%'Antisense Feature',]
colnames(p)[colnames(p)%in%'transition']='RNA_Type'
p$Difference=''

ggplot(p,aes(fill=Same_Comb_biotype_exon,x=factor(Oppo_Comb_biotype_exon)) )+geom_bar(stat="count")+theme_classic()+
  ggtitle("Opposite Strand Feature Features: \n Compare Antisense and Snese stand Annotation")+xlab("Antisense strand Annotation")+ylab("Peak Count")+labs(fill = "Sense Stand Annotation")+
theme(plot.title = element_text(hjust = 0.5,size=20),axis.title=element_text(size=15),axis.text=element_text(size=10),legend.position = "right",axis.text.x = element_text(angle = -45,vjust = .5))

 

# # chr13:49448962-49449025
# View(Peaksdata2_anno[Peaksdata2_anno$ID%in%'chr13:49448962-49449025',c('ID','ID2.x','Same_external_gene_name','Oppo_external_gene_name','Same_Comb_biotype_exon','Oppo_type_comb','Oppo_feature','Oppo_Comb_biotype_exon','Comb_biotype_exon_Oppo')])
# View(Peaksdata2_anno[,c('ID','ID2.x','Same_external_gene_name','Oppo_external_gene_name','Same_Comb_biotype_exon','Oppo_type_comb','Oppo_feature','Oppo_Comb_biotype_exon','Comb_biotype_exon_Oppo')])

# xxx
```

<br>  

```{r echo=F,warning=F, include=F,eval=F}
p=as.data.frame(((colSums(is.na(Peaksdata2_anno[,grep('Same_Repeat_',colnames(Peaksdata2_anno))])==F))))
p$Repeat_Type=rownames(p);colnames(p)=c('Frequency','Repeat_Type')
p$Repeat_Type=gsub("Same_Repeat_","",p$Repeat_Type)

ggplot(p )+geom_bar(aes(x=Repeat_Type,y=Frequency),stat="identity")+theme_classic()+#ggtitle("SNP Mutations")+
  xlab("")+ylab("number of peaks in \nrepeat regions")+ggtitle("Same Strand")+
theme(plot.title = element_text(hjust = 0.5,size = 15),axis.title=element_text(size=15),axis.text=element_text(size=10),legend.text = element_text( size=10))+theme(axis.text.x = element_text(angle = -45))
```
```{r echo=F,warning=F, include=T,eval=T}

p=as.data.frame(((colSums(is.na(Peaksdata2_anno[,grep('Oppo_Repeat_',colnames(Peaksdata2_anno))])==F))))
p$Repeat_Type=rownames(p);colnames(p)=c('Frequency','Repeat_Type')
p$Repeat_Type=gsub("Oppo_Repeat_","",p$Repeat_Type)

ggplot(p )+geom_bar(aes(x=Repeat_Type,y=Frequency),stat="identity")+theme_classic()+#ggtitle("SNP Mutations")+
  xlab("")+ylab("number of peaks in \nrepeat regions")+ggtitle("Opposite Strand")+
theme(plot.title = element_text(hjust = 0.5,size = 15),axis.title=element_text(size=15),axis.text=element_text(size=10),legend.text = element_text( size=10))+theme(axis.text.x = element_text(angle = -45))

```
<br>  

## Calculate average MAPQ for each peak   

We can plot the avarage Mapping Quality (MAPQ) of each peak to identify the confidence of read mapping. We see that repeat regions have generally lower mapping quality.    

<br>  
```{r fig.asp=.5, fig.align='center',fig.height=6, fig.width=12,echo=F,warning=F,eval=T,incude=T}

p=Peaksdata2_anno
# if (incd_rRNA==F) {
#   p=p[!p$Comb_biotype_exon_Oppo%in%'rRNA',]
# }

colnames(p)[colnames(p)%in%'Comb_biotype_exon_Oppo']='RNA_Type'
p$RNA=''
 gg=ggplot(p,aes(x=RNA_Type,y=Avg_mapq,fill=RNA_Type) )+geom_violin()+theme_classic()+geom_jitter(height = 0, width = 0.1,size=1)+
  ggtitle('Average MAPQ for Ro-Clip peak\nby Feature')+ylab("Avg_mapq - Ro-Clip")+xlab("")+
theme(plot.title = element_text(hjust = 0.5,size = 25),axis.title=element_text(size=20),axis.text=element_text(size=15))+theme(legend.position = "none") +
    theme(axis.text.x = element_text(angle = -45))
gg


############################################

p=Peaksdata2_anno
p=p[p$Comb_biotype_exon_Oppo%in%'Repeat Element',]

p=p[,c('ID','Avg_mapq',colnames(p)[grep('Same_Repeat_',colnames(p))])]
p2=p[,colnames(p)[grep('Same_Repeat_',colnames(p))]]
p2[,colnames(p2)[grep('Same_Repeat_',colnames(p2))]]=as.numeric(is.na(p2[,colnames(p2)[grep('Same_Repeat_',colnames(p2))]])==F)

p2[p2[,colnames(p2)[grep('Same_Repeat_',colnames(p2))]]==0]=NA
p2=(p2*p$Avg_mapq)

p2=melt.data.frame(p2,id.vars = NULL)
p2$variable=gsub("Same_Repeat_","",p2$variable)


 gg=ggplot(p2,aes(x=variable,y=value,fill=variable) )+geom_violin()+theme_classic()+geom_jitter( width = 0.1,size=1)+
  ggtitle('Same Strand: Repeat Element\nAverage MAPQ for Ro-Clip peak')+ylab("Avg_mapq - Ro-Clip")+xlab("")+
theme(plot.title = element_text(hjust = 0.5,size = 25),axis.title=element_text(size=20),axis.text=element_text(size=15))+theme(legend.position = "none") +
    theme(axis.text.x = element_text(angle = -45))+scale_fill_brewer(palette = "Paired")

gg
############################################

 
 
p=Peaksdata2_anno
p=p[p$Comb_biotype_exon_Oppo%in%'Repeat Element',]

p=p[,c('ID','Avg_mapq',colnames(p)[grep('Oppo_Repeat_',colnames(p))])]
p2=p[,colnames(p)[grep('Oppo_Repeat_',colnames(p))]]
p2[,colnames(p2)[grep('Oppo_Repeat_',colnames(p2))]]=as.numeric(is.na(p2[,colnames(p2)[grep('Oppo_Repeat_',colnames(p2))]])==F)

p2[p2[,colnames(p2)[grep('Oppo_Repeat_',colnames(p2))]]==0]=NA
p2=(p2*p$Avg_mapq)

p2=melt(p2,id.vars = NULL)
p2$variable=gsub("Oppo_Repeat_","",p2$variable)

gg= ggplot(p2,aes(x=variable,y=value,fill=variable) )+geom_violin()+theme_classic()+geom_jitter( width = 0.1,size=1)+
  ggtitle('Opposite strand feature: Repeat Element\nAverage MAPQ for Ro-Clip peak')+ylab("Avg_mapq - Ro-Clip")+xlab("")+
theme(plot.title = element_text(hjust = 0.5,size = 25),axis.title=element_text(size=20),axis.text=element_text(size=15))+theme(legend.position = "none") +
    theme(axis.text.x = element_text(angle = -45))
 
gg

```

<br>   

For comparison the average peak MAPQ is ploted for each ncRNA subtype.   

<br>  

```{r fig.asp=.5, fig.align='center',fig.height=6, fig.width=12,echo=F,warning=F,eval=T,incude=T}

 
############################################

 p=Peaksdata2_anno
## remove rRNA
  # p=p[!p$Comb_biotype_exon_Oppo%in%'rRNA',]


p=p[p$Comb_biotype_exon_Oppo%in%'ncRNA',]
colnames(p)[colnames(p)%in%'Same_Comb_biotype_ncRNA']='RNA_Type'


 gg=ggplot(p,aes(x=RNA_Type,y=Avg_mapq,fill=RNA_Type) )+geom_violin()+theme_classic()+geom_jitter( width = 0.1,size=1)+
  ggtitle('Average MAPQ for Ro-Clip peak\nby ncRNA Type')+ylab("Average MAPQ - Ro-Clip")+xlab("")+
theme(plot.title = element_text(hjust = 0.5,size = 25),axis.title=element_text(size=20),axis.text=element_text(size=15))+theme(legend.position = "none") +
    theme(axis.text.x = element_text(angle = -45))+scale_fill_brewer(palette = "Paired")

gg
 
 
 
```
 
 
 
```{r fig.asp=.5, fig.align='center',fig.height=6, fig.width=12,echo=F,warning=F,eval=F,incude=F}

 
############################################
####################################################################################################################################
############################################

p=Peaksdata2_anno
p=p[p$Comb_biotype_exon_Oppo%in%'Repeat Element',]

p=p[,c('ID','perc_mm',colnames(p)[grep('Same_Repeat_',colnames(p))])]
p2=p[,colnames(p)[grep('Same_Repeat_',colnames(p))]]
p2[,colnames(p2)[grep('Same_Repeat_',colnames(p2))]]=as.numeric(is.na(p2[,colnames(p2)[grep('Same_Repeat_',colnames(p2))]])==F)

p2[p2[,colnames(p2)[grep('Same_Repeat_',colnames(p2))]]==0]=NA
p2=(p2*p$perc_mm)

p2=melt(p2,id.vars = NULL)


 ggplot(p2,aes(x=variable,y=value) )+geom_violin()+theme_classic()+geom_jitter( width = 0.1,size=1)+
  ggtitle('Peak % Multimapped reads for Ro-Clip\nby ncRNA Type')+ylab("% multimapped reads - Ro-Clip")+xlab("")+
theme(plot.title = element_text(hjust = 0.5,size = 15),axis.title=element_text(size=15),axis.text=element_text(size=12))+theme(legend.position = "none") +
    theme(axis.text.x = element_text(angle = -45))+scale_fill_brewer(palette = "Paired")

      
#######################################      
      
      
p=Peaksdata2_anno
p=p[p$Comb_biotype_exon_Oppo%in%'Repeat Element',]

p=p[,c('ID','perc_mm',colnames(p)[grep('Oppo_Repeat_',colnames(p))])]
p2=p[,colnames(p)[grep('Oppo_Repeat_',colnames(p))]]
p2[,colnames(p2)[grep('Oppo_Repeat_',colnames(p2))]]=as.numeric(is.na(p2[,colnames(p2)[grep('Oppo_Repeat_',colnames(p2))]])==F)

p2[p2[,colnames(p2)[grep('Oppo_Repeat_',colnames(p2))]]==0]=NA
p2=(p2*p$perc_mm)

p2=melt(p2)

 ggplot(p2,aes(x=variable,y=value) )+geom_violin()+theme_classic()+geom_jitter( width = 0.1,size=1)+
  ggtitle('Peak % Multimapped reads for Ro-Clip\nby ncRNA Type')+ylab("% multimapped reads - Ro-Clip")+xlab("")+
theme(plot.title = element_text(hjust = 0.5,size = 15),axis.title=element_text(size=15),axis.text=element_text(size=12))+theme(legend.position = "none") +
    theme(axis.text.x = element_text(angle = -45))+scale_fill_brewer(palette = "Paired")
 



```  


```{r fig.asp=.5, fig.align='center',fig.height=4, fig.width=8,echo=F,warning=F,eval=T,include=F}
#   
# p=Peaksdata2_oppo
# p=p[p$Oppo_Comb_biotype_exon%in%'ncRNA',]
# p=paste0(p$type_comb,collapse = ",")
# p=as.data.frame(strsplit(p,','));colnames(p)='RNA_Type'
# p=p[!p$RNA_Type%in%c('pseudogene','protein_coding'),,drop=F]
# p$RNA=''
# gg=ggplot(p,aes(fill=RNA_Type,x=RNA) )+geom_bar(position="stack",width=.5)+theme_classic()+ggtitle("ncRNA only: Counted by catagory\n(Peak Can be counted multiple times)")+
# theme(plot.title = element_text(hjust = 0.5,size = 25),axis.title=element_text(size=20),axis.text=element_text(size=15),legend.text = element_text( size=30))+
 # scale_color_brewer(palette = "Paired")+scale_fill_brewer(palette = "Paired") 
# ggplotly(gg)

p=Peaksdata2_anno
if (incd_rRNA==F) {
  p=p[!p$Same_Comb_biotype_ncRNA%in%'rRNA',]
}

p=p[p$Oppo_Comb_biotype_exon%in%'ncRNA','Oppo_Comb_biotype_ncRNA',drop=F]
p$RNA=""
gg=ggplot(p,aes(fill=Oppo_Comb_biotype_ncRNA,x=RNA) )+geom_bar(width=.5)+theme_classic()+ggtitle("ncRNA only: # of peaks per catagory")+
 scale_color_brewer(palette = "Paired")+scale_fill_brewer(palette = "Paired") +  blank_theme+ 
  coord_polar("y",start=0)
(gg)

 
# colnames(Peaksdata3)
```       
<!-- <br>    -->
<!-- <br>    -->

```{r fig.asp=.5, fig.align='center',fig.height=6, fig.width=12,echo=F,warning=F,eval=T,include=F}
# **Total Read counts (incude MultiMapped and Unique reads)**  
# <br>


p=Peaksdata2_anno
if (incd_rRNA==F) {
  p=p[!p$Same_Comb_biotype_ncRNA%in%'rRNA',]
}

colnames(p)[colnames(p)%in%'Oppo_Comb_biotype_exon']='RNA_Type'
p$RNA=''
 gg=ggplot(p,aes(x=RNA_Type,y=Counts_fracMM,fill=RNA_Type) )+geom_violin()+theme_classic()+geom_jitter(height = 0, width = 0.1,size=1)+
  scale_y_continuous(trans = "log2")+
  ggtitle('Peak Counts (Unique +Sscaled MM) for Ro-Clip\nby Feature')+ylab("Counts - Ro-Clip")+xlab("")+
theme(plot.title = element_text(hjust = 0.5,size = 25),axis.title=element_text(size=20),axis.text=element_text(size=15))+theme(legend.position = "none") +
    theme(axis.text.x = element_text(angle = -45))
gg

############

p=Peaksdata2_anno
p=p[p$Oppo_Comb_biotype_exon%in%'ncRNA',]
colnames(p)[colnames(p)%in%'Oppo_Comb_biotype_ncRNA']='RNA_Type'


 gg=ggplot(p,aes(x=RNA_Type,y=Counts_fracMM,fill=RNA_Type) )+geom_violin()+theme_classic()+geom_jitter(height = 0, width = 0.1,size=1)+
  scale_y_continuous(trans = "log2")+
  ggtitle('Peak Counts (Unique + Scaled MM) for Ro-Clip\nby ncRNA Type')+ylab("Counts - Ro-Clip")+xlab("")+
theme(plot.title = element_text(hjust = 0.5,size = 25),axis.title=element_text(size=20),axis.text=element_text(size=15))+theme(legend.position = "none") +
    theme(axis.text.x = element_text(angle = -45))+scale_fill_brewer(palette = "Paired")
gg
```  
 
```{r fig.asp=.5, fig.align='center',fig.height=6, fig.width=12,echo=F,warning=F,eval=T,include=F}
# <br>   
# <br>
# **Unique Read counts (incude Unique reads only)**  
# <br> 

p=Peaksdata2_anno
if (incd_rRNA==F) {
  p=p[!p$Same_Comb_biotype_ncRNA%in%'rRNA',]
}

colnames(p)[colnames(p)%in%'Oppo_Comb_biotype_exon']='RNA_Type'
p$RNA=''
 gg=ggplot(p,aes(x=RNA_Type,y=Counts_Unique,fill=RNA_Type) )+geom_violin()+theme_classic()+geom_jitter(height = 0, width = 0.1,size=1)+
  scale_y_continuous(trans = "log2")+
  ggtitle('Peak Counts (Unique) for Ro-Clip\nby Feature')+ylab("Counts - Ro-Clip")+xlab("")+
theme(plot.title = element_text(hjust = 0.5,size = 25),axis.title=element_text(size=20),axis.text=element_text(size=15))+theme(legend.position = "none") +
    theme(axis.text.x = element_text(angle = -45))
# 
gg

##################

p=Peaksdata2_anno
if (incd_rRNA==F) {
  p=p[!p$Same_Comb_biotype_ncRNA%in%'rRNA',]
}

p=p[p$Oppo_Comb_biotype_exon%in%'ncRNA',]
colnames(p)[colnames(p)%in%'Oppo_Comb_biotype_ncRNA']='RNA_Type'


 gg=ggplot(p,aes(x=RNA_Type,y=Counts_Unique,fill=RNA_Type) )+geom_violin()+theme_classic()+geom_jitter(height = 0, width = 0.1,size=1)+
  scale_y_continuous(trans = "log2")+
  ggtitle('Peak Counts for Ro-Clip\nby ncRNA Type')+ylab("Counts - Ro-Clip")+xlab("")+
theme(plot.title = element_text(hjust = 0.5,size = 25),axis.title=element_text(size=20),axis.text=element_text(size=15))+theme(legend.position = "none") +
    theme(axis.text.x = element_text(angle = -45))+scale_fill_brewer(palette = "Paired")
gg
```  
<br>   
  
<br>   

<br>   

### Read Counts by peak for each catagory    
  
#### Unique Read counts (incude Unique reads only)   

<br>   

```{r fig.align='center',fig.height=6, fig.width=12,echo=F,warning=F,eval=T, include=T}
p=Peaksdata2_anno
p$Same_Comb_biotype_exon=p$Comb_biotype_exon_Oppo
if (incd_rRNA==F) {
  p=p[!p$Same_Comb_biotype_ncRNA%in%'rRNA',]
}

colnames(p)[colnames(p)%in%'Same_Comb_biotype_exon']='RNA_Type'
p$RNA=''
 gg=ggplot(p,aes(x=RNA_Type,y=Counts_Unique,fill=RNA_Type) )+geom_violin()+theme_classic()+geom_jitter(height = 0, width = 0.1,size=1)+
  scale_y_continuous(trans = "log2")+
  ggtitle('Peak Counts (Unique) for Ro-Clip\nby Feature')+ylab("Counts - Ro-Clip")+xlab("")+
theme(plot.title = element_text(hjust = 0.5,size = 25),axis.title=element_text(size=20),axis.text=element_text(size=15))+theme(legend.position = "none") +
    theme(axis.text.x = element_text(angle = -45))
# 
gg
```  
<!-- <font size="1">Columns: RNA_type_Classification </font>      -->
<br/>    

<!-- no Feature peaks with counts > 20 = `r nrow(Peaksdata2_anno[Peaksdata2_anno$Counts_Unique>20&(Peaksdata2_anno$Comb_biotype_exon_Oppo%in%'no Feature'),])`   -->

<br/>    

```{r fig.asp=.5, fig.align='center',fig.height=6, fig.width=8,echo=F,warning=F,eval=T, include=T}
p=Peaksdata2_anno
p$Same_Comb_biotype_exon=p$Comb_biotype_exon_Oppo
# if (incd_rRNA==F) {
#   p=p[!p$Same_Comb_biotype_ncRNA%in%'rRNA',]
# }

p=p[p$Same_Comb_biotype_exon%in%'ncRNA',]
p$ID_name=paste0(p$ID,"\n",p$Same_gene_name_comb)


colnames(p)[colnames(p)%in%'Same_Comb_biotype_ncRNA']='RNA_Type'


 gg=ggplot(p,aes(x=RNA_Type,y=Counts_Unique,fill=RNA_Type) )+geom_violin()+theme_classic()+geom_jitter(height = 0, width = 0.1,size=0.5,aes(text=ID_name))+
  scale_y_continuous(trans = "log2")+
  ggtitle('Peak Counts (Unique) for Ro-Clip\nby Feature')+ylab("Counts - Ro-Clip")+xlab("")+
theme(plot.title = element_text(hjust = 0.5,size = 15),axis.title=element_text(size=15),axis.text=element_text(size=12))+theme(legend.position = "none") +
    theme(axis.text.x = element_text(angle = -45))+scale_fill_brewer(palette = "Paired")
 ggplotly(gg,tooltip = c("text", "y"))

```  
<br>   
<br>   

#### Scaled Read counts (incudes Unique reads + [MultiMapped read / # of different mappings] )     

Does include rRNA
<br>   

```{r fig.asp=.5, fig.align='center',fig.height=6, fig.width=12,echo=F,warning=F,eval=T}
p=Peaksdata2_anno
p$Same_Comb_biotype_exon=p$Comb_biotype_exon_Oppo
# if (incd_rRNA==F) {
#   p=p[!p$Same_Comb_biotype_ncRNA%in%'rRNA',]
# }

colnames(p)[colnames(p)%in%'Same_Comb_biotype_exon']='RNA_Type'
p$RNA=''
 gg=ggplot(p,aes(x=RNA_Type,y=Counts_fracMM,fill=RNA_Type) )+geom_violin()+theme_classic()+geom_jitter(height = 0, width = 0.1,size=1)+
  scale_y_continuous(trans = "log2")+
  ggtitle('Peak Counts (Unique + scaled MM) for Ro-Clip\nby Feature')+ylab("Counts - Ro-Clip")+xlab("")+
theme(plot.title = element_text(hjust = 0.5,size = 25),axis.title=element_text(size=20),axis.text=element_text(size=15))+theme(legend.position = "none") +
    theme(axis.text.x = element_text(angle = -45))
# 
gg
```  
<!-- <font size="1">Columns: RNA_type_Classification </font>      -->
<br/>    

<!-- no Feature peaks with counts > 20 = `r nrow(Peaksdata2_anno[Peaksdata2_anno$Counts_fracMM>20&(Peaksdata2_anno$Comb_biotype_exon_Oppo%in%'no Feature'),])`   -->

<br/>    

```{r fig.asp=.5,fig.align='center',fig.height=6, fig.width=8,echo=F,warning=F,eval=T}
p=Peaksdata2_anno
p$Same_Comb_biotype_exon=p$Comb_biotype_exon_Oppo
# if (incd_rRNA==F) {
#   p=p[!p$Same_Comb_biotype_ncRNA%in%'rRNA',]
# }

p=p[p$Same_Comb_biotype_exon%in%'ncRNA',]
colnames(p)[colnames(p)%in%'Same_Comb_biotype_ncRNA']='RNA_Type'
p$ID_name=paste0(p$ID,"\n",p$Same_gene_name_comb)

 gg=ggplot(p,aes(x=RNA_Type,y=Counts_fracMM,fill=RNA_Type) )+geom_violin()+theme_classic()+geom_jitter(height = 0, width = 0.05,size=.5,aes(text=ID_name))+
  scale_y_continuous(trans = "log2")+
  ggtitle('Peak Counts (Unique + scaled MM) for Ro-Clip\nby ncRNA Type')+ylab("Counts - Ro-Clip")+xlab("")+
theme(plot.title = element_text(hjust = 0.5,size = 15),axis.title=element_text(size=15),axis.text=element_text(size=12))+theme(legend.position = "none") +
    theme(axis.text.x = element_text(angle = -45))+scale_fill_brewer(palette = "Paired")

 ggplotly(gg,tooltip = c("text", "y"))
```  
<br>   
<br>   

#### Best Mapped Read counts (incuden Unique reads + Best mapping location for MM reads )    

Multimapped reads are only assigned to best mapping location and get a count of 1.  

<br>   

```{r fig.asp=.5, fig.align='center',fig.height=6, fig.width=12,echo=F,warning=F,eval=T }
p=Peaksdata2_anno
p$Same_Comb_biotype_exon=p$Comb_biotype_exon_Oppo
# if (incd_rRNA==F) {
#   p=p[!p$Same_Comb_biotype_ncRNA%in%'rRNA',]
# }

colnames(p)[colnames(p)%in%'Same_Comb_biotype_exon']='RNA_Type'
p$RNA=''
 gg=ggplot(p,aes(x=RNA_Type,y=Counts_primary,fill=RNA_Type) )+geom_violin()+theme_classic()+geom_jitter(height = 0, width = 0.1,size=1)+
  scale_y_continuous(trans = "log2")+
  ggtitle('Peak Counts (Unique + Best Mapped MM) for Ro-Clip\nby Feature')+ylab("Counts - Ro-Clip")+xlab("")+
theme(plot.title = element_text(hjust = 0.5,size = 25),axis.title=element_text(size=20),axis.text=element_text(size=15))+theme(legend.position = "none") +
    theme(axis.text.x = element_text(angle = -45))
# 
gg
```  
<!-- <font size="1">Columns: RNA_type_Classification </font>      -->
<br/>    

<!-- no Feature peaks with counts > 20 = `r nrow(Peaksdata2_anno[Peaksdata2_anno$Counts_fracMM>20&(Peaksdata2_anno$Comb_biotype_exon_Oppo%in%'no Feature'),])`   -->

<br/>    

```{r fig.asp=.5, fig.align='center',fig.height=6, fig.width=8,echo=F,warning=F,eval=T}
p=Peaksdata2_anno
p$Same_Comb_biotype_exon=p$Comb_biotype_exon_Oppo
# if (incd_rRNA==F) {
#   p=p[!p$Same_Comb_biotype_ncRNA%in%'rRNA',]
# }

p=p[p$Same_Comb_biotype_exon%in%'ncRNA',]
colnames(p)[colnames(p)%in%'Same_Comb_biotype_ncRNA']='RNA_Type'
p$ID_name=paste0(p$ID,"\n",p$Same_gene_name_comb)

 gg=ggplot(p,aes(x=RNA_Type,y=Counts_primary,fill=RNA_Type) )+geom_violin()+theme_classic()+geom_jitter(height = 0, width = 0.05,size=.5,aes(text=ID_name))+
  scale_y_continuous(trans = "log2")+
  ggtitle('Peak Counts (Unique + Best Mapped MM) for Ro-Clip\nby ncRNA Type')+ylab("Counts - Ro-Clip")+xlab("")+
theme(plot.title = element_text(hjust = 0.5,size = 15),axis.title=element_text(size=15),axis.text=element_text(size=12))+theme(legend.position = "none") +
    theme(axis.text.x = element_text(angle = -45))+scale_fill_brewer(palette = "Paired")

 ggplotly(gg,tooltip = c("text", "y"))
```  
<br>   
<br> 
<!-- #### % Multimapped reads for each peak    -->
<!-- <br> -->
```{r fig.asp=.5, fig.align='center',fig.height=6, fig.width=12,echo=F,warning=F,eval=F,incude=F}
p=Peaksdata2_anno
p$Same_Comb_biotype_exon=p$Comb_biotype_exon_Oppo
if (incd_rRNA==F) {
  p=p[!p$Same_Comb_biotype_ncRNA%in%'rRNA',]
}

colnames(p)[colnames(p)%in%'Same_Comb_biotype_exon']='RNA_Type'
p$RNA=''
 gg=ggplot(p,aes(x=RNA_Type,y=perc_mm,fill=RNA_Type) )+geom_violin()+theme_classic()+geom_jitter( width = 0.1,size=1)+
  ggtitle('Peak % Multimapped reads for Ro-Clip\nby Feature')+ylab("% multimapped reads - Ro-Clip")+xlab("")+
theme(plot.title = element_text(hjust = 0.5,size = 25),axis.title=element_text(size=20),axis.text=element_text(size=15))+theme(legend.position = "none") +
    theme(axis.text.x = element_text(angle = -45))
gg

```  
<!-- <font size="1">Columns: RNA_type_Classification </font>      -->
<!-- <br/>     -->
```{r fig.asp=.5, fig.align='center',fig.height=6, fig.width=12,echo=F,warning=F,eval=F,incude=F}
p=Peaksdata2_anno
p$Same_Comb_biotype_exon=p$Comb_biotype_exon_Oppo
if (incd_rRNA==F) {
  p=p[!p$Same_Comb_biotype_ncRNA%in%'rRNA',]
}

p=p[p$Same_Comb_biotype_exon%in%'ncRNA',]
colnames(p)[colnames(p)%in%'Same_Comb_biotype_ncRNA']='RNA_Type'


 gg=ggplot(p,aes(x=RNA_Type,y=perc_mm,fill=RNA_Type) )+geom_violin()+theme_classic()+geom_jitter( width = 0.1,size=1)+
  ggtitle('Peak % Multimapped reads for Ro-Clip\nby ncRNA Type')+ylab("% multimapped reads - Ro-Clip")+xlab("")+
theme(plot.title = element_text(hjust = 0.5,size = 15),axis.title=element_text(size=15),axis.text=element_text(size=12))+theme(legend.position = "none") +
    theme(axis.text.x = element_text(angle = -45))+scale_fill_brewer(palette = "Paired")

gg
```  



<br>   

### Number of reads assigned to each catagory   

#### Counts for RNA biotype  

<br>   

```{r fig.asp=.5, fig.align='center',fig.height=4, fig.width=8,echo=F,warning=F,include=T,eval=T}
# brewer.pal(n = 8, name = "Dark2")

nc_color=c("#1B9E77","#1fb487","#2bdba6" ,"#D95F02", "#7570B3" ,"#E7298A", "#66A61E", "#E6AB02", "#A6761D", "#666666")

p=Peaksdata2_anno
p$Same_Comb_biotype_exon=p$Comb_biotype_exon_Oppo
# if (incd_rRNA==F) {
#   p=p[!p$Same_Comb_biotype_ncRNA%in%'rRNA',]
# }
p$Same_Comb_biotype_exon=as.character(p$Same_Comb_biotype_exon)
p[p$Same_Comb_biotype_ncRNA%in%'rRNA','Same_Comb_biotype_exon']='rRNA'
p[p$Same_Comb_biotype_ncRNA%in%'yRNA','Same_Comb_biotype_exon']='yRNA'

p$Same_Comb_biotype_exon=factor(p$Same_Comb_biotype_exon, levels = c("ncRNA","rRNA","yRNA", "protein_coding: exon", "Repeat Element","pseudogene","Antisense Feature","protein_coding: Intron","lncRNA","no Feature"))


u=unique(p$Same_Comb_biotype_exon)
pcount=data.frame(u,1:length(u));colnames(pcount)=c('RNA_Type','Counts_All');pcount$Counts_Unique=NA;pcount$Counts_MM=NA;pcount$Peak_count=NA
for (x in 1:length(u)) {
  pam=p[p$Same_Comb_biotype_exon%in%u[x],]
  pcount[x,1]=u[x]
  pcount[x,'Counts_All']=colSums(pam[,'Counts_All',drop=F],na.rm = T)
  pcount[x,'Counts_Unique']=colSums(pam[,'Counts_Unique',drop=F],na.rm = T)
  pcount[x,'Counts_MM']=colSums(pam[,'Counts_MM',drop=F],na.rm = T)
  pcount[x,'Counts_primary']=colSums(pam[,'Counts_primary',drop=F],na.rm = T)
  pcount[x,'Counts_fracMM']=colSums(pam[,'Counts_fracMM',drop=F],na.rm = T)
  pcount[x,'Counts_Primary_fracMM']=colSums(pam[,'Counts_Primary_fracMM',drop=F],na.rm = T)
    pcount[x,'Peak_count']=nrow(pam)

}
pcount$RNA=""

# gg=
# ggplot(pcount,aes(fill=RNA_Type,x=RNA,y=Counts_All) )+geom_bar(stat='identity',position="stack",width=.5)+theme_classic()+ggtitle("BioType - # of reads (mm+unique) by catagory)")+
#  scale_fill_brewer(palette = "Dark2") 
# ggplotly(gg)
# 

gg=
ggplot(pcount,aes(fill=RNA_Type,x=RNA,y=Counts_Unique) )+geom_bar(stat='identity',position="stack",width=.5)+theme_classic()+ggtitle("BioType - # of reads (unique) by catagory)")+
 scale_fill_manual(values=nc_color)
  # scale_fill_brewer(palette = "Dark2") 
  
ggplotly(gg)


gg=
ggplot(pcount,aes(fill=RNA_Type,x=RNA,y=Counts_fracMM) )+geom_bar(stat='identity',position="stack",width=.5)+theme_classic()+ggtitle("BioType - # of reads (unique + Scaled MM) by catagory)")+
scale_fill_manual(values=nc_color)
  # scale_fill_brewer(palette = "Dark2")
ggplotly(gg)

gg=
ggplot(pcount,aes(fill=RNA_Type,x=RNA,y=Counts_primary) )+geom_bar(stat='identity',position="stack",width=.5)+theme_classic()+ggtitle("BioType - # of reads (unique + Best Mapped MM) by catagory)")+
 scale_fill_manual(values=nc_color)
  # scale_fill_brewer(palette = "Dark2") 
ggplotly(gg)
```
<br>    

#### Counts for ncRNA      


```{r fig.asp=.5, fig.align='center',fig.height=4, fig.width=8,echo=F,warning=F,include=T}

p=Peaksdata2_anno
p$Same_Comb_biotype_exon=p$Comb_biotype_exon_Oppo
# if (incd_rRNA==F) {
#   p=p[!p$Same_Comb_biotype_ncRNA%in%'rRNA',]
# }
p$Same_Comb_biotype_exon=as.character(p$Same_Comb_biotype_exon)
p[p$Same_Comb_biotype_ncRNA%in%'rRNA','Same_Comb_biotype_exon']='rRNA'
p[p$Same_Comb_biotype_ncRNA%in%'yRNA','Same_Comb_biotype_exon']='yRNA'

p$Same_Comb_biotype_exon=factor(p$Same_Comb_biotype_exon, levels = c("ncRNA","rRNA","yRNA", "protein_coding: exon", "Repeat Element","pseudogene","Antisense Feature","protein_coding: Intron","lncRNA","no Feature"))

p=p[p$Same_Comb_biotype_exon%in%'ncRNA',]

u=unique(p$Same_Comb_biotype_ncRNA)
pcount=data.frame(u,1:length(u));colnames(pcount)=c('Same_RNA_Subtype','Counts_All');pcount$Counts_Unique=NA;pcount$Counts_MM=NA;pcount$Peak_count=NA
for (x in 1:length(u)) {
  pam=p[p$Same_Comb_biotype_ncRNA%in%u[x],]
  pcount[x,1]=u[x]
  pcount[x,'Counts_All']=colSums(pam[,'Counts_All',drop=F],na.rm = T)
  pcount[x,'Counts_Unique']=colSums(pam[,'Counts_Unique',drop=F],na.rm = T)
  pcount[x,'Counts_MM']=colSums(pam[,'Counts_MM',drop=F],na.rm = T)
    pcount[x,'Counts_primary']=colSums(pam[,'Counts_primary',drop=F],na.rm = T)
  pcount[x,'Counts_fracMM']=colSums(pam[,'Counts_fracMM',drop=F],na.rm = T)
  pcount[x,'Counts_Primary_fracMM']=colSums(pam[,'Counts_Primary_fracMM',drop=F],na.rm = T)
  pcount[x,'Peak_count']=nrow(pam)
}

pcount$RNA=''

# gg=ggplot(pcount,aes(fill=Same_RNA_Subtype,x=RNA,y=Counts_All) )+geom_bar(stat='identity',width=.5)+theme_classic()+ggtitle("ncRNA only: # of reads (mm+unique) by catagory")+
#  scale_color_brewer(palette = "Paired")+scale_fill_brewer(palette = "Paired") +  blank_theme+ 
#   coord_polar("y",start=0)
# (gg)

gg=ggplot(pcount,aes(fill=Same_RNA_Subtype,x=RNA,y=Counts_Unique) )+geom_bar(stat='identity',width=.5)+theme_classic()+ggtitle("ncRNA only: # of reads (unique) by catagory")+
 scale_color_brewer(palette = "Paired")+scale_fill_brewer(palette = "Paired") +  blank_theme+ 
  coord_polar("y",start=0)
(gg)

gg=ggplot(pcount,aes(fill=Same_RNA_Subtype,x=RNA,y=Counts_fracMM) )+geom_bar(stat='identity',width=.5)+theme_classic()+ggtitle("ncRNA only: # of reads (unique + Scaled MM) by catagory")+
 scale_color_brewer(palette = "Paired")+scale_fill_brewer(palette = "Paired") +  blank_theme+ 
  coord_polar("y",start=0)
(gg)

gg=ggplot(pcount,aes(fill=Same_RNA_Subtype,x=RNA,y=Counts_primary) )+geom_bar(stat='identity',width=.5)+theme_classic()+ggtitle("ncRNA only: # of reads (unique + Best Mapped MM) by catagory")+
 scale_color_brewer(palette = "Paired")+scale_fill_brewer(palette = "Paired") +  blank_theme+ 
  coord_polar("y",start=0)
(gg)
# colnames(Peaksdata3)
```      
<!-- <font size="1">Columns: RNA_type_Classification + Counts_Unique </font>      -->







<!-- <br>    -->

<!-- ### Number of reads for each catagory   -->

<br>   

```{r fig.asp=.5, fig.align='center',fig.height=4, fig.width=8,echo=F,warning=F,eval=T,include=F}

p=Peaksdata2_anno
if (incd_rRNA==F) {
  p=p[!p$Oppo_Comb_biotype_ncRNA%in%'rRNA',]
}

u=unique(p$Oppo_Comb_biotype_exon)
pcount=data.frame(u,1:length(u));colnames(pcount)=c('RNA_Type','Counts_All');pcount$Counts_Unique=NA;pcount$Counts_MM=NA;pcount$Peak_count=NA
for (x in 1:length(u)) {
  pam=p[p$Oppo_Comb_biotype_exon%in%u[x],]
  pcount[x,1]=u[x]
  pcount[x,'Counts_All']=colSums(pam[,'Counts_All',drop=F],na.rm = T)
  pcount[x,'Counts_Unique']=colSums(pam[,'Counts_Unique',drop=F],na.rm = T)
  pcount[x,'Counts_MM']=colSums(pam[,'Counts_MM',drop=F],na.rm = T)
  pcount[x,'Counts_primary']=colSums(pam[,'Counts_primary',drop=F],na.rm = T)
  pcount[x,'Counts_fracMM']=colSums(pam[,'Counts_fracMM',drop=F],na.rm = T)
  pcount[x,'Counts_Primary_fracMM']=colSums(pam[,'Counts_Primary_fracMM',drop=F],na.rm = T)
    pcount[x,'Peak_count']=nrow(pam)

}
pcount$RNA=""

# gg=
# ggplot(pcount,aes(fill=RNA_Type,x=RNA,y=Counts_All) )+geom_bar(stat='identity',position="stack",width=.5)+theme_classic()+ggtitle("BioType - # of reads (mm+unique) by catagory)")+
#  scale_fill_brewer(palette = "Dark2") 
# ggplotly(gg)

gg=
ggplot(pcount,aes(fill=RNA_Type,x=RNA,y=Counts_Unique) )+geom_bar(stat='identity',position="stack",width=.5)+theme_classic()+ggtitle("BioType - # of reads (unique) by catagory)")+
 scale_fill_brewer(palette = "Dark2") 
ggplotly(gg)

gg=
ggplot(pcount,aes(fill=RNA_Type,x=RNA,y=Counts_fracMM) )+geom_bar(stat='identity',position="stack",width=.5)+theme_classic()+ggtitle("BioType - # of reads (Unique + Scaled Multimap) by catagory)")+
 scale_fill_brewer(palette = "Dark2")
ggplotly(gg)
```


```{r fig.asp=.5, fig.align='center',fig.height=4, fig.width=8,echo=F,warning=F,eval=T,include=F}

p=Peaksdata2_anno
if (incd_rRNA==F) {
  p=p[!p$Oppo_Comb_biotype_ncRNA%in%'rRNA',]
}
p=p[p$Oppo_Comb_biotype_exon%in%'ncRNA',]

u=unique(p$Oppo_Comb_biotype_ncRNA)
pcount=data.frame(u,1:length(u));colnames(pcount)=c('RNA_Subtype','Counts_All');pcount$Counts_Unique=NA;pcount$Counts_MM=NA;pcount$Peak_count=NA
for (x in 1:length(u)) {
  pam=p[p$Oppo_Comb_biotype_ncRNA%in%u[x],]
  pcount[x,1]=u[x]
  pcount[x,'Counts_All']=colSums(pam[,'Counts_All',drop=F],na.rm = T)
  pcount[x,'Counts_Unique']=colSums(pam[,'Counts_Unique',drop=F],na.rm = T)
  pcount[x,'Counts_MM']=colSums(pam[,'Counts_MM',drop=F],na.rm = T)
  pcount[x,'Counts_primary']=colSums(pam[,'Counts_primary',drop=F],na.rm = T)
  pcount[x,'Counts_fracMM']=colSums(pam[,'Counts_fracMM',drop=F],na.rm = T)
  pcount[x,'Counts_Primary_fracMM']=colSums(pam[,'Counts_Primary_fracMM',drop=F],na.rm = T)
  pcount[x,'Peak_count']=nrow(pam)
}


pcount$RNA=''

# 
# gg=ggplot(pcount,aes(fill=RNA_Subtype,x=RNA,y=Counts_All) )+geom_bar(stat='identity',width=.5)+theme_classic()+ggtitle("ncRNA only: # of reads (mm+unique) by catagory")+
#  scale_color_brewer(palette = "Paired")+scale_fill_brewer(palette = "Paired") +  blank_theme+ 
#   coord_polar("y",start=0)
# (gg)


gg=ggplot(pcount,aes(fill=RNA_Subtype,x=RNA,y=Counts_Unique) )+geom_bar(stat='identity',width=.5)+theme_classic()+ggtitle("ncRNA only: # of reads (unique) by catagory")+
 scale_color_brewer(palette = "Paired")+scale_fill_brewer(palette = "Paired") +  blank_theme+ 
  coord_polar("y",start=0)
(gg)


gg=ggplot(pcount,aes(fill=RNA_Subtype,x=RNA,y=Counts_fracMM) )+geom_bar(stat='identity',width=.5)+theme_classic()+ggtitle("ncRNA only: # of reads (Scaled Multimap) by catagory")+
 scale_color_brewer(palette = "Paired")+scale_fill_brewer(palette = "Paired") +  blank_theme+ 
  coord_polar("y",start=0)
(gg)
```      
<!-- <font size="1">Columns: Opposite Strand: RNA_type </font>      -->



```{r fig.asp=.5, fig.align='center',fig.height=6, fig.width=12,echo=F,warning=F,eval=T,include=F}
# colnames(Peaksdata2_anno)

Peaksdata3_anno=Peaksdata2_anno[,c("ID","strand","Avg_mapq",'Counts_Unique','Counts_All','Counts_fracMM','Counts_primary','perc_mm',"CountsCntrl_Unique","CountsCntrl_All","CountsCntrl_frac_ftrCount","CountsCntrl_Primary","Same_gene_name_comb","Same_ensembl_gene_id","Same_type_comb","Same_repeat_comb",'Same_feature',"Same_Comb_biotype_exon","Same_Comb_biotype_ncRNA","Oppo_gene_name_comb","Oppo_ensembl_gene_id","Oppo_type_comb","Oppo_repeat_comb",'Oppo_feature',"Oppo_Comb_biotype_exon","Same_Comb_biotype_ncRNA","Comb_biotype_exon_Oppo")]

colnames(Peaksdata3_anno)=c("ID","strand","Average MAPQ", 'Counts_Unique', 'Counts_Multimappers_All', 'Counts_Multimappers_Scaled', 'Counts_Multimappers_BestMapping', 'Multimaped Reads %',"Control:Counts_Unique","Control:Counts_Multimappers_All","Control:Counts_Multimappers_Scaled","Control:Counts_Multimappers_BestMapping","Same Strand: Host_gene_name","Same Strand: Host_gene_ensembl_id","Same Strand: RNA_type",'Same Strand: Repeat_Type',"Same Strand: Intron Exon",'Same Strand: RNA_type_Classification','Same Strand: ncRNA_Classification',"Opposite Strand: Host_gene_name","Opposite Strand: Host_gene_ensembl_id","Opposite Strand: RNA_type",'Opposite Strand: Repeat_Type',"Opposite Strand: Intron Exon",'Opposite Strand: RNA_type_Classification','Opposite Strand: ncRNA_Classification','Both Strand: RNA_type_Classification')


Peaksdata4=Peaksdata3_anno

```


## Identify Alternative Mapping locations for Multimapped reads    


```{r fig.asp=.5, fig.align='center',fig.height=6, fig.width=12,echo=F,warning=F,eval=T,include=F}
library(GenomicAlignments)


BamMM=readGAlignments('./NovoAlign_umi/All/recomb_NHtag/bam/Ro_Clip_iCountcutadpt_all.unique.NH.mm.ddup.s.mm.bam',use.names=T)
mcols(BamMM)$qname=names(BamMM)

Peaksdata2_anno.GR = GRanges(seqnames = as.character(Peaksdata2_anno$chr), ranges=IRanges(start = as.numeric(Peaksdata2_anno$start), end = as.numeric(Peaksdata2_anno$end)),strand = Peaksdata2_anno$strand,
                   ID=Peaksdata2_anno$ID,
                   Same_Host_gene_name=Peaksdata2_anno$Same_gene_name_comb,
                   Same_Host_gene_ensembl_id=Peaksdata2_anno$Same_ensembl_gene_id,
                    counts_Ro=Peaksdata2_anno$Counts_Unique,
                   Same_Comb_biotype_ncRNA=Peaksdata2_anno$Same_Comb_biotype_ncRNA,
                   Same_Comb_biotype_exon=Peaksdata2_anno$Same_Comb_biotype_exon)

##############################################
        q = Peaksdata2_anno.GR
        s = BamMM
        
        # p = BamMM
        # s = Peaksdata4.GR 
        xo=as.data.frame(GenomicRanges::findOverlaps(q,s,type = "any",ignore.strand=T),row.names = NULL)
          
        qh=as.data.frame(q[xo$queryHits],row.names = NULL)
          colnames(qh)=paste0(colnames(qh),"_peaks")
        sh=as.data.frame(s[xo$subjectHits],row.names = NULL)
          colnames(sh)=paste0(colnames(sh),"_bam")
        bamMM_anno=cbind(qh,sh)
        
# sum(Peaksdata2$Counts_MM)
##############################################
        
Peaksdata2_MM=Peaksdata2_anno[Peaksdata2_anno$Counts_MM>1,]
Peaksdata2_MM=Peaksdata2_MM[Peaksdata2_MM$Same_Comb_biotype_exon%in%'ncRNA',]

out=as.data.frame(matrix(nrow=nrow(Peaksdata2_MM),ncol=12));
colnames(out)=c("ID","Peak_biotype","Number_ALLMM","MultiMappers_ncRNA_biotype",'Number_MMinpeak','Number_MMoutpeak',"Number_AltMappings",'Number_Altlocations',"Altlocations","Number_Altlocations_SamePeaktype","Number_Altlocations_DifferentPeaktype","Perc_DifferentPeaktype")
for (x in 1:nrow(Peaksdata2_MM)) {
  # for (x in 1:17) {
  PL=Peaksdata2_MM[x,]
  
  
  PL.GR=GRanges(seqnames = as.character(PL$chr), ranges=IRanges(start = as.numeric(PL$start), end = as.numeric(PL$end)),strand = PL$strand,
                   ID=PL$ID,
                   counts_Ro=PL$Counts_Unique,
                   Same_Host_gene_name=as.character(PL$Same_gene_name_comb),
                   Same_Host_gene_ensembl_id=as.character(PL$Same_ensembl_gene_id),
                   Same_Comb_biotype_ncRNA=as.character(PL$Same_Comb_biotype_ncRNA),
                   Same_Comb_biotype_exon=as.character(PL$Same_Comb_biotype_exon))
        q = PL.GR
        s = BamMM

        xo=as.data.frame(GenomicRanges::findOverlaps(q,s,type = "any",ignore.strand=T))
        qh=as.data.frame(q[xo$queryHits],row.names = NULL)
          colnames(qh)=paste0(colnames(qh),"_peaks")
        sh=as.data.frame(s[xo$subjectHits],row.names = NULL)
          colnames(sh)=paste0(colnames(sh),"_bam")
        bam_PL=cbind(qh,sh)
  
        Peak_MultiMappers=bamMM_anno[bamMM_anno$qname_bam%in%bam_PL$qname_bam,]
        Peak_MultiMappers[is.na(Peak_MultiMappers$Same_Comb_biotype_ncRNA_peaks),'Same_Comb_biotype_ncRNA_peaks']=as.character(Peak_MultiMappers[is.na(Peak_MultiMappers$Same_Comb_biotype_ncRNA_peaks),'Same_Comb_biotype_exon_peaks'])

        
       out[x,'ID']=PL$ID
       out[x,'Peak_biotype']=PL$Same_Comb_biotype_ncRNA
          m_NC_BT=unique(Peak_MultiMappers$Same_Comb_biotype_ncRNA_peaks)
          m_NC_BT=m_NC_BT[is.na(m_NC_BT)==F]
          # m_NC_BT=m_NC_BT[(m_NC_BT%in%PL$Same_Comb_biotype_ncRNA)==F]
        
       out[x,'MultiMappers_ncRNA_biotype']=paste0(m_NC_BT,collapse = ', ')
       out[x,'Number_ALLMM']=nrow(Peak_MultiMappers)
       out[x,'Number_MMinpeak']=length(bam_PL$qname_bam)
       out[x,'Number_MMoutpeak']=nrow(Peak_MultiMappers[!Peak_MultiMappers$ID%in%PL$ID,])
       out[x,'Number_AltMappings']=nrow(Peak_MultiMappers)-PL$Counts_MM
       out[x,'Number_Altlocations']=length(unique(Peak_MultiMappers[!Peak_MultiMappers$ID%in%PL$ID,"ID_peaks"]))
       
       out[x,c('Altlocations','Number_Altlocations_SamePeaktype','Number_Altlocations_DifferentPeaktype','Perc_DifferentPeaktype')]=NA    
       
              ALTLOC=((Peak_MultiMappers[!Peak_MultiMappers$ID_peaks%in%PL$ID,]))
       if (nrow(ALTLOC)>0) {
              ALTLOC$ALTLOCname=paste0(ALTLOC$ID_peaks,'_',ALTLOC$Same_Comb_biotype_ncRNA_peaks)
              ALTLOCname=unique(ALTLOC$ALTLOCname)
              ALTLOCname=ALTLOCname[is.na(ALTLOCname)==F]
       out[x,'Altlocations']=paste0(ALTLOCname,collapse = ", ")
       Peak_MultiMappers_AltLoc=Peak_MultiMappers[Peak_MultiMappers$ID_peaks%in%unique(ALTLOC$ID_peaks),]
      
       ###################################
       #      ALTLOC=unique(unique(Peak_MultiMappers[!Peak_MultiMappers$ID%in%PL$ID,"ID_peaks"]))
       #      ALTLOC=ALTLOC[is.na(ALTLOC)==F]
       # out[x,'Altlocations']=paste0(ALTLOC,collapse = ", ")
       # Peak_MultiMappers_AltLoc=Peak_MultiMappers[Peak_MultiMappers$ID_peaks%in%ALTLOC,]        
    #########################
       
       out[x,'Number_Altlocations_SamePeaktype']=nrow(Peak_MultiMappers_AltLoc[Peak_MultiMappers_AltLoc$Same_Comb_biotype_ncRNA_peaks%in%(PL$Same_Comb_biotype_ncRNA),])
       out[x,'Number_Altlocations_DifferentPeaktype']=nrow(Peak_MultiMappers_AltLoc[!Peak_MultiMappers_AltLoc$Same_Comb_biotype_ncRNA_peaks%in%(PL$Same_Comb_biotype_ncRNA),])
       out[x,'Perc_DifferentPeaktype']=(out[x,'Number_Altlocations_DifferentPeaktype']/out[x,'Number_MMoutpeak'])*100
       }
       if (x%in%seq(from = 0, to = 10000, by = 50)) {print(x/nrow(Peaksdata2_MM))       }
# remove(PL,q,s,xo,qh,sh,bam_PL,Peak_MultiMappers,ALTLOC,m_NC_BT)
}

# View(as.character(out[1,'Altlocations']))
# chr1:103844667-103844707


# out[,c("ID",'Number_MMinpeak','Number_MMoutpeak',"Number_AltMappings",'Number_Altlocations',"Number_Altlocations_SamePeaktype","Number_Altlocations_DifferentPeaktype","Perc_DifferentPeaktype")]
# out[,c("ID","MultiMappers_ncRNA_biotype",'Number_Altlocations_SamePeaktype','Number_Altlocations_DifferentPeaktype','Altlocations')]

out=out[!is.na(out$Altlocations),]

Peaksdata4_alt=out[,c("ID","Number_MMoutpeak",'Number_Altlocations_DifferentPeaktype','Number_Altlocations_SamePeaktype','MultiMappers_ncRNA_biotype','Perc_DifferentPeaktype','Altlocations')]
colnames(Peaksdata4_alt)=
  c("ID","MM_number",'MM_Alt_Peaktype','MM_Same_Peaktype','MM_biotype','MM_Prec_Alt_Peaktype','MM_Alt_Alignments')

Peaksdata4_alt=merge(Peaksdata4,Peaksdata4_alt,by='ID',all.x=T)


# Peaksdata2_alt[,c("ID","Number_MMoutpeak",'Number_Altlocations_DifferentPeaktype','Number_Altlocations_SamePeaktype','MultiMappers_ncRNA_biotype','Altlocations')] 

Peaksdata4=Peaksdata4_alt

```



```{r fig.asp=.5, fig.align='center',fig.height=6, fig.width=12,echo=F,warning=F}

# ```{r fig.asp=.5, fig.align='center',fig.height=6, fig.width=12,echo=F,warning=F}
# 
# p=Peaksdata2
# # p[is.na(p$Var_type),'Var_type']="No Mutation"
# # p[grepl("_",p$Var_type),'Var_type']="Multiple"
# #
# # p[grepl("\\|",p$Var_type),'Var_type']="Multiple"
#
# # p=p[p$repeat_region==0,]
#
# gg=ggplot(p,aes(fill=feature,x=1) )+geom_bar(position="stack")+theme_classic()+xlim(0,2)+ggtitle("Location of CLIP Peak")+
# theme(plot.title = element_text(hjust = 0.5,size = 25),axis.title=element_text(size=20),axis.text=element_text(size=15),legend.text = element_text( size=30))
# gg
#
# # colnames(Peaksdata3)
# ```
# ```{r fig.asp=.5, fig.align='center',fig.height=6, fig.width=12,echo=F,warning=F}
# p=Peaksdata2
#
#
# gg=ggplot(p,aes(x=feature,y=Counts_All,fill=feature) )+geom_violin()+theme_classic()+geom_jitter(height = 0, width = 0.1,size=1)+
#   scale_y_continuous(trans = "log2")+geom_hline(yintercept = 100,color ='red')+
#   ggtitle('Peak Counts for Ro-Clip\nby Gene Location')+ylab("Counts - Ro-Clip")+
# theme(plot.title = element_text(hjust = 0.5,size = 25),axis.title=element_text(size=20),axis.text=element_text(size=15),legend.text = element_text( size=30))+
#     annotate('text',label=paste0('100'),x=.5,y=120,size=5)
#
# gg
#
#
#
# pp=p
# # pp[pp$Perc_Control>50,"Perc_Control"]=50
# gg=ggplot(pp,aes(x=feature,y=Perc_Control,fill=feature) )+geom_violin()+theme_classic()+geom_jitter(height = 0, width = 0.1,size=1)+
#     scale_y_continuous(trans = "log2")+geom_hline(yintercept = 10,color ='red')+
#   ggtitle('Peaks with high background in\nControl sample')+ylab("Ratio of Control counts / Ro-Clip counts\n(%- Log2 scale)")+
# theme(plot.title = element_text(hjust = 0.5,size = 25),axis.title=element_text(size=20),axis.text=element_text(size=15),legend.text = element_text( size=30))+
#     annotate('text',label=paste0('10%'),x=.7,y=12,size=5)
#
#
# gg
#
```

<br>


**3. Calculate relative location of CLIP peak in Intron or Exon**

  Use 5'most nt in CLIP peak

  Calculate position of CLIP peak in exon/intron feature.  

<br>  

```{r fig.asp=.5, fig.align='center',fig.height=6, fig.width=12,echo=F,warning=F,eval=T,include=T}
# Peaksdata2_anno$Same_Intron_start_dist

pp=Peaksdata2_anno[ ,c("ID","Geneid","strand","Counts_Unique",'Same_Intron_start_dist','Same_Exn_start_dist','Comb_biotype_exon_Oppo','Same_type_comb','Same_repeat_comb','Same_Comb_biotype_exon','Same_Comb_biotype_ncRNA','Oppo_type_comb','Oppo_repeat_comb','Oppo_Comb_biotype_exon','Oppo_Comb_biotype_ncRNA') ]
pp=pp[is.na(pp$Same_repeat_comb),]

ggplot(pp,aes(x=Same_Intron_start_dist) )+geom_density(size=2)+theme_classic()+
  ggtitle("Location of CLIP Peak in Intron")+xlab("Relative Intron Location of CLIP Peak - 5' to 3' (%)")+
  xlim(-10,110)+
theme(plot.title = element_text(hjust = 0.5,size = 25),axis.title=element_text(size=20),axis.text=element_text(size=15),legend.text = element_text( size=30))

ggplot(pp,aes(x=Same_Exn_start_dist) )+geom_density(size=2)+theme_classic()+
  ggtitle("Location of CLIP Peak in Exon")+xlab("Relative Exon Location of CLIP Peak - 5' to 3' (%)")+
  xlim(-10,110)+
theme(plot.title = element_text(hjust = 0.5,size = 25),axis.title=element_text(size=20),axis.text=element_text(size=15),legend.text = element_text( size=30))

```
<br>
<br>

```{r fig.asp=.5, fig.align='center',fig.height=6, fig.width=12,echo=F,warning=F,eval =F}

# p=Peaksdata2[Peaksdata2$strand=="+",]
# p=Peaksdata2[Peaksdata2$strand=="-",]
p=Peaksdata2
# pp=p[p$overlap_class=='intron',];pp=pp[is.na(pp[,1])==F,]
# pp=unlist(stri_split_fixed(pp$Int_jx_updist,","))
# pp=gsub(" ","",pp);pp=pp=gsub("_NA","",pp);pp=pp=gsub("NA","",pp)
# pp=as.data.frame(pp[pp!=""])
# colnames(pp)[1]="pp"
# pp=pp[is.na(pp$pp)==F,,drop=F]
# pp$pp=as.numeric(as.character(pp$pp))
# pp=cbind(matrix(nrow = nrow(pp),'intron'),pp);colnames(pp)=c("Overlap_Type",'pp')

ppx=p[p$overlap_class=='exon',];ppx=ppx[is.na(ppx[,1])==F,]
ppx=unlist(stri_split_fixed(ppx$Exn_jx_updist,","))
ppx=gsub(" ","",ppx);ppx=ppx=gsub("_NA","",ppx);ppx=ppx=gsub("NA","",ppx)
ppx=as.data.frame(ppx[ppx!=""])
colnames(ppx)[1]="pp"
ppx=ppx[is.na(ppx$pp)==F,,drop=F]
ppx$pp=as.numeric(as.character(ppx$pp))
ppx=cbind(matrix(nrow = nrow(ppx),'exon'),ppx);colnames(ppx)=c("Overlap_Type",'pp')
# pp=rbind(pp,ppx)
pp=ppx
#
# ppx=p[p$overlap_class=='exon/intron',];ppx=ppx[is.na(ppx[,1])==F,]
# # ppx=unlist(stri_split_fixed(ppx$Int_jx_updist,","))
# # ppx=gsub(" ","",ppx);ppx=ppx=gsub("_NA","",ppx);ppx=ppx=gsub("NA","",ppx)
# # ppx=as.data.frame(ppx[ppx!=""])
# # colnames(ppx)[1]="ppx"
# # ppx=ppx[is.na(ppx$ppx)==F,,drop=F]
# # ppx$ppx=as.numeric(as.character(ppx$ppx))
# # ppx=cbind(matrix(nrow = nrow(ppx),'exon/intron'),ppx);colnames(ppx)=c("Overlap_Type",'pp')
# # pp=rbind(pp,ppx)
#
# ggplot(pp,aes(x=pp,color=Overlap_Type,line=Overlap_Type) )+geom_density(size=2)+theme_classic()+
#   ggtitle("Location of CLIP Peak in Exon")+xlab("Relative Exon Location of CLIP Peak - 5' to 3' (%)")+xlim(-10,110)+
# theme(plot.title = element_text(hjust = 0.5,size = 25),axis.title=element_text(size=20),axis.text=element_text(size=15),legend.text = element_text( size=30))


```


# Output Data:  
<br>  
Location:  
sftp://Biowulf.nih.gov/data/RBL_NCI/datashare/wolin/mESC_Clip/Novoalign_UMItools  

<br>  


<!-- # Compare to previous Peak detection    -->

<!-- <br>   -->
<!-- <br>   -->

```{r ,fig.asp=.5, fig.align='center',fig.height=6, fig.width=12,echo=F,warning=F,eval=T,include=F}

Peaksdata_original=fread('CLIP_Peaks_MD4.txt', header=T, sep="\t",stringsAsFactors = F,data.table=F)
Peaksdata_original=separate(Peaksdata_original,ID,into = c('chr','start'),sep=':',remove = F)
Peaksdata_original=separate(Peaksdata_original,start,into = c('start','end'),sep='-',remove = F)
Peaksdata_original$start=as.numeric(Peaksdata_original$start)
Peaksdata_original$end=as.numeric(Peaksdata_original$end)




#################################################
iCount=fread('CLIP_Peaks_MD4_iCounts.txt', header=T, sep="\t",stringsAsFactors = F,data.table=F)
iCount=separate(iCount,ID,into = c('chr','start'),sep=':',remove = F)
iCount=separate(iCount,start,into = c('start','end'),sep='-',remove = F)
iCount$start=as.numeric(iCount$start)
iCount$end=as.numeric(iCount$end)

#######
iClip_range=fread('CLIP_Peaks_MD4_iCounts_locations_range.txt', header=T, sep="\t",stringsAsFactors = F,data.table=F)
iClip_range$V1=paste0('chr',iClip_range$V1)

colnames(iClip_range)[colnames(iClip_range)%in%'V2']='start_range'
colnames(iClip_range)[colnames(iClip_range)%in%'V3']='end_range'



iCount=merge(iCount,iClip_range,by.x='ID',by.y='V1')

```






```{r ,echo=F,warning=F,warning=T,eval=T,include=F}

# Peaksdata4=fread('CLIP_Peaks_MD10.txt', header=T, sep="\t",stringsAsFactors = F,data.table=F)
Peaksdata4=separate(Peaksdata4,ID,into = c('chr','start'),sep=':',remove = F)
Peaksdata4=separate(Peaksdata4,start,into = c('start','end'),sep='-',remove = F)
Peaksdata4$start=as.numeric(Peaksdata4$start)
Peaksdata4$end=as.numeric(Peaksdata4$end)



Peaksdata4.GR = GRanges(seqnames = as.character(Peaksdata4$chr), ranges=IRanges(start = as.numeric(Peaksdata4$start), end = as.numeric(Peaksdata4$end)),strand = Peaksdata4$strand,
                   ID=Peaksdata4$ID,
                   Same_Host_gene_name=Peaksdata4$`Same Strand: Host_gene_name`,
                   Same_Host_gene_ensembl_id=Peaksdata4$`Same Strand: Host_gene_ensembl_id`,
                    counts_Ro=Peaksdata4$Counts_Unique)

```

```{r fig.asp=.5, fig.align='center',fig.height=6, fig.width=12,echo=F,warning=F,eval=T,include=F}

iCount_clust2_range=fread('CLIP_Peaks_MD4_iCounts_locations_range.txt', header=T, sep="\t",stringsAsFactors = F,data.table=F)
colnames(iCount_clust2_range)=c('ID','Start_range','End_range','chr','ID_range')
iCount_clust2_range$ID=paste0('chr',iCount_clust2_range$ID)
iCount_clust2_range=separate(iCount_clust2_range,col = ID,into =c( 'chr'),sep = ':',remove = F)
iCount_clust2_range$Start_range=iCount_clust2_range$Start_range-0
iCount_clust2_range$End_rangeiCount_clust2_range$End_range+0
iCount_clust2_range$ID_range =paste0(iCount_clust2_range$chr,':',iCount_clust2_range$Start_range,'-',iCount_clust2_range$End_range)

iCount_clust2_adj=merge(iCount,iCount_clust2_range[c('ID','Start_range','End_range','ID_range')],by='ID')
iCount_clust2_adj=iCount_clust2_adj[duplicated(iCount_clust2_adj$ID)==F,]

```
```{r fig.asp=.5, fig.align='center',fig.height=6, fig.width=12,echo=F,warning=F,eval=T,include=F}
iCount_clust=iCount_clust2_adj
iCount_clust.GR = GRanges(seqnames = as.character(iCount_clust$chr), ranges=IRanges(start = as.numeric(iCount_clust$start)-0, end = as.numeric(iCount_clust$end)+0),strand = iCount_clust$strand,
                   ID=iCount_clust$ID,
                   count=iCount_clust$counts_Ro)


        q = Peaksdata4.GR
        s = iCount_clust.GR
        xo=as.data.frame(GenomicRanges::findOverlaps(q,s,type = "any",ignore.strand=T))
          
        qh=as.data.frame(q[xo$queryHits],row.names = NULL)
          colnames(qh)=paste0(colnames(qh),"_peaks")
        sh=as.data.frame(s[xo$subjectHits],row.names = NULL)
          colnames(sh)=paste0(colnames(sh),"_iCount")

        peaks_iCount=cbind(qh,sh)
        
Peaksdata4$iCount_peak=0
# Peaksdata4[duplicated(Peaksdata4),]

Peaksdata4[Peaksdata4$ID%in%unique(peaks_iCount$ID_peaks),'iCount_peak']=1
# sum(Peaksdata4$iCount_peak)

        
iCount_clust$NovoCount=0
iCount_clust[iCount_clust$ID%in%qh$ID_iCounr==T,"NovoCount"]=1
# iCount_clust=merge(iCount_clust,rpeatinfo[,c('ID','width_repeat')],by='ID',all.x=T)
iCount_clust3=merge(iCount_clust,peaks_iCount[,c('ID_peaks','ID_iCount')],all.x=T,by.x='ID',by.y='ID_iCount')
  
Clip_Comb=merge(Peaksdata4,iCount_clust3,by.x='ID',by.y='ID_peaks',all=T,suffix=c('_Custom','_iCount'))
Clip_Comb=Clip_Comb[!duplicated(Clip_Comb),]

iCount_only=iCount_clust[!iCount_clust.GR$ID%in%peaks_iCount$ID_iCount,]
Peaks_only=Peaksdata4[!Peaksdata4$ID%in%peaks_iCount$ID_peaks,]
intersect=peaks_iCount
        


```







```{r  fig.align='center' ,echo=F,warning=F,eval=T,eval=T,include=F}
#

Clip_Comb_venn=Clip_Comb
Clip_Comb_venn$ID_ALL=paste0(Clip_Comb_venn$ID,'_',Clip_Comb_venn$ID_iCount)

a=Clip_Comb_venn[is.na(Clip_Comb_venn$ID)==F,'ID_ALL']
  a=a[is.na(a)==F]
  a=unique(a)
b=Clip_Comb_venn[is.na(Clip_Comb_venn$ID_iCount)==F,'ID_ALL']
  b=b[is.na(b)==F]
    b=unique(b)



OVLG_all=unique(c(a,b))

  OVLG_all=OVLG_all[is.na(OVLG_all)==F]

a=OVLG_all%in%a
b=OVLG_all%in%b

C3=cbind(a,b)

rownames(C3)=OVLG_all;colnames(C3)=c('Custom Peak Calling(Novoalign)','ICount peaks')

{vennDiagram(vennCounts(C3),
            circle.col = c("red", "blue", "green3",'orange'),
            cex = 1)
remove(a,b,c,OVLG_all,C3)
  title('Compare Peak calling methods')}



iCount_only=iCount_clust2_adj[!iCount_clust2_adj$ID%in%peaks_iCount$ID_iCount,]
Peaks_only=Peaksdata4[!Peaksdata4$ID%in%unique(peaks_iCount$ID_peaks),]
intersect=iCount_clust2_adj[iCount_clust2_adj$ID%in%peaks_iCount$ID_iCount,]
```


```{r fig.asp=.5, fig.align='center',fig.height=3, fig.width=6,echo=F,warning=F,eval=T,include=F}

iCount$Peak_width=abs(iCount$end_range-iCount$start_range)
# Peaksdata_original$Peak_width=abs(Peaksdata_original$end-Peaksdata_original$start)
Peaksdata2$Peak_width=abs(as.numeric(Peaksdata2$end)-as.numeric(Peaksdata2$start))


Peaksdata2$method='NovoAlign_UMItools'
# Peaksdata_original$method='NovoAlign_old'
iCount$method='iCount'


# CombMethod=rbind(Peaksdata2[,c('method','Peak_width')],Peaksdata_original[,c('method','Peak_width')],iCount[,c('method','Peak_width')])
CombMethod=rbind(Peaksdata2[,c('method','Peak_width')],iCount[,c('method','Peak_width')])

ggplot(CombMethod,aes(x=Peak_width,line=method,color=method) )+geom_density(size=2)+theme_classic()+
  ggtitle("Peak Width")+xlab("Width (nt)")+xlim(-10,110)+
theme(plot.title = element_text(hjust = 0.5,size = 25),axis.title=element_text(size=20),axis.text=element_text(size=15),legend.text = element_text( size=30))

```


```{r ,fig.asp=.5, fig.align='center',fig.height=6, fig.width=12,echo=F,warning=F,eval=T,include=F}
Peaksdata4=Peaksdata4[,colnames(Peaksdata4)[!colnames(Peaksdata4)%in%c('chr','start','end')]]

write.table(Peaksdata4,file=("CLIP_Peaks_MD10.txt"), sep = "\t", row.names = FALSE, col.names = T, append = F, quote= FALSE,na = "")
# colnames(Peaksdata4)

```

## Motif Discovery - MEME   


```{r ,fig.asp=.5, fig.align='center',fig.height=6, fig.width=12,echo=F,warning=F,eval=T,include=T}

Peaksdata4_motif=separate(Peaksdata4,ID,into = c('chr','start'),sep = ":",remove = F)
Peaksdata4_motif=separate(Peaksdata4_motif,start,into = c('start','end'),sep = "-")
Peaksdata4_motif$start=as.integer(Peaksdata4_motif$start)
Peaksdata4_motif$end=as.integer(Peaksdata4_motif$end)
Peaksdata4_motif$width=Peaksdata4_motif$end-Peaksdata4_motif$start
Peaksdata4_motif=Peaksdata4_motif[Peaksdata4_motif$`Both Strand: RNA_type_Classification`%in%c('protein_coding: exon','protein_coding: Intron','no Feature','lncRNA'),]
# unique(Peaksdata4_motif$`Both Strand: RNA_type_Classification`)

gg=ggplot(Peaksdata4_motif,aes(x=width,line=`Both Strand: RNA_type_Classification`,color=`Both Strand: RNA_type_Classification`))+geom_density(size=1) + theme_classic()+ylab("Peak width (nt)")+xlab("Density")+ggtitle("Width of CLIP peak")+scale_color_brewer(palette = "Dark2")+
geom_vline(xintercept = (40),color="black")+annotate('text',x=50,y=4.4,label="Peak Width\n40nt")+
  theme(plot.title = element_text(hjust = 0.5,size = 25),axis.title=element_text(size=15),axis.text=element_text(size=15),legend.text = element_text( size=10))+
scale_x_continuous(trans = "log10")#+theme(legend.position = c(0.8, 0.5),legend.key.size = 2)
gg
# ggplotly(gg,tooltip = c("x", "y"))
# ggplotly(gg)


```

```{r ,fig.asp=.5, fig.align='center',fig.height=6, fig.width=12,echo=F,warning=F,eval=T,include=F}
## create bed files for Meme

source("Peak_fa_bed.R")

Peak_fa_bed(Peaksdata4)


```





```{r ,fig.asp=.5, fig.align='center',fig.height=6, fig.width=12,echo=F,warning=F,eval=F,include=F}
#### calc GC content of whole genome
library(BSgenome)
library(BSgenome.Mmusculus.UCSC.mm10)
library(TTR)
remove('dna','xxx')
remove(dna,xxx)
dna=as(as.list(Mmusculus),'DNAStringSet')
   dna=dna[c(paste0('chr',1:19), "chrX", "chrY")]
   # dna=dna[c("chr1")]
        window=99
        xxx = lapply(dna, function(x) {
        gc = rowSums(Biostrings::letterFrequencyInSlidingView(x, window, c("G", "C"), as.prob = TRUE), na.rm = TRUE)
        })
        
# remove(dna)
# dna=as(as.list(Mmusculus),'DNAStringSet')
#    dna=dna[c(paste0('chr',1:19), "chrX", "chrY")]
#    # dna=dna[c("chr1")]
#         window=25
#         xxx25 = lapply(dna, function(x) {
#         gc = rowSums(Biostrings::letterFrequencyInSlidingView(x, window, c("G", "C"), as.prob = TRUE), na.rm = TRUE)
#         })   
#         
#         
#    # dna=dna[c("chr1")]
#         window=50
#         xxx50 = lapply(dna, function(x) {
#         gc = rowSums(Biostrings::letterFrequencyInSlidingView(x, window, c("G", "C"), as.prob = TRUE), na.rm = TRUE)
#         })        
```

```{r ,fig.asp=.5, fig.align='center',fig.height=6, fig.width=12,echo=F,warning=F,eval=T,include=F}
######################################################################################
####### using sliding winow output do identify structure directlu upstream and downstream
#####################################################################################

fasta='../../../../Resources/ref/mm10/Gencode_VM23/fromGencode/GRCm38.primary_assembly.genome.fa'

Peaksdata4_motif2=Peaksdata4[Peaksdata4$`Both Strand: RNA_type_Classification`%in%c('protein_coding: exon','protein_coding: Intron'),]
Peaksdata4_motif2=Peaksdata4[Peaksdata4$`Both Strand: RNA_type_Classification`%in%c('protein_coding: Intron'),]
# unique(Peaksdata4_motif$`Both Strand: RNA_type_Classification`)
Peaksdata4_motif2=separate(Peaksdata4_motif2,ID,into = c('chr','start'),sep=':',remove = F)
Peaksdata4_motif2=separate(Peaksdata4_motif2,start,into = c('start','end'),sep='-')
Peaksdata4_motif2$start=as.numeric(Peaksdata4_motif2$start)
Peaksdata4_motif2$end=as.numeric(Peaksdata4_motif2$end)
# unique(Peaksdata4_motif2$chr)

####################################################
### function to calculate Gaccess for peaks
####################################################
# input=Peaksdata4_motif2[,,drop=F]
# l= lapply(split(input, row.names(input)), unlist)
# p=l
# p=as.data.frame(t(p[[1]]))
figure=F

for (w in c(25,50,100)) {
  
CalcFunction=function(p){
   p=as.data.frame(t(p))
  
  fasta='../../../../Resources/ref/mm10/Gencode_VM23/fromGencode/GRCm38.primary_assembly.genome.fa'
# GCout=xxx
  structout=matrix(nrow = nrow(Peaksdata4_motif2),ncol=14)
  structout=matrix(nrow = 1,ncol=14)

 
chr=as.character(p$chr)
start=as.numeric(as.character(p$start))
end=as.numeric(as.character(p$end))
strand=as.character(p$strand)

window=w
window=window-1
buffer=window+50

##################
s1=start-buffer
e1=end+(buffer)

bedpos=matrix(ncol=6,nrow=1)
bedpos[1,1]=chr
bedpos[1,2]=s1
bedpos[1,3]=e1
bedpos[1,4]=".";bedpos[1,5]="."
bedpos[1,6]=strand
write.table(bedpos, file=paste0('./structure/bed/',as.character(p$ID),"_intronexon.bed"), quote=F, sep="\t", row.names=F, col.names=F)
system(paste0('bedtools getfasta -s -fi ',fasta,' -bed structure/bed/',as.character(p$ID),'_intronexon.bed > structure/bed/',as.character(p$ID),'_intronexon.fa'))
fa=read.delim(paste0('structure/bed/',as.character(p$ID),'_intronexon.fa'),comment.char = "#",header = F,sep = "\t")
fa=as.character(fa[2,])
fa=unlist(strsplit(fa,useBytes=T,split="",fixed=T))

system(paste0('/Users/homanpj/Documents/Tools/ParasoR/src/ParasoR --input structure/bed/',as.character(p$ID),'_intronexon.fa --pre --acc --cd structure/ --window ',window,' > structure/bed/',as.character(p$ID),'.out'))

str=read.delim(paste0('structure/bed/',as.character(p$ID),'.out'),comment.char = "#",header = F,sep = "\t")
str=separate(str,V2,into = c('start','end'),sep = '-',remove = F)
str$start=as.numeric(str$start)
str$end=as.numeric(str$end)
str$V4adj=str$V4/80
getseq=function(x){paste(fa[((as.numeric(x['start'])):as.numeric(x['end']))],collapse = "")}
str$seq=apply(str,1,FUN= getseq)

######################################################################
# a=as.data.frame(GCout[[chr]])
# a=cbind(c(1:nrow(a)),a)
# # look upsream - calculation starts at nt through length of window 
# # shift nt count so that Calc ends with windo at nt of interest
# # a5=a
# # a5[,1]=a5[,1]+window
# # colnames(a5)=c("nt",'GC')
# a3=a
# a3[,1]=a3[,1]
# colnames(a3)=c("nt",'GC')
######################################################################

if (strand%in%'+') {
## Start of table is start of selected sequence
  # str$start_adj=(str$start-1)+s1
  str$start_adj=(str$start)+s1
## Adjust count so that window ends at nt position
  str$end_adj=str$start_adj+(window)

# GCstruct2=merge(a3,str,by.x='nt',by.y='start_adj')
GCstruct2=str;colnames(GCstruct2)[colnames(GCstruct2)%in%'start_adj']='nt'
# GCstruct2$GC=SMA(GCstruct2$GC,n=10)
# GCstruct2$V4adj=SMA(GCstruct2$V4adj,n=10)
# GCstruct2$V4=SMA(GCstruct2$V4,n=10)

croslink=as.numeric(as.character(p$start))
structure_upstream=GCstruct2[GCstruct2$end_adj%in%croslink,'V4']
structure_downstream=GCstruct2[GCstruct2$nt%in%croslink,'V4']
# GC_upstream=GCstruct2[GCstruct2$end_adj%in%croslink,'GC']
# GC_downstream=GCstruct2[GCstruct2$nt%in%croslink,'GC']

coord_upstream=paste0(p$chr,':',(GCstruct2[GCstruct2$end_adj%in%croslink,'nt']),'-',GCstruct2[GCstruct2$end_adj%in%croslink,'end_adj'])
coord_downstream=paste0(p$chr,':',GCstruct2[GCstruct2$nt%in%(croslink),'nt'],'-',(GCstruct2[GCstruct2$nt%in%croslink,'end_adj']))
seq_upstream=GCstruct2[GCstruct2$end_adj%in%croslink,'seq']
seq_downstream=GCstruct2[GCstruct2$nt%in%(croslink),'seq']
}

##############
if (strand%in%'-') {
## Reverse order of table (so now oriented same direction as + strand sequence)
  str=str[sort(str$start,decreasing = T),]
  str$start_opp=c(1:nrow(str))
## Start of table is start of selected sequence
## Window goes from nt possition + window downstream
  str$start_adj=(str$start_opp-1)+s1

### look downfild of nucleotide
  str$end_adj=(str$start_adj-1)+(window+1)

# GCstruct2=merge(a3,str,by.x='nt',by.y='start_adj')
GCstruct2=str;colnames(GCstruct2)[colnames(GCstruct2)%in%'start_adj']='nt'
# GCstruct2$GC=SMA(GCstruct2$GC,n=10)
# GCstruct2$V4adj=SMA(GCstruct2$V4adj,n=10)
# GCstruct2$V4=SMA(GCstruct2$V4,n=10)

croslink=as.numeric(as.character(p$end))
structure_upstream=GCstruct2[GCstruct2$nt%in%croslink,'V4']
structure_downstream=GCstruct2[GCstruct2$end_adj%in%croslink,'V4']
# GC_upstream=GCstruct2[GCstruct2$nt%in%croslink,'GC']
# GC_downstream=GCstruct2[GCstruct2$end_adj%in%croslink,'GC']

coord_upstream=paste0(p$chr,':',GCstruct2[GCstruct2$nt%in%croslink,'nt'],'-',(GCstruct2[GCstruct2$nt%in%croslink,'end_adj']))
coord_downstream=paste0(p$chr,':',GCstruct2[GCstruct2$end_adj%in%croslink,'nt'],'-',GCstruct2[GCstruct2$end_adj%in%croslink,'end_adj'])
seq_upstream=GCstruct2[GCstruct2$nt%in%(croslink-1),'seq']
seq_downstream=GCstruct2[GCstruct2$end_adj%in%(croslink-1),'seq']
# cor(GCstruct2[is.na(GCstruct2$GC)==F,'GC'],GCstruct2[is.na(GCstruct2$V4adj)==F,'V4adj'])
}

####################################
structout[,1]=as.character(p$ID)
structout[,2]=as.character(p$strand)
# structout[,3]=GC_upstream*100
structout[,4]=structure_upstream
# structout[,5]=GC_downstream*100
structout[,6]=structure_downstream
structout[,9]=coord_upstream
structout[,10]=seq_upstream
structout[,11]=coord_downstream
structout[,12]=seq_downstream
structout[,13]=GC(unlist(strsplit(seq_upstream,useBytes=T,split="",fixed=T)))
structout[,14]=GC(unlist(strsplit(seq_downstream,useBytes=T,split="",fixed=T)))

structout[,7]=as.numeric(structout[,6])-as.numeric(structout[,4])
structout[,8]=as.numeric(structout[,5])-as.numeric(structout[,3])
# {plot(GCstruct2$GC,GCstruct2$V4,type='p',pch=16,col='red',xlab='Nucleotide Position',ylab='Value')}
  
remove('GCstruct2','croslink','a','str')
if(file.exists(paste0('structure/bed/',as.character(p$ID),'.out'))){file.remove(paste0('structure/bed/',as.character(p$ID),'.out'))}
if(file.exists(paste0('structure/bed/',as.character(p$ID),'_intronexon.bed'))){file.remove(paste0('structure/bed/',as.character(p$ID),'_intronexon.bed'))}
if(file.exists(paste0('structure/bed/',as.character(p$ID),'_intronexon.fa'))){file.remove(paste0('structure/bed/',as.character(p$ID),'_intronexon.fa'))}


return(structout[1,])
}

##################################################################################################################################
##################################################################################################################################

input=Peaksdata4_motif2[,,drop=F]
l= lapply(split(input, row.names(input)), unlist)

system.time({outmcl=mclapply(l,CalcFunction,mc.cores=8)}  )

outmcl=outmcl
 out=as.data.frame(matrix(nrow = 1,ncol = 15))

for (x in 1:length(outmcl)) {
 out= rbind(out,outmcl[[x]])
}
colnames(out)=c('ID','strand','GC content: Upstream','Access Energy: Upstream','GC content: Downstream','Access Energy: Downstream','Structure Difference','GC difference','coordinates_upstream','sequence_upstream','coordinates_Downstream','sequence_Downstream','GC_Upstream','GC_Downstream','ID2' )
out=out[is.na(out)==F,]
out=as.data.frame(out)
out[,3]=as.numeric(out[,3])
out[,4]=as.numeric(out[,4])
out[,5]=as.numeric(out[,5])
out[,6]=as.numeric(out[,6])
out[,7]=as.numeric(out[,7])
out[,8]=as.numeric(out[,8])
out[,13]=as.numeric(out[,13])
out[,14]=as.numeric(out[,14])
write.table(out, file=paste0('./structure/bed/structure_',w,'.txt'), quote=F, sep="\t", row.names=F, col.names=T)

}

```

```{r ,fig.asp=.5, fig.align='center',fig.height=6, fig.width=12,echo=F,warning=F,eval=F,include=F}
######################################################################################
####### check sliding window with exact selection of window (slower)
#####################################################################################


fasta='../../../../Resources/ref/mm10/Gencode_VM23/fromGencode/GRCm38.primary_assembly.genome.fa'

Peaksdata4_motif2=Peaksdata4[Peaksdata4$`Both Strand: RNA_type_Classification`%in%c('protein_coding: exon','protein_coding: Intron'),]
# unique(Peaksdata4_motif$`Both Strand: RNA_type_Classification`)
Peaksdata4_motif2=separate(Peaksdata4_motif2,ID,into = c('chr','start'),sep=':',remove = F)
Peaksdata4_motif2=separate(Peaksdata4_motif2,start,into = c('start','end'),sep='-')
Peaksdata4_motif2$start=as.numeric(Peaksdata4_motif2$start)
Peaksdata4_motif2$end=as.numeric(Peaksdata4_motif2$end)
# unique(Peaksdata4_motif2$chr)

####################################################
### function to calculate Gaccess for peaks
####################################################
# input=Peaksdata4_motif2[,,drop=F]
# l= lapply(split(input, row.names(input)), unlist)
# p=l
# p=as.data.frame(t(p[[1]]))
figure=F

CalcFunction=function(p){
  p=as.data.frame(t(p))
  
  fasta='../../../../Resources/ref/mm10/Gencode_VM23/fromGencode/GRCm38.primary_assembly.genome.fa'
  # GCout=xxx
  structout=matrix(nrow = nrow(Peaksdata4_motif2),ncol=14)
  structout=matrix(nrow = 1,ncol=14)
  
chr=as.character(p$chr)
start=as.numeric(as.character(p$start))
end=as.numeric(as.character(p$end))
strand=as.character(p$strand)

window=100
window=window-1
buffer=window+1

######################################################################
# a=as.data.frame(GCout[[chr]])
# a=cbind(c(1:nrow(a)),a)
# 
# a3=a
# # a3[,1]=a3[,1]
# colnames(a3)=c("nt",'GC')
######################################################################

if (strand%in%'+') {
e1=start
s1=start-buffer

bedpos=matrix(ncol=6,nrow=1)
bedpos[1,1]=chr
bedpos[1,2]=s1
bedpos[1,3]=e1
bedpos[1,4]=".";bedpos[1,5]="."
bedpos[1,6]=strand
write.table(bedpos, file=paste0('./structure/bed/',as.character(p$ID),"_intronexon.bed"), quote=F, sep="\t", row.names=F, col.names=F)
system(paste0('bedtools getfasta -s -fi ',fasta,' -bed structure/bed/',as.character(p$ID),'_intronexon.bed > structure/bed/',as.character(p$ID),'_intronexon.fa'))
fa=read.delim(paste0('structure/bed/',as.character(p$ID),'_intronexon.fa'),comment.char = "#",header = F,sep = "\t")
fa=as.character(fa[2,])
# fa=unlist(strsplit(fa,useBytes=T,split="",fixed=T))

system(paste0('/Users/homanpj/Documents/Tools/ParasoR/src/ParasoR --input structure/bed/',as.character(p$ID),'_intronexon.fa --pre --acc --cd structure/ --window ',window,' > structure/bed/',as.character(p$ID),'.out'))

str2=read.delim(paste0('structure/bed/',as.character(p$ID),'.out'),comment.char = "#",header = F,sep = "\t")
str2=separate(str2,V2,into = c('start','end'),sep = '-',remove = F)
str2$start=as.numeric(str2$start)
str2$end=as.numeric(str2$end)
str2$V4adj=str2$V4/80
str2$seq=fa

## Start of table is start of selected sequence
  str2$start_adj=(str2$start)+s1
## Adjust count so that window ends at nt position
  str2$end_adj=str2$start_adj+window

# GCstruct2=merge(a3,str,by.x='nt',by.y='start_adj')
GCstruct2=str2;colnames(GCstruct2)[colnames(GCstruct2)%in%'start_adj']='nt'

croslink=as.numeric(as.character(p$start))
structure_upstream=GCstruct2[GCstruct2$end_adj%in%croslink,'V4']
# GC_upstream=GCstruct2[GCstruct2$end_adj%in%croslink,'GC']
coord_upstream=paste0(p$chr,':',(GCstruct2[GCstruct2$end_adj%in%croslink,'nt']),'-',GCstruct2[GCstruct2$end_adj%in%croslink,'end_adj'])
seq_upstream=GCstruct2[GCstruct2$end_adj%in%croslink,'seq']

# structure_downstream=GCstruct2[GCstruct2$nt%in%croslink,'V4']
# GC_downstream=GCstruct2[GCstruct2$nt%in%croslink,'GC']
# coord_downstream=paste0(p$chr,':',GCstruct2[GCstruct2$nt%in%(croslink),'nt'],'-',(GCstruct2[GCstruct2$nt%in%croslink,'end_adj']-1))
# seq_downstream=GCstruct2[GCstruct2$nt%in%(croslink-1),'seq']

}
##############
if (strand%in%'-') {
s1=end-1
e1=end+buffer-1

bedpos=matrix(ncol=6,nrow=1)
bedpos[1,1]=chr
bedpos[1,2]=s1
bedpos[1,3]=e1
bedpos[1,4]=".";bedpos[1,5]="."
bedpos[1,6]=strand
write.table(bedpos, file=paste0('./structure/bed/',as.character(p$ID),"_intronexon.bed"), quote=F, sep="\t", row.names=F, col.names=F)
system(paste0('bedtools getfasta -s -fi ',fasta,' -bed structure/bed/',as.character(p$ID),'_intronexon.bed > structure/bed/',as.character(p$ID),'_intronexon.fa'))
fa=read.delim(paste0('structure/bed/',as.character(p$ID),'_intronexon.fa'),comment.char = "#",header = F,sep = "\t")
fa=as.character(fa[2,])
# fa=unlist(strsplit(fa,useBytes=T,split="",fixed=T))

system(paste0('/Users/homanpj/Documents/Tools/ParasoR/src/ParasoR --input structure/bed/',as.character(p$ID),'_intronexon.fa --pre --acc --cd structure/ --window ',window,' > structure/bed/',as.character(p$ID),'.out'))

str2=read.delim(paste0('structure/bed/',as.character(p$ID),'.out'),comment.char = "#",header = F,sep = "\t")
str2=separate(str2,V2,into = c('start','end'),sep = '-',remove = F)
str2$start=as.numeric(str2$start)
str2$end=as.numeric(str2$end)
str2$V4adj=str2$V4/80
str2$seq=fa

## Reverse order of table (so now oriented same direction as + strand sequence)
  str2=str2[sort(str2$start,decreasing = T),]
  str2$start_opp=c(1:nrow(str2))
## Start of table is start of selected sequence
## Window goes from nt possition + window downstream
  str2$start_adj=(str2$start_opp)+(s1)

### look downfild of nucleotide
  str2$end_adj=(str2$start_adj)+window

# GCstruct2=merge(a3,str2,by.x='nt',by.y='start_adj')
GCstruct2=str2;colnames(GCstruct2)[colnames(GCstruct2)%in%'start_adj']='nt'

croslink=as.numeric(as.character(p$end))
structure_upstream=GCstruct2[GCstruct2$nt%in%croslink,'V4']
# GC_upstream=GCstruct2[GCstruct2$nt%in%croslink,'GC']
coord_upstream=paste0(p$chr,':',GCstruct2[GCstruct2$nt%in%croslink,'nt'],'-',(GCstruct2[GCstruct2$nt%in%croslink,'end_adj']))
seq_upstream=GCstruct2[GCstruct2$nt%in%(croslink),'seq']

# structure_downstream=GCstruct2[GCstruct2$end_adj%in%croslink,'V4']
# GC_downstream=GCstruct2[GCstruct2$end_adj%in%croslink,'GC']
# coord_downstream=paste0(p$chr,':',GCstruct2[GCstruct2$end_adj%in%croslink,'nt'],'-',GCstruct2[GCstruct2$end_adj%in%croslink,'end_adj'])
# seq_downstream=GCstruct2[GCstruct2$end_adj%in%(croslink-1),'seq']

}


####################################
structout[,1]=as.character(p$ID)
structout[,2]=as.character(p$strand)
# structout[,3]=GC_upstream*100
structout[,4]=structure_upstream
# structout[,5]=GC_downstream*100
# structout[,6]=structure_downstream
structout[,9]=coord_upstream
structout[,10]=seq_upstream
# structout[,11]=coord_downstream
# structout[,12]=seq_downstream
structout[,13]=GC(unlist(strsplit(seq_upstream,useBytes=T,split="",fixed=T)))
# structout[,14]=GC(unlist(strsplit(seq_downstream,useBytes=T,split="",fixed=T)))


structout[,7]=as.numeric(structout[,6])-as.numeric(structout[,4])
structout[,8]=as.numeric(structout[,5])-as.numeric(structout[,3])
# {plot(GCstruct2$GC,GCstruct2$V4,type='p',pch=16,col='red',xlab='Nucleotide Position',ylab='Value')}
  
remove('GCstruct2','croslink','a','str2')
if(file.exists(paste0('structure/bed/',as.character(p$ID),'.out'))){file.remove(paste0('structure/bed/',as.character(p$ID),'.out'))}
if(file.exists(paste0('structure/bed/',as.character(p$ID),'_intronexon.bed'))){file.remove(paste0('structure/bed/',as.character(p$ID),'_intronexon.bed'))}
if(file.exists(paste0('structure/bed/',as.character(p$ID),'_intronexon.fa'))){file.remove(paste0('structure/bed/',as.character(p$ID),'_intronexon.fa'))}


return(structout[1,])
}


input=Peaksdata4_motif2[1,,drop=F]
l= lapply(split(input, row.names(input)), unlist)

system.time({outmcl2=mclapply(l,CalcFunction,mc.cores=8)}  )
```

```{r ,fig.asp=.5, fig.align='center',fig.height=6, fig.width=12,echo=F,warning=F,eval=F,include=F}
###############################################################
### process output from function to calculate Gaccess for peaks
###############################################################

input=Peaksdata4_motif2[1:5,,drop=F]
l= lapply(split(input, row.names(input)), unlist)
# p=l
# p=as.data.frame(t(p[[1]]))

# system.time({outl=lapply(l,CalcFunction)})
# outl2=outl
#  out1=as.data.frame(matrix(nrow = 1,ncol = 10))
# 
# for (x in 1:length(out1)) {
#  out1= rbind(out,out1[[x]])
# }
# colnames(out1)=c('ID','strand','GC content: Upstream','Minimum free Energy: Upstream','GC content: Downstream','Minimum free Energy: Downstream','Structure Difference','GC difference','ID2' )
# out1=out1[is.na(out1)==F,]
# out1=as.data.frame(out1)
# out1[,3]=as.numeric(out1[,3])
# out1[,4]=as.numeric(out1[,4])
# out1[,5]=as.numeric(out1[,5])
# out1[,6]=as.numeric(out1[,6])
# out1[,7]=as.numeric(out1[,7])
# out1[,8]=as.numeric(out1[,8])
# write.table(out1, file='./structure/bed/structure_list.txt', quote=F, sep="\t", row.names=F, col.names=F)


#######################################################


system.time({outmcl=mclapply(l,CalcFunction,mc.cores=8)}  )

outmcl2=outmcl
 out=as.data.frame(matrix(nrow = 1,ncol = 10))

for (x in 1:length(outmcl)) {
 out= rbind(out,outmcl[[x]])
}
colnames(out)=c('ID','strand','GC content: Upstream','Minimum free Energy: Upstream','GC content: Downstream','Minimum free Energy: Downstream','Structure Difference','GC difference','sequence','ID2' )
out=out[is.na(out)==F,]
out=as.data.frame(out)
out[,3]=as.numeric(out[,3])
out[,4]=as.numeric(out[,4])
out[,5]=as.numeric(out[,5])
out[,6]=as.numeric(out[,6])
out[,7]=as.numeric(out[,7])
out[,8]=as.numeric(out[,8])
write.table(out, file='./structure/bed/structure.txt', quote=F, sep="\t", row.names=F, col.names=F)

str=read.delim(paste0('structure/bed/',p$ID,'.out'),comment.char = "#",header = F,sep = "\t")

```

```{r ,fig.asp=.5, fig.align='center',fig.height=6, fig.width=12,echo=F,warning=F,eval=F,include=F}

out=read.delim('./structure/bed/structure.txt',comment.char = "#",header = F,sep = "\t")

colnames(out)=c("ID","strand","GC content: Upstream" , paste0("Gacces: Upstream") ,"GC content: Downstream","Gacces: Downstream", "Gacces Difference\n(downstream-upstream)\n","GC difference\n(downstream-upstream)\n","ID2")
out=out[is.na(out$ID)==F,]


```


```{r ,fig.asp=.5, fig.align='center',fig.height=3, fig.width=6,echo=F,warning=F,eval=T,include=F}

###############################################################
### calculate thermodynamic z-score for peaks
###############################################################

for (window in c(25,50,100)) {
# for (window in c(25)) {

bedpos_Z=as.data.frame(cbind(Peaksdata4_motif2$chr,Peaksdata4_motif2$start,Peaksdata4_motif2$end))
bedpos_Z$V4="."
bedpos_Z$V5="."
bedpos_Z=as.data.frame(cbind(bedpos_Z,Peaksdata4_motif2$strand))
colnames(bedpos_Z)=c('chr','start','end','V4','V5','strand')
bedpos_Z$strand=as.character(bedpos_Z$strand)
bedpos_Z$chr=as.character(bedpos_Z$chr)
bedpos_Z$start=as.numeric(as.character(bedpos_Z$start))
bedpos_Z$end=as.numeric(as.character(bedpos_Z$end))

###########

bedpos_Up=bedpos_Z
bedpos_Up_p=bedpos_Up[bedpos_Up$strand%in%"+",]
bedpos_Up_n=bedpos_Up[bedpos_Up$strand%in%"-",]

bedpos_Up_p[,"end"]=bedpos_Up_p[,"start"]
bedpos_Up_p[,"start"]=bedpos_Up_p[,"end"]-window

bedpos_Up_n[,"start"]=bedpos_Up_n[,"end"]-1
bedpos_Up_n[,"end"]=bedpos_Up_n[,"start"]+window

bedpos_Up=rbind(bedpos_Up_p,bedpos_Up_n)
write.table(bedpos_Up, file=paste0('./structure/bed_RNAz/up.bed'), quote=F, sep="\t", row.names=F, col.names=F)
# system(paste0('bedtools getfasta -s -fi ',fasta,' -bed structure/bed/',p$ID,'_updn.bed > structure/bed/',p$ID,'_updn.fa'))

fapos_Up=system(paste0('bedtools getfasta -s -fi ',fasta,' -bed structure/bed_RNAz/up.bed'),intern = T)
fapos_Up=as.data.frame(fapos_Up)
fapos_Up[,1]=as.character(fapos_Up[,1])
#########

fapos_Up2=as.data.frame(cbind(fapos_Up[seq(1, nrow(fapos_Up), 2),],fapos_Up[seq(2, nrow(fapos_Up), 2),]))
fapos_Up2$V1=gsub(">","",fapos_Up2$V1)

fapos_Up2=separate(fapos_Up2,"V1",into = c('chr','start'),sep = ":",remove = F)
fapos_Up2=separate(fapos_Up2,"start",into = c('start','strand'),sep = "\\(")
fapos_Up2=separate(fapos_Up2,"start",into = c('start','end'),sep = "-")
fapos_Up2$strand=gsub("\\)","",fapos_Up2$strand)
fapos_Up2$start=as.numeric(fapos_Up2$start)
fapos_Up2$end=as.numeric(fapos_Up2$end)

mafout_Up=as.data.frame(matrix(nrow=nrow(fapos_Up2),ncol=7))
  mafout_Up[,1]="s"
  mafout_Up[,2]=fapos_Up2$V1
  mafout_Up[,3]=fapos_Up2$start
  mafout_Up[,4]=abs(fapos_Up2$end - fapos_Up2$start)
  mafout_Up[,5]= fapos_Up2$strand
  mafout_Up[,6]= fapos_Up2$end
  mafout_Up[,7]= fapos_Up2$V2

  
    mafout_Up[mafout_Up$V5%in%"+",3]= mafout_Up[mafout_Up$V5%in%"+",3]+1
    mafout_Up[mafout_Up$V5%in%"-",3]= mafout_Up[mafout_Up$V5%in%"-",3]+1
    mafout_Up[mafout_Up$V5%in%"-",6]= mafout_Up[mafout_Up$V5%in%"-",6]

############################
inter=as.data.frame(matrix(nrow=2,ncol=7))
         inter[2,1]='a'
         inter[2,2]='score=1.000000'
#Every N rows after which empty rows should be inserted
N = 2
after_rows = 300

mafout_Up2=do.call(rbind, lapply(split(mafout_Up, ceiling(1:nrow(mafout_Up)/after_rows)),
                      function(a) rbind(a, replace(a[1:N,], TRUE, inter))))

mafout_Up2=rbind(inter[2,],mafout_Up2)
mafout_Up2=mafout_Up2[1:(nrow(mafout_Up2)-2),]

write.table(mafout_Up2, file=paste0('./structure/bed_RNAz/intronexon_rNAz_up.maf'),na = "", quote=F, sep="\t", row.names=F, col.names=F)


outmaf_Up=system(paste0('/Users/homanpj/Documents/Tools/RNAz-2.1/rnaz/RNAz ./structure/bed_RNAz/intronexon_rNAz_up.maf'),intern = T)
  if(file.exists(paste0('./structure/bed_RNAz/intronexon_rNAz_up.maf'))){file.remove(paste0('./structure/bed_RNAz/intronexon_rNAz_up.maf'))}

outmaf_Up_ov=as.data.table(outmaf_Up[outmaf_Up%in%""==F])
while (isEmpty(grep("#",outmaf_Up_ov$V1))==F) {
  g=grep("#",outmaf_Up_ov$V1)
  outmaf_Up_ov=(outmaf_Up_ov[-c(g[1]:g[2]),])
    outmaf_Up_ov=(outmaf_Up_ov)
}

outmaf_Up_ov2=cbind(outmaf_Up_ov[seq(3, nrow(outmaf_Up_ov), 3),],outmaf_Up_ov[seq(2, nrow(outmaf_Up_ov), 3),],outmaf_Up_ov[seq(1, nrow(outmaf_Up_ov), 3),])
colnames(outmaf_Up_ov2)=c('str','seq','info')
outmaf_Up_ov=separate(outmaf_Up_ov2,'str',into = c('str','data'),sep = window)

outmaf_Up_ov=separate(outmaf_Up_ov,'data',into = c('MFE','Zscore','direction'),sep = ', ')
outmaf_Up_ov$MFE=as.numeric(gsub("\\( ","",outmaf_Up_ov$MFE))
outmaf_Up_ov$Zscore=as.numeric(gsub("z-score = ","",outmaf_Up_ov$Zscore))

outmaf_Up_ov=separate(outmaf_Up_ov,'info',into = c('name','start','length','strand','end'),sep = ' ')
outmaf_Up_ov$start=as.numeric(outmaf_Up_ov$start)
outmaf_Up_ov$end=as.numeric(outmaf_Up_ov$end)

outmaf_Up_ov=outmaf_Up_ov[!is.na(outmaf_Up_ov$MFE),]
outmaf_Up_ov=cbind(outmaf_Up_ov,c(1:nrow(outmaf_Up_ov)))

write.table(outmaf_Up_ov, file=paste0('./structure/bed/Zscore_upstream_',window,'.txt'), quote=F, sep="\t", row.names=F, col.names=T)


########################################################################
  ########################################################################
  

bedpos_Dn=bedpos_Z
bedpos_Dn_p=bedpos_Dn[bedpos_Dn$strand%in%"+",]
bedpos_Dn_n=bedpos_Dn[bedpos_Dn$strand%in%"-",]

bedpos_Dn_p[,"start"]=bedpos_Dn_p[,"start"]-1
bedpos_Dn_p[,"end"]=bedpos_Dn_p[,"start"]+window

bedpos_Dn_n[,"end"]=bedpos_Dn_n[,"end"]
bedpos_Dn_n[,"start"]=bedpos_Dn_n[,"end"]-window

bedpos_Dn=rbind(bedpos_Dn_p,bedpos_Dn_n)
write.table(bedpos_Dn, file=paste0('./structure/bed_RNAz/dn.bed'), quote=F, sep="\t", row.names=F, col.names=F)

fapos_Dn=system(paste0('bedtools getfasta -s -fi ',fasta,' -bed structure/bed_RNAz/dn.bed'),intern = T)
fapos_Dn=as.data.frame(fapos_Dn)
fapos_Dn[,1]=as.character(fapos_Dn[,1])
#########

fapos_Dn2=as.data.frame(cbind(fapos_Dn[seq(1, nrow(fapos_Dn), 2),],fapos_Dn[seq(2, nrow(fapos_Dn), 2),]))
fapos_Dn2$V1=gsub(">","",fapos_Dn2$V1)

fapos_Dn2=separate(fapos_Dn2,"V1",into = c('chr','start'),sep = ":",remove = F)
fapos_Dn2=separate(fapos_Dn2,"start",into = c('start','strand'),sep = "\\(")
fapos_Dn2=separate(fapos_Dn2,"start",into = c('start','end'),sep = "-")
fapos_Dn2$strand=gsub("\\)","",fapos_Dn2$strand)
fapos_Dn2$start=as.numeric(fapos_Dn2$start)
fapos_Dn2$end=as.numeric(fapos_Dn2$end)

mafout_Dn=as.data.frame(matrix(nrow=nrow(fapos_Dn2),ncol=7))
  mafout_Dn[,1]="s"
  mafout_Dn[,2]=fapos_Dn2$V1
  mafout_Dn[,3]=fapos_Dn2$start
  mafout_Dn[,4]=abs(fapos_Dn2$end - fapos_Dn2$start)
  mafout_Dn[,5]= fapos_Dn2$strand
  mafout_Dn[,6]= fapos_Dn2$end
  mafout_Dn[,7]= fapos_Dn2$V2

  
    mafout_Dn[mafout_Dn$V5%in%"+",3]= mafout_Dn[mafout_Dn$V5%in%"+",3]+1
    mafout_Dn[mafout_Dn$V5%in%"-",3]= mafout_Dn[mafout_Dn$V5%in%"-",3]+1
    
########################################################################
  ########################################################################
    
 # mafout_UpDn=rbind(mafout_Up,mafout_Dn)
############################
inter=as.data.frame(matrix(nrow=2,ncol=7))
         inter[2,1]='a'
         inter[2,2]='score=1.000000'
#Every N rows after which empty rows should be inserted
N = 2
after_rows = 300

mafout_Dn2=do.call(rbind, lapply(split(mafout_Dn, ceiling(1:nrow(mafout_Dn)/after_rows)),
                      function(a) rbind(a, replace(a[1:N,], TRUE, inter))))

mafout_Dn2=rbind(inter[2,],mafout_Dn2)
mafout_Dn2=mafout_Dn2[1:(nrow(mafout_Dn2)-2),]

write.table(mafout_Dn2, file=paste0('./structure/bed_RNAz/intronexon_rNAz_Dn.maf'),na = "", quote=F, sep="\t", row.names=F, col.names=F)


outmaf_Dn=system(paste0('/Users/homanpj/Documents/Tools/RNAz-2.1/rnaz/RNAz ./structure/bed_RNAz/intronexon_rNAz_Dn.maf'),intern = T)
  if(file.exists(paste0('./structure/bed_RNAz/intronexon_rNAz_Dn.maf'))){file.remove(paste0('./structure/bed_RNAz/intronexon_rNAz_Dn.maf'))}

outmaf_Dn_ov=as.data.table(outmaf_Dn[outmaf_Dn%in%""==F])
while (isEmpty(grep("#",outmaf_Dn_ov$V1))==F) {
  g=grep("#",outmaf_Dn_ov$V1)
  outmaf_Dn_ov=(outmaf_Dn_ov[-c(g[1]:g[2]),])
    outmaf_Dn_ov=(outmaf_Dn_ov)
}

outmaf_Dn_ov2=cbind(outmaf_Dn_ov[seq(3, nrow(outmaf_Dn_ov), 3),],outmaf_Dn_ov[seq(2, nrow(outmaf_Dn_ov), 3),],outmaf_Dn_ov[seq(1, nrow(outmaf_Dn_ov), 3),])
colnames(outmaf_Dn_ov2)=c('str','seq','info')
outmaf_Dn_ov=separate(outmaf_Dn_ov2,'str',into = c('str','data'),sep = window)

outmaf_Dn_ov=separate(outmaf_Dn_ov,'data',into = c('MFE','Zscore','direction'),sep = ', ')
outmaf_Dn_ov$MFE=as.numeric(gsub("\\( ","",outmaf_Dn_ov$MFE))
outmaf_Dn_ov$Zscore=as.numeric(gsub("z-score = ","",outmaf_Dn_ov$Zscore))

outmaf_Dn_ov=separate(outmaf_Dn_ov,'info',into = c('name','start','length','strand','end'),sep = ' ')
outmaf_Dn_ov$start=as.numeric(outmaf_Dn_ov$start)
outmaf_Dn_ov$end=as.numeric(outmaf_Dn_ov$end)



outmaf_Dn_ov=outmaf_Dn_ov[!is.na(outmaf_Dn_ov$MFE),]
outmaf_Dn_ov=cbind(outmaf_Dn_ov,c(1:nrow(outmaf_Dn_ov)))
  
  
# outmaf_ov$MFE=SMA(outmaf_ov$MFE,n=10)
# outmaf_ov$Zscore=SMA(outmaf_ov$Zscore,n=10)


write.table(outmaf_Dn_ov, file=paste0('./structure/bed/Zscore_Downstream_',window,'.txt'), quote=F, sep="\t", row.names=F, col.names=T)
}
```

```{r ,fig.asp=1, fig.align='center',fig.height=3, fig.width=3,echo=F,warning=F,eval=T,include=F}

z25_upstream=read.delim('./structure/bed/Zscore_upstream_25.txt',comment.char = "#",header = T,sep = "\t")
z50_upstream=read.delim('./structure/bed/Zscore_upstream_50.txt',comment.char = "#",header = T,sep = "\t")
z100_upstream=read.delim('./structure/bed/Zscore_upstream_100.txt',comment.char = "#",header = T,sep = "\t")
z25_downstream=read.delim('./structure/bed/Zscore_downstream_25.txt',comment.char = "#",header = T,sep = "\t")
z50_downstream=read.delim('./structure/bed/Zscore_downstream_50.txt',comment.char = "#",header = T,sep = "\t")
z100_downstream=read.delim('./structure/bed/Zscore_downstream_100.txt',comment.char = "#",header = T,sep = "\t")

z25_upstream[z25_upstream$strand%in%"+",'crossslink']=z25_upstream[z25_upstream$strand%in%"+",'end']
z25_upstream[z25_upstream$strand%in%"-",'crossslink']=z25_upstream[z25_upstream$strand%in%"-",'start']
z50_upstream[z50_upstream$strand%in%"+",'crossslink']=z50_upstream[z50_upstream$strand%in%"+",'end']
z50_upstream[z50_upstream$strand%in%"-",'crossslink']=z50_upstream[z50_upstream$strand%in%"-",'start']
z100_upstream[z100_upstream$strand%in%"+",'crossslink']=z100_upstream[z100_upstream$strand%in%"+",'end']
z100_upstream[z100_upstream$strand%in%"-",'crossslink']=z100_upstream[z100_upstream$strand%in%"-",'start']

z25_downstream[z25_downstream$strand%in%"+",'crossslink']=z25_downstream[z25_downstream$strand%in%"+",'start']
z25_downstream[z25_downstream$strand%in%"-",'crossslink']=z25_downstream[z25_downstream$strand%in%"-",'end']
z50_downstream[z50_downstream$strand%in%"+",'crossslink']=z50_downstream[z50_downstream$strand%in%"+",'start']
z50_downstream[z50_downstream$strand%in%"-",'crossslink']=z50_downstream[z50_downstream$strand%in%"-",'end']
z100_downstream[z100_downstream$strand%in%"+",'crossslink']=z100_downstream[z100_downstream$strand%in%"+",'start']
z100_downstream[z100_downstream$strand%in%"-",'crossslink']=z100_downstream[z100_downstream$strand%in%"-",'end']
  
z25_upstream=separate(z25_upstream,'name',into = c('chr','name'),sep=":",remove=T);z25_upstream$name=paste0(z25_upstream$chr,":",z25_upstream$name)
z50_upstream=separate(z50_upstream,'name',into = c('chr','name'),sep=":",remove=T);z50_upstream$name=paste0(z50_upstream$chr,":",z50_upstream$name)
z100_upstream=separate(z100_upstream,'name',into = c('chr','name'),sep=":",remove=T);z100_upstream$name=paste0(z100_upstream$chr,":",z100_upstream$name)
z25_downstream=separate(z25_downstream,'name',into = c('chr','name'),sep=":",remove=T);z25_downstream$name=paste0(z25_downstream$chr,":",z25_downstream$name)
z50_downstream=separate(z50_downstream,'name',into = c('chr','name'),sep=":",remove=T);z50_downstream$name=paste0(z50_downstream$chr,":",z50_downstream$name)
z100_downstream=separate(z100_downstream,'name',into = c('chr','name'),sep=":",remove=T);z100_downstream$name=paste0(z100_downstream$chr,":",z100_downstream$name)
z25_upstream$coordinates=paste0(z25_upstream$chr,":",z25_upstream$start,"-",z25_upstream$end);z25_upstream$coordinates=gsub(">","",z25_upstream$coordinates)
z50_upstream$coordinates=paste0(z50_upstream$chr,":",z50_upstream$start,"-",z50_upstream$end);z50_upstream$coordinates=gsub(">","",z50_upstream$coordinates)
z100_upstream$coordinates=paste0(z100_upstream$chr,":",z100_upstream$start,"-",z100_upstream$end);z100_upstream$coordinates=gsub(">","",z100_upstream$coordinates)
z25_downstream$coordinates=paste0(z25_downstream$chr,":",z25_downstream$start,"-",z25_downstream$end);z25_downstream$coordinates=gsub(">","",z25_downstream$coordinates)
z50_downstream$coordinates=paste0(z50_downstream$chr,":",z50_downstream$start,"-",z50_downstream$end);z50_downstream$coordinates=gsub(">","",z50_downstream$coordinates)
z100_downstream$coordinates=paste0(z100_downstream$chr,":",z100_downstream$start,"-",z100_downstream$end);z100_downstream$coordinates=gsub(">","",z100_downstream$coordinates)


zscore_out=merge(z25_upstream[,c('crossslink','coordinates','MFE','Zscore','seq')],z50_upstream[,c('crossslink','coordinates','MFE','Zscore','seq')],by="crossslink",suffixes = c("_25nt_Upstream","_50nt_Upstream"))
zscore_out=merge(zscore_out,z100_upstream[,c('crossslink','coordinates','MFE','Zscore','seq')],by="crossslink")
zscore_out=merge(zscore_out,z25_downstream[,c('crossslink','coordinates','MFE','Zscore','seq')],by="crossslink",suffixes = c("_100nt_Upstream","_25nt_Downstream"))
zscore_out=merge(zscore_out,z50_downstream[,c('crossslink','coordinates','MFE','Zscore','seq')],by="crossslink")
zscore_out=merge(zscore_out,z100_downstream[,c('crossslink','coordinates','MFE','Zscore','seq')],by="crossslink",suffixes = c("_50nt_Downstream","_100nt_Downstream"))

####################################################

GC25=read.delim('./structure/bed/structure_25.txt',comment.char = "#",header = T,sep = "\t",stringsAsFactors = F)
GC50=read.delim('./structure/bed/structure_50.txt',comment.char = "#",header = T,sep = "\t",stringsAsFactors = F)
GC100=read.delim('./structure/bed/structure_100.txt',comment.char = "#",header = T,sep = "\t",stringsAsFactors = F)
GC25=GC25[is.na(GC25$ID)==F,]
GC50=GC50[is.na(GC50$ID)==F,]
GC100=GC100[is.na(GC100$ID)==F,]

GC25$sequence_upstream=gsub("T","U",GC25$sequence_upstream);GC25$sequence_Downstream=gsub("T","U",GC25$sequence_Downstream)
GC100$sequence_upstream=gsub("T","U",GC100$sequence_upstream);GC100$sequence_Downstream=gsub("T","U",GC100$sequence_Downstream)
GC50$sequence_upstream=gsub("T","U",GC50$sequence_upstream);GC50$sequence_Downstream=gsub("T","U",GC50$sequence_Downstream)

GC25=separate(GC25,'ID',into = c('chr','start'),sep=":",remove = F)
GC25=separate(GC25,'start',into = c('start','end'),sep="-",remove = T)
GC25$start=as.numeric(GC25$start)
GC25$end=as.numeric(GC25$end)
GC50=separate(GC50,'ID',into = c('chr','start'),sep=":",remove = F)
GC50=separate(GC50,'start',into = c('start','end'),sep="-",remove = T)
GC50$start=as.numeric(GC50$start)
GC50$end=as.numeric(GC50$end)
GC100=separate(GC100,'ID',into = c('chr','start'),sep=":",remove = F)
GC100=separate(GC100,'start',into = c('start','end'),sep="-",remove = T)
GC100$start=as.numeric(GC100$start)
GC100$end=as.numeric(GC100$end)
colnames(GC100)=paste0(colnames(GC100),"_100nt")


GCout=merge(GC25[,c('ID','strand','coordinates_upstream','coordinates_Downstream','sequence_upstream','sequence_Downstream','Access.Energy..Upstream','GC_Upstream','Access.Energy..Downstream','GC_Downstream')],GC50[,c('ID','strand','coordinates_upstream','coordinates_Downstream','sequence_upstream','sequence_Downstream','Access.Energy..Upstream','GC_Upstream','Access.Energy..Downstream','GC_Downstream')],by="ID",suffixes = c("_25nt","_50nt"))
GCout=merge(GCout,GC100[,paste0(c('ID','strand','coordinates_upstream','coordinates_Downstream','sequence_upstream','sequence_Downstream','Access.Energy..Upstream','GC_Upstream','Access.Energy..Downstream','GC_Downstream'),"_100nt")],by.x="ID",by.y="ID_100nt")


GCout=separate(GCout,'ID',into = c('chr','start'),sep=":",remove = F)
GCout=separate(GCout,'start',into = c('start','end'),sep="-",remove = T)
GCout$start=as.numeric(GCout$start)
GCout$end=as.numeric(GCout$end)
GCout[GCout$strand_25nt%in%"+",'crossslink']=GCout[GCout$strand_25nt%in%"+",'start']
GCout[GCout$strand_25nt%in%"-",'crossslink']=GCout[GCout$strand_25nt%in%"-",'end']

structure_outall=merge(GCout,zscore_out,by.x='crossslink')

structure_outall=structure_outall[,c("ID","chr","start","end",'strand_25nt','strand_50nt','strand_100nt',"crossslink",
                     "coordinates_upstream_25nt",'coordinates_25nt_Upstream',"sequence_upstream_25nt","seq_25nt_Upstream","Access.Energy..Upstream_25nt","GC_Upstream_25nt","MFE_25nt_Upstream","Zscore_25nt_Upstream",
                     "coordinates_Downstream_25nt",'coordinates_25nt_Downstream',"sequence_Downstream_25nt","seq_25nt_Downstream","Access.Energy..Downstream_25nt","GC_Downstream_25nt","MFE_25nt_Downstream","Zscore_25nt_Downstream",
                     "coordinates_upstream_50nt",'coordinates_50nt_Upstream','sequence_upstream_50nt',"seq_50nt_Upstream","Access.Energy..Upstream_50nt","GC_Upstream_50nt","MFE_50nt_Upstream","Zscore_50nt_Upstream",
                     "coordinates_Downstream_50nt",'coordinates_50nt_Downstream','sequence_Downstream_50nt',"seq_50nt_Downstream","Access.Energy..Downstream_50nt","GC_Downstream_50nt","MFE_50nt_Downstream","Zscore_50nt_Downstream",
                     "coordinates_upstream_100nt",'coordinates_100nt_Upstream','sequence_upstream_100nt',"seq_100nt_Upstream","Access.Energy..Upstream_100nt","GC_Upstream_100nt","MFE_100nt_Upstream","Zscore_100nt_Upstream",
                     "coordinates_Downstream_100nt",'coordinates_100nt_Downstream','sequence_Downstream_100nt',"seq_100nt_Downstream","Access.Energy..Downstream_100nt","GC_Downstream_100nt","MFE_100nt_Downstream","Zscore_100nt_Downstream")]

# structure_outall[(structure_outall$coordinates_upstream_25nt==structure_outall$coordinates_25nt_Upstream)==F,]
# structure_outall[(structure_outall$coordinates_upstream_50nt==structure_outall$coordinates_50nt_Upstream)==F,]
# structure_outall[(structure_outall$coordinates_upstream_100nt==structure_outall$coordinates_100nt_Upstream)==F,]
# structure_outall[(structure_outall$coordinates_Downstream_25nt==structure_outall$coordinates_25nt_Downstream)==F,]
# structure_outall[(structure_outall$coordinates_Downstream_50nt==structure_outall$coordinates_50nt_Downstream)==F,]
# structure_outall[(structure_outall$coordinates_Downstream_100nt==structure_outall$coordinates_100nt_Downstream)==F,]
# 
# structure_outall[(structure_outall$seq_25nt_Upstream==structure_outall$sequence_upstream_25nt)==F,]
# structure_outall[(structure_outall$seq_50nt_Upstream==structure_outall$sequence_upstream_50nt)==F,]
# structure_outall[(structure_outall$seq_100nt_Upstream==structure_outall$sequence_upstream_100nt)==F,]
# structure_outall[(structure_outall$seq_25nt_Downstream==structure_outall$sequence_Downstream_25nt)==F,]
# structure_outall[(structure_outall$seq_50nt_Downstream==structure_outall$sequence_Downstream_50nt)==F,]
# structure_outall[(structure_outall$seq_100nt_Downstream==structure_outall$sequence_Downstream_100nt)==F,]


structure_outall=structure_outall[,c("ID",'strand_25nt',"crossslink",
                     "coordinates_upstream_25nt","sequence_upstream_25nt","GC_Upstream_25nt","Access.Energy..Upstream_25nt","MFE_25nt_Upstream","Zscore_25nt_Upstream",
                     "coordinates_Downstream_25nt","sequence_Downstream_25nt","GC_Downstream_25nt","Access.Energy..Downstream_25nt","MFE_25nt_Downstream","Zscore_25nt_Downstream",
                     "coordinates_upstream_50nt",'sequence_upstream_50nt',"GC_Upstream_50nt","Access.Energy..Upstream_50nt","MFE_50nt_Upstream","Zscore_50nt_Upstream",
                     "coordinates_Downstream_50nt",'sequence_Downstream_50nt',"GC_Downstream_50nt","Access.Energy..Downstream_50nt","MFE_50nt_Downstream","Zscore_50nt_Downstream",
                     "coordinates_upstream_100nt",'sequence_upstream_100nt',"GC_Upstream_100nt","Access.Energy..Upstream_100nt","MFE_100nt_Upstream","Zscore_100nt_Upstream",
                     "coordinates_Downstream_100nt",'sequence_Downstream_100nt',"GC_Downstream_100nt","Access.Energy..Downstream_100nt","MFE_100nt_Downstream","Zscore_100nt_Downstream")]

colnames(structure_outall)=c("ID",'strand',"crossslink_site",
                     "coordinates_Upstream_25nt","sequence_Upstream_25nt","GC_Upstream_25nt","Access_Energy_Upstream_25nt","deltaG_Upstream_25nt","Zscore_Upstream_25nt",
                     "coordinates_Downstream_25nt","sequence_Downstream_25nt","GC_Downstream_25nt","Access_Energy_Downstream_25nt","deltaG_Downstream_25nt","Zscore_Downstream_25nt",
                     "coordinates_Upstream_50nt",'sequence_Upstream_50nt',"GC_Upstream_50nt","Access_Energy_Upstream_50nt","deltaG_Upstream_50nt","Zscore_Upstream_50nt",
                     "coordinates_Downstream_50nt",'sequence_Downstream_50nt',"GC_Downstream_50nt","Access_Energy_Downstream_50nt","deltaG_Downstream_50nt","Zscore_Downstream_50nt",
                     "coordinates_Upstream_100nt",'sequence_upstream_100nt',"GC_Upstream_100nt","Access_Energy_Upstream_100nt","deltaG_Upstream_100nt","Zscore_Upstream_100nt",
                     "coordinates_Downstream_100nt",'sequence_Downstream_100nt',"GC_Downstream_100nt","Access_Energy_Downstream_100nt","deltaG_Downstream_100nt","Zscore_Downstream_100nt")

write.table(structure_outall, file=paste0('./structure/bed/CLIP_peaks_MD12_structure.txt'), quote=F, sep="\t", row.names=F, col.names=T)

Peaksdata4_structure=merge(Peaksdata4,structure_outall,by="ID")

```

<!-- ++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++ -->
<!-- ++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++ -->
<!-- ++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++ -->


```{r ,fig.asp=1, fig.align='center',fig.height=3, fig.width=6,echo=F,warning=F,eval=T,include=F}

pp=Peaksdata4_structure
p=pp[,c('Access_Energy_Upstream_25nt','Access_Energy_Downstream_25nt','Access_Energy_Upstream_50nt','Access_Energy_Downstream_50nt','Access_Energy_Upstream_100nt','Access_Energy_Downstream_100nt')]
p=melt(p,id.vars = NULL)
ggplot(p,aes(x=value,line=variable,color=variable))+stat_density(size=1.5,geom='line',position="identity")+theme_classic()+
  scale_color_manual(values=c('red','orange','blue','cyan','chartreuse4','green'))+
# scale_linetype_manual(values=c(1,2,1,2,1,2))+
# guides(color = guide_legend(override.aes = list(linetype = c(1,2,1,2,1,2))))+
guides(color = guide_legend(override.aes = list(size=1)))


pp=Peaksdata4_structure
p=pp[,c('Zscore_Upstream_25nt','Zscore_Downstream_25nt','Zscore_Upstream_50nt','Zscore_Downstream_50nt','Zscore_Upstream_100nt','Zscore_Downstream_100nt')]
p=melt(p,id.vars = NULL)
ggplot(p,aes(x=value,line=variable,color=variable))+stat_density(size=1.5,geom='line',position="identity")+theme_classic()+
  scale_color_manual(values=c('red','orange','blue','cyan','chartreuse4','green'))+
# scale_linetype_manual(values=c(1,2,1,2,1,2))+
# guides(color = guide_legend(override.aes = list(linetype = c(1,2,1,2,1,2))))+
guides(color = guide_legend(override.aes = list(size=1)))


pp=Peaksdata4_structure
p=pp[,c('deltaG_Upstream_25nt','deltaG_Downstream_25nt','deltaG_Upstream_50nt','deltaG_Downstream_50nt','deltaG_Upstream_100nt','deltaG_Downstream_100nt')]
p=melt(p,id.vars = NULL)
ggplot(p,aes(x=value,line=variable,color=variable))+stat_density(size=1.5,geom='line',position="identity")+theme_classic()+
  scale_color_manual(values=c('red','orange','blue','cyan','chartreuse4','green'))+
# scale_linetype_manual(values=c(1,2,1,2,1,2))+
# guides(color = guide_legend(override.aes = list(linetype = c(1,2,1,2,1,2))))+
guides(color = guide_legend(override.aes = list(size=1)))



pp=Peaksdata4_structure
p=pp[,c('GC_Upstream_25nt','GC_Downstream_25nt','GC_Upstream_50nt','GC_Downstream_50nt','GC_Upstream_100nt','GC_Downstream_100nt')];
p=melt(p,id.vars = NULL)
ggplot(p,aes(x=value,line=variable,color=variable))+stat_density(size=1.5,geom='line',position="identity")+theme_classic()+
  scale_color_manual(values=c('red','orange','blue','cyan','chartreuse4','green'))+
# scale_linetype_manual(values=c(1,2,1,2,1,2))+
# guides(color = guide_legend(override.aes = list(linetype = c(1,2,1,2,1,2))))+
guides(color = guide_legend(override.aes = list(size=1)))


```

```{r ,fig.asp=1, fig.align='center',fig.height=3, fig.width=3,echo=F,warning=F,eval=T,include=F}

p=Peaksdata4_structure
ggscatter(p, x = "Access_Energy_Upstream_25nt", y = 'GC_Upstream_25nt', add = "reg.line",color = "red",   
          add.params = list(color = "black", fill = NA), # Customize reg. line
          ylab='GC content: Upstream_25nt',
          xlab=expression("G"["accessability"]*" :Upstream_25nt" )  
) +
 stat_cor(
   aes(label = paste(..rr.label.., ..p.label.., sep = "~`,`~")),
  label.x = 3
)
###############

ggscatter(p, x = "Access_Energy_Upstream_100nt", y = 'GC_Upstream_100nt', add = "reg.line",color = "red",   
          add.params = list(color = "black", fill = NA), # Customize reg. line
          ylab='GC content: Upstream_100nt',
          xlab=expression("G"["accessability"]*" :Upstream_100nt" )  
) +
 stat_cor(
   aes(label = paste(..rr.label.., ..p.label.., sep = "~`,`~")),
  label.x = 3
)


########################################################
p=Peaksdata4_structure
ggscatter(p, x = "Zscore_Upstream_25nt", y = 'GC_Upstream_25nt', add = "reg.line",color = "red",   
          add.params = list(color = "black", fill = NA), # Customize reg. line
          ylab='GC content: Upstream_25nt',
          xlab=expression("Zscore: Upstream_25nt" )  
) +
 stat_cor(
   aes(label = paste(..rr.label.., ..p.label.., sep = "~`,`~")),
  label.x = -5,label.y = .25
)
###############

ggscatter(p, x = "Zscore_Upstream_100nt", y = 'GC_Upstream_100nt', add = "reg.line",color = "red",   
          add.params = list(color = "black", fill = NA), # Customize reg. line
          ylab='GC content: Upstream_100nt',
          xlab=expression("Zscore: Upstream_100nt" )  
) +
 stat_cor(
   aes(label = paste(..rr.label.., ..p.label.., sep = "~`,`~")),
  label.x = -5,label.y = .25
)


########################################################
p=Peaksdata4_structure
ggscatter(p, x = "deltaG_Upstream_25nt", y = 'GC_Upstream_25nt', add = "reg.line",color = "red",   
          add.params = list(color = "black", fill = NA), # Customize reg. line
          ylab='GC content: Upstream_25nt',
          xlab=expression("deltaG: Upstream_25nt" )  
) +
 stat_cor(
   aes(label = paste(..rr.label.., ..p.label.., sep = "~`,`~")),
  label.x = -10,label.y = .25
)
###############

ggscatter(p, x = "deltaG_Upstream_100nt", y = 'GC_Upstream_100nt', add = "reg.line",color = "red",   
          add.params = list(color = "black", fill = NA), # Customize reg. line
          ylab='GC content: Upstream_100nt',
          xlab=expression("deltaG: Upstream_100nt" )  
) +
 stat_cor(
   aes(label = paste(..rr.label.., ..p.label.., sep = "~`,`~")),
  label.x = -40,label.y = .25
)




p=Peaksdata4_structure
# ggscatter(p, x = "Access_Energy_Upstream_100nt", y = 'deltaG_Upstream_100nt', add = "reg.line",color = "red",   
#           add.params = list(color = "black", fill = NA), # Customize reg. line
#           ylab="deltaG: Upstream_100nt",
#           xlab=expression("G"["accessability"]*" :Upstream_100nt" )  
# ) +
#  stat_cor(
#    aes(label = paste(..rr.label.., ..p.label.., sep = "~`,`~")),
#   label.x = -40,label.y = .25
# )


#  ggscatter(p, x = "Access_Energy_Upstream_100nt", y = 'Zscore_Upstream_100nt', add = "reg.line",color = "red",
#           add.params = list(color = "black", fill = NA), # Customize reg. line
#           ylab="zscore: Upstream_100nt",
#           xlab=expression("G"["accessability"]*" :Upstream_100nt" )
# ) +
#  stat_cor(
#    aes(label = paste(..rr.label.., ..p.label.., sep = "~`,`~")),
#   label.x = 0,label.y = 4.75
# )

 ggscatter(p, x = "Zscore_Upstream_25nt", y = 'Zscore_Upstream_100nt', add = "reg.line",color = "red",
          add.params = list(color = "black", fill = NA), # Customize reg. line
          ylab="zscore: Upstream_100nt",
          xlab=expression("Zscore_Upstream_25nt" )
) +
 stat_cor(
   aes(label = paste(..rr.label.., ..p.label.., sep = "~`,`~")),
  label.x = 0,label.y = 4.75
) 
 
 
 
 ###################### ###################### ###################### ###################### ###################### ###################### ###################### ###################### ###################### ######################
 
  ggscatter(p, x = "GC_Upstream_25nt", y = 'GC_Downstream_25nt', add = "reg.line",color = "red",
          add.params = list(color = "black", fill = NA), # Customize reg. line
          ylab="GC: Upstream_25nt",
          xlab=expression("GC: Downstream_25nt" )
) +xlim(0,1)+ylim(0,1)+
 stat_cor(
   aes(label = paste(..rr.label.., ..p.label.., sep = "~`,`~"))#,
  # label.x = 0,label.y = 4.75
) 
 
  ggscatter(p, x = "GC_Upstream_100nt", y = 'GC_Downstream_100nt', add = "reg.line",color = "red",
          add.params = list(color = "black", fill = NA), # Customize reg. line
          ylab="GC: Upstream_100nt",
          xlab=expression("GC: Downstream_25nt" )
) +xlim(0,1)+ylim(0,1)+
 stat_cor(
   aes(label = paste(..rr.label.., ..p.label.., sep = "~`,`~"))#,
  # label.x = 0,label.y = 4.75
) 
  
  
  cbind(seq(0,100,1),seq(0,100,1)+1,seq(0,100,1)-1)/100
    cbind(seq(0,25,1),seq(0,25,1)+1,seq(0,25,1)-1)/25
      
 ggplot(p, aes(x = GC_Upstream_25nt, y = GC_Downstream_25nt))+ 
    geom_point(color='red')+ theme_classic()+
    xlim(0,1)+ylim(0,1)+
    geom_smooth(method = lm)+
    geom_abline(slope=1,  intercept=0)+    
    geom_abline(slope=1,  intercept=.1,linetype=2)+
    geom_abline(slope=1,  intercept=-.1,linetype=2)
 
  ggplot(p, aes(x = GC_Upstream_100nt, y = GC_Downstream_100nt))+ 
    geom_point(color='red')+ theme_classic()+
    xlim(0,1)+ylim(0,1)+
    geom_smooth(method = lm)+
    geom_abline(slope=1,  intercept=0)+    
    geom_abline(slope=1,  intercept=.1,linetype=2)+
    geom_abline(slope=1,  intercept=-.1,linetype=2)

    

#  {plot(p$Zscore_Upstream_25nt,p$Zscore_Upstream_50nt,col='blue',pch=16)
# points(p$Zscore_Upstream_25nt,p$Zscore_Upstream_100nt,col='green',pch=16)}
#  
 # p$z50diff=abs(p$Zscore_Upstream_50nt-p$Zscore_Upstream_25nt)
 # p$z100diff=abs(p$Zscore_Upstream_100nt-p$Zscore_Upstream_25nt)
 # 
 # ggplot(p,aes(x=z50diff,y=z100diff))+geom_point()+theme_classic()+
 #   geom_vline(xintercept = 0)+geom_hline(yintercept = 0)
 
 ggplot(p,aes(x=Zscore_Upstream_25nt,y=Zscore_Upstream_100nt))+geom_point()+theme_classic()+
   geom_vline(xintercept = 0)+geom_hline(yintercept = 0)

 ####
  p$GC50diff=p$GC_Upstream_50nt-p$GC_Upstream_25nt
 p$GC100diff=p$GC_Upstream_100nt-p$GC_Upstream_25nt
 
 ggplot(p,aes(x=GC50diff,y=GC100diff))+geom_point()+theme_classic()+
   geom_vline(xintercept = 0)+geom_hline(yintercept = 0)
 
ggplot(p,aes(x=GC_Upstream_25nt,y=GC_Upstream_100nt))+geom_point()+theme_classic()+
   geom_vline(xintercept = 0)+geom_hline(yintercept = 0)
ggplot(p,aes(x=GC_Downstream_25nt,y=GC_Downstream_100nt))+geom_point()+theme_classic()+
   geom_vline(xintercept = 0)+geom_hline(yintercept = 0)
  
 ggplot(p,aes(x=GC_Upstream_25nt,y=GC_Upstream_50nt))+geom_point()+theme_classic()+
   geom_vline(xintercept = 0)+geom_hline(yintercept = 0)
ggplot(p,aes(x=GC_Downstream_25nt,y=GC_Downstream_50nt))+geom_point()+theme_classic()+
   geom_vline(xintercept = 0)+geom_hline(yintercept = 0)
  
 
 ggplot(p,aes(x=GC_Upstream_50nt,y=GC_Upstream_100nt))+geom_point()+theme_classic()+
   geom_vline(xintercept = 0)+geom_hline(yintercept = 0)
ggplot(p,aes(x=GC_Downstream_50nt,y=GC_Downstream_100nt))+geom_point()+theme_classic()+
   geom_vline(xintercept = 0)+geom_hline(yintercept = 0)
  
 
##################3########




#   ####
# p$Acc50diff=p$Access_Energy_Upstream_50nt-p$Access_Energy_Upstream_25nt
# p$Acc100diff=p$Access_Energy_Upstream_100nt-p$Access_Energy_Upstream_25nt
#  
#  ggplot(p,aes(x=Acc50diff,y=Acc100diff))+geom_point()+theme_classic()+
#    geom_vline(xintercept = 0)+geom_hline(yintercept = 0)
#  
#   ggplot(p,aes(x=Access_Energy_Upstream_50nt,y=Acc100diff))+geom_point()+theme_classic()+

 
```

```{r ,fig.asp=1, fig.align='center',fig.height=3, fig.width=3,echo=F,warning=F,eval=T,include=F}

p=Peaksdata4_structure
ggscatter(p, x = "Access_Energy_Downstream_25nt", y = 'GC_Downstream_25nt', add = "reg.line",color = "red",   
          add.params = list(color = "black", fill = NA), # Customize reg. line
          ylab='GC content: Downstream_25nt',
          xlab=expression("G"["accessability"]*" :Downstream_25nt" )  
) +
 stat_cor(
   aes(label = paste(..rr.label.., ..p.label.., sep = "~`,`~")),
  label.x = 10,label.y=.2
)
###############

ggscatter(p, x = "Access_Energy_Downstream_100nt", y = 'GC_Downstream_100nt', add = "reg.line",color = "red",   
          add.params = list(color = "black", fill = NA), # Customize reg. line
          ylab='GC content: Downstream_100nt',
          xlab=expression("G"["accessability"]*" :Downstream_100nt" )  
) +
 stat_cor(
   aes(label = paste(..rr.label.., ..p.label.., sep = "~`,`~")),
  label.x = 3
)


########################################################
p=Peaksdata4_structure
ggscatter(p, x = "Zscore_Downstream_25nt", y = 'GC_Downstream_25nt', add = "reg.line",color = "red",   
          add.params = list(color = "black", fill = NA), # Customize reg. line
          ylab='GC content: Downstream_25nt',
          xlab=expression("Zscore: Downstream_25nt" )  
) +
 stat_cor(
   aes(label = paste(..rr.label.., ..p.label.., sep = "~`,`~")),
  label.x = -5,label.y = .25
)
###############

ggscatter(p, x = "Zscore_Downstream_100nt", y = 'GC_Downstream_100nt', add = "reg.line",color = "red",   
          add.params = list(color = "black", fill = NA), # Customize reg. line
          ylab='GC content: Downstream_100nt',
          xlab=expression("Zscore: Downstream_100nt" )  
) +
 stat_cor(
   aes(label = paste(..rr.label.., ..p.label.., sep = "~`,`~")),
  label.x = -5,label.y = .25
)


########################################################
p=Peaksdata4_structure
ggscatter(p, x = "deltaG_Downstream_25nt", y = 'GC_Downstream_25nt', add = "reg.line",color = "red",   
          add.params = list(color = "black", fill = NA), # Customize reg. line
          ylab='GC content: Downstream_25nt',
          xlab=expression("deltaG: Downstream_25nt" )  
) +
 stat_cor(
   aes(label = paste(..rr.label.., ..p.label.., sep = "~`,`~")),
  label.x = -10,label.y = .25
)
###############

ggscatter(p, x = "deltaG_Downstream_100nt", y = 'GC_Downstream_100nt', add = "reg.line",color = "red",   
          add.params = list(color = "black", fill = NA), # Customize reg. line
          ylab='GC content: Downstream_100nt',
          xlab=expression("deltaG: Downstream_100nt" )  
) +
 stat_cor(
   aes(label = paste(..rr.label.., ..p.label.., sep = "~`,`~")),
  label.x = -50,label.y = .25
)

```



```{r ,fig.asp=1, fig.align='center',fig.height=3, fig.width=3,echo=F,warning=F,eval=T,include=F}

counts="Counts_Multimappers_All"
counts='Counts_Multimappers_Scaled'
# counts='Counts_Multimappers_BestMapping'
# counts='Counts_Unique'
p=Peaksdata4_structure
ggscatter(p, x = counts, y = 'GC_Upstream_50nt', add = "reg.line",color = "red",   
          add.params = list(color = "black", fill = NA), # Customize reg. line
          xlab='CLIP peak Counts',
          ylab=expression("GC_Upstream_50nt" )  
) +
 stat_cor(
   aes(label = paste(..rr.label.., ..p.label.., sep = "~`,`~")),
  label.x = 3
)

#################

p=Peaksdata4_structure
ggscatter(p, x = counts, y = 'deltaG_Upstream_50nt', add = "reg.line",color = "red",   
          add.params = list(color = "black", fill = NA), # Customize reg. line
          xlab='CLIP peak Counts',
          ylab=expression("deltaG_Upstream_50nt" )  
) +
 stat_cor(
   aes(label = paste(..rr.label.., ..p.label.., sep = "~`,`~")),
  label.x = 3
)

#################

p=Peaksdata4_structure
ggscatter(p, x = counts, y = 'Access_Energy_Upstream_50nt', add = "reg.line",color = "red",   
          add.params = list(color = "black", fill = NA), # Customize reg. line
          xlab='CLIP peak Counts',
          ylab=expression("Access_Energy_Upstream_50nt" )  
) +
 stat_cor(
   aes(label = paste(..rr.label.., ..p.label.., sep = "~`,`~")),
  label.x = 3
)


#################

p=Peaksdata4_structure
ggscatter(p, x = counts, y = 'Zscore_Upstream_50nt', add = "reg.line",color = "red",   
          add.params = list(color = "black", fill = NA), # Customize reg. line
          xlab='CLIP peak Counts',
          ylab=expression("Zscore_Upstream_50nt" )  
) +
 stat_cor(
   aes(label = paste(..rr.label.., ..p.label.., sep = "~`,`~")),
  label.x = 3
)

###########################################################################################################
###########################################################################################################

p=Peaksdata4_structure
ggscatter(p, x = counts, y = 'GC_Downstream_50nt', add = "reg.line",color = "red",   
          add.params = list(color = "black", fill = NA), # Customize reg. line
          xlab='CLIP peak Counts',
          ylab=expression("GC_Downstream_50nt" )  
) +
 stat_cor(
   aes(label = paste(..rr.label.., ..p.label.., sep = "~`,`~")),
  label.x = 3
)

#################

p=Peaksdata4_structure
ggscatter(p, x = counts, y = 'deltaG_Downstream_50nt', add = "reg.line",color = "red",   
          add.params = list(color = "black", fill = NA), # Customize reg. line
          xlab='CLIP peak Counts',
          ylab=expression("deltaG_Downstream_50nt" )  
) +
 stat_cor(
   aes(label = paste(..rr.label.., ..p.label.., sep = "~`,`~")),
  label.x = 3
)

#################

p=Peaksdata4_structure
ggscatter(p, x = counts, y = 'Access_Energy_Downstream_50nt', add = "reg.line",color = "red",   
          add.params = list(color = "black", fill = NA), # Customize reg. line
          xlab='CLIP peak Counts',
          ylab=expression("Access_Energy_Downstream_50nt" )  
) +
 stat_cor(
   aes(label = paste(..rr.label.., ..p.label.., sep = "~`,`~")),
  label.x = 3
)


#################

p=Peaksdata4_structure
ggscatter(p, x = counts, y = 'Zscore_Downstream_50nt', add = "reg.line",color = "red",   
          add.params = list(color = "black", fill = NA), # Customize reg. line
          xlab='CLIP peak Counts',
          ylab=expression("Zscore_Downstream_50nt" )  
) +
 stat_cor(
   aes(label = paste(..rr.label.., ..p.label.., sep = "~`,`~")),
  label.x = 3
)


```


```{r ,fig.asp=1, fig.align='center',fig.height=3, fig.width=3,echo=F,warning=F,eval=T,include=F}

counts="Counts_Multimappers_All"
counts='Counts_Multimappers_Scaled'
# counts='Counts_Multimappers_BestMapping'
# counts='Counts_Unique'

p=Peaksdata4_structure

p$diff_Access25nt=p$Access_Energy_Upstream_25nt-p$Access_Energy_Downstream_25nt
ggscatter(p, x = counts, y = 'diff_Access25nt', add = "reg.line",color = "red",   
          add.params = list(color = "black", fill = NA), # Customize reg. line
          xlab='CLIP peak Counts',
          ylab=expression("diff_Access_Upstream_25nt" )  
) +
 stat_cor(
   aes(label = paste(..rr.label.., ..p.label.., sep = "~`,`~")),
  label.x = 3
)

#################

p$diff_deltaG25nt=p$deltaG_Upstream_25nt-p$deltaG_Downstream_25nt
ggscatter(p, x = counts, y = 'diff_deltaG25nt', add = "reg.line",color = "red",   
          add.params = list(color = "black", fill = NA), # Customize reg. line
          xlab='CLIP peak Counts',
          ylab=expression("diff_deltaG_Upstream_25nt" )  
) +
 stat_cor(
   aes(label = paste(..rr.label.., ..p.label.., sep = "~`,`~")),
  label.x = 3
)

#################

p$diff_GC25nt=(p$GC_Upstream_25nt)*100-(p$GC_Downstream_25nt)*100
ggscatter(p, x = counts, y = 'diff_GC25nt', add = "reg.line",color = "red",   
          add.params = list(color = "black", fill = NA), # Customize reg. line
          xlab='CLIP peak Counts',
          ylab=expression("diff_GC_Upstream_25nt" )  
) +
 stat_cor(
   aes(label = paste(..rr.label.., ..p.label.., sep = "~`,`~")),
  label.x = 3
)
```

```{r ,fig.asp=1, fig.align='center',fig.height=3, fig.width=6,echo=F,warning=F,eval=T,include=F}

pp=p[,c('diff_Access25nt','diff_deltaG25nt','diff_GC25nt')]
pp=melt(pp,id.vars = NULL)
ggplot(pp,aes(x=value,line=variable,color=variable) )+geom_density(size=1)+theme_classic()+
  geom_vline(xintercept = 0)+
    # scale_color_manual(values=c('red','orange','blue','cyan','chartreuse4','green'))+
# scale_linetype_manual(values=c(1,2,1,2,1,2))+
# guides(color = guide_legend(override.aes = list(linetype = c(1,2,1,2,1,2))))+
guides(color = guide_legend(override.aes = list(size=1)))



```
```{r ,fig.asp=1, fig.align='center',fig.height=3, fig.width=3,echo=F,warning=F,eval=T,include=F}

#######################################################################################################################
#######################################################################################################################

p$diff_Access100nt=p$Access_Energy_Upstream_100nt-p$Access_Energy_Downstream_100nt
ggscatter(p, x = counts, y = 'diff_Access100nt', add = "reg.line",color = "red",   
          add.params = list(color = "black", fill = NA), # Customize reg. line
          xlab='CLIP peak Counts',
          ylab=expression("diff_Access_Upstream_100nt" )  
) +
 stat_cor(
   aes(label = paste(..rr.label.., ..p.label.., sep = "~`,`~")),
  label.x = 3
)

#################

p$diff_deltaG100nt=p$deltaG_Upstream_100nt-p$deltaG_Downstream_100nt
ggscatter(p, x = counts, y = 'diff_deltaG100nt', add = "reg.line",color = "red",   
          add.params = list(color = "black", fill = NA), # Customize reg. line
          xlab='CLIP peak Counts',
          ylab=expression("diff_deltaG_Upstream_100nt" )  
) +
 stat_cor(
   aes(label = paste(..rr.label.., ..p.label.., sep = "~`,`~")),
  label.x = 3
)

#################

p$diff_GC100nt=(p$GC_Upstream_100nt)*100-(p$GC_Downstream_100nt)*100
ggscatter(p, x = counts, y = 'diff_GC100nt', add = "reg.line",color = "red",   
          add.params = list(color = "black", fill = NA), # Customize reg. line
          xlab='CLIP peak Counts',
          ylab=expression("diff_GC_Upstream_100nt" )  
) +
 stat_cor(
   aes(label = paste(..rr.label.., ..p.label.., sep = "~`,`~")),
  label.x = 3
)

#######################################################################################################################
#######################################################################################################################

```

```{r ,fig.asp=1, fig.align='center',fig.height=3, fig.width=6,echo=F,warning=F,eval=T,include=F}

pp=p[,c('diff_Access100nt','diff_deltaG100nt','diff_GC100nt')]
pp=melt(pp,id.vars = NULL)
ggplot(pp,aes(x=value,line=variable,color=variable) )+stat_density(size=1.5,geom='line',position="identity")+theme_classic()+
  geom_vline(xintercept = 0)


pp=p[,c('diff_GC25nt','diff_GC100nt')]
pp=melt(pp,id.vars = NULL)
ggplot(pp,aes(x=value,line=variable,color=variable) )+stat_density(size=1.5,geom='line',position="identity")+theme_classic()+
  geom_vline(xintercept = 0)
```



<!-- ============================================================================================================================= -->

<!-- figure for specific peak for function to calculate Gaccess for peaks -->
```{r ,fig.asp=.5, fig.align='center',fig.height=6, fig.width=12,echo=F,warning=F,eval=F,include=F}

###############################################################
### figure for specific peak for function to calculate Gaccess for peaks
###############################################################
Peaksdata4_motif2=Peaksdata4[Peaksdata4$`Both Strand: RNA_type_Classification`%in%c('protein_coding: exon','protein_coding: Intron'),]
# unique(Peaksdata4_motif$`Both Strand: RNA_type_Classification`)
Peaksdata4_motif2=separate(Peaksdata4_motif2,ID,into = c('chr','start'),sep=':',remove = F)
Peaksdata4_motif2=separate(Peaksdata4_motif2,start,into = c('start','end'),sep='-')
Peaksdata4_motif2$start=as.numeric(Peaksdata4_motif2$start)
Peaksdata4_motif2$end=as.numeric(Peaksdata4_motif2$end)
# unique(Peaksdata4_motif2$chr)

fasta='../../../../Resources/ref/mm10/Gencode_VM23/fromGencode/GRCm38.primary_assembly.genome.fa'


# 
## chr1:128642762-128642837
## chr12:111251316-111251370
p=Peaksdata4[Peaksdata4$ID%in%'chr12:111251316-111251370',]
p=separate(p,ID,into = c('chr','start'),sep=':',remove = F)
p=separate(p,start,into = c('start','end'),sep='-')
p$start=as.numeric(p$start)
p$end=as.numeric(p$end)

######################################################################################################
structout=matrix(nrow = nrow(Peaksdata4_motif2),ncol=8)
figure=T
# for (x in 1:nrow(Peaksdata4_motif2)) {
for (x in 1) {
# if (x%in%seq(from=0,to=1000,by=1)){print(x)}
# p=Peaksdata4_motif2[x,]
  
chr=p$chr
start=p$start
end=p$end
strand=p$strand

buffer=1000
window=100
##################
s1=start-buffer
e1=end+(buffer-1)

bedpos=matrix(ncol=6,nrow=1)
bedpos[1,1]=chr
bedpos[1,2]=s1
bedpos[1,3]=e1
bedpos[1,4]=".";bedpos[1,5]="."
bedpos[1,6]=strand
write.table(bedpos, file=paste0('./structure/bed/',p$ID,"_intronexon.bed"), quote=F, sep="\t", row.names=F, col.names=F)
system(paste0('bedtools getfasta -s -fi ',fasta,' -bed structure/bed/',p$ID,'_intronexon.bed > structure/bed/',p$ID,'_intronexon.fa'))

# system(paste0('/Users/homanpj/Documents/Tools/ParasoR/src/ParasoR --input structure/bed/',p$ID,'_intronexon.fa --pre --bpp --mfe --acc --cd structure/ --window ',window,'  > structure/bed/',p$ID,'.out'))
system(paste0('/Users/homanpj/Documents/Tools/ParasoR/src/ParasoR --input structure/bed/',p$ID,'_intronexon.fa --pre --acc --bpp --mfe --cd structure/ --window ',window,' > structure/bed/',p$ID,'.out'))


str=read.delim(paste0('structure/bed/',p$ID,'.out'),comment.char = "#",header = F,sep = "\t")
str=separate(str,V2,into = c('start','end'),sep = '-',remove = F)
str$start=as.numeric(str$start)
str$end=as.numeric(str$end)
str$V4adj=str$V4/50
######################################################################

a=as.data.frame(xxx[[chr]])
a=cbind(c(1:nrow(a)),a)
# look upsream - calculation starts at nt through length of window 
# shift nt count so that Calc ends with windo at nt of interest
# a5=a
# a5[,1]=a5[,1]+window
# colnames(a5)=c("nt",'GC')
a3=a
a3[,1]=a3[,1]
colnames(a3)=c("nt",'GC')
######################################################################

if (strand%in%'+') {
## Start of table is start of selected sequence
  str$start_adj=(str$start-1)+s1
## Adjust count so that window ends at nt position
  str$end_adj=str$start_adj+window

GCstruct2=merge(a3,str,by.x='nt',by.y='start_adj')
GCstruct2$GC=SMA(GCstruct2$GC,n=10)
GCstruct2$V4adj=SMA(GCstruct2$V4adj,n=10)
GCstruct2$V4=SMA(GCstruct2$V4,n=10)

croslink=p$start
structure_upstream=GCstruct2[GCstruct2$end_adj%in%croslink,'V4']
structure_downstream=GCstruct2[GCstruct2$nt%in%croslink,'V4']
GC_upstream=GCstruct2[GCstruct2$end_adj%in%croslink,'GC']
GC_downstream=GCstruct2[GCstruct2$nt%in%croslink,'GC']

if (figure==T) {
{par(mar = c(5, 5, 3, 5))
  plot(GCstruct2$end_adj,GCstruct2$V4,type='l',col='red',xlab='Nucleotide Position',ylab=expression("G"["accessability"]),main=paste0('Strucutre Upstream\n',p$ID))
  par(new = TRUE)
  plot(GCstruct2$end_adj,GCstruct2$GC,type='l',col='black',ylab = "", xlab = "", xaxt = "n", yaxt = "n")
    axis(side = 4)
  mtext('GC content (%)', side = 4, line = 3)
  lines(start:end,matrix(max(GCstruct2$GC,na.rm=T)-.05,nrow=((end-start)+1),ncol=1)[,1],col='blue',type = "p")
  lines(croslink,max(GCstruct2$GC,na.rm=T)-.05,col='green',type = "p",pch=16)
    legend("bottomleft", legend=c("GC content", expression("G"["accessability"]),"CLIP peak",'5`end CLIP peak' ),
         col=c("black", "red","blue",'green'), lty=c(1,1,1,0),pch=c(NA,NA,NA,16), cex=1,lwd=3)} 
{par(mar = c(5, 5, 3, 5))
  plot(GCstruct2$nt,GCstruct2$V4,type='l',col='red',xlab='Nucleotide Position',ylab=expression("G"["accessability"]),main=paste0('Strucutre Downstream\n',p$ID))
  par(new = TRUE)
  plot(GCstruct2$nt,GCstruct2$GC,type='l',col='black',ylab = "", xlab = "", xaxt = "n", yaxt = "n")
    axis(side = 4)
  mtext('GC content (%)', side = 4, line = 3)
  lines(start:end,matrix(max(GCstruct2$GC,na.rm=T)-.05,nrow=((end-start)+1),ncol=1)[,1],col='blue',type = "p")
  lines(croslink,max(GCstruct2$GC,na.rm=T)-.05,col='green',type = "p",pch=16)
    legend("bottomleft", legend=c("GC content", expression("G"["accessability"]),"CLIP peak",'5`end CLIP peak' ),
         col=c("black", "red","blue",'green'), lty=c(1,1,1,0),pch=c(NA,NA,NA,16), cex=1,lwd=3)}
}
}
##############
if (strand%in%'-') {
## Reverse order of table (so now oriented same direction as + strand sequence)
  str=str[sort(str$start,decreasing = T),]
  str$start_opp=c(1:nrow(str))
## Start of table is start of selected sequence
## Window goes from nt possition + window downstream
  str$start_adj=(str$start_opp-1)+s1

### look downfild of nucleotide
  str$end_adj=(str$start_adj-1)+window

GCstruct2=merge(a3,str,by.x='nt',by.y='start_adj')
GCstruct2$GC=SMA(GCstruct2$GC,n=10)*100
GCstruct2$V4adj=SMA(GCstruct2$V4adj,n=10)
GCstruct2$V4=SMA(GCstruct2$V4,n=10)*100

croslink=p$end
structure_upstream=GCstruct2[GCstruct2$nt%in%croslink,'V4']
structure_downstream=GCstruct2[GCstruct2$end_adj%in%croslink,'V4']
GC_upstream=GCstruct2[GCstruct2$nt%in%croslink,'GC']
GC_downstream=GCstruct2[GCstruct2$end_adj%in%croslink,'GC']

if (figure==T) {
{par(mar = c(5, 5, 3, 5))
  plot(GCstruct2$nt,GCstruct2$V4,type='l',col='red',xlab='Nucleotide Position',ylab=expression("deltaG"["accessability"]),main=paste0('Strucutre Upstream\n',p$ID))
  par(new = TRUE)
  plot(GCstruct2$nt,GCstruct2$GC,type='l',col='black',ylab = "", xlab = "", xaxt = "n", yaxt = "n")
    axis(side = 4)
  mtext('GC content (%)', side = 4, line = 3)
  lines(start:end,matrix(max(GCstruct2$GC,na.rm=T)-.05,nrow=((end-start)+1),ncol=1)[,1],col='blue',type = "p")
  lines(croslink,max(GCstruct2$GC,na.rm=T)-.05,col='green',type = "p",pch=16)
  legend("bottomleft", legend=c("GC content", expression("G"["accessability"]),"CLIP peak",'5`end CLIP peak' ),
         col=c("black", "red","blue",'green'), lty=c(1,1,1,0),pch=c(NA,NA,NA,16), cex=1,lwd=3)}
{par(mar = c(5, 5, 3, 5))
  plot(GCstruct2$end_adj,GCstruct2$V4,type='l',col='red',xlab='Nucleotide Position',ylab=expression("deltaG"["accessability"]),main=paste0('Strucutre Downstream\n',p$ID))
  par(new = TRUE)
  plot(GCstruct2$end_adj,GCstruct2$GC,type='l',col='black',ylab = "", xlab = "", xaxt = "n", yaxt = "n")
  axis(side = 4)
  mtext('GC content (%)', side = 4, line = 3)
  lines(start:end,matrix(max(GCstruct2$GC,na.rm=T)-.05,nrow=((end-start)+1),ncol=1)[,1],col='blue',type = "p")
  lines(croslink,max(GCstruct2$GC,na.rm=T)-.05,col='green',type = "p",pch=16)
  legend("bottomleft", legend=c("GC content", expression("G"["accessability"]),"CLIP peak",'5`end CLIP peak' ),
         col=c("black", "red","blue",'green'), lty=c(1,1,1,0),pch=c(NA,NA,NA,16), cex=1,lwd=3)}
}
# cor(GCstruct2[is.na(GCstruct2$GC)==F,'GC'],GCstruct2[is.na(GCstruct2$V4adj)==F,'V4adj'])
}
# remove('GCstruct2','croslink','a3','a','str')

}
```

```{r ,fig.asp=.5, fig.align='center',fig.height=3, fig.width=6,echo=F,warning=F,eval=F,include=F}

p$Wstart=p$start-buffer
p$Wend=(p$end+buffer)-1
mafout=as.data.frame(matrix(nrow=abs(p$Wstart-(p$Wend-window)),ncol=7))
for (x in 1:nrow(mafout)) {
  
mafout[x,1]="s"
mafout[x,2]=paste0(p$ID,"_",x)
mafout[x,3]=p$Wstart+(x)
mafout[x,4]=abs(((p$Wstart+(x))+window) - mafout[x,3])
mafout[x,5]= p$strand
mafout[x,6]= (p$Wstart+(x-1))+window

bedpos=matrix(ncol=6,nrow=1)
bedpos[1,1]=p$chr
bedpos[1,2]=p$Wstart+(x-1)
bedpos[1,3]=(p$Wstart+(x-1))+window
bedpos[1,4]=".";bedpos[1,5]="."
bedpos[1,6]=p$strand
write.table(bedpos, file=paste0('./structure/bed_RNAz/',p$ID,"_rNAz.bed"), quote=F, sep="\t", row.names=F, col.names=F)
mafout[x,7]=system(paste0('bedtools getfasta -s -fi ',fasta,' -bed structure/bed_RNAz/',p$ID,'_rNAz.bed'),intern = T)[2]
if(file.exists(paste0('./structure/bed_RNAz/',p$ID,"_rNAz.bed"))){file.remove(paste0('./structure/bed_RNAz/',p$ID,"_rNAz.bed"))}
}


mafout=mafout[is.na(mafout$V1)==F,]
inter=as.data.frame(matrix(nrow=2,ncol=7))
inter[2,1]='a'
inter[2,2]='score=1.000000'
#Every N rows after which empty rows should be inserted
N = 2
after_rows = 300
mafout2=do.call(rbind, lapply(split(mafout, ceiling(1:nrow(mafout)/after_rows)),
function(a) rbind(a, replace(a[1:N,], TRUE, inter))))
mafout2=rbind(inter[2,],mafout2)
mafout2=mafout2[1:(nrow(mafout2)-2),]
write.table(mafout2, file=paste0('./structure/bed_RNAz/',p$ID,"_rNAz.maf"),na = "", quote=F, sep="\t", row.names=F, col.names=F)
# system(paste0('/Users/homanpj/Documents/Tools/RNAz-2.1/rnaz/RNAz ./structure/bed_RNAz/',p$ID,'_rNAz.maf > ./structure/bed_RNAz/',p$ID,'_rNAz.out'))
outmaf=system(paste0('/Users/homanpj/Documents/Tools/RNAz-2.1/rnaz/RNAz ./structure/bed_RNAz/',p$ID,'_rNAz.maf'),intern = T)
if(file.exists(paste0('./structure/bed_RNAz/',p$ID,'_rNAz.maf'))){file.remove(paste0('./structure/bed_RNAz/',p$ID,'_rNAz.maf'))}
outmaf_ov=as.data.table(outmaf[outmaf%in%""==F])
while (isEmpty(grep("#",outmaf_ov$V1))==F) {
g=grep("#",outmaf_ov$V1)
outmaf_ov=(outmaf_ov[-c(g[1]:g[2]),])
outmaf_ov=(outmaf_ov)
}
outmaf_ov2=cbind(outmaf_ov[seq(3, nrow(outmaf_ov), 3),],outmaf_ov[seq(1, nrow(outmaf_ov), 3),])
colnames(outmaf_ov2)=c('output','info')
outmaf_ov=separate(outmaf_ov2,'output',into = c('str','data'),sep = 100)
outmaf_ov=separate(outmaf_ov,'data',into = c('MFE','Zscore','direction'),sep = ', ')
outmaf_ov$MFE=as.numeric(gsub("\\( ","",outmaf_ov$MFE))
outmaf_ov$Zscore=as.numeric(gsub("z-score = ","",outmaf_ov$Zscore))
outmaf_ov=separate(outmaf_ov,'info',into = c('name','start','length','strand','end'),sep = ' ')
outmaf_ov$start=as.numeric(outmaf_ov$start)
outmaf_ov$end=as.numeric(outmaf_ov$end)
outmaf_ov=outmaf_ov[!is.na(outmaf_ov$MFE),]
outmaf_ov=cbind(outmaf_ov,c(1:nrow(outmaf_ov)))
outmaf_ov$MFE=SMA(outmaf_ov$MFE,n=10)
outmaf_ov$Zscore=SMA(outmaf_ov$Zscore,n=10)
if (strand%in%"+") {outmaf_ov2=merge(GCstruct2[,c('nt','GC','V4',"V4adj",'end_adj')],outmaf_ov,by.x="nt",by.y="start")}
if (strand%in%"-") {outmaf_ov2=merge(GCstruct2[,c('nt','GC','V4',"V4adj",'end_adj')],outmaf_ov,by.x="nt",by.y="start")}


```




```{r ,fig.asp=.5, fig.align='center',fig.height=6, fig.width=12,echo=F,warning=F,eval=F,include=F}

if (figure==T) {
{par(mar = c(5, 5, 3, 5))
  plot(GCstruct2$end_adj,GCstruct2$V4,type='l',col='red',xlab='Nucleotide Position',ylab=expression("G"["accessability"]),main=paste0('Strucutre Upstream\n',p$ID))
  par(new = TRUE)
  plot(GCstruct2$end_adj,GCstruct2$GC,type='l',col='black',ylab = "", xlab = "", xaxt = "n", yaxt = "n")
    axis(side = 4)
  mtext('GC content (%)', side = 4, line = 3)
  lines(start:end,matrix(max(GCstruct2$GC,na.rm=T)-.05,nrow=((end-start)+1),ncol=1)[,1],col='blue',type = "p")
  lines(croslink,max(GCstruct2$GC,na.rm=T)-.05,col='green',type = "p",pch=16)
    legend("bottomleft", legend=c("GC content", expression("G"["accessability"]),"CLIP peak",'5`end CLIP peak' ),
         col=c("black", "red","blue",'green'), lty=c(1,1,1,0),pch=c(NA,NA,NA,16), cex=1,lwd=3)} 
}


if (p$strand%in%c("+")) {
{par(mar = c(5, 5, 3, 5))
plot(outmaf_ov2$end_adj,-outmaf_ov2$MFE,type='l',col='orange',xlab='Nucleotide Position',ylab=expression("-deltaG"),main=paste0('Strucutre Upstream\n',p$ID))
par(new = TRUE)
plot(outmaf_ov2$end_adj,-outmaf_ov2$Zscore,type='l',col='purple',ylab = "", xlab = "", xaxt = "n", yaxt = "n",ylim=c(-3.5,3.5))
axis(side = 4)
abline(h=0,lty=2,col='purple')
abline(h=-1,lty=2,col='purple')
abline(h=1,lty=2,col='purple')
mtext('-Z-score', side = 4, line = 3)
text(max(outmaf_ov2$end_adj-100),3,"More Structured");  text(max(outmaf_ov2$end_adj-100),-3,"Less Structured")
lines(start:end,matrix(3,nrow=((end-start)+1),ncol=1)[,1],col='blue',type = "p")
lines(croslink,3,col='green',type = "p",pch=16)
legend("bottomleft", legend=c("Gibbs free engergy", 'Z-score',"CLIP peak",'5`end CLIP peak' ),
col=c("orange", "purple","blue",'green'), lty=c(1,1,1,0),pch=c(NA,NA,NA,16), cex=1,lwd=3)}
  
  
{par(mar = c(5, 5, 3, 5))
plot(outmaf_ov2$end_adj,-outmaf_ov2$MFE,type='l',col='orange',xlab='Nucleotide Position',ylab=expression("-deltaG"),main=paste0('Strucutre Upstream\n',p$ID))
par(new = TRUE)
plot(outmaf_ov2$end_adj,outmaf_ov2$V4,type='l',col='red',ylab = "", xlab = "", xaxt = "n", yaxt = "n")
axis(side = 4)
mtext(expression("G"["accessability"]), side = 4, line = 3)
lines(start:end,matrix(max(outmaf_ov2$V4,na.rm = T)-5,nrow=((end-start)+1),ncol=1)[,1],col='blue',type = "p")
lines(croslink,max(outmaf_ov2$V4,na.rm = T)-5,col='green',type = "p",pch=16)
legend("bottomleft", legend=c("Gibbs free engergy", expression("G"["accessability"]),"CLIP peak",'5`end CLIP peak' ),
col=c("orange", "red","blue",'green'), lty=c(1,1,1,0),pch=c(NA,NA,NA,16), cex=1,lwd=3)}
  
  
{par(mar = c(5, 5, 3, 5))
plot(outmaf_ov2$end_adj,-outmaf_ov2$MFE,type='l',col='orange',xlab='Nucleotide Position',ylab=expression("-deltaG"),main=paste0('Strucutre Upstream\n',p$ID))
lines(outmaf_ov2$end_adj,(outmaf_ov2$V4)/1.3,col='red')
par(new = TRUE)
plot(outmaf_ov2$end_adj,-outmaf_ov2$Zscore,type='l',col='purple',ylab = "", xlab = "", xaxt = "n", yaxt = "n",ylim=c(-3.5,3.5))
axis(side = 4)
abline(h=0,lty=2,col='purple')
abline(h=-1,lty=2,col='purple')
abline(h=1,lty=2,col='purple')
mtext(expression("-Z-score"), side = 4, line = 3)
text(max(outmaf_ov2$end_adj-100),3,"More Structured");  text(max(outmaf_ov2$end_adj-100),-3,"Less Structured")
lines(start:end,matrix(3,nrow=((end-start)+1),ncol=1)[,1],col='blue',type = "p")
lines(croslink,3,col='green',type = "p",pch=16)
legend("bottomleft", legend=c("Gibbs free engergy", expression("G"["accessability"]),'Z-Score',"CLIP peak",'5`end CLIP peak' ),
col=c("orange", "red","purple","blue",'green'), lty=c(1,1,1,1,0),pch=c(NA,NA,NA,NA,16), cex=1,lwd=3)}
  
  
{par(mar = c(5, 5, 3, 5))
plot(outmaf_ov2$end_adj,outmaf_ov2$V4,col='red',type='l',xlab='Nucleotide Position',ylab=expression("G"["accessability"]),main=paste0('Strucutre Upstream\n',p$ID))
par(new = TRUE)
plot(outmaf_ov2$end_adj,-outmaf_ov2$Zscore,type='l',col='purple',ylab = "", xlab = "", xaxt = "n", yaxt = "n",ylim=c(-3.5,3.5))
axis(side = 4)
abline(h=0,lty=2,col='purple')
abline(h=-1,lty=2,col='purple')
abline(h=1,lty=2,col='purple')
mtext(expression("-Z-score"), side = 4, line = 3)
text(max(outmaf_ov2$end_adj-100),3,"More Structured");  text(max(outmaf_ov2$end_adj-100),-3,"Less Structured")
lines(start:end,matrix(3,nrow=((end-start)+1),ncol=1)[,1],col='blue',type = "p")
lines(croslink,3,col='green',type = "p",pch=16)
legend("bottomleft", legend=c( expression("G"["accessability"]),'Z-Score',"CLIP peak",'5`end CLIP peak' ),
col=c( "red","purple","blue",'green'), lty=c(1,1,1,0),pch=c(NA,NA,NA,16), cex=1,lwd=3)}
}


if (p$strand%in%c("-")) {
{par(mar = c(5, 5, 3, 5))
plot(outmaf_ov2$nt,-outmaf_ov2$MFE,type='l',col='orange',xlab='Nucleotide Position',ylab=expression("-deltaG"),main=paste0('Strucutre Upstream\n',p$ID))
par(new = TRUE)
plot(outmaf_ov2$nt,-outmaf_ov2$Zscore,type='l',col='purple',ylab = "", xlab = "", xaxt = "n", yaxt = "n",ylim=c(-3.5,3.5))
axis(side = 4)
abline(h=0,lty=2,col='purple')
abline(h=-1,lty=2,col='purple')
abline(h=1,lty=2,col='purple')
mtext('-Z-score', side = 4, line = 3)
text(max(outmaf_ov2$nt-100),3,"More Structured");  text(max(outmaf_ov2$nt-100),-3,"Less Structured")
lines(start:end,matrix(3,nrow=((end-start)+1),ncol=1)[,1],col='blue',type = "p")
lines(croslink,3,col='green',type = "p",pch=16)
legend("bottomleft", legend=c("Gibbs free engergy", 'Z-score',"CLIP peak",'5`end CLIP peak' ),
col=c("orange", "purple","blue",'green'), lty=c(1,1,1,0),pch=c(NA,NA,NA,16), cex=1,lwd=3)}
  
  
{par(mar = c(5, 5, 3, 5))
plot(outmaf_ov2$nt,-outmaf_ov2$MFE,type='l',col='orange',xlab='Nucleotide Position',ylab=expression("-deltaG"),main=paste0('Strucutre Upstream\n',p$ID))
par(new = TRUE)
plot(outmaf_ov2$nt,outmaf_ov2$V4,type='l',col='red',ylab = "", xlab = "", xaxt = "n", yaxt = "n")
axis(side = 4)
mtext(expression("G"["accessability"]), side = 4, line = 3)
lines(start:end,matrix(max(outmaf_ov2$V4,na.rm = T)-5,nrow=((end-start)+1),ncol=1)[,1],col='blue',type = "p")
lines(croslink,max(outmaf_ov2$V4,na.rm = T)-5,col='green',type = "p",pch=16)
legend("bottomleft", legend=c("Gibbs free engergy", expression("G"["accessability"]),"CLIP peak",'5`end CLIP peak' ),
col=c("orange", "red","blue",'green'), lty=c(1,1,1,0),pch=c(NA,NA,NA,16), cex=1,lwd=3)}
  
  
{par(mar = c(5, 5, 3, 5))
plot(outmaf_ov2$nt,-outmaf_ov2$MFE,type='l',col='orange',xlab='Nucleotide Position',ylab=expression("-deltaG"),main=paste0('Strucutre Upstream\n',p$ID))
lines(outmaf_ov2$nt,(outmaf_ov2$V4)/1.3,col='red')
par(new = TRUE)
plot(outmaf_ov2$nt,-outmaf_ov2$Zscore,type='l',col='purple',ylab = "", xlab = "", xaxt = "n", yaxt = "n",ylim=c(-3.5,3.5))
axis(side = 4)
abline(h=0,lty=2,col='purple')
abline(h=-1,lty=2,col='purple')
abline(h=1,lty=2,col='purple')
mtext(expression("-Z-score"), side = 4, line = 3)
text(max(outmaf_ov2$nt-100),3,"More Structured");  text(max(outmaf_ov2$nt-100),-3,"Less Structured")
lines(start:end,matrix(3,nrow=((end-start)+1),ncol=1)[,1],col='blue',type = "p")
lines(croslink,3,col='green',type = "p",pch=16)
legend("bottomleft", legend=c("Gibbs free engergy", expression("G"["accessability"]),'Z-Score',"CLIP peak",'5`end CLIP peak' ),
col=c("orange", "red","purple","blue",'green'), lty=c(1,1,1,1,0),pch=c(NA,NA,NA,NA,16), cex=1,lwd=3)}
  
  
{par(mar = c(5, 5, 3, 5))
plot(outmaf_ov2$nt,outmaf_ov2$V4,col='red',type='l',xlab='Nucleotide Position',ylab=expression("G"["accessability"]),main=paste0('Strucutre Upstream\n',p$ID))
par(new = TRUE)
plot(outmaf_ov2$nt,-outmaf_ov2$Zscore,type='l',col='purple',ylab = "", xlab = "", xaxt = "n", yaxt = "n",ylim=c(-3.5,3.5))
axis(side = 4)
abline(h=0,lty=2,col='purple')
abline(h=-1,lty=2,col='purple')
abline(h=1,lty=2,col='purple')
mtext(expression("-Z-score"), side = 4, line = 3)
text(max(outmaf_ov2$nt-100),3,"More Structured");  text(max(outmaf_ov2$end_adj-100),-3,"Less Structured")
lines(start:end,matrix(3,nrow=((end-start)+1),ncol=1)[,1],col='blue',type = "p")
lines(croslink,3,col='green',type = "p",pch=16)
legend("bottomleft", legend=c( expression("G"["accessability"]),'Z-Score',"CLIP peak",'5`end CLIP peak' ),
col=c( "red","purple","blue",'green'), lty=c(1,1,1,0),pch=c(NA,NA,NA,16), cex=1,lwd=3)}
}

```

## WINDOW PAIRING PROBABILITY



<!-- Exons -->
```{r ,fig.asp=.5, fig.align='center',fig.height=3, fig.width=6,echo=F,warning=F,eval=T,include=F}
iso=read.delim('../mESC_102218/4-Analysis/Wolin.mECS_102218/inst/extdata/RSEM.isoforms.FPKM.all_samples.txt', header=T, sep="\t",stringsAsFactors = F)
iso$WT=rowMeans(iso[,c('WTQ3_FPKM','WTQ4_FPKM')])
iso$gene_id=removeVersion(iso$gene_id)
iso$transcript_id=removeVersion(iso$transcript_id)

gene=read.delim('../mESC_102218/4-Analysis/Wolin.mECS_102218/inst/extdata/RSEM.genes.FPKM.all_samples.txt', header=T, sep="\t",stringsAsFactors = F)
gene$WT=rowMeans(gene[,c('WTQ3_FPKM','WTQ4_FPKM')])
gene$gene_id=removeVersion(gene$gene_id)

canonical=fread("/Users/homanpj/OneDrive - National Institutes of Health/Resources/ref/mm10/Gencode_VM23/fromUCSC/KnownCanonical/KnownCanonical_GencodeM23_GRCm38.txt", header=T, sep="\t",stringsAsFactors = F,data.table=F)
  canonical$transcript=removeVersion(canonical$transcript)
  canonical$protein=removeVersion(canonical$protein)
 
 
  mm10all=fread("/Users/homanpj/OneDrive - National Institutes of Health/Resources/ref/mm10/Gencode_VM23/fromGencode/gencode.vM23.basic.annotation.gtf.txt", header=T, sep="\t",stringsAsFactors = F,data.table=F)
mm10all$transcript_id=removeVersion(mm10all$transcript_id)
mm10all$gene_id=removeVersion(mm10all$gene_id)

mm10_exon=mm10all[mm10all$feature%in%c("exon"),]



  Peaksdata4_chunkEx=Peaksdata4[Peaksdata4$`Both Strand: RNA_type_Classification`%in%c('protein_coding: exon'),]
Peaksdata4_chunkEx=separate(Peaksdata4_chunkEx,ID,into = c('chr','start'),sep=':',remove = F)
Peaksdata4_chunkEx=separate(Peaksdata4_chunkEx,start,into = c('start','end'),sep='-')
Peaksdata4_chunkEx$start=as.numeric(Peaksdata4_chunkEx$start)
Peaksdata4_chunkEx$end=as.numeric(Peaksdata4_chunkEx$end)
  
  
  isoout=as.data.frame(matrix(nrow = nrow(Peaksdata4_chunkEx),ncol=8));colnames(isoout)=c('ID','gene_id','maxiso','maxiso_count','maxiso2','maxiso2_count','diff',"IScanonical")
  for (e in 1:nrow(Peaksdata4_chunkEx)) {
    p=Peaksdata4_chunkEx[e,]
    g=p$`Same Strand: Host_gene_ensembl_id`
    
      if (length(strsplit(g,",")[[1]])>1) {
        g=(strsplit(g,",")[[1]]) 
        gexp=gene[gene$gene_id%in%g,]
        gexp=gexp[order(gexp$WT,decreasing = T),]
        if (nrow(gexp)>1) {
              if ((gexp[1,'WT']/gexp[2,'WT'])<1.5) {xxxxGene}
        g=gexp[1,'gene_id']
      }}  

    isog=iso[iso$gene_id%in%g,]
    isog=isog[order(isog$WT,decreasing = T),]
    isogMax=isog[1,'WT']
    isogMax2=isog[2,'WT']
    
   if  (is.na(isogMax2)==F&isogMax!=0){
      if (((isogMax-isogMax2)/isogMax)<.2) {
        isoout[e,'maxiso']=isog[1,'transcript_id']
        isoout[e,'maxiso_count']=isog[1,'WT']
        isoout[e,'maxiso2']=isog[2,'transcript_id']
        isoout[e,'maxiso2_count']=isog[2,'WT']
        isoout[e,'diff']=((isogMax-isogMax2)/isogMax)
      }}
  isoout[e,'ID']=p$ID
  isoout[e,'gene_id']=isog[1,'gene_id']
  isoout[e,'maxiso']=isog[1,'transcript_id']
  isoout[e,'maxiso_count']=isog[1,'WT']
  isoout[e,'IScanonical']=0
    i=isoout[e,c("maxiso","maxiso2"),drop=T]
    i=i[i%in%canonical$transcript]
      if (length(i)>0) {isoout[e,'IScanonical']=i}
  }
   isoout=merge(isoout,canonical[,c('transcript','protein')],by.x='gene_id',by.y='protein',all.x=T)

  isoout$USEtranscript=0
    ## if no close sexond use max
  isoout[is.na(isoout$maxiso2),'USEtranscript']=isoout[is.na(isoout$maxiso2),'maxiso']
    ## if close sexond  and one of top two is  Canonicol use Cannonical
  isoout[is.na(isoout$maxiso2)==F,'USEtranscript']=isoout[is.na(isoout$maxiso2)==F,'IScanonical']
    ## if close sexond  and one of top two is NOT Canonicol use Cannonical
  isoout[(isoout$USEtranscript)==0,c('USEtranscript')]=isoout[(isoout$USEtranscript)==0,c('transcript')]
```  


<!-- #### Exon Upstream Base Pairing Probability -->
```{r ,fig.asp=.5, fig.align='center',fig.height=3, fig.width=6,echo=F,warning=F,eval=T,include=F}

########################################################################################################################################################################################################################################################################################################################################################################################################################################################################################################################################################################################################################################################################################################################################################################################################################
  
  buffer=100
window=100
chsize=10

  chunkname3_exon=as.data.frame(matrix(nrow=1,ncol=nrow(Peaksdata4_chunkEx)+2))
  chunkout3_exon=as.data.frame(matrix(nrow=window,ncol=nrow(Peaksdata4_chunkEx)+2))
  chunkout3_exon[,1]=seq(1,window,1)
  chunkout3_max_exon=as.data.frame(matrix(nrow=window,ncol=nrow(Peaksdata4_chunkEx)+2))
  chunkout3_max_exon[,1]=seq(1,window,1)
seqout3_exon=as.data.frame(matrix(nrow=window,ncol=nrow(Peaksdata4_chunkEx)+2))
seqout3_exon[,1]=seq(1,window,1)
seqout3_max_exon=as.data.frame(matrix(nrow=window,ncol=nrow(Peaksdata4_chunkEx)+2))
seqout3_max_exon[,1]=seq(1,window,1)
  
  
    checkout=as.data.frame(matrix(nrow = nrow(Peaksdata4_chunkEx),ncol=14));colnames(checkout)=c('n','ID','SingleExon','crosslinkexon','s1exon','e1exon','lastexon','TranscChoise','numberTranscript','trancript','exon#','sequence','strand','deltaG')
    for (x in 1:nrow(isoout)) {
      if (x%in%seq(from=0,to=1000,by=25)){print(x/nrow(isoout))}
      
t=isoout[x,]
p=Peaksdata4_chunkEx[Peaksdata4_chunkEx$ID%in%t$ID,]

chr=p$chr
start=p$start
end=p$end
strand=p$strand

##################
if (strand%in%'+') {
croslink=start
s1=start-buffer#(selects base +1 of start site but ends at correct 5' site)
e1=start
}

if (strand%in%'-') {
croslink=end
s1=end-1#(selects base +1 of start site but ends at correct 5' site)
e1=end+buffer-1
}

  anno.GR=GRanges(seqnames=chr, ranges=IRanges(start=s1,end=e1),
                 strand =strand)
##################
   checkout[x,1]=x
   checkout[x,'ID']=p$ID
   
#### overlap with highest expresse
loc=mm10_exon[mm10_exon$transcript_id%in%t$USEtranscript,]
loc2=mm10all[mm10all$transcript_id%in%t$USEtranscript,]
loc2=mm10all[mm10all$gene_id%in%t$gene_id,]
loc2=mm10_exon[mm10_exon$gene_id%in%t$gene_id,]

if (nrow(loc2)==0) {checkout[x,'SingleExon']='transcript not in mm10'
  next}
s.GR = GRanges(seqnames=loc$seqname, ranges=IRanges(start=loc$start,end=loc$end),
                 strand=loc$strand)
  q =s.GR
  s=anno.GR
  xo=as.data.frame(GenomicRanges::findOverlaps(q,s,type = "any",ignore.strand=F))
  texn=loc[xo$queryHit,]
  checkout[x,'TranscChoise']="Highest Expression"
  checkout[x,'numberTranscript']=length(unique(texn$transcript_id))
  if (nrow(texn)>0) { 
    checkout[x,'trancript']=unique(texn$transcript_id)
    checkout[x,'exon#']=(texn$exon_number)
    checkout[x,'strand']=strand}
  
  remove(xo,loc,s.GR)

#### overlap with Canonical
if (nrow(texn)==0) {
  loc=mm10_exon[mm10_exon$gene_id%in%t$IScanonical,]
    if (length(unique(loc$transcript_id))>1) {xxxx1  }
  s.GR = GRanges(seqnames=loc$seqname, ranges=IRanges(start=loc$start,end=loc$end),
                 strand=loc$strand)
  q =s.GR
  s=anno.GR  
  xo=as.data.frame(GenomicRanges::findOverlaps(q,s,type = "any",ignore.strand=F))
  texn=loc[xo$queryHit,]
  checkout[x,'TranscChoise']="switch to Cannonical"
  checkout[x,'numberTranscript']=length(unique(texn$transcript_id))
  if (nrow(texn)>0) { 
    checkout[x,'trancript']=unique(texn$transcript_id)
    checkout[x,'exon#']=(texn$exon_number)
    checkout[x,'strand']=strand}

  remove(xo,loc,s.GR)
}
  
#### overlap with any
if (nrow(texn)==0) {
  loc=mm10_exon[mm10_exon$gene_id%in%t$gene_id,]
  s.GR = GRanges(seqnames=loc$seqname, ranges=IRanges(start=loc$start,end=loc$end),
                 strand=loc$strand)
  q =s.GR
  s=anno.GR
  xo=as.data.frame(GenomicRanges::findOverlaps(q,s,type = "any",ignore.strand=F))
  texn=loc[xo$queryHit,]
    checkout[x,'TranscChoise']="switch to other"
    checkout[x,'numberTranscript']=length(unique(texn$transcript_id))
    if (nrow(texn)>0) { 
      checkout[x,'trancript']=unique(texn$transcript_id)
      checkout[x,'exon#']=(texn$exon_number)
    checkout[x,'strand']=strand}
    remove(xo,loc,s.GR)
}
  
if (nrow(texn)==0) {
  checkout[x,'SingleExon']='does not cross exon'
  checkout[x,'TranscChoise']=NA
  checkout[x,'numberTranscript']=NA
  checkout[x,'trancript']=NA
  checkout[x,'exon#']=NA
  checkout[x,'sequence']=NA
  next}

  #### single exon Method if window starts and ends in entron or in chr1 crosslink (window 3') is in exon
  checkout[x,'SingleExon']=(texn$start<=s1&texn$end>=e1)|
                (strand=="+"&(texn$start>=s1&texn$end>=e1)&texn$exon_number==1)| 
                (strand=="-"&(texn$start<=s1&texn$end<=e1)&texn$exon_number==1)
  if ((strand=="+"&(texn$start>=s1&texn$end>=e1)&texn$exon_number==1)|
      (strand=="-"&(texn$start<=s1&texn$end<=e1)&texn$exon_number==1)) {checkout[x,'numberTranscript']=paste0(checkout[x,'numberTranscript'],"_out of bounds exon upstream only")}

  
  
    #### window region falls in exon
  checkout[x,'crosslinkexon']=texn$start<=croslink&texn$end>=croslink #crosslink
  checkout[x,'s1exon']=(strand=="+"&texn$start<=s1&texn$end>=s1)|(strand=="-"&texn$start<=e1&texn$end>=e1) # 5` end in exon
  checkout[x,'e1exon']=(strand=="+"&texn$start<=e1&texn$end>=e1)|(strand=="-"&texn$start<=s1&texn$end>=s1) # 3` end in exon
  checkout[x,'lastexon']=checkout[x,'exon#']/max(mm10_exon[mm10_exon$transcript_id%in%checkout[x,'trancript'],'exon_number'])
  
  if ((checkout[x,'SingleExon']==F)&(checkout[x,'exon#']>1)) {#run if not in single exon and not chr1 (if crosslink is in exon then upstream has to be upstream out of gene for exon1)
bedpos=matrix(ncol=6,nrow=2)

  locx=mm10_exon[mm10_exon$transcript_id%in%checkout[x,'trancript'],]
  locx=locx[order(locx$exon_number,decreasing = F),]
  if (strand=="+") {
    exns=locx[locx$exon_number%in%c(  checkout[x,'exon#'], as.character(as.numeric(checkout[x,'exon#'])-1)),]
    if (checkout[x,7,drop=F]==1) {next}
      
      # closest to peak
      bedpos[2,2]=exns[2,'start']-1
      bedpos[2,3]=croslink
      
      # Upstream exon
      bedpos[1,2]=exns[1,'end']-(window-abs(croslink-exns[2,'start']+1))
      bedpos[1,3]=exns[1,'end']
  }
  
   if (strand=="-") {
      exns=locx[locx$exon_number%in%c(  checkout[x,'exon#'], as.character(as.numeric(checkout[x,'exon#'])-1)),]
      if (checkout[x,'exon#']==1) {next}
      
      # closest to peak
      bedpos[2,2]=croslink-1
      bedpos[2,3]=exns[2,'end']
      
      # Upstream exon
      bedpos[1,2]=exns[1,'start']-1
      bedpos[1,3]=exns[1,'start']+(window-abs(croslink-exns[2,'end']-1))-1
   }
  

bedpos[,1]=chr
bedpos[,4]=".";bedpos[,5]="."
bedpos[,6]=strand

write.table(bedpos, file=paste0('./structure/chunkwindow/Exons/upstream/',p$ID,"_chunk.bed"), quote=F, sep="\t", row.names=F, col.names=F)
# system(paste0('bedtools getfasta -s -fi ',fasta,' -bed structure/chunkwindow/Exons/upstream/',p$ID,'_chunk.bed > structure/chunkwindow/Exons/upstream/',p$ID,'_chunk.fa'))
fa=system(paste0('bedtools getfasta -s -fi ',fasta,' -bed structure/chunkwindow/Exons/upstream/',p$ID,'_chunk.bed'),intern = T)
fa2=fa
fa2[1]=fa2[3]
fa2[2]=paste(fa[2],fa[4],collapse = "")
fa2[2]=gsub(" ","",fa2[2])
fa2=fa2[1:2]
if (nchar(fa[2])==window) {fa2=fa;checkout[x,'SingleExon']='Change-TRUE'}
if (nchar(fa2[2])!=window){xxxx}
fa=fa2
write.table(fa2, file=paste0('./structure/chunkwindow/Exons/upstream/',p$ID,"_chunk.fa"), quote=F, sep="\t", row.names=F, col.names=F)
  checkout[x,'sequence']=fa[2]
  remove(bedpos,fa2)
  }

if ((checkout[x,'SingleExon']==T)|(checkout[x,'exon#']==1)) {
  bedpos=matrix(ncol=6,nrow=1)
  bedpos[1,1]=chr
  bedpos[1,2]=s1
  bedpos[1,3]=e1
  bedpos[1,4]=".";bedpos[,5]="."
  bedpos[1,6]=strand
      
  write.table(bedpos, file=paste0('./structure/chunkwindow/Exons/upstream/',p$ID,"_chunk.bed"), quote=F, sep="\t", row.names=F, col.names=F)
# system(paste0('bedtools getfasta -s -fi ',fasta,' -bed structure/chunkwindow/Exons/upstream/',p$ID,'_chunk.bed > structure/chunkwindow/Exons/upstream/',p$ID,'_chunk.fa'))
fa=system(paste0('bedtools getfasta -s -fi ',fasta,' -bed structure/chunkwindow/Exons/upstream/',p$ID,'_chunk.bed'),intern = T)
  write.table(fa, file=paste0('./structure/chunkwindow/Exons/upstream/',p$ID,"_chunk.fa"), quote=F, sep="\t", row.names=F, col.names=F)
  checkout[x,'sequence']=fa[2]
    remove(bedpos)

}
  
 checkout$length=apply(checkout[,'sequence',drop=F], 1, nchar)
 
 
 ########################################################################################
  
  system(paste0('/Users/homanpj/Documents/Tools/ViennaRNA-2.4.14/bin/RNAfold --id-prefix ',paste0(p$chr,'_',s1,'-',e1), ' -p ./structure/chunkwindow/Exons/upstream/',p$ID,'_chunk.fa > out.fold.txt'))
system('mv *.ps ./structure/chunkwindow/Exons/upstream/')
system('mv out.fold.txt ./structure/chunkwindow/Exons/upstream/')
 
chunk=read.delim(paste0('./structure/chunkwindow/Exons/upstream/',p$chr,'_',s1,'-',e1,'_0001_dp.ps'),comment.char = "%",header = F,sep = "\t")
chunk=chunk[grep("ubox",chunk$V1),]
chunk=as.data.frame(chunk[-1],stringsAsFactors=F)
chunk$`chunk[-1]`=as.character(chunk$`chunk[-1]`);colnames(chunk)="V1"
chunk=separate(chunk,'V1',sep = " ",into = c('b1','b2','pvalue','ubox'))
chunk$pvalue=as.numeric(chunk$pvalue)
chunk$b1=as.numeric(chunk$b1)
chunk$b2=as.numeric(chunk$b2)

deltaG=read.delim(paste0('./structure/chunkwindow/Exons/upstream/out.fold.txt'),comment.char = "%",header = F,sep = "\t")
deltaG=as.data.frame(deltaG[3,])
deltaG=separate(deltaG,col = 1,into = c('str','DG'),sep = window+2)
deltaG$DG=gsub(")","",deltaG$DG)
deltaG$DG=as.numeric(deltaG$DG)
checkout[x,'deltaG']=deltaG$DG

###############
## BP probibilites for each chuck at each position
###############
for (y in 1:(window-9)) {
  ch=chunk[chunk$b1%in%seq(y,y+9,1)|chunk$b2%in%seq(y,y+9,1),]
  ch2=ch
    if (nrow(ch2)>0) {
   ch2$br1=rowMins(as.matrix(ch[,c('b1','b2')]))
    ch2$br2=rowMaxs(as.matrix(ch[,c('b1','b2')]))
      ch$b1=ch2$br1
      ch$b2=ch2$br2
      ch=ch[duplicated(ch)==F,]
  # colnames(chunkout2)[y+2]=paste0(y,'-',y+9)
      chunkout3_exon[y,1]=y
      chunkout3_exon[y,2]=paste0(y,'-',y+9)
      chunkout3_exon[y,x+2]=mean(ch$pvalue) ##average all basepairinng accross chunck
      
      chunkout3_max_exon[y,1]=y
      chunkout3_max_exon[y,2]=paste0(y,'-',y+9)
      chunkout3_max_exon[y,x+2]=max(ch$pvalue) ##average all basepairinng accross chunck
      
    }   
    if (nrow(ch2)==0) {
      chunkout3_exon[y,1]=y
      chunkout3_exon[y,2]=paste0(y,'-',y+9)
      chunkout3_exon[y,x+2]=0
      chunkout3_max_exon[y,1]=y
      chunkout3_max_exon[y,2]=paste0(y,'-',y+9)
      chunkout3_max_exon[y,x+2]=0}  
      
            chunkname3_exon[1,x+2]=p$ID
}

### count each base
  for (y in 1:(window)) {
  seq=chunk[chunk$b1%in%(y)|chunk$b2%in%(y),]
  seq2=seq
  
  if (nrow(seq2)>0) {
    seq2$br1=rowMins(as.matrix(seq[,c('b1','b2')]))
    seq2$br2=rowMaxs(as.matrix(seq[,c('b1','b2')]))
  seq$b1=seq2$br1
  seq$b2=seq2$br2
  seq=seq[duplicated(seq)==F,]
  # colnames(chunkout2)[y+2]=paste0(y,'-',y+9)
      seqout3_exon[y,1]=y
      # seqout3_exon[y,2]=paste0(y)
      seqout3_exon[y,x+2]=mean(seq$pvalue) ##average all basepairinng accross chunck
      
      seqout3_max_exon[y,1]=y
      # seqout3_max_exon[y,2]=paste0(y)
      seqout3_max_exon[y,x+2]=max(seq$pvalue) ##average all basepairinng accross chunck
  }    
   if (nrow(seq2)==0) {
      seqout3_exon[y,1]=y
      seqout3_exon[y,x+2]=0
      seqout3_max_exon[y,1]=y
      seqout3_max_exon[y,x+2]=0}
}


# paste0('./structure/chunkwindow/Exons/upstream/',p$chr,'_',s1,'-',e1,'_0001_dp.ps')


if (x<2) {chunkout2_matout=chunkout2_mat}
if (x>1) {chunkout2_matout=(chunkout2_matout+chunkout2_mat)}
######################################################################
if(file.exists(paste0('./structure/chunkwindow/Exons/upstream/',p$ID,'_chunk.fa'))){file.remove(paste0('./structure/chunkwindow/Exons/upstream/',p$ID,'_chunk.fa'))}
if(file.exists(paste0('./structure/chunkwindow/Exons/upstream/',p$ID,'_chunk.bed'))){file.remove(paste0('./structure/chunkwindow/Exons/upstream/',p$ID,'_chunk.bed'))}
if(file.exists(paste0('./structure/chunkwindow/Exons/upstream/',p$chr,'_',s1,'-',e1,'_0001_dp.ps'))){file.remove(paste0('./structure/chunkwindow/Exons/upstream/',p$chr,'_',s1,'-',e1,'_0001_dp.ps'))}
if(file.exists(paste0('./structure/chunkwindow/Exons/upstream/',p$chr,'_',s1,'-',e1,'_0001_ss.ps'))){file.remove(paste0('./structure/chunkwindow/Exons/upstream/',p$chr,'_',s1,'-',e1,'_0001_ss.ps'))}
if(file.exists(paste0('./structure/chunkwindow/Exons/upstream/out.fold.txt'))){file.remove(paste0('./structure/chunkwindow/Exons/upstream/out.fold.txt'))}

if(file.exists(paste0('./structure/chunkwindow/Exons/upstream/',p$chr,'_',s1,'-',e1,'\\(',strand,'\\)_dp.ps.mntn'))){file.remove(paste0('./structure/chunkwindow/Exons/upstream/',p$chr,'_',s1,'-',e1,'\\(',strand,'\\)_dp.ps.mntn'))}

if(file.exists(paste0(p$chr,'_',s1,'-',e1,'(',strand,').fold'))){file.remove(paste0(p$chr,'_',s1,'-',e1,'(',strand,').fold'))}
######################################################################
remove(p,t,fa)
    }


chunkname3_exon=t(chunkname3_exon)

######################
colnames(chunkout3_exon)[3:ncol(chunkout3_exon)]=chunkname3_exon[3:nrow(chunkname3_exon),1]

chunkout3_exon=chunkout3_exon[is.na(chunkout3_exon[,3])==F,]
chunkout3_exon[is.infinite(as.matrix(chunkout3_exon))==T]=0


colnames(chunkout3_max_exon)[3:ncol(chunkout3_max_exon)]=chunkname3_exon[3:nrow(chunkname3_exon),1]

chunkout3_max_exon=chunkout3_max_exon[is.na(chunkout3_max_exon[,3])==F,]
chunkout3_max_exon[is.infinite(as.matrix(chunkout3_max_exon))==T]=0

######################
colnames(seqout3_exon)[3:ncol(seqout3_exon)]=chunkname3_exon[3:nrow(chunkname3_exon),1]

seqout3_exon=seqout3_exon[is.na(seqout3_exon[,3])==F,]
seqout3_exon[is.infinite(as.matrix(seqout3_exon))==T]=0

colnames(seqout3_max_exon)[3:ncol(seqout3_max_exon)]=chunkname3_exon[3:nrow(chunkname3_exon),1]

seqout3_max_exon=seqout3_max_exon[is.na(seqout3_max_exon[,3])==F,]
seqout3_max_exon[is.infinite(as.matrix(seqout3_max_exon))==T]=0

######################

phex=as.matrix(chunkout3_exon[,-c(1:2)])
phex=(chunkout3_exon[,-c(1:2)])

phex=phex[,is.na(phex[1,])==F]
phex=phex[is.na(phex[,3])==F,]
phex[is.nan(as.matrix(phex))==T]=0
phex[is.infinite(as.matrix(phex))==T]=0
  phex=t(phex)
  phex=as.data.frame(phex)

  #### sort by row diff
# phex=phex[order(rowMaxs(as.matrix(phex))-rowMins(as.matrix(phex)),decreasing = T),]
phex=phex[order(rowMaxs(as.matrix(phex)),decreasing = T),]
 colnames(phex)=c(-window:(ncol(phex)-(window+1)))
 # phex= cbind(rowMaxs(as.matrix(phex))-rowMins(as.matrix(phex)),phex);colnames(phex)[1]='diff'
 phex= cbind(rowMaxs(as.matrix(phex)),phex);colnames(phex)[1]='diff'

write.table(phex, file=paste0('./structure/chunkwindow/Exons/upstream/phex.txt'), quote=F, sep="\t", row.names=T, col.names=NA)

######

phex_max=as.matrix(chunkout3_max_exon[,-c(1:2)])
phex_max=(chunkout3_max_exon[,-c(1:2)])

phex_max=phex_max[,is.na(phex_max[1,])==F]
phex_max=phex_max[is.na(phex_max[,3])==F,]
phex_max[is.nan(as.matrix(phex_max))==T]=0
phex_max[is.infinite(as.matrix(phex_max))==T]=0
  phex_max=t(phex_max)
  phex_max=as.data.frame(phex_max)

  #### sort by row diff
# phex_max=phex_max[order(rowMaxs(as.matrix(phex_max))-rowMins(as.matrix(phex_max)),decreasing = T),]
phex_max=phex_max[order(rowMaxs(as.matrix(phex_max)),decreasing = T),]
 colnames(phex_max)=c(-window:(ncol(phex_max)-(window+1)))
 # phex_max= cbind(rowMaxs(as.matrix(phex_max))-rowMins(as.matrix(phex_max)),phex_max);colnames(phex_max)[1]='diff'
 phex_max= cbind(rowMaxs(as.matrix(phex_max)),phex_max);colnames(phex_max)[1]='diff'

write.table(phex_max, file=paste0('./structure/chunkwindow/Exons/upstream/phex_max.txt'), quote=F, sep="\t", row.names=T, col.names=NA)

#######################################################

phexseq=seqout3_exon
  phexseq=as.data.frame(phexseq[,-c(1:2)])
  phexseq=t(phexseq)
  phexseq[is.na(phexseq)]=0
phexseq=phexseq[order(rowMaxs(as.matrix(phexseq)),decreasing = T),]
 colnames(phexseq)=c(-window:(ncol(phexseq)-(window+1)))
 # phexseq= cbind(rowMaxs(as.matrix(phexseq))-rowMins(as.matrix(phexseq)),phexseq);colnames(phexseq)[1]='diff'
 phexseq= cbind(rowMaxs(as.matrix(phexseq)),phexseq);colnames(phexseq)[1]='diff'

write.table(phexseq, file=paste0('./structure/chunkwindow/Exons/upstream/phexseq.txt'), quote=F, sep="\t", row.names=T, col.names=NA)

phexseq_max=seqout3_max_exon
  phexseq_max=as.data.frame(phexseq_max[,-c(1:2)])
  phexseq_max=t(phexseq_max)
  phexseq_max[is.na(phexseq_max)]=0
phexseq_max=phexseq_max[order(rowMaxs(as.matrix(phexseq_max)),decreasing = T),]
 colnames(phexseq_max)=c(-window:(ncol(phexseq_max)-(window+1)))
 # phexseq_max= cbind(rowMaxs(as.matrix(phexseq_max))-rowMins(as.matrix(phexseq_max)),phexseq_max);colnames(phexseq_max)[1]='diff'
 phexseq_max= cbind(rowMaxs(as.matrix(phexseq_max)),phexseq_max);colnames(phexseq_max)[1]='diff'

write.table(phexseq_max, file=paste0('./structure/chunkwindow/Exons/upstream/phexseq_max.txt'), quote=F, sep="\t", row.names=T, col.names=NA)


```


<!-- #### Exon Downstream Base Pairing Probability -->
```{r ,fig.asp=.5, fig.align='center',fig.height=3, fig.width=6,echo=F,warning=F,eval=T,include=F}
  
########################################################################################################################################################################################################################################################################################################################################################################################################################################################################################################################################################################################################################################################################################################################################################################################################################
  

  chunkname3_exondn=as.data.frame(matrix(nrow=1,ncol=nrow(Peaksdata4_chunkEx)+2))
  chunkout3_exondn=as.data.frame(matrix(nrow=window,ncol=nrow(Peaksdata4_chunkEx)+2))
  chunkout3_exondn[,1]=seq(1,window,1)
  chunkout3_max_exondn=as.data.frame(matrix(nrow=window,ncol=nrow(Peaksdata4_chunkEx)+2))
  chunkout3_max_exondn[,1]=seq(1,window,1)
seqout3_exondn=as.data.frame(matrix(nrow=window,ncol=nrow(Peaksdata4_chunkEx)+2))
seqout3_exondn[,1]=seq(1,window,1)
seqout3_max_exondn=as.data.frame(matrix(nrow=window,ncol=nrow(Peaksdata4_chunkEx)+2))
seqout3_max_exondn[,1]=seq(1,window,1)
  
  
    checkout_dn=as.data.frame(matrix(nrow = nrow(Peaksdata4_chunkEx),ncol=13));colnames(checkout_dn)=c('n','ID','SingleExon','crosslinkexon','s1exon','e1exon','lastexon','TranscChoise','numberTranscript','trancript','exon#','sequence','strand')
    for (x in 1:nrow(isoout)) {
    # for (x in 80) {
      if (x%in%seq(from=0,to=1000,by=25)){print(x/nrow(isoout))}
      
t=isoout[x,]
p=Peaksdata4_chunkEx[Peaksdata4_chunkEx$ID%in%t$ID,]

chr=p$chr
start=p$start
end=p$end
strand=p$strand

##################
if (strand%in%'+') {
croslink=start
s1=start-1
e1=start+buffer-1#(selects base +1 of start site but ends at correct 5' site)

}

if (strand%in%'-') {
croslink=end
s1=end-buffer
e1=end#(selects base +1 of start site but ends at correct 5' site)

}

  anno.GR=GRanges(seqnames=chr, ranges=IRanges(start=s1,end=e1),
                 strand =strand)
##################
   checkout_dn[x,1]=x
   checkout_dn[x,'ID']=p$ID
   
#### overlap with highest expresse
loc=mm10_exon[mm10_exon$transcript_id%in%t$USEtranscript,]
loc2=mm10all[mm10all$transcript_id%in%t$USEtranscript,]
loc2=mm10all[mm10all$gene_id%in%t$gene_id,]
loc2=mm10_exon[mm10_exon$gene_id%in%t$gene_id,]
 
if (nrow(loc2)==0) {checkout_dn[x,'SingleExon']='transcript not in mm10'
  next}
s.GR = GRanges(seqnames=loc$seqname, ranges=IRanges(start=loc$start,end=loc$end),
                 strand=loc$strand)
  q =s.GR
  s=anno.GR
  xo=as.data.frame(GenomicRanges::findOverlaps(q,s,type = "any",ignore.strand=F))
  texn=loc[xo$queryHit,]
  checkout_dn[x,'TranscChoise']="Highest Expression"
  checkout_dn[x,'numberTranscript']=length(unique(texn$transcript_id))
  if (nrow(texn)>0) { 
    checkout_dn[x,'trancript']=unique(texn$transcript_id)
    checkout_dn[x,'exon#']=(texn$exon_number)
    checkout_dn[x,'strand']=strand}
  
  remove(xo,loc,s.GR)

#### overlap with Canonical
if (nrow(texn)==0) {
  loc=mm10_exon[mm10_exon$gene_id%in%t$IScanonical,]
    if (length(unique(loc$transcript_id))>1) {xxxx1  }
  s.GR = GRanges(seqnames=loc$seqname, ranges=IRanges(start=loc$start,end=loc$end),
                 strand=loc$strand)
  q =s.GR
  s=anno.GR  
  xo=as.data.frame(GenomicRanges::findOverlaps(q,s,type = "any",ignore.strand=F))
  texn=loc[xo$queryHit,]
  checkout_dn[x,'TranscChoise']="switch to Cannonical"
  checkout_dn[x,'numberTranscript']=length(unique(texn$transcript_id))
  if (nrow(texn)>0) { 
    checkout_dn[x,'trancript']=unique(texn$transcript_id)
    checkout_dn[x,'exon#']=(texn$exon_number)
    checkout_dn[x,'strand']=strand}

  remove(xo,loc,s.GR)
}
  
#### overlap with any
if (nrow(texn)==0) {
  loc=mm10_exon[mm10_exon$gene_id%in%t$gene_id,]
  s.GR = GRanges(seqnames=loc$seqname, ranges=IRanges(start=loc$start,end=loc$end),
                 strand=loc$strand)
  q =s.GR
  s=anno.GR
  xo=as.data.frame(GenomicRanges::findOverlaps(q,s,type = "any",ignore.strand=F))
  texn=loc[xo$queryHit,]
  
  if (nrow(texn)>1){texn=texn[texn$transcript_id%in%t$transcript,]}
    checkout_dn[x,'TranscChoise']="switch to other"
    checkout_dn[x,'numberTranscript']=length(unique(texn$transcript_id))
    if (nrow(texn)>0) { 
      checkout_dn[x,'trancript']=unique(texn$transcript_id)
      checkout_dn[x,'exon#']=(texn$exon_number)
    checkout_dn[x,'strand']=strand}
    remove(xo,loc,s.GR)
}
  
if (nrow(texn)==0) {
  checkout_dn[x,'SingleExon']='does not cross exon'
  checkout_dn[x,'TranscChoise']=NA
  checkout_dn[x,'numberTranscript']=NA
  checkout_dn[x,'trancript']=NA
  checkout_dn[x,'exon#']=NA
  checkout_dn[x,'sequence']=NA
  next}

#### single exon if window starts and ends in entron or in chr1 crosslink (window 3') is in exon
  checkout_dn[x,'SingleExon']=(texn$start<=s1&texn$end>=e1)|
                (strand=="+"&(texn$start>=s1&texn$end>=e1)&texn$exon_number==1)|
                (strand=="-"&(texn$start<=s1&texn$end<=e1)&texn$exon_number==1)
  if ((strand=="+"&texn$start>=s1&texn$end>=e1&texn$exon_number==1)|
      (strand=="-"&texn$start<=s1&texn$end<=e1&texn$exon_number==1)) {checkout_dn[x,'numberTranscript']=paste0(checkout_dn[x,'numberTranscript'],"_Crosslink is out of bounds exon 1")}
  
    #### window region falls in exon
  checkout_dn[x,'crosslinkexon']=texn$start<=croslink&texn$end>=croslink #crosslink
  checkout_dn[x,'s1exon']=(strand=="+"&texn$start<=s1&texn$end>=s1)|(strand=="-"&texn$start<=e1&texn$end>=e1) # 5` end in exon
  checkout_dn[x,'e1exon']=(strand=="+"&texn$start<=e1&texn$end>=e1)|(strand=="-"&texn$start<=s1&texn$end>=s1) # 3` end in exon
  checkout_dn[x,'lastexon']=checkout_dn[x,'exon#']/max(mm10_exon[mm10_exon$transcript_id%in%checkout_dn[x,'trancript'],'exon_number'])
  
    locx=mm10_exon[mm10_exon$transcript_id%in%checkout_dn[x,'trancript'],]
  locx=locx[order(locx$exon_number,decreasing = F),]
  
  ##         not contained in exon               not Last exon                        start of window is in exon (if start not in exon then Upstream overlap)
  if (((checkout_dn[x,'SingleExon']==F)&(checkout_dn[x,'lastexon']<1)&(checkout_dn[x,'s1exon']==T))) {
bedpos=matrix(ncol=6,nrow=2)

  if (strand=="+") {
    exns=locx[locx$exon_number%in%c(  checkout_dn[x,'exon#'], as.character(as.numeric(checkout_dn[x,'exon#'])+1)),]
    if (nrow(exns)<2) {next}
       
      # closest to peak
      bedpos[2,2]=croslink-1
      bedpos[2,3]=exns[exns$exon_number%in%checkout_dn[x,'exon#'],'end']
      
      # downstream exon
      bedpos[1,2]=exns[(exns$exon_number%in%checkout_dn[x,'exon#'])==F,'start']-1
      bedpos[1,3]=exns[(exns$exon_number%in%checkout_dn[x,'exon#'])==F,'start']+(window-abs(croslink-exns[exns$exon_number%in%checkout_dn[x,'exon#'],'end']-1))-1
  }
  
   if (strand=="-") {
      exns=locx[locx$exon_number%in%c(  checkout_dn[x,'exon#'], as.character(as.numeric(checkout_dn[x,'exon#'])+1)),]
    if (nrow(exns)<2) {next}
      
      # closest to peak
      bedpos[2,2]=exns[exns$exon_number%in%checkout_dn[x,'exon#'],'start']-1
      bedpos[2,3]=croslink
      
      # downstream exon
      bedpos[1,2]= exns[(exns$exon_number%in%checkout_dn[x,'exon#'])==F,'end']-(window-abs(croslink-exns[(exns$exon_number%in%checkout_dn[x,'exon#']),'start'])-1)
      bedpos[1,3]=exns[(exns$exon_number%in%checkout_dn[x,'exon#'])==F,'end']
   }

bedpos[,1]=chr
bedpos[,4]=".";bedpos[,5]="."
bedpos[,6]=strand

exns[exns$exon_number%in%checkout_dn[x,'exon#'],]

write.table(bedpos, file=paste0('./structure/chunkwindow/Exons/downstream/',p$ID,"_chunk.bed"), quote=F, sep="\t", row.names=F, col.names=F)
# system(paste0('bedtools getfasta -s -fi ',fasta,' -bed structure/chunkwindow/Exons/downstream/',p$ID,'_chunk.bed > structure/chunkwindow/Exons/downstream/',p$ID,'_chunk.fa'))
fa=system(paste0('bedtools getfasta -s -fi ',fasta,' -bed structure/chunkwindow/Exons/downstream/',p$ID,'_chunk.bed'),intern = T)
fa2=fa
fa2[1]=fa2[3]
fa2[2]=paste(fa[4],fa[2],collapse = "")
fa2[2]=gsub(" ","",fa2[2])
fa2=fa2[1:2]
if (nchar(fa[2])==window) {fa2=fa;checkout_dn[x,'SingleExon']='Change-TRUE'}
if (nchar(fa2[2])!=window){xxxx}
fa=fa2
write.table(fa2, file=paste0('./structure/chunkwindow/Exons/downstream/',p$ID,"_chunk.fa"), quote=F, sep="\t", row.names=F, col.names=F)
  checkout_dn[x,'sequence']=fa[2]
  remove(bedpos,fa2)
  }

  
 ##        contained in exon               or last Last exon           with       start of window is in exon (if start not in exon then Upstream overlap)
if ((checkout_dn[x,'SingleExon']==T)|((checkout_dn[x,'lastexon']==1)&(checkout_dn[x,'s1exon']==T))) {  
  checkout_dn[x,'numberTranscript']=paste0(checkout_dn[x,'numberTranscript'],'_lastexon or internal exon')
  bedpos=matrix(ncol=6,nrow=1)
  bedpos[1,1]=chr
  bedpos[1,2]=s1
  bedpos[1,3]=e1
  bedpos[1,4]=".";bedpos[,5]="."
  bedpos[1,6]=strand
      
  write.table(bedpos, file=paste0('./structure/chunkwindow/Exons/downstream/',p$ID,"_chunk.bed"), quote=F, sep="\t", row.names=F, col.names=F)
# system(paste0('bedtools getfasta -s -fi ',fasta,' -bed structure/chunkwindow/Exons/downstream/',p$ID,'_chunk.bed > structure/chunkwindow/Exons/downstream/',p$ID,'_chunk.fa'))
fa=system(paste0('bedtools getfasta -s -fi ',fasta,' -bed structure/chunkwindow/Exons/downstream/',p$ID,'_chunk.bed'),intern = T)
  write.table(fa, file=paste0('./structure/chunkwindow/Exons/downstream/',p$ID,"_chunk.fa"), quote=F, sep="\t", row.names=F, col.names=F)
  checkout_dn[x,'sequence']=fa[2]
    remove(bedpos)

}
  
  ##         not conained in exon               start not in exon                        end in exon
if ((checkout_dn[x,'SingleExon']==F)&(checkout_dn[x,'s1exon']==F)&(checkout_dn[x,'e1exon']==T)) {  
  checkout_dn[x,'numberTranscript']=paste0(checkout_dn[x,'numberTranscript'],'_should be intron')
  bedpos=matrix(ncol=6,nrow=1)
  bedpos[1,1]=chr
  bedpos[1,2]=s1
  bedpos[1,3]=e1
  bedpos[1,4]=".";bedpos[,5]="."
  bedpos[1,6]=strand
      
  write.table(bedpos, file=paste0('./structure/chunkwindow/Exons/downstream/',p$ID,"_chunk.bed"), quote=F, sep="\t", row.names=F, col.names=F)
# system(paste0('bedtools getfasta -s -fi ',fasta,' -bed structure/chunkwindow/Exons/downstream/',p$ID,'_chunk.bed > structure/chunkwindow/Exons/downstream/',p$ID,'_chunk.fa'))
fa=system(paste0('bedtools getfasta -s -fi ',fasta,' -bed structure/chunkwindow/Exons/downstream/',p$ID,'_chunk.bed'),intern = T)
  write.table(fa, file=paste0('./structure/chunkwindow/Exons/downstream/',p$ID,"_chunk.fa"), quote=F, sep="\t", row.names=F, col.names=F)
    checkout_dn[x,'sequence']=fa[2]

    remove(bedpos)

}
  
  
 checkout_dn$length=apply(checkout_dn[,'sequence',drop=F], 1, nchar)
 
 
 ########################################################################################
  
      system(paste0('/Users/homanpj/Documents/Tools/ViennaRNA-2.4.14/bin/RNAfold --id-prefix ',paste0(p$chr,'_',s1,'-',e1), ' -p ./structure/chunkwindow/Exons/downstream/',p$ID,'_chunk.fa > out.fold_dn.txt'))
system('mv *.ps ./structure/chunkwindow/Exons/downstream/')
system('mv out.fold_dn.txt ./structure/chunkwindow/Exons/downstream/')
 
chunk=read.delim(paste0('./structure/chunkwindow/Exons/downstream/',p$chr,'_',s1,'-',e1,'_0001_dp.ps'),comment.char = "%",header = F,sep = "\t")
chunk=chunk[grep("ubox",chunk$V1),]
chunk=as.data.frame(chunk[-1],stringsAsFactors=F)
chunk$`chunk[-1]`=as.character(chunk$`chunk[-1]`);colnames(chunk)="V1"
chunk=separate(chunk,'V1',sep = " ",into = c('b1','b2','pvalue','ubox'))
chunk$pvalue=as.numeric(chunk$pvalue)
chunk$b1=as.numeric(chunk$b1)
chunk$b2=as.numeric(chunk$b2)

deltaG=read.delim(paste0('./structure/chunkwindow/Exons/downstream/out.fold_dn.txt'),comment.char = "%",header = F,sep = "\t")
deltaG=as.data.frame(deltaG[3,])
deltaG=separate(deltaG,col = 1,into = c('str','DG'),sep = window+2)
deltaG$DG=gsub(")","",deltaG$DG)
deltaG$DG=as.numeric(deltaG$DG)
checkout_dn[x,'deltaG']=deltaG$DG

###############
## BP probibilites for each chuck at each position
###############
for (y in 1:(window-9)) {
  ch=chunk[chunk$b1%in%seq(y,y+9,1)|chunk$b2%in%seq(y,y+9,1),]
  ch2=ch
    if (nrow(ch2)>0) {
   ch2$br1=rowMins(as.matrix(ch[,c('b1','b2')]))
    ch2$br2=rowMaxs(as.matrix(ch[,c('b1','b2')]))
      ch$b1=ch2$br1
      ch$b2=ch2$br2
      ch=ch[duplicated(ch)==F,]
  # colnames(chunkout2)[y+2]=paste0(y,'-',y+9)
      chunkout3_exondn[y,1]=y
      chunkout3_exondn[y,2]=paste0(y,'-',y+9)
      chunkout3_exondn[y,x+2]=mean(ch$pvalue) ##average all basepairinng accross chunck
      
      chunkout3_max_exondn[y,1]=y
      chunkout3_max_exondn[y,2]=paste0(y,'-',y+9)
      chunkout3_max_exondn[y,x+2]=max(ch$pvalue) ##average all basepairinng accross chunck
      
    }   
    if (nrow(ch2)==0) {
      chunkout3_exondn[y,1]=y
      chunkout3_exondn[y,2]=paste0(y,'-',y+9)
      chunkout3_exondn[y,x+2]=0
      chunkout3_max_exondn[y,1]=y
      chunkout3_max_exondn[y,2]=paste0(y,'-',y+9)
      chunkout3_max_exondn[y,x+2]=0}  
      
            chunkname3_exondn[1,x+2]=p$ID
}

### count each base
  for (y in 1:(window)) {
  seq=chunk[chunk$b1%in%(y)|chunk$b2%in%(y),]
  seq2=seq
  
  if (nrow(seq2)>0) {
    seq2$br1=rowMins(as.matrix(seq[,c('b1','b2')]))
    seq2$br2=rowMaxs(as.matrix(seq[,c('b1','b2')]))
  seq$b1=seq2$br1
  seq$b2=seq2$br2
  seq=seq[duplicated(seq)==F,]
  # colnames(chunkout2)[y+2]=paste0(y,'-',y+9)
      seqout3_exondn[y,1]=y
      # seqout3_exondn[y,2]=paste0(y)
      seqout3_exondn[y,x+2]=mean(seq$pvalue) ##average all basepairinng accross chunck
      
      seqout3_max_exondn[y,1]=y
      # seqout3_exondn[y,2]=paste0(y)
      seqout3_max_exondn[y,x+2]=max(seq$pvalue) ##average all basepairinng accross chunck
  }    
   if (nrow(seq2)==0) {
      seqout3_exondn[y,1]=y
      seqout3_exondn[y,x+2]=0
      seqout3_max_exondn[y,1]=y
      seqout3_max_exondn[y,x+2]=0}
}

paste0('./structure/chunkwindow/Exons/downstream/',p$chr,'_',s1,'-',e1,'_0001_dp.ps')


######################################################################
if(file.exists(paste0('./structure/chunkwindow/Exons/downstream/',p$ID,'_chunk.fa'))){file.remove(paste0('./structure/chunkwindow/Exons/downstream/',p$ID,'_chunk.fa'))}
if(file.exists(paste0('./structure/chunkwindow/Exons/downstream/',p$ID,'_chunk.bed'))){file.remove(paste0('./structure/chunkwindow/Exons/downstream/',p$ID,'_chunk.bed'))}
if(file.exists(paste0('./structure/chunkwindow/Exons/downstream/',p$chr,'_',s1,'-',e1,'_0001_dp.ps'))){file.remove(paste0('./structure/chunkwindow/Exons/downstream/',p$chr,'_',s1,'-',e1,'_0001_dp.ps'))}
if(file.exists(paste0('./structure/chunkwindow/Exons/downstream/',p$chr,'_',s1,'-',e1,'_0001_ss.ps'))){file.remove(paste0('./structure/chunkwindow/Exons/downstream/',p$chr,'_',s1,'-',e1,'_0001_ss.ps'))}
if(file.exists(paste0('./structure/chunkwindow/Exons/downstream/out.fold_dn.txt'))){file.remove(paste0('./structure/chunkwindow/Exons/downstream/out.fold_dn.txt'))}
if(file.exists(paste0('./structure/chunkwindow/Exons/downstream/',p$chr,'_',s1,'-',e1,'\\(',strand,'\\)_dp.ps.mntn'))){file.remove(paste0('./structure/chunkwindow/Exons/downstream/',p$chr,'_',s1,'-',e1,'\\(',strand,'\\)_dp.ps.mntn'))}

if(file.exists(paste0(p$chr,'_',s1,'-',e1,'(',strand,').fold'))){file.remove(paste0(p$chr,'_',s1,'-',e1,'(',strand,').fold'))}
######################################################################
remove(p,t,fa)
    }


chunkname3_exondn=t(chunkname3_exondn)

######################
colnames(chunkout3_exondn)[3:ncol(chunkout3_exondn)]=chunkname3_exondn[3:nrow(chunkname3_exondn),1]

chunkout3_exondn=chunkout3_exondn[is.na(chunkout3_exondn[,3])==F,]
chunkout3_exondn[is.infinite(as.matrix(chunkout3_exondn))==T]=0


colnames(chunkout3_max_exondn)[3:ncol(chunkout3_max_exondn)]=chunkname3_exondn[3:nrow(chunkname3_exondn),1]

chunkout3_max_exondn=chunkout3_max_exondn[is.na(chunkout3_max_exondn[,3])==F,]
chunkout3_max_exondn[is.infinite(as.matrix(chunkout3_max_exondn))==T]=0

######################
colnames(seqout3_exondn)[3:ncol(seqout3_exondn)]=chunkname3_exondn[3:nrow(chunkname3_exondn),1]

seqout3_exondn=seqout3_exondn[is.na(seqout3_exondn[,3])==F,]
seqout3_exondn[is.infinite(as.matrix(seqout3_exondn))==T]=0

colnames(seqout3_max_exondn)[3:ncol(seqout3_max_exondn)]=chunkname3_exondn[3:nrow(chunkname3_exondn),1]

seqout3_max_exondn=seqout3_max_exondn[is.na(seqout3_max_exondn[,3])==F,]
seqout3_max_exondn[is.infinite(as.matrix(seqout3_max_exondn))==T]=0

######################

phex_dn=as.matrix(chunkout3_exondn[,-c(1:2)])
phex_dn=(chunkout3_exondn[,-c(1:2)])

phex_dn=phex_dn[,is.na(phex_dn[1,])==F]
phex_dn=phex_dn[is.na(phex_dn[,3])==F,]
phex_dn[is.nan(as.matrix(phex_dn))==T]=0
phex_dn[is.infinite(as.matrix(phex_dn))==T]=0
  phex_dn=t(phex_dn)
  phex_dn=as.data.frame(phex_dn)

  #### sort by row diff
# phex_dn=phex_dn[order(rowMaxs(as.matrix(phex_dn))-rowMins(as.matrix(phex_dn)),decreasing = T),]
phex_dn=phex_dn[order(rowMaxs(as.matrix(phex_dn)),decreasing = T),]
 colnames(phex_dn)=c(-window:(ncol(phex_dn)-(window+1)))
 # phex_dn= cbind(rowMaxs(as.matrix(phex_dn))-rowMins(as.matrix(phex_dn)),phex_dn);colnames(phex_dn)[1]='diff'
 phex_dn= cbind(rowMaxs(as.matrix(phex_dn)),phex_dn);colnames(phex_dn)[1]='diff'

write.table(phex_dn, file=paste0('./structure/chunkwindow/Exons/downstream/phex_dn.txt'), quote=F, sep="\t", row.names=T, col.names=NA)

######

phex_max_dn=as.matrix(chunkout3_max_exondn[,-c(1:2)])
phex_max_dn=(chunkout3_max_exondn[,-c(1:2)])

phex_max_dn=phex_max_dn[,is.na(phex_max_dn[1,])==F]
phex_max_dn=phex_max_dn[is.na(phex_max_dn[,3])==F,]
phex_max_dn[is.nan(as.matrix(phex_max_dn))==T]=0
phex_max_dn[is.infinite(as.matrix(phex_max_dn))==T]=0
  phex_max_dn=t(phex_max_dn)
  phex_max_dn=as.data.frame(phex_max_dn)

  #### sort by row diff
# phex_max_dn=phex_max_dn[order(rowMaxs(as.matrix(phex_max_dn))-rowMins(as.matrix(phex_max_dn)),decreasing = T),]
phex_max_dn=phex_max_dn[order(rowMaxs(as.matrix(phex_max_dn)),decreasing = T),]
 colnames(phex_max_dn)=c(-window:(ncol(phex_max_dn)-(window+1)))
 # phex_max_dn= cbind(rowMaxs(as.matrix(phex_max_dn))-rowMins(as.matrix(phex_max_dn)),phex_max_dn);colnames(phex_max_dn)[1]='diff'
 phex_max_dn= cbind(rowMaxs(as.matrix(phex_max_dn)),phex_max_dn);colnames(phex_max_dn)[1]='diff'

write.table(phex_max_dn, file=paste0('./structure/chunkwindow/Exons/downstream/phex_max_dn.txt'), quote=F, sep="\t", row.names=T, col.names=NA)

#######################################################

phexseq_dn=seqout3_exondn
  phexseq_dn=as.data.frame(phexseq_dn[,-c(1:2)])
  phexseq_dn=t(phexseq_dn)
  phexseq_dn[is.na(phexseq_dn)]=0
phexseq_dn=phexseq_dn[order(rowMaxs(as.matrix(phexseq_dn)),decreasing = T),]
 colnames(phexseq_dn)=c(-window:(ncol(phexseq_dn)-(window+1)))
 # phexseq_dn= cbind(rowMaxs(as.matrix(phexseq_dn))-rowMins(as.matrix(phexseq_dn)),phexseq_dn);colnames(phexseq_dn)[1]='diff'
 phexseq_dn= cbind(rowMaxs(as.matrix(phexseq_dn)),phexseq_dn);colnames(phexseq_dn)[1]='diff'

write.table(phexseq_dn, file=paste0('./structure/chunkwindow/Exons/downstream/phexseq_dn.txt'), quote=F, sep="\t", row.names=T, col.names=NA)

phexseq_max_dn=seqout3_max_exondn
  phexseq_max_dn=as.data.frame(phexseq_max_dn[,-c(1:2)])
  phexseq_max_dn=t(phexseq_max_dn)
  phexseq_max_dn[is.na(phexseq_max_dn)]=0
phexseq_max_dn=phexseq_max_dn[order(rowMaxs(as.matrix(phexseq_max_dn)),decreasing = T),]
 colnames(phexseq_max_dn)=c(-window:(ncol(phexseq_max_dn)-(window+1)))
 # phexseq_max_dn= cbind(rowMaxs(as.matrix(phexseq_max_dn))-rowMins(as.matrix(phexseq_max_dn)),phexseq_max_dn);colnames(phexseq_max_dn)[1]='diff'
 phexseq_max_dn= cbind(rowMaxs(as.matrix(phexseq_max_dn)),phexseq_max_dn);colnames(phexseq_max_dn)[1]='diff'

write.table(phexseq_max_dn, file=paste0('./structure/chunkwindow/Exons/downstream/phexseq_max_dn.txt'), quote=F, sep="\t", row.names=T, col.names=NA)

```






<!-- Introns -->
```{r ,fig.asp=.5, fig.align='center',fig.height=3, fig.width=6,echo=F,warning=F,eval=T,include=F}
remove('mntnout3','chunkout3')

Peaksdata4_chunk2=Peaksdata4[Peaksdata4$`Both Strand: RNA_type_Classification`%in%c('protein_coding: exon','protein_coding: Intron'),]
Peaksdata4_chunk2=Peaksdata4[Peaksdata4$`Both Strand: RNA_type_Classification`%in%c('protein_coding: Intron'),]
# unique(Peaksdata4_motif$`Both Strand: RNA_type_Classification`)
Peaksdata4_chunk2=separate(Peaksdata4_chunk2,ID,into = c('chr','start'),sep=':',remove = F)
Peaksdata4_chunk2=separate(Peaksdata4_chunk2,start,into = c('start','end'),sep='-')
Peaksdata4_chunk2$start=as.numeric(Peaksdata4_chunk2$start)
Peaksdata4_chunk2$end=as.numeric(Peaksdata4_chunk2$end)
# unique(Peaksdata4_chunk2$chr)

fasta='../../../../Resources/ref/mm10/Gencode_VM23/fromGencode/GRCm38.primary_assembly.genome.fa'

chunkout2_Comb=as.data.frame(matrix(nrow = nrow(Peaksdata4_chunk2),ncol=91))
chunkout2_seq=as.data.frame(matrix(nrow = nrow(Peaksdata4_chunk2),ncol=4))
fig=F
```

<!-- #### Upstream Base Pairing Probability -->
```{r ,fig.asp=.5, fig.align='center',fig.height=3, fig.width=6,echo=F,warning=F,eval=T,include=F}
####################################################################
#### Upstream Base Pairing Probability
####################################################################
### no basepairing chr18:77457443-77457501
### no low basepairing chr14:105137760-105137796 (bottom of heatmap)
### high diff in basepair probability chr6:55054497-55054526

buffer=100 
window=100
chsize=10

  chunkname3=as.data.frame(matrix(nrow=1,ncol=nrow(Peaksdata4_chunk2)+2))
  chunkout3=as.data.frame(matrix(nrow=window,ncol=nrow(Peaksdata4_chunk2)+2))
  chunkout3[,1]=seq(1,window,1)
  chunkout3_max=as.data.frame(matrix(nrow=window,ncol=nrow(Peaksdata4_chunk2)+2))
  chunkout3_max[,1]=seq(1,window,1)
  
  mntnout3=as.data.frame(matrix(nrow=(window*3),ncol=nrow(Peaksdata4_chunk2)+2))
  mntnout3[,1]=rep(seq(1,window,1),3)
  mntnout3[,2]=c(rep('mntn1',window),rep('mntn2',window),rep('mntn3',window))
seqout3=as.data.frame(matrix(nrow=window,ncol=nrow(Peaksdata4_chunk2)+2))
seqout3[,1]=seq(1,window,1)
seqout3_max=as.data.frame(matrix(nrow=window,ncol=nrow(Peaksdata4_chunk2)+2))
seqout3_max[,1]=seq(1,window,1)
  
for (x in 1:nrow(Peaksdata4_chunk2)) {
# for (x in 1:5) {
if (x%in%seq(from=0,to=1000,by=25)){print(x/nrow(Peaksdata4_chunk2))}

p=Peaksdata4_chunk2[x,]
# p=Peaksdata4_chunk2[Peaksdata4_chunk2$ID%in%'chr6:55054497-55054526',]

chr=p$chr
start=p$start
end=p$end
strand=p$strand


##################
if (strand%in%'+') {
croslink=start
s1=start-buffer#(selects base +1 of start site but ends at correct 5' site)
e1=start

bedpos=matrix(ncol=6,nrow=1)
bedpos[1,1]=chr
bedpos[1,2]=s1
bedpos[1,3]=e1
bedpos[1,4]=".";bedpos[1,5]="."
bedpos[1,6]=strand
write.table(bedpos, file=paste0('./structure/chunkwindow/',p$ID,"_chunk.bed"), quote=F, sep="\t", row.names=F, col.names=F)
s0=start-buffer+1#(selects base +1 of start site but ends at correct 5' site)
e0=start

}

if (strand%in%'-') {
croslink=end
s1=end-1#(selects base +1 of start site but ends at correct 5' site)
e1=end+buffer-1
bedpos=matrix(ncol=6,nrow=1)
bedpos[1,1]=chr
bedpos[1,2]=s1
bedpos[1,3]=e1
bedpos[1,4]=".";bedpos[1,5]="."
bedpos[1,6]=strand
write.table(bedpos, file=paste0('./structure/chunkwindow/',p$ID,"_chunk.bed"), quote=F, sep="\t", row.names=F, col.names=F)
s0=end-0#(selects base +1 of start site but ends at correct 5' site)
e0=end+buffer-1
}

system(paste0('bedtools getfasta -s -fi ',fasta,' -bed structure/chunkwindow/',p$ID,'_chunk.bed > structure/chunkwindow/',p$ID,'_chunk.fa'))

system(paste0('/Users/homanpj/Documents/Tools/ViennaRNA-2.4.14/bin/RNAfold -p ./structure/chunkwindow/',p$ID,'_chunk.fa > out.fold'))
system('mv *.ps ./structure/chunkwindow/')

chunk=read.delim(paste0('./structure/chunkwindow/',p$chr,'_',s1,'-',e1,'(',strand,')_dp.ps'),comment.char = "%",header = F,sep = "\t")
chunk=chunk[grep("ubox",chunk$V1),]
chunk=as.data.frame(chunk[-1],stringsAsFactors=F)
chunk$`chunk[-1]`=as.character(chunk$`chunk[-1]`);colnames(chunk)="V1"
chunk=separate(chunk,'V1',sep = " ",into = c('b1','b2','pvalue','ubox'))
chunk$pvalue=as.numeric(chunk$pvalue)
chunk$b1=as.numeric(chunk$b1)
chunk$b2=as.numeric(chunk$b2)

# system(paste0('/Users/homanpj/Documents/Tools/ViennaRNA-2.4.14/src/Utils/mountain.pl ./structure/chunkwindow/',p$chr,'_',s1,'-',e1,'\\(',strand,'\\)_dp.ps > ./structure/chunkwindow/',p$chr,'_',s1,'-',e1,'\\(',strand,'\\)_dp.ps.mntn'))
# 
# mntn=read.delim(paste0('./structure/chunkwindow/',p$chr,'_',s1,'-',e1,'(',strand,')_dp.ps.mntn'),comment.char = "",header = F,sep = "\t")
# mntn$V1=as.character(mntn$V1)
# mntn=separate(mntn,V1,into = c('pos','score'),sep = 4)
# mntn1=mntn[1:which(mntn$pos%in%"&")[1],];
#     mntn1=mntn1[-which(mntn1$pos%in%"&"),];mntn1$class="mntn1"
# mntn3=mntn[which(mntn$pos%in%"&")[2]:nrow(mntn),]
#     mntn3=mntn3[-which(mntn3$pos%in%"&"),];mntn3$class="mntn3"
# mntn2=mntn[which(mntn$pos%in%"&")[1]:which(mntn$pos%in%"&")[2],]
#   mntn2=mntn2[-which(mntn2$pos%in%"&"),];mntn2$class="mntn2"
# 
# mntnall=rbind(mntn1,mntn2,mntn3)
# mntnall$pos=as.numeric(mntnall$pos)
# mntnall$score=as.numeric(mntnall$score)
#   # ggplot(mntnall,aes(x=pos,y=score,line=class,color=class))+geom_line()+theme_classic()
# 
#   mntnout3[,x+2]=mntnall$score

###############
## BP probibilites for each chuck at each position
###############
for (y in 1:(window-9)) {
  ch=chunk[chunk$b1%in%seq(y,y+9,1)|chunk$b2%in%seq(y,y+9,1),]
  ch2=ch
    if (nrow(ch2)>0) {
   ch2$br1=rowMins(as.matrix(ch[,c('b1','b2')]))
    ch2$br2=rowMaxs(as.matrix(ch[,c('b1','b2')]))
      ch$b1=ch2$br1
      ch$b2=ch2$br2
      ch=ch[duplicated(ch)==F,]
  # colnames(chunkout2)[y+2]=paste0(y,'-',y+9)
      chunkout3[y,1]=y
      chunkout3[y,2]=paste0(y,'-',y+9)
      chunkout3[y,x+2]=mean(ch$pvalue) ##average all basepairinng accross chunck
      
      chunkout3_max[y,1]=y
      chunkout3_max[y,2]=paste0(y,'-',y+9)
      chunkout3_max[y,x+2]=max(ch$pvalue) ##average all basepairinng accross chunck
      
    }   
    if (nrow(ch2)==0) {
      chunkout3[y,1]=y
      chunkout3[y,2]=paste0(y,'-',y+9)
      chunkout3[y,x+2]=0
      chunkout3_max[y,1]=y
      chunkout3_max[y,2]=paste0(y,'-',y+9)
      chunkout3_max[y,x+2]=0}  
      
            chunkname3[1,x+2]=p$ID
}


### count each base
  for (y in 1:(window)) {
  seq=chunk[chunk$b1%in%(y)|chunk$b2%in%(y),]
  seq2=seq
  
  if (nrow(seq2)>0) {
    seq2$br1=rowMins(as.matrix(seq[,c('b1','b2')]))
    seq2$br2=rowMaxs(as.matrix(seq[,c('b1','b2')]))
  seq$b1=seq2$br1
  seq$b2=seq2$br2
  seq=seq[duplicated(seq)==F,]
  # colnames(chunkout2)[y+2]=paste0(y,'-',y+9)
      seqout3[y,1]=y
      # seqout3[y,2]=paste0(y)
      seqout3[y,x+2]=mean(seq$pvalue) ##average all basepairinng accross chunck
      
      seqout3_max[y,1]=y
      # seqout3[y,2]=paste0(y)
      seqout3_max[y,x+2]=max(seq$pvalue) ##average all basepairinng accross chunck
  }    
   if (nrow(seq2)==0) {
      seqout3[y,1]=y
      seqout3[y,x+2]=0
      seqout3_max[y,1]=y
      seqout3_max[y,x+2]=0}
}


### count Mountain BPP
#   for (y in 1:(window)) {
#     # i=rowMins(as.matrix(chunk[,c('b1','b2')]))
#     # j=rowMaxs(as.matrix(chunk[,c('b1','b2')]))
#     i=chunk[,'b1']
#     j=chunk[,'b2']
#   seq=chunk[(i<y)&(j>y),]
# 
#   # colnames(chunkout2)[y+2]=paste0(y,'-',y+9)
#       seqout3[y,1]=y
#       # seqout3[y,x+2]=sum(seq[,'pvalue'])
#       # seqout3[y,x+2]=nrow(seq)*mean(seq[,'pvalue'])
#       seqout3[y,x+2]=mean(seq[,'pvalue'])*100
# }
# 
#  chx=chunkout3[,c(1,2,x+2)]
#   colnames(chx)=c('pos','chunk','score');chx$class='BPP'
#     # chx$score=chx$score*50
#     # chx$score=(-chx$score)*log(chx$score)
#     chx$score=chx$score*50
# 
#       seqx=seqout3[,c(1,2,x+2)]
#     colnames(seqx)=c('pos','chunk','score');seqx$class='BPP'
#         # seqx$score=seqx$score/4
# 
# 
#   ggplot(mntnall,aes(x=pos,y=score,line=class,color=class))+geom_line()+theme_classic()+geom_line(data=seqx,col='blue')#+geom_line(data=chx)#
#


######################################################################
if(file.exists(paste0('structure/chunkwindow/',p$ID,'_chunk.fa'))){file.remove(paste0('structure/chunkwindow/',p$ID,'_chunk.fa'))}
if(file.exists(paste0('structure/chunkwindow/',p$ID,'_chunk.bed'))){file.remove(paste0('structure/chunkwindow/',p$ID,'_chunk.bed'))}
if(file.exists(paste0('./structure/chunkwindow/',p$chr,'_',s1,'-',e1,'(',strand,')_dp.ps'))){file.remove(paste0('./structure/chunkwindow/',p$chr,'_',s1,'-',e1,'(',strand,')_dp.ps'))}
if(file.exists(paste0('./structure/chunkwindow/',p$chr,'_',s1,'-',e1,'(',strand,')_ss.ps'))){file.remove(paste0('./structure/chunkwindow/',p$chr,'_',s1,'-',e1,'(',strand,')_ss.ps'))}
if(file.exists(paste0('./structure/chunkwindow/',p$chr,'_',s1,'-',e1,'\\(',strand,'\\)_dp.ps.mntn'))){file.remove(paste0('./structure/chunkwindow/',p$chr,'_',s1,'-',e1,'\\(',strand,'\\)_dp.ps.mntn'))}

if(file.exists(paste0(p$chr,'_',s1,'-',e1,'(',strand,').fold'))){file.remove(paste0(p$chr,'_',s1,'-',e1,'(',strand,').fold'))}
######################################################################

}

chunkname3=t(chunkname3)

######################
colnames(chunkout3)[3:ncol(chunkout3)]=chunkname3[3:nrow(chunkname3),1]

chunkout3=chunkout3[is.na(chunkout3[,3])==F,]
chunkout3[is.infinite(as.matrix(chunkout3))==T]=0


colnames(chunkout3_max)[3:ncol(chunkout3_max)]=chunkname3[3:nrow(chunkname3),1]

chunkout3_max=chunkout3_max[is.na(chunkout3_max[,3])==F,]
chunkout3_max[is.infinite(as.matrix(chunkout3_max))==T]=0

######################
colnames(seqout3)[3:ncol(seqout3)]=chunkname3[3:nrow(chunkname3),1]

seqout3=seqout3[is.na(seqout3[,3])==F,]
seqout3[is.infinite(as.matrix(seqout3))==T]=0

colnames(seqout3_max)[3:ncol(seqout3_max)]=chunkname3[3:nrow(chunkname3),1]

seqout3_max=seqout3_max[is.na(seqout3_max[,3])==F,]
seqout3_max[is.infinite(as.matrix(seqout3_max))==T]=0

######################

ph=as.matrix(chunkout3[,-c(1:2)])
ph=(chunkout3[,-c(1:2)])

ph=ph[is.na(ph[,3])==F,]
ph[is.nan(as.matrix(ph))==T]=0
ph[is.infinite(as.matrix(ph))==T]=0
  ph=t(ph)
  ph=as.data.frame(ph)

  #### sort by row diff
# ph=ph[order(rowMaxs(as.matrix(ph))-rowMins(as.matrix(ph)),decreasing = T),]
ph=ph[order(rowMaxs(as.matrix(ph)),decreasing = T),]
 colnames(ph)=c(-window:(ncol(ph)-(window+1)))
 # ph= cbind(rowMaxs(as.matrix(ph))-rowMins(as.matrix(ph)),ph);colnames(ph)[1]='diff'
 ph= cbind(rowMaxs(as.matrix(ph)),ph);colnames(ph)[1]='diff'

write.table(ph, file=paste0('./structure/chunkwindow/genee/ph.txt'), quote=F, sep="\t", row.names=T, col.names=NA)

######

ph_max=as.matrix(chunkout3_max[,-c(1:2)])
ph_max=(chunkout3_max[,-c(1:2)])

ph_max=ph_max[is.na(ph_max[,3])==F,]
ph_max[is.nan(as.matrix(ph_max))==T]=0
ph_max[is.infinite(as.matrix(ph_max))==T]=0
  ph_max=t(ph_max)
  ph_max=as.data.frame(ph_max)

  #### sort by row diff
# ph_max=ph_max[order(rowMaxs(as.matrix(ph_max))-rowMins(as.matrix(ph_max)),decreasing = T),]
ph_max=ph_max[order(rowMaxs(as.matrix(ph_max)),decreasing = T),]
 colnames(ph_max)=c(-window:(ncol(ph_max)-(window+1)))
 # ph_max= cbind(rowMaxs(as.matrix(ph_max))-rowMins(as.matrix(ph_max)),ph_max);colnames(ph_max)[1]='diff'
 ph_max= cbind(rowMaxs(as.matrix(ph_max)),ph_max);colnames(ph_max)[1]='diff'

write.table(ph_max, file=paste0('./structure/chunkwindow/genee/ph_max.txt'), quote=F, sep="\t", row.names=T, col.names=NA)

#######################################################

phseq=seqout3
  phseq=as.data.frame(phseq[,-c(1:2)])
  phseq=t(phseq)
  phseq[is.na(phseq)]=0
phseq=phseq[order(rowMaxs(as.matrix(phseq)),decreasing = T),]
 colnames(phseq)=c(-window:(ncol(phseq)-(window+1)))
 # phseq= cbind(rowMaxs(as.matrix(phseq))-rowMins(as.matrix(phseq)),phseq);colnames(phseq)[1]='diff'
 phseq= cbind(rowMaxs(as.matrix(phseq)),phseq);colnames(phseq)[1]='diff'

write.table(phseq, file=paste0('./structure/chunkwindow/genee/phseq.txt'), quote=F, sep="\t", row.names=T, col.names=NA)

phseq_max=seqout3_max
  phseq_max=as.data.frame(phseq_max[,-c(1:2)])
  phseq_max=t(phseq_max)
  phseq_max[is.na(phseq_max)]=0
phseq_max=phseq_max[order(rowMaxs(as.matrix(phseq_max)),decreasing = T),]
 colnames(phseq_max)=c(-window:(ncol(phseq_max)-(window+1)))
 # phseq_max= cbind(rowMaxs(as.matrix(phseq_max))-rowMins(as.matrix(phseq_max)),phseq_max);colnames(phseq_max)[1]='diff'
 phseq_max= cbind(rowMaxs(as.matrix(phseq_max)),phseq_max);colnames(phseq_max)[1]='diff'

write.table(phseq_max, file=paste0('./structure/chunkwindow/genee/phseq_max.txt'), quote=F, sep="\t", row.names=T, col.names=NA)


#######################################################

# phmt=mntnout3[mntnout3$V2%in%'mntn1',]
#   phmt=as.data.frame(phmt[,-c(1:2)])
#   phmt=t(phmt)
# 
# phmt=phmt[order(rowMaxs(as.matrix(phmt)),decreasing = T),]
#  colnames(phmt)=c(-window:(ncol(phmt)-(window+1)))
#  # phmt= cbind(rowMaxs(as.matrix(phmt))-rowMins(as.matrix(phmt)),phmt);colnames(phmt)[1]='diff'
#  phmt= cbind(rowMaxs(as.matrix(phmt)),phmt);colnames(phmt)[1]='diff'
# 
# write.table(phmt, file=paste0('./structure/chunkwindow/phmt.txt'), quote=F, sep="\t", row.names=T, col.names=NA)

```


<!-- #### Downstream Base Pairing Probability -->
```{r ,fig.asp=.5, fig.align='center',fig.height=6, fig.width=12,echo=F,warning=F,eval=T,include=F}
chunkout2_dn_Comb=as.data.frame(matrix(nrow = nrow(Peaksdata4_chunk2),ncol=(window-9)))
fig=F
  chunkout2_dn_seq=as.data.frame(matrix(nrow = nrow(Peaksdata4_chunk2),ncol=4))



####################################################################
#### downstream Base Pairing Probability
####################################################################
### High base pair probability chr8:18628170-18628210
### low basepair probibility chr9:72581297-72581336
buffer=100
window=100

  chunkname3_dn=as.data.frame(matrix(nrow=1,ncol=nrow(Peaksdata4_chunk2)+2))

  chunkout3_dn=as.data.frame(matrix(nrow=window,ncol=nrow(Peaksdata4_chunk2)+2))
    chunkout3_dn[,1]=seq(1,window,1)
  chunkout3_dn_max=as.data.frame(matrix(nrow=window,ncol=nrow(Peaksdata4_chunk2)+2))
    chunkout3_dn_max[,1]=seq(1,window,1)

  seqout3_dn=as.data.frame(matrix(nrow=window,ncol=nrow(Peaksdata4_chunk2)+2))
    seqout3_dn[,1]=seq(1,window,1)
  seqout3_dn_max=as.data.frame(matrix(nrow=window,ncol=nrow(Peaksdata4_chunk2)+2))
    seqout3_dn_max[,1]=seq(1,window,1)

  
for (x in 1:nrow(Peaksdata4_chunk2)) {
# for (x in 1:5) {
if (x%in%seq(from=0,to=1000,by=25)){print(x/nrow(Peaksdata4_chunk2))}

p=Peaksdata4_chunk2[x,]
# p=Peaksdata4_chunk2[Peaksdata4_chunk2$ID%in%'chr9:72581297-72581336',]

chr=p$chr
start=p$start
end=p$end
strand=p$strand


##################
if (strand%in%'+') {
croslink=start
s1=start-1
e1=start+buffer-1#(selects base +1 of start site but ends at correct 5' site)

bedpos=matrix(ncol=6,nrow=1)
bedpos[1,1]=chr
bedpos[1,2]=s1
bedpos[1,3]=e1
bedpos[1,4]=".";bedpos[1,5]="."
bedpos[1,6]=strand
write.table(bedpos, file=paste0('./structure/chunkwindow/',p$ID,"_chunk.bed"), quote=F, sep="\t", row.names=F, col.names=F)
s0=start-buffer+1#(selects base +1 of start site but ends at correct 5' site)
e0=start

}
if (strand%in%'-') {
croslink=end
s1=end-buffer
e1=end#(selects base +1 of start site but ends at correct 5' site)

bedpos=matrix(ncol=6,nrow=1)
bedpos[1,1]=chr
bedpos[1,2]=s1
bedpos[1,3]=e1
bedpos[1,4]=".";bedpos[1,5]="."
bedpos[1,6]=strand
write.table(bedpos, file=paste0('./structure/chunkwindow/',p$ID,"_chunk.bed"), quote=F, sep="\t", row.names=F, col.names=F)
s0=end-0#(selects base +1 of start site but ends at correct 5' site)
e0=end+buffer-1
}

system(paste0('bedtools getfasta -s -fi ',fasta,' -bed structure/chunkwindow/',p$ID,'_chunk.bed > structure/chunkwindow/',p$ID,'_chunk.fa'))

system(paste0('/Users/homanpj/Documents/Tools/ViennaRNA-2.4.14/bin/RNAfold -p ./structure/chunkwindow/',p$ID,'_chunk.fa > out.fold'))
system('mv *.ps ./structure/chunkwindow/')

chunk=read.delim(paste0('./structure/chunkwindow/',p$chr,'_',s1,'-',e1,'(',strand,')_dp.ps'),comment.char = "%",header = F,sep = "\t")
chunk=chunk[grep("ubox",chunk$V1),]
chunk=as.data.frame(chunk[-1],stringsAsFactors=F)
chunk$`chunk[-1]`=as.character(chunk$`chunk[-1]`);colnames(chunk)="V1"
chunk=separate(chunk,'V1',sep = " ",into = c('b1','b2','pvalue','ubox'))
chunk$pvalue=as.numeric(chunk$pvalue)
chunk$b1=as.numeric(chunk$b1)
chunk$b2=as.numeric(chunk$b2)


###############
## BP probibilites for each chuck at each position
###############


for (y in 1:(window-9)) {
  ch=chunk[chunk$b1%in%seq(y,y+9,1)|chunk$b2%in%seq(y,y+9,1),]
  ch2=ch
    if (nrow(ch2)>0) {
   ch2$br1=rowMins(as.matrix(ch[,c('b1','b2')]))
    ch2$br2=rowMaxs(as.matrix(ch[,c('b1','b2')]))
      ch$b1=ch2$br1
      ch$b2=ch2$br2
      ch=ch[duplicated(ch)==F,]
  # colnames(chunkout2)[y+2]=paste0(y,'-',y+9)
      chunkout3_dn[y,1]=y
      chunkout3_dn[y,2]=paste0(y,'-',y+9)
      chunkout3_dn[y,x+2]=mean(ch$pvalue) ##average all basepairinng accross chunck
      
      chunkout3_dn_max[y,1]=y
      chunkout3_dn_max[y,2]=paste0(y,'-',y+9)
      chunkout3_dn_max[y,x+2]=max(ch$pvalue) ##average all basepairinng accross chunck
      
    }   
    if (nrow(ch2)==0) {
      chunkout3_dn[y,1]=y
      chunkout3_dn[y,2]=paste0(y,'-',y+9)
      chunkout3_dn[y,x+2]=0
      chunkout3_dn_max[y,1]=y
      chunkout3_dn_max[y,2]=paste0(y,'-',y+9)
      chunkout3_dn_max[y,x+2]=0}  
      
            chunkname3_dn[1,x+2]=p$ID
}

### count each base
  for (y in 1:(window)) {
  seq=chunk[chunk$b1%in%(y)|chunk$b2%in%(y),]
  seq2=seq
  
  if (nrow(seq2)>0) {
    seq2$br1=rowMins(as.matrix(seq[,c('b1','b2')]))
    seq2$br2=rowMaxs(as.matrix(seq[,c('b1','b2')]))
  seq$b1=seq2$br1
  seq$b2=seq2$br2
  seq=seq[duplicated(seq)==F,]
  # colnames(chunkout2)[y+2]=paste0(y,'-',y+9)
      seqout3_dn[y,1]=y
      # seqout3_dn[y,2]=paste0(y)
      seqout3_dn[y,x+2]=mean(seq$pvalue) ##average all basepairinng accross chunck
      
      seqout3_dn_max[y,1]=y
      # seqout3_dn[y,2]=paste0(y)
      seqout3_dn_max[y,x+2]=max(seq$pvalue) ##average all basepairinng accross chunck
  }    
   if (nrow(seq2)==0) {
      seqout3_dn[y,1]=y
      seqout3_dn[y,x+2]=0
      seqout3_dn_max[y,1]=y
      seqout3_dn_max[y,x+2]=0}
}

######################################################################
if(file.exists(paste0('structure/chunkwindow/',p$ID,'_chunk.fa'))){file.remove(paste0('structure/chunkwindow/',p$ID,'_chunk.fa'))}
if(file.exists(paste0('structure/chunkwindow/',p$ID,'_chunk.bed'))){file.remove(paste0('structure/chunkwindow/',p$ID,'_chunk.bed'))}
if(file.exists(paste0('./structure/chunkwindow/',p$chr,'_',s1,'-',e1,'(',strand,')_dp.ps'))){file.remove(paste0('./structure/chunkwindow/',p$chr,'_',s1,'-',e1,'(',strand,')_dp.ps'))}
if(file.exists(paste0('./structure/chunkwindow/',p$chr,'_',s1,'-',e1,'(',strand,')_ss.ps'))){file.remove(paste0('./structure/chunkwindow/',p$chr,'_',s1,'-',e1,'(',strand,')_ss.ps'))}
if(file.exists(paste0('./structure/chunkwindow/',p$chr,'_',s1,'-',e1,'\\(',strand,'\\)_dp.ps.mntn'))){file.remove(paste0('./structure/chunkwindow/',p$chr,'_',s1,'-',e1,'\\(',strand,'\\)_dp.ps.mntn'))}
if(file.exists(paste0(p$chr,'_',s1,'-',e1,'(',strand,').fold'))){file.remove(paste0(p$chr,'_',s1,'-',e1,'(',strand,').fold'))}
######################################################################

}


chunkname3_dn=t(chunkname3_dn)
colnames(chunkout3_dn)[3:ncol(chunkout3_dn)]=chunkname3_dn[3:nrow(chunkname3_dn),1]
colnames(chunkout3_dn_max)[3:ncol(chunkout3_dn_max)]=chunkname3_dn[3:nrow(chunkname3_dn),1]

colnames(seqout3_dn)[3:ncol(seqout3_dn)]=chunkname3_dn[3:nrow(chunkname3_dn),1]
colnames(seqout3_dn_max)[3:ncol(seqout3_dn_max)]=chunkname3_dn[3:nrow(chunkname3_dn),1]

##########
ph_dn=chunkout3_dn[,-c(1:2)]
ph_dn=ph_dn[is.na(ph_dn[,3])==F,]
ph_dn[is.nan(as.matrix(ph_dn))==T]=0
  ph_dn=t(ph_dn)
  ph_dn=as.data.frame(ph_dn)
  
#### sort by row diff
ph_dn=ph_dn[order(rowMaxs(as.matrix(ph_dn)),decreasing = T),]
ph_dn= cbind(rowMaxs(as.matrix(ph_dn)),ph_dn);colnames(ph_dn)[1]='diff'

write.table(ph_dn, file=paste0('./structure/chunkwindow/ph_dn_order.txt'), quote=F, sep="\t", row.names=T, col.names=NA)

##########
ph_dn_max=chunkout3_dn_max[,-c(1:2)]
ph_dn_max=ph_dn_max[is.na(ph_dn_max[,3])==F,]
ph_dn_max[is.nan(as.matrix(ph_dn_max))==T]=0
  ph_dn_max=t(ph_dn_max)
  ph_dn_max=as.data.frame(ph_dn_max)
  
#### sort by row diff
ph_dn_max=ph_dn_max[order(rowMaxs(as.matrix(ph_dn_max)),decreasing = T),]
ph_dn_max= cbind(rowMaxs(as.matrix(ph_dn_max)),ph_dn_max);colnames(ph_dn_max)[1]='diff'

write.table(ph_dn_max, file=paste0('./structure/chunkwindow/ph_dn_max_order.txt'), quote=F, sep="\t", row.names=T, col.names=NA)

#######################################################
##########
phseq_dn=seqout3_dn[,-c(1:2)]
phseq_dn=phseq_dn[is.na(phseq_dn[,3])==F,]
phseq_dn[is.nan(as.matrix(phseq_dn))==T]=0
  phseq_dn=t(phseq_dn)
  phseq_dn=as.data.frame(phseq_dn)
  
#### sort by row diff
phseq_dn=phseq_dn[order(rowMaxs(as.matrix(phseq_dn)),decreasing = T),]
phseq_dn= cbind(rowMaxs(as.matrix(phseq_dn)),phseq_dn);colnames(phseq_dn)[1]='diff'

write.table(phseq_dn, file=paste0('./structure/chunkwindow/genee/nt_mean/phseq_dn_order.txt'), quote=F, sep="\t", row.names=T, col.names=NA)


##########
phseq_dn_max=seqout3_dn_max[,-c(1:2)]
phseq_dn_max=phseq_dn_max[is.na(phseq_dn_max[,3])==F,]
phseq_dn_max[is.nan(as.matrix(phseq_dn_max))==T]=0
  phseq_dn_max=t(phseq_dn_max)
  phseq_dn_max=as.data.frame(phseq_dn_max)
  
#### sort by row diff
phseq_dn_max=phseq_dn_max[order(rowMaxs(as.matrix(phseq_dn_max)),decreasing = T),]
phseq_dn_max= cbind(rowMaxs(as.matrix(phseq_dn_max)),phseq_dn_max);colnames(phseq_dn_max)[1]='diff'

write.table(phseq_dn_max, file=paste0('./structure/chunkwindow/genee/nt_max/phseq_dn_max_order.txt'), quote=F, sep="\t", row.names=T, col.names=NA)



# phseq_dn=seqout3_dn
#   phseq_dn=as.data.frame(phseq_dn[,-c(1:2)])
#   phseq_dn=t(phseq_dn)
# 
# phseq_dn=phseq_dn[order(rowMaxs(as.matrix(phseq_dn)),decreasing = T),]
#  colnames(phseq_dn)=c(-window:(ncol(phseq_dn)-(window+1)))
#  # phseq_dn= cbind(rowMaxs(as.matrix(phseq_dn))-rowMins(as.matrix(phseq_dn)),phseq_dn);colnames(phseq_dn)[1]='diff'
#  phseq_dn= cbind(rowMaxs(as.matrix(phseq_dn)),phseq_dn);colnames(phseq_dn)[1]='diff'
# 
# write.table(phseq_dn, file=paste0('./structure/chunkwindow/genee/phseq_dn.txt'), quote=F, sep="\t", row.names=T, col.names=NA)

 

```

```{r ,fig.asp=.5, fig.align='center',fig.height=6, fig.width=12,echo=F,warning=F,eval=T,include=F}


pc=chunkout3[,-c(2)]
pc$V1=c(-window:(nrow(pc)-(window+1)))
pc=melt(pc,id.vars = 'V1')
ggplotly(ggplot(pc,aes(y=value,x=V1,line=variable,color=variable))+geom_line(na.rm=F)+xlab("Position of 10nt Window\n(Distance from 5'Clip peak)")+ylab('Average basepairing probability'))

pc2=rowMeans(chunkout3[,-c(1:2)],na.rm = T)
pc2=cbind(chunkout3[,c(1:2)],pc2)
pc2$V1=c(-window:(nrow(pc2)-(window+1)))

pc2=pc2[is.na(pc2$pc2)==F,]
ggplot(pc2,aes(y=pc2,x=V1))+geom_line(na.rm=F)+xlab("Position of 10nt Window\n(Distance from 5'Clip peak)")+ylab('Average basepairing probability')+ylim(.02,.1)
```



```{r ,fig.asp=1.5, fig.align='center',fig.height=6, fig.width=6,echo=F,warning=F,eval=T,include=F}
remove(consensus,pph)
##### 10nt Chuck mean
# consensus=read.delim(paste0("./structure/chunkwindow/genee/mean/100nt/ph.Hclustaverage.txt.txt"), header=T, sep="\t",stringsAsFactors = F,row.names=1)
# colnames(consensus[,-c(1:2)])=c(-window:(ncol(consensus[,-c(1:2)])-(window+2)))
# consensusorder=read.delim(paste0("./structure/chunkwindow/genee/mean/100nt/ph.Hclustaverage.hclustorder.txt"), header=F, sep="\t",stringsAsFactors = F,row.names=1)
# phx=ph
# ph_dnx=ph_dn
# region='intron'
# 
# phm=consensus[,-c(1:2)]
# ComplexHeatmap::Heatmap(phm,cluster_columns = F,cluster_rows =F)


##### 10nt Chuck max 
# consensus=read.delim(paste0("./structure/chunkwindow/genee/max/ph_4K.txt.txt"), header=T, sep="\t",stringsAsFactors = F,row.names=1)
# colnames(consensus[,-c(1:7)])=c(-window:(ncol(consensus[,-c(1:7)])-(window+1)))
# consensusorder=consensus[,c('X4.Clusters'),drop=F]
# phx=ph_max
# ph_dnx=ph_dn_max
# region='intron'

# phm=consensus[,-c(1:7)]
# ComplexHeatmap::Heatmap(phm,cluster_columns = F,cluster_rows =F)



##### Each nt mean
# consensus=read.delim(paste0("./structure/chunkwindow/genee/nt_mean/ph.Hclustaverage.txt.txt"), header=T, sep="\t",stringsAsFactors = F,row.names=1)
# colnames(consensus[,-c(1:2)])=c(-window:-1)
# consensusorder=read.delim(paste0("./structure/chunkwindow/genee/nt_mean/ph.Hclustaverage.hclustorder.txt"), header=F, sep="\t",stringsAsFactors = F,row.names=1)
# phx=phseq
# ph_dnx=phseq_dn
# region='intron'

# phm=consensus[,-c(1:2)]
# ComplexHeatmap::Heatmap(phm,cluster_columns = F,cluster_rows =F)


##### Exon 10nt Chuck mean 
# consensus=read.delim(paste0("./structure/chunkwindow/Exons/downstream/genee/mean/ph.Hclustaverage.txt.txt"), header=T, sep="\t",stringsAsFactors = F,row.names=1)
# colnames(consensus[,-c(1:2)])=c(-window:(ncol(consensus[,-c(1:2)])-(window+2)))
# consensusorder=read.delim(paste0("./structure/chunkwindow/Exons/downstream/genee/mean/ph.Hclustaverage.hclustorder.txt"), header=F, sep="\t",stringsAsFactors = F,row.names=1)
# phx=phex
# ph_dnx=phex_dn
# 
# phm=consensus[,-c(1:2)]
# ComplexHeatmap::Heatmap(phm,cluster_columns = F,cluster_rows =F)
# region='exon'



##### Exon 10nt Chuck mead_downstream 
consensus=read.delim(paste0("./structure/chunkwindow/Exons/downstream/genee/mean/ph_dn.Hclustaverage.txt.txt"), header=T, sep="\t",stringsAsFactors = F,row.names=1)
colnames(consensus[,-c(1:2)])=c(-window:(ncol(consensus[,-c(1:2)])-(window+2)))
consensusorder=read.delim(paste0("./structure/chunkwindow/Exons/downstream/genee/mean/ph_dn.Hclustaverage.hclustorder.txt"), header=F, sep="\t",stringsAsFactors = F,row.names=1)
phx=phex_dn
ph_dnx=phex

phm=consensus[,-c(1:2)]
ComplexHeatmap::Heatmap(phm,cluster_columns = F,cluster_rows =F)
region='exon'


#### Sort by Upstream order
ph_dn=ph_dn[rownames(consensus),]
# write.table(ph_dn, file=paste0('./structure/chunkwindow/ph_dn_uporder.txt'), quote=F, sep="\t", row.names=T, col.names=NA)
```

```{r ,fig.asp=.5, fig.align='center',fig.height=3, fig.width=6,echo=F,warning=F,eval=T,include=F}

checkoutx=checkout[checkout$SingleExon%in%'Change-TRUE'==F,]
checkoutx=merge(checkoutx,phex[,c('diff'),drop=F],by.x="ID",by.y='row.names')

checkoutx[checkoutx$SingleExon%in%'TRUE','SingleExon']='Single Exon'
checkoutx[checkoutx$SingleExon%in%'FALSE','SingleExon']='Exon Spanning'

checkout_dnx=checkout_dn
checkout_dnx=merge(checkout_dnx,phex_dn[,c('diff'),drop=F],by.x="ID",by.y='row.names')

checkout_dnx[checkout_dnx$SingleExon%in%'TRUE','SingleExon']='Single Exon'
checkout_dnx[checkout_dnx$SingleExon%in%'FALSE','SingleExon']='Exon Spanning'

ggplot(checkoutx,aes(x=-deltaG,line=SingleExon,color=SingleExon))+geom_density()
ggplot(checkout_dnx,aes(x=-deltaG,line=SingleExon,color=SingleExon))+geom_density()

ggplot(checkoutx,aes(x=diff,line=SingleExon,color=SingleExon))+geom_density()
ggplot(checkout_dnx,aes(x=diff,line=SingleExon,color=SingleExon))+geom_density()

nrow(checkoutx[checkoutx$SingleExon%in%'Exon Spanning',])
nrow(checkoutx[checkoutx$SingleExon%in%'Single Exon',])

nrow(checkout_dnx[checkout_dnx$SingleExon%in%'Exon Spanning',])
nrow(checkout_dnx[checkout_dnx$SingleExon%in%'Single Exon',])
```

```{r ,fig.asp=.5, fig.align='center',fig.height=6, fig.width=6,echo=F,warning=F,eval=T,include=F}

####################################################################
for (n in min(consensusorder[,1]):max(consensusorder[,1])) {
  

pph=as.data.frame(phx[rownames(consensusorder)[consensusorder[,1]%in%n],-1])

pph1=cbind(rownames(pph),pph);colnames(pph1)[1]='ID';pph1$ID=as.character(pph1$ID)
 pph1= melt(pph1,id.vars ="ID" );pph1$variable=as.numeric(as.character(pph1$variable))
print(ggplotly(ggplot(pph1,aes(y=value,x=variable,line=ID,color=ID))+geom_line(na.rm=F)+xlab("Position of 10nt Window\n(Distance from 5'Clip peak)")+ylab('Average basepairing probability')+theme_classic()+ggtitle(as.character(n))))
 
pph1_average=t(colMeans(pph))
pph1_average=(rbind(colnames(pph1_average),as.data.frame(pph1_average)))
pph1_average=as.data.frame(t(pph1_average))
pph1_average$V1=as.numeric(as.character(pph1_average$V1));pph1_average$V2=as.numeric(as.character(pph1_average$V2))
print(ggplot(pph1_average,aes(y=V2,x=V1))+geom_line(na.rm=F)+xlab("Position of 10nt Window\n(Distance from 5'Clip peak)")+ylab('Average basepairing probability')+theme_classic()+ggtitle(as.character(n)))
#+ylim(0.5,1))

remove(pph,pph1,pph1_average)

}
```




```{r ,fig.asp=.5, fig.align='center',fig.height=2, fig.width=4,echo=F,warning=F,eval=T,include=F}
##################################################################################################################
###### Plot Upstream clusters structure with corresponding Downstream 
##################################################################################################################
for (n in min(consensusorder[,1]):max(consensusorder[,1])) {

pph=as.data.frame(ph_dnx[rownames(consensusorder)[consensusorder[,1]%in%n],-1])

pph1=cbind(rownames(pph),pph);colnames(pph1)[1]='ID';pph1$ID=as.character(pph1$ID)
 pph1= melt(pph1);pph1$variable=as.numeric(pph1$variable)
print(ggplotly(ggplot(pph1,aes(y=value,x=variable,line=ID,color=ID))+geom_line(na.rm=F)+xlab("Position of 10nt Window\n(Distance from 5'Clip peak)")+ylab('Average basepairing probability')+theme_classic()+ggtitle(as.character(n))))

pph1_average=(colMeans(pph))
pph1_average=rbind(c(1:length(pph1_average)),pph1_average)
pph1_average=as.data.frame(t(pph1_average))

########
pph2=as.data.frame(phx[rownames(consensusorder)[consensusorder[,1]%in%n],-1])

pph2_average=(colMeans(pph2))
pph2_average=rbind(colnames(pph2),pph2_average)
pph2_average=as.data.frame(t(pph2_average))
pph2_average$V1=as.numeric(as.character(pph2_average$V1))
pph2_average$pph2_average=as.numeric(as.character(pph2_average$pph2_average))

##
pph1_average=cbind(pph1_average,pph2_average[nrow(pph2_average):1,])
colnames(pph1_average)=c("V1","Downstream",'V2',"Upstream")
rownames(pph1_average)=abs(pph1_average$V2)
pph1_average=melt(t(pph1_average[,-1]));
pph1_average=pph1_average[pph1_average$X1%in%c("V1","V2")==F,]

print(ggplot(pph1_average,aes(y=value,x=X2,line=X1,color=X1))+geom_line(na.rm=F)+xlab("Position of 10nt Window\n(Absolute Distance from 5'Clip peak)")+ylab('Average basepairing probability')+theme_classic()+xlim(window,1)+ggtitle(as.character(n)))
  # +ylim(.5,1))



if ((region)=='intron') {DG=structure_outall[, c('ID','deltaG_Upstream_100nt')];rownames(DG)=DG$ID
DG=DG[rownames(consensusorder)[consensusorder[,1]%in%n], ]}
if ((region)=='exon') {DG=merge(consensusorder,checkout[, c('ID','deltaG')],all.x=T,by.x='row.names',by.y='ID',sort = F)
  colnames(DG)=c('ID','V2','deltaG_Upstream_100nt');rownames(DG)=DG$ID;DG$ID=as.character(DG$ID)}

DG$deltaG_Upstream_100nt=-DG$deltaG_Upstream_100nt

print(ggplot(DG,aes(x=ID,y=deltaG_Upstream_100nt))+geom_bar(stat="identity")+theme_classic()+ylim(0,60)+ggtitle("0"))


remove(pph,pph1,pph1_average,pph2,pp2_average,DG)
}


####################################################################
####################################################################
## locations not Clustered

if ((region)=='intron') {DG=structure_outall[, c('ID','deltaG_Upstream_100nt')];rownames(DG)=DG$ID
DG=DG[rownames(consensus)[(rownames(consensus)%in%rownames(consensusorder))==F], ]
}

if ((region)=='exon') {DG=merge(consensus[,1:2],checkout[, c('ID','deltaG')],all.x=T,by.x='row.names',by.y='ID',sort = F)
  colnames(DG)[colnames(DG)%in%'deltaG']=c('deltaG_Upstream_100nt');
  colnames(DG)[colnames(DG)%in%'Row.names']=c('ID');
  rownames(DG)=DG$ID;DG$ID=as.character(DG$ID)}


DG$deltaG_Upstream_100nt=-DG$deltaG_Upstream_100nt

ggplot(DG,aes(x=ID,y=deltaG_Upstream_100nt))+geom_bar(stat="identity")+theme_classic()#+ylim(0,70)
 

####################################################################
## All locations

if ((region)=='intron') {DGall=structure_outall[, c('ID','deltaG_Upstream_100nt')];rownames(DGall)=DGall$ID}
if ((region)=='exon') {DGall=checkout[,c('ID',"deltaG")];colnames(DGall)[2]='deltaG_Upstream_100nt';rownames(DGall)=DGall$ID}

DGall=DGall[rownames(consensus), ]
DGall$deltaG_Upstream_100nt=-DGall$deltaG_Upstream_100nt

ggplot(DGall,aes(x=ID,y=deltaG_Upstream_100nt))+geom_bar(stat="identity")+theme_classic()+ylim(0,60)
``` 


```{r ,fig.asp=1, fig.align='center',fig.height=3, fig.width=3,echo=F,warning=F,eval=T,include=F}

########################################################################################################################################
########################################################################################################################################


DGallp=merge(DGall,phx,by.x='ID',by.y='row.names')
DGallp$deltaG_Upstream_100nt=DGallp$deltaG_Upstream_100nt

DGallp$diff=rowMaxs(as.matrix(DGallp[,-c(1:3)]))

ggscatter(DGallp, x = 'deltaG_Upstream_100nt', y = 'diff', add = "reg.line",color = "red",   
          add.params = list(color = "black", fill = NA), # Customize reg. line
          xlab='CLIP peak Upstream (100nt) \n-DeltaG',
          ylab=expression("Max row Base Pair Probability" )  
) +
 stat_cor(
   aes(label = paste(..rr.label.., ..p.label.., sep = "~`,`~")),
  label.x = 30,label.y=.65
)


##########################################

DGallp=merge(DGall,phx,by.x='ID',by.y='row.names')
DGallp$deltaG_Upstream_100nt=DGallp$deltaG_Upstream_100nt

DGallp$diff=rowMeans(as.matrix(DGallp[,-c(1:3)]))

ggscatter(DGallp, x = 'deltaG_Upstream_100nt', y = 'diff', add = "reg.line",color = "red",   
          add.params = list(color = "black", fill = NA), # Customize reg. line
          xlab='CLIP peak Upstream (100nt) \n-DeltaG',
          ylab=expression("Mean row Base Pair Probability" )  
) +
 stat_cor(
   aes(label = paste(..rr.label.., ..p.label.., sep = "~`,`~")),
  label.x = 30,label.y=.4
)
```



```{r ,fig.asp=.5, fig.align='center',fig.height=6, fig.width=6,echo=F,warning=F,eval=T,include=F}
##################################################################################################################
###### Plot Downstream clusters structure with corresponding Upstream 
##################################################################################################################


####################################################################
consensusorder_dn=read.delim(paste0("./structure/chunkwindow/genee/ph_dn_order.Hclustaverage.hclustorder.txt"), header=F, sep="\t",stringsAsFactors = F,row.names=1)

#### Sort Upstream by downstream order

consensus_dn=read.delim(paste0("./structure/chunkwindow/genee/ph_dn_order.Hclustaverage.txt.txt"), header=F, sep="\t",stringsAsFactors = F,row.names=1)
ph_dnorder=phx[rownames(consensus_dn),]

ComplexHeatmap::Heatmap(ph_dnorder,cluster_columns = F,cluster_rows =F, clustering_distance_columns ="pearson", row_dend_reorder = T)

write.table(ph_dnorder, file=paste0('./structure/chunkwindow/genee/ph_dnorder.txt'), quote=F, sep="\t", row.names=T, col.names=T)

####################################################################


pph_dn=ph_dn[rownames(consensusorder_dn)[consensusorder_dn$V2%in%0],-1]

pph_dn1=cbind(rownames(pph_dn),pph_dn);
colnames(pph_dn1)[1]='ID';
 pph_dn1= melt(pph_dn1);pph_dn1$variable=as.numeric(pph_dn1$variable)
ggplotly(ggplot(pph_dn1,aes(y=value,x=variable,line=ID,color=ID))+geom_line(na.rm=F)+xlab("Position of 10nt Window\n(Distance from 5'Clip peak)")+ylab('Average basepairing probability')+theme_classic())


####################################################################
pph_dn=ph_dn[rownames(consensusorder_dn)[consensusorder_dn$V2%in%1],-1]

pph_dn1=cbind(rownames(pph_dn),pph_dn);colnames(pph_dn1)[1]='ID';pph_dn1$ID=as.character(pph_dn1$ID)
 pph_dn1= melt(pph_dn1);pph_dn1$variable=as.numeric(pph_dn1$variable)
ggplotly(ggplot(pph_dn1,aes(y=value,x=variable,line=ID,color=ID))+geom_line(na.rm=F)+xlab("Position of 10nt Window\n(Distance from 5'Clip peak)")+ylab('Average basepairing probability')+theme_classic())


####################################################################
pph_dn=ph_dn[rownames(consensusorder_dn)[consensusorder_dn$V2%in%2],-1]

pph_dn1=cbind(rownames(pph_dn),pph_dn);colnames(pph_dn1)[1]='ID';pph_dn1$ID=as.character(pph_dn1$ID)
 pph_dn1= melt(pph_dn1);pph_dn1$variable=as.numeric(pph_dn1$variable)
ggplotly(ggplot(pph_dn1,aes(y=value,x=variable,line=ID,color=ID))+geom_line(na.rm=F)+xlab("Position of 10nt Window\n(Distance from 5'Clip peak)")+ylab('Average basepairing probability')+theme_classic())


####################################################################
pph_dn=ph_dn[rownames(consensusorder_dn)[consensusorder_dn$V2%in%3],-1]

pph_dn1=cbind(rownames(pph_dn),pph_dn);colnames(pph_dn1)[1]='ID';pph_dn1$ID=as.character(pph_dn1$ID)
 pph_dn1= melt(pph_dn1);pph_dn1$variable=as.numeric(pph_dn1$variable)
ggplotly(ggplot(pph_dn1,aes(y=value,x=variable,line=ID,color=ID))+geom_line(na.rm=F)+xlab("Position of 10nt Window\n(Distance from 5'Clip peak)")+ylab('Average basepairing probability')+theme_classic())


####################################################################
pph_dn=ph_dn[rownames(consensusorder_dn)[consensusorder_dn$V2%in%4],-1]

pph_dn1=cbind(rownames(pph_dn),pph_dn);colnames(pph_dn1)[1]='ID';pph_dn1$ID=as.character(pph_dn1$ID)
 pph_dn1= melt(pph_dn1);pph_dn1$variable=as.numeric(pph_dn1$variable)
ggplotly(ggplot(pph_dn1,aes(y=value,x=variable,line=ID,color=ID))+geom_line(na.rm=F)+xlab("Position of 10nt Window\n(Distance from 5'Clip peak)")+ylab('Average basepairing probability')+theme_classic())


####################################################################
```



```{r ,fig.asp=.5, fig.align='center',fig.height=6, fig.width=6,echo=F,warning=F,eval=T,include=F}

####################################################################
consensusorder_dn=read.delim(paste0("./structure/chunkwindow/genee/ph_dn_order.Hclustaverage.hclustorder.txt"), header=F, sep="\t",stringsAsFactors = F,row.names=1)

####################################################################


pph_dn=ph_dn[rownames(consensusorder_dn)[consensusorder_dn$V2%in%0],-1]

pph_dn1=cbind(rownames(pph_dn),pph_dn);colnames(pph_dn1)[1]='ID';pph_dn1$ID=as.character(pph_dn1$ID)
 pph_dn1= melt(pph_dn1);pph_dn1$variable=as.numeric(pph_dn1$variable)
ggplotly(ggplot(pph_dn1,aes(y=value,x=variable,line=ID,color=ID))+geom_line(na.rm=F)+xlab("Position of 10nt Window\n(Distance from 5'Clip peak)")+ylab('Average basepairing probability')+theme_classic())


pph_dn1_average=t(colMeans(pph_dn))
pph_dn1_average=rbind(c(1:length(pph_dn1_average)),pph_dn1_average)
pph_dn1_average=as.data.frame(t(pph_dn1_average))

pph2=ph[rownames(consensusorder_dn)[consensusorder_dn$V2%in%0],-1]

pph1=cbind(rownames(pph2),pph2);colnames(pph1)[1]='ID';pph1$ID=as.character(pph1$ID)
 pph1= melt(pph1);pph1$variable=as.numeric(as.character(pph1$variable));pph1$variable=pph1$variable+10
ggplotly(ggplot(pph1,aes(y=value,x=variable,line=ID,color=ID))+geom_line(na.rm=F)+xlab("Position of 10nt Window\n(Distance from 5'Clip peak)")+ylab('Average basepairing probability')+theme_classic()+xlim(-1,min(pph1$variable)))


pph2_average=t(colMeans(pph2))
pph2_average=rbind(colnames(pph2_average),pph2_average)
pph2_average=as.data.frame(t(pph2_average))
pph2_average$V1=as.numeric(as.character(pph2_average$V1))
pph2_average$V2=as.numeric(as.character(pph2_average$V2))

pph1_average=cbind(pph_dn1_average,pph2_average[nrow(pph2_average):1,])
colnames(pph1_average)=c("V1","Downstream",'V2',"Upstream")
pph1_average=melt(t(pph1_average[,-1]))
pph1_average=pph1_average[pph1_average$X1%in%c("V1","V2")==F,]

(ggplot(pph1_average,aes(y=value,x=X2,line=X1,color=X1))+geom_line(na.rm=F)+xlab("Position of 10nt Window\n(Absolute Distance from 5'Clip peak)")+ylab('Average basepairing probability')+theme_classic()+ylim(0,.3)+xlim(1,window))


####################################################################
pph_dn=ph_dn[rownames(consensusorder_dn)[consensusorder_dn$V2%in%1],-1]

pph_dn1=cbind(rownames(pph_dn),pph_dn);colnames(pph_dn1)[1]='ID';pph_dn1$ID=as.character(pph_dn1$ID)
 pph_dn1= melt(pph_dn1);pph_dn1$variable=as.numeric(pph_dn1$variable)
ggplotly(ggplot(pph_dn1,aes(y=value,x=variable,line=ID,color=ID))+geom_line(na.rm=F)+xlab("Position of 10nt Window\n(Distance from 5'Clip peak)")+ylab('Average basepairing probability')+theme_classic())


pph_dn1_average=t(colMeans(pph_dn))
pph_dn1_average=rbind(c(1:length(pph_dn1_average)),pph_dn1_average)
pph_dn1_average=as.data.frame(t(pph_dn1_average))

pph2=ph[rownames(consensusorder_dn)[consensusorder_dn$V2%in%1],-1]

pph1=cbind(rownames(pph2),pph2);colnames(pph1)[1]='ID';pph1$ID=as.character(pph1$ID)
 pph1= melt(pph1);pph1$variable=as.numeric(as.character(pph1$variable));pph1$variable=pph1$variable+10
ggplotly(ggplot(pph1,aes(y=value,x=variable,line=ID,color=ID))+geom_line(na.rm=F)+xlab("Position of 10nt Window\n(Distance from 5'Clip peak)")+ylab('Average basepairing probability')+theme_classic()+xlim(-1,min(pph1$variable)))


pph2_average=t(colMeans(pph2))
pph2_average=rbind(colnames(pph2_average),pph2_average)
pph2_average=as.data.frame(t(pph2_average))
pph2_average$V1=as.numeric(as.character(pph2_average$V1))
pph2_average$V2=as.numeric(as.character(pph2_average$V2))

pph1_average=cbind(pph_dn1_average,pph2_average[nrow(pph2_average):1,])
colnames(pph1_average)=c("V1","Downstream",'V2',"Upstream")
pph1_average=melt(t(pph1_average[,-1]))
pph1_average=pph1_average[pph1_average$X1%in%c("V1","V2")==F,]

(ggplot(pph1_average,aes(y=value,x=X2,line=X1,color=X1))+geom_line(na.rm=F)+xlab("Position of 10nt Window\n(Absolute Distance from 5'Clip peak)")+ylab('Average basepairing probability')+theme_classic()+ylim(0,.3)+xlim(1,window))


####################################################################
pph_dn=ph_dn[rownames(consensusorder_dn)[consensusorder_dn$V2%in%2],-1]

pph_dn1=cbind(rownames(pph_dn),pph_dn);colnames(pph_dn1)[1]='ID';pph_dn1$ID=as.character(pph_dn1$ID)
 pph_dn1= melt(pph_dn1);pph_dn1$variable=as.numeric(pph_dn1$variable)
ggplotly(ggplot(pph_dn1,aes(y=value,x=variable,line=ID,color=ID))+geom_line(na.rm=F)+xlab("Position of 10nt Window\n(Distance from 5'Clip peak)")+ylab('Average basepairing probability')+theme_classic())


pph_dn1_average=t(colMeans(pph_dn))
pph_dn1_average=rbind(c(1:length(pph_dn1_average)),pph_dn1_average)
pph_dn1_average=as.data.frame(t(pph_dn1_average))

pph2=ph[rownames(consensusorder_dn)[consensusorder_dn$V2%in%2],-1]

pph1=cbind(rownames(pph2),pph2);colnames(pph1)[1]='ID';pph1$ID=as.character(pph1$ID)
 pph1= melt(pph1);pph1$variable=as.numeric(as.character(pph1$variable));pph1$variable=pph1$variable+10
ggplotly(ggplot(pph1,aes(y=value,x=variable,line=ID,color=ID))+geom_line(na.rm=F)+xlab("Position of 10nt Window\n(Distance from 5'Clip peak)")+ylab('Average basepairing probability')+theme_classic()+xlim(-1,min(pph1$variable)))


pph2_average=t(colMeans(pph2))
pph2_average=rbind(colnames(pph2_average),pph2_average)
pph2_average=as.data.frame(t(pph2_average))
pph2_average$V1=as.numeric(as.character(pph2_average$V1))
pph2_average$V2=as.numeric(as.character(pph2_average$V2))

pph1_average=cbind(pph_dn1_average,pph2_average[nrow(pph2_average):1,])
colnames(pph1_average)=c("V1","Downstream",'V2',"Upstream")
pph1_average=melt(t(pph1_average[,-1]))
pph1_average=pph1_average[pph1_average$X1%in%c("V1","V2")==F,]

(ggplot(pph1_average,aes(y=value,x=X2,line=X1,color=X1))+geom_line(na.rm=F)+xlab("Position of 10nt Window\n(Absolute Distance from 5'Clip peak)")+ylab('Average basepairing probability')+theme_classic()+ylim(0,.3)+xlim(1,window))


####################################################################
pph_dn=ph_dn[rownames(consensusorder_dn)[consensusorder_dn$V2%in%3],-1]

pph_dn1=cbind(rownames(pph_dn),pph_dn);colnames(pph_dn1)[1]='ID';pph_dn1$ID=as.character(pph_dn1$ID)
 pph_dn1= melt(pph_dn1);pph_dn1$variable=as.numeric(pph_dn1$variable)
ggplotly(ggplot(pph_dn1,aes(y=value,x=variable,line=ID,color=ID))+geom_line(na.rm=F)+xlab("Position of 10nt Window\n(Distance from 5'Clip peak)")+ylab('Average basepairing probability')+theme_classic())


pph_dn1_average=t(colMeans(pph_dn))
pph_dn1_average=rbind(c(1:length(pph_dn1_average)),pph_dn1_average)
pph_dn1_average=as.data.frame(t(pph_dn1_average))

pph2=ph[rownames(consensusorder_dn)[consensusorder_dn$V2%in%3],-1]

pph1=cbind(rownames(pph2),pph2);colnames(pph1)[1]='ID';pph1$ID=as.character(pph1$ID)
 pph1= melt(pph1);pph1$variable=as.numeric(as.character(pph1$variable));pph1$variable=pph1$variable+10
ggplotly(ggplot(pph1,aes(y=value,x=variable,line=ID,color=ID))+geom_line(na.rm=F)+xlab("Position of 10nt Window\n(Distance from 5'Clip peak)")+ylab('Average basepairing probability')+theme_classic()+xlim(-1,min(pph1$variable)))


pph2_average=t(colMeans(pph2))
pph2_average=rbind(colnames(pph2_average),pph2_average)
pph2_average=as.data.frame(t(pph2_average))
pph2_average$V1=as.numeric(as.character(pph2_average$V1))
pph2_average$V2=as.numeric(as.character(pph2_average$V2))

pph1_average=cbind(pph_dn1_average,pph2_average[nrow(pph2_average):1,])
colnames(pph1_average)=c("V1","Downstream",'V2',"Upstream")
pph1_average=melt(t(pph1_average[,-1]))
pph1_average=pph1_average[pph1_average$X1%in%c("V1","V2")==F,]

(ggplot(pph1_average,aes(y=value,x=X2,line=X1,color=X1))+geom_line(na.rm=F)+xlab("Position of 10nt Window\n(Absolute Distance from 5'Clip peak)")+ylab('Average basepairing probability')+theme_classic()+ylim(0,.3)+xlim(1,window))


####################################################################


pph_dn=ph_dn[rownames(consensusorder_dn)[consensusorder_dn$V2%in%4],-1]

pph_dn1=cbind(rownames(pph_dn),pph_dn);colnames(pph_dn1)[1]='ID';pph_dn1$ID=as.character(pph_dn1$ID)
 pph_dn1= melt(pph_dn1);pph_dn1$variable=as.numeric(pph_dn1$variable)
ggplotly(ggplot(pph_dn1,aes(y=value,x=variable,line=ID,color=ID))+geom_line(na.rm=F)+xlab("Position of 10nt Window\n(Distance from 5'Clip peak)")+ylab('Average basepairing probability')+theme_classic())


pph_dn1_average=t(colMeans(pph_dn))
pph_dn1_average=rbind(c(1:length(pph_dn1_average)),pph_dn1_average)
pph_dn1_average=as.data.frame(t(pph_dn1_average))

pph2=ph[rownames(consensusorder_dn)[consensusorder_dn$V2%in%4],-1]

pph1=cbind(rownames(pph2),pph2);colnames(pph1)[1]='ID';pph1$ID=as.character(pph1$ID)
 pph1= melt(pph1);pph1$variable=as.numeric(as.character(pph1$variable));pph1$variable=pph1$variable+10
ggplotly(ggplot(pph1,aes(y=value,x=variable,line=ID,color=ID))+geom_line(na.rm=F)+xlab("Position of 10nt Window\n(Distance from 5'Clip peak)")+ylab('Average basepairing probability')+theme_classic()+xlim(-1,min(pph1$variable)))


pph2_average=t(colMeans(pph2))
pph2_average=rbind(colnames(pph2_average),pph2_average)
pph2_average=as.data.frame(t(pph2_average))
pph2_average$V1=as.numeric(as.character(pph2_average$V1))
pph2_average$V2=as.numeric(as.character(pph2_average$V2))

pph1_average=cbind(pph_dn1_average,pph2_average[nrow(pph2_average):1,])
colnames(pph1_average)=c("V1","Downstream",'V2',"Upstream")
pph1_average=melt(t(pph1_average[,-1]))
pph1_average=pph1_average[pph1_average$X1%in%c("V1","V2")==F,]

(ggplot(pph1_average,aes(y=value,x=X2,line=X1,color=X1))+geom_line(na.rm=F)+xlab("Position of 10nt Window\n(Absolute Distance from 5'Clip peak)")+ylab('Average basepairing probability')+theme_classic()+ylim(0,.3)+xlim(1,window))


```

```{r ,fig.asp=.5, fig.align='center',fig.height=6, fig.width=12,echo=F,warning=F,eval=T,include=F}

Peaksdata4_chunk2_out=Peaksdata4_chunk2[,c('ID','strand',"Same Strand: Host_gene_name","Same Strand: Host_gene_ensembl_id","Same Strand: Repeat_Type","Both Strand: RNA_type_Classification")]

Peaksdata4_chunk2_out=merge(consensusorder,Peaksdata4_chunk2_out,all.y=T,by.y='ID',by.x="row.names")
Peaksdata4_chunk2_out$V2=Peaksdata4_chunk2_out$V2+1
  colnames(Peaksdata4_chunk2_out)[colnames(Peaksdata4_chunk2_out)%in%"V2"]="Upstream pairing probability Cluster"
  
Peaksdata4_chunk2_out=merge(consensusorder_dn,Peaksdata4_chunk2_out,all.y=T,by.y='Row.names',by.x="row.names")
Peaksdata4_chunk2_out$V2=Peaksdata4_chunk2_out$V2+1
  colnames(Peaksdata4_chunk2_out)[colnames(Peaksdata4_chunk2_out)%in%"V2"]="Downstream pairing probability Cluster"  
  colnames(Peaksdata4_chunk2_out)[colnames(Peaksdata4_chunk2_out)%in%"Row.names"]="ID"
  Peaksdata4_chunk2_out$ID=as.character(Peaksdata4_chunk2_out$ID)
  Peaksdata4_chunk2_out=Peaksdata4_chunk2_out[order(Peaksdata4_chunk2_out$`Downstream pairing probability Cluster`),]
  
write.table(Peaksdata4_chunk2_out, file=paste0('./structure/chunkwindow/genee/CLIP_Peaks_BasePairing.txt'), quote=F, sep="\t", row.names=T, col.names=T)



############### Exonics ######################

checkoutx=checkout
checkoutx[checkoutx$SingleExon=='TRUE',"SingleExon"]='Single Exon'
checkoutx[checkoutx$SingleExon=='Change-TRUE',"SingleExon"]='Single Exon'
checkoutx[checkoutx$SingleExon=='FALSE',"SingleExon"]='Multiple Exons'

checkoutx=merge(checkoutx[,c('ID','trancript',"SingleExon","exon#",'deltaG','sequence')],phex[,'diff',drop=F],by.x='ID',by.y='row.names',sort=F)
checkoutx=checkoutx[,c('ID',"SingleExon","exon#",'diff','deltaG','sequence')]
colnames(checkoutx)=c('ID',"Upstream:SingleExon","Upstream:exon#",'Upstream:Max BPP','Upstream:deltaG','Upstream:sequence')

#######

checkout_dnx=checkout_dn
checkout_dnx[checkout_dnx$SingleExon=='TRUE',"SingleExon"]='Single Exon'
checkout_dnx[checkout_dnx$SingleExon=='Change-TRUE',"SingleExon"]='Single Exon'
checkout_dnx[checkout_dnx$SingleExon=='FALSE',"SingleExon"]='Multiple Exons'

checkout_dnx=merge(checkout_dnx[,c('ID','trancript',"SingleExon","exon#",'deltaG','sequence')],phex_dn[,'diff',drop=F],by.x='ID',by.y='row.names',sort=F)
checkout_dnx=checkout_dnx[,c('ID',"SingleExon","exon#",'diff','deltaG','sequence')]
colnames(checkout_dnx)=c('ID',"Downstream:SingleExon","Downstream:exon#",'Downstream:Max BPP','Downstream:deltaG','Downstream:sequence')

#######

Peaksdata4_chunkEx_out=Peaksdata4_chunkEx[,c('ID','strand',"Same Strand: Host_gene_name","Same Strand: Host_gene_ensembl_id","Same Strand: Repeat_Type","Both Strand: RNA_type_Classification")]

Peaksdata4_chunkEx_out=merge(Peaksdata4_chunkEx_out,checkoutx,all.y=T,by='ID',all.x=F)
Peaksdata4_chunkEx_out=merge(Peaksdata4_chunkEx_out,checkout_dnx,all.y=T,by='ID',all.x=F)

Peaksdata4_chunkEx_out=Peaksdata4_chunkEx_out[is.na(Peaksdata4_chunkEx_out$strand)==F,]

write.table(Peaksdata4_chunk2_out, file=paste0('./structure/chunkwindow/Exons/downstream/CLIP_Peaks_exons_BasePairing.txt'), quote=F, sep="\t", row.names=T, col.names=T)



```


### random Intron sequence location

```{r ,fig.asp=.5, fig.align='center',fig.height=6, fig.width=12,echo=F,warning=F,eval=T,include=F}
#################################################
# Function to process RNAfold output
#################################################

RNAfold_output=function(fls){
    chunkRout3=as.data.frame(matrix(nrow=window,ncol=3))
  chunkRname3=as.data.frame(matrix(nrow=1,ncol=3))
  chunkRout3[,1]=seq(1,window,1)
# for (x in 1:length(fls)) {
  # if (x%in%seq(from=0,to=1000,by=25)){print(x/length(fls))}
  # fls=as.list(fls)
  # fls=fls[117]
  fls=unlist(fls)

chunkR=read.delim(paste0('./structure/chunkwindow/scr/',fls[1]),comment.char = "%",header = F,sep = "\t")
chunkR=chunkR[grep("ubox",chunkR$V1),]
chunkR=as.data.frame(chunkR[-1],stringsAsFactors=F)
chunkR$`chunkR[-1]`=as.character(chunkR$`chunkR[-1]`);colnames(chunkR)="V1"
chunkR=separate(chunkR,'V1',sep = " ",into = c('b1','b2','pvalue','ubox'))
chunkR$pvalue=as.numeric(chunkR$pvalue)
chunkR$b1=as.numeric(chunkR$b1)
chunkR$b2=as.numeric(chunkR$b2)
  
     # for (y in 1:(window-9)) {
    #   ch1=chunkR[chunkR$b1%in%seq(y,y+9,1),]
    #   ch2=chunkR[chunkR$b2%in%seq(y,y+9,1),];
    #     b1r=ch2$b2;b2r=ch2$b1
    #     ch2$b1=b1r;ch2$b2=b2r
    #   ch=rbind(ch1,ch2)
    #   ch=ch[duplicated(ch)==F,]
    #       chunkRout3[y,1]=y
    #       chunkRout3[y,2]=paste0(y,'-',y+9)
    #       if (nrow(ch)==0) {chunkRout3[y,3]=0}
    #       if (nrow(ch)>0) {chunkRout3[y,3]=max(ch$pvalue)} ##average all basepairinng accross chunck
    #         chunkRname3[1,3]=gsub("_dp.ps","",fls[1])
    # }


### count each base
  for (y in 1:(window)) {
  seq=chunk[chunkR$b1%in%(y)|chunk$b2%in%(y),]
  seq2=seq
  
  if (nrow(seq2)>0) {
    seq2$br1=rowMins(as.matrix(seq[,c('b1','b2')]))
    seq2$br2=rowMaxs(as.matrix(seq[,c('b1','b2')]))
  seq$b1=seq2$br1
  seq$b2=seq2$br2
  seq=seq[duplicated(seq)==F,]
  # colnames(chunkout2)[y+2]=paste0(y,'-',y+9)
      chunkRout3[y,1]=y
      # chunkRout3[y,2]=paste0(y)
      chunkRout3[y,x+2]=mean(seq$pvalue) ##average all basepairinng accross chunck
      
      # chunkRout3_max[y,1]=y
      # chunkRout3[y,2]=paste0(y)
      # chunkRout3_max[y,x+2]=max(seq$pvalue) ##average all basepairinng accross chunck
  }    
   if (nrow(seq2)==0) {
      chunkRout3[y,1]=y
      chunkRout3[y,x+2]=0
      # chunkRout3_max[y,1]=y
      # chunkRout3_max[y,x+2]=0
      }
}

chunkRname3=t(chunkRname3)
colnames(chunkRout3)[3:ncol(chunkRout3)]=chunkRname3[3:nrow(chunkRname3),1]

chunkRout3=chunkRout3[is.na(chunkRout3[,3])==F,]
chunkRout3[is.nan(as.matrix(chunkRout3))==T]=0

return(chunkRout3)
} 
######################################################################################################################################################################################################################################################################################################################


introns=fread("/Users/homanpj/OneDrive - National Institutes of Health/Resources/ref/mm10/Gencode_VM23/fromUCSC/KnownGene/KnownGene_GRCm38_introns.bed", header=F, sep="\t",stringsAsFactors = F,data.table=F)
  introns=introns[grepl("_",introns$V1)==F,]
  colnames(introns)=c('chr','start','end','attribute','V5','strand')
  introns=separate(introns,attribute,into=c('transcript_id','feature','exon_number','level','chr2','intronnumber','dir'),remove = T,sep = "_")
  introns$transcript_id=removeVersion(introns$transcript_id)
  # introns_canon=introns[introns$transcript_id%in%canonical$transcript_id,]
  introns$start=introns$start+1
  introns$exon_number=as.numeric(introns$exon_number)+1
  introns$diff=introns$end-introns$start
  # introns[introns$diff>300,]
  
fasta='../../../../Resources/ref/mm10/Gencode_VM23/fromGencode/GRCm38.primary_assembly.genome.fa'
fig=F
buffer=100
window=100
nrepeat=100
maxBPP=.9
outchunck_rowaverage=matrix(nrow=nrepeat,ncol = (window-9) )
outchunck_frac=matrix(nrow=nrepeat,ncol = 1 )
outchunck_rowaverage_dn=matrix(nrow=nrepeat,ncol = (window-9) )
outchunck_frac_dn=matrix(nrow=nrepeat,ncol = 1 )

system.time({
for (r in 1:nrepeat) {
print(r)
  
rand=sample(1:nrow(introns), nrow(Peaksdata4_chunk2_out), replace=TRUE)
introns_rand=introns[rand,]
introns_rand$CL=NA
for (x in 1:nrow(introns_rand)) {
  strt=introns_rand[x,'start']
  end=introns_rand[x,'end']
  introns_rand[x,'CL']=sample((strt-window):(end+window), 1, replace=TRUE)
}

####################################################################
#### Upstream Base Pairing Probability
#################################################################### 
##################
introns_rand_p=introns_rand[introns_rand$strand%in%"+",]
s1p=introns_rand_p$CL-buffer#(selects base +1 of start site but ends at correct 5' site)
e1p=introns_rand_p$CL
chr=introns_rand_p$chr
strand=introns_rand_p$strand
  bedpos=matrix(ncol=6,nrow=nrow(introns_rand_p))
  bedpos[,1]=chr
  bedpos[,2]=s1p
  bedpos[,3]=e1p
  bedpos[,4]=".";bedpos[,5]="."
  bedpos[,6]=strand
#################

introns_rand_n=introns_rand[introns_rand$strand%in%"-",]
s1n=introns_rand_n$CL-1#(selects base +1 of start site but ends at correct 5' site)
e1n=introns_rand_n$CL+buffer-1
chr=introns_rand_n$chr
strand=introns_rand_n$strand
s0=introns_rand_n$CL-0#(selects base +1 of start site but ends at correct 5' site)
e0=introns_rand_n$CL+buffer-1
  bedneg=matrix(ncol=6,nrow=nrow(introns_rand_n))
  bedneg[,1]=chr
  bedneg[,2]=s1n
  bedneg[,3]=e1n
  bedneg[,4]=".";bedneg[,5]="."
  bedneg[,6]=strand

#################
bedpos=rbind(bedpos,bedneg)
#################

write.table(bedpos[,], file=paste0("./structure/chunkwindow/scr/bed_chunkR.bed"), quote=F, sep="\t", row.names=F, col.names=F)

  system(paste0('rm -f ./structure/chunkwindow/scr/*.ps'))


system(paste0('bedtools getfasta -s -fi ',fasta,' -bed structure/chunkwindow/scr/bed_chunkR.bed | /Users/homanpj/Documents/Tools/ViennaRNA-2.4.14/bin/RNAfold -p  > out.fold'))
system('mv *.ps ./structure/chunkwindow/scr')

fls=list.files("./structure/chunkwindow/scr/", pattern="dp.ps", all.files=FALSE,
    full.names=FALSE)


out=mclapply(as.list(fls),RNAfold_output,mc.cores=8)
out2=out[[1]][,1:2,drop=F]
for (o in 1:length(out)) {
  if (nrow(out[[o]][,3,drop=F])==0) {out2$Xx=0;colnames(out2)[colnames(out2)%in%"Xx"]=colnames(out[[o]][,3,drop=F])}
  if (nrow(out[[o]][,3,drop=F])>0) {out2=cbind(out2,out[[o]][,3,drop=F])}
}

system(paste0('rm -f ./structure/chunkwindow/scr/*.ps'))
system(paste0('rm -f ./structure/chunkwindow/scr/*.fa'))
system(paste0('rm -f ./structure/chunkwindow/scr/*.bed'))

# #########################################
ph=as.matrix(out2[,-c(1:2)])
ph=ph[is.na(ph[,3])==F,]
ph[is.nan(as.matrix(ph))==T]=0

  ph=t(ph)
  ph=as.data.frame(ph)

  #### sort by row diff
ph=ph[order(rowMaxs(as.matrix(ph)),decreasing = T),]
 colnames(ph)=c(-window:(ncol(ph)-(window+1)))
 ph= cbind(rowMaxs(as.matrix(ph)),ph);colnames(ph)[1]='diff'

outchunck_rowaverage[r,]=colMeans(ph[ph$diff>=maxBPP,-1,drop=F])
outchunck_frac[r,1]=nrow(ph[ph$diff>=maxBPP,-1,drop=F])




####################################################################
#### downstream Base Pairing Probability
####################################################################
### High base pair probability chr8:18628170-18628210
### low basepair probibility chr9:72581297-72581336

##################

s1dnp=introns_rand_p$CL-1
e1dnp=introns_rand_p$CL+buffer-1#(selects base +1 of start site but ends at correct 5' site)
chr=introns_rand_p$chr
strand=introns_rand_p$strand
  bedposdn=matrix(ncol=6,nrow=nrow(introns_rand_p))
  bedposdn[,1]=chr
  bedposdn[,2]=s1dnp
  bedposdn[,3]=e1dnp
  bedposdn[,4]=".";bedposdn[,5]="."
  bedposdn[,6]=strand
#################

s1dnn=introns_rand_n$CL-buffer
e1dnn=introns_rand_n$CL
chr=introns_rand_n$chr
strand=introns_rand_n$strand
  bednegdn=matrix(ncol=6,nrow=nrow(introns_rand_n))
  bednegdn[,1]=chr
  bednegdn[,2]=s1dnn
  bednegdn[,3]=e1dnn
  bednegdn[,4]=".";bednegdn[,5]="."
  bednegdn[,6]=strand

#################
bedposdn=rbind(bedposdn,bednegdn)
#################
  
write.table(bedposdn[,], file=paste0("./structure/chunkwindow/scr/bed_chunkRdn.bed"), quote=F, sep="\t", row.names=F, col.names=F)
system(paste0('rm -f ./structure/chunkwindow/scr/*.ps'))


system(paste0('bedtools getfasta -s -fi ',fasta,' -bed structure/chunkwindow/scr/bed_chunkRdn.bed | /Users/homanpj/Documents/Tools/ViennaRNA-2.4.14/bin/RNAfold -p  > out.fold'))
system('mv *.ps ./structure/chunkwindow/scr')

flsdn=list.files("./structure/chunkwindow/scr", pattern="dp.ps", all.files=FALSE,
    full.names=FALSE)  


outdn=mclapply(as.list(flsdn),RNAfold_output,mc.cores=8)
outdn2=outdn[[1]][,1:2,drop=F]
for (o in 1:length(outdn)) {
  if (nrow(outdn[[o]][,3,drop=F])==0) {outdn2$Xx=0;colnames(outdn2)[colnames(outdn2)%in%"Xx"]=colnames(outdn[[o]][,3,drop=F])}
  if (nrow(outdn[[o]][,3,drop=F])>0) {outdn2=cbind(outdn2,outdn[[o]][,3,drop=F])}
}

system(paste0('rm -f ./structure/chunkwindow/scr/*.ps'))
system(paste0('rm -f ./structure/chunkwindow/scr/*.fa'))
system(paste0('rm -f ./structure/chunkwindow/scr/*dn.bed'))

# #########################################
phdn=as.matrix(outdn2[,-c(1:2)])
phdn=phdn[is.na(phdn[,3])==F,]
phdn[is.nan(as.matrix(phdn))==T]=0
  phdn=t(phdn)
  phdn=as.data.frame(phdn)

  #### sort by row diff
phdn=phdn[order(rowMaxs(as.matrix(phdn)),decreasing = T),]
 colnames(phdn)=c(-window:(ncol(phdn)-(window+1)))
 phdn= cbind(rowMaxs(as.matrix(phdn)),phdn);colnames(phdn)[1]='diff'

outchunck_rowaverage_dn[r,]=colMeans(phdn[phdn$diff>=maxBPP,-1,drop=F])
outchunck_frac_dn[r,1]=nrow(phdn[phdn$diff>=maxBPP,-1,drop=F])

}
})


rownames(outchunck_rowaverage)=c(1:100)
colnames(outchunck_rowaverage)=c(-100:-10)

write.table(outchunck_rowaverage, file=paste0('./structure/chunkwindow/Random_outchunck_rowaverage_',window,'nt4.txt'), quote=F, sep="\t", row.names=T, col.names=NA)
write.table(outchunck_frac, file=paste0('./structure/chunkwindow/Random_outchunck_frac_',window,'nt4.txt'), quote=F, sep="\t", row.names=F, col.names=F)

rownames(outchunck_rowaverage_dn)=c(1:100)
colnames(outchunck_rowaverage_dn)=c(1:91)

write.table(outchunck_rowaverage_dn, file=paste0('./structure/chunkwindow/Random_dn_outchunck_rowaverage_',window,'nt4.txt'), quote=F, sep="\t", row.names=T, col.names=NA)
write.table(outchunck_frac_dn, file=paste0('./structure/chunkwindow/Random_dn_outchunck_frac_',window,'nt4.txt'), quote=F, sep="\t", row.names=F, col.names=F)
```

```{r ,fig.asp=.75, fig.align='center',fig.height=3, fig.width=6,echo=F,warning=F,eval=T,include=F}

######################################################################################################################################################################################################################################################################################################################
Rowaverage1=read.delim(paste0('./structure/chunkwindow/Random/Random_max/Random_outchunck_rowaverage_',window,'nt.txt'),comment.char = "%",header = F,sep = "\t")
Rowaverage2=read.delim(paste0('./structure/chunkwindow/Random/Random_max/Random_outchunck_rowaverage_',window,'nt2.txt'),comment.char = "%",header = F,sep = "\t")
Rowaverage3=read.delim(paste0('./structure/chunkwindow/Random/Random_max/Random_outchunck_rowaverage_',window,'nt3.txt'),comment.char = "%",header = F,sep = "\t")
Rowaverage4=read.delim(paste0('./structure/chunkwindow/Random/Random_max/Random_outchunck_rowaverage_',window,'nt4.txt'),comment.char = "%",header = F,sep = "\t")
outchunck_rowaverage=as.matrix(rbind(Rowaverage1[-1,-1],Rowaverage2[-1,-1],Rowaverage3[-1,-1]))
rownames(outchunck_rowaverage)=c(1:nrow(outchunck_rowaverage))
colnames(outchunck_rowaverage)=c(-100:-10)

# chunkout3
ph_mean=colMeans(ph[ph$diff>maxBPP,-1]);ph_mean=cbind(ph_mean,c(-100:-10));colnames(ph_mean)=c("value","position")
ph_mean=as.data.frame(ph_mean);ph_mean$X1="Real Data"

p=melt(outchunck_rowaverage);
p$X1=as.character(p$X1);
colnames(p)=c("X1","position","value")
ggplot(p,aes(y=value,x=position,line=X1))+
  geom_line(data=p,color="grey",na.rm=F)+
  geom_line(data=ph_mean,colour="red")+
  xlab("Position of 10nt Window\n(Absolute Distance from 5'Clip peak)")+ylab('Average basepairing probability')+
  theme_classic()+
  ylim(.5,1)+
  xlim(-window,1)
  


RowFrac1=read.delim(paste0('./structure/chunkwindow/Random/Random_max/Random_outchunck_frac_',window,'nt2.txt'),comment.char = "%",header = F,sep = "\t")
RowFrac2=read.delim(paste0('./structure/chunkwindow/Random/Random_max/Random_outchunck_frac_',window,'nt3.txt'),comment.char = "%",header = F,sep = "\t")
RowFrac3=read.delim(paste0('./structure/chunkwindow/Random/Random_max/Random_outchunck_frac_',window,'nt4.txt'),comment.char = "%",header = F,sep = "\t")
outchunck_fracx=as.matrix(rbind(RowFrac1,RowFrac2,RowFrac3))

p=as.data.frame(outchunck_fracx);colnames(p)="BPdiff"
ggplot(p,aes(x=BPdiff))+geom_density(size=1) + theme_classic()+ylab("Density")+xlab("# Locations with Max Basepair probibility > 0.7")+
  geom_vline(xintercept=nrow(ph[ph$diff>maxBPP,]),colour="red")



#####################################################################################################################################
RowaverageDn1=read.delim(paste0('./structure/chunkwindow/Random/Random_max/Random_dn_outchunck_rowaverage_',window,'nt.txt'),comment.char = "%",header = F,sep = "\t")
RowaverageDn2=read.delim(paste0('./structure/chunkwindow/Random/Random_max/Random_dn_outchunck_rowaverage_',window,'nt2.txt'),comment.char = "%",header = F,sep = "\t")
RowaverageDn3=read.delim(paste0('./structure/chunkwindow/Random/Random_max/Random_dn_outchunck_rowaverage_',window,'nt3.txt'),comment.char = "%",header = F,sep = "\t")
RowaverageDn4=read.delim(paste0('./structure/chunkwindow/Random/Random_max/Random_dn_outchunck_rowaverage_',window,'nt4.txt'),comment.char = "%",header = F,sep = "\t")
outchunck_RowaverageDnx=as.matrix(rbind(RowaverageDn1[-1,-1],RowaverageDn2[-1,-1],RowaverageDn3[-1,-1],RowaverageDn4[-1,-1]))
rownames(outchunck_RowaverageDnx)=c(1:nrow(outchunck_RowaverageDnx))
colnames(outchunck_RowaverageDnx)=c(1:91)

# ph_dn
ph_dn_mean=colMeans(ph_dn[ph_dn$diff>maxBPP,-1]);ph_dn_mean=cbind(ph_dn_mean,c(1:91));colnames(ph_dn_mean)=c("value","position")
ph_dn_mean=as.data.frame(ph_dn_mean);ph_dn_mean$X1="Real Data"


p=melt(outchunck_RowaverageDnx);p$X1=as.character(p$X1);colnames(p)=c("X1","position","value")
ggplot(p,aes(y=value,x=position,line=X1))+
  geom_line(data=p,color="grey",na.rm=F)+
  geom_line(data=ph_dn_mean,colour="red")+
  xlab("Position of 10nt Window\n(Absolute Distance from 5'Clip peak)")+ylab('Average basepairing probability')+theme_classic()+
  ylim(.5,.9)+
  xlim(1,window)


RowFracDn1=read.delim(paste0('./structure/chunkwindow/Random/Random_max/Random_dn_outchunck_frac_',window,'nt.txt'),comment.char = "%",header = F,sep = "\t")
RowFracDn2=read.delim(paste0('./structure/chunkwindow/Random/Random_max/Random_dn_outchunck_frac_',window,'nt2.txt'),comment.char = "%",header = F,sep = "\t")
RowFracDn3=read.delim(paste0('./structure/chunkwindow/Random/Random_max/Random_dn_outchunck_frac_',window,'nt3.txt'),comment.char = "%",header = F,sep = "\t")
RowFracDn4=read.delim(paste0('./structure/chunkwindow/Random/Random_max/Random_dn_outchunck_frac_',window,'nt4.txt'),comment.char = "%",header = F,sep = "\t")
outchunck_fracx=as.matrix(rbind(RowFracDn2,RowFracDn3))

p=as.data.frame(outchunck_frac);colnames(p)="BPdiff"
ggplot(p,aes(x=BPdiff))+geom_density(size=1) + theme_classic()+ylab("Density")+xlab("# Locations with Max Basepair probibility > 0.7")+
  geom_vline(xintercept=nrow(ph_dn[ph_dn$diff>maxBPP,]),colour="red")


```

## G-Quadraplex identification


```{r ,fig.asp=.5, fig.align='center',fig.height=2, fig.width=4,echo=F,warning=F,eval=T,include=F}

Peaksdata4_quad2=Peaksdata4[Peaksdata4$`Both Strand: RNA_type_Classification`%in%c('protein_coding: exon','protein_coding: Intron'),]
Peaksdata4_quad2=Peaksdata4[Peaksdata4$`Both Strand: RNA_type_Classification`%in%c('protein_coding: Intron'),]
# unique(Peaksdata4_motif$`Both Strand: RNA_type_Classification`)
Peaksdata4_quad2=separate(Peaksdata4_quad2,ID,into = c('chr','start'),sep=':',remove = F)
Peaksdata4_quad2=separate(Peaksdata4_quad2,start,into = c('start','end'),sep='-')
Peaksdata4_quad2$start=as.numeric(Peaksdata4_quad2$start)
Peaksdata4_quad2$end=as.numeric(Peaksdata4_quad2$end)
# unique(Peaksdata4_quad2$chr)

fasta='../../../../Resources/ref/mm10/Gencode_VM23/fromGencode/GRCm38.primary_assembly.genome.fa'

quadout2_Comb=as.data.frame(matrix(nrow = nrow(Peaksdata4_quad2),ncol=91))
quadout2_seq=as.data.frame(matrix(nrow = nrow(Peaksdata4_quad2),ncol=6))
quadout3=as.data.frame(matrix(nrow = 1,ncol=15))
colnames(quadout3)=c("V1","chr","V2","start","V3","end","strand","V4","V5","V6","V7","ID" ,"Gstart", "Gend","GID")

quadout2_dn_Comb=as.data.frame(matrix(nrow = nrow(Peaksdata4_quad2),ncol=91))
quadout2_dn_seq=as.data.frame(matrix(nrow = nrow(Peaksdata4_quad2),ncol=6))
quadout3_dn=as.data.frame(matrix(nrow = 1,ncol=15))
colnames(quadout3_dn)=c("V1","chr","V2","start","V3","end","strand","V4","V5","V6","V7","ID" ,"Gstart", "Gend","GID")

####################################################################
#### Upstream gQuad
####################################################################
### no basepairing chr18:77457443-77457501
### no low basepairing chr14:105137760-105137796 (bottom of heatmap)
### high diff in basepair probability chr6:55054497-55054526

window=100
buffer=100

 
for (x in 1:nrow(Peaksdata4_quad2)) {
# for (x in 1:5) {
if (x%in%seq(from=0,to=1000,by=25)){print(x/nrow(Peaksdata4_quad2))}
# {print(x)}

p=Peaksdata4_quad2[x,]
# p=Peaksdata4_quad2[Peaksdata4_quad2$ID%in%'chr6:55054497-55054526',]

chr=p$chr
start=p$start
end=p$end
strand=p$strand


##################
if (strand%in%'+') {
croslink=start
s1=start-buffer#(selects base +1 of start site but ends at correct 5' site)
e1=start

bedpos=matrix(ncol=6,nrow=1)
bedpos[1,1]=chr
bedpos[1,2]=s1
bedpos[1,3]=e1
bedpos[1,4]=".";bedpos[1,5]="."
bedpos[1,6]=strand
write.table(bedpos, file=paste0('./Gquad/',p$ID,"_quad.bed"), quote=F, sep="\t", row.names=F, col.names=F)
s0=start-buffer+1#(selects base +1 of start site but ends at correct 5' site)
e0=start

}

if (strand%in%'-') {
croslink=end
s1=end-1#(selects base +1 of start site but ends at correct 5' site)
e1=end+buffer-1
bedpos=matrix(ncol=6,nrow=1)
bedpos[1,1]=chr
bedpos[1,2]=s1
bedpos[1,3]=e1
bedpos[1,4]=".";bedpos[1,5]="."
bedpos[1,6]=strand
write.table(bedpos, file=paste0('./Gquad/',p$ID,"_quad.bed"), quote=F, sep="\t", row.names=F, col.names=F)
s0=end-0#(selects base +1 of start site but ends at correct 5' site)
e0=end+buffer-1
}

system(paste0('bedtools getfasta -s -fi ',fasta,' -bed Gquad/',p$ID,'_quad.bed > Gquad/',p$ID,'_quad.fa'))

quad=system(paste0('/Users/homanpj/Documents/Tools/fastaRegexFinder/fastaRegexFinder.py -f ./Gquad/',p$ID,'_quad.fa -r "([gG]{2,4}\\w{1,7}){3,}[gG]{2,4}" --noreverse > ./Gquad/out.fold.txt'),intern = T,ignore.stderr = T)


info = file.info(paste0('./Gquad/out.fold.txt'))
if (info$size>0){;quad=read.delim(paste0('./Gquad/out.fold.txt'),comment.char = "%",header = F,sep = "\t")}



quadout2_seq[x,1]=p$ID
quadout2_seq[x,2]=p$strand
quadout2_seq[x,3]=0


    (system(paste0('/Users/homanpj/Documents/Tools/qgrs-cpp/qgrs -s -csv -i ./Gquad/',p$ID,'_quad.fa -o quad_qgrs_all.txt'),intern = F,ignore.stderr = T));
          quad_qgrs_all=read.delim(paste0('quad_qgrs_all.txt'),comment.char = "%",header = T,sep = ",") 
        if (nrow(quad_qgrs_all)>0){
          quadout2_seq[x,6]=paste(quad_qgrs_all$GS,collapse = ", ")}


    

if (info$size>0) {
  
    quad=separate(quad,V1,into = c("chr",'strand'),remove = F,sep = "\\(")
    quad=separate(quad,chr,into = c("chr",'start'),remove = F,sep = ":")
    quad=separate(quad,start,into = c("start",'end'),remove = F,sep = "-")
    quad$start=as.numeric(quad$start)
    quad$end=as.numeric(quad$end)
    quad$ID=p$ID
 if (strand%in%'+') {      
    quad$Gstart=(quad$start+1)+quad$V2
    quad$Gend=(quad$start+1)+quad$V3}
if (strand%in%'-') {      
    quad$Gstart=(quad$end)-quad$V2
    quad$Gend=(quad$end)-(quad$V3-1)}    
    
    quad$GID=paste0(quad$chr,":",quad$Gstart,"-",quad$Gend)
    quadout3=rbind(quadout3,quad)  
    
  quadout2_seq[x,3]=nrow(quad)
  # quadout2_seq[x,4]=paste(quad$V1,collapse = )

       quadx=matrix(ncol=1,nrow=nrow(quad))
       quadx=matrix(ncol=1,nrow=1)
  for (n in 1){#:nrow(quad)) {
    # if (nrow(quad)>1) {xxx}
    write.table(as.character(quad[n,"V7"]), file=paste0('quad.txt'), quote=F, sep="\t", row.names=F, col.names=F)

    (system(paste0('/Users/homanpj/Documents/Tools/qgrs-cpp/qgrs -csv -i quad.txt -o quad_qgrs.txt'),intern = F,ignore.stderr = T));
    quad_qgrs=read.delim(paste0('quad_qgrs.txt'),comment.char = "%",header = T,sep = ",")  
    quadx[n,1]=quad_qgrs$GS
    
######################################################################
    if(file.exists(paste0('quad_qgrs.txt'))){file.remove(paste0('quad_qgrs.txt'))}
######################################################################
}
quadout2_seq[x,5]=paste(quadx[,1],collapse = ", ")

if (quadout2_seq[x,5]!=quadout2_seq[x,6]) {}

}

######################################################################
    if(file.exists(paste0('quad_qgrs_all.txt'))){file.remove(paste0('quad_qgrs_all.txt'))}

if(file.exists(paste0('Gquad/',p$ID,'_quad.fa'))){file.remove(paste0('Gquad/',p$ID,'_quad.fa'))}
if(file.exists(paste0('Gquad/',p$ID,'_quad.bed'))){file.remove(paste0('Gquad/',p$ID,'_quad.bed'))}
if(file.exists(paste0(p$chr,'_',s1,'-',e1,'(',strand,').fold.txt'))){file.remove(paste0(p$chr,'_',s1,'-',e1,'(',strand,').fold.txt'))}
######################################################################

}
quadout3=quadout3[is.na(quadout3[,3])==F,]
quadout3=separate(quadout3,ID,into = c('Pchr','Pstart'),sep = ":",remove = F)
quadout3=separate(quadout3,Pstart,into = c('Pstart','Pend'),sep = "-",remove = F)
# colnames(quad)

quadout3_p=quadout3[quadout3$strand%in%"+)",]
  quadout3_p$distance=abs(quadout3_p$Gend-as.numeric(quadout3_p$Pstart))
quadout3_n=quadout3[quadout3$strand%in%"-)",]
  quadout3_n$distance=abs(quadout3_n$Gstart-as.numeric(quadout3_n$Pend))
quadout3=rbind(quadout3_p,quadout3_n)

p=(as.data.frame(quadout3))
ggplot(p,aes(x=distance))+geom_density(size=1) + theme_classic()+ylab("Density")+xlab("Gquad Distance from 5' CLIP peak")
  # geom_vline(xintercept=nrow(ph_dn[ph_dn$diff>.7,]),colour="red")
  
quadout2_seq$comp=quadout2_seq$V5==quadout2_seq$V6
  

  
####################################################################
#### Downstream gQuad
####################################################################
for (x in 1:nrow(Peaksdata4_quad2)) {
# for (x in 1:5) {
if (x%in%seq(from=0,to=1000,by=25)){print(x/nrow(Peaksdata4_quad2))}
# {print(x)}

p=Peaksdata4_quad2[x,]
# p=Peaksdata4_quad2[Peaksdata4_quad2$ID%in%'chr6:55054497-55054526',]

chr=p$chr
start=p$start
end=p$end
strand=p$strand

##################
if (strand%in%'+') {
croslink=start
s1=start-1
e1=start+buffer-1#(selects base +1 of start site but ends at correct 5' site)


bedpos_dn=matrix(ncol=6,nrow=1)
bedpos_dn[1,1]=chr
bedpos_dn[1,2]=s1
bedpos_dn[1,3]=e1
bedpos_dn[1,4]=".";bedpos_dn[1,5]="."
bedpos_dn[1,6]=strand
write.table(bedpos_dn, file=paste0('./Gquad/',p$ID,"_quad_dn.bed"), quote=F, sep="\t", row.names=F, col.names=F)


}

if (strand%in%'-') {
croslink=end
s1=end-buffer
e1=end#(selects base +1 of start site but ends at correct 5' site)

bedpos_dn=matrix(ncol=6,nrow=1)
bedpos_dn[1,1]=chr
bedpos_dn[1,2]=s1
bedpos_dn[1,3]=e1
bedpos_dn[1,4]=".";bedpos_dn[1,5]="."
bedpos_dn[1,6]=strand
write.table(bedpos_dn, file=paste0('./Gquad/',p$ID,"_quad_dn.bed"), quote=F, sep="\t", row.names=F, col.names=F)
}

system(paste0('bedtools getfasta -s -fi ',fasta,' -bed Gquad/',p$ID,'_quad_dn.bed > Gquad/',p$ID,'_quad_dn.fa'))

quad=system(paste0('/Users/homanpj/Documents/Tools/fastaRegexFinder/fastaRegexFinder.py -f ./Gquad/',p$ID,'_quad_dn.fa -r "([gG]{2,4}\\w{1,7}){3,}[gG]{2,4}" --noreverse > ./Gquad/out.fold_dn.txt'),intern = T,ignore.stderr = T)


info = file.info(paste0('./Gquad/out.fold_dn.txt'))
if (info$size>0){quad=read.delim(paste0('./Gquad/out.fold_dn.txt'),comment.char = "%",header = F,sep = "\t")}


quadout2_dn_seq[x,1]=p$ID
quadout2_dn_seq[x,2]=p$strand
quadout2_dn_seq[x,3]=0

if (info$size>0) {
    quad=separate(quad,V1,into = c("chr",'strand'),remove = F,sep = "\\(")
    quad=separate(quad,chr,into = c("chr",'start'),remove = F,sep = ":")
    quad=separate(quad,start,into = c("start",'end'),remove = F,sep = "-")
    quad$start=as.numeric(quad$start)
    quad$end=as.numeric(quad$end)
    quad$ID=p$ID
 if (strand%in%'+') {      
    quad$Gstart=(quad$start+1)+quad$V2
    quad$Gend=(quad$start+1)+quad$V3}
if (strand%in%'-') {      
    quad$Gstart=(quad$end)-quad$V2
    quad$Gend=(quad$end)-(quad$V3-1)}    
    
    quad$GID=paste0(quad$chr,":",quad$Gstart,"-",quad$Gend)
    quadout3_dn=rbind(quadout3_dn,quad)  
    
  quadout2_dn_seq[x,3]=nrow(quad)
  # quadout2_dn_seq[x,4]=paste(quad$V1,collapse = )
  
         quadx_dn=matrix(ncol=1,nrow=nrow(quad))
         quadx_dn=matrix(ncol=1,nrow=1)
  for (n in 1){#:nrow(quad)) {
    # if (nrow(quad)>1) {xxx}
    write.table(as.character(quad[n,"V7"]), file=paste0('quad.txt'), quote=F, sep="\t", row.names=F, col.names=F)

    (system(paste0('/Users/homanpj/Documents/Tools/qgrs-cpp/qgrs -csv -i quad.txt -o quad_qgrs.txt'),intern = F,ignore.stderr = T));
    quad_qgrs_dn=read.delim(paste0('quad_qgrs.txt'),comment.char = "%",header = T,sep = ",")  
    quadx_dn[n,1]=quad_qgrs_dn$GS
    
    if(file.exists(paste0('quad_qgrs.txt'))){file.remove(paste0('quad_qgrs.txt'))}
}
quadout2_dn_seq[x,5]=paste(quadx_dn[,1],collapse = ", ")

}

######################################################################
if(file.exists(paste0('Gquad/',p$ID,'_quad_dn.fa'))){file.remove(paste0('Gquad/',p$ID,'_quad_dn.fa'))}
if(file.exists(paste0('Gquad/',p$ID,'_quad_dn.bed'))){file.remove(paste0('Gquad/',p$ID,'_quad_dn.bed'))}
if(file.exists(paste0(p$chr,'_',s1,'-',e1,'(',strand,').fold_dn.txt'))){file.remove(paste0(p$chr,'_',s1,'-',e1,'(',strand,').fold_dn.txt'))}
######################################################################

}

quadout3_Dn=quadout3_dn[is.na(quadout3_dn[,3])==F,]
quadout3_Dn=separate(quadout3_Dn,ID,into = c('Pchr','Pstart'),sep = ":",remove = F)
quadout3_Dn=separate(quadout3_Dn,Pstart,into = c('Pstart','Pend'),sep = "-",remove = F)
# colnames(quad)

quadout3_Dn_p=quadout3_Dn[quadout3_Dn$strand%in%"+)",]
  quadout3_Dn_p$distance=abs(quadout3_Dn_p$Gstart-as.numeric(quadout3_Dn_p$Pstart))
  
quadout3_Dn_n=quadout3_Dn[quadout3_Dn$strand%in%"-)",]
  quadout3_Dn_n$distance=abs(quadout3_Dn_n$Gend-as.numeric(quadout3_Dn_n$Pend))
quadout3_Dn=rbind(quadout3_Dn_p,quadout3_Dn_n)

p=(as.data.frame(quadout3_Dn))
ggplot(p,aes(x=distance))+geom_density(size=1) + theme_classic()+ylab("Density")+xlab("Gquad Distance from 5' CLIP peak")#+
  # geom_vline(xintercept=nrow(ph_dn[ph_dn$diff>.7,]),colour="red")

p1=as.data.frame(quadout3_Dn$distance);p1$region="Downstream";colnames(p1)[1]='distance'
p2=as.data.frame(quadout3$distance);p2$region="Upstream";colnames(p2)[1]='distance'
p=rbind(p1,p2)
ggplot(p,aes(x=distance,line=region,color=region))+geom_density(size=1) + theme_classic()+ylab("Density")+xlab("Gquad Distance from 5' CLIP peak")#+


p1=as.data.frame(quadout2_dn_seq$V5)
p1=separate(p1,1,into = c('V1','V2'),sep= ', ');p1=c(p1[,'V1',drop=T],p1[,'V2',drop=T]);p1=as.numeric(as.character(p1))
p1=as.data.frame(p1)
p1$region="Downstream";colnames(p1)[1]='Score'

p2=as.data.frame(quadout2_seq$V5)
p2=separate(p2,1,into = c('V1','V2'),sep= ', ');p2=c(p2[,'V1',drop=T],p2[,'V2',drop=T]);p2=as.numeric(as.character(p2))
p2=as.data.frame(p2)
p2$region="Upstream";colnames(p2)[1]='Score'
p=rbind(p1,p2)

ggplot(p,aes(x=Score,line=region,color=region))+geom_density(size=1,na.rm = T) + theme_classic()+ylab("Density")+xlab("Gquad score")#+

```

```{r ,fig.asp=.5, fig.align='center',fig.height=6, fig.width=12,echo=F,warning=F,eval=T,include=F}

Peaksdata4_quad2_out=Peaksdata4_quad2[,c('ID','strand',"Same Strand: Host_gene_name","Same Strand: Host_gene_ensembl_id","Same Strand: Repeat_Type","Both Strand: RNA_type_Classification")]

quadout3x=quadout3[,c('ID','GID','V7','distance','strand')];colnames(quadout3x)=c('ID','Upstream_Gquad_ID','Upstream_Gquad_sequence','Upstream_Distance','strandG')
 quadout3x=merge(quadout3x,quadout2_seq[,c('V1','V5')],all.x=T,by.x="ID",by.y='V1')
colnames(quadout3x)[colnames(quadout3x)%in%"V5"]='Upstream_Gquad_Score'

Peaksdata4_quad2_out=merge(quadout3x[,c('ID','Upstream_Gquad_ID','Upstream_Gquad_sequence','Upstream_Distance','Upstream_Gquad_Score')],Peaksdata4_quad2_out,all.y=T,by.y='ID',by.x="ID")

  
quadout3_dnx=quadout3_Dn[,c('ID','GID','V7','distance','strand')];colnames(quadout3_dnx)=c('ID','Downstream_Gquad_ID','Downstream_Gquad_sequence','Downstream_Distance','strandG')
quadout3_dnx=merge(quadout3_dnx,quadout2_dn_seq[,c('V1','V5')],all.x=T,by.x="ID",by.y='V1')
colnames(quadout3_dnx)[colnames(quadout3_dnx)%in%"V5"]='Dwonstream_Gquad_Score'

Peaksdata4_quad2_out=merge(quadout3_dnx[,c('ID','Downstream_Gquad_ID','Downstream_Gquad_sequence','Downstream_Distance','Dwonstream_Gquad_Score')],Peaksdata4_quad2_out,all.y=T,by.y='ID',by.x="ID")


write.table(Peaksdata4_quad2_out, file=paste0('./Gquad/CLIP_Peaks_Gquad_100nt.txt'), quote=F, sep="\t", row.names=T, col.names=T)

```



```{r ,fig.asp=.5, fig.align='center',fig.height=6, fig.width=12,echo=F,warning=F,eval=T,include=F}
#################################################
# Function to process RNAfold output
#################################################

######################################################################################################################################################################################################################################################################################################################


introns=fread("/Users/homanpj/OneDrive - National Institutes of Health/Resources/ref/mm10/Gencode_VM23/fromUCSC/KnownGene/KnownGene_GRCm38_introns.bed", header=F, sep="\t",stringsAsFactors = F,data.table=F)
  introns=introns[grepl("_",introns$V1)==F,]
  colnames(introns)=c('chr','start','end','attribute','V5','strand')
  introns=separate(introns,attribute,into=c('transcript_id','feature','exon_number','level','chr2','intronnumber','dir'),remove = T,sep = "_")
  introns$transcript_id=removeVersion(introns$transcript_id)
  # introns_canon=introns[introns$transcript_id%in%canonical$transcript_id,]
  introns$start=introns$start+1
  introns$exon_number=as.numeric(introns$exon_number)+1
  introns$diff=introns$end-introns$start
  # introns[introns$diff>300,]
  
fasta='../../../../Resources/ref/mm10/Gencode_VM23/fromGencode/GRCm38.primary_assembly.genome.fa'
fig=F
buffer=100
window=100  



quadout3_dist=matrix(nrow=buffer,ncol = (421) )
quadout3_count=matrix(nrow=buffer,ncol = 1 )
quadout3_Dn_dist=matrix(nrow=buffer,ncol = (421) )
quadout3_Dn_count=matrix(nrow=buffer,ncol = 1 )


for (r in 1:2) {
print(r)
  
quadout2_seq=as.data.frame(matrix(nrow = nrow(Peaksdata4_quad2),ncol=4))
quadout3=as.data.frame(matrix(nrow = 1,ncol=15))
colnames(quadout3)=c("V1","chr","V2","start","V3","end","strand","V4","V5","V6","V7","ID" ,"Gstart", "Gend","GID")

quadout2_dn_seq=as.data.frame(matrix(nrow = nrow(Peaksdata4_quad2),ncol=4))
quadout3_dn=as.data.frame(matrix(nrow = 1,ncol=15))
colnames(quadout3_dn)=c("V1","chr","V2","start","V3","end","strand","V4","V5","V6","V7","ID" ,"Gstart", "Gend","GID")
  
  
rand=sample(1:nrow(introns), nrow(Peaksdata4_chunk2_out), replace=TRUE)
introns_rand=introns[rand,]
introns_rand$CL=NA
for (x in 1:nrow(introns_rand)) {
  strt=introns_rand[x,'start']
  end=introns_rand[x,'end']
  introns_rand[x,'CL']=sample((strt-window):(end+window), 1, replace=TRUE)
}

####################################################################
#### Upstream Base Pairing Probability
#################################################################### 
##################
introns_rand_p=introns_rand[introns_rand$strand%in%"+",]
s1p=introns_rand_p$CL-buffer#(selects base +1 of start site but ends at correct 5' site)
e1p=introns_rand_p$CL
chr=introns_rand_p$chr
strand=introns_rand_p$strand
  bedpos=matrix(ncol=6,nrow=nrow(introns_rand_p))
  bedpos[,1]=chr
  bedpos[,2]=s1p
  bedpos[,3]=e1p
  bedpos[,4]=".";bedpos[,5]="."
  bedpos[,6]=strand
#################

introns_rand_n=introns_rand[introns_rand$strand%in%"-",]
s1n=introns_rand_n$CL-1#(selects base +1 of start site but ends at correct 5' site)
e1n=introns_rand_n$CL+buffer-1
chr=introns_rand_n$chr
strand=introns_rand_n$strand
s0=introns_rand_n$CL-0#(selects base +1 of start site but ends at correct 5' site)
e0=introns_rand_n$CL+buffer-1
  bedneg=matrix(ncol=6,nrow=nrow(introns_rand_n))
  bedneg[,1]=chr
  bedneg[,2]=s1n
  bedneg[,3]=e1n
  bedneg[,4]=".";bedneg[,5]="."
  bedneg[,6]=strand

#################
bedpos=rbind(bedpos,bedneg)
#################

write.table(bedpos, file=paste0('./Gquad/scr/',p$ID,"_quad.bed"), quote=F, sep="\t", row.names=F, col.names=F)

for (x in 1:nrow(bedpos)) {
  if (x%in%seq(from=0,to=1000,by=25)){print(x/nrow(Peaksdata4_quad2))}

 
system(paste0('bedtools getfasta -s -fi ',fasta,' -bed Gquad/scr/',p$ID,'_quad.bed > Gquad/',p$ID,'_quad.fa'))

quad=system(paste0('/Users/homanpj/Documents/Tools/fastaRegexFinder/fastaRegexFinder.py -f ./Gquad/scr/',p$ID,'_quad.fa -r "([gG]{2,4}\\w{1,7}){3,}[gG]{2,4}" --noreverse > ./Gquad/scr/out.fold.txt'),intern = T,ignore.stderr = T)

info = file.info(paste0('./Gquad/out.fold.txt'))
if (info$size>0){quad=read.delim(paste0('./Gquad/out.fold.txt'),comment.char = "%",header = F,sep = "\t")}


quadout2_seq[x,1]=p$ID
quadout2_seq[x,2]=p$strand
quadout2_seq[x,3]=0

if (info$size>0) {
    quad=separate(quad,V1,into = c("chr",'strand'),remove = F,sep = "\\(")
    quad=separate(quad,chr,into = c("chr",'start'),remove = F,sep = ":")
    quad=separate(quad,start,into = c("start",'end'),remove = F,sep = "-")
    quad$start=as.numeric(quad$start)
    quad$end=as.numeric(quad$end)
    quad$ID=p$ID
 if (strand%in%'+') {      
    quad$Gstart=(quad$start+1)+quad$V2
    quad$Gend=(quad$start+1)+quad$V3}
if (strand%in%'-') {      
    quad$Gstart=(quad$end)-quad$V2
    quad$Gend=(quad$end)-(quad$V3-1)}    
    
    quad$GID=paste0(quad$chr,":",quad$Gstart,"-",quad$Gend)
    quadout3=rbind(quadout3,quad)  
    
  quadout2_seq[x,3]=nrow(quad)
  # quadout2_seq[x,4]=paste(quad$V1,collapse = )
# system(paste0('rm -f ./structure/chunkwindow/scr/*.ps'))
system(paste0('rm -f ./structure/chunkwindow/scr/*.fa'))
system(paste0('rm -f ./structure/chunkwindow/scr/*.bed'))
} 

}
# #########################################

# #########################################
quadout3_dist[r,]=quadout3_Dn$distance
quadout3_count[r,1]=nrow(quadout2_Dn_seq[quadout2_Dn_seq$V3>0,])






####################################################################
#### downstream Base Pairing Probability
####################################################################
### High base pair probability chr8:18628170-18628210
### low basepair probibility chr9:72581297-72581336

##################

s1dnp=introns_rand_p$CL-1
e1dnp=introns_rand_p$CL+buffer-1#(selects base +1 of start site but ends at correct 5' site)
chr=introns_rand_p$chr
strand=introns_rand_p$strand
  bedposdn=matrix(ncol=6,nrow=nrow(introns_rand_p))
  bedposdn[,1]=chr
  bedposdn[,2]=s1dnp
  bedposdn[,3]=e1dnp
  bedposdn[,4]=".";bedposdn[,5]="."
  bedposdn[,6]=strand
#################

s1dnn=introns_rand_n$CL-buffer
e1dnn=introns_rand_n$CL
chr=introns_rand_n$chr
strand=introns_rand_n$strand
  bednegdn=matrix(ncol=6,nrow=nrow(introns_rand_n))
  bednegdn[,1]=chr
  bednegdn[,2]=s1dnn
  bednegdn[,3]=e1dnn
  bednegdn[,4]=".";bednegdn[,5]="."
  bednegdn[,6]=strand

#################
bedposdn=rbind(bedposdn,bednegdn)
#################
  
system(paste0('bedtools getfasta -s -fi ',fasta,' -bed Gquad/',p$ID,'_quad_dn.bed > Gquad/',p$ID,'_quad_dn.fa'))

quad=system(paste0('/Users/homanpj/Documents/Tools/fastaRegexFinder/fastaRegexFinder.py -f ./Gquad/',p$ID,'_quad_dn.fa -r "([gG]{2,4}\\w{1,7}){3,}[gG]{2,4}" --noreverse > ./Gquad/out.fold_dn.txt'),intern = T,ignore.stderr = T)


info = file.info(paste0('./Gquad/out.fold_dn.txt'))
if (info$size>0){quad=read.delim(paste0('./Gquad/out.fold_dn.txt'),comment.char = "%",header = F,sep = "\t")}


quadout2_Dn_seq[x,1]=p$ID
quadout2_Dn_seq[x,2]=p$strand
quadout2_Dn_seq[x,3]=0

if (info$size>0) {
    quad=separate(quad,V1,into = c("chr",'strand'),remove = F,sep = "\\(")
    quad=separate(quad,chr,into = c("chr",'start'),remove = F,sep = ":")
    quad=separate(quad,start,into = c("start",'end'),remove = F,sep = "-")
    quad$start=as.numeric(quad$start)
    quad$end=as.numeric(quad$end)
    quad$ID=p$ID
 if (strand%in%'+') {      
    quad$Gstart=(quad$start+1)+quad$V2
    quad$Gend=(quad$start+1)+quad$V3}
if (strand%in%'-') {      
    quad$Gstart=(quad$end)-quad$V2
    quad$Gend=(quad$end)-(quad$V3-1)}    
    
    quad$GID=paste0(quad$chr,":",quad$Gstart,"-",quad$Gend)
    quadout3_dn=rbind(quadout3_dn,quad)  
    
  quadout2_Dn_seq[x,3]=nrow(quad)
  # quadout2_Dn_seq[x,4]=paste(quad$V1,collapse = )

}
}
# system(paste0('rm -f ./structure/chunkwindow/scr/*.ps'))
system(paste0('rm -f ./structure/chunkwindow/scr/*.fa'))
system(paste0('rm -f ./structure/chunkwindow/scr/*dn.bed'))

# #########################################
quadout3_Dn_dist[r,]=quadout3_Dn$distance
quadout3_Dn_count[r,1]=nrow(quadout2_Dn_seq[quadout2_Dn_seq$V3>0,])



}


rownames(outchunck_rowaverage)=c(1:100)
colnames(outchunck_rowaverage)=c(-100:-10)

write.table(outchunck_rowaverage, file=paste0('./structure/chunkwindow/Random_outchunck_rowaverage_',window,'nt4.txt'), quote=F, sep="\t", row.names=T, col.names=T)
write.table(outchunck_frac, file=paste0('./structure/chunkwindow/Random_outchunck_frac_',window,'nt4.txt'), quote=F, sep="\t", row.names=F, col.names=F)

rownames(outchunck_rowaverage_dn)=c(1:100)
colnames(outchunck_rowaverage_dn)=c(1:91)

write.table(outchunck_rowaverage_dn, file=paste0('./structure/chunkwindow/Random_dn_outchunck_rowaverage_',window,'nt4.txt'), quote=F, sep="\t", row.names=T, col.names=T)
write.table(outchunck_frac_dn, file=paste0('./structure/chunkwindow/Random_dn_outchunck_frac_',window,'nt4.txt'), quote=F, sep="\t", row.names=F, col.names=F)
```

## DISTANCE TO SINE/LINE

```{r ,fig.asp=.5, fig.align='center',fig.height=6, fig.width=12,echo=F,warning=F,eval=T,include=F}
xxxxx
##################################
# BiocManager::install("regioneR")
library(regioneR)
# Peaksdata4_LS=Peaksdata4[Peaksdata4$`Both Strand: RNA_type_Classification`%in%c('protein_coding: exon','protein_coding: Intron'),]
# unique(Peaksdata4_motif$`Both Strand: RNA_type_Classification`)
Peaksdata4_LS=separate(Peaksdata4,ID,into = c('chr','start'),sep=':',remove = F)
Peaksdata4_LS=separate(Peaksdata4_LS,start,into = c('start','end'),sep='-')
Peaksdata4_LS$start=as.numeric(Peaksdata4_LS$start)
Peaksdata4_LS$end=as.numeric(Peaksdata4_LS$end)
peaks.GR = GRanges(seqnames=Peaksdata4_LS$chr, ranges=IRanges(start=Peaksdata4_LS$start,
end=Peaksdata4_LS$end),
strand=Peaksdata4_LS$strand)

rmsk_GRCm38=fread(paste0("/Users/homanpj/OneDrive - National Institutes of Health/Resources/ref/mm10/repeatmasker/rmsk_GRCm38.txt"), header=T, sep="\t",stringsAsFactors = F,data.table=F)
rmsk_GRCm38_LISI=rmsk_GRCm38[rmsk_GRCm38$repClass%in%c('LINE','SINE'),]
LISI.GR = GRanges(seqnames=rmsk_GRCm38_LISI$genoName, ranges=IRanges(start=rmsk_GRCm38_LISI$genoStart,
end=rmsk_GRCm38_LISI$genoEnd),
strand=rmsk_GRCm38_LISI$strand)
pt2 <- overlapPermTest(A=peaks.GR, B=LISI.GR, ntimes=100)
LZ2=localZScore(pt=pt2,A=peaks.GR,B=LISI.GR,B=LISI.GR, step=50,count.once=TRUE)



Peaksdata4_LS_IE=Peaksdata4_LS[Peaksdata4_LS$`Both Strand: RNA_type_Classification`%in%c('protein_coding: exon','protein_coding: Intron'),]
IntExn.GR = GRanges(seqnames=Peaksdata4_LS_IE$chr, ranges=IRanges(start=Peaksdata4_LS_IE$start,
end=Peaksdata4_LS_IE$end),
strand=Peaksdata4_LS_IE$strand)
# ptIntExn <- overlapPermTest(A=IntExn.GR, B=LISI.GR, ntimes=1000)
# LZ_IntExon=localZScore(pt=ptIntExn,A=peaks.GR,B=LISI.GR)
# plot(LZ_IntExon)
ptIntExn2 <- overlapPermTest(A=IntExn.GR, B=LISI.GR, ntimes=100)
LZ_IntExon2=localZScore(pt=ptIntExn2,A=peaks.GR,B=LISI.GR,window=1000, step=50)
plot(LZ_IntExon2)


##########
Peaksdata4_NC_IE=Peaksdata4_LS[Peaksdata4_LS$`Both Strand: RNA_type_Classification`%in%c('ncRNA'),]
NC.GR = GRanges(seqnames=Peaksdata4_NC_IE$chr, ranges=IRanges(start=Peaksdata4_NC_IE$start,
end=Peaksdata4_NC_IE$end),
strand=Peaksdata4_NC_IE$strand)
ptNC2 <- overlapPermTest(A=NC.GR, B=LISI.GR, ntimes=100)
LZ_NC2=localZScore(pt=ptNC2,A=peaks.GR,B=LISI.GR,window=2000, step=50)
plot(LZ_NC2)

```

## Counts Table : "CLIP_Peaks_MD10.txt"

**Columns:**  
"ID" - CLIP peak location    
"strand" - CLIP peak stand    
"Average MAPQ" - Average Mapping quality of reads aligned to peak region.  
"Counts_Unique" - Number of Unique reads aligned to CLIP peak  
"Counts_Multimappers_All" - Count of Unique + Multimapped read for each peak (Each occurrence of Multimapped read is counted separately)  
"Counts_Multimappers_Scaled"  - number of reads with Unique reads =1 and Multimapped =1/#MM locations             
"Counts_Multimappers_Best Mapping" - Number of Unique reads + best mapQ Multimapped Read aligned to CLIP peak  
"Multimaped Reads %" - Percent of Multimapped reads that align to CLIP peak   

Control:Counts_Unique" - Number of Unique reads aligned to CLIP peak in Control Sample  
"Control:Counts_Multimappers_All" - Count of Unique + Multimapped read for each peak in Control Sample (Each occurrence of Multimapped read is counted)    
"Control:Counts_Multimappers_Scaled"  - number of reads with Unique reads =1 and Multimapped =1/#MM locations in Control Sample      
"Control:Counts_Multimappers_BestMapping" - Number of Unique reads + best mapQ Multimapped Read aligned to CLIP peak in Control Sample  

"Same Strand: Host_gene_name" - Name of Host Gene on same strand as CLIP reads   
"Same Strand: Host_gene_ensembl_id" - Ensemble Name of Host Gene on same strand as CLIP reads   
"Same Strand: RNA_type" - Transcript type of CLIP peak host gene on same strand as CLIP reads   
"Same Strand: Repeat_Type" - Type of repeat in Clip peak (determined by repeat masker) on same strand as CLIP reads   
"Same Strand: Intron Exon" - Gene feature of CLIP peak location in protein coding Gene on same strand as CLIP reads   
"Same Strand: RNA_type_Classification" - Summary of Peak Classification 
(ncRNA > Protein coding : Exonic > repeats > Pseudogene > Protein Coding : Intronic > lncRNA > no Feature) 
"Same Strand: ncRNA_Classification" - Summary of ncRNA Peak assignments

"Opposite Strand: Host_gene_name" - Name of Host Gene on opposite strand as CLIP reads   
"Opposite Strand: Host_gene_ensembl_id" - Ensemble Name of Host Gene on opposite strand as CLIP reads   
"Opposite Strand: RNA_type" - Transcript type of CLIP peak host gene on opposite strand as CLIP reads   
"Opposite Strand: Repeat_Type" - Type of repeat in Clip peak (determined by repeat masker) on opposite strand as CLIP reads   
"Opposite Strand: Intron Exon" - Gene feature of CLIP peak location in protein coding Gene on opposite strand as CLIP reads   
"Opposite Strand: RNA_type_Classification" - Summary of Peak Classification 
(ncRNA > Protein coding : Exonic > repeats > Pseudogene > Protein Coding : Intronic > lncRNA > no Feature)  
"Opposite Strand: ncRNA_Classification" - Summary of ncRNA Peak assignments

"Both Strand: RNA_type_Classification" - Summary of Peak Classification including both strand annotations 
(ncRNA > Protein coding : Exonic > repeats > Pseudogene > Antisense Feature > Protein Coding : Intronic > lncRNA > no Feature)  

"MM_number" - # of reads in CLIP peak that aligns to other locations Genome                               
"MM_Alt_Peaktype" - # of Multimapped reads with Alternative Peak Classifications different from original Peak Classification                          
"MM_Same_Peaktype" - # of Multimapped reads with the same Peak Classification as the original Peak Classification      
"MM_biotype" - Peak Classifications for alternative mappings of multimapped reads                            
"MM_Prec_Alt_Peaktype" - % of Multimapped reads with Alternative Peak Classifications different from original Peak Classification                      
"MM_Alt_Alignments" - Location alternate alignments for Multimapped reads   

"iCount_peak" - 1= peak is identified by iCount  
<br>   



## Alignment Tracks   

**Ro_Clip_iCountcutadpt_ddup.s.all.bam** - MultiMapped + Uniquely aligned CLIP Reads   
**Ro_Clip_iCountcutadpt_ddup.s.unique.bam** - Uniquely aligned CLIP Reads   
**Ro_Clip_iCountcutadpt_ddup.s.multimapped.bam** - MultiMapped aligned CLIP Reads   
**Control_Clip_iCountcutadpt_ddup.s.all.bam** - MultiMapped + Uniquely aligned CLIP Reads in Control Sample     


<br>  
<br>   

## Annotation Tracks    
  
sftp://Biowulf.nih.gov/data/RBL_NCI/datashare/wolin/mESC_Clip/annotation_files/

   
<br>  
<br>  












