---
title: "Untitled" 
output: html_document
---
# Analysis of Intron + Exon CLIP peaks   

## Set Analysis mode  
CLIPtype:   
    Introns - Select peaks that are classified as "Protein Coding : Exons"   
    Exons - Select peaks that are classified as "Protein Coding : Introns"   
    Introns_LINESINETLR - Select peaks that are classified as Protein Coding : Introns and Intron peaks that are classified as "Repeat Elements" but overelap with LINE SINE or TLR   
       
RunMode:   
    Genomic - Include Intronic and Exonic Sequences when finding 100nt region Upsteram and Downstream of CLIP peak   
    Transcriptomic - Only include Exonic Sequences when finding 100nt region Upsteram and Downstream of CLIP peak. If Upstream/Downstream region crosses Intronic region, the intronic sequence is skpped and the next exon in included.   
       
TranscLoc: select subset of Protein Coding CLIP 
       
```{r ,echo=F, include=FALSE}
list=ls()
remove(list=list[list%in%c('Peaksdata4')==F])

##### Set Analysis mode

CLIPtype='Introns_LINESINETLR' #Introns or #Exons #Introns_LINESINETLR
RunMode='Genomic' #Genomic or Transcriptomic
TrasncLoc='all' # all, 3UTR, 5UTR, exons_CDS, introns



  scratch=paste0('./structure/chunkwindow/',CLIPtype,'/',TrasncLoc,'/scr')
  scratch_UP=paste0('./structure/chunkwindow/',CLIPtype,'/',TrasncLoc,'/upstream/scr')
  scratch_DN=paste0('./structure/chunkwindow/',CLIPtype,'/',TrasncLoc,'/downstream/scr')
  outdir_UP=paste0('./structure/chunkwindow/',CLIPtype,'/',TrasncLoc,'/upstream')
  outdir_DN=paste0('./structure/chunkwindow/',CLIPtype,'/',TrasncLoc,'/downstream')
  qgrs='/Users/homanpj/Documents/Tools/qgrs-cpp/qgrs'
  fastaRegexFinder= '/Users/homanpj/Documents/Tools/fastaRegexFinder/fastaRegexFinder.py'
  RNAfold=  '/Users/homanpj/Documents/Tools/ViennaRNA-2.4.14/bin/RNAfold'
  
  inRandom='./structure/Random/Random_Biowulf/V2'
  # inRandom='./structure/Random/Random_Biowulf/Exon_includeintron'
  
  LINESINEout=paste0('./structure/LINESINE/',CLIPtype,'/',TrasncLoc)
  
  
  window=100
  nrepeat=1
  meanBPP=.2
  maxBPP=.9
```  
  
```{r include=F ,echo=F,warning=F,message=FALSE}


# Peaksdata4=write.table(Peaksdata4, file=paste0('./structure/chunkwindow/Peaksdata4'), quote=F, sep="\t", row.names=F, col.names=T)

    Peaksdata4=read.delim('./structure/chunkwindow/Peaksdata4',header = T,sep = "\t")
colnames(Peaksdata4)=c("ID", "strand", "Average MAPQ",  "Counts_Unique", "Counts_Multimappers_All", "Counts_Multimappers_Scaled",  "Counts_Multimappers_BestMapping",
                       "Multimaped Reads %",
                       "Control:Counts_Unique", 
                       "Control:Counts_Multimappers_All", "Control:Counts_Multimappers_Scaled",
                       "Control:Counts_Multimappers_BestMapping",
                       "Same Strand: Host_gene_name",
                       "Same Strand: Host_gene_ensembl_id",
                       "Same Strand: RNA_type", 
                       "Same Strand: Repeat_Type",
                       "Same Strand: Intron Exon",
                       "Same Strand: RNA_type_Classification",
                       "Same.Strand: ncRNA_Classification",
                       "Opposite Strand: Host_gene_name",
                       "Opposite Strand: Host_gene_ensembl_id",
                       "Opposite Strand: RNA_type",
                       "Opposite Strand: Repeat_Type",
                       "Opposite Strand: Intron Exon",
                       "Opposite Strand: RNA_type_Classification",
                       "Opposite Strand: ncRNA_Classification",
                       "Both Strand: RNA_type_Classification",
                       "MM_number", "MM_Alt_Peaktype", "MM_Same_Peaktype", "MM_biotype", "MM_Prec_Alt_Peaktype", "MM_Alt_Alignments", "iCount_peak" )


  fasta='../../../../Resources/ref/mm10/Gencode_VM23/fromGencode/GRCm38.primary_assembly.genome.fa'
  intronbed="/Users/homanpj/OneDrive - National Institutes of Health/Resources/ref/mm10/Gencode_VM23/fromUCSC/KnownGene/KnownGene_GRCm38_exons.bed"
  
    library("ComplexHeatmap") ## version should be >=2.5
    # packageVersion("ComplexHeatmap")
    library(VariantAnnotation,quietly = T,verbose = F,warn.conflicts = F,logical.return = F)
    library(GenomicRanges,quietly = T,verbose = F,warn.conflicts = F,logical.return = F)
    library(trackViewer)
    library("pheatmap")
    library(vcfR)
    library(seqinr,quietly = T,verbose = F,warn.conflicts = F,logical.return = F)
    library(ggplot2,quietly = T,verbose = F,warn.conflicts = F,logical.return = F)
    library("viridis",quietly = T,verbose = F,warn.conflicts = F,logical.return = F)
    library('GenomicFeatures',quietly = T,verbose = F,warn.conflicts = F,logical.return = F)
    library('rtracklayer',quietly = T,verbose = F,warn.conflicts = F,logical.return = F)
    #library(GeneStructureTools)
    library(matrixStats,quietly = T,verbose = F,warn.conflicts = F,logical.return = F)
    library(plyr,quietly = T,verbose = F,warn.conflicts = F,logical.return = F)
    library(tidyr,quietly = T,verbose = F,warn.conflicts = F,logical.return = F)
    library(fitdistrplus,quietly = T,verbose = F,warn.conflicts = F,logical.return = F)
    library(stringr,quietly = T,verbose = F,warn.conflicts = F,logical.return = F)
    library(data.table)
    library(reshape)
    library(knitr)
library(stringi)
library(biomaRt)
library(plotly)
library(tidyr)
library(GenomicRanges)
library(RColorBrewer)
library('gplots')
library(ggpubr)
     library(circlize)
library('regioneR')
library(GeneStructureTools)

source("PARCLIP_Annotation_MD10.R")


blank_theme <- theme_minimal()+
  theme(
  axis.title.x = element_blank(),
  axis.title.y = element_blank(),
  panel.border = element_blank(),
  panel.grid=element_blank(),
  axis.ticks = element_blank(),
  plot.title=element_text(size=14, face="bold")
  )

incd_rRNA=T
```

```{r , include=FALSE}
  
  #################################################
  # Get Peaks data
  #################################################
 
########################################
# All Coding Transcripts
########################################
# Peaksdata4_chunkEx=Peaksdata4[Peaksdata4$`Both Strand: RNA_type_Classification`%in%'protein_coding: Intron'|
# Peaksdata4$`Both Strand: RNA_type_Classification`%in%'protein_coding: exon'|
# Peaksdata4$`Both Strand: RNA_type_Classification`%in%'Repeat Element'&Peaksdata4$`Same Strand: Repeat_Type`%in%c('LINE', 'SINE','LTR'),]

  
### Select peaks that are classified as Intron or Exon  
Peaksdata4_chunkEx1=Peaksdata4[(Peaksdata4$`Both Strand: RNA_type_Classification`)%in%c('protein_coding: Intron','protein_coding: exon') ,] 
  
### Select peaks that are classified as Repeat Element and OL with LINE SINE LTR only  
Peaksdata4_chunkEx2a=Peaksdata4[Peaksdata4$`Both Strand: RNA_type_Classification`%in%c('Repeat Element')&
                               grepl(c('LINE|SINE|LTR'),Peaksdata4$`Same Strand: Repeat_Type`),]
Peaksdata4_chunkEx2a=Peaksdata4_chunkEx2a[-grep(c('repeat'),Peaksdata4_chunkEx2a$`Same Strand: Repeat_Type`),]
Peaksdata4_chunkEx2a=Peaksdata4_chunkEx2a[-grep(c('complexity'),Peaksdata4_chunkEx2a$`Same Strand: Repeat_Type`),]
Peaksdata4_chunkEx2a=Peaksdata4_chunkEx2a[-grep(c('Satellite'),Peaksdata4_chunkEx2a$`Same Strand: Repeat_Type`),]
Peaksdata4_chunkEx2a=Peaksdata4_chunkEx2a[-grep(c('DNA'),Peaksdata4_chunkEx2a$`Same Strand: Repeat_Type`),]
Peaksdata4_chunkEx2a=Peaksdata4_chunkEx2a[grepl(c('Unknown'),Peaksdata4_chunkEx2a$`Opposite Strand: Repeat_Type`)==F,]
Peaksdata4_chunkEx2a=Peaksdata4_chunkEx2a[grepl(c('Other'),Peaksdata4_chunkEx2a$`Opposite Strand: Repeat_Type`)==F,]

### Combine Regions
Peaksdata4_chunkEx=rbind(Peaksdata4_chunkEx1,Peaksdata4_chunkEx2a)
  Peaksdata4_chunkEx=Peaksdata4_chunkEx[duplicated(Peaksdata4_chunkEx$ID)==F,]
  
### PEAK MUST OVERLAP WITH INTRON OR EXON  
Peaksdata4_chunkEx=Peaksdata4_chunkEx[is.na(Peaksdata4_chunkEx$`Same Strand: Intron Exon`)==F,]

########################################

### Select peaks that are classified as Antisense Feature 
Peaksdata4_chunkEx2b=Peaksdata4[Peaksdata4$`Both Strand: RNA_type_Classification`%in%c('Antisense Feature')&
                               grepl(c('LINE|SINE|LTR'),Peaksdata4$`Opposite Strand: Repeat_Type`),]

### Select peaks that are classified as Antisense Feature and remove repeeat type except  LINE SINE LTR
Peaksdata4_chunkEx2b=Peaksdata4_chunkEx2b[-grep(c('repeat'),Peaksdata4_chunkEx2b$`Opposite Strand: Repeat_Type`),]
Peaksdata4_chunkEx2b=Peaksdata4_chunkEx2b[-grep(c('complexity'),Peaksdata4_chunkEx2b$`Opposite Strand: Repeat_Type`),]
Peaksdata4_chunkEx2b=Peaksdata4_chunkEx2b[-grep(c('Satellite'),Peaksdata4_chunkEx2b$`Opposite Strand: Repeat_Type`),]
Peaksdata4_chunkEx2b=Peaksdata4_chunkEx2b[grepl(c('DNA'),Peaksdata4_chunkEx2b$`Opposite Strand: Repeat_Type`)==F,]
Peaksdata4_chunkEx2b=Peaksdata4_chunkEx2b[grepl(c('Unknown'),Peaksdata4_chunkEx2b$`Opposite Strand: Repeat_Type`)==F,]
Peaksdata4_chunkEx2b=Peaksdata4_chunkEx2b[grepl(c('Other'),Peaksdata4_chunkEx2b$`Opposite Strand: Repeat_Type`)==F,]

Peaksdata4_chunkEx2b=Peaksdata4_chunkEx2b[Peaksdata4_chunkEx2b$ID%in%Peaksdata4_chunkEx$ID==F,]

Peaksdata4_chunkEx=rbind(Peaksdata4_chunkEx,Peaksdata4_chunkEx2b)
Peaksdata4_chunkEx=Peaksdata4_chunkEx[duplicated(Peaksdata4_chunkEx$ID)==F,]

### PEAK MUST OVERLAP WITH INTRON OR EXON on either strand
# Peaksdata4_chunkEx=Peaksdata4_chunkEx[(is.na(Peaksdata4_chunkEx$`Same Strand: Host_gene_ensembl_id`)==F)|(is.na(Peaksdata4_chunkEx$`Opposite Strand: Host_gene_ensembl_id`)==F),]
Peaksdata4_chunkEx=Peaksdata4_chunkEx[(is.na(Peaksdata4_chunkEx$`Same Strand: Intron Exon`)==F)|(is.na(Peaksdata4_chunkEx$`Opposite Strand: Intron Exon`)==F),]


Peaksdata4_chunkEx=separate(Peaksdata4_chunkEx,col = ID,into = c('chr','start'),sep=':',remove = F)
Peaksdata4_chunkEx=separate(Peaksdata4_chunkEx,start,into = c('start','end'),sep='-')
Peaksdata4_chunkEx$start=as.numeric(Peaksdata4_chunkEx$start)
Peaksdata4_chunkEx$end=as.numeric(Peaksdata4_chunkEx$end)

########################################################################################################################################################################

  chunkname3_exon=as.data.frame(matrix(nrow=1,ncol=nrow(Peaksdata4_chunkEx)+2))
  chunkout3_exon=as.data.frame(matrix(nrow=window,ncol=nrow(Peaksdata4_chunkEx)+2))
  chunkout3_exon[,1]=seq(1,window,1)
  chunkout3_max_exon=as.data.frame(matrix(nrow=window,ncol=nrow(Peaksdata4_chunkEx)+2))
  chunkout3_max_exon[,1]=seq(1,window,1)
seqout3_exon=as.data.frame(matrix(nrow=window,ncol=nrow(Peaksdata4_chunkEx)+2))
seqout3_exon[,1]=seq(1,window,1)
seqout3_max_exon=as.data.frame(matrix(nrow=window,ncol=nrow(Peaksdata4_chunkEx)+2))
seqout3_max_exon[,1]=seq(1,window,1)    
  
   #################################################
  # Function to process RNAfold output
  #################################################
  
  RNAfold_output=function(fls){
    chunkRout3=as.data.frame(matrix(nrow=window,ncol=4))
    chunkRname3=as.data.frame(matrix(nrow=1,ncol=3))
    chunkRout3[,1]=seq(1,window,1)
    chunkRout3_max=as.data.frame(matrix(nrow=window,ncol=4))
    
    # for (x in 1:length(fls)) {
    # if (x%in%seq(from=0,to=1000,by=25)){print(x/length(fls))}
    # fls=as.list(fls)
    # fls=fls[117]
    fls=unlist(fls)
    
    chunkR=read.delim(paste0(fls[1]),comment.char = "%",header = F,sep = "\t")
    chunkR=chunkR[grep("ubox",chunkR$V1),]
    chunkR=as.data.frame(chunkR[-1],stringsAsFactors=F)
    chunkR$`chunkR[-1]`=as.character(chunkR$`chunkR[-1]`);colnames(chunkR)="V1"
    chunkR=separate(chunkR,'V1',sep = " ",into = c('b1','b2','pvalue','ubox'))
    chunkR$pvalue=as.numeric(chunkR$pvalue)
    chunkR$b1=as.numeric(chunkR$b1)
    chunkR$b2=as.numeric(chunkR$b2)
    
    
    ### count each base
    for (y in 1:(window-9)) {
      ch=chunkR[chunkR$b1%in%seq(y,y+9,1)|chunkR$b2%in%seq(y,y+9,1),]
      ch2=ch
      #   
      if (nrow(ch2)>0) {
        ch2$br1=rowMins(as.matrix(ch[,c('b1','b2')]))
        ch2$br2=rowMaxs(as.matrix(ch[,c('b1','b2')]))
        ch$b1=ch2$br1
        ch$b2=ch2$br2
        ch=ch[duplicated(ch)==F,]
        
        chunkRout3[y,1]=y
        chunkRout3[y,2]=paste0(y,'-',y+9)
        chunkRout3[y,3]=mean(ch$pvalue) ##average all basepairinng accross chunck
        #     
        chunkRout3_max[y,1]=y
        chunkRout3_max[y,2]=paste0(y,'-',y+9)
        chunkRout3[y,4]=max(ch$pvalue) ##max all basepairinng accross chunck
      }
      if (nrow(ch2)==0) {
        chunkRout3[y,1]=y
        chunkRout3[y,2]=paste0(y,'-',y+9)
        chunkRout3[y,3]=0
        chunkRout3_max[y,1]=y
        chunkRout3_max[y,2]=paste0(y,'-',y+9)
        chunkRout3[y,4]=0
      }
      chunkRname3[1,3]=gsub("_dp.ps","",fls[1])
      
    }
    
    chunkRname3=t(chunkRname3)
    colnames(chunkRout3)[3:ncol(chunkRout3)]=chunkRname3[3:nrow(chunkRname3),1]
    
    chunkRout3=chunkRout3[is.na(chunkRout3[,3])==F,]
    chunkRout3[is.nan(as.matrix(chunkRout3))==T]=0
    
    return(chunkRout3)
  } 
  
  
  #################################################
  # Get Genome files
  #################################################
  
  
canonical=fread("/Users/homanpj/OneDrive - National Institutes of Health/Resources/ref/mm10/Gencode_VM23/fromUCSC/KnownCanonical/KnownCanonical_GencodeM23_GRCm38.txt", header=T, sep="\t",stringsAsFactors = F,data.table=F)
  canonical$transcript=removeVersion(canonical$transcript)
  canonical$protein=removeVersion(canonical$protein)
  
mm10all=fread("/Users/homanpj/OneDrive - National Institutes of Health/Resources/ref/mm10/Gencode_VM23/fromGencode/gencode.vM23.basic.annotation.gtf.txt", header=T, sep="\t",stringsAsFactors = F,data.table=F)
  mm10all$transcript_id=removeVersion(mm10all$transcript_id)
  mm10all$gene_id=removeVersion(mm10all$gene_id)
  
mm10_trans=mm10all[mm10all$feature%in%c("transcript"),]  
  
  mm10_exon=mm10all[mm10all$feature%in%c("exon"),]
  mm10_exon$start=as.numeric(as.character(mm10_exon$start))
  mm10_exon$end=as.numeric(as.character(mm10_exon$end))
  mm10_exon=mm10_exon[mm10_exon$transcript_id%in%mm10_trans$transcript_id,]



introns=fread("/Users/homanpj/OneDrive - National Institutes of Health/Resources/ref/mm10/Gencode_VM23/fromUCSC/KnownGene/KnownGene_GRCm38_introns.bed", header=F, sep="\t",stringsAsFactors = F,data.table=F)
  introns=introns[grepl("_",introns$V1)==F,]
  colnames(introns)=c('seqname','start','end','attribute','V5','strand')
  introns=separate(introns,attribute,into=c('transcript_id','feature','exon_number','level','chr2','intronnumber','dir'),remove = T,sep = "_")
  introns$transcript_id=removeVersion(introns$transcript_id)
      # introns_canon=introns[introns$transcript_id%in%canonical$transcript_id,]
  introns$start=introns$start+1
  introns$exon_number=as.numeric(introns$exon_number)+1

 introns=introns[introns$transcript_id%in%mm10_exon$transcript_id,]
 introns=merge(introns,mm10_trans[,c('transcript_id','gene_id')],by='transcript_id',all.x=T)
  
###################################  
  iso=read.delim('../mESC_102218/4-Analysis/Wolin.mECS_102218/inst/extdata/RSEM.isoforms.FPKM.all_samples.txt', header=T, sep="\t",stringsAsFactors = F)
iso$WT=rowMeans(iso[,c('WTQ3_FPKM','WTQ4_FPKM')])
iso$gene_id=removeVersion(iso$gene_id)
iso$transcript_id=removeVersion(iso$transcript_id)
iso=merge(iso,mm10_trans[,c('transcript_id','start','end')],by='transcript_id',all.x=T)
colnames(iso)[colnames(iso)%in%c('start','end')]=c('start_iso','end_iso')

gene=read.delim('../mESC_102218/4-Analysis/Wolin.mECS_102218/inst/extdata/RSEM.genes.FPKM.all_samples.txt', header=T, sep="\t",stringsAsFactors = F)
gene$WT=rowMeans(gene[,c('WTQ3_FPKM','WTQ4_FPKM')])
gene$gene_id=removeVersion(gene$gene_id)


 if (CLIPtype=='Introns'){mm10_USE=introns}
 if (CLIPtype=='Introns_LINESINETLR'){mm10_USE=introns}
   if (CLIPtype=='Exons'){mm10_USE=mm10_exon}


  
  #################################################
  # Select apropriate isoforms for each
  #################################################
  
  isoout=as.data.frame(matrix(nrow = nrow(Peaksdata4_chunkEx),ncol=16));colnames(isoout)=c('ID','strand','gene_id','CL','maxiso','maxiso_count','maxiso_start','maxiso_end','maxiso_OL','maxiso2','maxiso2_count','maxiso2_start','maxiso2_end','maxiso2_OL','diff',"IScanonical")
  
  for (e in 1:nrow(Peaksdata4_chunkEx)) {
    p=Peaksdata4_chunkEx[e,]
    
    if (p$strand=='+') {isoout[e,"CL"]=p$start }
    if (p$strand=='-') {isoout[e,"CL"]=p$end }

    g=p$`Same Strand: Host_gene_ensembl_id`
    if (is.na(g)) {g=p$`Opposite Strand: Host_gene_ensembl_id` }
    
      if (length(strsplit(g,",")[[1]])>1) { ##split if multiple genes
        g=(strsplit(g,",")[[1]]) 
        gexp=gene[gene$gene_id%in%g,]
        gexp=gexp[order(gexp$WT,decreasing = T),]
        if (nrow(gexp)>1) {
              # if ((gexp[1,'WT']/gexp[2,'WT'])<1.5) {xxxxGene}
        g=gexp[1,'gene_id']
      }}  

    isog=iso[iso$gene_id%in%g,]
    isog=isog[order(isog$WT,decreasing = T),]
    isogMax=isog[1,'WT']
    isogMax2=isog[2,'WT']
    
   if  (is.na(isogMax2)==F&isogMax!=0){
      if (((isogMax-isogMax2)/isogMax)<.2) {
        isoout[e,'maxiso']=isog[1,'transcript_id']
        isoout[e,'maxiso_count']=isog[1,'WT']
        isoout[e,'maxiso_start']=isog[1,'start_iso']
        isoout[e,'maxiso_end']=isog[1,'end_iso']

        isoout[e,'maxiso2']=isog[2,'transcript_id']
        isoout[e,'maxiso2_count']=isog[2,'WT']
        isoout[e,'maxiso2_start']=isog[2,'start_iso']
        isoout[e,'maxiso2_end']=isog[2,'end_iso']
        isoout[e,'diff']=((isogMax-isogMax2)/isogMax)
      }}
  isoout[e,'ID']=p$ID
  isoout[e,'strand']=p$strand
  isoout[e,'gene_id']=isog[1,'gene_id']
  isoout[e,'maxiso']=isog[1,'transcript_id']
  isoout[e,'maxiso_count']=isog[1,'WT']
  isoout[e,'maxiso_start']=isog[1,'start_iso']
  isoout[e,'maxiso_end']=isog[1,'end_iso']
  isoout[e,'IScanonical']=0
    i=isoout[e,c("maxiso","maxiso2"),drop=T]
    i=i[i%in%canonical$transcript]
      if (length(i)>0) {isoout[e,'IScanonical']=i}
  }
        isoout[,'maxiso_OL']=(isoout[,'CL']>isoout[,'maxiso_start'])&(isoout[,'CL']<isoout[,'maxiso_end'])
        isoout[,'maxiso2_OL']=(isoout[,'CL']>isoout[,'maxiso2_start'])&(isoout[,'CL']<isoout[,'maxiso2_end'])

   isoout=merge(isoout,canonical[,c('transcript','protein','chromStart','chromEnd')],by.x='gene_id',by.y='protein',all.x=T)
   isoout$CanOL=(isoout[,'CL']>isoout[,'chromStart'])&(isoout[,'CL']<isoout[,'chromEnd'])
   
   isoout[isoout$maxiso_OL==T|isoout$maxiso2_OL==T|isoout$CanOL==T,]


  isoout$USEtranscript=0
    ## if no close sexond use max
      criteria=which(is.na(isoout$maxiso2)&(isoout$maxiso_OL==T))
  isoout[criteria,'USEtranscript']=isoout[criteria,'maxiso']
    ## if close sexond  and one of top two is  Canonicol use Cannonical
        criteria=which((is.na(isoout$maxiso2)==F)&(isoout$maxiso2_OL==T))
  isoout[criteria,'USEtranscript']=isoout[criteria,'IScanonical']
    ## if close sexond  and one of top two is NOT Canonicol use Cannonical
          criteria=which((isoout$USEtranscript)%in%'0'&(isoout$CanOL==T))
  isoout[criteria,c('USEtranscript')]=isoout[criteria,c('transcript')]
  
   ## if still no CanOL chec to see if max still OL
        criteria=which((isoout$USEtranscript=="0")&(isoout$maxiso_OL==T))
  isoout[criteria,'USEtranscript']=isoout[criteria,'maxiso']
          criteria=which((isoout$USEtranscript=="0")&(isoout$maxiso2_OL==T))
  isoout[criteria,'USEtranscript']=isoout[criteria,'maxiso2']
  
  mm10_transGR=GRanges(seqnames=mm10_trans$seqname, 
                          ranges=IRanges(start=mm10_trans$start,end=mm10_trans$end),
                          strand=mm10_trans$strand,
                          width2=(mm10_trans$end-mm10_trans$start),
                          transcript_id=mm10_trans$transcript_id)
  
  
  isoutcorrect=isoout[isoout$USEtranscript=="0",]
  isoutcorrect=separate(isoutcorrect,ID,into=c('CLIPchr','CLIPstart'),sep = ":",remove=F)
  isoutcorrect=separate(isoutcorrect,CLIPstart,into=c('CLIPstart','CLIPend'),sep = "-",remove=T)
  isoutcorrect$CLIPstart=as.numeric(isoutcorrect$CLIPstart)
  isoutcorrect$CLIPend=as.numeric(isoutcorrect$CLIPend)
  for (x in 1:nrow(isoutcorrect)) {
    a=isoutcorrect[x,]
    a_gR=GRanges(seqnames=a$CLIPchr, 
                          ranges=IRanges(start=as.numeric(a$CLIPstart),end=as.numeric(a$CLIPend)),
                          strand=a$strand                   )
    
              gr = a_gR
              gr2 = mm10_transGR
              xo=as.data.frame(GenomicRanges::findOverlaps(gr,gr2,type = "any",ignore.strand=F))
          
              TransOut_GR=as.data.frame(gr2[xo$subjectHits])
              aGR=as.data.frame(gr[xo$queryHits])
              TransOut_GR=TransOut_GR[order(TransOut_GR$width,decreasing = T),]
              isoutcorrect[x,'USEtranscript']=TransOut_GR[1,'transcript_id']
              
  }
  
  isoout=rbind(isoout[(isoout$ID%in%isoutcorrect$ID)==F,],isoutcorrect[,colnames(isoutcorrect)%in%colnames(isoout)])
  
```
  



```{r , include=FALSE}
## Overlap with Transcript Component  

  Nlocations=nrow(Peaksdata4_chunkEx)
  UseCores=8

isoout=separate(isoout,ID,into = c('chr','start'),sep = ":",remove = F)
isoout=separate(isoout,start,into = c('start','end'),sep = "-",remove = T)
 
CrossLink_GR=GRanges(seqnames=isoout$chr, ranges=IRanges(start=isoout$CL,
end=isoout$CL),
strand=isoout$strand,
ID=isoout$ID,
USEtranscript=isoout$USEtranscript)  

################################################
#### create Genomic files
################################################


  mm10clip_trans=mm10all[mm10all$feature%in%'transcript',]
  mm10clip_UTR=mm10all[mm10all$feature%in%'UTR',]
  mm10clip_exon=mm10all[mm10all$feature%in%c("exon"),]

  
  
  ###############
  
mm10clip_3UTRall=fread("/Users/homanpj/OneDrive - National Institutes of Health/Resources/ref/mm10/Gencode_VM23/fromUCSC/KnownGene/KnownGene_GRCm38_GencodeV23_3UTR.bed", header=F, sep="\t",stringsAsFactors = F,data.table=F)
  mm10clip_3UTRall=mm10clip_3UTRall[grepl("_",mm10clip_3UTRall$V1)==F,]
  colnames(mm10clip_3UTRall)=c('chr','start','end','attribute','V5','strand')
  mm10clip_3UTRall=separate(mm10clip_3UTRall,attribute,into=c('transcript_id','feature','exon_number','level','chr2','start+1','dir'),remove = T,sep = "_")
  mm10clip_3UTRall$transcript_id=removeVersion(mm10clip_3UTRall$transcript_id)
  mm10clip_3UTRall$start=mm10clip_3UTRall$start+1
  mm10clip_3UTRall$exon_number=as.numeric(mm10clip_3UTRall$exon_number)+1
  mm10clip_3UTRall$width=(mm10clip_3UTRall$end-mm10clip_3UTRall$start)+1
          mm10clip_3UTR=mm10clip_3UTRall[mm10clip_3UTRall$transcript_id%in%mm10clip_trans$transcript_id,]
  

############################################################################################

mm10clip_intron=fread("/Users/homanpj/OneDrive - National Institutes of Health/Resources/ref/mm10/Gencode_VM23/fromUCSC/KnownGene/KnownGene_GRCm38_introns.bed", header=F, sep="\t",stringsAsFactors = F,data.table=F)
  mm10clip_intron=mm10clip_intron[grepl("_",mm10clip_intron$V1)==F,]
  colnames(mm10clip_intron)=c('chr','start','end','attribute','V5','strand')
  mm10clip_intron=separate(mm10clip_intron,attribute,into=c('transcript_id','feature','exon_number','level','chr2','intronnumber','dir'),remove = T,sep = "_")
  mm10clip_intron$transcript_id=removeVersion(mm10clip_intron$transcript_id)
  mm10clip_intron$start=mm10clip_intron$start+1
  mm10clip_intron$exon_number=as.numeric(mm10clip_intron$exon_number)+1
  mm10clip_intron=merge(mm10clip_intron,mm10_trans[,c('transcript_id','gene_id')],by='transcript_id',all.x=T)

  
################################################
  
  mm10clip_trans=mm10clip_trans[mm10clip_trans$transcript_id%in%isoout$USEtranscript,]
  mm10clip_UTR=mm10clip_UTR[mm10clip_UTR$transcript_id%in%isoout$USEtranscript,]
  mm10clip_3UTR=mm10clip_3UTRall[mm10clip_3UTRall$transcript_id%in%isoout$USEtranscript,]
  mm10clip_exon=mm10clip_exon[mm10clip_exon$transcript_id%in%isoout$USEtranscript,]
  mm10clip_intron=mm10clip_intron[mm10clip_intron$transcript_id%in%isoout$USEtranscript,]
  
################################################
intronsclip_GR=GRanges(seqnames=mm10clip_intron$chr,
                   ranges=IRanges(start=mm10clip_intron$start,end=mm10clip_intron$end),
                   strand=mm10clip_intron$strand,
                   transcript_id=mm10clip_intron$transcript_id,
                   exon_number=mm10clip_intron$exon_number
                   )
  

############################################################################################
 
exonsclip_GR=GRanges(seqnames=mm10clip_exon$seqname,
                   ranges=IRanges(start=mm10clip_exon$start,end=mm10clip_exon$end),
                   strand=mm10clip_exon$strand,
                   transcript_id=mm10clip_exon$transcript_id,
                   gene_id=mm10clip_exon$gene_id,
                   exon_number=mm10clip_exon$exon_number
                   )


############################

 TranscUTRclip_GR=GRanges(seqnames=mm10clip_UTR$seqname,
                   ranges=IRanges(start=mm10clip_UTR$start,end=mm10clip_UTR$end),
                   strand=mm10clip_UTR$strand,
                   transcript_id=mm10clip_UTR$transcript_id,
                   gene_id=mm10clip_UTR$gene_id,
                   exon_number=mm10clip_UTR$exon_number
                   )


############################

 Transc3UTRclip_GR=GRanges(seqnames=mm10clip_3UTR$chr,
                   ranges=IRanges(start=mm10clip_3UTR$start,end=mm10clip_3UTR$end),
                   strand=mm10clip_3UTR$strand,
                   transcript_id=mm10clip_3UTR$transcript_id,
                   # gene_id=mm10clip_3UTR$gene_id,
                   exon_number=mm10clip_3UTR$exon_number
                   )


############################

Transcclip_GR=GRanges(seqnames=mm10clip_trans$seqname,
                   ranges=IRanges(start=mm10clip_trans$start,end=mm10clip_trans$end),
                   strand=mm10clip_trans$strand,
                   transcript_id=mm10clip_trans$transcript_id,
                  exon_number=mm10clip_trans$exon_number,
                   gene_id=mm10clip_trans$gene_id
                   )


################################################
#### create Genomic overlaps
################################################


### Get Transc that overlap with CLIP peaks
        gr = CrossLink_GR
        gr2 = Transcclip_GR
        xo=as.data.frame(GenomicRanges::findOverlaps(gr,gr2,type = "any",ignore.strand=F))
          
        Transc_CrossLink_GR=gr2[xo$subjectHits]
                  # qh=as.data.frame(gr[xo$queryHits],row.names = NULL)
                  # colnames(qh)=paste0(colnames(qh),"_mm10")
                  # sh=as.data.frame(gr2[xo$subjectHits],row.names = NULL)
        Crosslink_Transc_GR=gr[xo$queryHits]  
    Crosslink_Transc_GR_DF=as.data.frame(Crosslink_Transc_GR)
    Crosslink_Transc_GR_DF$Transc='Transc'
    
    isoout$Transc=NA
    isoout[which(isoout$ID%in%Crosslink_Transc_GR_DF$ID),'Transc']="Transc"
    

### Get Exon that overlap with CLIP peaks
        gr = CrossLink_GR
        gr2 = exonsclip_GR
        xo=as.data.frame(GenomicRanges::findOverlaps(gr,gr2,type = "any",ignore.strand=F))
          
        exons_CrossLink_GR=gr2[xo$subjectHits]
                  # qh=as.data.frame(gr[xo$queryHits],row.names = NULL)
                  # colnames(qh)=paste0(colnames(qh),"_mm10")
                  # sh=as.data.frame(gr2[xo$subjectHits],row.names = NULL)
        Crosslink_exonsclip_GR=gr[xo$queryHits]
        
      
        exons_CrossLink_GR_DF=as.data.frame(exons_CrossLink_GR)
        colnames(exons_CrossLink_GR_DF)=paste(colnames(exons_CrossLink_GR_DF),'_exons')
        
    Crosslink_exonsclip_GR_DF=cbind(as.data.frame(Crosslink_exonsclip_GR),exons_CrossLink_GR_DF)
    Crosslink_exonsclip_GR_DF$exons='exons'
    
    isoout$exons=NA
    isoout[which(isoout$ID%in%Crosslink_exonsclip_GR_DF$ID),'exons']="exons"


### Get UTR that overlap with CLIP peaks
        gr = CrossLink_GR
        gr2 = TranscUTRclip_GR
        xo=as.data.frame(GenomicRanges::findOverlaps(gr,gr2,type = "any",ignore.strand=F))
          
        UTR_CrossLink_GR=gr2[xo$subjectHits]
                  # qh=as.data.frame(gr[xo$queryHits],row.names = NULL)
                  # colnames(qh)=paste0(colnames(qh),"_mm10")
                  # sh=as.data.frame(gr2[xo$subjectHits],row.names = NULL)
        Crosslink_UTR_GR=gr[xo$queryHits]
        
      
        UTR_CrossLink_GR_DF=as.data.frame(UTR_CrossLink_GR)
        colnames(UTR_CrossLink_GR_DF)=paste0(colnames(UTR_CrossLink_GR_DF),'_UTR')
        
    Crosslink_UTR_GR_DF=cbind(as.data.frame(Crosslink_UTR_GR),UTR_CrossLink_GR_DF)
    Crosslink_UTR_GR_DF$UTR='UTR'
    Crosslink_UTR_GR_DF$`5UTR`=NA
    Crosslink_UTR_GR_DF[Crosslink_UTR_GR_DF$exon_number_UTR==1,'5UTR']="5UTR"
    Crosslink_UTR_GR_DF_5UTR=Crosslink_UTR_GR_DF[is.na(Crosslink_UTR_GR_DF$`5UTR`)==F,]

    
### Get 3UTR that overlap with CLIP peaks
        gr = CrossLink_GR
        gr2 = Transc3UTRclip_GR
        xo=as.data.frame(GenomicRanges::findOverlaps(gr,gr2,type = "any",ignore.strand=F))
          
        UTR3CrossLink_GR=gr2[xo$subjectHits]
                  # qh=as.data.frame(gr[xo$queryHits],row.names = NULL)
                  # colnames(qh)=paste0(colnames(qh),"_mm10")
                  # sh=as.data.frame(gr2[xo$subjectHits],row.names = NULL)
        Crosslink_3UTR_GR=gr[xo$queryHits]
        
      
        UTR3CrossLink_GR_DF=as.data.frame(UTR3CrossLink_GR)
        colnames(UTR3CrossLink_GR_DF)=paste0(colnames(UTR3CrossLink_GR_DF),'_3UTR')
        
    Crosslink_UTR_GR_DF_3UTR=cbind(as.data.frame(Crosslink_3UTR_GR),UTR3CrossLink_GR_DF)
    Crosslink_UTR_GR_DF_3UTR$`3UTR`='3UTR'
    
    
### Get introns that overlap with CLIP peaks
        gr = CrossLink_GR
        gr2 = intronsclip_GR
        xo=as.data.frame(GenomicRanges::findOverlaps(gr,gr2,type = "any",ignore.strand=F))
          
        introns_CrossLink_GR=gr2[xo$subjectHits]
                  # qh=as.data.frame(gr[xo$queryHits],row.names = NULL)
                  # colnames(qh)=paste0(colnames(qh),"_mm10")
                  # sh=as.data.frame(gr2[xo$subjectHits],row.names = NULL)
        Crosslink_intronsclip_GR=gr[xo$queryHits]
        
      
        introns_CrossLink_GR_DF=as.data.frame(introns_CrossLink_GR)
        colnames(introns_CrossLink_GR_DF)=paste0(colnames(introns_CrossLink_GR_DF),'_introns')
        
    Crosslink_intronsclip_GR_DF=cbind(as.data.frame(Crosslink_intronsclip_GR),introns_CrossLink_GR_DF)
    if (nrow(Crosslink_intronsclip_GR_DF)>0) {
          Crosslink_intronsclip_GR_DF$introns='introns'
    }
    
    
################################################
#### create Genomic files
################################################    
############################################################################################

    isoout$UTR=NA
    isoout[which(isoout$ID%in%Crosslink_UTR_GR_DF$ID),'UTR']="UTR"
    
    isoout$`3UTR`=NA
    isoout[which(isoout$ID%in%Crosslink_UTR_GR_DF_3UTR$ID),'3UTR']="3UTR"
    
    isoout$`5UTR`=NA
    isoout[which(isoout$ID%in%Crosslink_UTR_GR_DF_5UTR$ID),'5UTR']="5UTR"
    
    isoout$`introns`=NA
    isoout[which(isoout$ID%in%Crosslink_intronsclip_GR_DF$ID),'introns']="introns"
    
    isoout[is.na(isoout$exons)==F,'TranscRegion']='exons'
    isoout[is.na(isoout$`introns`)==F,'TranscRegion']='introns'
    isoout[is.na(isoout$`5UTR`)==F,'TranscRegion']='5UTR'
    isoout[is.na(isoout$`3UTR`)==F,'TranscRegion']='3UTR'
    isoout$TranscRegion=gsub("exons","exons_CDS",isoout$TranscRegion)
      
  

```  



```{r , include=FALSE}


outchunck_rowaverage=matrix(nrow=nrepeat,ncol = (window-9) )
  outchunck_rowaverage_max=matrix(nrow=nrepeat,ncol = (window-9) )
  # outchunck_rowaverage=matrix(nrow=nrepeat,ncol = (window) )
  outchunck_frac=matrix(nrow=nrepeat,ncol = 2 )
  outchunck_frac_max=matrix(nrow=nrepeat,ncol = 2 )
  outchunck_rowaverage_dn=matrix(nrow=nrepeat,ncol = (window-9) )
  outchunck_rowaverage_maxdn=matrix(nrow=nrepeat,ncol = (window-9) )
  # outchunck_rowaverage_dn=matrix(nrow=nrepeat,ncol = (window) )
  outchunck_frac_dn=matrix(nrow=nrepeat,ncol = 2 )
  outchunck_frac_maxdn=matrix(nrow=nrepeat,ncol = 2 )
  
  
  n_UP=matrix(nrow=Nlocations,ncol=nrepeat)
  ID_UP=matrix(nrow=Nlocations,ncol=nrepeat)
  seq_UP=matrix(nrow=Nlocations,ncol=nrepeat)
  SingleExon_UP=matrix(nrow=Nlocations,ncol=nrepeat)
  Exon_UP=matrix(nrow=Nlocations,ncol=nrepeat)
  transcript_UP=matrix(nrow=Nlocations,ncol=nrepeat)
  LastExon_UP=matrix(nrow=Nlocations,ncol=nrepeat)
  Gscore_UP=matrix(nrow=Nlocations,ncol=nrepeat)
  GscoreDist_UP=matrix(nrow=Nlocations,ncol=nrepeat)
  FinderSeq_UP=matrix(nrow=Nlocations,ncol=nrepeat)
  qgrsSeq_UP=matrix(nrow=Nlocations,ncol=nrepeat)
  maxBPPmean_UP=matrix(nrow=Nlocations,ncol=nrepeat)
  maxBPPmax_UP=matrix(nrow=Nlocations,ncol=nrepeat)
  MFE_UP=matrix(nrow=Nlocations,ncol=nrepeat)
  
  n_DN=matrix(nrow=Nlocations,ncol=nrepeat)
  ID_DN=matrix(nrow=Nlocations,ncol=nrepeat)
  seq_DN=matrix(nrow=Nlocations,ncol=nrepeat)
  SingleExon_DN=matrix(nrow=Nlocations,ncol=nrepeat)
  Exon_DN=matrix(nrow=Nlocations,ncol=nrepeat)
  transcript_DN=matrix(nrow=Nlocations,ncol=nrepeat)
  LastExon_DN=matrix(nrow=Nlocations,ncol=nrepeat)
  Gscore_DN=matrix(nrow=Nlocations,ncol=nrepeat)
  GscoreDist_DN=matrix(nrow=Nlocations,ncol=nrepeat)
  FinderSeq_DN=matrix(nrow=Nlocations,ncol=nrepeat)
  qgrsSeq_DN=matrix(nrow=Nlocations,ncol=nrepeat)
  maxBPPmean_DN=matrix(nrow=Nlocations,ncol=nrepeat)
  maxBPPmax_DN=matrix(nrow=Nlocations,ncol=nrepeat)
  MFE_DN=matrix(nrow=Nlocations,ncol=nrepeat)

########################################################################################################################################################################################################################
  
  
  checkout2=as.data.frame(matrix(nrow = Nlocations,ncol=17));colnames(checkout2)=c('n','ID','CLIP_ID','crosslink','SingleExon','numberExCrossed','crosslinkexon','s1exon','e1exon','lastexon','TranscChoise','numberTranscript','transcript_id','exon#','sequence','strand','DeltaG')

  checkout2_dn=as.data.frame(matrix(nrow = Nlocations,ncol=17));colnames(checkout2_dn)=c('n','ID','CLIP_ID','crosslink','SingleExon','numberExCrossed','crosslinkexon','s1dnexon','e1dnexon','lastexon','TranscChoise','numberTranscript','transcript_id','exon#','sequence','strand','DeltaG')
      
        faout=matrix(nrow = (Nlocations*2),ncol=1)
        faout_dn=matrix(nrow = (Nlocations*2),ncol=1)

  checkout2_summary=matrix(nrow=nrepeat,ncol=11)
  colnames(checkout2_summary)=c('SingleExon','numberExCrossed','lastexon','lastexonCrossexExon','firstexon','firstexonCrossexExon','total Locations','Gquad','Avg_GquadScore','MeanBPP','MaxBPP')
  checkout2_dn_summary=matrix(nrow=nrepeat,ncol=11)
  colnames(checkout2_dn_summary)=c('SingleExon','numberExCrossed','lastexon','lastexonCrossexExon','firstexon','firstexonCrossexExon',"total Locations",'Gquad','Avg_GquadScore','MeanBPP','MaxBPP')
  
  isoout=separate(isoout,col = "ID",into = c('chr','start'),sep = ":",remove = F)
  isoout=separate(isoout,col = "start",into = c('start','end'),sep = "-",remove = F)
  isoout$start=as.numeric(isoout$start)
  isoout$end=as.numeric(isoout$end)
  

  
  isoout_transcAll=isoout ##column USEtranscript
  Peaksdata4_chunkInExall=Peaksdata4_chunkEx
 #### Intronic peaks only    
  if (CLIPtype=='Introns') {
   Peaksdata4_chunkInall=Peaksdata4_chunkEx[Peaksdata4_chunkEx$`Both Strand: RNA_type_Classification`%in%c('protein_coding: exon','Antisense Feature')==F,]
   Peaksdata4_chunkEx=Peaksdata4_chunkEx[Peaksdata4_chunkEx$`Both Strand: RNA_type_Classification`%in%c('protein_coding: Intron'),]
  }
  
    if (CLIPtype=='Introns_LINESINETLR') {
### will include peaks that do not OL with intron on Same strand but does on opposite strand
   Peaksdata4_chunkInall1=Peaksdata4_chunkEx[Peaksdata4_chunkEx$`Both Strand: RNA_type_Classification`%in%c('protein_coding: exon')==F,] 
### Includes any peak that only overlaps with introns on same strand
   Peaksdata4_chunkInall= Peaksdata4_chunkEx[grepl('intron',Peaksdata4_chunkEx$`Same Strand: Intron Exon`),]
### takes away CLIP peaks classified as opposite strand features and still could be interesting
   Peaksdata4_chunkInall3= Peaksdata4_chunkEx[Peaksdata4_chunkEx$`Both Strand: RNA_type_Classification`%in%c('protein_coding: exon','Antisense Feature')==F,]

     Peaksdata4_chunkEx=Peaksdata4_chunkInall
  }
  
 #### exonic peaks only
    if (CLIPtype=='Exons'&TrasncLoc=='all') {
   Peaksdata4_chunkExall=Peaksdata4_chunkEx[Peaksdata4_chunkEx$`Both Strand: RNA_type_Classification`%in%c('protein_coding: exon'),]
   Peaksdata4_chunkEx=Peaksdata4_chunkEx[Peaksdata4_chunkEx$`Both Strand: RNA_type_Classification`%in%c('protein_coding: exon'),]
    }
   if (CLIPtype=='Exons'&TrasncLoc=='3UTR') {
   Peaksdata4_chunkEx=Peaksdata4_chunkEx[(Peaksdata4_chunkEx$ID%in%isoout_transcAll[isoout_transcAll$TranscRegion%in%'3UTR','ID']),]
    }

Peaksdata4_chunkEx=separate(Peaksdata4_chunkEx,ID,into = c('chr','start'),sep=':',remove = F)
Peaksdata4_chunkEx=separate(Peaksdata4_chunkEx,start,into = c('start','end'),sep='-')
Peaksdata4_chunkEx$start=as.numeric(Peaksdata4_chunkEx$start)
Peaksdata4_chunkEx$end=as.numeric(Peaksdata4_chunkEx$end)
      
############################################################################################################
## Checked Intron,exon peaks to find more Exon peak these are listed below
ExonAdd=c('chr1:71620722-71620766', 'chr10:115341383-115341422', 'chr10:122921578-122921615', 'chr10:53610110-53610170', 'chr11:100504854-100504876', 'chr11:115916957-115917009', 'chr15:80251044-80251086', 'chr16:17626516-17626585', 'chr17:45592176-45592225', 'chr17:83463473-83463526', 'chr2:129112779-129112816', 'chr3:108078674-108078713', 'chr8:124048273-124048334', 'chr9:106285553-106285575')

# ExonAdd=Peaksdata4[Peaksdata4$ID%in%ExonAdd,]
# ExonAdd=separate(ExonAdd,ID,into = c('chr','start'),sep=':',remove = F)
# ExonAdd=separate(ExonAdd,start,into = c('start','end'),sep='-')
# ExonAdd$start=as.numeric(ExonAdd$start)
# ExonAdd$end=as.numeric(ExonAdd$end)
# 
# Peaksdata4_chunkEx=as.data.frame(rbind(Peaksdata4_chunkEx,ExonAdd))


## These couldbe exons depending on transcript

# ExonMaybe=c('chr12:69372320-69372375', 'chr2:71820222-71820258', 'chr4:138239647-138239709', 'chr5:110430755-110430813', 'chr7:133014567-133014602', 'chrX:38419895-38419931', 'chrX:48277218-48277276')
############################################################################################################


isoout=isoout[isoout$ID%in%Peaksdata4_chunkEx$ID,]


if (CLIPtype=='Introns_LINESINETLR') {CLIPtype='Introns'}

  
```       

## Overlap with LINE/SINE

Select Any CLIP peaks that overlap with intronic regions. If Peak also Overlaps with a Repeat Element it is removed, except for peaks that overlap with LINE, SINE or LTR.  


  

```{r ,echo=F,warning=F,eval=T,include=F}
 
################################################################################################################################################################       
###### AllIntrons 
       
checkout2_LS=Peaksdata4_chunkInExall[Peaksdata4_chunkInExall$`Both Strand: RNA_type_Classification`%in%c('protein_coding: exon')==F,]
# unique(checkout2_LS$`Both Strand: RNA_type_Classification`)

checkout2_LS$start=as.numeric(checkout2_LS$start)
checkout2_LS$end=as.numeric(checkout2_LS$end)
checkout2_LS[checkout2_LS$strand%in%"+",'crosslink']=checkout2_LS[checkout2_LS$strand%in%"+",'start']
checkout2_LS[checkout2_LS$strand%in%"-",'crosslink']=checkout2_LS[checkout2_LS$strand%in%"-",'end']
checkout2_LS=merge(checkout2_LS,isoout_transcAll[,c('ID','USEtranscript')],by='ID',all.x=T)
colnames(checkout2_LS)[colnames(checkout2_LS)%in%'USEtranscript']='transcript_id'


CrossLink_GR=
  GRanges(seqnames=checkout2_LS$chr, 
          ranges=IRanges(start=checkout2_LS$crosslink,
                         end=checkout2_LS$crosslink),
          strand=checkout2_LS$strand,
          ID=checkout2_LS$ID,
          transcript_id=checkout2_LS$transcript_id)  

###################################################################
            
mm10_transall=mm10all[mm10all$feature%in%'transcript',]
    
mm10_transUSE=mm10_transall[mm10_transall$transcript_id%in%canonical$transcript,]
mm10_transUSE=mm10_transUSE[mm10_transUSE$transcript_type%in%'protein_coding',]


    ###########################
    ### Make sure that Transcripts from CLIP peaks are selected. 
    ###########################
Uptrancripts=checkout2_LS$transcript_id
Uptrancripts_mm10_trans=mm10_transall[mm10_transall$transcript_id%in%Uptrancripts,]
mm10_transUSE=rbind(mm10_transUSE[mm10_transUSE$gene_id%in%unique(Uptrancripts_mm10_trans$gene_id)==F,],
            Uptrancripts_mm10_trans)
 

trans_GR=GRanges(seqnames=mm10_transUSE$seqname, 
                 ranges=IRanges(start=mm10_transUSE$start,end=mm10_transUSE$end),
                 strand=mm10_transUSE$strand,
                 transcript_id=mm10_transUSE$transcript_id,
                 gene_id=mm10_transUSE$gene_id
                 )

mm10_transUSE=mm10_transUSE[,c('transcript_id','seqname','start','end','strand')]



# ########################################################################################        
#  #### adjust CLIP coordinates
# #############
# 
checkout2_LS=merge(checkout2_LS,mm10_transUSE,by='transcript_id',all.x=T,suffixes=c('_CLIP','_mm10'))
# 
# checkout2_LS$CL_adj=NA
# 
# checkout2_LS[checkout2_LS$strand_CLIP%in%"+",'CL_adj']=
#   checkout2_LS[checkout2_LS$strand_CLIP%in%"+",'crosslink']-
#   checkout2_LS[checkout2_LS$strand_CLIP%in%"+",'start']
# 
# checkout2_LS[checkout2_LS$strand_CLIP%in%"-",'CL_adj']=
#   checkout2_LS[checkout2_LS$strand_CLIP%in%"-",'end']-
#   checkout2_LS[checkout2_LS$strand_CLIP%in%"-",'crosslink']

# checkout2_LS$TranscWidth=checkout2_LS$end-checkout2_LS$start
# checkout2_LS[checkout2_LS$CL_adj+1>=checkout2_LS$TranscWidth,]

# CrossLinkADJ=checkout2_LS[,c('transcript_id','chr','crosslink',"strand_CLIP",'ID','gene_id','gene_name','start','end','CL_adj')]


checkout2_LS[checkout2_LS$strand_CLIP%in%"+",'CL_up']=checkout2_LS[checkout2_LS$strand_CLIP%in%"+",'crosslink']-(window-1)
checkout2_LS[checkout2_LS$strand_CLIP%in%"-",'CL_up']=checkout2_LS[checkout2_LS$strand_CLIP%in%"-",'crosslink']+(window-1)


checkout2_LS[checkout2_LS$strand_CLIP%in%"+",'CL_dn']=checkout2_LS[checkout2_LS$strand_CLIP%in%"+",'crosslink']+(window-1)
checkout2_LS[checkout2_LS$strand_CLIP%in%"-",'CL_dn']=checkout2_LS[checkout2_LS$strand_CLIP%in%"-",'crosslink']-(window-1)


CL_GR = GRanges(seqnames=checkout2_LS$chr, 
                  ranges=IRanges(start=checkout2_LS$crosslink,
                  end=checkout2_LS$crosslink),
                  strand=checkout2_LS$strand_CLIP,
                  ID=checkout2_LS$ID,
                  transcript_id=checkout2_LS$transcript_id
                  )
CLup_GR = GRanges(seqnames=checkout2_LS$chr, 
                  ranges=IRanges(start=rowMins(as.matrix(checkout2_LS[,c('crosslink','CL_up')])),
                  end=rowMaxs(as.matrix(checkout2_LS[,c('crosslink','CL_up')]))),
                  strand=checkout2_LS$strand_CLIP,
                  ID=checkout2_LS$ID,
                  transcript_id=checkout2_LS$transcript_id
                  )
CLdn_GR = GRanges(seqnames=checkout2_LS$chr, 
                  ranges=IRanges(start=rowMins(as.matrix(checkout2_LS[,c('crosslink','CL_dn')])),
                  end=rowMaxs(as.matrix(checkout2_LS[,c('crosslink','CL_dn')]))),
                  strand=checkout2_LS$strand_CLIP,
                  ID=checkout2_LS$ID,
                  transcript_id=checkout2_LS$transcript_id
                  )
# View(CrossLinkADJ[CrossLinkADJ$transcript_id%in%CrossLinkADJ[duplicated(CrossLinkADJ$transcript_id),'transcript_id'],])

###################################################################

rmsk_GRCm38=fread(paste0("/Users/homanpj/OneDrive - National Institutes of Health/Resources/ref/mm10/repeatmasker/rmsk_GRCm38.txt"), header=T, sep="\t",stringsAsFactors = F,data.table=F)
rmsk_GRCm38_LISI=rmsk_GRCm38[rmsk_GRCm38$repClass%in%c('LINE','SINE'),]
rmsk_GRCm38_LISI=rmsk_GRCm38_LISI[rmsk_GRCm38_LISI$genoName%in%c('chrUn_GL456239', 'chrUn_GL456359', 'chrUn_GL456360', 'chrUn_GL456366', 'chrUn_GL456367', 'chrUn_GL456368', 'chrUn_GL456370', 'chrUn_GL456372', 'chrUn_GL456378', 'chrUn_GL456379', 'chrUn_GL456381', 'chrUn_GL456382', 'chrUn_GL456385', 'chrUn_GL456387', 'chrUn_GL456389', 'chrUn_GL456393', 'chrUn_GL456394', 'chrUn_JH584304', 'chr1_GL456210_random', 'chr1_GL456211_random', 'chr1_GL456212_random', 'chr1_GL456213_random', 'chr1_GL456221_random', 'chr4_GL456216_random', 'chr4_GL456350_random', 'chr4_JH584292_random', 'chr4_JH584293_random', 'chr4_JH584294_random', 'chr5_GL456354_random', 'chr5_JH584296_random', 'chr5_JH584297_random', 'chr5_JH584298_random', 'chr5_JH584299_random', 'chr7_GL456219_random', 'chrX_GL456233_random', 'chrY_JH584300_random', 'chrY_JH584301_random', 'chrY_JH584302_random', 'chrY_JH584303_random')==F,]
rmsk_GRCm38_LISI$ID=paste0(rmsk_GRCm38_LISI$genoName,":",rmsk_GRCm38_LISI$genoStart,'-',rmsk_GRCm38_LISI$genoEnd)


LISI_GR = GRanges(seqnames=rmsk_GRCm38_LISI$genoName, 
                  ranges=IRanges(start=rmsk_GRCm38_LISI$genoStart,
                  end=rmsk_GRCm38_LISI$genoEnd),
                  strand=rmsk_GRCm38_LISI$strand,
                  repClass=rmsk_GRCm38_LISI$repClass,
                  repName=rmsk_GRCm38_LISI$repName,
                  ID=rmsk_GRCm38_LISI$ID
                  )


### Get LINE/SINES that overlap with Transcripts
        gr = trans_GR
        gr2 = LISI_GR
        xo=as.data.frame(GenomicRanges::findOverlaps(gr,gr2,type = "any",ignore.strand=T))
          
        LISI_Trans_GR=as.data.frame(gr2[xo$subjectHits])
                  # qh=as.data.frame(gr[xo$queryHits],row.names = NULL)
                  # colnames(qh)=paste0(colnames(qh),"_mm10")
                  # sh=as.data.frame(gr2[xo$subjectHits],row.names = NULL)
        Trans_LISI_GR=as.data.frame(gr[xo$queryHits])
          colnames(LISI_Trans_GR)=paste0(colnames(LISI_Trans_GR),"_LINESINE")
          colnames(Trans_LISI_GR)=paste0(colnames(Trans_LISI_GR),"_Trans")
        
        LISI_Trans=cbind(as.data.frame(LISI_Trans_GR),as.data.frame(Trans_LISI_GR))
      

       
 
 
########################
### Upstream
########################    
 
        gr = CLup_GR
        gr2 = LISI_GR
        xo=as.data.frame(GenomicRanges::findOverlaps(gr,gr2,type = "any",ignore.strand=T))

        LISI_CLup_GR=as.data.frame(gr2[xo$subjectHits])
                  # qh=as.data.frame(gr[xo$queryHits],row.names = NULL)
                  # colnames(qh)=paste0(colnames(qh),"_mm10")
                  # sh=as.data.frame(gr2[xo$subjectHits],row.names = NULL)
        CLup_LISI_GR=as.data.frame(gr[xo$queryHits])

 colnames(LISI_CLup_GR)=paste0(colnames(LISI_CLup_GR),"_LISI")
 CLup_LISI_GR=as.data.frame(cbind(CLup_LISI_GR,LISI_CLup_GR))
 
 dup=unique(CLup_LISI_GR[duplicated(CLup_LISI_GR$ID),"ID"])
 CLup_LISI_GR_dup=CLup_LISI_GR[CLup_LISI_GR$ID%in%dup==T,]
  CLup_LISI_GR_uniqe=CLup_LISI_GR[CLup_LISI_GR$ID%in%dup==F,]

 
 CLup_LISI_GR_dupout=as.data.frame(matrix(nrow = length(dup),ncol=ncol(CLup_LISI_GR_dup)))
 colnames(CLup_LISI_GR_dupout)=colnames(CLup_LISI_GR_dup)
 for (d in 1:length(dup)) {
  dm= CLup_LISI_GR_dup[CLup_LISI_GR_dup$ID%in%dup[d],]
   
  dmm=(dm$strand==dm$strand_LISI)
   if (length(unique(dmm))==1) {dm=dm[1,]}
   if (length(unique(dmm))>1) {dm=dm[dmm%in%T,]
   if (nrow(dm)>1) {dm=dm[1,]}}
CLup_LISI_GR_dupout[d,]=as.matrix(dm)
 }
 
 CLup_LISI_GR=rbind(CLup_LISI_GR_dupout,CLup_LISI_GR_uniqe)
 
 # length(unique(CLup_LISI_GR$ID) )
 #  nrow(CLup_LISI_GR) 
 #  length(unique(CLup_LISI_GR$ID) )

 isoout$LISI_OLup=NA
 isoout[isoout$ID%in%CLup_LISI_GR$ID,'LISI_OLup']=1

  isoout=merge(isoout,CLup_LISI_GR[,c("ID","strand_LISI")],by='ID',all.x=T)
  colnames(isoout)[colnames(isoout)%in%'strand_LISI']='strand_LISIup'
  
  # op=which(checkout2$strand!=checkout2$strand_LISI.x)
  # checkout2_op=checkout2[op,]
  # checkout2_same=checkout2[-op,]
  # 
  # for (o in 1:nrow(checkout2_op)) {
  #   checkout2_op[o,'sequence']= as.character(reverseComplement(RNAString(checkout2_op[o,'sequence',drop=T])))}
  # checkout2=rbind(checkout2_op,checkout2_same)
  # 

########################
### Downstream
########################    
 
          gr = CLdn_GR
        gr2 = LISI_GR
        xo=as.data.frame(GenomicRanges::findOverlaps(gr,gr2,type = "any",ignore.strand=T))

        LISI_CLdn_GR=as.data.frame(gr2[xo$subjectHits])
                  # qh=as.data.frame(gr[xo$queryHits],row.names = NULL)
                  # colnames(qh)=paste0(colnames(qh),"_mm10")
                  # sh=as.data.frame(gr2[xo$subjectHits],row.names = NULL)
        CLdn_LISI_GR=as.data.frame(gr[xo$queryHits])
        
 colnames(LISI_CLdn_GR)=paste0(colnames(LISI_CLdn_GR),"_LISI")
 CLdn_LISI_GR=as.data.frame(cbind(CLdn_LISI_GR,LISI_CLdn_GR))
 
 ddn=unique(CLdn_LISI_GR[duplicated(CLdn_LISI_GR$ID),"ID"])
 CLdn_LISI_GR_ddn=CLdn_LISI_GR[CLdn_LISI_GR$ID%in%ddn==T,]
  CLdn_LISI_GR_uniqe=CLdn_LISI_GR[CLdn_LISI_GR$ID%in%ddn==F,]

 
 CLdn_LISI_GR_ddnout=as.data.frame(matrix(nrow = length(ddn),ncol=ncol(CLdn_LISI_GR_ddn)))
 colnames(CLdn_LISI_GR_ddnout)=colnames(CLdn_LISI_GR_ddn)
 for (d in 1:length(ddn)) {
  dm= CLdn_LISI_GR_ddn[CLdn_LISI_GR_ddn$ID%in%ddn[d],]
   
  dmm=(dm$strand==dm$strand_LISI)
   if (length(unique(dmm))==1) {dm=dm[1,]}
   if (length(unique(dmm))>1) {dm=dm[dmm%in%T,]
   if (nrow(dm)>1) {dm=dm[1,]}}
CLdn_LISI_GR_ddnout[d,]=as.matrix(dm)
 }
 
 CLdn_LISI_GR=rbind(CLdn_LISI_GR_ddnout,CLdn_LISI_GR_uniqe)        


 isoout$LISI_OLdn=NA
 isoout[isoout$ID%in%CLdn_LISI_GR$ID,'LISI_OLdn']=1

  isoout=merge(isoout,CLdn_LISI_GR[,c("ID","strand_LISI")],by='ID',all.x=T)
    colnames(isoout)[colnames(isoout)%in%'strand_LISI']='strand_LISIdn'

  # opdn=which(checkout2_dn$strand!=checkout2_dn$strand_LISI.x)
  # checkout2_dn_op=checkout2_dn[opdn,]
  # checkout2_dn_same=checkout2_dn[-opdn,]
  # 
  # for (o in 1:nrow(checkout2_dn_op)) {
  #   checkout2_dn_op[o,'sequence']= as.character(reverseComplement(RNAString(checkout2_dn_op[o,'sequence',drop=T])))}
  # checkout2_dn=rbind(checkout2_dn_op,checkout2_dn_same)
  
 
```


## Get Sequences  

sequences are taken `window` nts Upstream and Downstream of the 5' end of the CLIP peak.   

For RunMode='Genomic' sequences are taken from the Genomic sequence irregardless whether it contains Intronic/Exonic regions.  
For RunMode='Transcriptomic' sequences are taken from the spliced transcript sequence. If 100nt region crosses into Intronic region. The Intron sequence is not considered and taken from the next Exon.  
  
  
```{r , echo=F, include=FALSE}
  
  
  for (x in 1:nrow(isoout)) {
            ##### select location in exon
         isx=isoout[x,]

         
         isx=separate(isx,ID,into = c('chr','Clipstart'),sep=":",remove = F)
         isx=separate(isx,Clipstart,into = c('Clipstart','Clipend'),sep="-",remove = F)
         
         checkout2[x,"CLIP_ID"]=isx$ID
         checkout2_dn[x,"CLIP_ID"]=isx$ID
         
         anno.GR=GRanges(seqnames=isx$chr, ranges=IRanges(start=isx$start,end=isx$end),
                 strand =isx$strand)
         
         texn=mm10_USE[mm10_USE$transcript_id%in%isx$USEtranscript,]



#########################
         
loc2=mm10_USE[mm10_USE$gene_id%in%isx$gene_id,]
if (nrow(loc2)==0) {checkout2[x,'SingleExon']='transcript not in mm10'
  next}


s.GR = GRanges(seqnames=texn$seqname, ranges=IRanges(start=texn$start,end=texn$end),
                 strand=texn$strand)
  q =s.GR
  s=anno.GR
  xo=as.data.frame(GenomicRanges::findOverlaps(q,s,type = "any",ignore.strand=F))
  texn=texn[xo$queryHit,]
  checkout2[x,'TranscChoise']="Highest Expression"
  checkout2[x,'numberTranscript']=length(unique(texn$transcript_id))
  if (nrow(texn)>0) {
    checkout2[x,'transcript_id']=unique(texn$transcript_id)
    checkout2[x,'exon#']=(texn$exon_number)
    checkout2[x,'strand']=texn$strand}

  remove(xo,s.GR)

#### overlap with Canonical
if (nrow(texn)==0) {
  texn=mm10_USE[mm10_USE$gene_id%in%isx$IScanonical,]
    if (length(unique(texn$transcript_id))>1) {xxxx1  }
  s.GR = GRanges(seqnames=texn$seqname, ranges=IRanges(start=texn$start,end=texn$end),
                 strand=texn$strand)
  q =s.GR
  s=anno.GR
  xo=as.data.frame(GenomicRanges::findOverlaps(q,s,type = "any",ignore.strand=F))
  texn=texn[xo$queryHit,]
  checkout2[x,'TranscChoise']="switch to Cannonical"
  checkout2[x,'numberTranscript']=length(unique(texn$transcript_id))
  if (nrow(texn)>0) {
    checkout2[x,'transcript_id']=unique(texn$transcript_id)
    checkout2[x,'exon#']=(texn$exon_number)
    checkout2[x,'strand']=strand}

  remove(xo,s.GR)
}

#### overlap with any
if (nrow(texn)==0) {
  texn=mm10_USE[mm10_USE$gene_id%in%isx$gene_id,]
  s.GR = GRanges(seqnames=texn$seqname, ranges=IRanges(start=texn$start,end=texn$end),
                 strand=texn$strand)
  q =s.GR
  s=anno.GR
  xo=as.data.frame(GenomicRanges::findOverlaps(q,s,type = "any",ignore.strand=F))
  texn=texn[xo$queryHit,]
    checkout2[x,'TranscChoise']="switch to other"
    checkout2[x,'numberTranscript']=length(unique(texn$transcript_id))
    
   if (nrow(texn)>0) {
     isotexn=iso[iso$transcript_id%in% texn$transcript_id,]
     isotexn=isotexn[order(isotexn$WT,decreasing = T),]
     texn=texn[texn$transcript_id%in%isotexn[1,'transcript_id'],]}
    
    if (nrow(texn)>0) {
      checkout2[x,'transcript_id']=unique(texn$transcript_id)
      checkout2[x,'exon#']=(texn$exon_number)
    checkout2[x,'strand']=strand}
    remove(xo,s.GR)
}
#          
         
if (nrow(texn)==0) {
  checkout2[x,'SingleExon']='does not cross exon'
  checkout2[x,'TranscChoise']=NA
  checkout2[x,'numberTranscript']=NA
  checkout2[x,'transcript_id']=NA
  checkout2[x,'exon#']=NA
  checkout2[x,'sequence']=NA
  next}
        
        if (length(unique(texn$transcript_id))>1) {xxxxxxxxcxcxcx}
        if (nrow(texn)!=1) {xxxxxxxxcxcxcx}
  
  
        texn$Clipstart=isx$Clipstart
          texn$Clipend=isx$Clipend

  
        strt=texn$start
        end=texn$end
        strand=texn$strand
        chr=texn$seqname
        
        texn$CL=isx$CL
      
        if (strand=="+") {
          texn$s1=texn$CL-window#(selects base +1 of start site but ends at correct 5' site)
          texn$e1=texn$CL
          
          texn$s1dn=texn$CL-1
          texn$e1dn=texn$CL+window-1
        }
        if (strand=="-") {
          texn$s1=texn$CL-1#(selects base +1 of start site but ends at correct 5' site)
          texn$e1=texn$CL+window-1
          
          texn$s1dn=texn$CL-window
          texn$e1dn=texn$CL
        }
        
        s1=texn$s1
        e1=texn$e1
        
        s1dn=texn$s1dn
        e1dn=texn$e1dn
        
        checkout2[x,'transcript_id']=unique(texn$transcript_id)
        checkout2[x,'exon#']=(texn$exon_number)
        checkout2[x,'strand']=strand
        checkout2[x,'ID']=paste0(chr,":",texn$CL,"_",texn$strand)
        checkout2[x,'crosslink']=texn$CL
        
        checkout2_dn[x,'transcript_id']=unique(texn$transcript_id)
        checkout2_dn[x,'exon#']=(texn$exon_number)
        checkout2_dn[x,'strand']=strand
        checkout2_dn[x,'ID']=paste0(chr,":",texn$CL,"_",texn$strand)
        checkout2_dn[x,'crosslink']=texn$CL
        
                         # if (isx$ID=='chr18:35179784-35179821') {xxxxxxx}

        ### UPSTREAM
        #############################################################################################
       
         checkout2[x,'SingleExon']=(texn$start<=texn$s1 & texn$end>=texn$e1)|
          (strand=="+"&(texn$start>=texn$s1&texn$end>=texn$e1)&texn$exon_number==1)|
          (strand=="-"&(texn$start<=texn$s1&texn$end<=texn$e1)&texn$exon_number==1)
        
        # checkout2[x,'SingleExon']=(texn$start<=texn$s1 & texn$end>=texn$e1)|
        #   (strand=="+"&(texn$start>=texn$s1&texn$end>=texn$e1)&texn$exon_number==1)| 
        #   (strand=="-"&(texn$start<=texn$s1&texn$end<=texn$e1)&texn$exon_number==1)|
        #   (strand=="+"&(texn$end>=texn$e1))| 
        #   (strand=="-"&(texn$start<=texn$s1))
        
        
             if (RunMode=='Genomic') {checkout2[x,'SingleExon']=T}

        
        ## si and e1 out ofbounds but clip end is in 
        
        if (
          checkout2[x,'SingleExon']==F&
          ((strand=="+"&texn$e1<=texn$start&texn$Clipend>=texn$start)| 
          (strand=="-"&texn$s1>=texn$end&texn$Clipstart<=texn$end))
          ) {
          checkout2[x,'SingleExon']=T
          checkout2[x,'numberTranscript']=paste0(checkout2[x,'numberTranscript'],"Window out of bounds but CLIP in")}
            
              
        if ((strand=="+"&(texn$start>=texn$s1&texn$end>=texn$e1)&texn$exon_number==1)|
            (strand=="-"&(texn$start<=texn$s1&texn$end<=texn$e1)&texn$exon_number==1)) {checkout2[x,'numberTranscript']=paste0(checkout2[x,'numberTranscript'],"_out of bounds exon upstream only")}
        
        #### window region falls in exon
        
        checkout2[x,'crosslinkexon']=texn$start<=texn$CL&texn$end>=texn$CL #crosslink
        checkout2[x,'s1exon']=(strand=="+"&texn$start<=s1&texn$end>=s1)|(strand=="-"&texn$start<=e1&texn$end>=e1) # 5` end in exon
        checkout2[x,'e1exon']=(strand=="+"&texn$start<=e1&texn$end>=e1)|(strand=="-"&texn$start<=s1&texn$end>=s1) # 3` end in exon
        checkout2[x,'lastexon']=checkout2[x,'exon#']/max(mm10_USE[mm10_USE$transcript_id%in%checkout2[x,'transcript_id'],'exon_number'])
        
        checkout2_dn[x,'lastexon']=checkout2_dn[x,'exon#']/max(mm10_USE[mm10_USE$transcript_id%in%checkout2_dn[x,'transcript_id'],'exon_number'])
        checkout2_dn[x,'crosslinkexon']=texn$start<=texn$CL&texn$end>=texn$CL #crosslink
        
        ###########################################################
        #run if not in single exon and not exon 1 (if crosslink is in exon then upstream has to be upstream out of gene for exon1)
        if ((checkout2[x,'SingleExon']==F)&(checkout2[x,'exon#']>1)) {
          bedpos=matrix(ncol=6,nrow=2)
          
          locx=mm10_USE[mm10_USE$transcript_id%in%texn$transcript_id,]
          # locx=locx[order(locx$exon_number,decreasing = F),]
          if (strand=="+") {
            if (checkout2[x,'exon#',drop=F]==1) {next}
            
            # closest to peak
            bedpos[2,2]=locx[(locx$exon_number %in%texn$exon_number),'start']-1
            bedpos[2,3]=texn$CL
            
            # Upstream exon
            bedpos[1,2]=locx[(locx$exon_number %in%(texn$exon_number-1)),'end']-(window-abs(texn$CL-(locx[(locx$exon_number %in%texn$exon_number),'start'])+1))
            bedpos[1,3]=locx[(locx$exon_number %in%(texn$exon_number-1)),'end']
            
            checkout2[x,'numberExCrossed']=nrow(bedpos)
            
            # exon number of bedpos[1,]
            exnum=(texn$exon_number)-(nrow(bedpos)-1)
            while ((exnum>1)&(locx[(locx$exon_number %in%(exnum)),'start']>bedpos[1,2])) {#xnxnxnx

              bedpos[1,2]=locx[(locx$exon_number %in%(exnum)),'start']-1
              bedpos[1,3]=locx[(locx$exon_number %in%(exnum)),'end']
              
              bedpos2=matrix(nrow=1,ncol = 6)
              bedpos2[1,2]=locx[(locx$exon_number %in%((exnum-1))),'end']-((window-1)-sum(rowDiffs(bedpos[,c(2,3)]))+1)
              bedpos2[1,3]=locx[(locx$exon_number %in%((exnum-1))),'end']
              
              bedpos=rbind(bedpos2,bedpos)
              checkout2[x,'numberExCrossed']=nrow(bedpos)
              exnum=(texn$exon_number)-(nrow(bedpos)-1)-1
              if (exnum==0) {break}
            }
          }
          
          if (strand=="-") {
            if (checkout2[x,'exon#']==1) {next}
            
            # closest to peak
            bedpos[2,2]=texn$CL-1
            bedpos[2,3]=locx[(locx$exon_number %in%texn$exon_number),'end']
            
            # Upstream exon
            bedpos[1,2]=locx[(locx$exon_number %in%(texn$exon_number-1)),'start']-1
            bedpos[1,3]=locx[(locx$exon_number %in%(texn$exon_number-1)),'start']+((window-1)-abs(texn$CL-locx[(locx$exon_number %in%(texn$exon_number)),'end']))-1
            
            checkout2[x,'numberExCrossed']=nrow(bedpos)
            
            
                  exnum=(texn$exon_number)-(nrow(bedpos)-1)
            while ((exnum>1)&(locx[(locx$exon_number %in%(exnum)),'end']<bedpos[1,3])) {#xnegxneg
              
              bedpos[1,2]=locx[(locx$exon_number %in%(exnum)),'start']-1
              bedpos[1,3]=locx[(locx$exon_number %in%(exnum)),'end']
              bedpos2=matrix(nrow=1,ncol = 6)
              bedpos2[1,2]=locx[(locx$exon_number %in%((exnum-1))),'start']-1
              bedpos2[1,3]=locx[(locx$exon_number %in%((exnum-1))),'start']+((window)-sum(rowDiffs(bedpos[,c(2,3)]))-1)

              bedpos=rbind(bedpos2,bedpos)
              checkout2[x,'numberExCrossed']=nrow(bedpos)
              exnum=(texn$exon_number)-(nrow(bedpos)-1)-1
              if (exnum==0) {break}
            }
          }
          
          bedposx=as.data.frame(bedpos)
          bedposx[,2]=as.numeric(as.character(bedposx[,2]))
          bedposx[,3]=as.numeric(as.character(bedposx[,3]))
          sum(rowDiffs(as.matrix(bedposx[,c(2,3)])) )       
          (rowDiffs(as.matrix(bedposx[,c(2,3)])) )       
          
          bedpos[,1]=chr
          bedpos[,4]=".";bedpos[,5]="."
          bedpos[,6]=strand
          
          write.table(bedpos, file=paste0(scratch_UP,'/',texn$seqname,'_',texn$CL,'_chunk.bed'), quote=F, sep="\t", row.names=F, col.names=F)
          # system(paste0('bedtools getfasta -s -fi ',fasta,' -bed structure/chunkwindow/exon/',texn$ID,'_chunk.bed > structure/chunkwindow/exon/',texn$ID,'_chunk.fa'))
          fa=system(paste0('bedtools getfasta -s -fi ',fasta,' -bed ',scratch_UP,'/',texn$seqname,'_',texn$CL,'_chunk.bed'),intern = T)
          
          
          if (nchar(fa[2])==window) {fa2=fa;checkout2[x,'SingleExon']='Change-TRUE'}
          
          fa2=fa
          for (f in 2:nrow(bedpos)) {
          fa2[2]=paste(fa2[2],fa2[f*2],collapse = "")
          fa2[2]=gsub(" ","",fa2[2])
          
          if ((nchar(fa2[2])==window)&f<nrow(bedpos)){
            # checkout2=TRUE
            checkout2[x,'TranscChoise']='Change Multiple exons'
            fa2=fa2[1:2]
            checkout2[x,'numberExCrossed']=checkout2[x,'numberExCrossed']-1
            break}
          }
          
          
          fa2=fa2[1:2]
          if (nchar(fa[2])==window){
            checkout2[x,'SingleExon']=TRUE
            checkout2[x,'TranscChoise']='Change'
            checkout2[x,'numberExCrossed']=checkout2[x,'numberExCrossed']-1
            fa2=fa[1:2]}
          
          if (nchar(fa2[2])!=window){print(nchar(fa2[2]));shrtstrand}
          fa=fa2
          write.table(fa2, file=paste0(scratch_UP,'/',texn$seqname,'_',texn$CL,'_chunk.fa'), quote=F, sep="\t", row.names=F, col.names=F)
          checkout2[x,'sequence']=gsub('T','U',fa[2])
          checkout2[x,'n']=fa[1]
          # faout=c(faout,fa[1:2])
                    faout[(2*x-1):(2*x),1]=(as.matrix(fa[1:2]))
          remove(bedpos,fa2)
        }
        
        if ((checkout2[x,'SingleExon']==T)|(checkout2[x,'exon#']==1)) {
          bedpos=matrix(ncol=6,nrow=1)
          bedpos[1,1]=chr
          bedpos[1,2]=s1
          bedpos[1,3]=e1
          bedpos[1,4]=".";bedpos[,5]="."
          bedpos[1,6]=strand
          
          write.table(bedpos, file=paste0(scratch_UP,'/',texn$seqname,'_',texn$CL,'_chunk.bed'), quote=F, sep="\t", row.names=F, col.names=F)
          # system(paste0('bedtools getfasta -s -fi ',fasta,' -bed structure/chunkwindow/exon/',texn$ID,'_chunk.bed > structure/chunkwindow/exon/',texn$ID,'_chunk.fa'))
          fa=system(paste0('bedtools getfasta -s -fi ',fasta,' -bed ',scratch_UP,'/',texn$seqname,'_',texn$CL,'_chunk.bed'),intern = T)
          write.table(fa, file=paste0(scratch_UP,'/',texn$seqname,'_',texn$CL,'_chunk.fa'), quote=F, sep="\t", row.names=F, col.names=F)
          checkout2[x,'sequence']=gsub('T','U',fa[2])
          checkout2[x,'n']=fa[1]
          checkout2[x,'numberExCrossed']=0
          
          # faout=c(faout,fa[1:2])
               faout[(2*x-1):(2*x),1]=(as.matrix(fa[1:2]))
          remove(bedpos)
        }
      
        if(file.exists(paste0(scratch_UP,'/',texn$seqname,'_',texn$CL,'_chunk.fa'))){file.remove(paste0(scratch_UP,'/',texn$seqname,'_',texn$CL,'_chunk.fa'))}
        if(file.exists(paste0(scratch_UP,'/',texn$seqname,'_',texn$CL,'_chunk.bed'))){file.remove(paste0(scratch_UP,'/',texn$seqname,'_',texn$CL,'_chunk.bed'))}
  
        
        
      ### DOWNSTREA
      #################################################################################################################################################################################
      #################################################################################################################################################################################
        
      checkout2_dn[x,'SingleExon']=(texn$start<=texn$s1dn & texn$end>=texn$e1dn)|
        (strand=="+"&(texn$start<=texn$s1dn&texn$end<=texn$e1dn)&checkout2_dn[x,'lastexon']==1)|
        (strand=="-"&(texn$start>=texn$s1dn&texn$end>=texn$e1dn)&checkout2_dn[x,'lastexon']==1)
      
        ### CLIP peak in window so dont need extra qualifiers
     # checkout2_dn[x,'SingleExon']=(texn$start<=texn$s1dn & texn$end>=texn$e1dn)|
     #    (strand=="+"&(texn$start<=texn$s1dn&texn$end<=texn$e1dn)&checkout2_dn[x,'lastexon']==1)|
     #    (strand=="-"&(texn$start>=texn$s1dn&texn$end>=texn$e1dn)&checkout2_dn[x,'lastexon']==1)|
     #    (strand=="+"&texn$end>=texn$e1dn)|
     #    (strand=="-"&texn$start<=texn$s1dn)
        
        
         if (RunMode=='Genomic') {checkout2_dn[x,'SingleExon']=T}

        
            if ( ###start is out of bounds but window end is in
              checkout2_dn[x,'SingleExon']==F&
              ((strand=="+"&texn$start>=texn$s1dn&texn$end>=texn$e1dn)|
              (strand=="-"&texn$end<=texn$e1dn&texn$start<=texn$s1dn))
              ) {
          checkout2_dn[x,'SingleExon']=T
          checkout2_dn[x,'numberTranscript']=paste0(checkout2_dn[x,'numberTranscript'],"5` end is out of bounds but single exon")}
          #   
      
      #### window region faDNlls in exon
      checkout2_dn[x,'s1dnexon']=(strand=="+"&texn$start<=s1dn&texn$end>=s1dn)|(strand=="-"&texn$start<=e1dn&texn$end>=e1dn) # 5` end in exon
      checkout2_dn[x,'e1dnexon']=(strand=="+"&texn$start<=e1dn&texn$end>=e1dn)|(strand=="-"&texn$start<=s1dn&texn$end>=s1dn) # 3` end in exon

      ###########################################################
      #run if not in single exon and not exon 1 (if crosslink is in exon then upstream has to be upstream out of gene for exon1)
      if ((checkout2_dn[x,'SingleExon']==F)&(checkout2_dn[x,'lastexon']<1)) {
        bedposDN=matrix(ncol=6,nrow=2)
        
        locx=mm10_USE[mm10_USE$transcript_id%in%texn$transcript_id,]
        if (strand=="+") {
          if (checkout2_dn[x,'lastexon',drop=F]==1) {next}
          
          # closest to peak
          bedposDN[2,2]=texn$CL-1
          bedposDN[2,3]=locx[(locx$exon_number %in%texn$exon_number),'end']
          
          
          # Upstream exon
          bedposDN[1,2]=locx[(locx$exon_number %in%(texn$exon_number+1)),'start']-1
          bedposDN[1,3]=locx[(locx$exon_number %in%(texn$exon_number+1)),'start']+(window-1-abs(texn$CL-(locx[(locx$exon_number %in%texn$exon_number),'end'])))-1
          
          checkout2_dn[x,'numberExCrossed']=nrow(bedposDN)
          
          # exon number of bedposDN[1,]
          exnum=(texn$exon_number)+(nrow(bedposDN)-1)
          while ((exnum<max(locx$exon_number))&(locx[(locx$exon_number %in%(exnum)),'end']<bedposDN[1,3])) {#xnxnxnx
            
            bedposDN[1,2]=locx[(locx$exon_number %in%(exnum)),'start']-1
            bedposDN[1,3]=locx[(locx$exon_number %in%(exnum)),'end']
            
            bedposDN2=matrix(nrow=1,ncol = 6)
            bedposDN2[1,2]=locx[(locx$exon_number %in%((exnum+1))),'start']-1
            bedposDN2[1,3]=locx[(locx$exon_number %in%((exnum+1))),'start']+((window-1)-sum(rowDiffs(bedposDN[,c(2,3)])))
            
            bedposDN=rbind(bedposDN2,bedposDN)
            checkout2_dn[x,'numberExCrossed']=nrow(bedposDN)
            exnum=exnum+1
            if (exnum==max(locx$exon_number)) {break}
          }
        }
        
        if (strand=="-") {
          if (checkout2_dn[x,'lastexon',drop=F]==1) {next}
          
          # closest to peak
          bedposDN[2,2]=locx[(locx$exon_number %in%texn$exon_number),'start']-1
          bedposDN[2,3]=texn$CL
          
          # downstream exon
          bedposDN[1,2]=locx[(locx$exon_number %in%(texn$exon_number+1)),'end']-((window-1)-abs(texn$CL-locx[(locx$exon_number %in%(texn$exon_number)),'start']))
          bedposDN[1,3]=locx[(locx$exon_number %in%(texn$exon_number+1)),'end']
          
          checkout2_dn[x,'numberExCrossed']=nrow(bedposDN)
          
          
          exnum=(texn$exon_number)+(nrow(bedposDN)-1)
          while ((exnum<max(locx$exon_number))&(locx[(locx$exon_number %in%(exnum)),'start']>bedposDN[1,2])) {#xnegxneg
            
            bedposDN[1,2]=locx[(locx$exon_number %in%(exnum)),'start']-1
            bedposDN[1,3]=locx[(locx$exon_number %in%(exnum)),'end']
            bedposDN2=matrix(nrow=1,ncol = 6)
            bedposDN2[1,2]=locx[(locx$exon_number %in%((exnum+1))),'end']-((window)-sum(rowDiffs(bedposDN[,c(2,3)])))
            bedposDN2[1,3]=locx[(locx$exon_number %in%((exnum+1))),'end']
            
            bedposDN=rbind(bedposDN2,bedposDN)
            checkout2_dn[x,'numberExCrossed']=nrow(bedposDN)
            exnum=exnum+1
            if (exnum==max(locx$exon_number)) {break}
          }
        }
        
        bedposDNx=as.data.frame(bedposDN)
        bedposDNx[,2]=as.numeric(as.character(bedposDNx[,2]))
        bedposDNx[,3]=as.numeric(as.character(bedposDNx[,3]))
        sum(rowDiffs(as.matrix(bedposDNx[,c(2,3)])) )       
        (rowDiffs(as.matrix(bedposDNx[,c(2,3)])) )       
        
        bedposDN[,1]=chr
        bedposDN[,4]=".";bedposDN[,5]="."
        bedposDN[,6]=strand
        
        write.table(bedposDN, file=paste0(scratch_DN,'/',texn$seqname,'_',texn$CL,'_chunkDN.bed'), quote=F, sep="\t", row.names=F, col.names=F)
        # system(paste0('bedtools getfasta -s -fi ',fasta,' -bed structure/chunkwindow/exon/',texn$ID,'_chunkDN.bed > structure/chunkwindow/exon/',texn$ID,'_chunk.faDN'))
        faDN=system(paste0('bedtools getfasta -s -fi ',fasta,' -bed ',scratch_DN,'/',texn$seqname,'_',texn$CL,'_chunkDN.bed'),intern = T)
        if (nchar(faDN[2])==window) {
          faDN2=faDN[1:2];faDN=faDN[1:2];
          checkout2_dn[x,'SingleExon']='Change-TRUE'
          checkout2_dn[x,'numberExCrossed']=checkout2_dn[x,'numberExCrossed']-1
        }
        
        if (length(faDN)>2) {
        faDN2=faDN[c(length(faDN):1)]
        for (f in seq(3,length(faDN2),2)) {
          faDN2[1]=paste(faDN2[1],faDN2[f],collapse = "")
          faDN2[1]=gsub(" ","",faDN2[1])
          
          if ((nchar(faDN2[1])==window)&f<max(seq(3,length(faDN2),2)) ){
            # checkout2_dn=TRUE
            checkout2_dn[x,'TranscChoise']='Change Multiple exons'
            faDN2=faDN2[1:2]
            checkout2_dn[x,'numberExCrossed']=checkout2_dn[x,'numberExCrossed']-1
            break}
        }
            faDN2=faDN2[2:1]
        }
        
        if (nchar(faDN2[2])!=window){print(nchar(faDN2[2]));shrtstrand}
         
        write.table(faDN2, file=paste0(scratch_DN,'/',texn$seqname,'_',texn$CL,'_chunkDN.fa'), quote=F, sep="\t", row.names=F, col.names=F)
        checkout2_dn[x,'sequence']=gsub('T','U',faDN2[2])
        checkout2_dn[x,'n']=faDN2[1]
        
        # faout_dn=c(faout_dn,faDN2[1:2])
           faout_dn[(2*x-1):(2*x),1]=(as.matrix(faDN2[1:2]))
        remove(bedposDN,faDN2)
      }
 
      if ((checkout2_dn[x,'SingleExon']==T)|(checkout2_dn[x,'lastexon']==1)) {
        bedposDN=matrix(ncol=6,nrow=1)
        bedposDN[1,1]=chr
        bedposDN[1,2]=s1dn
        bedposDN[1,3]=e1dn
        bedposDN[1,4]=".";bedposDN[,5]="."
        bedposDN[1,6]=strand
        
        write.table(bedposDN, file=paste0(scratch_DN,'/',texn$seqname,'_',texn$CL,'_chunkDN.bed'), quote=F, sep="\t", row.names=F, col.names=F)
        # system(paste0('bedtools getfasta -s -fi ',fasta,' -bed structure/chunkwindow/exon/',texn$ID,'_chunkDN.bed > structure/chunkwindow/exon/',texn$ID,'_chunk.faDN'))
        faDN=system(paste0('bedtools getfasta -s -fi ',fasta,' -bed ',scratch_DN,'/',texn$seqname,'_',texn$CL,'_chunkDN.bed'),intern = T)
        write.table(faDN, file=paste0(scratch_DN,'/',texn$seqname,'_',texn$CL,'_chunkDN.fa'), quote=F, sep="\t", row.names=F, col.names=F)
        checkout2_dn[x,'sequence']=gsub('T','U',faDN[2])
        checkout2_dn[x,'n']=faDN[1]
        checkout2_dn[x,'numberExCrossed']=0
        # faout_dn=c(faout_dn,faDN[1:2])
          faout_dn[(2*x-1):(2*x),1]=(as.matrix(faDN[1:2]))
        remove(bedposDN)
      }
      
      
####################################################################################################
#### reverse compliment of sequence if LISI on opposite strand
          if(is.na(isx$strand_LISIup)==F){if(isx$strand!=isx$strand_LISIup){
      faout[(2*x),1]=as.character(reverseComplement(DNAString(fa[2])))
      checkout2[x,'sequence']=as.character(reverseComplement(DNAString(fa[2])))
              checkout2[x,'sequence']=gsub('T','U',checkout2[x,'sequence'])
      }}
      
      if(is.na(isx$strand_LISIdn)==F){if(isx$strand!=isx$strand_LISIdn){
      faout_dn[(2*x),1]=as.character(reverseComplement(DNAString(faDN[2])))
      checkout2_dn[x,'sequence']=as.character(reverseComplement(DNAString(faDN[2])))
              checkout2_dn[x,'sequence']=gsub('T','U',checkout2_dn[x,'sequence'])
      }}
      
#############################################################################################################################
#############################################################################################################################
      
      if(file.exists(paste0(scratch_DN,'/',texn$seqname,'_',texn$CL,'_chunkDN.fa'))){file.remove(paste0(scratch_DN,'/',texn$seqname,'_',texn$CL,'_chunkDN.fa'))}
      if(file.exists(paste0(scratch_DN,'/',texn$seqname,'_',texn$CL,'_chunkDN.bed'))){file.remove(paste0(scratch_DN,'/',texn$seqname,'_',texn$CL,'_chunkDN.bed'))}
     
    }
    
      checkout2$length=apply(checkout2[,'sequence',drop=F], 1, nchar)
      checkout2$n=gsub(">","",checkout2$n)
      
    checkout2_dn$length=apply(checkout2_dn[,'sequence',drop=F], 1, nchar)
    checkout2_dn$n=gsub(">","",checkout2_dn$n)
    
  
  ##########################################################################################  
  ##### Create Fasta file for all locations
  ##########################################################################################  
    
    faout=faout[is.na(faout)==F,] 
    # faout=faout[duplicated(faout)==F]
    # faout[duplicated(faout)]

      faout=gsub('T','U',faout)
      faout=gsub('\\(\\+)',"",faout)
      faout=gsub('\\(\\-)',"",faout)
      
    faout_dn=faout_dn[is.na(faout_dn)==F,] 
    # faout_dn=faout_dn[duplicated(faout_dn)==F]

      faout_dn=gsub('T','U',faout_dn)
      faout_dn=gsub('\\(\\+)',"",faout_dn)
      faout_dn=gsub('\\(\\-)',"",faout_dn)
      
    write.table(faout, file=paste0(outdir_UP, '/AllLocation_Random.fa'), quote=F, sep="\t", row.names=F, col.names=F)
    write.table(faout_dn, file=paste0(outdir_DN, '/AllLocation_DN_Random.fa'), quote=F, sep="\t", row.names=F, col.names=F)
    
    write.table(faout, file=paste0('./structure/MEME/fasta/',CLIPtype,'_',RunMode,'_',TrasncLoc,'_UP.fa'), quote=F, sep="\t", row.names=F, col.names=F)
    write.table(faout_dn, file=paste0('./structure/MEME/fasta/',CLIPtype,'_',RunMode,'_',TrasncLoc,'_DN.fa'), quote=F, sep="\t", row.names=F, col.names=F)
    
    checkout2$n=gsub('\\(\\+)',"",checkout2$n)
    checkout2$n=gsub('\\(\\-)',"",checkout2$n) 
    
    checkout2_dn$n=gsub('\\(\\+)',"",checkout2_dn$n)
    checkout2_dn$n=gsub('\\(\\-)',"",checkout2_dn$n)
    
    
    
      checkout2=checkout2[is.na(checkout2$CLIP_ID)==F,]
      checkout2_dn=checkout2_dn[is.na(checkout2_dn$CLIP_ID)==F,]
      
   faCentout=as.data.frame(matrix(nrow=(nrow(checkout2)*2),ncol = 2))
   faCentout100=as.data.frame(matrix(nrow=(nrow(checkout2)*2),ncol = 2))
        for (f in 1:nrow(checkout2)) {
        # for (f in 1:3) {

          fid=checkout2[f,'CLIP_ID']
          fseq=checkout2[f,'sequence']
          fseq1=substr(fseq,((window/2)+1),nchar(fseq))

          fseqDN=checkout2_dn[checkout2_dn$CLIP_ID%in%fid,'sequence']
          fseqDN1=substr(fseqDN,2,((window/2)+1))
          fseqDN2=substr(fseqDN,2,nchar(fseqDN))

          fseqCent=paste0(fseq1,fseqDN1)
          faCentout[(f+(f-1)),1]=paste0(">",fid)
          faCentout[f+(f),1]=fseqCent
          
          fseqCent100=paste0(fseq,fseqDN2)
          faCentout100[(f+(f-1)),1]=paste0(">",fid)
          faCentout100[f+(f),1]=fseqCent100
          
            
        }
    
    write.table(faCentout$V1, file=paste0('./structure/MEME/fasta/',CLIPtype,'_',RunMode,'_',TrasncLoc,'_Centr.fa'), quote=F, sep="\t", row.names=F, col.names=F)
    write.table(faCentout100$V1, file=paste0('./structure/MEME/fasta/',CLIPtype,'_',RunMode,'_',TrasncLoc,'_Centr100.fa'), quote=F, sep="\t", row.names=F, col.names=F)

    
    # MD10=fread(paste0('./structure/chunkwindow/genee/CLIP_peaks_MD12_structure.txt') ,header = F,sep = "\t",stringsAsFactors = F,data.table=F)


```

 

```{r , include=FALSE, eval=F}
## Files for MEME 
  #################################################
  # Create fasta file for each Transcript component
  #################################################

mm10_trans_meme=mm10all[mm10all$feature%in%'transcript',]
mm10_trans_meme=mm10_trans_meme[mm10_trans_meme$transcript_type%in%'protein_coding',]
mm10_trans_meme=mm10_trans_meme[mm10_trans_meme$transcript_id%in%canonical$transcript,]

mm10_exons_meme=mm10all[mm10all$feature%in%'exon',]
mm10_exons_meme=mm10_exons_meme[mm10_exons_meme$transcript_id%in%mm10_trans_meme$transcript_id,]

mm10_UTR3meme=mm10all[mm10all$feature%in%'UTR',]
mm10_UTR3meme=mm10_UTR3meme[mm10_UTR3meme$transcript_id%in%mm10_trans_meme$transcript_id,]
mm10_UTR3meme=mm10_UTR3meme[mm10_UTR3meme$exon_number!=1,]

mm10_introns_meme=fread("/Users/homanpj/OneDrive - National Institutes of Health/Resources/ref/mm10/Gencode_VM23/fromUCSC/KnownGene/KnownGene_GRCm38_introns.bed", header=F, sep="\t",stringsAsFactors = F,data.table=F)
  mm10_introns_meme=mm10_introns_meme[grepl("_",mm10_introns_meme$V1)==F,]
  colnames(mm10_introns_meme)=c('seqname','start','end','attribute','V5','strand')
  mm10_introns_meme=separate(mm10_introns_meme,attribute,into=c('transcript_id','feature','exon_number','level','chr2','intronnumber','dir'),remove = T,sep = "_")
  mm10_introns_meme$transcript_id=removeVersion(mm10_introns_meme$transcript_id)
  mm10_introns_meme$start=as.numeric(mm10_introns_meme$start)
  mm10_introns_meme$end=as.numeric(mm10_introns_meme$end)
    mm10_introns_meme$start=mm10_introns_meme$start+1
    mm10_introns_meme$exon_number=as.numeric(mm10_introns_meme$exon_number)+1
  mm10_introns_meme=mm10_introns_meme[mm10_introns_meme$transcript_id%in%mm10_exons_meme$transcript_id,]


mm10_trans_meme_bed=mm10_trans_meme[,c('seqname','start','end','transcript_id','score','strand')]
mm10_exons_meme_bed=mm10_exons_meme[,c('seqname','start','end','transcript_id','score','strand')]
mm10_UTR3meme_bed=mm10_UTR3meme[,c('seqname','start','end','transcript_id','score','strand')]
mm10_introns_meme_bed=mm10_introns_meme[,c('seqname','start','end','transcript_id','level','strand')]
  mm10_introns_meme_bed$level="."
  


            write.table(mm10_trans_meme_bed, file=paste0('./structure/MEME/bed/mm10_trans_meme.bed'), quote=F, sep="\t", row.names=F, col.names=F)
            write.table(mm10_exons_meme_bed, file=paste0('./structure/MEME/bed/mm10_exons_meme.bed'), quote=F, sep="\t", row.names=F, col.names=F)
            write.table(mm10_UTR3meme_bed, file=paste0('./structure/MEME/bed/mm10_UTR3meme.bed'), quote=F, sep="\t", row.names=F, col.names=F)
            write.table(mm10_introns_meme_bed, file=paste0('./structure/MEME/bed/mm10_introns_meme.bed'), quote=F, sep="\t", row.names=F, col.names=F)
            
 system(paste0("bedtools getfasta -s -fi ",fasta," -bed ./structure/MEME/bed/mm10_trans_meme.bed | sed 's/T/U/g'  | sed 's/t/u/g'  > ./structure/MEME/fasta/mm10_trans_meme.fa"),
                          intern = F,ignore.stderr = T,ignore.stdout = F)
 system(paste0("bedtools getfasta -s -fi ",fasta," -bed ./structure/MEME/bed/mm10_exons_meme.bed | sed 's/T/U/g'  | sed 's/t/u/g'  > ./structure/MEME/fasta/mm10_exons_meme.fa"),
                          intern = F,ignore.stderr = T,ignore.stdout = F)
 system(paste0("bedtools getfasta -s -fi ",fasta," -bed ./structure/MEME/bed/mm10_UTR3meme.bed | sed 's/T/U/g'  | sed 's/t/u/g'  > ./structure/MEME/fasta/mm10_UTR3meme.fa"),
                         intern = F,ignore.stderr = T,ignore.stdout = F)
 system(paste0("bedtools getfasta -s -fi ",fasta," -bed ./structure/MEME/bed/mm10_introns_meme.bed | sed 's/T/U/g'  | sed 's/t/u/g'  > ./structure/MEME/fasta/mm10_introns_meme.fa"),
                         intern = F,ignore.stderr = T,ignore.stdout = F)

system(paste0("bedtools getfasta -s -fi ",fasta," -bed ./structure/MEME/bed/mm10_trans_meme.bed  > ./structure/MEME/fasta/mm10_trans_homer.fa"),
                          intern = F,ignore.stderr = T,ignore.stdout = F)
system(paste0("bedtools getfasta -s -fi ",fasta," -bed ./structure/MEME/bed/mm10_exons_meme.bed  > ./structure/MEME/fasta/mm10_exons_homer.fa"),
                          intern = F,ignore.stderr = T,ignore.stdout = F)
system(paste0("bedtools getfasta -s -fi ",fasta," -bed ./structure/MEME/bed/mm10_UTR3meme.bed  > ./structure/MEME/fasta/mm10_UTR3homer.fa"),
                         intern = F,ignore.stderr = T,ignore.stdout = F)
system(paste0("bedtools getfasta -s -fi ",fasta," -bed ./structure/MEME/bed/mm10_introns_meme.bed > ./structure/MEME/fasta/mm10_introns_homer.fa"),
                         intern = F,ignore.stderr = T,ignore.stdout = F)
  
        
         # length(unique(mm10_UTR3meme_bed$transcript_id))
         # length(unique(mm10_trans_meme_bed$transcript_id))

```


## Gquad identification  

G-quadroplex were identified by two methods.  
1) fastaRegexFinder  
fastaRegexFinder Identifies G-quadroplex based on the supplied Regular Expression for each sequence supplied in the fasta file.  
  
fastaRegexFinder.py -f CLIPsequences.fa -r ([gG]{2,4}w{1,7}){3,}[gG]{2,4} --noreverse > output.txt  
  
2) QGRS-Mapper  
QGRS-Mapper Identifies and scores G-quads based on GxNy1GxNy2GxNy3Gx.  
  
qgrs -s -csv -i SingleCLIPseqence.fa -o output.txt   
  
     
      

```{r , include=FALSE}     
    ##########################################################################################   
    ##### Gquad idnettification 
    ##########################################################################################   
    


    ###############################################################
    ## Upstream ###
    
    FAupstream=fread(paste0(outdir_UP,'/AllLocation_Random.fa') ,header = F,sep = "\t",stringsAsFactors = F,data.table=F)
    Gquadout=as.data.frame(matrix(nrow=(nrow(FAupstream)/2),ncol=5))
    colnames(Gquadout)=c('ID','qgrsSEQ','qgrsScore','qgrsStart','qgrsX')
    
    
    for (f in seq(1,nrow(FAupstream),2)) {
      fax=FAupstream[c(f,f+1),]
      nm=fax[1]
      nm=gsub(">","",nm);
      # nm=gsub("\\+","",nm);
      # nm=gsub("-","",nm);
      # nm=gsub("\\()","",nm);
      write.table(fax, file=paste0(scratch_UP,'/',nm,'.fa'), quote=F, sep="\t", row.names=F, col.names=F)
      ###############
      ## qgrs
      ###############
      (system(paste0(qgrs,' -s -csv -i ',scratch_UP,'/',nm,'.fa -o ',scratch_UP,'/',nm,'.txt'),intern = F,ignore.stderr = T));
    
        Gquadout[f,'ID']=fax[1]
      quad_qgrs=read.delim(paste0(scratch_UP,'/',nm,'.txt'),comment.char = "%",header = T,sep = ",") 
      if (nrow(quad_qgrs)>0){
        Gquadout[f,'qgrsSEQ']=paste(quad_qgrs$SEQ,collapse = ", ")
        Gquadout[f,'qgrsScore']=paste(quad_qgrs$GS,collapse = ", ")
        # Gquadout[f,'qgrsStart']=paste(quad_qgrs$T1,collapse = ", ")
        Gquadout[f,'qgrsX']=paste(quad_qgrs$X,collapse = ", ")}
      
      if(file.exists(paste0(scratch_UP,'/',nm,'.txt'))){file.remove(paste0(scratch_UP,'/',nm,'.txt'))}
      if(file.exists(paste0(scratch_UP,'/',nm,'.fa'))){file.remove(paste0(scratch_UP,'/',nm,'.fa'))}
    }
    Gquadout=Gquadout[is.na(Gquadout$ID)==F,]
    Gquadout$ID=gsub(">","",Gquadout$ID);
    
    
    ############### 
    ### fastaRegexFinder
    ###############
    system(paste0(fastaRegexFinder,' -f ',outdir_UP,'/AllLocation_Random.fa -r "([gG]{2,4}\\w{1,7}){3,}[gG]{2,4}" --noreverse > ',outdir_UP,'/out.Random.fold.txt'),intern = T,ignore.stderr = T)
    # system(paste0(fastaRegexFinder,' -f ',outdir_UP,'/AllLocation_Random.fa -r "([gG]{2,4}\\w{1,7}){3,}[gG]{2,4}" --noreverse '),intern = T,ignore.stderr = T)
    
    info = file.info(paste0(outdir_UP,'/out.Random.fold.txt'))
    if (info$size>0){
      quad=read.delim(paste0(outdir_UP,'/out.Random.fold.txt'),comment.char = "%",header = F,sep = "\t",stringsAsFactors = F)
    
    quadx2=quad
    dup=quadx2[duplicated(quadx2$V1),]
    
    for (d in 1:nrow(dup)) {
      dd=dup[d,]
      quadx=quadx2[quad$V1%in%dd$V1,]
      quadx2[quadx2$V1%in%dd$V1,'V2']=paste(quadx$V2,collapse = ",")
      quadx2[quadx2$V1%in%dd$V1,'V3']=paste(quadx$V3,collapse = ",")
      quadx2[quadx2$V1%in%dd$V1,'V5']=paste(quadx$V5,collapse = ",")
      quadx2[quadx2$V1%in%dd$V1,'V7']=paste(quadx$V7,collapse = ",")
    }
    quadx2=quadx2[duplicated(quadx2$V1)==F,]
      # if (nrow(quad)!=nrow(quadx2)){badquad}
    quad=quadx2
    Gquadout=merge(quad[,c('V1','V2','V3','V5','V6','V7')],Gquadout,by.x='V1',by.y='ID',all=T)
    colnames(Gquadout)[colnames(Gquadout)%in%'V1']='ID'
    
    if(file.exists(paste0(outdir_UP,'/out.Random.fold.txt'))){file.remove(paste0(outdir_UP,'/out.Random.fold.txt'))}
    
    }
    
    checkout2=merge(checkout2,Gquadout,by.x="n",by.y="ID",all.x=T)
  
          checkout2$qgrsSEQ= gsub(" ","",checkout2$qgrsSEQ)
          checkout2=cbind(checkout2,str_locate(string=checkout2[,'sequence'],pattern=checkout2[,'qgrsSEQ']))

          
    ## upstream - start of sequence is farthest away from crosslink start-------------------------CL
          checkout2$qgrs_maxScore=checkout2$qgrsScore
          checkout2$qgrsdist=100-checkout2$end
          checkout2$qgrsdist_maxScore=100-checkout2$end
          
          com=checkout2[grep(",",checkout2$qgrsSEQ),]
          if (nrow(com)>0) {
          for (cm in 1:nrow(com)) {
            com1=com[cm,]
            sq=str_split(com1$qgrsSEQ,",")[[1]]
            scr=str_split(com1$qgrsScore,",")[[1]]
                checkout2[checkout2$ID%in%com1$ID,'qgrs_maxScore']=max(as.numeric(scr))
            maxn=which(as.numeric(scr)%in%max(as.numeric(scr)))
            
            loc=str_locate(string=as.character(com1$sequence),pattern=sq)
            checkout2[checkout2$ID%in%com1$ID,'Qstart'] = paste(loc[,'start'],collapse = ',')
            checkout2[checkout2$ID%in%com1$ID,'Qend'] = paste(loc[,'end'],collapse = ',')
            
            checkout2[checkout2$ID%in%com1$ID,'qgrsdist'] = paste(100-loc[,'end'],collapse = ',')
            checkout2[checkout2$ID%in%com1$ID,'qgrsdist_maxScore']=min(100-loc[maxn,'end'])
          
          }}


    ########################################################################
    ## downstream ###
    
    FAdownstream=fread(paste0(outdir_DN,'/AllLocation_DN_Random.fa'),header = F,sep = "\t",stringsAsFactors = F,data.table=F)
    Gquadout_dn=as.data.frame(matrix(nrow=(nrow(FAdownstream)/2),ncol=5))
    colnames(Gquadout_dn)=c('ID','qgrsSEQ','qgrsScore','qgrsStart','qgrsX')
    
    ###############
      ## qgrs
    ###############
    for (f in seq(1,nrow(FAdownstream),2)) {
      fax=FAdownstream[c(f,f+1),]
      nm=fax[1]
      nm=gsub(">","",nm);
      # nm=gsub("\\+","",nm);
      # nm=gsub("-","",nm);
      # nm=gsub("\\()","",nm);
      write.table(fax, file=paste0(scratch_DN,'/',nm,'.fa'), quote=F, sep="\t", row.names=F, col.names=F)
     
      (system(paste0(qgrs,' -s -csv -i ',scratch_DN,'/',nm,'.fa -o ',scratch_DN,'/',nm,'.txt'),intern = F,ignore.stderr = T));
      
      Gquadout_dn[f,'ID']=fax[1]
      quad_qgrs_dn=read.delim(paste0(scratch_DN,'/',nm,'.txt'),comment.char = "%",header = T,sep = ",") 
      if (nrow(quad_qgrs_dn)>0){
        Gquadout_dn[f,'qgrsSEQ']=paste(quad_qgrs_dn$SEQ,collapse = ", ")
        Gquadout_dn[f,'qgrsScore']=paste(quad_qgrs_dn$GS,collapse = ", ")
        # Gquadout_dn[f,'qgrsStart']=paste(quad_qgrs_dn$T1,collapse = ", ")
        Gquadout_dn[f,'qgrsX']=paste(quad_qgrs_dn$X,collapse = ", ")}
      
      if(file.exists(paste0(scratch_DN,'/',nm,'.txt'))){file.remove(paste0(scratch_DN,'/',nm,'.txt'))}
      if(file.exists(paste0(scratch_DN,'/',nm,'.fa'))){file.remove(paste0(scratch_DN,'/',nm,'.fa'))}
    }
    Gquadout_dn=Gquadout_dn[is.na(Gquadout_dn$ID)==F,]
    Gquadout_dn$ID=gsub(">","",Gquadout_dn$ID);
    
    ############### 
    ### fastaRegexFinder
    ###############    
    system(paste0(fastaRegexFinder,' -f ',outdir_DN,'/AllLocation_DN_Random.fa -r "([gG]{2,4}\\w{1,7}){3,}[gG]{2,4}" --noreverse > ',outdir_DN,'/out.Random.fold.txt'),intern = F,ignore.stderr = T)
    # system(paste0(fastaRegexFinder,' -f ',outdir_DN,'/AllLocation_DN_Random.fa -r "([gG]{2,4}\\w{1,7}){3,}[gG]{2,4}" --noreverse '),intern = T,ignore.stderr = T)
    
    info = file.info(paste0(outdir_DN,'/out.Random.fold.txt'))
    if (info$size>0){
      quad_dn=read.delim(paste0(outdir_DN,'/out.Random.fold.txt'),comment.char = "%",header = F,sep = "\t",stringsAsFactors = F)
    
    dup=quad_dn[duplicated(quad_dn$V1),]
    quad_dnx2=quad_dn
     
    for (d in 1:nrow(dup)) {
      dd=dup[d,]
      quad_dnx=quad_dn[quad_dn$V1%in%dd$V1,]
      quad_dnx2[quad_dnx2$V1%in%dd$V1,'V2']=paste(quad_dnx$V2,collapse = ",")
      quad_dnx2[quad_dnx2$V1%in%dd$V1,'V3']=paste(quad_dnx$V3,collapse = ",")
      quad_dnx2[quad_dnx2$V1%in%dd$V1,'V5']=paste(quad_dnx$V5,collapse = ",")
      quad_dnx2[quad_dnx2$V1%in%dd$V1,'V7']=paste(quad_dnx$V7,collapse = ",")
    }
    quad_dn=quad_dnx2[duplicated(quad_dnx2$V1)==F,]
  
      Gquadout_dn=merge(quad_dn[,c('V1','V2','V3','V5','V6','V7')],Gquadout_dn,by.x='V1',by.y='ID',all=T)
      colnames(Gquadout_dn)[colnames(Gquadout_dn)%in%'V1']='ID'
      
    if(file.exists(paste0(outdir_DN,'/out.Random.fold.txt'))){file.remove(paste0(outdir_DN,'/out.Random.fold.txt'))}
    
    }
    
    checkout2_dn=merge(checkout2_dn,Gquadout_dn,by.x="n",by.y="ID",all.x=T)
    
    checkout2_dn$qgrsSEQ= gsub(" ","",checkout2_dn$qgrsSEQ)
    checkout2_dn=cbind(checkout2_dn,str_locate(string=checkout2_dn[,'sequence'],pattern=checkout2_dn[,'qgrsSEQ']))
    
    ## upstream - start of sequence is farthest away from crosslink CL-------------------------end
      checkout2_dn$qgrs_maxScore=checkout2_dn$qgrsScore
    checkout2_dn$qgrsdist=checkout2_dn$start
    checkout2_dn$qgrsdist_maxScore=checkout2_dn$start
    
    com=checkout2_dn[grep(",",checkout2_dn$qgrsSEQ),]
    if (nrow(com)>0) {
    for (cm in 1:nrow(com)) {
      com1=com[cm,]
      sq=str_split(com1$qgrsSEQ,",")[[1]]
      scr=str_split(com1$qgrsScore,",")[[1]]
           checkout2_dn[checkout2_dn$ID%in%com1$ID,'qgrs_maxScore']=max(as.numeric(scr))
      maxn=which(as.numeric(scr)%in%max(as.numeric(scr)))
      
      loc=str_locate(string=as.character(com1$sequence),pattern=sq)
      checkout2_dn[checkout2_dn$ID%in%com1$ID,'Qstart'] = paste(loc[,'start'],collapse = ',')
      checkout2_dn[checkout2_dn$ID%in%com1$ID,'Qend'] = paste(loc[,'end'],collapse = ',')
      
      checkout2_dn[checkout2_dn$ID%in%com1$ID,'qgrsdist'] = paste(loc[,'start'],collapse = ',')
      checkout2_dn[checkout2_dn$ID%in%com1$ID,'qgrsdist_maxScore']=min(loc[maxn,'start'])
    }}
    
    
    
    # } #eliminate to calculate structure : repeat loop
    # }) #eliminate to calculate structure : RUNtime
```

 
### GQuad Analysis  

```{r ,echo=Ffig.asp=.5, include=T, fig.align='center',fig.height=3, fig.width=3, figures-side, fig.show="hold"}
##```{r , include=T,fig.asp=1.25, fig.align='center',fig.height=4, fig.width=3}
    

#######################################       
### # of gquad identified
#######################################       

      GdistRando=as.matrix(read.delim(paste0(inRandom,'/Random1/',CLIPtype,'/',TrasncLoc,'/',RunMode,'/upstream/Table/Col_GscoreDist_UP.txt'),comment.char = "%",header = F,sep = "\t",stringsAsFactors = F))
    
      for (rand in c(2:12)) {

            GdistRando2=as.matrix(read.delim(paste0(inRandom,'/Random',rand,'/',CLIPtype,'/',TrasncLoc,'/',RunMode,'/upstream/Table/Col_GscoreDist_UP.txt'),comment.char = "%",header = F,sep = "\t",stringsAsFactors = F))
           GdistRando=as.data.frame(cbind(GdistRando,GdistRando2))
      }

            GdistRandoP=as.data.frame(colSums(is.na(GdistRando)==F))
            colnames(GdistRandoP)=c("Number_Seq_with_GQuads")

            GdistRandoP=GdistRandoP[GdistRandoP$Number_Seq_with_GQuads>0,,drop=F]
            GdistRandoP[,1]=as.numeric(GdistRandoP[,1])
      
       ggplot(GdistRandoP,aes(x=Number_Seq_with_GQuads))+geom_density(size= .6) + theme_classic()+ylab("Density")+xlab("# Locations with identified G quadruplex")+ggtitle('Upstream:# of sequencens with quadruplex' )+
       geom_vline(xintercept=nrow(checkout2[is.na(checkout2$qgrs_maxScore)==F,]),colour="red") +xlim(15,70)


####################

      GdistRando_dn=as.matrix(read.delim(paste0(inRandom,'/Random1/',CLIPtype,'/',TrasncLoc,'/',RunMode,'/downstream/Table/Col_GscoreDist_DN.txt'),comment.char = "%",header = F,sep = "\t",stringsAsFactors = F))
     
      for (rand in c(2:12)) {
            GdistRando_dn2=as.matrix(read.delim(paste0(inRandom,'/Random',rand,'/',CLIPtype,'/',TrasncLoc,'/',RunMode,'/downstream/Table/Col_GscoreDist_DN.txt'),comment.char = "%",header = F,sep = "\t",stringsAsFactors = F))
       GdistRando_dn=as.data.frame(cbind(GdistRando_dn,GdistRando_dn2));
      }
       
               GdistRando_dnP=as.data.frame(colSums(is.na(GdistRando_dn)==F))
               colnames(GdistRando_dnP)=c("Number_Seq_with_GQuads")
               GdistRando_dnP=GdistRando_dnP[GdistRando_dnP$Number_Seq_with_GQuads>0,,drop=F]

               GdistRando_dnP[,1]=as.numeric(GdistRando_dnP[,1])
      
       ggplot(GdistRando_dnP,aes(x=Number_Seq_with_GQuads))+geom_density(size= .6) + theme_classic()+ylab("Density")+xlab("# Locations with identified G quadruplex")+ggtitle('Downstream:# of sequencens with quadruplex' )+
       geom_vline(xintercept=nrow(checkout2_dn[is.na(checkout2_dn$qgrs_maxScore)==F,]),colour="red") +xlim(15,70)

# intersect(checkout2$sequence,checkout2_dn$sequence)
# intersect(checkout2$V7,checkout2_dn$V7)
#        


#######################################       
### Distance of Gquad from CLIP start       
#######################################       
       
        GscrRando=as.matrix(read.delim(paste0(inRandom,'/Random1/',CLIPtype,'/',TrasncLoc,'/',RunMode,'/upstream/Table/Col_Gscore_UP.txt'),comment.char = "%",header = F,sep = "\t",stringsAsFactors = F))
                   for (rand in c(2:12)) {

      GdscrRando2=as.matrix(read.delim(paste0(inRandom,'/Random',rand,'/',CLIPtype,'/',TrasncLoc,'/',RunMode,'/upstream/Table/Col_Gscore_UP.txt'),comment.char = "%",header = F,sep = "\t",stringsAsFactors = F))
      GscrRando=as.data.frame(cbind(GscrRando,GdscrRando2));
                   }
       
      DistnaceDist=as.data.frame(matrix(nrow = nrow(GscrRando),ncol = ncol(GscrRando)))
      GscoreMax=as.data.frame(matrix(nrow = nrow(GscrRando),ncol = ncol(GscrRando)))
      GscoreMaxfrac=as.data.frame(matrix(nrow = nrow(GscrRando),ncol = ncol(GscrRando)))

      for (col in 1:ncol(GscrRando)) {
                  colscr1=((str_split(GscrRando[,col],pattern = ",")))
                  colscr1max=unlist(lapply(colscr1,  function(x) unique(max(as.numeric(x))) ))
                  
                  coldist1=((str_split(GdistRando[,col],pattern = ",")))

                  rowmaxscore=as.data.frame(matrix(nrow=length(colscr1),ncol=1))
                  rowmaxscoredist=as.data.frame(matrix(nrow=length(colscr1),ncol=1))
              for (r in 1:length(colscr1)) {
                        maxscr=grep(colscr1max[r],colscr1[[r]])
                        maxscrdist=as.numeric(coldist1[[r]][maxscr])
                       if (length(maxscr>1)) {
                         maxscr=maxscr[1]
                         maxscrdist=min(as.numeric(coldist1[[r]]))
                         }
                    rowmaxscore[r,]=maxscr
                    rowmaxscoredist[r,]=maxscrdist
              }     
                  DistnaceDist[,col]=rowmaxscoredist[,1]
                   GscoreMax[,col]=colscr1max
                   GscoreMaxfrac[1,col]=sum(colscr1max>21,na.rm = T)/(sum(is.na(colscr1max)==F,na.rm = T))
                   # GscoreMaxfrac[1,col]=sum(colscr1max>21,na.rm = T)

              # xxx=cbind(rowmaxscore,rowdist,GscrRando[,col],GdistRando[,col])
      }
       
      R=melt(DistnaceDist);colnames(R)[2]="qgrsdist_maxScore"
      R$variable=as.character(R$variable)
      
      D=checkout2[,'qgrsdist_maxScore',drop=F]
      D$variable="upstream"
      
ggplot(R,aes(x=qgrsdist_maxScore,line=variable))+
   geom_density(data=R,size= .6,color='grey') +
  geom_density(data=D,size= .6,color='red')+
  theme_classic()+ylab("Density")+xlab("G-Quadruplexes Distance from 5' CLIP peak (nt)")+ggtitle('Upstream G-Quadruplexes Distance')
       #geom_vline(xintercept=nrow(checkout2[is.na(checkout2$qgrs_maxScore)==F,]),colour="red") +xlim(0,100)

       
       R=melt(GscoreMax);colnames(R)[2]="qgrs_maxScore"
      R$variable=as.character(R$variable)
      
      D=as.data.frame(as.numeric(checkout2[,'qgrs_maxScore',drop=T]))
      colnames(D)[1]="qgrs_maxScore"
      D$variable="downstream"
      
ggplot(R,aes(x=qgrs_maxScore,line=variable))+
   geom_density(data=R,size= .6,color='grey') +
  geom_density(data=D,size= .6,color='red')+
  theme_classic()+ylab("Density")+xlab("G-Quadruplexes Score (qgrs)")+ggtitle('Upstream G-Quadruplexes score')
       #geom_vline(xintercept=nrow(checkout2[is.na(checkout2$qgrs_maxScore)==F,]),colour="red") +xlim(0,100)
 

########

 R=as.data.frame(t(GscoreMaxfrac[1,]));
 colnames(R)="qgrs_maxScoreFrac"
      # R$variable=as.character(R$variable)
      
   
ggplot(R,aes(x=qgrs_maxScoreFrac))+geom_density(size= .6) + theme_classic()+ylab("Density")+xlab("Fraction of G-Quadruplexs with scores >21")+ggtitle('Upstream: Fraction of G-Quadruplexs with High scores' )+
       geom_vline(xintercept=sum(as.numeric(checkout2$qgrs_maxScore)>21,na.rm = T)/sum(is.na(checkout2$qgrs_maxScore)==F,na.rm = T),colour="red") 
      
 # ggplot(R,aes(x=qgrs_maxScoreFrac))+geom_density(size= .6) + theme_classic()+ylab("Density")+xlab("Upstream: Fraction of G-Quadruplexs with scores >21")+ggtitle('# of G-Quadruplexs with High scores' )+
 #       geom_vline(xintercept=sum(as.numeric(checkout2$qgrs_maxScore)>21,na.rm = T),colour="red")
   

      
#########

GscrRando_dn=as.matrix(read.delim(paste0(inRandom,'/Random1/',CLIPtype,'/',TrasncLoc,'/',RunMode,'/downstream/Table/Col_Gscore_DN.txt'),comment.char = "%",header = F,sep = "\t",stringsAsFactors = F))
            for (rand in c(2:12)) {

      GdscrRando_dn2=as.matrix(read.delim(paste0(inRandom,'/Random',rand,'/',CLIPtype,'/',TrasncLoc,'/',RunMode,'/downstream/Table/Col_Gscore_DN.txt'),comment.char = "%",header = F,sep = "\t",stringsAsFactors = F))
      GscrRando_dn=as.data.frame(cbind(GscrRando_dn,GdscrRando_dn2));
            }
      
      DistnaceDist_dn=as.data.frame(matrix(nrow = nrow(GscrRando_dn),ncol = ncol(GscrRando_dn)))
      GscoreMax_dn=as.data.frame(matrix(nrow = nrow(GscrRando_dn),ncol = ncol(GscrRando_dn)))
      GscoreMaxfrac_dn=as.data.frame(matrix(nrow = nrow(GscrRando_dn),ncol = ncol(GscrRando_dn)))
      for (col in 1:ncol(GscrRando_dn)) {
                  colscr1=((str_split(GscrRando_dn[,col],pattern = ",")))
                  colscr1max=unlist(lapply(colscr1,  function(x) unique(max(as.numeric(x))) ))
                  
                  coldist1=((str_split(GdistRando_dn[,col],pattern = ",")))

                  rowmaxscore=as.data.frame(matrix(nrow=length(colscr1),ncol=1))
                  rowmaxscoredist=as.data.frame(matrix(nrow=length(colscr1),ncol=1))
              for (r in 1:length(colscr1)) {
                        maxscr=grep(colscr1max[r],colscr1[[r]])
                        maxscrdist=as.numeric(coldist1[[r]][maxscr])
                       if (length(maxscr>1)) {
                         maxscr=maxscr[1]
                         maxscrdist=min(as.numeric(coldist1[[r]]))
                         }
                    rowmaxscore[r,]=maxscr
                    rowmaxscoredist[r,]=maxscrdist
              }     
                  DistnaceDist_dn[,col]=rowmaxscoredist[,1]
                  GscoreMax_dn[,col]=colscr1max
                  GscoreMaxfrac_dn[,col]=sum(colscr1max>21,na.rm = T)/sum(is.na(colscr1max)==F,na.rm = T)

              # xxx=cbind(rowmaxscore,rowdist,GscrRando_dn[,col],GdistRando_dn[,col])
      }
       
      R=melt(DistnaceDist_dn);colnames(R)[2]="qgrsdist_maxScore"
      R$variable=as.character(R$variable)
      
      D=checkout2_dn[,'qgrsdist_maxScore',drop=F]
      D$variable="downstream"
      
ggplot(R,aes(x=qgrsdist_maxScore,line=variable))+
   geom_density(data=R,size= .6,color='grey') +
  geom_density(data=D,size= .6,color='red')+
  theme_classic()+ylab("Density")+xlab("G-Quadruplexes Distance from 5' CLIP peak (nt)")+ggtitle('Downstream G-Quadruplexes Distance')
       #geom_vline(xintercept=nrow(checkout2_dn[is.na(checkout2_dn$qgrs_maxScore)==F,]),colour="red") #+xlim(0,100)



 R=melt(GscoreMax_dn);colnames(R)[2]="qgrs_maxScore"
      R$variable=as.character(R$variable)
      
D=as.data.frame(as.numeric(checkout2_dn[,'qgrs_maxScore',drop=T]))
      colnames(D)[1]="qgrs_maxScore"
      D$variable="downstream"
      
ggplot(R,aes(x=qgrs_maxScore,line=variable))+
   geom_density(data=R,size= .6,color='grey') +
  geom_density(data=D,size= .6,color='red')+
  theme_classic()+ylab("Density")+xlab("G-Quadruplexes Score (qgrs)")+ggtitle('Downstream G-Quadruplexes score')
       #geom_vline(xintercept=nrow(checkout2_dn[is.na(checkout2_dn$qgrs_maxScore)==F,]),colour="red") #+xlim(0,100)



########

 R=as.data.frame(t(GscoreMaxfrac_dn[1,]));
 colnames(R)="qgrs_maxScoreFrac"
      # R$variable=as.character(R$variable)
      
   
ggplot(R,aes(x=qgrs_maxScoreFrac))+geom_density(size= .6) + theme_classic()+ylab("Density")+xlab("Fraction of G-Quadruplexs with scores >21")+ggtitle('Downstream: Fraction of G-Quadruplexs with High scores' )+
       geom_vline(xintercept=sum(as.numeric(checkout2_dn$qgrs_maxScore)>21,na.rm = T)/sum(is.na(checkout2_dn$qgrs_maxScore)==F,na.rm = T),colour="red") 
      



```   

    
 
## Nucleotide Content    

Individual nucleotide content for the Upstream and Dwonstream sequences of the supplied CLIP location.  

```{r ,echo=F,include=FALSE}
 ##########################################################################################   
  ##### Count Nucleotide Content 
  ##########################################################################################  

   
  NucHeatPlot =function(DF1,DF2,ordermatrix,nuc,region){
    # DF1=chucnkseqG
    # DF2=checkout2
        col_fun = colorRamp2(c(0, .25, .50), c("blue", "white", "red"))
    ################
    nucmap=DF1[rownames(ordermatrix),]
      GscrAnno=DF2
      rownames(GscrAnno)=GscrAnno$ID
      GscrAnno=GscrAnno[rownames(nucmap),'qgrs_maxScore',drop=F]
              # rownames(nucmap)[33]
              # rownames(GscrAnno)[33]
      GscrAnno=rowAnnotation(Gquad_score = row_anno_barplot(as.numeric(GscrAnno$qgrs_maxScore)))
          
       p= ComplexHeatmap::Heatmap(as.matrix(nucmap),col = col_fun,name = "mat", 
                                  column_title = "Distance from CLIP peak",column_title_side = "bottom",
                                  row_title = paste0(CLIPtype," CLIP peaks"),row_title_side = "left",show_row_names=F,
                                  cluster_columns = F,cluster_rows =F,right_annotation = GscrAnno)
        draw(p,column_title= paste0('% ',nuc,' in ',ch,'nt sliding window - ',region,' 100nt'),column_title_side = "top")
        #################
  }

    checkout2=checkout2[is.na(checkout2$sequence)==F,]
    checkout2$sequence=gsub('T','U',checkout2$sequence)
    
    checkout2_dn=checkout2_dn[is.na(checkout2_dn$sequence)==F,]
    checkout2_dn$sequence=gsub('T','U',checkout2_dn$sequence)
```

### Upstream  
```{r , include=T, fig.align='center',fig.height=2, fig.width=2, figures-side, fig.show="hold"}
# ```{r , include=T,fig.asp=1.25, fig.align='center',fig.height=4, fig.width=3,eval=T}

  # upstream   
    
    for (ch in c(10,20,30,40,50)) { 
    # for (ch in c(20)) { 
    chadj=ch-1

    chucnkseqA=as.data.frame(matrix(nrow=nrow(checkout2),ncol=(nchar(checkout2[1,'sequence'])-chadj)))
    rownames(chucnkseqA)=checkout2$ID;colnames(chucnkseqA)=seq(-(nchar(checkout2[1,'sequence'])-chadj),-1)
    chucnkseqU=as.data.frame(matrix(nrow=nrow(checkout2),ncol=(nchar(checkout2[1,'sequence'])-chadj)))
    rownames(chucnkseqU)=checkout2$ID;colnames(chucnkseqU)=seq(-(nchar(checkout2[1,'sequence'])-chadj),-1)
    chucnkseqG=as.data.frame(matrix(nrow=nrow(checkout2),ncol=(nchar(checkout2[1,'sequence'])-chadj)))
    rownames(chucnkseqG)=checkout2$ID;colnames(chucnkseqG)=seq(-(nchar(checkout2[1,'sequence'])-chadj),-1)
    chucnkseqC=as.data.frame(matrix(nrow=nrow(checkout2),ncol=(nchar(checkout2[1,'sequence'])-chadj)))
    rownames(chucnkseqC)=checkout2$ID;colnames(chucnkseqC)=seq(-(nchar(checkout2[1,'sequence'])-chadj),-1)
    
    for (l in 1:(nchar(checkout2[1,'sequence'])-chadj)) {
      ss= substr(checkout2[,'sequence'], start = l, stop = (l+chadj))
      chucnkseqA[,l]=str_count(ss, "A") /ch
      chucnkseqU[,l]=str_count(ss, "U") /ch
      chucnkseqG[,l]=str_count(ss, "G") /ch
      chucnkseqC[,l]=str_count(ss, "C") /ch
    }
    
        chucnkseqG=chucnkseqG[order(rowMaxs(as.matrix(chucnkseqG)),decreasing = T),]

      # NucHeatPlot(chucnkseqA,checkout2,chucnkseqG,"A",'upstream')
      # NucHeatPlot(chucnkseqU,checkout2,chucnkseqG,"U",'upstream')
      # NucHeatPlot(chucnkseqG,checkout2,chucnkseqG,"G",'upstream')
      # NucHeatPlot(chucnkseqC,checkout2,chucnkseqG,"C",'upstream')
        
      

      cht=cbind((colMeans(chucnkseqA)),(colMeans(chucnkseqU)),(colMeans(chucnkseqC)),(colMeans(chucnkseqG)))
colnames(cht)=c('A','U','C','G')
cht=melt(cht)
colnames(cht)=c('distance','Nucleotide','Percent')

####### plot CLIP peak data
# print(ggplot(cht,aes(y=Percent,x=distance,line=Nucleotide,color=Nucleotide))+
#                  xlab("Position of Window\n(Distance from 5'Clip peak)")+ylab('Nucleotide content ')+theme_classic()+ggtitle(paste0('window size=',ch,' - upstream 100nt'))+
#       geom_line(data=cht,na.rm=F)+  ylim(.1,.4))

########### Import Random data
    
      GRandom_A=as.matrix(read.delim(paste0(inRandom,'/Random1/',CLIPtype,'/',TrasncLoc,'/',RunMode,'/upstream/chucnkseqA_window',ch,'.txt'),comment.char = "%",header = T,sep = "\t",stringsAsFactors = F))
                        colnames(GRandom_A)=seq(-(nchar(checkout2[1,'sequence'])-chadj),-1)

      GRandom_U=as.matrix(read.delim(paste0(inRandom,'/Random1/',CLIPtype,'/',TrasncLoc,'/',RunMode,'/upstream/chucnkseqU_window',ch,'.txt'),comment.char = "%",header = T,sep = "\t",stringsAsFactors = F))
                              colnames(GRandom_U)=seq(-(nchar(checkout2[1,'sequence'])-chadj),-1)  
                              
      GRandom_G=as.matrix(read.delim(paste0(inRandom,'/Random1/',CLIPtype,'/',TrasncLoc,'/',RunMode,'/upstream/chucnkseqG_window',ch,'.txt'),comment.char = "%",header = T,sep = "\t",stringsAsFactors = F))
                              colnames(GRandom_G)=seq(-(nchar(checkout2[1,'sequence'])-chadj),-1) 
      
       GRandom_C=as.matrix(read.delim(paste0(inRandom,'/Random1/',CLIPtype,'/',TrasncLoc,'/',RunMode,'/upstream/chucnkseqC_window',ch,'.txt'),comment.char = "%",header = T,sep = "\t",stringsAsFactors = F))
                              colnames(GRandom_C)=seq(-(nchar(checkout2[1,'sequence'])-chadj),-1)                                                
                              
for (rand in c(2:12)) {
  GRandom2_Ax=as.matrix(read.delim(paste0(inRandom,'/Random',rand,'/',CLIPtype,'/',TrasncLoc,'/',RunMode,'/upstream/chucnkseqA_window',ch,'.txt'),comment.char = "%",header = T,sep = "\t",stringsAsFactors = F))
                        colnames(GRandom2_Ax)=seq(-(nchar(checkout2[1,'sequence'])-chadj),-1)
                         GRandom_A=rbind(GRandom_A,GRandom2_Ax)
      
  GRandom2_Ux=as.matrix(read.delim(paste0(inRandom,'/Random',rand,'/',CLIPtype,'/',TrasncLoc,'/',RunMode,'/upstream/chucnkseqU_window',ch,'.txt'),comment.char = "%",header = T,sep = "\t",stringsAsFactors = F))
                        colnames(GRandom2_Ux)=seq(-(nchar(checkout2[1,'sequence'])-chadj),-1)
                         GRandom_U=rbind(GRandom_U,GRandom2_Ux)
                         
  GRandom2_Gx=as.matrix(read.delim(paste0(inRandom,'/Random',rand,'/',CLIPtype,'/',TrasncLoc,'/',RunMode,'/upstream/chucnkseqG_window',ch,'.txt'),comment.char = "%",header = T,sep = "\t",stringsAsFactors = F))
                        colnames(GRandom2_Gx)=seq(-(nchar(checkout2[1,'sequence'])-chadj),-1)
                         GRandom_G=rbind(GRandom_G,GRandom2_Gx)
                         
  GRandom2_Cx=as.matrix(read.delim(paste0(inRandom,'/Random',rand,'/',CLIPtype,'/',TrasncLoc,'/',RunMode,'/upstream/chucnkseqC_window',ch,'.txt'),comment.char = "%",header = T,sep = "\t",stringsAsFactors = F))
                        colnames(GRandom2_Cx)=seq(-(nchar(checkout2[1,'sequence'])-chadj),-1)
                         GRandom_C=rbind(GRandom_C,GRandom2_Cx)
  
}
      
                colnames(GRandom_A)=seq(-(nchar(checkout2[1,'sequence'])-chadj),-1)
            GRandom_A_avg=colMeans(GRandom_A)    
                GRandom_A= melt(t(GRandom_A))
                GRandom_A$X2=as.character(GRandom_A$X2)
                colnames(GRandom_A)=c('distance','Nucleotide','Percent')

                colnames(GRandom_U)=seq(-(nchar(checkout2[1,'sequence'])-chadj),-1)
            GRandom_U_avg=colMeans(GRandom_U)    
                GRandom_U= melt(t(GRandom_U))
                GRandom_U$X2=as.character(GRandom_U$X2)
                colnames(GRandom_U)=c('distance','Nucleotide','Percent')
      
                colnames(GRandom_G)=seq(-(nchar(checkout2[1,'sequence'])-chadj),-1)
            GRandom_G_avg=colMeans(GRandom_G)    
                GRandom_G= melt(t(GRandom_G))
                GRandom_G$X2=as.character(GRandom_G$X2)
                colnames(GRandom_G)=c('distance','Nucleotide','Percent')
                
                colnames(GRandom_C)=seq(-(nchar(checkout2[1,'sequence'])-chadj),-1)
            GRandom_C_avg=colMeans(GRandom_C)    
                GRandom_C= melt(t(GRandom_C))
                GRandom_C$X2=as.character(GRandom_C$X2) 
                colnames(GRandom_C)=c('distance','Nucleotide','Percent')


                
 Rht=cbind(((GRandom_A_avg)),((GRandom_U_avg)),((GRandom_C_avg)),((GRandom_G_avg)))
colnames(Rht)=c('A','U','C','G')
Rht=melt(Rht)
colnames(Rht)=c('distance','Nucleotide','Percent')

####### plot Random Averages
# print(ggplot(Rht,aes(y=Percent,x=distance,line=Nucleotide,color=Nucleotide))+
#                  xlab("Position of Window\n(Distance from 5'Clip peak)")+ylab('Nucleotide content')+theme_classic()+ggtitle(paste0('Random Sequenc:window size=',ch,' - upstream 100nt'))+
#       geom_line(na.rm=F))


####### plot all nuc with all random
# print(ggplot(cht,aes(y=Percent,x=distance,line=Nucleotide,color=Nucleotide))+
#                  xlab("Position of Window\n(Distance from 5'Clip peak)")+ylab('Nucleotide content')+theme_classic()+ggtitle(paste0('window size=',ch,' - upstream 100nt'))+
#       geom_line(data = GRandom_A,color='red',alpha=0.05)+
#       geom_line(data = GRandom_U,color='purple',alpha=0.05)+
#       geom_line(data = GRandom_G,color='blue',alpha=0.05)+
#       geom_line(data = GRandom_C,color='green',alpha=0.05)+
#       geom_line(data=cht,na.rm=F))


###
print(ggplot(cht,aes(y=Percent,x=distance,line=Nucleotide,color=Nucleotide))+
                 xlab("Position of Window\n(Distance from 5'Clip peak)")+ylab('Nucleotide content ')+theme_classic()+ggtitle(paste0('window size=',ch,' - upstream 100nt'))+
      geom_line(data=cht,na.rm=F)+
        geom_line(data=Rht,na.rm=F,alpha=0.25,size=2)+  ylim(.1,.4))

################# PLOT INDIVIDUAL Nuc with Random backgound
chtx=cht[cht$Nucleotide%in%"A",]
print(ggplot(chtx,aes(y=Percent,x=distance,line=Nucleotide,color=Nucleotide))+
                 xlab("Position of Window\n(Distance from 5'Clip peak)")+ylab('Nucleotide content (A)')+theme_classic()+ggtitle(paste0('window size=',ch,' - upstream 100nt'))+
      geom_line(data = GRandom_A,color='grey',alpha=0.05)+
      geom_line(data=chtx,na.rm=F,color='red')+  ylim(.1,.4))

chtx=cht[cht$Nucleotide%in%"U",]
print(ggplot(chtx,aes(y=Percent,x=distance,line=Nucleotide,color=Nucleotide))+
                 xlab("Position of Window\n(Distance from 5'Clip peak)")+ylab('Nucleotide content (U)')+theme_classic()+ggtitle(paste0('window size=',ch,' - upstream 100nt'))+
      geom_line(data = GRandom_U,color='grey',alpha=0.05)+
      geom_line(data=chtx,na.rm=F,color='purple')+  ylim(.1,.4))

chtx=cht[cht$Nucleotide%in%"G",]
print(ggplot(chtx,aes(y=Percent,x=distance,line=Nucleotide,color=Nucleotide))+
                 xlab("Position of Window\n(Distance from 5'Clip peak)")+ylab('Nucleotide content (G)')+theme_classic()+ggtitle(paste0('window size=',ch,' - upstream 100nt'))+
      geom_line(data = GRandom_G,color='grey',alpha=0.05)+
      geom_line(data=chtx,na.rm=F,color='blue')+  ylim(.1,.4))

chtx=cht[cht$Nucleotide%in%"C",]
print(ggplot(chtx,aes(y=Percent,x=distance,line=Nucleotide,color=Nucleotide))+
                 xlab("Position of Window\n(Distance from 5'Clip peak)")+ylab('Nucleotide content (C)')+theme_classic()+ggtitle(paste0('window size=',ch,' - upstream 100nt'))+
      geom_line(data = GRandom_C,color='grey',alpha=0.05)+
      geom_line(data=chtx,na.rm=F,color='green')+  ylim(.1,.4))
      
      
      write.table(chucnkseqA, file=paste0(outdir_UP,'/chucnkseqA_window',ch,'.txt'), quote=F, sep="\t", row.names=F, col.names=T)
      write.table(chucnkseqU, file=paste0(outdir_UP,'/chucnkseqU_window',ch,'.txt'), quote=F, sep="\t", row.names=F, col.names=T)
      write.table(chucnkseqG, file=paste0(outdir_UP,'/chucnkseqG_window',ch,'.txt'), quote=F, sep="\t", row.names=F, col.names=T)
      write.table(chucnkseqC, file=paste0(outdir_UP,'/chucnkseqC_window',ch,'.txt'), quote=F, sep="\t", row.names=F, col.names=T)
    
      remove(chucnkseqA,chucnkseqU,chucnkseqG,chucnkseqC)  
      
      
    }
```

### Downstream   
     
```{r ,echo=F,include=T, fig.align='center',fig.height=2, fig.width=2, figures-side, fig.show="hold"}
##```{r , include=T,fig.asp=1.25, fig.align='center',fig.height=4, fig.width=3}
    
    # downstream   

    for (ch in c(10,20,30,40,50)) {
      chadj=ch-1
    
    chucnkseq_dnA=as.data.frame(matrix(nrow=nrow(checkout2_dn),ncol=(nchar(checkout2_dn[1,'sequence'])-chadj)))
    rownames(chucnkseq_dnA)=checkout2_dn$ID;colnames(chucnkseq_dnA)=seq(1,(nchar(checkout2_dn[1,'sequence'])-chadj))
    chucnkseq_dnU=as.data.frame(matrix(nrow=nrow(checkout2_dn),ncol=(nchar(checkout2_dn[1,'sequence'])-chadj)))
    rownames(chucnkseq_dnU)=checkout2_dn$ID;colnames(chucnkseq_dnU)=seq(1,(nchar(checkout2_dn[1,'sequence'])-chadj))
    chucnkseq_dnG=as.data.frame(matrix(nrow=nrow(checkout2_dn),ncol=(nchar(checkout2_dn[1,'sequence'])-chadj)))
    rownames(chucnkseq_dnG)=checkout2_dn$ID;colnames(chucnkseq_dnG)=seq(1,(nchar(checkout2_dn[1,'sequence'])-chadj))
    chucnkseq_dnC=as.data.frame(matrix(nrow=nrow(checkout2_dn),ncol=(nchar(checkout2_dn[1,'sequence'])-chadj)))
    rownames(chucnkseq_dnC)=checkout2_dn$ID;colnames(chucnkseq_dnC)=seq(1,(nchar(checkout2_dn[1,'sequence'])-chadj))
    
    for (l in 1:(nchar(checkout2_dn[1,'sequence'])-chadj)) {
      ss= substr(checkout2_dn[,'sequence'], start = l, stop = (l+chadj))
      chucnkseq_dnA[,l]=str_count(ss, "A") /ch
      chucnkseq_dnU[,l]=str_count(ss, "U") /ch
      chucnkseq_dnG[,l]=str_count(ss, "G") /ch
      chucnkseq_dnC[,l]=str_count(ss, "C") /ch
    }

   chucnkseq_dnG=chucnkseq_dnG[order(rowMaxs(as.matrix(chucnkseq_dnG)),decreasing = T),]
        


      # NucHeatPlot(chucnkseq_dnA,checkout2_dn,chucnkseq_dnG,"A",'downstream')
      # NucHeatPlot(chucnkseq_dnU,checkout2_dn,chucnkseq_dnG,"U",'downstream')
      # NucHeatPlot(chucnkseq_dnG,checkout2_dn,chucnkseq_dnG,"G",'downstream')
      # NucHeatPlot(chucnkseq_dnC,checkout2_dn,chucnkseq_dnG,"C",'downstream')


      cht=cbind((colMeans(chucnkseq_dnA)),(colMeans(chucnkseq_dnU)),(colMeans(chucnkseq_dnC)),(colMeans(chucnkseq_dnG)))
colnames(cht)=c('A','U','C','G')
cht=melt(cht)
colnames(cht)=c('distance','Nucleotide','Percent')

####### plot CLIP peak data
# print(ggplot(cht,aes(y=Percent,x=distance,line=Nucleotide,color=Nucleotide))+
#                  xlab("Position of Window\n(Distance from 5'Clip peak)")+ylab('Nucleotide content ')+theme_classic()+ggtitle(paste0('window size=',ch,' - downstream 100nt'))+
#       geom_line(data=cht,na.rm=F)+  ylim(.1,.4))

########### Import Random data
    
      GRandom_A=as.matrix(read.delim(paste0(inRandom,'/Random1/',CLIPtype,'/',TrasncLoc,'/',RunMode,'/downstream/chucnkseqA_window',ch,'.txt'),comment.char = "%",header = T,sep = "\t",stringsAsFactors = F))
                        colnames(GRandom_A)=seq(1,(nchar(checkout2_dn[1,'sequence'])-chadj))

      GRandom_U=as.matrix(read.delim(paste0(inRandom,'/Random1/',CLIPtype,'/',TrasncLoc,'/',RunMode,'/downstream/chucnkseqU_window',ch,'.txt'),comment.char = "%",header = T,sep = "\t",stringsAsFactors = F))
                              colnames(GRandom_U)=seq(1,(nchar(checkout2_dn[1,'sequence'])-chadj))  
                              
      GRandom_G=as.matrix(read.delim(paste0(inRandom,'/Random1/',CLIPtype,'/',TrasncLoc,'/',RunMode,'/downstream/chucnkseqG_window',ch,'.txt'),comment.char = "%",header = T,sep = "\t",stringsAsFactors = F))
                              colnames(GRandom_G)=seq(1,(nchar(checkout2_dn[1,'sequence'])-chadj)) 
      
       GRandom_C=as.matrix(read.delim(paste0(inRandom,'/Random1/',CLIPtype,'/',TrasncLoc,'/',RunMode,'/downstream/chucnkseqC_window',ch,'.txt'),comment.char = "%",header = T,sep = "\t",stringsAsFactors = F))
                              colnames(GRandom_C)=seq(1,(nchar(checkout2_dn[1,'sequence'])-chadj))                                                
                              
for (rand in c(2:12)) {
  GRandom2_Ax=as.matrix(read.delim(paste0(inRandom,'/Random',rand,'/',CLIPtype,'/',TrasncLoc,'/',RunMode,'/downstream/chucnkseqA_window',ch,'.txt'),comment.char = "%",header = T,sep = "\t",stringsAsFactors = F))
                        colnames(GRandom2_Ax)=seq(1,(nchar(checkout2_dn[1,'sequence'])-chadj))
                         GRandom_A=rbind(GRandom_A,GRandom2_Ax)
      
  GRandom2_Ux=as.matrix(read.delim(paste0(inRandom,'/Random',rand,'/',CLIPtype,'/',TrasncLoc,'/',RunMode,'/downstream/chucnkseqU_window',ch,'.txt'),comment.char = "%",header = T,sep = "\t",stringsAsFactors = F))
                        colnames(GRandom2_Ux)=seq(1,(nchar(checkout2_dn[1,'sequence'])-chadj))
                         GRandom_U=rbind(GRandom_U,GRandom2_Ux)
                         
  GRandom2_Gx=as.matrix(read.delim(paste0(inRandom,'/Random',rand,'/',CLIPtype,'/',TrasncLoc,'/',RunMode,'/downstream/chucnkseqG_window',ch,'.txt'),comment.char = "%",header = T,sep = "\t",stringsAsFactors = F))
                        colnames(GRandom2_Gx)=seq(1,(nchar(checkout2_dn[1,'sequence'])-chadj))
                         GRandom_G=rbind(GRandom_G,GRandom2_Gx)
                         
  GRandom2_Cx=as.matrix(read.delim(paste0(inRandom,'/Random',rand,'/',CLIPtype,'/',TrasncLoc,'/',RunMode,'/downstream/chucnkseqC_window',ch,'.txt'),comment.char = "%",header = T,sep = "\t",stringsAsFactors = F))
                        colnames(GRandom2_Cx)=seq(1,(nchar(checkout2_dn[1,'sequence'])-chadj))
                         GRandom_C=rbind(GRandom_C,GRandom2_Cx)
  
}
      
                colnames(GRandom_A)=seq(1,(nchar(checkout2_dn[1,'sequence'])-chadj))
            GRandom_A_avg=colMeans(GRandom_A)    
                GRandom_A= melt(t(GRandom_A))
                GRandom_A$X2=as.character(GRandom_A$X2)
                colnames(GRandom_A)=c('distance','Nucleotide','Percent')

                colnames(GRandom_U)=seq(1,(nchar(checkout2_dn[1,'sequence'])-chadj))
            GRandom_U_avg=colMeans(GRandom_U)    
                GRandom_U= melt(t(GRandom_U))
                GRandom_U$X2=as.character(GRandom_U$X2)
                colnames(GRandom_U)=c('distance','Nucleotide','Percent')
      
                colnames(GRandom_G)=seq(1,(nchar(checkout2_dn[1,'sequence'])-chadj))
            GRandom_G_avg=colMeans(GRandom_G)    
                GRandom_G= melt(t(GRandom_G))
                GRandom_G$X2=as.character(GRandom_G$X2)
                colnames(GRandom_G)=c('distance','Nucleotide','Percent')
                
                colnames(GRandom_C)=seq(1,(nchar(checkout2_dn[1,'sequence'])-chadj))
            GRandom_C_avg=colMeans(GRandom_C)    
                GRandom_C= melt(t(GRandom_C))
                GRandom_C$X2=as.character(GRandom_C$X2) 
                colnames(GRandom_C)=c('distance','Nucleotide','Percent')

                                
 Rht=cbind(((GRandom_A_avg)),((GRandom_U_avg)),((GRandom_C_avg)),((GRandom_G_avg)))
colnames(Rht)=c('A','U','C','G')
Rht=melt(Rht)
colnames(Rht)=c('distance','Nucleotide','Percent')

####### plot Random Averages
# print(ggplot(Rht,aes(y=Percent,x=distance,line=Nucleotide,color=Nucleotide))+
#                  xlab("Position of Window\n(Distance from 5'Clip peak)")+ylab('Nucleotide content')+theme_classic()+ggtitle(paste0('Random Sequenc:window size=',ch,' - downstream 100nt'))+
#       geom_line(na.rm=F))


####### plot all nuc with all random
# print(ggplot(cht,aes(y=Percent,x=distance,line=Nucleotide,color=Nucleotide))+
#                  xlab("Position of Window\n(Distance from 5'Clip peak)")+ylab('Nucleotide content')+theme_classic()+ggtitle(paste0('window size=',ch,' - downstream 100nt'))+
#       geom_line(data = GRandom_A,color='red',alpha=0.05)+
#       geom_line(data = GRandom_U,color='purple',alpha=0.05)+
#       geom_line(data = GRandom_G,color='blue',alpha=0.05)+
#       geom_line(data = GRandom_C,color='green',alpha=0.05)+
#       geom_line(data=cht,na.rm=F))


### plot CLIP avg with Random AVG
print(ggplot(cht,aes(y=Percent,x=distance,line=Nucleotide,color=Nucleotide))+
                 xlab("Position of Window\n(Distance from 5'Clip peak)")+ylab('Nucleotide content ')+theme_classic()+ggtitle(paste0('window size=',ch,' - downstream 100nt'))+
      geom_line(data=cht,na.rm=F)+
        geom_line(data=Rht,na.rm=F,alpha=0.25,size=2)+  ylim(.1,.4))

################# PLOT INDIVIDUAL Nuc with Random backgound
chtx=cht[cht$Nucleotide%in%"A",]
print(ggplot(chtx,aes(y=Percent,x=distance,line=Nucleotide,color=Nucleotide))+
                 xlab("Position of Window\n(Distance from 5'Clip peak)")+ylab('Nucleotide content (A)')+theme_classic()+ggtitle(paste0('window size=',ch,' - downstream 100nt'))+
      geom_line(data = GRandom_A,color='grey',alpha=0.05)+
      geom_line(data=chtx,na.rm=F,color='red')+  ylim(.1,.4))

chtx=cht[cht$Nucleotide%in%"U",]
print(ggplot(chtx,aes(y=Percent,x=distance,line=Nucleotide,color=Nucleotide))+
                 xlab("Position of Window\n(Distance from 5'Clip peak)")+ylab('Nucleotide content (U)')+theme_classic()+ggtitle(paste0('window size=',ch,' - downstream 100nt'))+
      geom_line(data = GRandom_U,color='grey',alpha=0.05)+
      geom_line(data=chtx,na.rm=F,color='purple')+  ylim(.1,.4))

chtx=cht[cht$Nucleotide%in%"G",]
print(ggplot(chtx,aes(y=Percent,x=distance,line=Nucleotide,color=Nucleotide))+
                 xlab("Position of Window\n(Distance from 5'Clip peak)")+ylab('Nucleotide content (G)')+theme_classic()+ggtitle(paste0('window size=',ch,' - downstream 100nt'))+
      geom_line(data = GRandom_G,color='grey',alpha=0.05)+
      geom_line(data=chtx,na.rm=F,color='blue')+  ylim(.1,.4))

chtx=cht[cht$Nucleotide%in%"C",]
print(ggplot(chtx,aes(y=Percent,x=distance,line=Nucleotide,color=Nucleotide))+
                 xlab("Position of Window\n(Distance from 5'Clip peak)")+ylab('Nucleotide content (C)')+theme_classic()+ggtitle(paste0('window size=',ch,' - downstream 100nt'))+
      geom_line(data = GRandom_C,color='grey',alpha=0.05)+
      geom_line(data=chtx,na.rm=F,color='green')+  ylim(.1,.4))
      
    
    write.table(chucnkseq_dnA, file=paste0(outdir_DN,'/chucnkseq_dnA_window',ch,'.txt'), quote=F, sep="\t", row.names=T, col.names=T)
    write.table(chucnkseq_dnU, file=paste0(outdir_DN,'/chucnkseq_dnU_window',ch,'.txt'), quote=F, sep="\t", row.names=T, col.names=T)
    write.table(chucnkseq_dnG, file=paste0(outdir_DN,'/chucnkseq_dnG_window',ch,'.txt'), quote=F, sep="\t", row.names=T, col.names=T)
    write.table(chucnkseq_dnC, file=paste0(outdir_DN,'/chucnkseq_dnC_window',ch,'.txt'), quote=F, sep="\t", row.names=T, col.names=T)
    
    
    
    
    
    }
```
  
## Nucleotide Content Heatmap   

Heat map is clustered by %G content (first plot)  
All heat maps are in same row order  
Line plots show average nucleotide content across upstream/downstream region for each cluster  
First heat plot is annotated with G-quadruplex score   

  
```{r ,echo=F,fig.asp=.75, include=F, fig.align='center'}
# ```{r ,fig.asp=.5, include=T, fig.align='center',fig.height=2, fig.width=2}
# ```{r , include=T,fig.asp=1.25, fig.align='center',fig.height=4, fig.width=3}

    
    # for (ch in c(10,20,30,40,50)) { 
    ch =20
    chadj=ch-1

# upstream   

    chucnkseqA=as.data.frame(matrix(nrow=nrow(checkout2),ncol=(nchar(checkout2[1,'sequence'])-chadj)))
    rownames(chucnkseqA)=checkout2$ID;colnames(chucnkseqA)=seq(-(nchar(checkout2[1,'sequence'])-chadj),-1)
    chucnkseqU=as.data.frame(matrix(nrow=nrow(checkout2),ncol=(nchar(checkout2[1,'sequence'])-chadj)))
    rownames(chucnkseqU)=checkout2$ID;colnames(chucnkseqU)=seq(-(nchar(checkout2[1,'sequence'])-chadj),-1)
    chucnkseqG=as.data.frame(matrix(nrow=nrow(checkout2),ncol=(nchar(checkout2[1,'sequence'])-chadj)))
    rownames(chucnkseqG)=checkout2$ID;colnames(chucnkseqG)=seq(-(nchar(checkout2[1,'sequence'])-chadj),-1)
    chucnkseqC=as.data.frame(matrix(nrow=nrow(checkout2),ncol=(nchar(checkout2[1,'sequence'])-chadj)))
    rownames(chucnkseqC)=checkout2$ID;colnames(chucnkseqC)=seq(-(nchar(checkout2[1,'sequence'])-chadj),-1)
    
    for (l in 1:(nchar(checkout2[1,'sequence'])-chadj)) {
      ss= substr(checkout2[,'sequence'], start = l, stop = (l+chadj))
      chucnkseqA[,l]=str_count(ss, "A") /ch
      chucnkseqU[,l]=str_count(ss, "U") /ch
      chucnkseqG[,l]=str_count(ss, "G") /ch
      chucnkseqC[,l]=str_count(ss, "C") /ch
    }
    
        chucnkseqG=chucnkseqG[order(rowMaxs(as.matrix(chucnkseqG)),decreasing = T),]



 # downstream 
    
    chucnkseq_dnA=as.data.frame(matrix(nrow=nrow(checkout2_dn),ncol=(nchar(checkout2_dn[1,'sequence'])-chadj)))
    rownames(chucnkseq_dnA)=checkout2_dn$ID;colnames(chucnkseq_dnA)=seq(1,(nchar(checkout2_dn[1,'sequence'])-chadj))
    chucnkseq_dnU=as.data.frame(matrix(nrow=nrow(checkout2_dn),ncol=(nchar(checkout2_dn[1,'sequence'])-chadj)))
    rownames(chucnkseq_dnU)=checkout2_dn$ID;colnames(chucnkseq_dnU)=seq(1,(nchar(checkout2_dn[1,'sequence'])-chadj))
    chucnkseq_dnG=as.data.frame(matrix(nrow=nrow(checkout2_dn),ncol=(nchar(checkout2_dn[1,'sequence'])-chadj)))
    rownames(chucnkseq_dnG)=checkout2_dn$ID;colnames(chucnkseq_dnG)=seq(1,(nchar(checkout2_dn[1,'sequence'])-chadj))
    chucnkseq_dnC=as.data.frame(matrix(nrow=nrow(checkout2_dn),ncol=(nchar(checkout2_dn[1,'sequence'])-chadj)))
    rownames(chucnkseq_dnC)=checkout2_dn$ID;colnames(chucnkseq_dnC)=seq(1,(nchar(checkout2_dn[1,'sequence'])-chadj))
    
    for (l in 1:(nchar(checkout2_dn[1,'sequence'])-chadj)) {
      ss= substr(checkout2_dn[,'sequence'], start = l, stop = (l+chadj))
      chucnkseq_dnA[,l]=str_count(ss, "A") /ch
      chucnkseq_dnU[,l]=str_count(ss, "U") /ch
      chucnkseq_dnG[,l]=str_count(ss, "G") /ch
      chucnkseq_dnC[,l]=str_count(ss, "C") /ch
    }

   chucnkseq_dnG=chucnkseq_dnG[order(rowMaxs(as.matrix(chucnkseq_dnG)),decreasing = T),]
  
   #######################################################   #######################################################   #######################################################   #######################################################   #######################################################   #######################################################
   
             col_fun = colorRamp2(c(0, .25, .50), c("blue", "white", "red"))

   
      
chucnkseqGp=chucnkseqG
chucnkseqCp=chucnkseqC
chucnkseqAp=chucnkseqA
chucnkseqUp=chucnkseqU

   chucnkseqGp$'0'=NA
   chucnkseqCp$'0'=NA
   chucnkseqAp$'0'=NA
   chucnkseqUp$'0'=NA
   
   
  ##################### 

  chucnkseqG_UPDN=merge(chucnkseqGp,chucnkseq_dnG,by='row.names',reorder=F)
  rownames(chucnkseqG_UPDN)=chucnkseqG_UPDN$Row.names
  chucnkseqG_UPDN=chucnkseqG_UPDN[,-1]
  chucnkseqG_UPDN=chucnkseqG_UPDN[rownames(chucnkseqG[order(rowMaxs(as.matrix(chucnkseqG)),decreasing = T),]),]
     chucnkseqG_UPDNp=chucnkseqG_UPDN
     Gname=as.matrix(colnames(chucnkseqG_UPDNp))
     Gname[,1]=NA
     Gname[which(colnames(chucnkseqG_UPDNp)%in%colnames(chucnkseqG_UPDNp)[seq(1,ncol(chucnkseqG_UPDN),15)]),1]=colnames(chucnkseqG_UPDNp)[seq(1,ncol(chucnkseqG_UPDN),15)]
     Gname=as.numeric(Gname)
     Gname[which(is.na(Gname))]=""

     nucmap=chucnkseqG_UPDN
      GscrAnno_UP=checkout2
      
      rownames(GscrAnno_UP)=GscrAnno_UP$ID
      GscrAnno_UP=GscrAnno_UP[rownames(nucmap),'qgrs_maxScore',drop=F]
      GscrAnno_UP$qgrs_maxScore=-as.numeric(GscrAnno_UP$qgrs_maxScore)
      GscrAnno_UP=rowAnnotation(Gquad_score = row_anno_barplot(as.numeric(GscrAnno_UP$qgrs_maxScore)))

      GscrAnno_DN=checkout2_dn
      rownames(GscrAnno_DN)=GscrAnno_DN$ID
      GscrAnno_DN=GscrAnno_DN[rownames(nucmap),'qgrs_maxScore',drop=F]
      GscrAnno_DN$qgrs_maxScore=as.numeric(GscrAnno_DN$qgrs_maxScore)
      GscrAnno_DN=rowAnnotation(Gquad_score = row_anno_barplot(as.numeric(GscrAnno_DN$qgrs_maxScore)))
        

chucnkseqC_UPDN=merge(chucnkseqCp,chucnkseq_dnC,by='row.names',reorder=F)
  rownames(chucnkseqC_UPDN)=chucnkseqC_UPDN$Row.names
  chucnkseqC_UPDN=chucnkseqC_UPDN[,-1]
     chucnkseqC_UPDN=chucnkseqC_UPDN[rownames(chucnkseqG_UPDN),]

  chucnkseqA_UPDN=merge(chucnkseqAp,chucnkseq_dnA,by='row.names',reorder=F)
  rownames(chucnkseqA_UPDN)=chucnkseqA_UPDN$Row.names
  chucnkseqA_UPDN=chucnkseqA_UPDN[,-1]
     chucnkseqA_UPDN=chucnkseqA_UPDN[rownames(chucnkseqG_UPDN),]

  chucnkseqU_UPDN=merge(chucnkseqUp,chucnkseq_dnU,by='row.names',reorder=F)
  rownames(chucnkseqU_UPDN)=chucnkseqU_UPDN$Row.names
  chucnkseqU_UPDN=chucnkseqU_UPDN[,-1]
     chucnkseqU_UPDN=chucnkseqU_UPDN[rownames(chucnkseqG_UPDN),]


###################  
  # ComplexHeatmap::Heatmap(as.matrix(chucnkseqG_UPDN),col = col_fun,name = "mat", 
  #                                 column_title = "Distance from CLIP peak",column_title_side = "bottom",
  #                                 row_title = "Exonic CLIP peaks",row_title_side = "left",show_row_names=F,
  #                                 cluster_columns = F,cluster_rows =F,left_annotation = GscrAnno_UP,right_annotation = GscrAnno_DN )
  # 
  # ComplexHeatmap::Heatmap(as.matrix(chucnkseqG_UPDN),col = col_fun,name = "mat", 
  #                                 column_title = "Distance from CLIP peak",column_title_side = "bottom",
  #                                 row_title = "Exonic CLIP peaks",row_title_side = "left",show_row_names=F,
  #                                 cluster_columns = F,cluster_rows =T,
  #                         left_annotation = GscrAnno_UP,right_annotation = GscrAnno_DN )
  # 
  # { p=ComplexHeatmap::Heatmap(as.matrix(chucnkseqG_UPDN),col = col_fun,name = "mat", 
  #                                 column_title = "Distance from CLIP peak",column_title_side = "bottom",
  #                                 row_title = "Exonic CLIP peaks",row_title_side = "left",show_row_names=F,
  #                                 cluster_columns = F,cluster_rows =F,row_km = 3,
  #                         left_annotation = GscrAnno_UP,right_annotation = GscrAnno_DN )
  #       draw(p,column_title= paste0('% G in ',ch,'nt sliding window - ',region,' 100nt'),column_title_side = "top")
  # 
  # op= row_order(p)
  #  
  #  for (clst in 1:length(op)) {
  #    opavg=colMeans(chucnkseqG_UPDN[op[[clst]],])
  #    opavg=melt(opavg)
  #    opavg$distance=as.numeric(rownames(opavg))
  #    print(ggplot(opavg,aes(y=value,x=distance))+
  #                xlab("Position of Window\n(Distance from 5'Clip peak)")+ylab('Nucleotide content ')+theme_classic()+ggtitle(paste0('window size=',ch,' - upstream/downstream 100nt'))+
  #     geom_line(na.rm=F)+ylim(.1,.6))
  #  }
  #          }   
  #         
  #   }   
  
  # { pG=ComplexHeatmap::Heatmap(as.matrix(chucnkseqG_UPDN),col = col_fun,name = "matG", 
  #                                 column_title = "Distance from CLIP peak",column_title_side = "bottom",
  #                                 row_title = "Exonic CLIP peaks",row_title_side = "left",show_row_names=F,
  #                                 cluster_columns = F,cluster_rows =T,row_split = 3,row_gap =unit(3,"mm"),
  #                         left_annotation = GscrAnno_UP,right_annotation = GscrAnno_DN, 
  #                         width = unit(2, "cm"),heatmap_height = unit(5, "cm")
  #                         )
  #          draw(p,column_title= paste0('% G in ',ch,'nt sliding window - ',region,' 100nt'),column_title_side = "top")
  # 
  #   op= row_order(p)
  #  
  #  for (clst in 1:length(op)) {
  #    opavg=colMeans(chucnkseqG_UPDN[op[[clst]],])
  #    opavg=melt(opavg)
  #    opavg$distance=as.numeric(rownames(opavg))
  #    print(ggplot(opavg,aes(y=value,x=distance))+
  #                xlab("Position of Window\n(Distance from 5'Clip peak)")+ylab('Nucleotide content ')+theme_classic()+ggtitle(paste0(clst,' window size=',ch,' - upstream/downstream 100nt'))+
  #     geom_line(na.rm=F)+ylim(.1,.6) 
  #     )
  #  }
  # 
  #          }   

h=20;w=2

{ pG=ComplexHeatmap::Heatmap(as.matrix(chucnkseqG_UPDN),col = col_fun,name = "matG", 
                                  column_title = paste0("% G content in \n",ch,"nt window"),column_title_side = "top",    heatmap_legend_param = list(title = "Nucleotide\ncontent (%)"),

                                  row_title = paste0(CLIPtype," Clip peaks\nNucleotide Content"),row_title_side = "left",
                             show_row_names=F,  row_names_side ="left", row_names_gp = gpar(fontsize = 5),clustering_method_rows='complete',
                                  cluster_columns = F,cluster_rows =T,row_split = 4,row_gap =unit(3,"mm"),
                              column_labels = Gname,
                          left_annotation = GscrAnno_UP,right_annotation = GscrAnno_DN, 
                          width = unit(w, "in"),heatmap_height = unit(h, "cm"), column_names_gp = gpar(fontsize = 8)
                          )

op= row_order(pG)
opn=rownames(chucnkseqG_UPDN)[unlist(op)]
  pC=ComplexHeatmap::Heatmap(as.matrix(chucnkseqC_UPDN[,]),col = col_fun,name = "matC", 
                                  column_title = paste0("% C content in \n",ch,"nt window"),column_title_side = "top",show_heatmap_legend = F,
                                  row_title = paste0(CLIPtype," Clip peaks\nNucleotide Content"),row_title_side = "left",show_row_names=F,
                                  cluster_columns = F,cluster_rows =F,
                              column_labels = Gname,
                             # row_order=opn,
                          width = unit(w, "in"),heatmap_height = unit(h, "cm"), column_names_gp = gpar(fontsize = 8)#,
                          )

    pA=ComplexHeatmap::Heatmap(as.matrix(chucnkseqA_UPDN[,]),col = col_fun,name = "matA", 
                                  column_title = paste0("% A content in \n",ch,"nt window"),column_title_side = "top",show_heatmap_legend = F,
                                  row_title = paste0(CLIPtype," Clip peaks\nNucleotide Content"),row_title_side = "left",show_row_names=F,
                                  cluster_columns = F,cluster_rows =F,
                              column_labels = Gname,
                              # row_order=opn,
                          width = unit(w, "in"),heatmap_height = unit(h, "cm"), column_names_gp = gpar(fontsize = 8)#,

                                                    )
  pU=ComplexHeatmap::Heatmap(as.matrix(chucnkseqU_UPDN[,]),col = col_fun,name = "matU", 
                                  column_title = paste0("% U content in \n",ch,"nt window"),column_title_side = "top",show_heatmap_legend = F,
                                  row_title = paste0(CLIPtype," Clip peaks\nNucleotide Content"),row_title_side = "left",
                             show_row_names=F, row_names_gp = gpar(fontsize = 5),
                                  cluster_columns = F,cluster_rows =F,
                              column_labels = Gname,
                             # row_order=opn,
                          width = unit(w, "in"),heatmap_height = unit(h, "cm"), column_names_gp = gpar(fontsize = 8)#,

                                                    )
           ht_list = pG + pC + pA + pU
}
     
```
```{r ,echo=F,fig.asp=.75, include=T, fig.align='center', figures-side, fig.show="hold"}
                draw(ht_list, ht_gap = unit(.5, "in"),main_heatmap = "matG",column_title = "Distance from CLIP peak",column_title_side="bottom")


```
```{r ,echo=F,fig.asp=.5, include=T, fig.align='center',fig.height=3, fig.width=3, figures-side, fig.show="hold"}
   

for (clst in 1:length(op)) {
  
  opn2=rownames(chucnkseqG_UPDN)[unlist(op[[clst]])]
     
     opavg=cbind(colMeans(chucnkseqA_UPDN[opn2,]),colMeans(chucnkseqU_UPDN[opn2,]),colMeans(chucnkseqC_UPDN[opn2,]),colMeans(chucnkseqG_UPDN[opn2,]))
     colnames(opavg)=c('A','U','C','G')
     opavg=melt(opavg)
     colnames(opavg)=c("distance","nucleotide","value")
     
     print(ggplot(opavg,aes(y=value,x=distance,line=nucleotide,color=nucleotide))+
                 xlab("Position of Window\n(Distance from 5'Clip peak)")+ylab('Nucleotide content ')+theme_classic()+ggtitle(paste0('Cluster ',clst,': window size=',ch,' - upstream/downstream 100nt'))+
      geom_line(na.rm=F)+ylim(.0,.6) 
      )
   }
  
           
             
```


  

```{r ,echo=F,fig.asp=.5, include=T, fig.align='center',fig.height=3, fig.width=3,eval=F}
      ## Distance of Splicejunction from CLIP start
#######################################       
#### Distance of Splicejunction from CLIP start       
#######################################       
if (CLIPtype=='Introns'){name='Intron'}
   if (CLIPtype=='Exons'){name='Exon'}

       
      GexnRando=as.matrix(read.delim(paste0(inRandom,'/Random1/',CLIPtype,'/',TrasncLoc,'/',RunMode,'/upstream/Table/Col_exon_UP.txt'),comment.char = "%",header = F,sep = "\t",stringsAsFactors = F))
            for (rand in c(2:12)) {

      GexnRando2=as.matrix(read.delim(paste0(inRandom,'/Random',rand,'/',CLIPtype,'/',TrasncLoc,'/',RunMode,'/upstream/Table/Col_exon_UP.txt'),comment.char = "%",header = F,sep = "\t",stringsAsFactors = F))
      GexnRando=as.data.frame(cbind(GexnRando,GexnRando2),stringsAsFactors = F);
            }
      GexnRando=GexnRando[is.na(GexnRando$V1)==F,]

      
######################
      GtnsxRando=as.matrix(read.delim(paste0(inRandom,'/Random1/',CLIPtype,'/',TrasncLoc,'/',RunMode,'/upstream/Table/Col_transcript_UP.txt'),comment.char = "%",header = F,sep = "\t",stringsAsFactors = F))
      
            for (rand in c(2:12)) {
      GtnsxRando2=as.matrix(read.delim(paste0(inRandom,'/Random',rand,'/',CLIPtype,'/',TrasncLoc,'/',RunMode,'/upstream/Table/Col_transcript_UP.txt'),comment.char = "%",header = F,sep = "\t",stringsAsFactors = F))
      GtnsxRando=as.data.frame(cbind(GtnsxRando,GtnsxRando2),stringsAsFactors = F);
            }
      GtnsxRando=GtnsxRando[is.na(GtnsxRando$V1)==F,]

######################
      GidRando=as.matrix(read.delim(paste0(inRandom,'/Random1/',CLIPtype,'/',TrasncLoc,'/',RunMode,'/upstream/Table/Col_ID_UP.txt'),comment.char = "%",header = F,sep = "\t",stringsAsFactors = F))
                  
      for (rand in c(2:12)) {
      GidRando2=as.matrix(read.delim(paste0(inRandom,'/Random',rand,'/',CLIPtype,'/',TrasncLoc,'/',RunMode,'/upstream/Table/Col_ID_UP.txt'),comment.char = "%",header = F,sep = "\t",stringsAsFactors = F))
      GidRando=as.data.frame(cbind(GidRando,GidRando2),stringsAsFactors = F);
                  }
      GidRando=GidRando[is.na(GidRando$V1)==F,]
######################
      
      
        exons=mm10_USE[mm10_USE$transcript_id%in%canonical$transcript,]
      
exons.GR=GRanges(seqnames=exons$seqname, ranges=IRanges(start=exons$start,end=exons$end),
strand=exons$strand,transcript_id=exons$transcript_id,exon_number=exons$exon_number)

ExonDist=as.data.frame(matrix(nrow=nrow(GidRando),ncol=ncol(GidRando)))
ExonDist_dn=as.data.frame(matrix(nrow=nrow(GidRando),ncol=ncol(GidRando)))
ExonTransc=as.data.frame(matrix(nrow=nrow(GidRando),ncol=ncol(GidRando)))
      # for (col in 1:ncol(GidRando)) {
      for (col in 1:1) {
                  id=GidRando[,col]
                  trnx=GtnsxRando[,col]
                  exn=GexnRando[,col]
                  RandInfo=as.data.frame(cbind(id,trnx,exn),stringsAsFactors = F)
                      RandInfo$exn=as.numeric(RandInfo$exn)
                  RandInfo=separate(RandInfo,id,into = c('start','strand'),sep="_",remove = F)    
                  RandInfo=separate(RandInfo,start,into = c('chr','start'),sep=":",remove = F)    
                      RandInfo$start=as.numeric(RandInfo$start)
              for (r in 1:length(id)) {
                    RandInfoR=RandInfo[r,]

                 RandInfoTrans= mm10_USE[mm10_USE$transcript_id%in%RandInfo[r,'trnx']&mm10_USE$exon_number%in%as.numeric(RandInfo[r,'exn']),]
                 
                if (nrow(RandInfoTrans)>1) {toomannyrows}
        
    
                  
       if (unique(RandInfoR$strand)=="+") {ExonDist[r,col]=RandInfoR$start-RandInfoTrans$start
                                    ExonDist_dn[r,col]=(RandInfoTrans$end-RandInfoR$start)}
                  
       if (unique(RandInfoR$strand)=="-") {ExonDist[r,col]=(RandInfoTrans$end-RandInfoR$start)
                                    ExonDist_dn[r,col]=(RandInfoR$start-RandInfoTrans$start)}
        
                  ExonTransc[r,col]=RandInfoR$trnx
        # peaks_Anno_mpile=cbind(qh,sh)
        # peaks_Anno_mpile=merge(peaks2,peaks_Anno_mpile,by.x="ID",by.y="ID_peaks",all.x=T)

              }
                      print(col)
      }

    # ExonDistx=ExonDist  
# ExonDist=ExonDistx
ExonDist=ExonDist[is.na(ExonDist[1,])==F,]

      R=melt(ExonDist);colnames(R)[2]="ExonDist"
      R$variable=as.character(R$variable)
      
     
       
      checkout2$upstream_exon_dist=NA
      checkout2_dn$downstream_exon_dist=NA
      for (r in 1:nrow(checkout2)) {

         D=checkout2[r,]
         M=mm10_USE[(mm10_USE$transcript_id%in%D$transcript_id)&(mm10_USE$exon_number%in%D$`exon#`),]
        colnames(M)=paste0(colnames(M),"_mm10")
       
        if (unique(D$strand)=="+") {checkout2[r,'upstream_exon_dist']=unique(((D$crosslink-M$start_mm10)))
                             checkout2_dn[r,'downstream_exon_dist']=unique(max((M$end_mm10-D$crosslink)))}
                  
       if (unique(D$strand)=="-") {checkout2[r,'upstream_exon_dist']=unique(((M$end_mm10-D$crosslink)))
                                    checkout2_dn[r,'downstream_exon_dist']=unique(((D$crosslink-M$start_mm10)))}      
      }
      
 D=checkout2 
 D=(D[,'upstream_exon_dist',drop=F])
 colnames(D)='ExonDist'
 D$variable='CLIP_Exon'
ggplot(R,aes(x=ExonDist,line=variable))+
   geom_density(data=R,size= .6,color='grey') +
  geom_density(data=D,size= .6,color='red')+
  theme_classic()+ylab("Density")+xlab("Upstream Exon junction Distance from 5' CLIP peak (nt)")+ggtitle("Distance of 5' CLIP peak (nt) from Upstream Exon Junction")+
  scale_x_continuous(trans = 'log10')
       #geom_vline(xintercept=nrow(checkout2[is.na(checkout2$qgrs_maxScore)==F,]),colour="red") +xlim(0,100)


D=checkout2_dn 
 D=(D[,'downstream_exon_dist',drop=F])
 colnames(D)='ExonDist'
 D$variable='CLIP_Exon'
ggplot(R,aes(x=ExonDist,line=variable))+
   geom_density(data=R,size= .6,color='grey') +
  geom_density(data=D,size= .6,color='red')+
  theme_classic()+ylab("Density")+xlab("Downstream Exon junction Distance from 5' CLIP peak (nt)")+ggtitle("Distance of 5' CLIP peak (nt) from Downstream Exon Junction")+
  scale_x_continuous(trans = 'log10')
       #geom_vline(xintercept=nrow(checkout2[is.na(checkout2$qgrs_maxScore)==F,]),colour="red") +xlim(0,100)

```



## CLIP peak Local Structure - Basepairing Probability  

All possible Base Pairing probabilities were calculated for the 100nt upstream/downstream region of Intronic CLIP peaks.  

"RNAfold -p -i CLIPsequences.fa > out.fold"  

The base pairing probabilities were then averaged for all possible pairings in a 10nt sliding window. 
Base paring probabilities are taken from RNAfold_out.dp.ps   
    
Gibbs free energy for each 100 nt Sequence was also calculated and taken from out.fold   



```{r ,echo=F,fig.asp=.75, include=T, fig.align='center', figures-side, fig.show="hold"}
    
      ####################################################################
      #### Upstream Base Pairing Probability
      ####################################################################

    system(paste0('rm -f ',scratch_UP,'/*.ps'),intern = F)
    system(paste0('rm -f ',scratch_UP,'/out.fold'),intern = F)
    
      # system(paste0('RNAfold -p -i ',outdir_UP,'/AllLocation_Random.fa > out.fold'),intern = F)
      system(paste0(RNAfold,' -p -i ',outdir_UP,'/AllLocation_Random.fa > out.fold'),intern = F)
      system(paste0('mv *.ps ',scratch_UP),intern = F)
      system(paste0('mv out.fold ',scratch_UP),intern = F)
      
      
      #### get deltaG
      enrg=read.delim(paste0(scratch_UP,'/out.fold'),comment.char = "%",header = F,sep = "\t",stringsAsFactors = F)
    
      for (e in 1:nrow(checkout2)) {
        s=as.character(checkout2[e,'n'])
        s=gsub("\\(-)","",s)
        s=gsub("\\(+)","",s)
        loc=grep(s,enrg$V1)
          if (isEmpty(loc)) {next ;xxx  }
      enrgx=enrg[c(loc:(loc+5)),]
      enrgx=as.data.frame(enrgx[4])
      enrgx=separate(enrgx,col=1,into = c('seq','DG'),window+2)
      enrgx$DG=gsub("]","",enrgx$DG)
      checkout2[e,'DeltaG']=as.numeric(enrgx$DG)
      }
      

      #### get BPP

      fls=list.files(paste0(scratch_UP,'/'), pattern="dp.ps", all.files=FALSE,full.names=FALSE)
      fls=paste0(scratch_UP,'/',fls)
      fls=as.list(fls)

      # xxx=gsub("\\(","\\\\(",fls)
      # xxx=str_replace_all(fls,"\\(","/\\(")
      # str_replace_all(xxx,"/","\\\\")
      

      out=mclapply(fls,RNAfold_output,mc.cores=UseCores)
      out2=out[[1]][,1:2,drop=F]
      out2_max=out[[1]][,1:2,drop=F]
      for (o in 1:length(out)) {
        if (nrow(out[[o]][,3,drop=F])==0) {
          out2$Xx=0;colnames(out2)[colnames(out2)%in%"Xx"]=colnames(out[[o]][,3,drop=F])
          out2_max$Xx=0;colnames(out2_max)[colnames(out2_max)%in%"Xx"]=colnames(out[[o]][,4,drop=F])
        }
        if (nrow(out[[o]][,3,drop=F])>0) {
          out2=cbind(out2,out[[o]][,3,drop=F])
          out2_max=cbind(out2_max,out[[o]][,4,drop=F])
        }
      }

      system(paste0('rm -f ',scratch_UP,'/*.ps'))
      system(paste0('rm -f ',scratch_UP,'/*.fa'))
      system(paste0('rm -f ',scratch_UP,'/*.bed'))
      system(paste0('rm -f ',scratch_UP,'/out.dn.fold'))

##################################################################################
      ph=as.matrix(out2[,-c(1:2)])
      ph=ph[is.na(ph[,3])==F,]
      ph[is.nan(as.matrix(ph))==T]=0

      ph=t(ph)
      ph=as.data.frame(ph)

      #### sort by row diff
      ph=ph[order(rowMaxs(as.matrix(ph)),decreasing = T),]
      colnames(ph)=c(-window:(ncol(ph)-(window+1)))
      ph= cbind(rowMaxs(as.matrix(ph)),ph);colnames(ph)[1]='diff'
      
      
      
      ph$n=rownames(ph)
      ph$n=gsub(paste0(scratch_UP,'/'),"",ph$n)
      ph$n=gsub("_",":",ph$n)
            rownames(ph)=ph$n

      checkout2=merge(checkout2,ph[,c("n","diff"),drop=F],by='n',all.x=T)
      colnames(checkout2)[colnames(checkout2)%in%'diff']='RowMax_MeanChuckBPP'
      
      
      #### #### #### #### #### #### #### #### #### #### #### #### #### #### ####
      #### #### #### #### #### #### #### #### #### #### #### #### #### #### ####

      ph_max=as.matrix(out2_max[,-c(1:2)])
      ph_max=ph_max[is.na(ph_max[,3])==F,]
      ph_max[is.nan(as.matrix(ph_max))==T]=0

      ph_max=t(ph_max)
      ph_max=as.data.frame(ph_max)
      
      
      #### sort by row diff
      ph_max=ph_max[order(rowMaxs(as.matrix(ph_max)),decreasing = T),]
      colnames(ph_max)=c(-window:(ncol(ph_max)-(window+1)))
      ph_max= cbind(rowMaxs(as.matrix(ph_max)),ph_max);colnames(ph_max)[1]='diff'

      
      ph_max$n=rownames(ph_max)
      ph_max$n=gsub(paste0(scratch_UP,'/'),"",ph_max$n)
      ph_max$n=gsub("_",":",ph_max$n)
      rownames(ph_max)=ph_max$n
      
      checkout2=merge(checkout2,ph_max[,c("n","diff"),drop=F],by='n',all.x=T)
      colnames(checkout2)[colnames(checkout2)%in%'diff']='RowMax_MaxChuckBPP'
      
      
      
  
      
      ####################################################################
      #### downstream Base Pairing Probability
      ####################################################################
      ### High base pair probability chr8:18628170-18628210
      ### low basepair probibility chr9:72581297-72581336

      system(paste0('rm -f ',scratch_DN,'/*.ps'))
      system(paste0('rm -f ',scratch_DN,'/out.dn.fold'))
      
      system(paste0(RNAfold,' -p -i ',outdir_DN,'/AllLocation_DN_Random.fa > out.dn.fold'),intern=F)
      system(paste0('mv *.ps ',scratch_DN))
      system(paste0('mv out.dn.fold ',scratch_DN),intern = F)
      
      energDN=read.delim(paste0(scratch_DN,'/out.dn.fold'),comment.char = "%",header = F,sep = "\t",stringsAsFactors = F)
       
      for (e in 1:nrow(checkout2_dn)) {
        s=as.character(checkout2_dn[e,'n'])
        s=gsub("\\(-)","",s)
        s=gsub("\\(+)","",s)
        loc=grep(s,energDN$V1)
        if (isEmpty(loc)) {next ;xxx  }
        
        energDNx=energDN[c(loc:(loc+5)),]
        energDNx=as.data.frame(energDNx[4])
        energDNx=separate(energDNx,col=1,into = c('seq','DG'),window+2)
        energDNx$DG=gsub("]","",energDNx$DG)
        checkout2_dn[e,'DeltaG']=as.numeric(energDNx$DG)
      }
    
      
      
      flsdn=list.files(paste0(scratch_DN,'/'), pattern="dp.ps", all.files=FALSE, full.names=FALSE)
      flsdn=paste0(scratch_DN,'/',flsdn)
      
      flsdn= as.list(flsdn)

      outdn=mclapply((flsdn),RNAfold_output,mc.cores=UseCores)
      outdn2=outdn[[1]][,1:2,drop=F]
      outdn2_max=outdn[[1]][,1:2,drop=F]
      for (o in 1:length(outdn)) {
        if (nrow(outdn[[o]][,3,drop=F])==0) {
          outdn2$Xx=0;colnames(outdn2)[colnames(outdn2)%in%"Xx"]=colnames(outdn[[o]][,3,drop=F])
          outdn2_max$Xx=0;colnames(outdn2_max)[colnames(outdn2_max)%in%"Xx"]=colnames(outdn[[o]][,4,drop=F])
        }
        if (nrow(outdn[[o]][,3,drop=F])>0) {
          outdn2=cbind(outdn2,outdn[[o]][,3,drop=F])
          outdn2_max=cbind(outdn2_max,outdn[[o]][,4,drop=F])
        }
      }

      system(paste0('rm -f ',scratch_DN,'/*.ps'))
      system(paste0('rm -f ',scratch_DN,'/*.fa'))
      system(paste0('rm -f ',scratch_DN,'/*dn.bed'))
      system(paste0('rm -f ',scratch_DN,'/out.dn.fold'))

      # #########################################
      phdn=as.matrix(outdn2[,-c(1:2)])
      phdn=phdn[is.na(phdn[,3])==F,]
      phdn[is.nan(as.matrix(phdn))==T]=0
      phdn=t(phdn)
      phdn=as.data.frame(phdn)

      #### sort by row diff
      phdn=phdn[order(rowMaxs(as.matrix(phdn)),decreasing = T),]
      colnames(phdn)=c(((window+1)-(ncol(phdn))):window)
      phdn= cbind(rowMaxs(as.matrix(phdn)),phdn);colnames(phdn)[1]='diff'

      
      
      phdn$n=rownames(phdn)
      phdn$n=gsub(paste0(scratch_DN,'/'),"",phdn$n)
      phdn$n=gsub("_",":",phdn$n)
                  rownames(phdn)=phdn$n

      checkout2_dn=merge(checkout2_dn,phdn[,c("n","diff"),drop=F],by='n',all.x=T)
      colnames(checkout2_dn)[colnames(checkout2_dn)%in%'diff']='RowMax_MeanChuckBPP'

      # #########################################
      # #########################################
      phdn_max=as.matrix(outdn2_max[,-c(1:2)])
      phdn_max=phdn_max[is.na(phdn_max[,3])==F,]
      phdn_max[is.nan(as.matrix(phdn_max))==T]=0
      phdn_max=t(phdn_max)
      phdn_max=as.data.frame(phdn_max)

      #### sort by row diff
      phdn_max=phdn_max[order(rowMaxs(as.matrix(phdn_max)),decreasing = T),]
      colnames(phdn_max)=c(((window+1)-(ncol(phdn_max))):window)
      phdn_max= cbind(rowMaxs(as.matrix(phdn_max)),phdn_max);colnames(phdn_max)[1]='diff'
      
     
      phdn_max$n=rownames(phdn_max)
            phdn_max$n=gsub(paste0(scratch_DN,'/'),"",phdn_max$n)
            phdn_max$n=gsub("_",":",phdn_max$n)
                        rownames(phdn_max)=phdn_max$n

            
      checkout2_dn=merge(checkout2_dn,phdn_max[,c("n","diff"),drop=F],by='n',all.x=T)
      colnames(checkout2_dn)[colnames(checkout2_dn)%in%'diff']='RowMax_MaxChuckBPP'
      

  

  rownames(outchunck_rowaverage)=c(1:nrepeat)
  # colnames(outchunck_rowaverage)=c(-window:-1)
  colnames(outchunck_rowaverage)=c(10:window)

```


### Basepairing probability Heatmap   

Upstream heatmap was sorted by maximum average Base pairing probability for each Row.  

Downstream heat map is sorted by same order as Upstream heatmap.  

Heat map is split by Upstream sequences with row max BPP `>` 0.2.  

Heatmaps are annotated with each sequences:  
DeltaG  Gibbs free energy for 100nt region  
Gquad_score- G-quad score determined 	by QGRS if G-quad is identified  
IntronDist- distance (nt) of CLIP peak 	from corresponding Upstream/Downstream Intron junction  


```{r ,echo=F,fig.asp=.75, include=T, fig.align='center', figures-side, fig.show="hold"}
  
  checkout2_HM=merge(checkout2,Peaksdata4[,c('ID','Counts_Unique','Counts_Multimappers_All','Counts_Multimappers_Scaled','Counts_Multimappers_BestMapping')],by.y="ID",by.x="CLIP_ID")
  checkout2_dn_HM=merge(checkout2_dn,Peaksdata4[,c('ID','Counts_Unique','Counts_Multimappers_All','Counts_Multimappers_Scaled','Counts_Multimappers_BestMapping')],by.y="ID",by.x="CLIP_ID")
########################################################################################################
  ################################################################################################################################################

   
hm=as.matrix(ph[,colnames(ph)%in%c('diff','n')==F])
      hmp=hm
      Gname=as.matrix(colnames(hm))
     Gname[,1]=NA
     Gname[which(colnames(hm)%in%colnames(hmp)[seq(1,ncol(hm),15)]),1]=colnames(hm)[seq(1,ncol(hm),15)]
     Gname=as.numeric(Gname)
     Gname[which(is.na(Gname))]=""
     
     ########Z
     hm=as.matrix(ph[,colnames(ph)%in%c('n')==F])

      nucmap=hm
      DGAnno_UP=checkout2_HM
      
      rownames(DGAnno_UP)=DGAnno_UP$n
      DGAnno_UP=DGAnno_UP[rownames(nucmap),c('DeltaG','CLIP_ID','qgrs_maxScore','Counts_Unique','Counts_Multimappers_Scaled','Counts_Multimappers_All'),drop=F]
      # setdiff(c('DeltaG','CLIP_ID','qgrs_maxScore','upstream_exon_dist','Counts_Unique','Counts_Multimappers_Scaled','Counts_Multimappers_All'),colnames('DGAnno_UP'))

            hm=merge(DGAnno_UP,hm,by="row.names")
            hm=hm[order(hm$diff,decreasing = T),]
            rownames(hm)=hm[,'CLIP_ID']
                  split=as.matrix(hm$diff<meanBPP)    

      DGAnno_UP=hm[,colnames(hm)%in%c('Row.names','DeltaG','CLIP_ID','upstream_exon_dist','qgrs_maxScore','diff','Counts_Unique','Counts_Multimappers_Scaled','Counts_Multimappers_All')]
      hm=as.matrix(hm[,colnames(hm)%in%c('Row.names','DeltaG','CLIP_ID','qgrs_maxScore','upstream_exon_dist','diff','Counts_Unique','Counts_Multimappers_Scaled','Counts_Multimappers_All')==F])
            
      DGAnno_UP$DeltaG=-as.numeric(DGAnno_UP$DeltaG)
      # DGAnno_UP=rowAnnotation(DeltaG = row_anno_barplot(as.numeric(DGAnno_UP$DeltaG)))
      Anno_UP=rowAnnotation( Gibbs_energy = row_anno_barplot(as.numeric(DGAnno_UP$DeltaG)),
                             Gquad_score = row_anno_barplot(as.numeric(DGAnno_UP$qgrs_maxScore)),
                             ExonDistance = row_anno_barplot(as.numeric(DGAnno_UP$upstream_exon_dist)),
                             # Counts_UniqueReads = row_anno_barplot(as.numeric(DGAnno_UP$Counts_Unique)),
                             Counts_ScaledMM = row_anno_barplot(as.numeric(DGAnno_UP$Counts_Multimappers_Scaled)),
                             # Counts_AllMM = row_anno_barplot(as.numeric(DGAnno_UP$Counts_Multimappers_All)),
                             annotation_name_rot=270)



      col_fun = colorRamp2(c(0, .2, .4), c("blue", "white", "red"))

 BPP_UP=ComplexHeatmap::Heatmap(hm,col = col_fun,name = "matBPP_UP", 
                                  column_title = "Upstream",column_title_side = "top",    heatmap_legend_param = list(title = "BasePairing\nProbability"),
                                  row_title = paste0(CLIPtype," CLIP peaks\nBasepairing Probability"),row_title_side = "left",
                                show_row_names=F,    row_names_side = "left", row_names_gp = gpar(fontsize = 8),
                                  cluster_columns = F,cluster_rows =F,
                         row_split = split,row_gap =unit(3,"mm"),
                              column_labels = Gname,
                          #left_annotation = GscrAnno_UP,
                         right_annotation = Anno_UP, 
                        cluster_row_slices = FALSE,
                          width = unit(w, "in"),heatmap_height = unit(h, "cm"), column_names_gp = gpar(fontsize = 8)
                          )  
  
 op= row_order(BPP_UP)
opn=rownames(hm)[unlist(op)]
 



hmDN=as.matrix(phdn[,colnames(phdn)%in%c('diff','n')==F])
      hmDNp=hmDN
      Gname=as.matrix(colnames(hmDN))
     Gname[,1]=NA
     Gname[which(colnames(hmDN)%in%colnames(hmDNp)[seq(1,ncol(hmDN),15)]),1]=colnames(hmDN)[seq(1,ncol(hmDN),15)]
     Gname=as.numeric(Gname)
     Gname[which(is.na(Gname))]=""
     
     ########Z
     hmDN=as.matrix(phdn[,colnames(phdn)%in%c('n')==F])
      nucmap=hmDN
      DGAnno_DN=checkout2_dn_HM
      
      rownames(DGAnno_DN)=DGAnno_DN$n
      # rownames(DGAnno_DN)=DGAnno_DN$CLIP_ID
      DGAnno_DN=DGAnno_DN[rownames(nucmap),c('DeltaG','CLIP_ID','qgrs_maxScore','Counts_Unique','Counts_Multimappers_Scaled','Counts_Multimappers_All'),drop=F]
      
            hmDN=merge(DGAnno_DN,hmDN,by="row.names")
            hmDN=hmDN[order(hmDN$diff,decreasing = T),]
            rownames(hmDN)=hmDN[,'CLIP_ID']
                  split=as.matrix(hmDN$diff<meanBPP)    

      DGAnno_DN=hmDN[,colnames(hmDN)%in%c('Row.names','DeltaG','CLIP_ID','qgrs_maxScore','downstream_exon_dist','diff','Counts_Unique','Counts_Multimappers_Scaled','Counts_Multimappers_All')]
      hmDN=as.matrix(hmDN[,colnames(hmDN)%in%c('Row.names','DeltaG','CLIP_ID','downstream_exon_dist','qgrs_maxScore','diff','Counts_Unique','Counts_Multimappers_Scaled','Counts_Multimappers_All')==F])
            
      DGAnno_DN$DeltaG=-as.numeric(DGAnno_DN$DeltaG)
      # DGAnno_DN=rowAnnotation(DeltaG = row_anno_barplot(as.numeric(DGAnno_DN$DeltaG)))
      Anno_DN=rowAnnotation(DeltaG = row_anno_barplot(as.numeric(DGAnno_DN$DeltaG)),
                            Gquad_score = row_anno_barplot(as.numeric(DGAnno_DN$qgrs_maxScore)),
                            ExonDist = row_anno_barplot(as.numeric(DGAnno_DN$downstream_exon_dist)),
                            # Counts_UniqueReads = row_anno_barplot(as.numeric(DGAnno_DN$Counts_Unique)),
                             Counts_ScaledMM = row_anno_barplot(as.numeric(DGAnno_DN$Counts_Multimappers_Scaled)),
                             # Counts_AllMM = row_anno_barplot(as.numeric(DGAnno_DN$Counts_Multimappers_All)),
                            annotation_name_rot=270)

      
     hmDN=hmDN[opn,]
  BPP_DN=ComplexHeatmap::Heatmap(hmDN,col = col_fun,name = "matBPP_DN", 
                                  column_title = "Downstream",column_title_side = "top",show_heatmap_legend = F,
                                  row_title = paste0(CLIPtype," CLIP peaks"),row_title_side = "left",show_row_names=F, row_names_gp = gpar(fontsize = 8),
                                  cluster_columns = F,cluster_rows =F,
                                   right_annotation = Anno_DN, 
                              column_labels = Gname,
                          width = unit(w, "in"),heatmap_height = unit(h, "cm"), column_names_gp = gpar(fontsize = 8)#,
                                     )
  
  ht_list = BPP_UP + BPP_DN
          
          # BPP_UP
          # BPP_DN
           # draw(ht_list, ht_gap = unit(1, "in"),main_heatmap = "matBPP_UP",column_title = "Distance from CLIP peak",column_title_side="bottom")
                
           draw(ht_list, ht_gap = unit(1, "in"),main_heatmap = "matBPP_UP",column_title = "Distance from CLIP peak",column_title_side="bottom",heatmap_height = unit((20), "cm"),heatmap_width = unit(7, "cm"))


```           
           
  
```{r ,echo=F,fig.asp=.5, include=T, fig.align='center',fig.height=3, fig.width=3, figures-side, fig.show="hold"}
   
     op= row_order(BPP_UP)

for (clst in 1:length(op)) {
     
      opn=rownames(hm)[unlist(op[[clst]])]
 
     
     opavg=cbind(colMeans(hm[opn,c(ncol(hm):1)]),colMeans(hmDN[opn,]))
     colnames(opavg)=c("Upstream","Downstream")
     opavg=melt(opavg)
     colnames(opavg)=c("distance","nucleotide","value")
     
     print(ggplot(opavg,aes(y=value,x=distance,line=nucleotide,color=nucleotide))+
                 xlab("Position of Window\n(Distance from 5'Clip peak)")+ylab('Average basepairing probabilty')+theme_classic()+ggtitle(paste0(clst,' window size=',ch,' - upstream/downstream 100nt'))+
      geom_line(na.rm=F)#+ylim(.0,.6) 
      )
   }
  
           
             
```



The first heatmap shows all CLIP locations and is split row max BPP `>` 0.2 (TOP).  
    
The second heatmap shows only the rows with row max BPP `>` 0.2 and clustered using hierarchical clustering.      
The second heatmap is scaled so that it shoud dirrectly be superimposed over the TOP unclusted rows in the first heatmap    
       
          

```{r ,echo=F,fig.asp=.75, include=T, fig.align='center', figures-side, fig.show="hold"}
robust_dist = function(x, y) {
    qx = quantile(x, c(0.1, 0.9))
    qy = quantile(y, c(0.1, 0.9))
    l = x > qx[1] & x < qx[2] & y > qy[1] & y < qy[2]
    x = x[l]
    y = y[l]
    sqrt(sum((x - y)^2))
}
  
  checkout2_HM=merge(checkout2,Peaksdata4[,c('ID','Counts_Unique','Counts_Multimappers_All','Counts_Multimappers_Scaled','Counts_Multimappers_BestMapping')],by.y="ID",by.x="CLIP_ID")
  checkout2_dn_HM=merge(checkout2_dn,Peaksdata4[,c('ID','Counts_Unique','Counts_Multimappers_All','Counts_Multimappers_Scaled','Counts_Multimappers_BestMapping')],by.y="ID",by.x="CLIP_ID")
########################################################################################################
  ################################################################################################################################################

#### Upstream Heatmap
hm=as.matrix(ph[,colnames(ph)%in%c('diff','n')==F])
      hmp=hm
      Gname=as.matrix(colnames(hm))
     Gname[,1]=NA
     Gname[which(colnames(hm)%in%colnames(hmp)[seq(1,ncol(hm),15)]),1]=colnames(hm)[seq(1,ncol(hm),15)]
     Gname=as.numeric(Gname)
     Gname[which(is.na(Gname))]=""
     
     ########Z
     hm=as.matrix(ph[,colnames(ph)%in%c('n')==F])

      nucmap=hm
      DGAnno_UP=checkout2_HM
      
      rownames(DGAnno_UP)=DGAnno_UP$n
      DGAnno_UP=DGAnno_UP[rownames(nucmap),c('DeltaG','CLIP_ID','qgrs_maxScore','Counts_Unique','Counts_Multimappers_Scaled','Counts_Multimappers_All'),drop=F]
      # setdiff(c('DeltaG','CLIP_ID','qgrs_maxScore','upstream_exon_dist','Counts_Unique','Counts_Multimappers_Scaled','Counts_Multimappers_All'),colnames('DGAnno_UP'))
      
            hm=merge(DGAnno_UP,hm,by="row.names")
            hm=hm[order(hm$diff,decreasing = T),]
            hm=hm[hm$diff>meanBPP,]
            rownames(hm)=hm[,'CLIP_ID']
                  split=as.matrix(hm$diff<meanBPP)    

      DGAnno_UP=hm[,colnames(hm)%in%c('Row.names','DeltaG','CLIP_ID','qgrs_maxScore','diff','Counts_Unique','Counts_Multimappers_Scaled','Counts_Multimappers_All')]
      hm=as.matrix(hm[,colnames(hm)%in%c('Row.names','DeltaG','CLIP_ID','qgrs_maxScore','upstream_exon_dist','diff','Counts_Unique','Counts_Multimappers_Scaled','Counts_Multimappers_All')==F])
            
      DGAnno_UP$DeltaG=-as.numeric(DGAnno_UP$DeltaG)
      # DGAnno_UP=rowAnnotation(DeltaG = row_anno_barplot(as.numeric(DGAnno_UP$DeltaG)))
      Anno_UP=rowAnnotation( Gibbs_energy = row_anno_barplot(as.numeric(DGAnno_UP$DeltaG)),
                             Gquad_score = row_anno_barplot(as.numeric(DGAnno_UP$qgrs_maxScore)),
                             ExonDistance = row_anno_barplot(as.numeric(DGAnno_UP$upstream_exon_dist)),
                             # Counts_UniqueReads = row_anno_barplot(as.numeric(DGAnno_UP$Counts_Unique)),
                             Counts_ScaledMM = row_anno_barplot(as.numeric(DGAnno_UP$Counts_Multimappers_Scaled)),
                             # Counts_AllMM = row_anno_barplot(as.numeric(DGAnno_UP$Counts_Multimappers_All)),
                             annotation_name_rot=270)

### Heatmap


      col_fun = colorRamp2(c(0, .2, .4), c("blue", "white", "red"))
      
{     xx= hclust(dist(cor(hm)),method = 'average',members=NULL)
BPP_UP= ComplexHeatmap::Heatmap(hm,col = col_fun,name = "matBPP_UP", 
                                  column_title = "Upstream",column_title_side = "top",
                                heatmap_legend_param = list(title = "BasePairing\nProbability"),
                                  # row_title = paste0(CLIPtype," CLIP peaks\nBasepairing Probability"),row_title_side = "left",
                                show_row_names=F,    row_names_side = "left", row_names_gp = gpar(fontsize = 8),
                                  cluster_columns = F,
                                # cluster_rows =T,
                                # clustering_distance_rows = "pearson",
                                clustering_distance_rows = function(x, y) 1 - cor(x, y,method = 'pearson'),
                                # clustering_distance_rows = robust_dist,
                                # cluster_rows = xx,
                                 row_split = 4,row_gap =unit(3,"mm"),
                              column_labels = F,
                              show_column_names = F,
                          #left_annotation = GscrAnno_UP,
                         right_annotation = Anno_UP, 
                        # cluster_row_slices = FALSE,
                          width = unit(w, "in"),heatmap_height = unit(h, "cm"), column_names_gp = gpar(fontsize = 8)
                          )  
  
      }    
 op= row_order(BPP_UP)
opn=rownames(hm)[unlist(op)]



####################################################
#### Downstream Heatmap
############################################################
hmDN=as.matrix(phdn[,colnames(phdn)%in%c('diff','n')==F])
      hmDNp=hmDN
      Gname=as.matrix(colnames(hmDN))
     Gname[,1]=NA
     Gname[which(colnames(hmDN)%in%colnames(hmDNp)[seq(1,ncol(hmDN),15)]),1]=colnames(hmDN)[seq(1,ncol(hmDN),15)]
     Gname=as.numeric(Gname)
     Gname[which(is.na(Gname))]=""
     
     ########
     hmDN=as.matrix(phdn[,colnames(phdn)%in%c('n')==F])
      nucmap=hmDN
      DGAnno_DN=checkout2_dn_HM
      
      rownames(DGAnno_DN)=DGAnno_DN$n
      # rownames(DGAnno_DN)=DGAnno_DN$CLIP_ID
      DGAnno_DN=DGAnno_DN[rownames(nucmap),c('DeltaG','CLIP_ID','qgrs_maxScore','Counts_Unique','Counts_Multimappers_Scaled','Counts_Multimappers_All'),drop=F]
      
            hmDN=merge(DGAnno_DN,hmDN,by="row.names")
            hmDN=hmDN[order(hmDN$diff,decreasing = T),]
            hmDN=hmDN[hmDN$CLIP_ID%in%rownames(hm),]
            rownames(hmDN)=hmDN[,'CLIP_ID']
                  split=as.matrix(hmDN$diff<.2)

      DGAnno_DN=hmDN[,colnames(hmDN)%in%c('Row.names','DeltaG','CLIP_ID','qgrs_maxScore','diff','Counts_Unique','Counts_Multimappers_Scaled','Counts_Multimappers_All')]
      hmDN=as.matrix(hmDN[,colnames(hmDN)%in%c('Row.names','DeltaG','CLIP_ID','qgrs_maxScore','diff','Counts_Unique','Counts_Multimappers_Scaled','Counts_Multimappers_All')==F])
            
      DGAnno_DN$DeltaG=-as.numeric(DGAnno_DN$DeltaG)
      # DGAnno_DN=rowAnnotation(DeltaG = row_anno_barplot(as.numeric(DGAnno_DN$DeltaG)))
      Anno_DN=rowAnnotation(DeltaG = row_anno_barplot(as.numeric(DGAnno_DN$DeltaG)),
                            Gquad_score = row_anno_barplot(as.numeric(DGAnno_DN$qgrs_maxScore)),
                            ExonDist = row_anno_barplot(as.numeric(DGAnno_DN$downstream_exon_dist)),
                            # Counts_UniqueReads = row_anno_barplot(as.numeric(DGAnno_DN$Counts_Unique)),
                             Counts_ScaledMM = row_anno_barplot(as.numeric(DGAnno_DN$Counts_Multimappers_Scaled)),
                             # Counts_AllMM = row_anno_barplot(as.numeric(DGAnno_DN$Counts_Multimappers_All)),
                            annotation_name_rot=270)




### Heatmap
      
     hmDN=hmDN[rownames(hm),]
  BPP_DN=ComplexHeatmap::Heatmap(hmDN,col = col_fun,name = "matBPP_DN", 
                                  column_title = "Downstream",column_title_side = "top",
                                 show_heatmap_legend = F,
                                  # row_title = paste0(CLIPtype," CLIP peaks"),row_title_side = "left",
                                 show_row_names=F, row_names_gp = gpar(fontsize = 8),
                                  cluster_columns = F,cluster_rows =F,
                                   right_annotation = Anno_DN, 
                              column_labels = Gname,
                          width = unit(w, "in"),heatmap_height = unit(h, "cm"), column_names_gp = gpar(fontsize = 8)#,
                                     )

  ht_list = BPP_UP + BPP_DN
          
          # BPP_UP
          # BPP_DN
           draw(ht_list, ht_gap = unit(1, "in"),main_heatmap = "matBPP_UP",column_title = "Distance from CLIP peak",column_title_side="bottom")
           # 
           # draw(ht_list[1:length(split[split[,]==F,])], ht_gap = unit(1, "in"),main_heatmap = "matBPP_UP",column_title = "Distance from CLIP peak",column_title_side="bottom",heatmap_height = unit(30, "cm"),heatmap_width = unit(7, "cm"))
           
           draw(ht_list, ht_gap = unit(1, "in"),main_heatmap = "matBPP_UP",column_title_side="bottom",heatmap_height = unit((20*(nrow(ph[ph$diff>meanBPP,])/nrow(ph))), "cm"),heatmap_width = unit(7, "cm"))

                
  # Heatmap(ht_list[1:length(split[split[,]==F,])])
           
```           
           
  
```{r ,echo=F,fig.asp=.5, include=T, fig.align='center',fig.height=4, fig.width=4, figures-side, fig.show="hold"}
   
     op= row_order(BPP_UP)

for (clst in 1:length(op)) {
     
      opn=rownames(hm)[unlist(op[[clst]])]
 
     
     opavg=cbind(colMeans(hm[opn,c(ncol(hm):1)]),colMeans(hmDN[opn,]))
     colnames(opavg)=c("Upstream","Downstream")
     opavg=melt(opavg)
     colnames(opavg)=c("distance","nucleotide","value")
     
     print(ggplot(opavg,aes(y=value,x=distance,line=nucleotide,color=nucleotide))+
                 xlab("Position of Window\n(Distance from 5'Clip peak)")+ylab('Average basepairing probabilty')+theme_classic()+ggtitle(paste0('Cluster ',clst,': window size=',ch,' - upstream/downstream 100nt'))+
      geom_line(na.rm=F)#+ylim(.0,.6) 
      )
   }
  
           
             
```




## Local Structure compared to Random   
      

```{r ,echo=F,fig.asp=.5, include=T, fig.align='center',fig.height=4, fig.width=4, figures-side, fig.show="hold"}

      BPPfracRando=as.matrix(read.delim(paste0(inRandom,'/Random1/',CLIPtype,'/',TrasncLoc,'/',RunMode,'/upstream/Random_outchunck_frac_100chunk4.txt'),comment.char = "%",header = F,sep = "\t",stringsAsFactors = F))
            for (rand in c(2:12)) {

      BPPfracRando2=as.matrix(read.delim(paste0(inRandom,'/Random',rand,'/',CLIPtype,'/',TrasncLoc,'/',RunMode,'/upstream/Random_outchunck_frac_100chunk4.txt'),comment.char = "%",header = F,sep = "\t",stringsAsFactors = F))

 BPPfracRando=rbind(BPPfracRando,BPPfracRando2) 
            }
 ###################################

 BPPfracRando=as.data.frame(BPPfracRando[,1]/BPPfracRando[,2])
 colnames(BPPfracRando)='frac_high_BPP'
 
 
 
 ggplot(BPPfracRando,aes(x=frac_high_BPP))+geom_density(size= .6) + theme_classic()+ylab("Density")+xlab("Fraction of Locations with with local BPP > 0.2")+ggtitle('Upstream: Fraction of Locations with High Local BasePairing Probabilities' )+
       geom_vline(xintercept=sum(as.numeric(checkout2$RowMax_MeanChuckBPP)>.2,na.rm = T)/nrow(checkout2),colour="red") 
      
 
 ################################### ################################### ###################################
      BPPfracRando_dn=as.matrix(read.delim(paste0(inRandom,'/Random1/',CLIPtype,'/',TrasncLoc,'/',RunMode,'/downstream/Random_dn_outchunck_frac_100chunk4.txt'),comment.char = "%",header = F,sep = "\t",stringsAsFactors = F))
            for (rand in c(2:12)) {

      BPPfracRando2_dn=as.matrix(read.delim(paste0(inRandom,'/Random',rand,'/',CLIPtype,'/',TrasncLoc,'/',RunMode,'/downstream/Random_dn_outchunck_frac_100chunk4.txt'),comment.char = "%",header = F,sep = "\t",stringsAsFactors = F))
      
 BPPfracRando_dn=rbind(BPPfracRando_dn,BPPfracRando2_dn)       
            }
########
 
 BPPfracRando_dn=as.data.frame(BPPfracRando_dn[,1]/BPPfracRando_dn[,2])
 colnames(BPPfracRando_dn)='frac_high_BPP'
 
 
 
 ggplot(BPPfracRando_dn,aes(x=frac_high_BPP))+geom_density(size= .6) + theme_classic()+ylab("Density")+xlab("Fraction of Locations with with local BPP > 0.2")+ggtitle('Downstream: Fraction of Locations with High Local BasePairing Probabilities' )+
       geom_vline(xintercept=sum(as.numeric(checkout2_dn$RowMax_MeanChuckBPP)>.2,na.rm = T)/nrow(checkout2_dn),colour="red") 
 
 
 
 
      BPPavgRando=as.matrix(read.delim(paste0(inRandom,'/Random1/',CLIPtype,'/',TrasncLoc,'/',RunMode,'/upstream/Random_outchunck_rowaverage_100chuck4.txt'),comment.char = "%",header = T,row.names=1,sep = "\t",stringsAsFactors = F))
      
      for (rand in c(2:12)) {

      BPPavgRando2=as.matrix(read.delim(paste0(inRandom,'/Random',rand,'/',CLIPtype,'/',TrasncLoc,'/',RunMode,'/upstream/Random_outchunck_rowaverage_100chuck4.txt'),comment.char = "%",header = T,row.names = 1, sep = "\t",stringsAsFactors = F))
      
       BPPavgRando=rbind(as.data.frame(BPPavgRando),
                            as.data.frame(BPPavgRando2))
      
      }
     
      
BPPavgRando=BPPavgRando[is.na(BPPavgRando[,1])==F,]
# BPPavgRando=as.numeric(as.character(BPPavgRando))
  colnames(BPPavgRando)=seq(-window,(ncol(BPPavgRando)-(window+1)),1)
 rownames(BPPavgRando)=paste0('n',rownames(BPPavgRando))
 BPPavgRando= melt(as.matrix(BPPavgRando))
       colnames(BPPavgRando)=c('Random','location','avg_high_BPP')
BPPavgRando$avg_high_BPP=as.numeric(as.character(BPPavgRando$avg_high_BPP))
 
   
 CLIP=(ph[ph$diff>.2,])
 CLIP=(colMeans(CLIP[,colnames(CLIP)%in%c('diff','n')==F]))
 CLIP=as.data.frame(rbind(CLIP,colnames(CLIP)))
 CLIP=t(CLIP)
 CLIP=as.data.frame(cbind(CLIP,rownames(CLIP)))
   colnames(CLIP)=c('avg_high_BPP','location')
CLIP$avg_high_BPP=as.numeric(as.matrix(CLIP$avg_high_BPP))
 CLIP$location=as.numeric(as.matrix(CLIP$location))
CLIP$Random='upstream'


ggplot(BPPavgRando,aes(y=avg_high_BPP,x=location,line=Random))+
   geom_line(data=BPPavgRando,size= .6,color='grey') +
  geom_line(data=CLIP,size= .6,color='red')+
  theme_classic()+ylab("Average Basepairing Probability")+xlab("Position of 10nt Window")+ggtitle('Upstream: Average BPP of Locations with local BPP >0.2')+ylim(0,.3)




########################################################################################################
########################################################################################################
########################################################################################################

      BPPavgRando_dn=as.matrix(read.delim(paste0(inRandom,'/Random1/',CLIPtype,'/',TrasncLoc,'/',RunMode,'/downstream/Random_dn_outchunck_rowaverage_100chucnk4.txt'),comment.char = "%",header = T,row.names=1,sep = "\t",stringsAsFactors = F))

for (rand in c(2:12)) {
  
      BPPavgRando_dn2=as.matrix(read.delim(paste0(inRandom,'/Random',rand,'/',CLIPtype,'/',TrasncLoc,'/',RunMode,'/downstream/Random_dn_outchunck_rowaverage_100chucnk4.txt'),comment.char = "%",header = T,row.names = 1, sep = "\t",stringsAsFactors = F))
      
      BPPavgRando_dn=rbind(as.data.frame(BPPavgRando_dn),
                            as.data.frame(BPPavgRando_dn2))
      
      }
       

BPPavgRando_dn=BPPavgRando_dn[is.na(BPPavgRando_dn[,1])==F,]
# BPPavgRando_dn=as.numeric(as.character(BPPavgRando_dn))
  colnames(BPPavgRando_dn)=seq(1,ncol(BPPavgRando_dn),1)
 rownames(BPPavgRando_dn)=paste0('n',rownames(BPPavgRando_dn))
 BPPavgRando_dn= melt(as.matrix(BPPavgRando_dn))
       colnames(BPPavgRando_dn)=c('Random','location','avg_high_BPP')
BPPavgRando_dn$avg_high_BPP=as.numeric(as.character(BPPavgRando_dn$avg_high_BPP))
 
 CLIP=(phdn[phdn$diff>.2,])
 CLIP=CLIP[,colnames(CLIP)%in%c('diff','n')==F]
  colnames(CLIP)=seq(1,length(CLIP),1)
 CLIP=(colMeans(CLIP[,colnames(CLIP)%in%c('diff','n')==F]))
 CLIP=as.data.frame(rbind(CLIP,colnames(CLIP)))
 CLIP=t(CLIP)
 CLIP=as.data.frame(cbind(CLIP,rownames(CLIP)))
   colnames(CLIP)=c('avg_high_BPP','location')
CLIP$avg_high_BPP=as.numeric(as.matrix(CLIP$avg_high_BPP))
 CLIP$location=as.numeric(as.matrix(CLIP$location))
CLIP$Random='downstream'


# ggplot(BPPavgRando_dn1,aes(y=avg_high_BPP,x=location,line=Random))+
#    geom_line(data=BPPavgRando_dn1,size= .6,color='grey') +
#      geom_line(data=BPPavgRando_dn2,size= .6,color='grey') +
#    geom_line(data=BPPavgRando_dn3,size= .6,color='grey') +
#    geom_line(data=BPPavgRando_dn4,size= .6,color='grey') +
#   geom_line(data=CLIP,size= .6,color='red')+
#   theme_classic()+ylab("Average Basepairing Probability")+xlab("Position of 10nt Window")+ggtitle('Downsteam: Average BPP of Locations with local BPP >0.2')+ylim(0,.3)


ggplot(BPPavgRando_dn,aes(y=avg_high_BPP,x=location,line=Random))+
   geom_line(data=BPPavgRando_dn,size= .6,color='grey') +
  geom_line(data=CLIP,size= .6,color='red')+
  theme_classic()+ylab("Average Basepairing Probability")+xlab("Position of 10nt Window")+ggtitle('Downsteam: Average BPP of Locations with local BPP >0.2')+ylim(0,.3)

```
  
## Structure/Count Correlations           


### Upstream  
      
     
```{r ,echo=F,fig.asp=.75, include=T, fig.align='center',fig.height=3, fig.width=3, figures-side, fig.show="hold"}
           
##########################
## Upstream
##########################

   sctrplot=checkout2_HM
  sctrplot[is.na(sctrplot$qgrs_maxScore),'qgrs_maxScore']=0
  sctrplot$qgrs_maxScore=as.numeric(sctrplot$qgrs_maxScore)  
  sctrplot$DeltaG=-sctrplot$DeltaG
  ggscatter(sctrplot, x = 'qgrs_maxScore', y = 'DeltaG', add = "reg.line",color = "red",   
          add.params = list(color = "black", fill = NA), # Customize reg. line
          xlab=expression("GQuad score (qgrs)" ), 
           ylab='CLIP peak Upstream (100nt) \n-DeltaG'
) +
 stat_cor(
   aes(label = paste(..rr.label.., ..p.label.., sep = "~`,`~")),
  label.x = 30,label.y=.65
)
  
  
  ggscatter(sctrplot, x = 'qgrs_maxScore', y = 'RowMax_MeanChuckBPP', add = "reg.line",color = "red",   
          add.params = list(color = "black", fill = NA), # Customize reg. line
          xlab=expression("GQuad score (qgrs)" ), 
           ylab='CLIP peak maximum 10nt window Base pairing probability'
) +
 stat_cor(
   aes(label = paste(..rr.label.., ..p.label.., sep = "~`,`~")),
  label.x = 30,label.y=.65
)
    
  #########  #########  #########  #########  #########  #########  #########  #########  #########  #########

 ggscatter(sctrplot, x = 'qgrs_maxScore', y = 'Counts_Multimappers_Scaled', add = "reg.line",color = "red",   
          add.params = list(color = "black", fill = NA), # Customize reg. line
          xlab=expression("CLIP peak Upstream (100nt) \nGQuad score (qgrs)" ), 
           ylab='CLIP peak Read Count (scaled MultiMappers)'
) +
 stat_cor(
   aes(label = paste(..rr.label.., ..p.label.., sep = "~`,`~"))#,
  # label.x = 30,label.y=.65
)  
  
   ggscatter(sctrplot, x = 'DeltaG', y = 'Counts_Multimappers_Scaled', add = "reg.line",color = "red",   
          add.params = list(color = "black", fill = NA), # Customize reg. line
          xlab=expression("CLIP peak Upstream (100nt) \n-DeltaG" ), 
           ylab='CLIP peak Read Count (scaled MultiMappers)'
) +
 stat_cor(
   aes(label = paste(..rr.label.., ..p.label.., sep = "~`,`~"))#,
  # label.x = 30,label.y=.65
)
  

 ggscatter(sctrplot, x = 'RowMax_MeanChuckBPP', y = 'Counts_Multimappers_Scaled', add = "reg.line",color = "red",   
          add.params = list(color = "black", fill = NA), # Customize reg. line
          xlab=expression("CLIP peak Upstream (100nt) \nmaximum 10nt window Base pairing probability" ), 
           ylab='CLIP peak Read Count (scaled MultiMappers)'
) +
 stat_cor(
   aes(label = paste(..rr.label.., ..p.label.., sep = "~`,`~"))#,
  # label.x = 30,label.y=.65
)

```

### Downstream  
      
     
```{r ,echo=F,fig.asp=.75, include=T, fig.align='center',fig.height=3, fig.width=3, figures-side, fig.show="hold"}
 
 
##########################
## Downstream
##########################
  
  sctrplot_dn=checkout2_dn_HM
  sctrplot_dn[is.na(sctrplot_dn$qgrs_maxScore),'qgrs_maxScore']=0
  sctrplot_dn$qgrs_maxScore=as.numeric(sctrplot_dn$qgrs_maxScore)  
  sctrplot_dn$DeltaG=-sctrplot_dn$DeltaG
  ggscatter(sctrplot_dn, x = 'qgrs_maxScore', y = 'DeltaG', add = "reg.line",color = "red",   
          add.params = list(color = "black", fill = NA), # Customize reg. line
          xlab=expression("GQuad score (qgrs)" ), 
           ylab='CLIP peak downstream (100nt) \n-DeltaG'
) +
 stat_cor(
   aes(label = paste(..rr.label.., ..p.label.., sep = "~`,`~")),
  label.x = 10,label.y=.65
)
  
  
  ggscatter(sctrplot_dn, x = 'qgrs_maxScore', y = 'RowMax_MeanChuckBPP', add = "reg.line",color = "red",   
          add.params = list(color = "black", fill = NA), # Customize reg. line
          xlab=expression("GQuad score (qgrs)" ), 
           ylab='CLIP peak maximum 10nt window Base pairing probability'
) +
 stat_cor(
   aes(label = paste(..rr.label.., ..p.label.., sep = "~`,`~")),
  label.x = 30,label.y=.65
)
    
  #########  #########  #########  #########  #########  #########  #########  #########  #########  #########

 ggscatter(sctrplot_dn, x = 'qgrs_maxScore', y = 'Counts_Multimappers_Scaled', add = "reg.line",color = "red",   
          add.params = list(color = "black", fill = NA), # Customize reg. line
          xlab=expression("CLIP peak downstream (100nt) \nGQuad score (qgrs)" ), 
           ylab='CLIP peak Read Count (scaled MultiMappers)'
) +
 stat_cor(
   aes(label = paste(..rr.label.., ..p.label.., sep = "~`,`~"))#,
  # label.x = 30,label.y=.65
)  
  
   ggscatter(sctrplot_dn, x = 'DeltaG', y = 'Counts_Multimappers_Scaled', add = "reg.line",color = "red",   
          add.params = list(color = "black", fill = NA), # Customize reg. line
          xlab=expression("CLIP peak downstream (100nt) \n-DeltaG" ), 
           ylab='CLIP peak Read Count (scaled MultiMappers)'
) +
 stat_cor(
   aes(label = paste(..rr.label.., ..p.label.., sep = "~`,`~"))#,
  # label.x = 30,label.y=.65
)
  

 ggscatter(sctrplot_dn, x = 'RowMax_MeanChuckBPP', y = 'Counts_Multimappers_Scaled', add = "reg.line",color = "red",   
          add.params = list(color = "black", fill = NA), # Customize reg. line
          xlab=expression("CLIP peak downstream (100nt) \nmaximum 10nt window Base pairing probability" ), 
           ylab='CLIP peak Read Count (scaled MultiMappers)'
) +
 stat_cor(
   aes(label = paste(..rr.label.., ..p.label.., sep = "~`,`~"))#,
  # label.x = 30,label.y=.65
)
 
  # length(Gname)
  # ncol(hmDN)
 
```         

     
       
## Structure associated with LINE/SINE CLIP peaks   

Select CLIP peaks that overlap with Introns on the same strand as CLIP peak. 
Identify CLIP peaks that overlap with LINE/SINE elements on either strand in 100nt window Upstream and Downstream.  
Compare deltaG, local base paring probability and Gquad identification/score for Intronic regions.   
        
### Upstream Structure Associated With Intronic CLIP peaks that overlap with LINE/SINES    

#### On either strand   

```{r ,echo=F,fig.asp=.75, include=T, fig.align='center',fig.height=3, fig.width=3, figures-side, fig.show="hold"}

################################################################################        
#         Does not exclude peaks thatoverlap with simplerepeats/Low_complexity on opposite strand,
#         but if CLIP-intron does overlap with simplerepeats/Low_complexity on same strand it is removed
        
## will include peaks that do not OL with intron on Same strand but does on opposite strand
LS1=Peaksdata4_chunkInExall[Peaksdata4_chunkInExall$`Both Strand: RNA_type_Classification`%in%c('protein_coding: exon')==F,] 
### Includes any peak that only overlaps with introns on same strand
LS2= Peaksdata4_chunkInExall[grepl('intron',Peaksdata4_chunkInExall$`Same Strand: Intron Exon`),]
### takes away CLIP peaks classified as opposite strand features and still could be interesting
LS3= Peaksdata4_chunkInExall[Peaksdata4_chunkInExall$`Both Strand: RNA_type_Classification`%in%c('protein_coding: exon','Antisense Feature')==F,]

################################################################################        

LISIplot=merge(checkout2[checkout2$CLIP_ID%in%LS3$ID,],isoout[,c('ID','LISI_OLup','strand_LISIup','LISI_OLdn','strand_LISIdn')],by.x='CLIP_ID',by.y='ID')
LISIplot$LISI_OLup=gsub(1,"OL with LISI",LISIplot$LISI_OLup)
LISIplot[is.na(LISIplot$LISI_OLup),'LISI_OLup']="No overlap"
LISIplot$qgrs_maxScore=as.numeric(LISIplot$qgrs_maxScore)
LISIplot[is.na(LISIplot$qgrs_maxScore),'qgrs_maxScore']=0
LISIplot=merge(LISIplot,Peaksdata4[,c('ID','Counts_Unique','Counts_Multimappers_Scaled','Counts_Multimappers_BestMapping')],by.x='CLIP_ID',by.y='ID')



ggplot(LISIplot,aes(x=RowMax_MeanChuckBPP,line=LISI_OLup,color=LISI_OLup))+
   geom_density(size= .6) +
  theme_classic()+xlab("Maximum Window Basepairing Probability")+ylab("Density")+ggtitle('Upstream: Maximum BPP of 10nt Chuck')

ggplot(LISIplot,aes(x=-DeltaG,line=LISI_OLup,color=LISI_OLup))+
   geom_density(size= .6) +
  theme_classic()+xlab("-Delta G")+ylab("Density")+ggtitle('Upstream: Folding Energy of 100nt ')


ggplot(LISIplot,aes(x=qgrs_maxScore,line=LISI_OLup,color=LISI_OLup))+
   geom_density(size= .6) +
  theme_classic()+xlab("-Delta G")+ylab("Density")+ggtitle('Upstream: Gquad Score ')

ggplot(LISIplot,aes(x=Counts_Unique,line=LISI_OLup,color=LISI_OLup))+
   geom_density(size= .6) +
  theme_classic()+xlab("CLIP counts:Unique")+ylab("Density")+ggtitle('Upstream: CLIP counts:Unique ')


ggplot(LISIplot,aes(x=Counts_Multimappers_Scaled,line=LISI_OLup,color=LISI_OLup))+
   geom_density(size= .6) +
  theme_classic()+xlab("CLIP counts:Unique")+ylab("Density")+ggtitle('Upstream: CLIP counts:Multimappers_Scaled ')
```


#### On same strand   
    
```{r ,echo=F,fig.asp=.75, include=T, fig.align='center',fig.height=3, fig.width=3, figures-side, fig.show="hold"}

p=LISIplot[-which(LISIplot$strand==LISIplot$strand_LISIup),]

ggplot(p,aes(x=RowMax_MeanChuckBPP,line=LISI_OLup,color=LISI_OLup))+
   geom_density(size= .6) +
  theme_classic()+xlab("Maximum Window Basepairing Probability")+ylab("Density")+ggtitle('Upstream SameStrand: Maximum BPP of 10nt Chuck')

ggplot(p,aes(x=-DeltaG,line=LISI_OLup,color=LISI_OLup))+
   geom_density(size= .6) +
  theme_classic()+xlab("-Delta G")+ylab("Density")+ggtitle('Upstream SameStrand: Folding Energy of 100nt ')


ggplot(p,aes(x=qgrs_maxScore,line=LISI_OLup,color=LISI_OLup))+
   geom_density(size= .6) +
  theme_classic()+xlab("-Delta G")+ylab("Density")+ggtitle('Upstream SameStrand: Gquad Score ')

ggplot(p,aes(x=Counts_Unique,line=LISI_OLup,color=LISI_OLup))+
   geom_density(size= .6) +
  theme_classic()+xlab("CLIP counts:Unique")+ylab("Density")+ggtitle('Upstream SameStrand: CLIP counts:Unique ')

ggplot(p,aes(x=Counts_Multimappers_Scaled,line=LISI_OLup,color=LISI_OLup))+
   geom_density(size= .6) +
  theme_classic()+xlab("CLIP counts:Unique")+ylab("Density")+ggtitle('Upstream SameStrand: CLIP counts:Multimappers_Scaled ')

```

### Downstream Structure Associated With Intronic CLIP peaks that overlap with LINE/SINES   

#### On either strand   
      
```{r ,echo=F,fig.asp=.75, include=T, fig.align='center',fig.height=3, fig.width=3, figures-side, fig.show="hold"}

LISIplot_dn=merge(checkout2_dn[checkout2_dn$CLIP_ID%in%LS3$ID,],isoout[,c('ID','LISI_OLup','strand_LISIup','LISI_OLdn','strand_LISIdn')],by.x='CLIP_ID',by.y='ID')
LISIplot_dn$LISI_OLdn=gsub(1,"OL with LISI",LISIplot_dn$LISI_OLdn)
LISIplot_dn[is.na(LISIplot_dn$LISI_OLdn),'LISI_OLdn']="No overlap"
LISIplot_dn$qgrs_maxScore=as.numeric(LISIplot_dn$qgrs_maxScore)
LISIplot_dn[is.na(LISIplot_dn$qgrs_maxScore),'qgrs_maxScore']=0
LISIplot_dn=merge(LISIplot_dn,Peaksdata4[,c('ID','Counts_Unique','Counts_Multimappers_Scaled','Counts_Multimappers_BestMapping')],by.x='CLIP_ID',by.y='ID')


ggplot(LISIplot_dn,aes(x=RowMax_MeanChuckBPP,line=LISI_OLdn,color=LISI_OLdn))+
   geom_density(size= .6) +
  theme_classic()+xlab("Maximum Window Basepairing Probability")+ylab("Density")+ggtitle('Downstream: Maximum BPP of 10nt Chuck')

ggplot(LISIplot_dn,aes(x=-DeltaG,line=LISI_OLdn,color=LISI_OLdn))+
   geom_density(size= .6) +
  theme_classic()+xlab("-Delta G")+ylab("Density")+ggtitle('Downstream: Folding Energy of 100nt ')


ggplot(LISIplot_dn,aes(x=qgrs_maxScore,line=LISI_OLdn,color=LISI_OLdn))+
   geom_density(size= .6) +
  theme_classic()+xlab("-Delta G")+ylab("Density")+ggtitle('Downstream: Gquad Score ')


ggplot(LISIplot_dn,aes(x=Counts_Unique,line=LISI_OLdn,color=LISI_OLdn))+
   geom_density(size= .6) +
  theme_classic()+xlab("CLIP counts:Unique")+ylab("Density")+ggtitle('Downstream: CLIP counts:Unique ')

ggplot(LISIplot_dn,aes(x=Counts_Multimappers_Scaled,line=LISI_OLdn,color=LISI_OLdn))+
   geom_density(size= .6) +
  theme_classic()+xlab("CLIP counts:Unique")+ylab("Density")+ggtitle('Downstream: CLIP counts:Multimappers_Scaled ')

```

#### On same strand    
       
```{r ,echo=F,fig.asp=.75, include=T, fig.align='center',fig.height=3, fig.width=3, figures-side, fig.show="hold"}

pdn=LISIplot_dn[-which(LISIplot_dn$strand==LISIplot_dn$strand_LISIdn),]

ggplot(pdn,aes(x=RowMax_MeanChuckBPP,line=LISI_OLdn,color=LISI_OLdn))+
   geom_density(size= .6) +
  theme_classic()+xlab("Maximum Window Basepairing Probability")+ylab("Density")+ggtitle('Downstream SameStrand: Maximum BPP of 10nt Chuck')

ggplot(pdn,aes(x=-DeltaG,line=LISI_OLdn,color=LISI_OLdn))+
   geom_density(size= .6) +
  theme_classic()+xlab("-Delta G")+ylab("Density")+ggtitle('Downstream SameStrand: Folding Energy of 100nt ')


ggplot(pdn,aes(x=qgrs_maxScore,line=LISI_OLdn,color=LISI_OLdn))+
   geom_density(size= .6) +
  theme_classic()+xlab("-Delta G")+ylab("Density")+ggtitle('Downstream SameStrand: Gquad Score ')

ggplot(pdn,aes(x=Counts_Unique,line=LISI_OLdn,color=LISI_OLdn))+
   geom_density(size= .6) +
  theme_classic()+xlab("CLIP counts:Unique")+ylab("Density")+ggtitle('Downstream SameStrand: CLIP counts:Unique ')


ggplot(pdn,aes(x=Counts_Multimappers_Scaled,line=LISI_OLdn,color=LISI_OLdn))+
   geom_density(size= .6) +
  theme_classic()+xlab("CLIP counts:Unique")+ylab("Density")+ggtitle('Downstream SameStrand: CLIP counts:Multimappers_Scaled ')
```

## DISTANCE TO SINE/LINE  

CLIP peak 100nt Upstream/Downstream regions were selected from full transcript with introns  
      
Any CLIP peaks that overlap with intronic regions were included in in Intronic CLIP peak analysis. If Peak also overlaps with a Repeat Element  it is removed, except for peaks that overlap with LINE, SINE or LTR.  
     
All LINE/SINE in same transcript as CLIP peak , or on the opposite strands, were considered for closest LINE/SINE element   
      
Random peaks were taken from Intronic regions only  
      

### INTRONS  
#### NEW  

```{r ,echo=F,warning=F,eval=F,include=F}
         # remove(checkout2_LS,mm10_mRNAome,CLIP_mRNAome,CLIP_mRNAome_up,CLIP_mRNAome_dn,CLIP_mRNAome_GR,CLIP_mRNAome_GRstrand,LISI_mRNAome,LISI_mRNAome_GR,LISI_mRNAome_GRstrand)
        # remove(mm10_CLIP_mRNAome_transc,CLIP_CLIP_mRNAome_transc,CLIP_CLIP_mRNAome_transcUP,CLIP_CLIP_mRNAome_transcDN,CLIP_CLIP_mRNAome_transc_GR,CLIP_CLIP_mRNAome_transc_GRstrand,LISI_CLIP_mRNAome_transc,LISI_CLIP_mRNAome_transc_GR,LISI_CLIP_mRNAome_transc_GRstrand)   
remove(mm10all,mm10_trans,mm10_exon,mm10_UTR,introns,mm10_intron,CrossLinkADJ)
       
################################################################################################################################################################       
# install.packages("BiocManager")
# BiocManager::install("regioneR")
# library(regioneR)

# crosslink_tbl=checkout2_LSBS_XLnk
# window=window
# LISIstrand='both' #same, opposite, both
MakeTranscriptome =function(crosslink_tbl,window,LISIstrand){
  QC_table=as.data.frame(matrix(nrow=1,ncol=5));colnames(QC_table)= c('CL_trans_OL','LISI_CL_OL','outDist_LISI_CL_OL','outDist_LISI_CL_OLUP','outDist_LISI_CL_OLDN')
  
#################
  
 mm10all=fread("/Users/homanpj/OneDrive - National Institutes of Health/Resources/ref/mm10/Gencode_VM23/fromGencode/gencode.vM23.basic.annotation.gtf.txt", header=T, sep="\t",stringsAsFactors = F,data.table=F)

    mm10all$transcript_id=removeVersion(mm10all$transcript_id)
    mm10all$gene_id=removeVersion(mm10all$gene_id)
    mm10_transall=mm10all[mm10all$feature%in%'transcript',]
    
mm10_trans=mm10_transall[mm10_transall$transcript_id%in%canonical$transcript,]
mm10_trans=mm10_trans[mm10_trans$transcript_type%in%'protein_coding',]


    ###########################
    ### Make sure that Transcripts from Exonic CLIP peaks are selected. 
    ###########################
Uptrancripts=crosslink_tbl$transcript_id
Uptrancripts_mm10_trans=mm10_transall[mm10_transall$transcript_id%in%Uptrancripts,]



mm10_trans=rbind(mm10_trans[mm10_trans$gene_id%in%unique(Uptrancripts_mm10_trans$gene_id)==F,],
            Uptrancripts_mm10_trans)
 
# setdiff(crosslink_tbl$transcript_id,exons$transcript_id)

trans_GR=GRanges(seqnames=mm10_trans$seqname, 
                 ranges=IRanges(start=mm10_trans$start,end=mm10_trans$end),
                 strand=mm10_trans$strand,
                 transcript_id=mm10_trans$transcript_id,
                 gene_id=mm10_trans$gene_id
                 )

mm10_trans_genome=mm10_trans[,c('transcript_id','seqname','start','end')]
mm10_trans_genome$length=(mm10_trans_genome$end-mm10_trans_genome$start)+1
mm10_trans_genome=mm10_trans_genome[,c('transcript_id','length')]
# mm10_trans_genome$V2=as.integer(mm10_trans_genome$V2)

rownames(mm10_trans_genome)=NULL
colnames(mm10_trans_genome)=c('V1','V2')

########################################################################################        
### Adjust Exon Coordinates
##############
    mm10_exonall=mm10all[mm10all$feature%in%'exon',]
    colnames(mm10_exonall)=paste0(colnames(mm10_exonall),'_exons')
    mm10_exonall$transcript_id_exons=removeVersion(mm10_exonall$transcript_id_exons)

   mm10_exon=merge(mm10_exonall,mm10_trans,by.x='transcript_id_exons',by.y='transcript_id')

   
   mm10_exon[mm10_exon$strand%in%"+",'strtAdj']=mm10_exon[mm10_exon$strand%in%"+",'start_exons']-mm10_exon[mm10_exon$strand%in%"+",'start']
   mm10_exon[mm10_exon$strand%in%"+",'endAdj']=mm10_exon[mm10_exon$strand%in%"+",'end_exons']-mm10_exon[mm10_exon$strand%in%"+",'start']
   
   mm10_exon[mm10_exon$strand%in%"-",'strtAdj']=mm10_exon[mm10_exon$strand%in%"-",'end']-mm10_exon[mm10_exon$strand%in%"-",'end_exons']
   mm10_exon[mm10_exon$strand%in%"-",'endAdj']=mm10_exon[mm10_exon$strand%in%"-",'end']-mm10_exon[mm10_exon$strand%in%"-",'start_exons']

   mm10_exo_mask=mm10_exon[,c('transcript_id_exons','strtAdj','endAdj')]

##############   
   mm10_exo_maskcheck=merge(mm10_exo_mask,mm10_trans[,c('transcript_id','seqname','start','end','strand')],by.x='transcript_id_exons',by.y='transcript_id')
   
   mm10_exo_maskcheck[mm10_exo_maskcheck$strand%in%"+",'start_OG']=mm10_exo_maskcheck[mm10_exo_maskcheck$strand%in%"+",'start']+mm10_exo_maskcheck[mm10_exo_maskcheck$strand%in%"+",'strtAdj']
   mm10_exo_maskcheck[mm10_exo_maskcheck$strand%in%"+",'end_OG']=mm10_exo_maskcheck[mm10_exo_maskcheck$strand%in%"+",'start']+mm10_exo_maskcheck[mm10_exo_maskcheck$strand%in%"+",'endAdj']
   
   mm10_exo_maskcheck[mm10_exo_maskcheck$strand%in%"-",'end_OG']=mm10_exo_maskcheck[mm10_exo_maskcheck$strand%in%"-",'end']-mm10_exo_maskcheck[mm10_exo_maskcheck$strand%in%"-",'strtAdj']
   mm10_exo_maskcheck[mm10_exo_maskcheck$strand%in%"-",'start_OG']=mm10_exo_maskcheck[mm10_exo_maskcheck$strand%in%"-",'end']-mm10_exo_maskcheck[mm10_exo_maskcheck$strand%in%"-",'endAdj']
   
   mm10_exo_maskcheck$loc=paste0(mm10_exo_maskcheck$seqname,':',mm10_exo_maskcheck$start_OG,'-',mm10_exo_maskcheck$end_OG)
   
  # mm10_exo_maskcheck[sample(1:nrow(mm10_exo_maskcheck),1),]
   
# View(mm10_exon[,c('seqname_exons','start_exons','end_exons','strand_exons','transcript_id_exons','strtAdj','endAdj')])

########################################################################################        
   #### adjust CLIP coordinates
#############

crosslink_tbl=merge(crosslink_tbl,mm10_trans,by='transcript_id',all.x=T,suffixes=c('_CLIP','_mm10'))

crosslink_tbl$CL_adj=NA

crosslink_tbl[crosslink_tbl$strand_CLIP%in%"+",'CL_adj']=
  crosslink_tbl[crosslink_tbl$strand_CLIP%in%"+",'crosslink']-
  crosslink_tbl[crosslink_tbl$strand_CLIP%in%"+",'start']

crosslink_tbl[crosslink_tbl$strand_CLIP%in%"-",'CL_adj']=
  crosslink_tbl[crosslink_tbl$strand_CLIP%in%"-",'end']-
  crosslink_tbl[crosslink_tbl$strand_CLIP%in%"-",'crosslink']

crosslink_tbl$TranscWidth=crosslink_tbl$end-crosslink_tbl$start
# crosslink_tbl[crosslink_tbl$CL_adj+1>=crosslink_tbl$TranscWidth,]

CrossLinkADJ=crosslink_tbl[,c('transcript_id','chr','crosslink',"strand_CLIP",'ID','gene_id','gene_name','start','end','CL_adj')]


crosslink_tbl[crosslink_tbl$strand_CLIP%in%"+",'CL_up']=crosslink_tbl[crosslink_tbl$strand_CLIP%in%"+",'crosslink']-(window-1)
crosslink_tbl[crosslink_tbl$strand_CLIP%in%"-",'CL_up']=crosslink_tbl[crosslink_tbl$strand_CLIP%in%"-",'crosslink']+(window-1)


crosslink_tbl[crosslink_tbl$strand_CLIP%in%"+",'CL_dn']=crosslink_tbl[crosslink_tbl$strand_CLIP%in%"+",'crosslink']+(window-1)
crosslink_tbl[crosslink_tbl$strand_CLIP%in%"-",'CL_dn']=crosslink_tbl[crosslink_tbl$strand_CLIP%in%"-",'crosslink']-(window-1)


CL_GR = GRanges(seqnames=crosslink_tbl$chr, 
                  ranges=IRanges(start=crosslink_tbl$crosslink,
                  end=crosslink_tbl$crosslink),
                  strand=crosslink_tbl$strand_CLIP,
                  ID=crosslink_tbl$ID,
                  transcript_id=crosslink_tbl$transcript_id,
                  crosslink=crosslink_tbl$crosslink
                  )

###################################################################

rmsk_GRCm38=fread(paste0("/Users/homanpj/OneDrive - National Institutes of Health/Resources/ref/mm10/repeatmasker/rmsk_GRCm38.txt"), header=T, sep="\t",stringsAsFactors = F,data.table=F)
rmsk_GRCm38_LISI=rmsk_GRCm38[rmsk_GRCm38$repClass%in%c('LINE','SINE'),]
rmsk_GRCm38_LISI=rmsk_GRCm38_LISI[rmsk_GRCm38_LISI$genoName%in%c('chrUn_GL456239', 'chrUn_GL456359', 'chrUn_GL456360', 'chrUn_GL456366', 'chrUn_GL456367', 'chrUn_GL456368', 'chrUn_GL456370', 'chrUn_GL456372', 'chrUn_GL456378', 'chrUn_GL456379', 'chrUn_GL456381', 'chrUn_GL456382', 'chrUn_GL456385', 'chrUn_GL456387', 'chrUn_GL456389', 'chrUn_GL456393', 'chrUn_GL456394', 'chrUn_JH584304', 'chr1_GL456210_random', 'chr1_GL456211_random', 'chr1_GL456212_random', 'chr1_GL456213_random', 'chr1_GL456221_random', 'chr4_GL456216_random', 'chr4_GL456350_random', 'chr4_JH584292_random', 'chr4_JH584293_random', 'chr4_JH584294_random', 'chr5_GL456354_random', 'chr5_JH584296_random', 'chr5_JH584297_random', 'chr5_JH584298_random', 'chr5_JH584299_random', 'chr7_GL456219_random', 'chrX_GL456233_random', 'chrY_JH584300_random', 'chrY_JH584301_random', 'chrY_JH584302_random', 'chrY_JH584303_random')==F,]
rmsk_GRCm38_LISI$ID=paste0(rmsk_GRCm38_LISI$genoName,":",rmsk_GRCm38_LISI$genoStart,'-',rmsk_GRCm38_LISI$genoEnd)

if (LISIstrand=='opposite') {
rmsk_GRCm38_LISI[rmsk_GRCm38_LISI$strand%in%"+","strand"]='+_-'
rmsk_GRCm38_LISI[rmsk_GRCm38_LISI$strand%in%"-","strand"]='+'
rmsk_GRCm38_LISI[rmsk_GRCm38_LISI$strand%in%"+_-","strand"]='-'
}

LISI_GR = GRanges(seqnames=rmsk_GRCm38_LISI$genoName, 
                  ranges=IRanges(start=rmsk_GRCm38_LISI$genoStart,
                  end=rmsk_GRCm38_LISI$genoEnd),
                  strand=rmsk_GRCm38_LISI$strand,
                  repClass=rmsk_GRCm38_LISI$repClass,
                  repName=rmsk_GRCm38_LISI$repName,
                  ID=rmsk_GRCm38_LISI$ID
                  )



### Get LINE/SINES that overlap with Transcripts
        gr = trans_GR
        gr2 = LISI_GR
        xo=as.data.frame(GenomicRanges::findOverlaps(gr,gr2,type = "any",ignore.strand=F))
        if (LISIstrand=='both') {xo=as.data.frame(GenomicRanges::findOverlaps(gr,gr2,type = "any",ignore.strand=T))}

        LISI_Trans_GR=as.data.frame(gr2[xo$subjectHits])
                  # qh=as.data.frame(gr[xo$queryHits],row.names = NULL)
                  # colnames(qh)=paste0(colnames(qh),"_mm10")
                  # sh=as.data.frame(gr2[xo$subjectHits],row.names = NULL)
        Trans_LISI_GR=as.data.frame(gr[xo$queryHits])
          colnames(LISI_Trans_GR)=paste0(colnames(LISI_Trans_GR),"_LINESINE")
          colnames(Trans_LISI_GR)=paste0(colnames(Trans_LISI_GR),"_Trans")
        
        LISI_Trans=cbind(as.data.frame(LISI_Trans_GR),as.data.frame(Trans_LISI_GR))
        
        
##############################        
# Adjust LINE/SINE coordinates
        LISI_Trans$LISI_startadj=NA
        LISI_Trans$LISI_endadj=NA
        
        LISI_Trans[LISI_Trans$strand_Trans%in%"+",'LISI_startadj']=
  LISI_Trans[LISI_Trans$strand_Trans%in%"+",'start_LINESINE']-
  LISI_Trans[LISI_Trans$strand_Trans%in%"+",'start_Trans']
        
          LISI_Trans[LISI_Trans$strand_Trans%in%"+",'LISI_endadj']=
  LISI_Trans[LISI_Trans$strand_Trans%in%"+",'end_LINESINE']-
  LISI_Trans[LISI_Trans$strand_Trans%in%"+",'start_Trans'] 
          
##########    
          LISI_Trans[LISI_Trans$strand_Trans%in%"-",'LISI_startadj']=
  LISI_Trans[LISI_Trans$strand_Trans%in%"-",'end_Trans']-
  LISI_Trans[LISI_Trans$strand_Trans%in%"-",'end_LINESINE']

          LISI_Trans[LISI_Trans$strand_Trans%in%"-",'LISI_endadj']=
  LISI_Trans[LISI_Trans$strand_Trans%in%"-",'end_Trans']-
  LISI_Trans[LISI_Trans$strand_Trans%in%"-",'start_LINESINE']               

LISI_Trans$LISI_widthadj=(LISI_Trans$LISI_endadj-LISI_Trans$LISI_startadj)+1
          
# LISI_Trans[LISI_Trans$width_LINESINE!=LISI_Trans$LISI_widthadj,]
# LISI_Trans[,c('ID_LINESINE','width_LINESINE','LISI_widthadj')]
# LISI_Trans[duplicated(LISI_Trans),]

LISI_mRNAome_GRstrand=GRanges(seqnames=LISI_Trans$transcript_id_Trans, 
                          ranges=IRanges(start=LISI_Trans$LISI_startadj,end=LISI_Trans$LISI_endadj),
                          # strand=as.character(LISI_Trans$strand_LINESINE),
                          ID=LISI_Trans$ID_LINESINE
                          # chrm=LISI_Trans$seqname
                          )   


  setDx=setdiff(CrossLinkADJ$transcript_id,as.data.frame(LISI_mRNAome_GRstrand)$seqnames)
##############################        
CrossLinkADJ=CrossLinkADJ[CrossLinkADJ$transcript_id%in%setDx==F,]
CLIP_mRNAome_GRstrand=GRanges(seqnames=CrossLinkADJ$transcript_id, 
                          ranges=IRanges(start=CrossLinkADJ$CL_adj, end=CrossLinkADJ$CL_adj),
                          # strand=as.character(CrossLinkADJ$strand_CLIP),
                          ID=CrossLinkADJ$ID
                          # chrm=CLIP_mRNAome$seqname
                          )


################################################################################################################################################
### check transcripts that overlap with CLIP
        gr = CL_GR
        gr2 = trans_GR
        xo=as.data.frame(GenomicRanges::findOverlaps(gr,gr2,type = "any",ignore.strand=F))
          
        Trans_CL_GR=as.data.frame(gr2[xo$subjectHits])
                  # qh=as.data.frame(gr[xo$queryHits],row.names = NULL)
                  # colnames(qh)=paste0(colnames(qh),"_mm10")
                  # sh=as.data.frame(gr2[xo$subjectHits],row.names = NULL)
        CL_Trans_GR=as.data.frame(gr[xo$queryHits])
        CL_Trans_comb=cbind(CL_Trans_GR,Trans_CL_GR)
# length(unique(CL_Trans_comb$ID))
# CL_Trans_comb[CL_Trans_comb$ID%in%CL_Trans_comb[duplicated(CL_Trans_comb$ID),'ID'],]
QC_table[1,"CL_trans_OL"]=length(unique(CL_Trans_comb$ID))
        
        
### check LINE/SINES that overlap with CLIP
        gr = CL_GR
        gr2 = LISI_GR
        xo=as.data.frame(GenomicRanges::findOverlaps(gr,gr2,type = "any",ignore.strand=F))
          
        LISI_CL_GR=as.data.frame(gr2[xo$subjectHits])
                  # qh=as.data.frame(gr[xo$queryHits],row.names = NULL)
                  # colnames(qh)=paste0(colnames(qh),"_mm10")
                  # sh=as.data.frame(gr2[xo$subjectHits],row.names = NULL)
        CL_LISI_GR=as.data.frame(gr[xo$queryHits])
        
         CL_LISI_GR[CL_LISI_GR$strand%in%"+",'CL_ID']=paste0(CL_LISI_GR[CL_LISI_GR$strand%in%"+",'seqnames'],':',CL_LISI_GR[CL_LISI_GR$strand%in%"+",'start'],'-',CL_LISI_GR[CL_LISI_GR$strand%in%"+",'end']+1)
         CL_LISI_GR[CL_LISI_GR$strand%in%"-",'CL_ID']=paste0(CL_LISI_GR[CL_LISI_GR$strand%in%"-",'seqnames'],':',CL_LISI_GR[CL_LISI_GR$strand%in%"-",'start'],'-',CL_LISI_GR[CL_LISI_GR$strand%in%"-",'end']+1)
QC_table[1,"LISI_CL_OL"]=length(unique(CL_LISI_GR$ID))
         
         
         CL_GR_tbl=as.data.frame(CL_GR)
          CL_GR_tbl[CL_GR_tbl$strand%in%"+",'CL_ID']=paste0(CL_GR_tbl[CL_GR_tbl$strand%in%"+",'seqnames'],':',CL_GR_tbl[CL_GR_tbl$strand%in%"+",'start'],'-',CL_GR_tbl[CL_GR_tbl$strand%in%"+",'end']+1)
         CL_GR_tbl[CL_GR_tbl$strand%in%"-",'CL_ID']=paste0(CL_GR_tbl[CL_GR_tbl$strand%in%"-",'seqnames'],':',CL_GR_tbl[CL_GR_tbl$strand%in%"-",'start'],'-',CL_GR_tbl[CL_GR_tbl$strand%in%"-",'end']+1)

         
#######################
### Count distances
########################
        
        outDist=as.data.frame(matrix(nrow(CrossLinkADJ),ncol=8));colnames(outDist)=c('Transcript','CLIP_ID','CLIP_LISI_Dist','CLIP_LISI_Dist_UP','CLIP_LISI_Dist_DN','ID_LINESINE','repClass_LINESINE','repName_LINESINE')
        for (CL in 1:nrow(CrossLinkADJ)) {

          cl1=CrossLinkADJ[CL,]
          si1=LISI_Trans[LISI_Trans$transcript_id_Trans%in%cl1$transcript_id,]
          
          outDist[CL,'Transcript']=cl1[1,'transcript_id']
          outDist[CL,'CLIP_ID']=cl1[1,'ID']
          
          if (nrow(si1)>0) {
        
            si1$OL=(si1$LISI_startadj<=cl1$CL_adj)&(si1$LISI_endadj>=cl1$CL_adj)
            
            if (nrow(si1[si1$OL==T,])>0) {
                        outDist[CL,'CLIP_LISI_Dist']=0
            outDist[CL,'ID_LINESINE']=si1[1,'ID_LINESINE']
            outDist[CL,'repClass_LINESINE']=si1[1,'repClass_LINESINE']
            outDist[CL,'repName_LINESINE']=si1[1,'repName_LINESINE']
                        next}  
            
            si1$distStart=cl1$CL_adj-si1$LISI_startadj
            si1$distEnd=cl1$CL_adj-si1$LISI_endadj
            si1$minDist=rowMins(as.matrix(abs(si1[,c('distStart','distEnd'),drop=F])))
            si1$Dir=as.character(si1$distStart>0|si1$distEnd>0)
            si1$Dir=gsub("FALSE","DN", si1$Dir)
            si1$Dir=gsub("TRUE","UP", si1$Dir)


            if (nrow(si1[si1$Dir%in%'UP',])>0) {outDist[CL,'CLIP_LISI_Dist_UP']=-min(si1[si1$Dir%in%'UP','minDist'])}
            if (nrow(si1[si1$Dir%in%'DN',])>0) {outDist[CL,'CLIP_LISI_Dist_DN']=min(si1[si1$Dir%in%'DN','minDist'])}
            
            si1=si1[si1$minDist%in%min(si1$minDist),]
            
### If equal distance up and down then default is to make it downstream
            if (T%in%(si1$Dir%in%'UP')) {outDist[CL,'CLIP_LISI_Dist']=-si1[si1$Dir%in%'UP','minDist']}
            if (T%in%(si1$Dir%in%'DN')) {outDist[CL,'CLIP_LISI_Dist']=si1[si1$Dir%in%'DN','minDist']
            si1=si1[si1$Dir%in%'DN',]}
              
              
              
            outDist[CL,'ID_LINESINE']=si1$ID_LINESINE
            outDist[CL,'repClass_LINESINE']=si1$repClass_LINESINE
            outDist[CL,'repName_LINESINE']=si1$repName_LINESINE
            
          }
          
        }

         outDist[outDist$CLIP_LISI_Dist%in%0,'CLIP_LISI_Dist_UP']=0
         outDist[outDist$CLIP_LISI_Dist%in%0,'CLIP_LISI_Dist_DN']=0

QC_table[1,"outDist_LISI_CL_OL"]=nrow(outDist[outDist$CLIP_LISI_Dist==0,])
QC_table[1,"outDist_LISI_CL_OLUP"]=nrow(outDist[outDist$CLIP_LISI_Dist<=0&outDist$CLIP_LISI_Dist>=-window,])
QC_table[1,"outDist_LISI_CL_OLDN"]=nrow(outDist[outDist$CLIP_LISI_Dist>=0&outDist$CLIP_LISI_Dist<=window,])

########################
### Upstream
########################
CLIP_mRNAome_up=as.data.frame(CLIP_mRNAome_GRstrand)
CLIP_mRNAome_up=merge(CLIP_mRNAome_up,mm10_trans_genome,by.x='seqnames',by.y='V1')

      CLIP_mRNAome_up[,'end']=CLIP_mRNAome_up[,'start']
      CLIP_mRNAome_up[,'start']=CLIP_mRNAome_up[,'start']-(window-1)
      
      
      CLIP_mRNAome_up[(CLIP_mRNAome_up$start<1),'start']=1
      CLIP_mRNAome_up[(CLIP_mRNAome_up$end>CLIP_mRNAome_up$V2),'end']= CLIP_mRNAome_up[(CLIP_mRNAome_up$V3>CLIP_mRNAome_up$V2_Genome),'V2_Genome']
  

       CLIP_mRNAome_GRstrand_UP=GRanges(seqnames=CLIP_mRNAome_up$seqnames, 
                            ranges=IRanges(start=CLIP_mRNAome_up$start,end=CLIP_mRNAome_up$end),
                          strand=CLIP_mRNAome_up$strand,
                          ID=CLIP_mRNAome_up$ID
                          # chrm=CLIP_mRNAome_up$seqname
                          )
 # setdiff(CLIP_mRNAome_up$V1,out$genome$V1)   
 
 
########################
### downstream
########################
CLIP_mRNAome_dn = as.data.frame(CLIP_mRNAome_GRstrand)
CLIP_mRNAome_dn=merge(CLIP_mRNAome_dn,mm10_trans_genome,by.x='seqnames',by.y='V1')

      CLIP_mRNAome_dn[,'end']=CLIP_mRNAome_dn[,'start']+(window-1)
      
      
      CLIP_mRNAome_dn[(CLIP_mRNAome_dn$V2<1),'start']=1
      CLIP_mRNAome_dn[(CLIP_mRNAome_dn$end>CLIP_mRNAome_dn$V2_Genome),'end']= CLIP_mRNAome_dn[(CLIP_mRNAome_dn$end>CLIP_mRNAome_dn$V2_Genome),'V2']
        


       CLIP_mRNAome_GRstrand_DN=GRanges(seqnames=CLIP_mRNAome_dn$seqnames, 
                          ranges=IRanges(start=CLIP_mRNAome_dn$start,end=CLIP_mRNAome_dn$end),
                          strand=CLIP_mRNAome_dn$strand,
                          ID=CLIP_mRNAome_dn$ID
                          # chrm=CLIP_mRNAome_dn$seqname
                          )
 # setdiff(CLIP_mRNAome_dn$V1,mm10_mRNAome$V1)    
         
       
return(list(
  CLIP_GR=CLIP_mRNAome_GRstrand,
  CLIP_GR_UP=CLIP_mRNAome_GRstrand_UP,
  CLIP_GR_DN=CLIP_mRNAome_GRstrand_DN,
  LISI_GR=LISI_mRNAome_GRstrand,
  genome=mm10_trans_genome,
  exon_mask=mm10_exo_mask,
  LISIdist=outDist,
  trans_GR=trans_GR,
  QC_table=QC_table))
}

```

#### Include same only Strand LISI


```{r ,echo=F,warning=F ,eval=F, include=F}
remove(mm10all,mm10_trans,mm10_exon,mm10_UTR,introns,mm10_intron)
       # 

########################################################################################        
###### AllIntrons 
        USELISIstrand='same'
        recalcPT=T; if(recalcPT==T){remove(pt,pt2,LZ2,pt_dn,pt2_dn,LZ2_dn,pt_up,pt2_up,LZ2_up)}
        pTntimes=10

################################################################################        
#         Does not exclude peaks thatoverlap with simplerepeats/Low_complexity on opposite strand,
#         but if CLIP-intron does overlap with simplerepeats/Low_complexity on same strand it is removed
        
## will include peaks that do not OL with intron on Same strand but does on opposite strand
checkout2_LS1=Peaksdata4_chunkInExall[Peaksdata4_chunkInExall$`Both Strand: RNA_type_Classification`%in%c('protein_coding: exon')==F,] 
### Includes any peak that only overlaps with introns on same strand
checkout2_LS2= Peaksdata4_chunkInExall[grepl('intron',Peaksdata4_chunkInExall$`Same Strand: Intron Exon`),]
### takes away CLIP peaks classified as opposite strand features and still could be interesting
checkout2_LS3= Peaksdata4_chunkInExall[Peaksdata4_chunkInExall$`Both Strand: RNA_type_Classification`%in%c('protein_coding: exon','Antisense Feature')==F,]


# as.character(unique(checkout2_LS1$`Both Strand: RNA_type_Classification`))
# as.character(unique(checkout2_LS1$`Same Strand: Repeat_Type`))
# as.character(unique(checkout2_LS1$`Opposite Strand: Repeat_Type`))

checkout2_LS=checkout2_LS2
################################################################################        


checkout2_LS=merge(checkout2_LS,isoout_transcAll[,c('ID','USEtranscript','CL')],by="ID")
colnames(checkout2_LS)[colnames(checkout2_LS)%in%"USEtranscript"]='transcript_id'
colnames(checkout2_LS)[colnames(checkout2_LS)%in%"CL"]='crosslink'

checkout2_LS_XLnk=checkout2_LS[,c('chr','crosslink','strand','ID','transcript_id')]
Intron_LISI_SS=MakeTranscriptome(checkout2_LS_XLnk,window,USELISIstrand)

# Intron_LISI_SS$CLIP_GR_UP
# Intron_LISI_SS$CLIP_GR_DN

#####################################################################################################       
##### Adjusted Coordinates


if (recalcPT==T) {
  
pt= permTest(
  A=Intron_LISI_SS$CLIP_GR,
  B=Intron_LISI_SS$LISI_GR,
  genome=Intron_LISI_SS$genome,#mm10_mRNAome,
    mask=Intron_LISI_SS$exon_mask,
  randomize.function=randomizeRegions,
  evaluate.function=meanDistance,ntimes=pTntimes,count.once=F,verbose=F)
# # 
# #   as.data.frame(CLIP_mRNAome_GRstrand)
# #   as.data.frame(LISI_mRNAome_GRstrand)
# 
# ################################# ################################# ################################# 

pt2= overlapPermTest(
  A=Intron_LISI_SS$CLIP_GR,
  B=Intron_LISI_SS$LISI_GR,
  genome=Intron_LISI_SS$genome,#mm10_mRNAome,
    mask=Intron_LISI_SS$exon_mask,
  non.overlapping=T,
  ntimes=pTntimes,count.once=TRUE)
# 
# ################################# ################################# ################################# 

# LZ2=localZScore(pt=pt2,
#   A=Intron_LISI_SS$CLIP_GR,
#   B=Intron_LISI_SS$LISI_GR,
#   genome=Intron_LISI_SS$genome,#mm10_mRNAome,
#     mask=Intron_LISI_SS$exon_mask,
#   window=500,
#   step=10,count.once=TRUE)
# 
}
# ################################# ################################# ################################# 
  QC_table_PT=as.data.frame(matrix(nrow=1,ncol=5));colnames(QC_table_PT)= c('CL_trans_OL','LISI_CL_OL','outDist_LISI_CL_OL','outDist_LISI_CL_OLUP','outDist_LISI_CL_OLDN')

{par(mar = c(5, 5, 5, 5)) # Set the margin on all sides to 6
plot(pt)
mtext(paste0("LINE/SINE overlap with Intronic CLIP: ",USELISIstrand," Strand "))}
{par(mar = c(5, 5, 5, 5)) # Set the margin on all sides to 6
plot(pt2)
mtext(paste0("LINE/SINE overlap with Intronic CLIP: ",USELISIstrand," Strand "))}
# {plot(LZ2)
# mtext(paste0("LINE/SINE overlap with Intronic CLIP: ",USELISIstrand," Strand "))}


        QC_table_PT[,'outDist_LISI_CL_OL']=pt2[["numOverlaps"]][["observed"]]   

########################################################################################   
  QC_table_check=as.data.frame(matrix(nrow=1,ncol=5));colnames(QC_table_check)= c('CL_trans_OL','LISI_CL_OL','outDist_LISI_CL_OL','outDist_LISI_CL_OLUP','outDist_LISI_CL_OLDN')


        gr = Intron_LISI_SS$CLIP_GR
        gr2 = Intron_LISI_SS$LISI_GR
        xo=as.data.frame(GenomicRanges::findOverlaps(gr,gr2,type = "any",ignore.strand=F))
          
        LISI_CLIP_VISH=gr2[xo$subjectHits]
                  # qh=as.data.frame(gr[xo$queryHits],row.names = NULL)
                  # colnames(qh)=paste0(colnames(qh),"_mm10")
                  # sh=as.data.frame(gr2[xo$subjectHits],row.names = NULL)
        CLIP_LISI_VISH=gr[xo$queryHits] 

QC_table_check[,'outDist_LISI_CL_OL']=length(unique(CLIP_LISI_VISH$ID ))    
       
        
        gr = Intron_LISI_SS$LISI_GR
        gr2 = Intron_LISI_SS$trans_GR
        xo=as.data.frame(GenomicRanges::findOverlaps(gr,gr2,type = "any",ignore.strand=F))
          
        Transc_CLIP_VISH=gr2[xo$subjectHits]
                  # qh=as.data.frame(gr[xo$queryHits],row.names = NULL)
                  # colnames(qh)=paste0(colnames(qh),"_mm10")
                  # sh=as.data.frame(gr2[xo$subjectHits],row.names = NULL)
        CLIP_Transc_VISH=gr[xo$queryHits]      
# intersect(LISI_OL$V1,as.data.frame(LISI_CLIP_VISH)$seqnames)
        QC_table_check[,'CL_trans_OL']=length(unique(CLIP_LISI_VISH$seqnames ))    


        
########################################################################################        
######### Get Vishal data
IntronLISI_detail=fread(paste0("./structure/LINESINE/for_PhilHoman2/for_PhilHoman/CLIP_in_introns_with_nearest_SINE_LINE.premmRNA.txt"), header=F, sep="\t",stringsAsFactors = F,data.table=F)
colnames(IntronLISI_detail)=c('CLIPpeak_mmRNA_chrom','CLIPpeak_mmRNA_start','CLIPpeak_mmRNA_end',	'CLIPpeak_original_coordinates','V5','V6',	'SINE_LINE_mmRNA_chrom',	'SINE_LINE_mmRNA_start',	'SINE_LINE_mmRNA_end', 'SINE_LINE_original_coordinates' , 'V11','V12',	'Distance_on_mmRNA','SINE_LINE_original_Chr','SINE_LINE_original_start','SINE_LINE_original_end','SINE_LINE_id', 'V18',	'SINE_LINE_strand','exon_transcript_chr','exon_transcript_start','exon_transcript_end',	'exon_transcript_id','V24',	'transcript_strand')

#### select oposite strand LINE/SINE
IntronLISI_detail=IntronLISI_detail[IntronLISI_detail$transcript_strand==IntronLISI_detail$SINE_LINE_strand,]

# IntronLISI_detail[IntronLISI_detail$CLIPpeak_original_coordinates%in%intersect(CL_LISI_GR$CL_ID,IntronLISI_detail$CLIPpeak_original_coordinates),]

IntronLISI_detail$CLIPpeak_mmRNA_chrom=removeVersion(IntronLISI_detail$CLIPpeak_mmRNA_chrom)

# IntronLISI_detail=IntronLISI_detail[IntronLISI_detail$CLIPpeak_mmRNA_chrom%in%checkIntron_LISI_SS2_LS$transcript_id,]

#############################################
  QC_table_vish=as.data.frame(matrix(nrow=1,ncol=5));colnames(QC_table_vish)= c('CL_trans_OL','LISI_CL_OL','outDist_LISI_CL_OL','outDist_LISI_CL_OLUP','outDist_LISI_CL_OLDN')

QC_table_vish[,"outDist_LISI_CL_OL"]=nrow(IntronLISI_detail[duplicated(IntronLISI_detail$CLIPpeak_original_coordinates)==F&IntronLISI_detail$Distance_on_mmRNA==0,])
QC_table_vish[,"outDist_LISI_CL_OLUP"]=nrow(IntronLISI_detail[duplicated(IntronLISI_detail$CLIPpeak_original_coordinates)==F&IntronLISI_detail$Distance_on_mmRNA<=0&IntronLISI_detail$Distance_on_mmRNA>=-window,])
QC_table_vish[,"outDist_LISI_CL_OLDN"]=nrow(IntronLISI_detail[duplicated(IntronLISI_detail$CLIPpeak_original_coordinates)==F&IntronLISI_detail$Distance_on_mmRNA>=0&IntronLISI_detail$Distance_on_mmRNA<=window,])

rbind(Intron_LISI_SS$QC_table,QC_table_vish)

        # nrow(IntronLISI_detail[duplicated(IntronLISI_detail$CLIPpeak_original_coordinates)==F&IntronLISI_detail$Distance_on_mmRNA<=0,])
        # nrow(IntronLISI_detail[duplicated(IntronLISI_detail$CLIPpeak_original_coordinates)==F&IntronLISI_detail$Distance_on_mmRNA>=0,])
        

        
        {plot(density(IntronLISI_detail$Distance_on_mmRNA))
        lines(density((Intron_LISI_SS$LISIdist$CLIP_LISI_Dist)),col='red')
        }
        #     Intron_LISI_SS$LISIdist=merge(Intron_LISI_SS$LISIdist,CL_GR_tbl[,c('ID','CL_ID')],by.x='CLIP_ID',by.y='ID')
        #     
        #     pp2=Intron_LISI_SS$LISIdist
        #     colnames(pp2)[colnames(pp2)%in%'CLIP_LISI_Dist']='Distance_on_mmRNA'
        # 
        # ggplot(IntronLISI_detail,aes(x=Distance_on_mmRNA))+geom_density()+geom_density(data=pp2,color='red')+scale_y_continuous(trans='log10')+xlim(-5000,5000)
        #     
        #     
        # comp=merge(IntronLISI_detail,Intron_LISI_SS$LISIdist,by.x='CLIPpeak_original_coordinates',by.y='CL_ID')
        # {plot(comp$Distance_on_mmRNA,comp$CLIP_LISI_Dist)
        # }
        
        # View(IntronLISI_detail[duplicated(IntronLISI_detail$CLIPpeak_original_coordinates)==F&IntronLISI_detail$Distance_on_mmRNA==0,])
        # View(IntronLISI_detail[duplicated(IntronLISI_detail$CLIPpeak_original_coordinates)==F&IntronLISI_detail$Distance_on_mmRNA<0&IntronLISI_detail$Distance_on_mmRNA>=-100,])
        # View(IntronLISI_detail[duplicated(IntronLISI_detail$CLIPpeak_original_coordinates)==F&IntronLISI_detail$Distance_on_mmRNA>0&IntronLISI_detail$Distance_on_mmRNA<=100,])
                
```

```{r ,echo=F,warning=F,eval=F,include=F}
        
########################
### Upstream
########################
if (recalcPT==T) {

pt_up= permTest(
  A=Intron_LISI_SS$CLIP_GR_UP,
  B=Intron_LISI_SS$LISI_GR,
  genome=Intron_LISI_SS$genome,#mm10_mRNAome,
    mask=Intron_LISI_SS$exon_mask,
  randomize.function=randomizeRegions,
  evaluate.function=meanDistance,ntimes=pTntimes,count.once=TRUE,verbose=F)

################################# ################################# ################################# 


pt2_up= overlapPermTest(
  A=Intron_LISI_SS$CLIP_GR_UP,
  B=Intron_LISI_SS$LISI_GR,
  genome=Intron_LISI_SS$genome,#mm10_mRNAome,
    mask=Intron_LISI_SS$exon_mask,
  ntimes=pTntimes,count.once=TRUE)
  # non.overlapping=T)

# LZ2_up=localZScore(pt=pt2_up,
#   A=Intron_LISI_SS$CLIP_GR_UP,
#   B=Intron_LISI_SS$LISI_GR,
#   genome=Intron_LISI_SS$genome,#mm10_mRNAome,
#     mask=Intron_LISI_SS$exon_mask,
#   window=500,
#   step=10,count.once=TRUE)

}

{par(mar = c(5, 5, 5, 5)) # Set the margin on all sides to 6
plot(pt_up)
mtext(paste0("LINE/SINE overlap with Intronic CLIP-100nt Upstream: ",USELISIstrand," Strand "))}
{par(mar = c(5, 5, 5, 5)) # Set the margin on all sides to 6
plot(pt2_up)
mtext(paste0("LINE/SINE overlap with Intronic CLIP-100nt Upstream: ",USELISIstrand," Strand "))}

# {plot(LZ2_up)
# mtext(paste0("LINE/SINE overlap with Intronic CLIP-100nt Upstream: ",USELISIstrand," Strand "))}

        
        
        gr = Intron_LISI_SS$CLIP_GR_UP
        gr2 = Intron_LISI_SS$LISI_GR
        xo=as.data.frame(GenomicRanges::findOverlaps(gr,gr2,type = "any",ignore.strand=F))
          
        LISI_CLIP_VISHup=gr2[xo$subjectHits]
                  # qh=as.data.frame(gr[xo$queryHits],row.names = NULL)
                  # colnames(qh)=paste0(colnames(qh),"_mm10")
                  # sh=as.data.frame(gr2[xo$subjectHits],row.names = NULL)
        CLIP_LISI_VISHup=gr[xo$queryHits] 
        
        # length(unique(as.data.frame(CLIP_LISI_VISHup)$ID))
        # View(as.data.frame(CLIP_LISI_VISHup))
        
        QC_table_check[,'outDist_LISI_CL_OLUP']=length(unique(CLIP_LISI_VISHup$ID ))    

            
#############################################################################################################################################################        
### downstream
###########################        

if (recalcPT==T) {

pt_dn= permTest(
  A=Intron_LISI_SS$CLIP_GR_DN,
  B=Intron_LISI_SS$LISI_GR,
  genome=Intron_LISI_SS$genome,#mm10_mRNAome,
    mask=Intron_LISI_SS$exon_mask,
  randomize.function=randomizeRegions,
  evaluate.function=meanDistance,ntimes=pTntimes,count.once=TRUE,verbose=F)

################################# ################################# ################################# 


pt2_dn= overlapPermTest(
  A=Intron_LISI_SS$CLIP_GR_DN,
  B=Intron_LISI_SS$LISI_GR,
  genome=Intron_LISI_SS$genome,#mm10_mRNAome,
    mask=Intron_LISI_SS$exon_mask,
  # non.overlapping=T,
  ntimes=pTntimes,count.once=T)

# LZ2_dn=localZScore(pt=pt2_dn,
#   A=Intron_LISI_SS$CLIP_GR_DN,
#   B=Intron_LISI_SS$LISI_GR,
#   genome=Intron_LISI_SS$genome,#mm10_mRNAome,
#     mask=Intron_LISI_SS$exon_mask,
#   window=500,
#   step=10,count.once=TRUE)

}

###########################################################################
{par(mar = c(5, 5, 5, 5)) # Set the margin on all sides to 6
plot(pt_dn)
mtext(paste0("LINE/SINE overlap with Intronic CLIP-100nt Downstream: ",USELISIstrand," Strand "))}
{par(mar = c(5, 5, 5, 5)) # Set the margin on all sides to 6
plot(pt2_dn)
mtext(paste0("LINE/SINE overlap with Intronic CLIP-100nt Downstream: ",USELISIstrand," Strand "))}
# {plot(LZ2_dn)
# mtext(paste0("LINE/SINE overlap with Intronic CLIP-100nt Downstream: ",USELISIstrand," Strand "))}

        QC_table_PT[,'outDist_LISI_CL_OLUP']=pt2_up[["numOverlaps"]][["observed"]]   
        QC_table_PT[,'outDist_LISI_CL_OLDN']=pt2_dn[["numOverlaps"]][["observed"]]   
###########################################################################
        
        
        gr = Intron_LISI_SS$CLIP_GR_DN
        gr2 = Intron_LISI_SS$LISI_GR
        xo=as.data.frame(GenomicRanges::findOverlaps(gr,gr2,type = "any",ignore.strand=F))
          
        LISI_CLIP_VISHdn=gr2[xo$subjectHits]
                  # qh=as.data.frame(gr[xo$queryHits],row.names = NULL)
                  # colnames(qh)=paste0(colnames(qh),"_mm10")
                  # sh=as.data.frame(gr2[xo$subjectHits],row.names = NULL)
        CLIP_LISI_VISHdn=gr[xo$queryHits] 
           # length(unique(as.data.frame(CLIP_LISI_VISHdn)$seqnames))
        QC_table_check[,'outDist_LISI_CL_OLDN']=length(unique(CLIP_LISI_VISHdn$ID ))    
        
        
       QC_table= rbind(Intron_LISI_SS$QC_table,QC_table_vish,QC_table_check,QC_table_PT)

 setdiff(unique(CLIP_LISI_VISHup$ID),Intron_LISI_SS$LISIdist[(Intron_LISI_SS$LISIdist$CLIP_LISI_Dist<=0&Intron_LISI_SS$LISIdist$CLIP_LISI_Dist>-100),'CLIP_ID'])
 setdiff(unique(CLIP_LISI_VISHup$ID),Intron_LISI_SS$LISIdist[(Intron_LISI_SS$LISIdist$CLIP_LISI_Dist<=100&Intron_LISI_SS$LISIdist$CLIP_LISI_Dist>-100),'CLIP_ID'])
 # setdiff(unique(CLIP_LISI_VISH$ID),Intron_LISI_SS$LISIdist[(Intron_LISI_SS$LISIdist$CLIP_LISI_Dist==0),'CLIP_ID'])
 
##############################
 ## the extra locations from the GR overlap that do not match the output from the Intron_LISI_SS$LISIdist table are locations that have LINE/SINEs < 100 nt away both upstream and downstream, they occure because the table only has one LISI per location however GR separatly identifies overlap upstrream and downstream
 
 Intron_LISI_SS$LISIdist[which((Intron_LISI_SS$LISIdist$CLIP_LISI_Dist_UP<=0&Intron_LISI_SS$LISIdist$CLIP_LISI_Dist_UP>-100)&
                           (Intron_LISI_SS$LISIdist$CLIP_LISI_Dist_DN>=0&Intron_LISI_SS$LISIdist$CLIP_LISI_Dist_DN<100)),'CLIP_ID']
 
 # View(Intron_LISI_SS$LISIdist)
```


#### Include Opposite only Strand LISI


```{r ,echo=F,warning=F,eval=F,include=F}

        remove(checkout2_LS,mm10_mRNAome,CLIP_mRNAome,CLIP_mRNAome_up,CLIP_mRNAome_dn,CLIP_mRNAome_GR,CLIP_mRNAome_GRstrand,LISI_mRNAome,LISI_mRNAome_GR,LISI_mRNAome_GRstrand)
        # remove(mm10_CLIP_mRNAome_transc,CLIP_CLIP_mRNAome_transc,CLIP_CLIP_mRNAome_transcUP,CLIP_CLIP_mRNAome_transcDN,CLIP_CLIP_mRNAome_transc_GR,CLIP_CLIP_mRNAome_transc_GRstrand,LISI_CLIP_mRNAome_transc,LISI_CLIP_mRNAome_transc_GR,LISI_CLIP_mRNAome_transc_GRstrand)   
remove(mm10all,mm10_trans,mm10_exon,mm10_UTR,introns,mm10_intron)
       
################################################################################################################################################################       
###### AllIntrons 

USELISIstrand='opposite'
        recalcPT=T; if(recalcPT==T){remove(pt_OS,pt2_OS,LZ2_OS,pt_OS_dn,pt2_OS_dn,LZ2_OS_dn,pt_OS_up,pt2_OS_up,LZ2_OS_up)}
        pTntimes=10
       
################################################################################        
#         Does not exclude peaks thatoverlap with simplerepeats/Low_complexity on opposite strand,
#         but if CLIP-intron does overlap with simplerepeats/Low_complexity on same strand it is removed
        
## will include peaks that do not OL with intron on Same strand but does on opposite strand
checkout2_LSOS1=Peaksdata4_chunkInExall[Peaksdata4_chunkInExall$`Both Strand: RNA_type_Classification`%in%c('protein_coding: exon')==F,] 
### Includes any peak that only overlaps with introns on same strand
checkout2_LSOS2= Peaksdata4_chunkInExall[grepl('intron',Peaksdata4_chunkInExall$`Same Strand: Intron Exon`),]
### takes away CLIP peaks classified as opposite strand features and still could be interesting
checkout2_LSOS3= Peaksdata4_chunkInExall[Peaksdata4_chunkInExall$`Both Strand: RNA_type_Classification`%in%c('protein_coding: exon','Antisense Feature')==F,]


# as.character(unique(checkout2_LSOS1$`Both Strand: RNA_type_Classification`))
# as.character(unique(checkout2_LSOS1$`Same Strand: Repeat_Type`))
# as.character(unique(checkout2_LSOS1$`Opposite Strand: Repeat_Type`))

checkout2_LSOS=checkout2_LSOS2
################################################################################   


checkout2_LSOS=merge(checkout2_LSOS,isoout_transcAll[,c('ID','USEtranscript','CL')],by="ID")
colnames(checkout2_LSOS)[colnames(checkout2_LSOS)%in%"USEtranscript"]='transcript_id'
colnames(checkout2_LSOS)[colnames(checkout2_LSOS)%in%"CL"]='crosslink'

checkout2_LSOS_XLnk=checkout2_LSOS[,c('chr','crosslink','strand','ID','transcript_id')]
Intron_LISI_OS=MakeTranscriptome(checkout2_LSOS_XLnk,window,USELISIstrand)

# Intron_LISI_OS$CLIP_GR_UP
# Intron_LISI_OS$CLIP_GR_DN


#####################################################################################################       
##### Adjusted Coordinates


if (recalcPT==T) {
  

pt_OS= permTest(
  A=Intron_LISI_OS$CLIP_GR,
  B=Intron_LISI_OS$LISI_GR,
  genome=Intron_LISI_OS$genome,#mm10_mRNAome,
    mask=Intron_LISI_OS$exon_mask,
  randomize.function=randomizeRegions,
  evaluate.function=meanDistance,ntimes=pTntimes,count.once=F,verbose=F)
# # 
# #   as.data.frame(CLIP_mRNAome_GRstrand)
# #   as.data.frame(LISI_mRNAome_GRstrand)
# 
# ################################# ################################# ################################# 

pt2_OS= overlapPermTest(
  A=Intron_LISI_OS$CLIP_GR,
  B=Intron_LISI_OS$LISI_GR,
  genome=Intron_LISI_OS$genome,#mm10_mRNAome,
    mask=Intron_LISI_OS$exon_mask,
  non.overlapping=T,
  ntimes=pTntimes,count.once=TRUE)
# 
# ################################# ################################# ################################# 

# LZ2_OS=localZScore(pt=pt2_OS,
#   A=Intron_LISI_OS$CLIP_GR,
#   B=Intron_LISI_OS$LISI_GR,
#   genome=Intron_LISI_OS$genome,#mm10_mRNAome,
#     mask=Intron_LISI_OS$exon_mask,
#   window=500,
#   step=10,count.once=TRUE)
# 
}
# ################################# ################################# ################################# 
  QC_table_PT=as.data.frame(matrix(nrow=1,ncol=5));colnames(QC_table_PT)= c('CL_trans_OL','LISI_CL_OL','outDist_LISI_CL_OL','outDist_LISI_CL_OLUP','outDist_LISI_CL_OLDN')

{par(mar = c(5, 5, 5, 5)) # Set the margin on all sides to 6
plot(pt_OS)
mtext(paste0("LINE/SINE overlap with Intronic CLIP: ",USELISIstrand," Strand "))}
{par(mar = c(5, 5, 5, 5)) # Set the margin on all sides to 6
plot(pt2_OS)
mtext(paste0("LINE/SINE overlap with Intronic CLIP: ",USELISIstrand," Strand "))}

# {plot(LZ2_OS)
# mtext(paste0("LINE/SINE overlap with Intronic CLIP: ",USELISIstrand," Strand "))}

        QC_table_PT[,'outDist_LISI_CL_OL']=pt2_OS[["numOverlaps"]][["observed"]]   

########################################################################################   
  QC_table_check=as.data.frame(matrix(nrow=1,ncol=5));colnames(QC_table_check)= c('CL_trans_OL','LISI_CL_OL','outDist_LISI_CL_OL','outDist_LISI_CL_OLUP','outDist_LISI_CL_OLDN')


        gr = Intron_LISI_OS$CLIP_GR
        gr2 = Intron_LISI_OS$LISI_GR
        xo=as.data.frame(GenomicRanges::findOverlaps(gr,gr2,type = "any",ignore.strand=F))
          
        LISI_CLIP_VISH=gr2[xo$subjectHits]
                  # qh=as.data.frame(gr[xo$queryHits],row.names = NULL)
                  # colnames(qh)=paste0(colnames(qh),"_mm10")
                  # sh=as.data.frame(gr2[xo$subjectHits],row.names = NULL)
        CLIP_LISI_VISH=gr[xo$queryHits] 

QC_table_check[,'outDist_LISI_CL_OL']=length(unique(CLIP_LISI_VISH$ID ))    
       
        
        gr = Intron_LISI_OS$LISI_GR
        gr2 = Intron_LISI_OS$trans_GR
        xo=as.data.frame(GenomicRanges::findOverlaps(gr,gr2,type = "any",ignore.strand=F))
          
        Transc_CLIP_VISH=gr2[xo$subjectHits]
                  # qh=as.data.frame(gr[xo$queryHits],row.names = NULL)
                  # colnames(qh)=paste0(colnames(qh),"_mm10")
                  # sh=as.data.frame(gr2[xo$subjectHits],row.names = NULL)
        CLIP_Transc_VISH=gr[xo$queryHits]      
# intersect(LISI_OL$V1,as.data.frame(LISI_CLIP_VISH)$seqnames)
        QC_table_check[,'CL_trans_OL']=length(unique(CLIP_LISI_VISH$seqnames ))    


        
########################################################################################        
               
```

```{r ,echo=F,warning=F,eval=F,include=F}
        
########################
### Upstream
########################
if (recalcPT==T) {

pt_OS_up= permTest(
  A=Intron_LISI_OS$CLIP_GR_UP,
  B=Intron_LISI_OS$LISI_GR,
  genome=Intron_LISI_OS$genome,#mm10_mRNAome,
    mask=Intron_LISI_OS$exon_mask,
  randomize.function=randomizeRegions,
  evaluate.function=meanDistance,ntimes=pTntimes,count.once=TRUE,verbose=F)

################################# ################################# ################################# 


pt2_OS_up= overlapPermTest(
  A=Intron_LISI_OS$CLIP_GR_UP,
  B=Intron_LISI_OS$LISI_GR,
  genome=Intron_LISI_OS$genome,#mm10_mRNAome,
    mask=Intron_LISI_OS$exon_mask,
  ntimes=pTntimes,count.once=TRUE)
  # non.overlapping=T)

# LZ2_OS_up=localZScore(pt=pt2_OS_up,
#   A=Intron_LISI_OS$CLIP_GR_UP,
#   B=Intron_LISI_OS$LISI_GR,
#   genome=Intron_LISI_OS$genome,#mm10_mRNAome,
#     mask=Intron_LISI_OS$exon_mask,
#   window=500,
#   step=10,count.once=TRUE)

}

################################################
{par(mar = c(5, 5, 5, 5)) # Set the margin on all sides to 6
plot(pt_OS_up)
mtext(paste0("LINE/SINE overlap with Intronic CLIP-100nt Upstream: ",USELISIstrand," Strand "))}
{par(mar = c(5, 5, 5, 5)) # Set the margin on all sides to 6
plot(pt2_OS_up)
mtext(paste0("LINE/SINE overlap with Intronic CLIP-100nt Upstream: ",USELISIstrand," Strand "))}

# {plot(LZ2_OS_up)
# mtext(paste0("LINE/SINE overlap with Intronic CLIP-100nt Upstream: ",USELISIstrand," Strand "))}

################################################

        gr = Intron_LISI_OS$CLIP_GR_UP
        gr2 = Intron_LISI_OS$LISI_GR
        xo=as.data.frame(GenomicRanges::findOverlaps(gr,gr2,type = "any",ignore.strand=F))
          
        LISI_CLIP_VISHup=gr2[xo$subjectHits]
                  # qh=as.data.frame(gr[xo$queryHits],row.names = NULL)
                  # colnames(qh)=paste0(colnames(qh),"_mm10")
                  # sh=as.data.frame(gr2[xo$subjectHits],row.names = NULL)
        CLIP_LISI_VISHup=gr[xo$queryHits] 
        
        # length(unique(as.data.frame(CLIP_LISI_VISHup)$ID))
        # View(as.data.frame(CLIP_LISI_VISHup))
        
        QC_table_check[,'outDist_LISI_CL_OLUP']=length(unique(CLIP_LISI_VISHup$ID ))    

            
#############################################################################################################################################################        
### downstream
###########################        

if (recalcPT==T) {

pt_OS_dn= permTest(
  A=Intron_LISI_OS$CLIP_GR_DN,
  B=Intron_LISI_OS$LISI_GR,
  genome=Intron_LISI_OS$genome,#mm10_mRNAome,
    mask=Intron_LISI_OS$exon_mask,
  randomize.function=randomizeRegions,
  evaluate.function=meanDistance,ntimes=pTntimes,count.once=TRUE,verbose=F)

################################# ################################# ################################# 


pt2_OS_dn= overlapPermTest(
  A=Intron_LISI_OS$CLIP_GR_DN,
  B=Intron_LISI_OS$LISI_GR,
  genome=Intron_LISI_OS$genome,#mm10_mRNAome,
    mask=Intron_LISI_OS$exon_mask,
  # non.overlapping=T,
  ntimes=pTntimes,count.once=T)

# LZ2_dn=localZScore(pt=pt2_OS_dn,
#   A=Intron_LISI_OS$CLIP_GR_DN,
#   B=Intron_LISI_OS$LISI_GR,
#   genome=Intron_LISI_OS$genome,#mm10_mRNAome,
#     mask=Intron_LISI_OS$exon_mask,
#   window=500,
#   step=10,count.once=TRUE)

}

################################################
{par(mar = c(5, 5, 5, 5)) # Set the margin on all sides to 6
plot(pt_OS_dn)
mtext(paste0("LINE/SINE overlap with Intronic CLIP-100nt Downstream: ",USELISIstrand," Strand "))}
{par(mar = c(5, 5, 5, 5)) # Set the margin on all sides to 6
plot(pt2_OS_dn)
mtext(paste0("LINE/SINE overlap with Intronic CLIP-100nt Downstream: ",USELISIstrand," Strand "))}
        
# {plot(LZ2_OS_dn)
# mtext(paste0("LINE/SINE overlap with Intronic CLIP-100nt Downstream: ",USELISIstrand," Strand "))}

      QC_table_PT[,'outDist_LISI_CL_OLUP']=pt2_OS_up[["numOverlaps"]][["observed"]]   
        QC_table_PT[,'outDist_LISI_CL_OLDN']=pt2_OS_dn[["numOverlaps"]][["observed"]] 
################################################
        
        gr = Intron_LISI_OS$CLIP_GR_DN
        gr2 = Intron_LISI_OS$LISI_GR
        xo=as.data.frame(GenomicRanges::findOverlaps(gr,gr2,type = "any",ignore.strand=F))
          
        LISI_CLIP_VISHdn=gr2[xo$subjectHits]
                  # qh=as.data.frame(gr[xo$queryHits],row.names = NULL)
                  # colnames(qh)=paste0(colnames(qh),"_mm10")
                  # sh=as.data.frame(gr2[xo$subjectHits],row.names = NULL)
        CLIP_LISI_VISHdn=gr[xo$queryHits] 
           # length(unique(as.data.frame(CLIP_LISI_VISHdn)$seqnames))
        QC_table_check[,'outDist_LISI_CL_OLDN']=length(unique(CLIP_LISI_VISHdn$ID ))    
        
        
       QC_table_OS= rbind(Intron_LISI_OS$QC_table,QC_table_vish,QC_table_check,QC_table_PT)

 setdiff(unique(CLIP_LISI_VISHup$ID),Intron_LISI_OS$LISIdist[(Intron_LISI_OS$LISIdist$CLIP_LISI_Dist<=0&Intron_LISI_OS$LISIdist$CLIP_LISI_Dist>-100),'CLIP_ID'])
 setdiff(unique(CLIP_LISI_VISHup$ID),Intron_LISI_OS$LISIdist[(Intron_LISI_OS$LISIdist$CLIP_LISI_Dist<=100&Intron_LISI_OS$LISIdist$CLIP_LISI_Dist>-100),'CLIP_ID'])
 # setdiff(unique(CLIP_LISI_VISH$ID),Intron_LISI_OS$LISIdist[(Intron_LISI_OS$LISIdist$CLIP_LISI_Dist==0),'CLIP_ID'])
 
##############################
 ## the extra locations from the GR overlap that do not match the output from the Intron_LISI_OS$LISIdist table are locations that have LINE/SINEs < 100 nt away both upstream and downstream, they occure because the table only has one LISI per location however GR separatly identifies overlap upstrream and downstream
 
 Intron_LISI_OS$LISIdist[which((Intron_LISI_OS$LISIdist$CLIP_LISI_Dist_UP<=0&Intron_LISI_OS$LISIdist$CLIP_LISI_Dist_UP>-100)&
                           (Intron_LISI_OS$LISIdist$CLIP_LISI_Dist_DN>=0&Intron_LISI_OS$LISIdist$CLIP_LISI_Dist_DN<100)),'CLIP_ID']
 
 # View(Intron_LISI_OS$LISIdist)
```


#### Include Opposite and same Strand

```{r ,echo=F,warning=F,eval=F,include=F}

        remove(checkout2_LS,mm10_mRNAome,CLIP_mRNAome,CLIP_mRNAome_up,CLIP_mRNAome_dn,CLIP_mRNAome_GR,CLIP_mRNAome_GRstrand,LISI_mRNAome,LISI_mRNAome_GR,LISI_mRNAome_GRstrand)
        # remove(mm10_CLIP_mRNAome_transc,CLIP_CLIP_mRNAome_transc,CLIP_CLIP_mRNAome_transcUP,CLIP_CLIP_mRNAome_transcDN,CLIP_CLIP_mRNAome_transc_GR,CLIP_CLIP_mRNAome_transc_GRstrand,LISI_CLIP_mRNAome_transc,LISI_CLIP_mRNAome_transc_GR,LISI_CLIP_mRNAome_transc_GRstrand)   
remove(mm10all,mm10_trans,mm10_exon,mm10_UTR,introns,mm10_intron)
       
       
################################################################################################################################################################       
###### AllIntrons 
USELISIstrand='both'
        recalcPT=T; if(recalcPT==T){remove(pt_BS,pt2_BS,LZ2_BS,pt_BS_dn,pt2_BS_dn,LZ2_BS_dn,pt_BS_up,pt2_BS_up,LZ2_BS_up)}
        pTntimes=10

################################################################################        
#         Does not exclude peaks thatoverlap with simplerepeats/Low_complexity on opposite strand,
#         but if CLIP-intron does overlap with simplerepeats/Low_complexity on same strand it is removed
        
## will include peaks that do not OL with intron on Same strand but does on opposite strand
checkout2_LSBS1=Peaksdata4_chunkInExall[Peaksdata4_chunkInExall$`Both Strand: RNA_type_Classification`%in%c('protein_coding: exon')==F,] 
### Includes any peak that only overlaps with introns on same strand
checkout2_LSBS2= Peaksdata4_chunkInExall[grepl('intron',Peaksdata4_chunkInExall$`Same Strand: Intron Exon`),]
### takes away CLIP peaks classified as opposite strand features and still could be interesting
checkout2_LSBS3= Peaksdata4_chunkInExall[Peaksdata4_chunkInExall$`Both Strand: RNA_type_Classification`%in%c('protein_coding: exon','Antisense Feature')==F,]


# as.character(unique(checkout2_LSBS1$`Both Strand: RNA_type_Classification`))
# as.character(unique(checkout2_LSBS1$`Same Strand: Repeat_Type`))
# as.character(unique(checkout2_LSBS1$`Opposite Strand: Repeat_Type`))

checkout2_LSBS=checkout2_LSBS2
################################################################################  

checkout2_LSBS=separate(checkout2_LSBS,ID,into = c('chr','start'),sep=':',remove = F)
checkout2_LSBS=separate(checkout2_LSBS,start,into = c('start','end'),sep='-')
checkout2_LSBS$start=as.numeric(checkout2_LSBS$start)
checkout2_LSBS$end=as.numeric(checkout2_LSBS$end)
checkout2_LSBS[checkout2_LSBS$strand%in%"+",'crosslink']=checkout2_LSBS[checkout2_LSBS$strand%in%"+",'start']
checkout2_LSBS[checkout2_LSBS$strand%in%"-",'crosslink']=checkout2_LSBS[checkout2_LSBS$strand%in%"-",'end']
checkout2_LSBS=merge(checkout2_LSBS,isoout_transcAll[,c('ID','USEtranscript')],by='ID',all.x=T)
colnames(checkout2_LSBS)[colnames(checkout2_LSBS)%in%'USEtranscript']='transcript_id'


checkout2_LSBS_XLnk=checkout2_LSBS[,c('chr','crosslink','strand','ID','transcript_id')]
Intron_LISI_BS=MakeTranscriptome(checkout2_LSBS_XLnk,window,USELISIstrand)

# Intron_LISI_BS$CLIP_GR_UP
# Intron_LISI_BS$CLIP_GR_DN


#####################################################################################################       
##### Adjusted Coordinates


if (recalcPT==T) {
  

pt_BS= permTest(
  A=Intron_LISI_BS$CLIP_GR,
  B=Intron_LISI_BS$LISI_GR,
  genome=Intron_LISI_BS$genome,#mm10_mRNAome,
    mask=Intron_LISI_BS$exon_mask,
  randomize.function=randomizeRegions,
  evaluate.function=meanDistance,ntimes=pTntimes,count.once=F,verbose=F)
# # 
# #   as.data.frame(CLIP_mRNAome_GRstrand)
# #   as.data.frame(LISI_mRNAome_GRstrand)
# 
# ################################# ################################# ################################# 

pt2_BS= overlapPermTest(
  A=Intron_LISI_BS$CLIP_GR,
  B=Intron_LISI_BS$LISI_GR,
  genome=Intron_LISI_BS$genome,#mm10_mRNAome,
    mask=Intron_LISI_BS$exon_mask,
  non.overlapping=T,
  ntimes=pTntimes,count.once=TRUE)
# 
# ################################# ################################# ################################# 

# LZ2_BS=localZScore(pt=pt2_BS,
#   A=Intron_LISI_BS$CLIP_GR,
#   B=Intron_LISI_BS$LISI_GR,
#   genome=Intron_LISI_BS$genome,#mm10_mRNAome,
#     mask=Intron_LISI_BS$exon_mask,
#   window=500,
#   step=10,count.once=TRUE)
# 
}
# ################################# ################################# ################################# 
  QC_table_PT=as.data.frame(matrix(nrow=1,ncol=5));colnames(QC_table_PT)= c('CL_trans_OL','LISI_CL_OL','outDist_LISI_CL_OL','outDist_LISI_CL_OLUP','outDist_LISI_CL_OLDN')

{par(mar = c(5, 5, 5, 5)) # Set the margin on all sides to 6
plot(pt_BS)
mtext(paste0("LINE/SINE overlap with Intronic CLIP: ",USELISIstrand," Strand "))}
{par(mar = c(5, 5, 5, 5)) # Set the margin on all sides to 6
plot(pt2_BS)
mtext(paste0("LINE/SINE overlap with Intronic CLIP: ",USELISIstrand," Strand "))}

# {plot(LZ2_BS)
# mtext(paste0("LINE/SINE overlap with Intronic CLIP: ",USELISIstrand," Strand "))}

      QC_table_PT[,'outDist_LISI_CL_OL']=pt2_BS[["numOverlaps"]][["observed"]]

########################################################################################   
  QC_table_check=as.data.frame(matrix(nrow=1,ncol=5));colnames(QC_table_check)= c('CL_trans_OL','LISI_CL_OL','outDist_LISI_CL_OL','outDist_LISI_CL_OLUP','outDist_LISI_CL_OLDN')


        gr = Intron_LISI_BS$CLIP_GR
        gr2 = Intron_LISI_BS$LISI_GR
        xo=as.data.frame(GenomicRanges::findOverlaps(gr,gr2,type = "any",ignore.strand=F))
          
        LISI_CLIP_VISH=gr2[xo$subjectHits]
                  # qh=as.data.frame(gr[xo$queryHits],row.names = NULL)
                  # colnames(qh)=paste0(colnames(qh),"_mm10")
                  # sh=as.data.frame(gr2[xo$subjectHits],row.names = NULL)
        CLIP_LISI_VISH=gr[xo$queryHits] 

QC_table_check[,'outDist_LISI_CL_OL']=length(unique(CLIP_LISI_VISH$ID ))    
       
        
        gr = Intron_LISI_BS$LISI_GR
        gr2 = Intron_LISI_BS$trans_GR
        xo=as.data.frame(GenomicRanges::findOverlaps(gr,gr2,type = "any",ignore.strand=F))
          
        Transc_CLIP_VISH=gr2[xo$subjectHits]
                  # qh=as.data.frame(gr[xo$queryHits],row.names = NULL)
                  # colnames(qh)=paste0(colnames(qh),"_mm10")
                  # sh=as.data.frame(gr2[xo$subjectHits],row.names = NULL)
        CLIP_Transc_VISH=gr[xo$queryHits]      
# intersect(LISI_OL$V1,as.data.frame(LISI_CLIP_VISH)$seqnames)
        QC_table_check[,'CL_trans_OL']=length(unique(CLIP_LISI_VISH$seqnames ))    


        
########################################################################################        
              
```

```{r ,echo=F,warning=F,eval=F,include=F}
        
########################
### Upstream
########################
if (recalcPT==T) {

pt_BS_up= permTest(
  A=Intron_LISI_BS$CLIP_GR_UP,
  B=Intron_LISI_BS$LISI_GR,
  genome=Intron_LISI_BS$genome,#mm10_mRNAome,
    mask=Intron_LISI_BS$exon_mask,
  randomize.function=randomizeRegions,
  evaluate.function=meanDistance,ntimes=pTntimes,count.once=TRUE,verbose=F)

################################# ################################# ################################# 


pt2_BS_up= overlapPermTest(
  A=Intron_LISI_BS$CLIP_GR_UP,
  B=Intron_LISI_BS$LISI_GR,
  genome=Intron_LISI_BS$genome,#mm10_mRNAome,
    mask=Intron_LISI_BS$exon_mask,
  ntimes=pTntimes,count.once=TRUE)
  # non.overlapping=T)

# LZ2_BS_up=localZScore(pt=pt2_BS_up,
#   A=Intron_LISI_BS$CLIP_GR_UP,
#   B=Intron_LISI_BS$LISI_GR,
#   genome=Intron_LISI_BS$genome,#mm10_mRNAome,
#     mask=Intron_LISI_BS$exon_mask,
#   window=500,
#   step=10,count.once=TRUE)

}

#############################################
{par(mar = c(5, 5, 5, 5)) # Set the margin on all sides to 6
plot(pt_BS_up)
mtext(paste0("LINE/SINE overlap with Intronic CLIP-100nt Upstream: ",USELISIstrand," Strand "))}
{par(mar = c(5, 5, 5, 5)) # Set the margin on all sides to 6
plot(pt2_BS_up)
mtext(paste0("LINE/SINE overlap with Intronic CLIP-100nt Upstream: ",USELISIstrand," Strand "))}

# {plot(LZ2_BS_up)
# mtext(paste0("LINE/SINE overlap with Intronic CLIP-100nt Upstream: ",USELISIstrand," Strand "))}
##############################################

        gr = Intron_LISI_BS$CLIP_GR_UP
        gr2 = Intron_LISI_BS$LISI_GR
        xo=as.data.frame(GenomicRanges::findOverlaps(gr,gr2,type = "any",ignore.strand=F))
          
        LISI_CLIP_VISHup=gr2[xo$subjectHits]
                  # qh=as.data.frame(gr[xo$queryHits],row.names = NULL)
                  # colnames(qh)=paste0(colnames(qh),"_mm10")
                  # sh=as.data.frame(gr2[xo$subjectHits],row.names = NULL)
        CLIP_LISI_VISHup=gr[xo$queryHits] 
        
        # length(unique(as.data.frame(CLIP_LISI_VISHup)$ID))
        # View(as.data.frame(CLIP_LISI_VISHup))
        
        QC_table_check[,'outDist_LISI_CL_OLUP']=length(unique(CLIP_LISI_VISHup$ID ))    

            
#############################################################################################################################################################        
### downstream
###########################        

if (recalcPT==T) {

pt_BS_dn= permTest(
  A=Intron_LISI_BS$CLIP_GR_DN,
  B=Intron_LISI_BS$LISI_GR,
  genome=Intron_LISI_BS$genome,#mm10_mRNAome,
    mask=Intron_LISI_BS$exon_mask,
  randomize.function=randomizeRegions,
  evaluate.function=meanDistance,ntimes=pTntimes,count.once=TRUE,verbose=F)

################################# ################################# ################################# 


pt2_BS_dn= overlapPermTest(
  A=Intron_LISI_BS$CLIP_GR_DN,
  B=Intron_LISI_BS$LISI_GR,
  genome=Intron_LISI_BS$genome,#mm10_mRNAome,
    mask=Intron_LISI_BS$exon_mask,
  # non.overlapping=T,
  ntimes=pTntimes,count.once=T)

# LZ2_dn=localZScore(pt=pt2_BS_dn,
#   A=Intron_LISI_BS$CLIP_GR_DN,
#   B=Intron_LISI_BS$LISI_GR,
#   genome=Intron_LISI_BS$genome,#mm10_mRNAome,
#     mask=Intron_LISI_BS$exon_mask,
#   window=500,
#   step=10,count.once=TRUE)

}

############################################################
{par(mar = c(5, 5, 5, 5)) # Set the margin on all sides to 6
plot(pt_BS_dn)
mtext(paste0("LINE/SINE overlap with Intronic CLIP-100nt Downstream: ",USELISIstrand," Strand "))}
{par(mar = c(5, 5, 5, 5)) # Set the margin on all sides to 6
plot(pt2_BS_dn)
mtext(paste0("LINE/SINE overlap with Intronic CLIP-100nt Downstream: ",USELISIstrand," Strand "))}
        
# {plot(LZ2_BS_dn)
# mtext(paste0("LINE/SINE overlap with Intronic CLIP-100nt Downstream: ",USELISIstrand," Strand "))}

        QC_table_PT[,'outDist_LISI_CL_OLUP']=pt2_BS_up[["numOverlaps"]][["observed"]]   
        QC_table_PT[,'outDist_LISI_CL_OLDN']=pt2_BS_dn[["numOverlaps"]][["observed"]] 
############################################################
        
        gr = Intron_LISI_BS$CLIP_GR_DN
        gr2 = Intron_LISI_BS$LISI_GR
        xo=as.data.frame(GenomicRanges::findOverlaps(gr,gr2,type = "any",ignore.strand=F))
          
        LISI_CLIP_VISHdn=gr2[xo$subjectHits]
                  # qh=as.data.frame(gr[xo$queryHits],row.names = NULL)
                  # colnames(qh)=paste0(colnames(qh),"_mm10")
                  # sh=as.data.frame(gr2[xo$subjectHits],row.names = NULL)
        CLIP_LISI_VISHdn=gr[xo$queryHits] 
           # length(unique(as.data.frame(CLIP_LISI_VISHdn)$seqnames))
        QC_table_check[,'outDist_LISI_CL_OLDN']=length(unique(CLIP_LISI_VISHdn$ID ))    
        
        
       QC_table_BS= rbind(Intron_LISI_BS$QC_table,QC_table_vish,QC_table_check,QC_table_PT)

 # setdiff(unique(CLIP_LISI_VISHup$ID),Intron_LISI_BS$LISIdist[(Intron_LISI_BS$LISIdist$CLIP_LISI_Dist<=0&Intron_LISI_BS$LISIdist$CLIP_LISI_Dist>-100),'CLIP_ID'])
 # setdiff(unique(CLIP_LISI_VISHup$ID),Intron_LISI_BS$LISIdist[(Intron_LISI_BS$LISIdist$CLIP_LISI_Dist<=100&Intron_LISI_BS$LISIdist$CLIP_LISI_Dist>-100),'CLIP_ID'])
 # setdiff(unique(CLIP_LISI_VISH$ID),Intron_LISI_BS$LISIdist[(Intron_LISI_BS$LISIdist$CLIP_LISI_Dist==0),'CLIP_ID'])
 
##############################
 ## the extra locations from the GR overlap that do not match the output from the Intron_LISI_BS$LISIdist table are locations that have LINE/SINEs < 100 nt away both upstream and downstream, they occure because the table only has one LISI per location however GR separatly identifies overlap upstrream and downstream
 
 # Intron_LISI_BS$LISIdist[which((Intron_LISI_BS$LISIdist$CLIP_LISI_Dist_UP<=0&Intron_LISI_BS$LISIdist$CLIP_LISI_Dist_UP>-100)&
                           # (Intron_LISI_BS$LISIdist$CLIP_LISI_Dist_DN>=0&Intron_LISI_BS$LISIdist$CLIP_LISI_Dist_DN<100)),'CLIP_ID']
 
 # View(Intron_LISI_BS$LISIdist)    
 
 
```

```{r ,echo=F,warning=F,eval=F,include=F}
####create table -Check
# intOL1= checkout2_LS1
# intOL2= checkout2_LS2
# intOL3= checkout2_LS3
# 
# unique(Intron_LISI_BS$CLIP_GR_UP$ID)
# unique(Intron_LISI_OS$CLIP_GR_UP$ID)
# unique(Intron_LISI_SS$CLIP_GR_UP$ID)
# 
# # bothstrands
# length(unique(isoout[(isoout$LISI_OLup%in%1)&(isoout$ID%in%intOL2$ID),"ID"])) #352
# isoout[(isoout$LISI_OLup%in%1)&(isoout$ID%in%intOL1$ID),]
# length(unique(isoout[(isoout$LISI_OLdn%in%1)&(isoout$ID%in%intOL2$ID),"ID"])) #368
# isoout[(isoout$LISI_OLdn%in%1)&(isoout$ID%in%intOL1$ID),]
# 
# 
# ###########################
# 
# length(unique(Intron_LISI_BS$LISIdist[(Intron_LISI_BS$LISIdist$CLIP_LISI_Dist)%in%c(0)|(Intron_LISI_BS$LISIdist$CLIP_LISI_Dist_UP)%in%c(-100:0),'CLIP_ID'])) #350
# length(unique(Intron_LISI_BS$LISIdist[(Intron_LISI_BS$LISIdist$CLIP_LISI_Dist)%in%c(0)|(Intron_LISI_BS$LISIdist$CLIP_LISI_Dist_DN)%in%c(100:0),'CLIP_ID'])) #366
# 
# (Intron_LISI_BS$LISIdist[(Intron_LISI_BS$LISIdist$CLIP_LISI_Dist_UP)%in%c(-100:0),])
# (Intron_LISI_BS$LISIdist[(Intron_LISI_BS$LISIdist$CLIP_LISI_Dist_DN)%in%c(100:0),])


############################
####create table 
colnames(Intron_LISI_BS$LISIdist)=paste0(colnames(Intron_LISI_BS$LISIdist),' BothStrand')

LISIdistTable=merge(Intron_LISI_SS$LISIdist[,colnames(Intron_LISI_SS$LISIdist)[colnames(Intron_LISI_SS$LISIdist)%in%c('ID_LINESINE')==F]],
                    Intron_LISI_OS$LISIdist[,colnames(Intron_LISI_OS$LISIdist)[colnames(Intron_LISI_OS$LISIdist)%in%c('Transcript','ID_LINESINE')==F]],
                    by="CLIP_ID",all=T,suffixes=c(" SameStrand"," OppositeStrand"))

LISIdistTable=merge(LISIdistTable,
                    Intron_LISI_BS$LISIdist[,colnames(Intron_LISI_BS$LISIdist)[colnames(Intron_LISI_BS$LISIdist)%in%c('Transcript BothStrand','ID_LINESINE BothStrand')==F]],
                    by.x="CLIP_ID",by.y="CLIP_ID BothStrand",all=T)

colnames(LISIdistTable)=gsub('repName_LINESINE','Closest repName',colnames(LISIdistTable))
colnames(LISIdistTable)=gsub('repClass_LINESINE','Closest repClass',colnames(LISIdistTable))
colnames(LISIdistTable)=gsub('ID_LINESINE','Closest LINESINE ID',colnames(LISIdistTable))
colnames(LISIdistTable)=gsub('CLIP_LISI_Dist_DN','Downstream LISI_Dist',colnames(LISIdistTable))
colnames(LISIdistTable)=gsub('CLIP_LISI_Dist_UP','Upstream LISI_Dist',colnames(LISIdistTable))
colnames(LISIdistTable)=gsub('CLIP_LISI_Dist','Closest LISI_Dist',colnames(LISIdistTable))

        write.table(LISIdistTable, file=paste0('./structure/chunkwindow/',CLIPtype,'/CLIP_Peaks_MD14_Introns_LINESINEdist.txt'), quote=F, sep="\t", row.names=T, col.names=NA)



```




### create mmRNAome  

```{r ,echo=F,warning=F,eval=F,include=F}
awk -F"\t" -v OFS="\t" '{print $1,$2,$3,$1":"$2"-"$3,".",$5}' TranscLINE_SINE.txt > TranscLINE_SINE.bed

cut -f1-3 TranscLineSINE.txt > TranscLINE_SINE.bed
bedtools intersect -wa -a transcLINE_SINE.bed -wb -b CLIP_introns_Coordinates.transcripts.bed > SINE_LINE_in_intron.transcripts.txt
while read a:do grep -m1@a transcLINE_SINE.bed ;done < tmp1 > tmp2

python make_mRNAome.py canonicalTranscripts_protein_coding.bed gencode.vM25.annotation.genePred mm10.fa mm10.mRNAome.fa mm10.mRNAome.genome
python convert_coordinates_premmRNAome.py transcripts.bed SINE_LINE_in_intron.transcripts.txt SINE_LINE_in_transcripts.new_coords.premmRNA.bed

################
# wget ftp://ftp.ebi.ac.uk/pub/databases/gencode/Gencode_mouse/release_M25/gencode.vM25.annotation.gtf.gz
# gzip -d gencode.vM25.annotation.gtf.gz
# module load ucsc
# gtfToGenePred gencode.vM25.annotation.gtf gencode.vM25.annotation.genePred
##### not on sinteractive
# mysql -h genome-mysql.soe.ucsc.edu -u genome -Ne "select name,chrom,txStart,txEnd,strand from wgEncodeGencodeBasicVM25" mm10|awk -v OFS="\t" '{print $2,$3,$4,$1,".",$5}'> wgEncodeGencodeBasicVM25.transcripts.bed
# grep -w 'chr1\|chr2|chr3\|chr4\|chr5\|chr6\|chr7\|chr8\|chr9\|chr10\|chr11\|chr12\|chr13\|chr14\|chr15\|chr16\|chr17\|chr18\|chr19\|chrX\|chrY\|chrM\|chr_rDNA' ../wgEncodeGencodeBasicVM25.transcripts.bed > ../transcripts.bed
# cp /data/RBL_NCI/Phil/Reference/fasta/mm10/GRCm38.primary_assembly.gencodeM23.genome.fa ./mm10.fa

tail -n +2 ExonicLINE_SINE.txt | awk -v OFS="\t" '{print $1,$2,$3,$1":"$2"_"$3,".",$5}' - > ExonicLINE_SINE_TEST.bed
tail -n +2 CLIPCoordinates.txt | awk -v OFS="\t" '{print $1,$2,$2,$1":"$2"_"$3,".",$3}' - > CLIPCoordinates.transcripts_TEST.bed
bedtools intersect -wa -a ExonicLINE_SINE_TEST.bed -wb -b CLIPCoordinates.transcripts_TEST.bed > ExonicLINE_SINE.transcripts_TEST.txt



# awk -v OFS="\t" '{print $2,$4,$5,$1,".",$3}' ../gencode.vM25.annotation.genePred > ../gencode.vM25_TEST.bed
# tail -n +2 ExonicCoordinates_Canonical.txt | awk -v OFS="\t" '{print $1,$4,$5,$15,".",$7}' - > ExonicCoordinates_Canonical_TEST.bed
tail -n +2 ExonicCoordinates_Canonical.txt | awk -v OFS="\t" '{print $15}' - > ExonicCoordinates_Canonical_TEST.lst

cut -f4 ExonicCoordinates_Canonical_TEST.bed | uniq | grep -v ../transcripts.bed > ExonicCoordinates.transcripts_TEST.txt




python ../make_mRNAome.v2.py ExonicCoordinates_Canonical_TEST.lst ../gencode.vM25.annotation.genePred /data/RBL_NCI/Phil/Reference/fasta/mm10/GRCm38.primary_assembly.gencodeM23.genome.fa mm10.mRNAome.fa mm10.mRNAome_TEST.genome

python ../convert_coordinates.py ExonicLINE_SINE.transcripts_TEST.txt ExonicLINE_SINE.transcripts_TEST.txt ExonicLINE_SINE.transcripts.new_coords.bed






#convert ExonicCoordinates_Canonical.txt to list of transcript ids
cut -f15 ExonicCoordinates_Canonical.txt |grep -v transcript_id|sort|uniq > ExonicCoordinates_Canonical.transcript_ids
#convert GTF to genePred
module load ucsc
gtfToGenePred gencode.vM25.annotation.gtf gencode.vM25.annotation.genePred
#run python script to make the .genome file
module load bedtools
python ../make_mRNAome.v2.py ExonicCoordinates_Canonical.transcript_ids ../gencode.vM25.annotation.genePred ../mm10.fa mm10.mRNAome.fa mm10.mRNAome.genome
#convert crosslink coordinatest to bed format
grep -v -i crosslink CLIP_crosslink.txt|awk -F"\t" -v OFS="\t" '{print $1,$2,$2+1,$1":"$2"-"$2+1,".",$3}' > CLIP_crosslink.bed
#convert exons from canonical transcripts to bed format
awk -F"\t" -v OFS="\t" '{if ($3=="exon"){print $1,$4,$5,$1":"$4"-"$5"|"$15,".",$7}}' ExonicCoordinates_Canonical.txt > ExonicCoordinates_Canonical.bed
#sort both bedfiles
bedSort CLIP_crosslink.bed CLIP_crosslink.bed
bedSort ExonicCoordinates_Canonical.bed ExonicCoordinates_Canonical.bed
#create input file for python script for converting the coordinates
bedtools intersect -wa -a CLIP_crosslink.bed -wb -b ExonicCoordinates_Canonical.bed > CLIP_crosslink_in_ExonicCoordinates_Canonical.txt
#convert the coordinates
python convert_coordinates.py gencode.vM25.annotation.genePred CLIP_crosslink_in_ExonicCoordinates_Canonical.txt CLIP_crosslink_in_ExonicCoordinates_Canonical.new_coords.bed


```

```{r ,echo=F,warning=F,eval=F,include=F}

## will include peaks that do not OL with intron on Same strand but does on opposite strand
checkout2_exon=Peaksdata4_chunkInExall[Peaksdata4_chunkInExall$`Both Strand: RNA_type_Classification`%in%c('protein_coding: exon'),] 


################################################################################        


checkout2_exon=merge(checkout2_exon,isoout_transcAll[,c('ID','USEtranscript','CL')],by="ID")
colnames(checkout2_exon)[colnames(checkout2_exon)%in%"USEtranscript"]='transcript_id'
colnames(checkout2_exon)[colnames(checkout2_exon)%in%"CL"]='crosslink'

checkout2_exon_XLnk=checkout2_exon[,c('chr','crosslink','strand','ID','transcript_id')]
Exon_LISI_SS=MakeTranscriptome(checkout2_exon_XLnk,window,"same")


```


## UTR


```{r ,echo=F,warning=F,eval=F,include=F}
# STOP
##################################
### prepare CLIP data
   checkout2_LS=Peaksdata4_chunkEx[(Peaksdata4_chunkEx$ID%in%isoout_transcAll[isoout_transcAll$TranscRegion%in%'3UTR','ID']),]
  
checkout2_LS=merge(checkout2_LS,isoout_transcAll[,c('ID','USEtranscript','CL')],by="ID")
colnames(checkout2_LS)[colnames(checkout2_LS)%in%"USEtranscript"]='transcript_id'
colnames(checkout2_LS)[colnames(checkout2_LS)%in%"CL"]='crosslink'


checkout2_LS_XLnk=checkout2_LS[,c('chr','crosslink','strand','ID','transcript_id')]

  write.table(checkout2_LS_XLnk, file=paste0(LINESINEout,'/UTRCLIPCoordinates.txt'), quote=F, sep="\t", row.names=F)

  
CrossLink_GR=GRanges(seqnames=checkout2_LS_XLnk$chr, ranges=IRanges(start=checkout2_LS_XLnk$crosslink,
end=checkout2_LS_XLnk$crosslink),
strand=checkout2_LS_XLnk$strand,
ID=checkout2_LS_XLnk$ID,
transcript_id=checkout2_LS_XLnk$transcript_id)  



#######################
rmsk_GRCm38=fread(paste0("/Users/homanpj/OneDrive - National Institutes of Health/Resources/ref/mm10/repeatmasker/rmsk_GRCm38.txt"), header=T, sep="\t",stringsAsFactors = F,data.table=F)
rmsk_GRCm38_LISI=rmsk_GRCm38[rmsk_GRCm38$repClass%in%c('LINE','SINE'),]
rmsk_GRCm38_LISI=rmsk_GRCm38_LISI[rmsk_GRCm38_LISI$genoName%in%c('chrUn_GL456239', 'chrUn_GL456359', 'chrUn_GL456360', 'chrUn_GL456366', 'chrUn_GL456367', 'chrUn_GL456368', 'chrUn_GL456370', 'chrUn_GL456372', 'chrUn_GL456378', 'chrUn_GL456379', 'chrUn_GL456381', 'chrUn_GL456382', 'chrUn_GL456385', 'chrUn_GL456387', 'chrUn_GL456389', 'chrUn_GL456393', 'chrUn_GL456394', 'chrUn_JH584304', 'chr1_GL456210_random', 'chr1_GL456211_random', 'chr1_GL456212_random', 'chr1_GL456213_random', 'chr1_GL456221_random', 'chr4_GL456216_random', 'chr4_GL456350_random', 'chr4_JH584292_random', 'chr4_JH584293_random', 'chr4_JH584294_random', 'chr5_GL456354_random', 'chr5_JH584296_random', 'chr5_JH584297_random', 'chr5_JH584298_random', 'chr5_JH584299_random', 'chr7_GL456219_random', 'chrX_GL456233_random', 'chrY_JH584300_random', 'chrY_JH584301_random', 'chrY_JH584302_random', 'chrY_JH584303_random')==F,]

LISI_GR = GRanges(seqnames=rmsk_GRCm38_LISI$genoName, 
                  ranges=IRanges(start=rmsk_GRCm38_LISI$genoStart,
                  end=rmsk_GRCm38_LISI$genoEnd),
                  strand=rmsk_GRCm38_LISI$strand,
                  repClass=rmsk_GRCm38_LISI$repClass,
                  repName=rmsk_GRCm38_LISI$repName
                  )

############################################################################################
### Get Protein coding Exons

  mm10all=fread("/Users/homanpj/OneDrive - National Institutes of Health/Resources/ref/mm10/Gencode_VM23/fromGencode/gencode.vM23.basic.annotation.gtf.txt", header=T, sep="\t",stringsAsFactors = F,data.table=F)
  # mm10all=fread("/data/RBL_NCI/Phil/Reference/gtf/mouse/mm10/Gencode_VM23/gencode.vM23.basic.annotation.gtf.txt", header=T, sep="\t",stringsAsFactors = F,data.table=F)
  mm10all$transcript_id=removeVersion(mm10all$transcript_id)
  mm10all$gene_id=removeVersion(mm10all$gene_id)
  mm10all_trans=mm10all[mm10all$feature%in%'transcript',]
mm10_trans=mm10all_trans[mm10all_trans$transcript_id%in%canonical$transcript,]
mm10_trans=mm10_trans[mm10_trans$transcript_type%in%'protein_coding',]

  mm10_exon=mm10all[mm10all$feature%in%c("exon"),]


  mm10_UTR=mm10all[mm10all$feature%in%'UTR',]
  mm10_3UTR=mm10_UTR[mm10_UTR$exon_number!=1,]
  mm10_3UTR=mm10_3UTR[mm10_3UTR$transcript_id%in%mm10_trans$transcript_id,  ]

# setdiff(checkout2_LS$transcript_id,exons$transcript_id)

    ###########################
    ### Make sure that Transcripts from Exonic CLIP peaks are selected. 
    ###########################
Uptrancripts=checkout2_LS$transcript_id
Uptrancripts_mm10_UTR=mm10_3UTR[mm10_3UTR$transcript_id%in%Uptrancripts,]

# length(unique(exons[exons$gene_id%in%unique(mm10_exon_Uptrancripts$gene_id)==T,'gene_id']))
# length(unique(mm10_exon_Uptrancripts$gene_id))

mm10_3UTR=rbind(mm10_3UTR[mm10_3UTR$gene_id%in%unique(Uptrancripts_mm10_UTR$gene_id)==F,],
            Uptrancripts_mm10_UTR)

# setdiff(checkout2_LS$transcript_id,exons$transcript_id)
  write.table(mm10_3UTR, file=paste0(LINESINEout,'/UTRCoordinates_Canonical.txt'), quote=F, sep="\t", row.names=F)

UTR_GR=GRanges(seqnames=mm10_3UTR$seqname, 
                 ranges=IRanges(start=mm10_3UTR$start,end=mm10_3UTR$end),
                 strand=mm10_3UTR$strand,
                 transcript_id=mm10_3UTR$transcript_id,
                 exon_number=mm10_3UTR$exon_number,
                 gene_id=mm10_3UTR$gene_id
                 )


# mm10_trans[duplicated(mm10_trans$transcript_id),]

############################################################################################

  mm10_trans=mm10all_trans[mm10all_trans$transcript_id%in%unique(mm10_3UTR$transcript_id),]

    write.table(mm10_trans, file=paste0(LINESINEout,'/TranscCoordinates_Canonical.txt'), quote=F, sep="\t", row.names=F)

 Transc_GR=GRanges(seqnames=mm10_trans$seqname,
                   ranges=IRanges(start=mm10_trans$start,end=mm10_trans$end),
                   strand=mm10_trans$strand,
                   transcript_id=mm10_trans$transcript_id,
                   gene_id=mm10_trans$gene_id
                   )
############################################################################################
mm10_exon=mm10_exon[mm10_exon$transcript_id%in%unique(mm10_3UTR$transcript_id),]

    write.table(mm10_exon, file=paste0(LINESINEout,'/ExonCoordinates_Canonical.txt'), quote=F, sep="\t", row.names=F)

 exon_GR=GRanges(seqnames=mm10_exon$seqname,
                   ranges=IRanges(start=mm10_exon$start,end=mm10_exon$end),
                   strand=mm10_exon$strand,
                   transcript_id=mm10_exon$transcript_id,
                   gene_id=mm10_exon$gene_id
                   )

############################################################################################
# exons_GR=c(introns_GR,exons_GR)

### Get LINE/SINES that overlap with Exons
        gr = UTR_GR
        gr2 = LISI_GR
        xo=as.data.frame(GenomicRanges::findOverlaps(gr,gr2,type = "any",ignore.strand=F))
          
        LISI_exn_GR=gr2[xo$subjectHits]
                  # qh=as.data.frame(gr[xo$queryHits],row.names = NULL)
                  # colnames(qh)=paste0(colnames(qh),"_mm10")
                  # sh=as.data.frame(gr2[xo$subjectHits],row.names = NULL)
        exn_LISI_GR=gr[xo$queryHits]

        
  write.table(as.data.frame(LISI_exn_GR), file=paste0(LINESINEout,'/UTRLINE_SINE.txt'), quote=F, sep="\t", row.names=F)
          
#
### Get LINE/SINES that overlap with CLIP peaks
        gr = CrossLink_GR
        gr2 = LISI_GR
        xo=as.data.frame(GenomicRanges::findOverlaps(gr,gr2,type = "any",ignore.strand=F))
          
        LISI_CrossLink_GR=gr2[xo$subjectHits]
                  # qh=as.data.frame(gr[xo$queryHits],row.names = NULL)
                  # colnames(qh)=paste0(colnames(qh),"_mm10")
                  # sh=as.data.frame(gr2[xo$subjectHits],row.names = NULL)
        Crosslink_LISI_GR=gr[xo$queryHits]
        
        

#convert ExonicCoordinates_Canonical.txt to list of transcript ids
cut -f15 ExonCoordinates_Canonical.txt |grep -v transcript_id|sort|uniq > ExonCoordinates_Canonical.transcript_ids
#convert GTF to genePred
module load ucsc
gtfToGenePred gencode.vM25.annotation.gtf gencode.vM25.annotation.genePred
#run python script to make the .genome file
module load bedtools
python ../make_mRNAome.v2.py ExonCoordinates_Canonical.transcript_ids ../gencode.vM25.annotation.genePred ../mm10.fa mm10.ExonmRNAome.fa mm10.ExonmRNAome.genome
#convert crosslink coordinatest to bed format
grep -v -i crosslink UTRCLIPCoordinates.txt|awk -F"\t" -v OFS="\t" '{print $1,$2,$2+1,$1":"$2"-"$2+1,".",$3}' > UTRCLIPCoordinates.bed
#convert exons from canonical transcripts to bed format
awk -F"\t" -v OFS="\t" '{if ($3=="exon"){print $1,$4,$5,$1":"$4"-"$5"|"$15,".",$7}}' ExonCoordinates_Canonical.txt > ExonCoordinates_Canonical.bed
#sort both bedfiles
bedSort UTRCLIPCoordinates.bed UTRCLIPCoordinates.bed
bedSort ExonCoordinates_Canonical.bed ExonCoordinates_Canonical.bed
#create input file for python script for converting the coordinates
bedtools intersect -wa -a UTRCLIPCoordinates.bed -wb -b ExonCoordinates_Canonical.bed > UTRCLIPCoordinates_in_ExonCoordinates_Canonical.txt
#convert the coordinates
python convert_coordinates.py gencode.vM25.annotation.genePred CLIP_crosslink_in_ExonicCoordinates_Canonical.txt CLIP_crosslink_in_ExonicCoordinates_Canonical.new_coords.bed

    
    
LISIdist=fread(paste0("./structure/LINESINE/for_PhilHoman2/for_PhilHoman/CLIP_crosslink_canonicalTranscripts_protein_coding_intersect.filtered.new_coords.closest_SINE_LINE.txt"), header=F, sep="\t",stringsAsFactors = F,data.table=F)
      LISIdist$V1=removeVersion(LISIdist$V1)

mean(LISIdist$V8)    
mean(abs(LISIdist$V8))    
mean(LISIdist$V9)    
mean(abs(LISIdist$V9))  

nrow(LISIdist[LISIdist$V13==0,])

LISI_OL=LISIdist[LISIdist$V13==0,]

length(unique((Crosslink_LISI_GR$transcript_id)))



```  


## old  

```{r ,echo=F,warning=F,eval=F,include=F}
# STOP
##################################
### prepare CLIP data
  
# BiocManager::install("regioneR")
library(regioneR)

checkout2_LS=separate(checkout2,CLIP_ID,into = c('chr','start'),sep=':',remove = F)
checkout2_LS=separate(checkout2_LS,start,into = c('start','end'),sep='-')
checkout2_LS$start=as.numeric(checkout2_LS$start)
checkout2_LS$end=as.numeric(checkout2_LS$end)

checkout2_LS=separate(checkout2,n,into = c('chr','start'),sep=':',remove = F)
checkout2_LS=separate(checkout2_LS,start,into = c('start','end'),sep='-')
checkout2_LS$start=as.numeric(checkout2_LS$start)
checkout2_LS$end=as.numeric(checkout2_LS$end)

checkout2_LS_XLnk=checkout2_LS[,c('chr','crosslink','strand','ID','transcript_id')]
  write.table(checkout2_LS_XLnk, file=paste0(LINESINEout,'/CLIPCoordinates.txt'), quote=F, sep="\t", row.names=F)

  
CrossLink_GR=GRanges(seqnames=checkout2_LS_XLnk$chr, ranges=IRanges(start=checkout2_LS_XLnk$crosslink,
end=checkout2_LS_XLnk$crosslink),
strand=checkout2_LS_XLnk$strand,
ID=checkout2_LS_XLnk$ID,
transcript_id=checkout2_LS_XLnk$transcript_id)  

# peaks_GR = GRanges(seqnames=checkout2_LS$chr, ranges=IRanges(start=checkout2_LS$start,
# end=checkout2_LS$end),
# strand=checkout2_LS$strand,exon_number=checkout2_LS$`exon#`)

# peaks2_GR = GRanges(seqnames=checkout2_LS$chr, ranges=IRanges(start=checkout2_LS$end,
# end=checkout2_LS$end+10000),
# strand=checkout2_LS$strand,
# exon_number=checkout2_LS$`exon#`,lastexon=checkout2_LS$lastexon,numberTranscript=checkout2_LS$numberTranscript)



#######################
rmsk_GRCm38=fread(paste0("/Users/homanpj/OneDrive - National Institutes of Health/Resources/ref/mm10/repeatmasker/rmsk_GRCm38.txt"), header=T, sep="\t",stringsAsFactors = F,data.table=F)
rmsk_GRCm38_LISI=rmsk_GRCm38[rmsk_GRCm38$repClass%in%c('LINE','SINE'),]
rmsk_GRCm38_LISI=rmsk_GRCm38_LISI[rmsk_GRCm38_LISI$genoName%in%c('chrUn_GL456239', 'chrUn_GL456359', 'chrUn_GL456360', 'chrUn_GL456366', 'chrUn_GL456367', 'chrUn_GL456368', 'chrUn_GL456370', 'chrUn_GL456372', 'chrUn_GL456378', 'chrUn_GL456379', 'chrUn_GL456381', 'chrUn_GL456382', 'chrUn_GL456385', 'chrUn_GL456387', 'chrUn_GL456389', 'chrUn_GL456393', 'chrUn_GL456394', 'chrUn_JH584304', 'chr1_GL456210_random', 'chr1_GL456211_random', 'chr1_GL456212_random', 'chr1_GL456213_random', 'chr1_GL456221_random', 'chr4_GL456216_random', 'chr4_GL456350_random', 'chr4_JH584292_random', 'chr4_JH584293_random', 'chr4_JH584294_random', 'chr5_GL456354_random', 'chr5_JH584296_random', 'chr5_JH584297_random', 'chr5_JH584298_random', 'chr5_JH584299_random', 'chr7_GL456219_random', 'chrX_GL456233_random', 'chrY_JH584300_random', 'chrY_JH584301_random', 'chrY_JH584302_random', 'chrY_JH584303_random')==F,]

LISI_GR = GRanges(seqnames=rmsk_GRCm38_LISI$genoName, 
                  ranges=IRanges(start=rmsk_GRCm38_LISI$genoStart,
                  end=rmsk_GRCm38_LISI$genoEnd),
                  strand=rmsk_GRCm38_LISI$strand,
                  repClass=rmsk_GRCm38_LISI$repClass,
                  repName=rmsk_GRCm38_LISI$repName
                  )

############################################################################################
### Get Protein coding Exons

  mm10all=fread("/Users/homanpj/OneDrive - National Institutes of Health/Resources/ref/mm10/Gencode_VM23/fromGencode/gencode.vM23.basic.annotation.gtf.txt", header=T, sep="\t",stringsAsFactors = F,data.table=F)
  # mm10all=fread("/data/RBL_NCI/Phil/Reference/gtf/mouse/mm10/Gencode_VM23/gencode.vM23.basic.annotation.gtf.txt", header=T, sep="\t",stringsAsFactors = F,data.table=F)
  mm10all$transcript_id=removeVersion(mm10all$transcript_id)
  mm10all$gene_id=removeVersion(mm10all$gene_id)
  mm10_trans=mm10all[mm10all$feature%in%'transcript',]
  mm10_UTR=mm10all[mm10all$feature%in%'UTR',]
  mm10_exon=mm10all[mm10all$feature%in%c("exon"),]

exons=mm10_exon[mm10_exon$transcript_id%in%canonical$transcript,]
exons=exons[exons$transcript_type%in%'protein_coding',]


# setdiff(checkout2_LS$transcript_id,exons$transcript_id)

    ###########################
    ### Make sure that Transcripts from Exonic CLIP peaks are selected. 
    ###########################
Uptrancripts=checkout2_LS$transcript_id
Uptrancripts_mm10_exon=mm10_exon[mm10_exon$transcript_id%in%Uptrancripts,]

# length(unique(exons[exons$gene_id%in%unique(mm10_exon_Uptrancripts$gene_id)==T,'gene_id']))
# length(unique(mm10_exon_Uptrancripts$gene_id))

exons=rbind(exons[exons$gene_id%in%unique(Uptrancripts_mm10_exon$gene_id)==F,],
            Uptrancripts_mm10_exon)

# setdiff(checkout2_LS$transcript_id,exons$transcript_id)

exons_GR=GRanges(seqnames=exons$seqname, 
                 ranges=IRanges(start=exons$start,end=exons$end),
                 strand=exons$strand,
                 transcript_id=exons$transcript_id,
                 exon_number=exons$exon_number,
                 gene_id=exons$gene_id
                 )

  write.table(exons, file=paste0(LINESINEout,'/ExonicCoordinates_Canonical.txt'), quote=F, sep="\t", row.names=F)

# mm10_trans[duplicated(mm10_trans$transcript_id),]

############################################################################################

  TranscUTR=mm10_UTR[mm10_UTR$transcript_id%in%exons$transcript_id,]

 TranscUTR_GR=GRanges(seqnames=TranscUTR$seqname,
                   ranges=IRanges(start=TranscUTR$start,end=TranscUTR$end),
                   strand=TranscUTR$strand,
                   transcript_id=TranscUTR$transcript_id,
                   gene_id=TranscUTR$gene_id
                   )
############################################################################################

introns=fread("/Users/homanpj/OneDrive - National Institutes of Health/Resources/ref/mm10/Gencode_VM23/fromUCSC/KnownGene/KnownGene_GRCm38_introns.bed", header=F, sep="\t",stringsAsFactors = F,data.table=F)
  introns=introns[grepl("_",introns$V1)==F,]
  colnames(introns)=c('chr','start','end','attribute','V5','strand')
  introns=separate(introns,attribute,into=c('transcript_id','feature','exon_number','level','chr2','intronnumber','dir'),remove = T,sep = "_")
  introns$transcript_id=removeVersion(introns$transcript_id)
  # introns_canon=introns[introns$transcript_id%in%canonical$transcript_id,]
  introns$start=introns$start+1
  introns$exon_number=as.numeric(introns$exon_number)+1

 introns=introns[introns$transcript_id%in%exons$transcript_id,]
 
   # setdiff(exons$transcript_id,introns$transcript_id)
   # intersect(checkout2_LS$transcript_id,exons$transcript_id)
   # intersect(checkout2_LS$transcript_id,introns$transcript_id)
   # intersect(checkout2_LS$transcript_id,TranscUTR$transcript_id)


  write.table(introns, file=paste0(LINESINEout,'/IntronsCoordinates_Canonical.txt'), quote=F, sep="\t", row.names=F)


introns_GR=GRanges(seqnames=introns$chr,
                   ranges=IRanges(start=introns$start,end=introns$end),
                   strand=introns$strand,
                   transcript_id=introns$transcript_id,
                   exon_number=introns$exon_number
                   )



  intronsExons=mm10_trans[mm10_trans$transcript_id%in%exons$transcript_id,]


intronsExons_GR=GRanges(seqnames=intronsExons$seqname,
                   ranges=IRanges(start=intronsExons$start,end=intronsExons$end),
                   strand=intronsExons$strand,
                   transcript_id=intronsExons$transcript_id,
                   gene_id=intronsExons$gene_id
                   )

  write.table(intronsExons, file=paste0(LINESINEout,'TranscCoordinates_Canonical.txt'), quote=F, sep="\t", row.names=F)



############################################################################################
# exons_GR=c(introns_GR,exons_GR)

### Get LINE/SINES that overlap with Exons
        gr = exons_GR
        gr2 = LISI_GR
        xo=as.data.frame(GenomicRanges::findOverlaps(gr,gr2,type = "any",ignore.strand=F))
          
        LISI_exn_GR=gr2[xo$subjectHits]
                  # qh=as.data.frame(gr[xo$queryHits],row.names = NULL)
                  # colnames(qh)=paste0(colnames(qh),"_mm10")
                  # sh=as.data.frame(gr2[xo$subjectHits],row.names = NULL)
        exn_LISI_GR=gr[xo$queryHits]

        
  write.table(as.data.frame(LISI_exn_GR), file=paste0(LINESINEout,'/ExonicLINE_SINE.txt'), quote=F, sep="\t", row.names=F)
          
  
### Get LINE/SINES that overlap with Introns
        gr = introns_GR
        gr2 = LISI_GR
        xo=as.data.frame(GenomicRanges::findOverlaps(gr,gr2,type = "any",ignore.strand=F))
          
        LISI_intrns_GR=gr2[xo$subjectHits]
                  # qh=as.data.frame(gr[xo$queryHits],row.names = NULL)
                  # colnames(qh)=paste0(colnames(qh),"_mm10")
                  # sh=as.data.frame(gr2[xo$subjectHits],row.names = NULL)
        intrns_LISI_GR=gr[xo$queryHits]

        
  write.table(as.data.frame(LISI_intrns_GR), file=paste0(LINESINEout,'/IntronicLINE_SINE.txt'), quote=F, sep="\t", row.names=F)
          
  
### Get LINE/SINES that overlap with Transcript
        gr = intronsExons_GR
        gr2 = LISI_GR
        xo=as.data.frame(GenomicRanges::findOverlaps(gr,gr2,type = "any",ignore.strand=F))
          
        LISI_Transc_GR=gr2[xo$subjectHits]
                  # qh=as.data.frame(gr[xo$queryHits],row.names = NULL)
                  # colnames(qh)=paste0(colnames(qh),"_mm10")
                  # sh=as.data.frame(gr2[xo$subjectHits],row.names = NULL)
        Transc_LISI_GR=gr[xo$queryHits]

        
  write.table(as.data.frame(LISI_Transc_GR), file=paste0(LINESINEout,'/TrasncLINE_SINE.txt'), quote=F, sep="\t", row.names=F)
          
  
#
### Get LINE/SINES that overlap with CLIP peaks
        gr = CrossLink_GR
        gr2 = LISI_GR
        xo=as.data.frame(GenomicRanges::findOverlaps(gr,gr2,type = "any",ignore.strand=F))
          
        LISI_CrossLink_GR=gr2[xo$subjectHits]
                  # qh=as.data.frame(gr[xo$queryHits],row.names = NULL)
                  # colnames(qh)=paste0(colnames(qh),"_mm10")
                  # sh=as.data.frame(gr2[xo$subjectHits],row.names = NULL)
        Crosslink_LISI_GR=gr[xo$queryHits]
        
        
### Get UTR that overlap with CLIP peaks
        gr = CrossLink_GR
        gr2 = TranscUTR_GR
        xo=as.data.frame(GenomicRanges::findOverlaps(gr,gr2,type = "any",ignore.strand=F))
          
        UTR_CrossLink_GR=gr2[xo$subjectHits]
                  # qh=as.data.frame(gr[xo$queryHits],row.names = NULL)
                  # colnames(qh)=paste0(colnames(qh),"_mm10")
                  # sh=as.data.frame(gr2[xo$subjectHits],row.names = NULL)
        Crosslink_UTR_GR=gr[xo$queryHits]  
    Crosslink_UTR_GR_DF=as.data.frame(Crosslink_UTR_GR)
    Crosslink_UTR_GR_DF$UTR='UTR'
    
    checkout2$UTR=NA
    checkout2[which(checkout2$ID%in%Crosslink_UTR_GR_DF$ID),'UTR']="UTR"

    # checkout2=checkout2[,(colnames(checkout2)%in%"UTR")==F] 
    # Crosslink_UTR_GR_DF=as.data.frame(Crosslink_UTR_GR)
    # Crosslink_UTR_GR_DF$UTR='UTR' 
    # checkout2=merge(checkout2, Crosslink_UTR_GR_DF[,c('ID','UTR')],by="ID",all.x=T)
    

### Get Transc that overlap with CLIP peaks
        gr = CrossLink_GR
        gr2 = intronsExons_GR
        xo=as.data.frame(GenomicRanges::findOverlaps(gr,gr2,type = "any",ignore.strand=F))
          
        Transc_CrossLink_GR=gr2[xo$subjectHits]
                  # qh=as.data.frame(gr[xo$queryHits],row.names = NULL)
                  # colnames(qh)=paste0(colnames(qh),"_mm10")
                  # sh=as.data.frame(gr2[xo$subjectHits],row.names = NULL)
        Crosslink_Transc_GR=gr[xo$queryHits]  
    Crosslink_Transc_GR_DF=as.data.frame(Crosslink_Transc_GR)
    Crosslink_Transc_GR_DF$Transc='Transc'
    
    checkout2$Transc=NA
    checkout2[which(checkout2$ID%in%Crosslink_Transc_GR_DF$ID),'Transc']="Transc"
    length(unique(checkout2$ID))

    # checkout2=checkout2[,(colnames(checkout2)%in%"UTR")==F] 
    # Crosslink_UTR_GR_DF=as.data.frame(Crosslink_UTR_GR)
    # Crosslink_UTR_GR_DF$UTR='UTR' 
    # checkout2=merge(checkout2, Crosslink_UTR_GR_DF[,c('ID','UTR')],by="ID",all.x=T)
    
############################################################################################        
#         gr = peaks2_GR
#         gr2 = LISI_exn_GR
#         xo=as.data.frame(GenomicRanges::findOverlaps(gr,gr2,type = "any",ignore.strand=F))
#           
#         LISI_exn_peaks2_GR=gr2[xo$subjectHits]
#                   # qh=as.data.frame(gr[xo$queryHits],row.names = NULL)
#                   # colnames(qh)=paste0(colnames(qh),"_mm10")
#                   # sh=as.data.frame(gr2[xo$subjectHits],row.names = NULL)
#         peaks2_LISI_exn_GR=gr[xo$queryHits]
# peaks2_LISI_exn_GR
# View(as.data.frame(peaks2_LISI_exn_GR))
    
    
    
LISIdist=fread(paste0("./structure/LINESINE/for_PhilHoman2/for_PhilHoman/CLIP_crosslink_canonicalTranscripts_protein_coding_intersect.filtered.new_coords.closest_SINE_LINE.txt"), header=F, sep="\t",stringsAsFactors = F,data.table=F)
      LISIdist$V1=removeVersion(LISIdist$V1)

mean(LISIdist$V8)    
mean(abs(LISIdist$V8))    
mean(LISIdist$V9)    
mean(abs(LISIdist$V9))  

nrow(LISIdist[LISIdist$V13==0,])

LISI_OL=LISIdist[LISIdist$V13==0,]

length(unique((Crosslink_LISI_GR$transcript_id)))



```  


```{r ,echo=F,warning=F,eval=F,include=F}
        remove(mm10_mRNAome,CLIP_mRNAome,CLIP_mRNAome_up,CLIP_mRNAome_dn,CLIP_mRNAome_GR,CLIP_mRNAome_GRstrand,LISI_mRNAome,LISI_mRNAome_GR,LISI_mRNAome_GRstrand)
        # remove(mm10_CLIP_mRNAome_transc,CLIP_CLIP_mRNAome_transc,CLIP_CLIP_mRNAome_transcUP,CLIP_CLIP_mRNAome_transcDN,CLIP_CLIP_mRNAome_transc_GR,CLIP_CLIP_mRNAome_transc_GRstrand,LISI_CLIP_mRNAome_transc,LISI_CLIP_mRNAome_transc_GR,LISI_CLIP_mRNAome_transc_GRstrand)   

##################################
### After convert Genome to Transcriptome coordinates
  
mm10_mRNAome=fread(paste0("./structure/LINESINE/for_PhilHoman2/for_PhilHoman/mm10.mRNAome.genome"), header=F, sep="\t",stringsAsFactors = F,data.table=F)
      mm10_mRNAome$V1=removeVersion(mm10_mRNAome$V1)

############################################################################################################      
      # setdiff(mm10_mRNAome1$V1,mm10_mRNAome$V1)

CLIP_mRNAome=fread(paste0("./structure/LINESINE/for_PhilHoman2/for_PhilHoman/CLIP_crosslink_in_ExonicCoordinates_Canonical.new_coords.bed"), header=F, sep="\t",stringsAsFactors = F,data.table=F)
      CLIP_mRNAome$V1=removeVersion(CLIP_mRNAome$V1)
CLIP_mRNAome=merge(CLIP_mRNAome,mm10_trans[,c("transcript_id","gene_id","strand")],by.x="V1",by.y='transcript_id')
    CLIP_mRNAome[( CLIP_mRNAome$V3-CLIP_mRNAome$V2 >1)&(CLIP_mRNAome$strand=="+"),'V3']=CLIP_mRNAome[( CLIP_mRNAome$V3-CLIP_mRNAome$V2 >1)&(CLIP_mRNAome$strand=="+"),'V2']+1
    CLIP_mRNAome[( CLIP_mRNAome$V3-CLIP_mRNAome$V2 >1)&(CLIP_mRNAome$strand=="-"),'V2']=CLIP_mRNAome[( CLIP_mRNAome$V3-CLIP_mRNAome$V2 >1)&(CLIP_mRNAome$strand=="-"),'V3']-1
  
    
checkout2_LISI=checkout2
checkout2_LISI=separate(checkout2_LISI,CLIP_ID,into = c('chr','start'),sep = ":",remove = F)
checkout2_LISI=separate(checkout2_LISI,start,into = c('start','end'),sep = "-",remove = T)
checkout2_LISI$Region_ID=paste0(checkout2_LISI$chr,":",checkout2_LISI$crosslink,"-",(checkout2_LISI$crosslink+1))

CLIP_mRNAome=CLIP_mRNAome[CLIP_mRNAome$V4%in%checkout2_LISI$Region_ID,]

# CLIP_mRNAome[CLIP_mRNAome$V4%in%checkout2_LISI$Region_ID==F,]
# checkout2_LISI$Region_ID2=paste0(checkout2_LISI$chr,":",checkout2_LISI$crosslink-1,"-",(checkout2_LISI$crosslink))
# checkout2_LISI$Region_ID3=paste0(checkout2_LISI$chr,":",checkout2_LISI$crosslink-2,"-",(checkout2_LISI$crosslink-1))
# checkout2_LISI$Region_ID4=paste0(checkout2_LISI$chr,":",checkout2_LISI$crosslink+1,"-",(checkout2_LISI$crosslink+2))
# CLIP_mRNAome[CLIP_mRNAome$V4%in%checkout2_LISI$Region_ID2,]
# CLIP_mRNAome[CLIP_mRNAome$V4%in%checkout2_LISI$Region_ID3,]
# CLIP_mRNAome[CLIP_mRNAome$V4%in%checkout2_LISI$Region_ID4,]
# chr10:98915198-98915199	chr7:19006283-19006284	chr5:110105007-110105008	chr9:57504110-57504111	chr9:58220066-58220067	chr10:75643989-75643990	chr15:100280858-100280859	chr1:180681123-180681124	
    
    
    SetD= setdiff(CLIP_mRNAome$V1,(mm10_mRNAome)$V1)    
    CLIP_mRNAome=CLIP_mRNAome[CLIP_mRNAome$V1%in%SetD==F,]

       CLIP_mRNAome_GRstrand=GRanges(seqnames=CLIP_mRNAome$V1, 
                          ranges=IRanges(start=CLIP_mRNAome$V2,end=CLIP_mRNAome$V3),
                          strand=CLIP_mRNAome$strand,
                          ID=CLIP_mRNAome$V4
                          # chrm=CLIP_mRNAome$seqname
                          )

######################################################################################################################## 

 LISI_mRNAome=fread(paste0("./structure/LINESINE/for_PhilHoman2/for_PhilHoman/SINE_LINE_in_ExonicCoordinates_Canonical.new_coords.bed"), header=F, sep="\t",stringsAsFactors = F,data.table=F)
      LISI_mRNAome$V1=removeVersion(LISI_mRNAome$V1)
      LISI_mRNAome=merge(LISI_mRNAome,mm10_trans[,c("transcript_id","gene_id","strand")],by.x="V1",by.y='transcript_id')
rmsk_GRCm38_LISI$rmskID=paste0(rmsk_GRCm38_LISI$genoName,':',rmsk_GRCm38_LISI$genoStart,'-',rmsk_GRCm38_LISI$genoEnd)
      LISI_mRNAome= merge(LISI_mRNAome,rmsk_GRCm38_LISI[,c("rmskID","strand")],by.x="V4",by.y='rmskID',all.x=T,suffixes=c('_Transc','_rmks'))
      LISI_mRNAome=LISI_mRNAome[LISI_mRNAome$strand_rmks==LISI_mRNAome$strand_Transc,]
      LISI_mRNAome=LISI_mRNAome[duplicated(LISI_mRNAome)==F,]
      # LISI_mRNAome=merge(LISI_mRNAome,mm10_exon[,c("transcript_id","strand","seqname")],by.x='V1',by.y="transcript_id",all.x=T,no.dups = TRUE)
      # LISI_mRNAome=LISI_mRNAome[duplicated(LISI_mRNAome)==F,]
      # LISI_mRNAome[is.na(LISI_mRNAome$strand),'strand']="*"
      # LISI_mRNAome=LISI_mRNAome[LISI_mRNAome$V1%in%CLIP_mRNAome$V1,]
      
          SetD= setdiff(LISI_mRNAome$V1,(mm10_mRNAome)$V1)    
    LISI_mRNAome=LISI_mRNAome[LISI_mRNAome$V1%in%SetD==F,]


       LISI_mRNAome_GRstrand=GRanges(seqnames=LISI_mRNAome$V1, 
                          ranges=IRanges(start=LISI_mRNAome$V2,end=LISI_mRNAome$V3),
                          strand=LISI_mRNAome$strand_Transc,
                          ID=LISI_mRNAome$V4
                          # chrm=LISI_mRNAome$seqname
                          )     

################################################################################################################################################################       
       

       
########################################      
##### Adjusted Coordinates

# randomizeRegions(A=CLIP_mRNAome_GR,genome= mm10_mRNAome,allow.overlaps = F,per.chromosome = F,mask = F)

# pt= permTest(
#   A=CLIP_mRNAome_GR,
#   B=LISI_mRNAome_GR,
#   genome=mm10_mRNAome,
#   randomize.function=randomizeRegions,
#   evaluate.function=meanDistance,ntimes=100,count.once=TRUE,verbose=F)
# plot(pt)
# summary(pt)
# plot(pt, plotType="Tailed")
# plot(pt, plotType="area")
# plotRegions(list(CLIP_mRNAome_GR,LISI_mRNAome_GR),chromosome = 'ENSMUST00000228543')

        # setdiff(as.data.frame(CLIP_mRNAome_GRstrand)$seqnames,as.data.frame(mm10_mRNAome)$V1)    
        # setdiff(as.data.frame(LISI_mRNAome_GRstrand)$seqnames,as.data.frame(mm10_mRNAome)$V1)   
        # 
        # setdiff(as.data.frame(mm10_mRNAome)$V1,as.data.frame(CLIP_mRNAome_GRstrand)$seqnames)    
        # setdiff(as.data.frame(mm10_mRNAome)$V1,as.data.frame(LISI_mRNAome_GRstrand)$seqnames)   


pt= permTest(
  A=CLIP_mRNAome_GRstrand,
  B=LISI_mRNAome_GRstrand,
  genome=mm10_mRNAome,
  randomize.function=randomizeRegions,
  evaluate.function=meanDistance,ntimes=1000,count.once=TRUE,verbose=F)


################################# ################################# ################################# 

# pt2= overlapPermTest(
#   A=CLIP_mRNAome_GR,
#   B=LISI_mRNAome_GR,
#   genome=mm10_mRNAome,
#   non.overlapping=T,
#   ntimes=100,count.once=TRUE)
# plot(pt2)


pt2= overlapPermTest(
  A=CLIP_mRNAome_GRstrand,
  B=LISI_mRNAome_GRstrand,
  genome=mm10_mRNAome,
  non.overlapping=T,
  ntimes=1000,count.once=TRUE)

    
# LZ2=localZScore(pt=pt2,
#   A=CLIP_mRNAome_GR, 
#   B=LISI_mRNAome_GR, 
#   genome=mm10_mRNAome,
#   window=2000,
#   step=10,count.once=TRUE)
# plot(LZ2)

plot(pt)
plot(pt2)

        gr = CLIP_mRNAome_GRstrand
        gr2 = LISI_mRNAome_GRstrand
        xo=as.data.frame(GenomicRanges::findOverlaps(gr,gr2,type = "any",ignore.strand=F))
          
        CLIP_LISI_VISH=gr2[xo$subjectHits]
                  # qh=as.data.frame(gr[xo$queryHits],row.names = NULL)
                  # colnames(qh)=paste0(colnames(qh),"_mm10")
                  # sh=as.data.frame(gr2[xo$subjectHits],row.names = NULL)
        LISI_CLIP_VISH=gr[xo$queryHits] 

# intersect(LISI_OL$V1,as.data.frame(LISI_CLIP_VISH)$seqnames)

# intersect(CLIP_mRNAome$V1,LISI_mRNAome$V1)
# grep('ENSMUST00000002914',LISI_mRNAome$V1)
```

```{r ,echo=F,warning=F,eval=F,include=F}

### Upstream
CLIP_mRNAome_up=CLIP_mRNAome
      CLIP_mRNAome_up[,'V2']=CLIP_mRNAome_up[,'V3']-window
      
      CLIP_mRNAome_up[(CLIP_mRNAome_up$V2<1),'V2']=1
      CLIP_mRNAome_up[(CLIP_mRNAome_up$V3>CLIP_mRNAome_up$V2_Genome),'V3']= CLIP_mRNAome_up[(CLIP_mRNAome_up$V3>CLIP_mRNAome_up$V2_Genome),'V2_Genome']
        

  

       CLIP_mRNAome_GRstrand=GRanges(seqnames=CLIP_mRNAome_up$V1, 
                          ranges=IRanges(start=CLIP_mRNAome_up$V2,end=CLIP_mRNAome_up$V3),
                          strand=CLIP_mRNAome_up$strand,
                          ID=CLIP_mRNAome_up$V4
                          # chrm=CLIP_mRNAome_up$seqname
                          )
 setdiff(CLIP_mRNAome_up$V1,mm10_mRNAome$V1)    


 
LISI_mRNAome=fread(paste0("./structure/LINESINE/for_PhilHoman2/for_PhilHoman/SINE_LINE_in_canonicalTranscripts_protein_coding.exonic.new_coords.bed"), header=F, sep="\t",stringsAsFactors = F,data.table=F)
      LISI_mRNAome$V1=removeVersion(LISI_mRNAome$V1)
      LISI_mRNAome=merge(LISI_mRNAome,mm10_trans[,c("transcript_id","gene_id","strand")],by.x="V1",by.y='transcript_id')
rmsk_GRCm38_LISI$rmskID=paste0(rmsk_GRCm38_LISI$genoName,':',rmsk_GRCm38_LISI$genoStart,'-',rmsk_GRCm38_LISI$genoEnd)
      LISI_mRNAome= merge(LISI_mRNAome,rmsk_GRCm38_LISI[,c("rmskID","strand")],by.x="V4",by.y='rmskID',all.x=T,suffixes=c('_Transc','_rmks'))
      LISI_mRNAome=LISI_mRNAome[LISI_mRNAome$strand_rmks==LISI_mRNAome$strand_Transc,]
      

       LISI_mRNAome_GRstrand=GRanges(seqnames=LISI_mRNAome$V1, 
                          ranges=IRanges(start=LISI_mRNAome$V2,end=LISI_mRNAome$V3),
                          strand=LISI_mRNAome$strand_Transc,
                          ID=LISI_mRNAome$V4
                          # chrm=LISI_mRNAome$seqname
                          )     

########################################      
##### Adjusted Coordinates

pt_up= permTest(
  A=CLIP_mRNAome_GRstrand,
  B=LISI_mRNAome_GRstrand,
  genome=mm10_mRNAome,
  randomize.function=randomizeRegions,
  evaluate.function=meanDistance,ntimes=1000,count.once=TRUE,verbose=F)

################################# ################################# ################################# 


pt2_up= overlapPermTest(
  A=CLIP_mRNAome_GRstrand,
  B=LISI_mRNAome_GRstrand,
  genome=mm10_mRNAome,
  non.overlapping=T,
  ntimes=1000,count.once=TRUE)

plot(pt_up)
plot(pt2_up)


        gr = CLIP_mRNAome_GRstrand
        gr2 = LISI_mRNAome_GRstrand
        xo=as.data.frame(GenomicRanges::findOverlaps(gr,gr2,type = "any",ignore.strand=F))
          
        CLIP_LISI_VISHup=gr2[xo$subjectHits]
                  # qh=as.data.frame(gr[xo$queryHits],row.names = NULL)
                  # colnames(qh)=paste0(colnames(qh),"_mm10")
                  # sh=as.data.frame(gr2[xo$subjectHits],row.names = NULL)
        LISI_CLIP_VISHup=gr[xo$queryHits] 
        
```

```{r ,echo=F,warning=F,eval=F,include=F}

### downstream
CLIP_mRNAome_dn=CLIP_mRNAome

      CLIP_mRNAome_dn[,'V3']=CLIP_mRNAome_dn[,'V2']+window
      
      
      CLIP_mRNAome_dn[(CLIP_mRNAome_dn$V2<1),'V2']=1
      CLIP_mRNAome_dn[(CLIP_mRNAome_dn$V3>CLIP_mRNAome_dn$V2_Genome),'V3']= CLIP_mRNAome_dn[(CLIP_mRNAome_dn$V3>CLIP_mRNAome_dn$V2_Genome),'V2_Genome']
        

      CLIP_mRNAome_GR=GRanges(seqnames=CLIP_mRNAome_dn$V1, 
                          ranges=IRanges(start=CLIP_mRNAome_dn$V2,end=CLIP_mRNAome_dn$V3),
                          # strand=CLIP_mRNAome_dn$strand
                          ID=CLIP_mRNAome_dn$V4
                          # chrm=CLIP_mRNAome_dn$seqname
                          )
       CLIP_mRNAome_GRstrand=GRanges(seqnames=CLIP_mRNAome_dn$V1, 
                          ranges=IRanges(start=CLIP_mRNAome_dn$V2,end=CLIP_mRNAome_dn$V3),
                          strand=CLIP_mRNAome_dn$strand,
                          ID=CLIP_mRNAome_dn$V4
                          # chrm=CLIP_mRNAome_dn$seqname
                          )
          
      
 setdiff(CLIP_mRNAome_dn$V1,mm10_mRNAome$V1)    
 

        
########################################      
##### Adjusted Coordinates


pt_dn= permTest(
  A=CLIP_mRNAome_GRstrand,
  B=LISI_mRNAome_GRstrand,
  genome=mm10_mRNAome,
  randomize.function=randomizeRegions,
  evaluate.function=meanDistance,ntimes=100,count.once=TRUE,verbose=F)

################################# ################################# ################################# 

pt2_dn= overlapPermTest(
  A=CLIP_mRNAome_GRstrand,
  B=LISI_mRNAome_GRstrand,
  genome=mm10_mRNAome,
  non.overlapping=T,
  ntimes=100,count.once=TRUE)


plot(pt_dn)
plot(pt2_dn)

        gr = CLIP_mRNAome_GRstrand
        gr2 = LISI_mRNAome_GRstrand
        xo=as.data.frame(GenomicRanges::findOverlaps(gr,gr2,type = "any",ignore.strand=F))
          
        CLIP_LISI_VISHdn=gr2[xo$subjectHits]
                  # qh=as.data.frame(gr[xo$queryHits],row.names = NULL)
                  # colnames(qh)=paste0(colnames(qh),"_mm10")
                  # sh=as.data.frame(gr2[xo$subjectHits],row.names = NULL)
        LISI_CLIP_VISHdn=gr[xo$queryHits] 
        
```


### SINE/LINE Transcriptomic Distance  

```{r ,echo=F,warning=F,eval=F,include=F}
        # remove(mm10_mRNAome,CLIP_mRNAome,CLIP_mRNAome_up,CLIP_mRNAome_dn,CLIP_mRNAome_GR,CLIP_mRNAome_GRstrand,LISI_mRNAome,LISI_mRNAome_GR,LISI_mRNAome_GRstrand)   

##################################
### After convert Genome to Transcriptome coordinates
  

mm10_mRNAome_transc=fread(paste0("./structure/LINESINE/for_PhilHoman2/for_PhilHoman/mm10.premmRNAome.genome"), header=F, sep="\t",stringsAsFactors = F,data.table=F)
      mm10_mRNAome_transc$V1=removeVersion(mm10_mRNAome_transc$V1)
mm10_mRNAome_transc$V3=mm10_mRNAome_transc$V2
mm10_mRNAome_transc$V2=1

######################################################  
# ExonCoord=fread(paste0("./structure/LINESINE/ExonicCoordinates_Canonical.txt"), header=T, sep="\t",stringsAsFactors = F,data.table=F)
# IntronCoord=fread(paste0("./structure/LINESINE/IntronsCoordinates_Canonical.txt"), header=T, sep="\t",stringsAsFactors = F,data.table=F)
# TranscCoord=fread(paste0("./structure/LINESINE/TranscCoordinates_Canonical.txt"), header=T, sep="\t",stringsAsFactors = F,data.table=F)
#       ExonCoord$gene_id=removeVersion(ExonCoord$gene_id)
#       IntronCoord$gene_id=removeVersion(IntronCoord$gene_id)
#       TranscCoord$gene_id=removeVersion(TranscCoord$gene_id)
# INTID=unique(IntronCoord$transcript_id)
# 
# IntronCoord$start_adj=NA
# IntronCoord$end_adj=NA
# IntronCoord$Trans_length=NA
# 
# for (tr in 1:length(GID)) {
#   INTIDX=INTID[tr]
#   trX=TranscCoord[TranscCoord$transcript_id%in%INTIDX,]
#   inX=IntronCoord[IntronCoord$transcript_id%in%INTIDX,]
#   exX=ExonCoord[ExonCoord$transcript_id%in%INTIDX,]
#   if (length((exX$start))==0|length(trX$start)==0) {next }
#   if (min(exX$start)!=trX$start) {next}
# 
#   if (unique(trX$strand=="+")) {
# 
#     IntronCoord[IntronCoord$transcript_id%in%INTIDX,'start_adj']=IntronCoord[IntronCoord$transcript_id%in%INTIDX,'start']-trX$start
#     IntronCoord[IntronCoord$transcript_id%in%INTIDX,'end_adj']=IntronCoord[IntronCoord$transcript_id%in%INTIDX,'end']-trX$start
#     IntronCoord[IntronCoord$transcript_id%in%INTIDX,'Trans_length']=trX$end-trX$start
#   }
# 
# if (unique(trX$strand=="-")) {
#    IntronCoord[IntronCoord$transcript_id%in%INTIDX,'start_adj']= trX$end-IntronCoord[IntronCoord$transcript_id%in%INTIDX,'end']
#    IntronCoord[IntronCoord$transcript_id%in%INTIDX,'end_adj']=trX$end-IntronCoord[IntronCoord$transcript_id%in%INTIDX,'start']
#    IntronCoord[IntronCoord$transcript_id%in%INTIDX,'Trans_length']=trX$end-trX$start
#   }
# 
# }
# 
#   write.table(IntronCoord, file=paste0('./structure/LINESINE/IntronsCoordinatesADJ_Canonical.txt'), quote=F, sep="\t", row.names=F)
  
######################################################  
  
IntronCoordADJ=fread(paste0("./structure/LINESINE/IntronsCoordinatesADJ_Canonical.txt"), header=T, sep="\t",stringsAsFactors = F,data.table=F)
IntronCoordADJ$outside=IntronCoordADJ$Trans_length-IntronCoordADJ$end_adj
 Intron_mRNAome_transcGRstrand=GRanges(seqnames=IntronCoordADJ$transcript_id, 
                          ranges=IRanges(start=IntronCoordADJ$start_adj,end=IntronCoordADJ$end_adj),
                          strand=IntronCoordADJ$strand,
                          ID=paste0(IntronCoordADJ$chr,':',IntronCoordADJ$start,'-',IntronCoordADJ$end)
                          # chrm=IntronCoordADJ$seqname
                          )
Intron_mRNAome_transcMASK=IntronCoordADJ[,c("transcript_id","start_adj","end_adj")]
colnames(Intron_mRNAome_transcMASK)=NA
Intron_mRNAome_transcMASK=as.data.frame((Intron_mRNAome_transcMASK))
colnames(Intron_mRNAome_transcMASK)=c('V1','V2','V3')
########################################################################################################################################

CLIP_mRNAome_transc=fread(paste0("./structure/LINESINE/for_PhilHoman2/for_PhilHoman/CLIP_crosslink_in_ExonicCoordinates_Canonical.new_coords.premmRNA.bed"), header=F, sep="\t",stringsAsFactors = F,data.table=F)
      CLIP_mRNAome_transc$V1=removeVersion(CLIP_mRNAome_transc$V1)
CLIP_mRNAome_transc=merge(CLIP_mRNAome_transc,mm10_trans[,c("transcript_id","gene_id","strand")],by.x="V1",by.y='transcript_id')
    CLIP_mRNAome_transc[( CLIP_mRNAome_transc$V3-CLIP_mRNAome_transc$V2 >1)&(CLIP_mRNAome_transc$strand=="+"),'V3']=CLIP_mRNAome_transc[( CLIP_mRNAome_transc$V3-CLIP_mRNAome_transc$V2 >1)&(CLIP_mRNAome_transc$strand=="+"),'V2']+1
    CLIP_mRNAome_transc[( CLIP_mRNAome_transc$V3-CLIP_mRNAome_transc$V2 >1)&(CLIP_mRNAome_transc$strand=="-"),'V2']=CLIP_mRNAome_transc[( CLIP_mRNAome_transc$V3-CLIP_mRNAome_transc$V2 >1)&(CLIP_mRNAome_transc$strand=="-"),'V3']-1

checkout2_LISI=checkout2
checkout2_LISI=separate(checkout2_LISI,CLIP_ID,into = c('chr','start'),sep = ":",remove = F)
checkout2_LISI=separate(checkout2_LISI,start,into = c('start','end'),sep = "-",remove = T)
checkout2_LISI$Region_ID=paste0(checkout2_LISI$chr,":",checkout2_LISI$crosslink,"-",(checkout2_LISI$crosslink+1))

CLIP_mRNAome_transc=CLIP_mRNAome_transc[CLIP_mRNAome_transc$V4%in%checkout2_LISI$Region_ID,]

# CLIP_mRNAome_transc[CLIP_mRNAome_transc$V4%in%checkout2_LISI$Region_ID==F,]
# checkout2_LISI$Region_ID2=paste0(checkout2_LISI$chr,":",checkout2_LISI$crosslink-1,"-",(checkout2_LISI$crosslink))
# checkout2_LISI$Region_ID3=paste0(checkout2_LISI$chr,":",checkout2_LISI$crosslink-2,"-",(checkout2_LISI$crosslink-1))
# checkout2_LISI$Region_ID4=paste0(checkout2_LISI$chr,":",checkout2_LISI$crosslink+1,"-",(checkout2_LISI$crosslink+2))
# CLIP_mRNAome_transc[CLIP_mRNAome_transc$V4%in%checkout2_LISI$Region_ID2,]
# CLIP_mRNAome_transc[CLIP_mRNAome_transc$V4%in%checkout2_LISI$Region_ID3,]
# CLIP_mRNAome_transc[CLIP_mRNAome_transc$V4%in%checkout2_LISI$Region_ID4,]
# chr10:98915198-98915199	chr7:19006283-19006284	chr5:110105007-110105008	chr9:57504110-57504111	chr9:58220066-58220067	chr10:75643989-75643990	chr15:100280858-100280859	chr1:180681123-180681124	



      CLIP_mRNAome_transcGR=GRanges(seqnames=CLIP_mRNAome_transc$V1, 
                          ranges=IRanges(start=CLIP_mRNAome_transc$V2,end=CLIP_mRNAome_transc$V3),
                          # strand=CLIP_mRNAome_transc$strand
                          ID=CLIP_mRNAome_transc$V4
                          # chrm=CLIP_mRNAome_transc$seqname
                          )
       CLIP_mRNAome_transcGRstrand=GRanges(seqnames=CLIP_mRNAome_transc$V1, 
                          ranges=IRanges(start=CLIP_mRNAome_transc$V2,end=CLIP_mRNAome_transc$V3),
                          strand=CLIP_mRNAome_transc$strand,
                          ID=CLIP_mRNAome_transc$V4
                          # chrm=CLIP_mRNAome_transc$seqname
                          )
           

      
 setdiff(CLIP_mRNAome_transc$V1,mm10_mRNAome_transc$V1)    


 ######################################################################################################################## 
LISI_mRNAome_transc=fread(paste0("./structure/LINESINE/for_PhilHoman2/for_PhilHoman/SINE_LINE_in_canonical.transcripts.new_coords.premmRNA.bed"), header=F, sep="\t",stringsAsFactors = F,data.table=F)
      LISI_mRNAome_transc$V1=removeVersion(LISI_mRNAome_transc$V1)
      LISI_mRNAome_transc=merge(LISI_mRNAome_transc,mm10_trans[,c("transcript_id","gene_id","strand")],by.x="V1",by.y='transcript_id')

      
      
rmsk_GRCm38=fread(paste0("/Users/homanpj/OneDrive - National Institutes of Health/Resources/ref/mm10/repeatmasker/rmsk_GRCm38.txt"), header=T, sep="\t",stringsAsFactors = F,data.table=F)
rmsk_GRCm38_LISI=rmsk_GRCm38[rmsk_GRCm38$repClass%in%c('LINE','SINE'),]
rmsk_GRCm38_LISI=rmsk_GRCm38_LISI[rmsk_GRCm38_LISI$genoName%in%c('chrUn_GL456239', 'chrUn_GL456359', 'chrUn_GL456360', 'chrUn_GL456366', 'chrUn_GL456367', 'chrUn_GL456368', 'chrUn_GL456370', 'chrUn_GL456372', 'chrUn_GL456378', 'chrUn_GL456379', 'chrUn_GL456381', 'chrUn_GL456382', 'chrUn_GL456385', 'chrUn_GL456387', 'chrUn_GL456389', 'chrUn_GL456393', 'chrUn_GL456394', 'chrUn_JH584304', 'chr1_GL456210_random', 'chr1_GL456211_random', 'chr1_GL456212_random', 'chr1_GL456213_random', 'chr1_GL456221_random', 'chr4_GL456216_random', 'chr4_GL456350_random', 'chr4_JH584292_random', 'chr4_JH584293_random', 'chr4_JH584294_random', 'chr5_GL456354_random', 'chr5_JH584296_random', 'chr5_JH584297_random', 'chr5_JH584298_random', 'chr5_JH584299_random', 'chr7_GL456219_random', 'chrX_GL456233_random', 'chrY_JH584300_random', 'chrY_JH584301_random', 'chrY_JH584302_random', 'chrY_JH584303_random')==F,]
rmsk_GRCm38_LISI$ID=paste0(rmsk_GRCm38_LISI$genoName,":",rmsk_GRCm38_LISI$genoStart,'-',rmsk_GRCm38_LISI$genoEnd)

rmsk_GRCm38_LISI$rmskID=paste0(rmsk_GRCm38_LISI$genoName,':',rmsk_GRCm38_LISI$genoStart,'-',rmsk_GRCm38_LISI$genoEnd)
      LISI_mRNAome_transc= merge(LISI_mRNAome_transc,rmsk_GRCm38_LISI[,c("rmskID","strand")],by.x="V4",by.y='rmskID',all.x=T,suffixes=c('_Transc','_rmks'))
      LISI_mRNAome_transc=LISI_mRNAome_transc[LISI_mRNAome_transc$strand_rmks==LISI_mRNAome_transc$strand_Transc,]
      LISI_mRNAome_transc=LISI_mRNAome_transc[duplicated(LISI_mRNAome_transc)==F,]

      # LISI_mRNAome_transc=merge(LISI_mRNAome_transc,mm10_exon[,c("transcript_id","strand","seqname")],by.x='V1',by.y="transcript_id",all.x=T,no.dups = TRUE)
      # LISI_mRNAome_transc=LISI_mRNAome_transc[duplicated(LISI_mRNAome_transc)==F,]
      # LISI_mRNAome_transc[is.na(LISI_mRNAome_transc$strand),'strand']="*"
      # LISI_mRNAome_transc=LISI_mRNAome_transc[LISI_mRNAome_transc$V1%in%CLIP_mRNAome_transc$V1,]
      
      LISI_mRNAome_GR=GRanges(seqnames=LISI_mRNAome_transc$V1, 
                          ranges=IRanges(start=LISI_mRNAome_transc$V2,end=LISI_mRNAome_transc$V3),
                          # strand=LISI_mRNAome_transc$strand
                          # ID=LISI_mRNAome_transc$V4,
                          # chrm=LISI_mRNAome_transc$seqname
                          )

       LISI_mRNAome_GRstrand=GRanges(seqnames=LISI_mRNAome_transc$V1, 
                          ranges=IRanges(start=LISI_mRNAome_transc$V2,end=LISI_mRNAome_transc$V3),
                          strand=LISI_mRNAome_transc$strand_Transc,
                          ID=LISI_mRNAome_transc$V4
                          # chrm=LISI_mRNAome_transc$seqname
                          )     

########################################################################################################################
       


 
 
 ########################################################################################################################        

########################################      
##### Adjusted Coordinates
maskx=Intron_mRNAome_transcMASK[Intron_mRNAome_transcMASK$V1%in%mm10_mRNAome_transc$V1,]
 # randomizeRegions(A=CLIP_mRNAome_transcGRstrand,genome= mm10_mRNAome_transc,mask = maskx,
                  # allow.overlaps = F,per.chromosome = F,pruning.mode="fine")


# pt= permTest(
#   A=CLIP_mRNAome_transcGRstrand,
#   B=LISI_mRNAome_GRstrand,
#   genome=mm10_mRNAome_transc,
#   randomize.function=randomizeRegions,
#   evaluate.function=meanDistance,ntimes=100,count.once=TRUE,verbose=F,mask=maskx)
# plot(pt)
# 
# ptx= permTest(
#   A=CLIP_mRNAome_transcGRstrand,
#   B=LISI_mRNAome_GRstrand,
#   genome=mm10_mRNAome_transc,
#   randomize.function=randomizeRegions,
#   evaluate.function=meanDistance,ntimes=100,count.once=TRUE,verbose=F)
# plot(ptx)
# 
# ################################# ################################# ################################# 
# 
# 
# pt2= overlapPermTest(
#   A=CLIP_mRNAome_transcGRstrand,
#   B=LISI_mRNAome_GRstrand,
#   genome=mm10_mRNAome_transc,
#   non.overlapping=T,
#   ntimes=100,count.once=TRUE,mask=maskx)
# plot(pt2)
# 
# pt2x= overlapPermTest(
#   A=CLIP_mRNAome_transcGRstrand,
#   B=LISI_mRNAome_GRstrand,
#   genome=mm10_mRNAome_transc,
#   non.overlapping=T,
#   ntimes=100,count.once=TRUE)
# plot(pt2x)
# 
# plot(pt)
# plot(ptx)
# plot(pt2)
# plot(pt2x)

        gr = CLIP_mRNAome_transcGRstrand
        gr2 = LISI_mRNAome_GRstrand
        xo=as.data.frame(GenomicRanges::findOverlaps(gr,gr2,type = "any",ignore.strand=F))
          
        CLIP_LISI_VISH=gr2[xo$subjectHits]
                  # qh=as.data.frame(gr[xo$queryHits],row.names = NULL)
                  # colnames(qh)=paste0(colnames(qh),"_mm10")
                  # sh=as.data.frame(gr2[xo$subjectHits],row.names = NULL)
        LISI_CLIP_VISH=gr[xo$queryHits] 


```

```{r ,echo=F,warning=F,eval=F,include=F}
     
      
### Upstream
CLIP_mRNAome_transcUP=CLIP_mRNAome_transc
      CLIP_mRNAome_transcUP[,'V2']=CLIP_mRNAome_transcUP[,'V3']-window
      
      CLIP_mRNAome_transcUP[(CLIP_mRNAome_transcUP$V2<1),'V2']=1
      CLIP_mRNAome_transcUP[(CLIP_mRNAome_transcUP$V3>CLIP_mRNAome_transcUP$V2_Genome),'V3']= CLIP_mRNAome_transcUP[(CLIP_mRNAome_transcUP$V3>CLIP_mRNAome_transcUP$V2_Genome),'V2_Genome']
        

  

       CLIP_mRNAome_transcGRstrand=GRanges(seqnames=CLIP_mRNAome_transcUP$V1, 
                          ranges=IRanges(start=CLIP_mRNAome_transcUP$V2,end=CLIP_mRNAome_transcUP$V3),
                          strand=CLIP_mRNAome_transcUP$strand,
                          ID=CLIP_mRNAome_transcUP$V4
                          # chrm=CLIP_mRNAome_transcUP$seqname
                          )
 # setdiff(CLIP_mRNAome_transcUP$V1,mm10_mRNAome_transc$V1)    


########################################      
##### Adjusted Coordinates

pt_up= permTest(
  A=CLIP_mRNAome_transcGRstrand,
  B=LISI_mRNAome_GRstrand,
  genome=mm10_mRNAome_transc,
  randomize.function=randomizeRegions,
  evaluate.function=meanDistance,ntimes=500,mask=maskx,
  count.once=TRUE,verbose=F)

################################# ################################# ################################# 


pt2_up= overlapPermTest(
  A=CLIP_mRNAome_transcGRstrand,
  B=LISI_mRNAome_GRstrand,
  genome=mm10_mRNAome_transc,
  mask=maskx,
  non.overlapping=T,ntimes=500,count.once=TRUE)

plot(pt_up)
plot(pt2_up)


        gr = CLIP_mRNAome_transcGRstrand
        gr2 = LISI_mRNAome_GRstrand
        xo=as.data.frame(GenomicRanges::findOverlaps(gr,gr2,type = "any",ignore.strand=F))
          
        CLIP_LISI_VISHup=gr2[xo$subjectHits]
                  # qh=as.data.frame(gr[xo$queryHits],row.names = NULL)
                  # colnames(qh)=paste0(colnames(qh),"_mm10")
                  # sh=as.data.frame(gr2[xo$subjectHits],row.names = NULL)
        LISI_CLIP_VISHup=gr[xo$queryHits] 
        
```

```{r ,echo=F,warning=F,eval=F,include=F}


### downstream
CLIP_mRNAome_transcDN=CLIP_mRNAome_transc

      CLIP_mRNAome_transcDN[,'V3']=CLIP_mRNAome_transcDN[,'V2']+window
      
      
      CLIP_mRNAome_transcDN[(CLIP_mRNAome_transcDN$V2<1),'V2']=1
      CLIP_mRNAome_transcDN[(CLIP_mRNAome_transcDN$V3>CLIP_mRNAome_transcDN$V2_Genome),'V3']= CLIP_mRNAome_transcDN[(CLIP_mRNAome_transcDN$V3>CLIP_mRNAome_transcDN$V2_Genome),'V2_Genome']
        

      CLIP_mRNAome_transcGR=GRanges(seqnames=CLIP_mRNAome_transcDN$V1, 
                          ranges=IRanges(start=CLIP_mRNAome_transcDN$V2,end=CLIP_mRNAome_transcDN$V3),
                          # strand=CLIP_mRNAome_transcDN$strand
                          ID=CLIP_mRNAome_transcDN$V4
                          # chrm=CLIP_mRNAome_transcDN$seqname
                          )
       CLIP_mRNAome_transcGRstrand=GRanges(seqnames=CLIP_mRNAome_transcDN$V1, 
                          ranges=IRanges(start=CLIP_mRNAome_transcDN$V2,end=CLIP_mRNAome_transcDN$V3),
                          strand=CLIP_mRNAome_transcDN$strand,
                          ID=CLIP_mRNAome_transcDN$V4
                          # chrm=CLIP_mRNAome_transcDN$seqname
                          )
          
      
 setdiff(CLIP_mRNAome_transcDN$V1,mm10_mRNAome_transc$V1)    
 

        
########################################      
##### Adjusted Coordinates


pt_dn= permTest(
  A=CLIP_mRNAome_transcGRstrand,
  B=LISI_mRNAome_GRstrand,
  genome=mm10_mRNAome_transc,
  randomize.function=randomizeRegions,
  evaluate.function=meanDistance,
  mask=maskx,
  ntimes=500,count.once=TRUE,verbose=F)

################################# ################################# ################################# 


pt2_dn= overlapPermTest(
  A=CLIP_mRNAome_transcGRstrand,
  B=LISI_mRNAome_GRstrand,
  genome=mm10_mRNAome_transc,
  mask=maskx,
  non.overlapping=T,ntimes=500,count.once=TRUE)


plot(pt_dn)
plot(pt2_dn)

        gr = CLIP_mRNAome_transcGRstrand
        gr2 = LISI_mRNAome_GRstrand
        xo=as.data.frame(GenomicRanges::findOverlaps(gr,gr2,type = "any",ignore.strand=F))
          
        CLIP_LISI_VISHdn=gr2[xo$subjectHits]
                  # qh=as.data.frame(gr[xo$queryHits],row.names = NULL)
                  # colnames(qh)=paste0(colnames(qh),"_mm10")
                  # sh=as.data.frame(gr2[xo$subjectHits],row.names = NULL)
        LISI_CLIP_VISHdn=gr[xo$queryHits] 
        
```


### LISI locations

```{r ,echo=F,warning=F,eval=F,include=F}

LISI_detail=fread(paste0("./structure/LINESINE/for_PhilHoman2/for_PhilHoman/CLIP_crosslink_in_ExonicCoordinates_Canonical.new_coords.closest_SINE_LINE.filtered.w_details.txt"), header=F, sep="\t",stringsAsFactors = F,data.table=F)
colnames(LISI_detail)=c('CLIPpeak_mmRNA_chrom','CLIPpeak_mmRNA_start','CLIPpeak_mmRNA_end',	'CLIPpeak_original_coordinates',	'SINE_LINE_mmRNA_chrom',	'SINE_LINE_mmRNA_start',	'SINE_LINE_mmRNA_end','SINE_LINE_original_coordinates',	'Distance_on_mmRNA','SINE_LINE_original_Chr','SINE_LINE_original_start','SINE_LINE_original_end','SINE_LINE_id',	'SINE_LINE_strand','exon_transcript_chr','exon_transcript_start','exon_transcript_end',	'exon_transcript_id',	'transcript_strand')

LISI_detail=as.data.frame(LISI_detail)
 LISI_detail= separate(LISI_detail,'CLIPpeak_original_coordinates',into=c('CLIP_chr','CLIP_start'),sep = ":",remove = F)
 LISI_detail= separate(LISI_detail,'CLIP_start',into=c('CLIP_start','CLIP_end'),sep = "-",remove = T)
LISI_detail$Crosslink_ID=paste0(LISI_detail$CLIP_chr,':',LISI_detail$CLIP_start,'_',LISI_detail$transcript_strand)
 
LISI_detail=LISI_detail[LISI_detail$SINE_LINE_strand==LISI_detail$transcript_strand,]


LISI_detail_transc=fread(paste0("./structure/LINESINE/for_PhilHoman2/for_PhilHoman/CLIP_crosslink.new_coords.premmRNA.closest_transcript_SINE_LINE.w_detail.txt"), header=F, sep="\t",stringsAsFactors = F,data.table=F)
colnames(LISI_detail_transc)=c('CLIPpeak_mmRNA_chrom','CLIPpeak_mmRNA_start','CLIPpeak_mmRNA_end',	'CLIPpeak_original_coordinates','V5','V6',	'SINE_LINE_mmRNA_chrom',	'SINE_LINE_mmRNA_start',	'SINE_LINE_mmRNA_end', 'SINE_LINE_original_coordinates' , 'V11','V12',	'Distance_on_mmRNA','SINE_LINE_original_Chr','SINE_LINE_original_start','SINE_LINE_original_end','SINE_LINE_id', 'V18',	'SINE_LINE_strand','exon_transcript_chr','exon_transcript_start','exon_transcript_end',	'exon_transcript_id','V24',	'transcript_strand')

LISI_detail_transc=as.data.frame(LISI_detail_transc)
 LISI_detail_transc= separate(LISI_detail_transc,'CLIPpeak_original_coordinates',into=c('CLIP_chr','CLIP_start'),sep = ":",remove = F)
 LISI_detail_transc= separate(LISI_detail_transc,'CLIP_start',into=c('CLIP_start','CLIP_end'),sep = "-",remove = T)
LISI_detail_transc$Crosslink_ID=paste0(LISI_detail_transc$CLIP_chr,':',LISI_detail_transc$CLIP_start,'_',LISI_detail_transc$transcript_strand)
 
LISI_detail_transc=LISI_detail_transc[LISI_detail_transc$SINE_LINE_strand==LISI_detail_transc$transcript_strand,]


clmns=c('Crosslink_ID','SINE_LINE_id','Distance_on_mmRNA')
# LISI_detail=merge(LISI_detail[,clmns],LISI_detail_transc[,clmns],by='Crosslink_ID',suffixes=c('_exonsOnly',"_Exons-Introns"))
# 
# colnames(LISI_detail)[colnames(LISI_detail)%in%c('Distance_on_mmRNA_Exons-Introns')]=c('Distance_including_Exons-Introns')
```








## Distibutioin of CLIP Transcript location

### 
```{r ,echo=F,warning=F,eval=T,include=F,fig.asp=1.1,fig.height=2, fig.width=2.5}
# intersect(checkout2$transcript_id,isoout_transcAll$USEtranscript)
remove(mm10all,mm10_trans,mm10_exon,mm10_UTR,introns,mm10_intron)

checkout2_LS=Peaksdata4_chunkInExall
checkout2_LS=merge(isoout_transcAll[,c('ID','USEtranscript')],checkout2_LS,by='ID')
colnames(checkout2_LS)[colnames(checkout2_LS)%in%'USEtranscript']='transcript_id'

TranscComponent='Exons' ## Exons, IntronsExonsStrict, All Protein Coding


if (TranscComponent=='All Protein Coding') {checkout2_LS=checkout2_LS[checkout2_LS$`Both Strand: RNA_type_Classification`%in%c('protein_coding: Intron','protein_coding: exon'),]}
if (TranscComponent=='Exons') {checkout2_LS=checkout2_LS[checkout2_LS$`Both Strand: RNA_type_Classification`%in%c('protein_coding: exon'),]}

checkout2_LS=merge(checkout2_LS,isoout_transcAll)

checkout2_LS$crosslink=NA
checkout2_LS[checkout2_LS$strand=="+",'crosslink']=checkout2_LS[checkout2_LS$strand=="+",'start']
checkout2_LS[checkout2_LS$strand=="-",'crosslink']=checkout2_LS[checkout2_LS$strand=="-",'end']

checkout2_LS_XLnk=checkout2_LS[,c('chr','crosslink','strand','ID','transcript_id')]
  # write.table(checkout2_LS_XLnk, file=paste0(LINESINEout,'/CLIPCoordinates.txt'), quote=F, sep="\t", row.names=F)

 
CrossLink_GR=GRanges(seqnames=checkout2_LS_XLnk$chr, ranges=IRanges(start=checkout2_LS_XLnk$crosslink,
end=checkout2_LS_XLnk$crosslink),
strand=checkout2_LS_XLnk$strand,
ID=checkout2_LS_XLnk$ID,
transcript_id=checkout2_LS_XLnk$transcript_id)  

################################################
#### create Genomic files
################################################

mm10all=fread("/Users/homanpj/OneDrive - National Institutes of Health/Resources/ref/mm10/Gencode_VM23/fromGencode/gencode.vM23.basic.annotation.gtf.txt", header=T, sep="\t",stringsAsFactors = F,data.table=F)
  # mm10all=fread("/data/RBL_NCI/Phil/Reference/gtf/mouse/mm10/Gencode_VM23/gencode.vM23.basic.annotation.gtf.txt", header=T, sep="\t",stringsAsFactors = F,data.table=F)
  mm10all$transcript_id=removeVersion(mm10all$transcript_id)
  mm10all$gene_id=removeVersion(mm10all$gene_id)
  
  mm10_trans=mm10all[mm10all$feature%in%'transcript',]
  mm10_UTR=mm10all[mm10all$feature%in%'UTR',]
  mm10_exon=mm10all[mm10all$feature%in%c("exon"),]

  
############################################################################################

mm10_intron=fread("/Users/homanpj/OneDrive - National Institutes of Health/Resources/ref/mm10/Gencode_VM23/fromUCSC/KnownGene/KnownGene_GRCm38_introns.bed", header=F, sep="\t",stringsAsFactors = F,data.table=F)
  mm10_intron=mm10_intron[grepl("_",mm10_intron$V1)==F,]
  colnames(mm10_intron)=c('chr','start','end','attribute','V5','strand')
  mm10_intron=separate(mm10_intron,attribute,into=c('transcript_id','feature','exon_number','level','chr2','intronnumber','dir'),remove = T,sep = "_")
  mm10_intron$transcript_id=removeVersion(mm10_intron$transcript_id)
  mm10_intron$start=mm10_intron$start+1
  mm10_intron$exon_number=as.numeric(mm10_intron$exon_number)+1
  mm10_intron=merge(mm10_intron,mm10_trans[,c('transcript_id','gene_id')],by='transcript_id',all.x=T)

  
###############
  
mm10_3UTRall=fread("/Users/homanpj/OneDrive - National Institutes of Health/Resources/ref/mm10/Gencode_VM23/fromUCSC/KnownGene/KnownGene_GRCm38_GencodeV23_3UTR.bed", header=F, sep="\t",stringsAsFactors = F,data.table=F)
  mm10_3UTRall=mm10_3UTRall[grepl("_",mm10_3UTRall$V1)==F,]
  colnames(mm10_3UTRall)=c('chr','start','end','attribute','V5','strand')
  mm10_3UTRall=separate(mm10_3UTRall,attribute,into=c('transcript_id','feature','exon_number','level','chr2','start+1','dir'),remove = T,sep = "_")
  mm10_3UTRall$transcript_id=removeVersion(mm10_3UTRall$transcript_id)
  mm10_3UTRall$start=mm10_3UTRall$start+1
  mm10_3UTRall$exon_number=as.numeric(mm10_3UTRall$exon_number)+1
  mm10_3UTRall$width=(mm10_3UTRall$end-mm10_3UTRall$start)+1

  
  
################################################
  
  mm10_trans=mm10_trans[mm10_trans$transcript_id%in%isoout_transcAll$USEtranscript,]
  mm10_UTR=mm10_UTR[mm10_UTR$transcript_id%in%isoout_transcAll$USEtranscript,]
  mm10_3UTR=mm10_3UTRall[mm10_3UTRall$transcript_id%in%isoout_transcAll$USEtranscript,]
  mm10_exon=mm10_exon[mm10_exon$transcript_id%in%isoout_transcAll$USEtranscript,]
  mm10_intron=mm10_intron[mm10_intron$transcript_id%in%isoout_transcAll$USEtranscript,]
  
################################################
introns_GR=GRanges(seqnames=mm10_intron$chr,
                   ranges=IRanges(start=mm10_intron$start,end=mm10_intron$end),
                   strand=mm10_intron$strand,
                   transcript_id=mm10_intron$transcript_id,
                   exon_number=mm10_intron$exon_number
                   )
  
  

############################################################################################
 
exons_GR=GRanges(seqnames=mm10_exon$seqname,
                   ranges=IRanges(start=mm10_exon$start,end=mm10_exon$end),
                   strand=mm10_exon$strand,
                   transcript_id=mm10_exon$transcript_id,
                   gene_id=mm10_exon$gene_id,
                   exon_number=mm10_exon$exon_number
                   )


############################

 TranscUTR_GR=GRanges(seqnames=mm10_UTR$seqname,
                   ranges=IRanges(start=mm10_UTR$start,end=mm10_UTR$end),
                   strand=mm10_UTR$strand,
                   transcript_id=mm10_UTR$transcript_id,
                   gene_id=mm10_UTR$gene_id,
                   exon_number=mm10_UTR$exon_number
                   )

############################

 Transc3UTR_GR=GRanges(seqnames=mm10_3UTR$chr,
                   ranges=IRanges(start=mm10_3UTR$start,end=mm10_3UTR$end),
                   strand=mm10_3UTR$strand,
                   transcript_id=mm10_3UTR$transcript_id,
                   # gene_id=mm10_3UTR$gene_id,
                   exon_number=mm10_3UTR$exon_number
                   )

############################

Transc_GR=GRanges(seqnames=mm10_trans$seqname,
                   ranges=IRanges(start=mm10_trans$start,end=mm10_trans$end),
                   strand=mm10_trans$strand,
                   transcript_id=mm10_trans$transcript_id,
                  exon_number=mm10_trans$exon_number,
                   gene_id=mm10_trans$gene_id
                   )


################################################
#### create Genomic overlaps
################################################


### Get Transc that overlap with CLIP peaks
        gr = CrossLink_GR
        gr2 = Transc_GR
        xo=as.data.frame(GenomicRanges::findOverlaps(gr,gr2,type = "any",ignore.strand=F))
          
        Transc_CrossLink_GR=gr2[xo$subjectHits]
                  # qh=as.data.frame(gr[xo$queryHits],row.names = NULL)
                  # colnames(qh)=paste0(colnames(qh),"_mm10")
                  # sh=as.data.frame(gr2[xo$subjectHits],row.names = NULL)
        Crosslink_Transc_GR=gr[xo$queryHits]  
    Crosslink_Transc_GR_DF=as.data.frame(Crosslink_Transc_GR)
    Crosslink_Transc_GR_DF$Transc='Transc'
    
    checkout2_LS$Transc=NA
    checkout2_LS[which(checkout2_LS$ID%in%Crosslink_Transc_GR_DF$ID),'Transc']="Transc"
    

### Get Exon that overlap with CLIP peaks
        gr = CrossLink_GR
        gr2 = exons_GR
        xo=as.data.frame(GenomicRanges::findOverlaps(gr,gr2,type = "any",ignore.strand=F))
          
        exons_CrossLink_GR=gr2[xo$subjectHits]
                  # qh=as.data.frame(gr[xo$queryHits],row.names = NULL)
                  # colnames(qh)=paste0(colnames(qh),"_mm10")
                  # sh=as.data.frame(gr2[xo$subjectHits],row.names = NULL)
        Crosslink_exons_GR=gr[xo$queryHits]
        
      
        exons_CrossLink_GR_DF=as.data.frame(exons_CrossLink_GR)
        colnames(exons_CrossLink_GR_DF)=paste(colnames(exons_CrossLink_GR_DF),'_exons')
        
    Crosslink_exons_GR_DF=cbind(as.data.frame(Crosslink_exons_GR),exons_CrossLink_GR_DF)
    Crosslink_exons_GR_DF$exons='exons'
    
    checkout2_LS$exons=NA
    checkout2_LS[which(checkout2_LS$ID%in%Crosslink_exons_GR_DF$ID),'exons']="exons"


### Get UTR that overlap with CLIP peaks
        gr = CrossLink_GR
        gr2 = TranscUTR_GR
        xo=as.data.frame(GenomicRanges::findOverlaps(gr,gr2,type = "any",ignore.strand=F))
          
        UTR_CrossLink_GR=gr2[xo$subjectHits]
                  # qh=as.data.frame(gr[xo$queryHits],row.names = NULL)
                  # colnames(qh)=paste0(colnames(qh),"_mm10")
                  # sh=as.data.frame(gr2[xo$subjectHits],row.names = NULL)
        Crosslink_UTR_GR=gr[xo$queryHits]
        
      
        UTR_CrossLink_GR_DF=as.data.frame(UTR_CrossLink_GR)
        colnames(UTR_CrossLink_GR_DF)=paste0(colnames(UTR_CrossLink_GR_DF),'_UTR')
        
    Crosslink_UTR_GR_DF=cbind(as.data.frame(Crosslink_UTR_GR),UTR_CrossLink_GR_DF)
    Crosslink_UTR_GR_DF$UTR='UTR'
    Crosslink_UTR_GR_DF$`5UTR`=NA
    Crosslink_UTR_GR_DF[Crosslink_UTR_GR_DF$exon_number_UTR==1,'5UTR']="5UTR"
    Crosslink_UTR_GR_DF_5UTR=Crosslink_UTR_GR_DF[is.na(Crosslink_UTR_GR_DF$`5UTR`)==F,]
    

### Get 3UTR that overlap with CLIP peaks
        gr = CrossLink_GR
        gr2 = Transc3UTR_GR
        xo=as.data.frame(GenomicRanges::findOverlaps(gr,gr2,type = "any",ignore.strand=F))
          
        UTR3_CrossLink_GR=gr2[xo$subjectHits]
                  # qh=as.data.frame(gr[xo$queryHits],row.names = NULL)
                  # colnames(qh)=paste0(colnames(qh),"_mm10")
                  # sh=as.data.frame(gr2[xo$subjectHits],row.names = NULL)
        Crosslink_3UTR_GR=gr[xo$queryHits]
        
      
        UTR3_CrossLink_GR_DF=as.data.frame(UTR3_CrossLink_GR)
        colnames(UTR3_CrossLink_GR_DF)=paste0(colnames(UTR3_CrossLink_GR_DF),'_3UTR')
        
    Crosslink_UTR_GR_DF_3UTR=cbind(as.data.frame(Crosslink_3UTR_GR),UTR3_CrossLink_GR_DF)
    Crosslink_UTR_GR_DF_3UTR$`3UTR`='3UTR'
    
        
### Get introns that overlap with CLIP peaks
        gr = CrossLink_GR
        gr2 = introns_GR
        xo=as.data.frame(GenomicRanges::findOverlaps(gr,gr2,type = "any",ignore.strand=F))
          
        introns_CrossLink_GR=gr2[xo$subjectHits]
                  # qh=as.data.frame(gr[xo$queryHits],row.names = NULL)
                  # colnames(qh)=paste0(colnames(qh),"_mm10")
                  # sh=as.data.frame(gr2[xo$subjectHits],row.names = NULL)
        Crosslink_introns_GR=gr[xo$queryHits]
        
      
        introns_CrossLink_GR_DF=as.data.frame(introns_CrossLink_GR)
        colnames(introns_CrossLink_GR_DF)=paste0(colnames(introns_CrossLink_GR_DF),'_introns')
        
    Crosslink_introns_GR_DF=cbind(as.data.frame(Crosslink_introns_GR),introns_CrossLink_GR_DF)
    if (nrow(Crosslink_introns_GR_DF)>0) {
          Crosslink_introns_GR_DF$introns='introns'
    }
    

################################################
#### create Genomic files
################################################    
############################################################################################

    checkout2_LS$UTR=NA
    checkout2_LS[which(checkout2_LS$ID%in%Crosslink_UTR_GR_DF$ID),'UTR']="UTR"
    
    checkout2_LS$`3UTR`=NA
    checkout2_LS[which(checkout2_LS$ID%in%Crosslink_UTR_GR_DF_3UTR$ID),'3UTR']="3UTR"
    
     checkout2_LS$`5UTR`=NA
    checkout2_LS[which(checkout2_LS$ID%in%Crosslink_UTR_GR_DF_5UTR$ID),'5UTR']="5UTR"

    checkout2_LS$`introns`=NA
    checkout2_LS[which(checkout2_LS$ID%in%Crosslink_introns_GR_DF$ID),'introns']="introns"
    
    checkout2_LS[is.na(checkout2_LS$exons)==F,'TranscRegion']='exons'
    checkout2_LS[is.na(checkout2_LS$`introns`)==F,'TranscRegion']='introns'
    checkout2_LS[is.na(checkout2_LS$`5UTR`)==F,'TranscRegion']='5UTR'
    checkout2_LS[is.na(checkout2_LS$`3UTR`)==F,'TranscRegion']='3UTR'
    checkout2_LS$TranscRegion=gsub("exons","exons_CDS",checkout2_LS$TranscRegion)
    
    # checkout2_LS=merge(checkout2_LS,mm10_trans)
    
    p=checkout2_LS
    p=p[is.na(p$TranscRegion)==F,]
          p$Data_Set=paste0(TranscComponent," CLIP Peaks")
          
gg=ggplot(p,aes(fill=TranscRegion,x=Data_Set) )+
    # geom_bar(position="stack",width=.5)+
    # ylab("# of Locations")+
  geom_bar(aes(y = (..count..)/sum(..count..)),position="stack",width=.5)+
    ylab("Percent of Locations")+
    xlab("")+
    theme_classic()+ggtitle(paste0("Transcript Location of ",TranscComponent," CLIP peaks"))+
# theme(plot.title = element_text(hjust = 0.5,size = 25),axis.title=element_text(size=20),axis.text=element_text(size=15),legend.text = element_text( size=30))+
 scale_fill_brewer(palette = "Dark2")
gg
# ggplotly(gg)   
```

### Random CLIP locations 

```{r ,echo=F,warning=F,eval=T,include=F,fig.asp=1.1,fig.height=2, fig.width=2.5}
################################################
#### RANDOM
 
TranscComponent='Exons'  ## Exons, IntronsExonsStrict, All Protein Coding

################################################
#### create Genomic files
################################################


mm10all=fread("/Users/homanpj/OneDrive - National Institutes of Health/Resources/ref/mm10/Gencode_VM23/fromGencode/gencode.vM23.basic.annotation.gtf.txt", header=T, sep="\t",stringsAsFactors = F,data.table=F)
  # mm10all=fread("/data/RBL_NCI/Phil/Reference/gtf/mouse/mm10/Gencode_VM23/gencode.vM23.basic.annotation.gtf.txt", header=T, sep="\t",stringsAsFactors = F,data.table=F)
  mm10all$transcript_id=removeVersion(mm10all$transcript_id)
  mm10all$gene_id=removeVersion(mm10all$gene_id)
  
  
introns=fread("/Users/homanpj/OneDrive - National Institutes of Health/Resources/ref/mm10/Gencode_VM23/fromUCSC/KnownGene/KnownGene_GRCm38_introns.bed", header=F, sep="\t",stringsAsFactors = F,data.table=F)
  introns=introns[grepl("_",introns$V1)==F,]
  colnames(introns)=c('seqname','start','end','attribute','V5','strand')
  introns=separate(introns,attribute,into=c('transcript_id','feature','exon_number','level','chr2','intronnumber','dir'),remove = T,sep = "_")
  introns$transcript_id=removeVersion(introns$transcript_id)
      # introns_canon=introns[introns$transcript_id%in%canonical$transcript_id,]
  introns$start=introns$start+1
  introns$exon_number=as.numeric(introns$exon_number)+1  
  
  
  
  mm10_trans=mm10all[mm10all$feature%in%'transcript',]
  mm10_trans=mm10_trans[mm10_trans$transcript_id%in%canonical$transcript,]
    mm10_trans=mm10_trans[mm10_trans$gene_type%in%'protein_coding',]

  
  mm10_UTR=mm10all[mm10all$feature%in%'UTR',]
  mm10_UTR=mm10_UTR[mm10_UTR$transcript_id%in%mm10_trans$transcript_id,]  
  mm10_exon=mm10all[mm10all$feature%in%c("exon"),]
  mm10_exon=mm10_exon[mm10_exon$transcript_id%in%mm10_trans$transcript_id,]  
  mm10_introns=introns[introns$transcript_id%in%mm10_trans$transcript_id,]
  mm10_introns=mm10_introns[mm10_introns$transcript_id%in%mm10_trans$transcript_id,]  

              
TranscCompOut=as.data.frame(matrix(nrow=100,ncol=7))
colnames(TranscCompOut)=c('exonCDS','UTR','5UTR','3UTR','intron','total','total2')

################################################
#### SELECT Locations
################################################


ntimes=1

Rano_fracUTR=matrix(nrow=ntimes,ncol=2)
Rano_frac3UTR=matrix(nrow=ntimes,ncol=2)
Rano_frac5UTR=matrix(nrow=ntimes,ncol=2)
Rano_fracExon=matrix(nrow=ntimes,ncol=2)
Rano_fracIntron=matrix(nrow=ntimes,ncol=2)

for (r in 1:1) {
# for (r in 1:1) {

randTrans=sample(1:nrow(mm10_trans),( nrow(checkout2_LS)*ntimes), replace=TRUE)
# randTrans=sample(1:nrow(mm10_trans),( nrow(checkout2_LS)), replace=TRUE)
# randTrans=sample(1:nrow(mm10_trans),136, replace=TRUE)
      trans_rand=mm10_trans[randTrans,]
      
            trans_rand$randLoc=NA
            trans_rand$UTROL=NA
            trans_rand$exonOL=NA
            trans_rand$intronOL=NA
  for (x in 1:(nrow(checkout2_LS)*ntimes)) {
  # for (x in 1:(nrow(checkout2_LS))) {
  # for (x in 1:136) {

    if (TranscComponent=='Exons') {
            ExonRandX=mm10_exon[mm10_exon$transcript_id%in%trans_rand[x,'transcript_id'],]
        
            locationRange=c(ExonRandX[1,'start']:ExonRandX[1,'end'])
            if (nrow(ExonRandX)>1) {
                for (n in 2:nrow(ExonRandX)) {
                  locationRange=c(locationRange,c(ExonRandX[n,'start']:ExonRandX[n,'end'])) }
            }
            }
 
     if (TranscComponent=='All Protein Coding') {
            locationRange=c(trans_rand[x,'start']:trans_rand[x,'end'])
     }
    
    randLoc=sample(locationRange, 1, replace=TRUE)
    trans_rand[x,'randLoc']=randLoc
    trans_randX=trans_rand[x,]
    
              trans_randXGR=GRanges(seqnames=trans_randX$seqname, 
                          ranges=IRanges(start=randLoc,end=randLoc),
                          strand=trans_randX$strand,
                          exon=trans_randX$exon_number
                          )
              
############################################    
### UTR          
    UTR_rand=mm10_UTR[mm10_UTR$transcript_id%in%trans_rand[x,'transcript_id'],]
      if (nrow(UTR_rand)==0) {trans_rand[x,'UTROL']=NA}
      if (nrow(UTR_rand)>0) {
              UTR_randGR=GRanges(seqnames=UTR_rand$seqname, 
                          ranges=IRanges(start=UTR_rand$start,end=UTR_rand$end),
                          strand=UTR_rand$strand,
                          exon=UTR_rand$exon_number,
                          transcript_id=UTR_rand$transcript_id,
                          UTR='UTR'
                          )

              gr = trans_randXGR
              gr2 = UTR_randGR
              xo=as.data.frame(GenomicRanges::findOverlaps(gr,gr2,type = "any",ignore.strand=F))
          
              UTR_Trans_GR=as.data.frame(gr2[xo$subjectHits])
              trans_UTR_GR=gr[xo$queryHits] 
              
              if (nrow(UTR_Trans_GR)==0) {trans_rand[x,'UTROL']=NA}
              if (nrow(UTR_Trans_GR)>0) {trans_rand[x,'UTROL']=unique(UTR_Trans_GR$exon)}
      }
    
############################################    
### Exons              
    exon_rand=mm10_exon[mm10_exon$transcript_id%in%trans_rand[x,'transcript_id'],]
      if (nrow(exon_rand)==0) {trans_rand[x,'exonOL']=NA}
      if (nrow(exon_rand)>0) {    
              exon_randGR=GRanges(seqnames=exon_rand$seqname, 
                          ranges=IRanges(start=exon_rand$start,end=exon_rand$end),
                          strand=exon_rand$strand,
                          exon=exon_rand$exon_number,
                          transcript_id=exon_rand$transcript_id,
                          exons='exons'
                          )
              gr = trans_randXGR
              gr2 = exon_randGR
              xo=as.data.frame(GenomicRanges::findOverlaps(gr,gr2,type = "any",ignore.strand=F))
          
              exon_Trans_GR=as.data.frame(gr2[xo$subjectHits])
              Trans_exon_GR=gr[xo$queryHits] 
      if (nrow(exon_Trans_GR)==0) {trans_rand[x,'exonOL']=NA}
      if (nrow(exon_Trans_GR)>0) {trans_rand[x,'exonOL']=exon_Trans_GR$exon}
            }

############################################    
########### Introns              
    introns_rand=mm10_introns[mm10_introns$transcript_id%in%trans_rand[x,'transcript_id'],]
      if (nrow(introns_rand)==0) {trans_rand[x,'intronOL']=NA }
      if (nrow(introns_rand)>0) {
             introns_randGR=GRanges(seqnames=introns_rand$chr,
                          ranges=IRanges(start=introns_rand$start,end=introns_rand$end),
                          strand=introns_rand$strand,
                          exon=introns_rand$exon_number,
                          transcript_id=introns_rand$transcript_id,
                          introns="introns"
                          )
              gr = trans_randXGR
              gr2 = introns_randGR
              xo=as.data.frame(GenomicRanges::findOverlaps(gr,gr2,type = "any",ignore.strand=F))

              introns_Trans_GR=as.data.frame(gr2[xo$subjectHits])
              trans_intron_GR=gr[xo$queryHits]

              if (nrow(introns_Trans_GR)==0) {trans_rand[x,'intronOL']=NA }
              if (nrow(introns_Trans_GR)>0) { trans_rand[x,'intronOL']=introns_Trans_GR$exon}
          }
              trans_rand$TranscComp=rowSums((trans_rand[,c("intronOL","exonOL","UTROL")]>0),na.rm = T)
       # remove(introns_rand,exon_rand,UTR_rand,trans_randX,trans_randXGR)      
       remove(exon_rand,UTR_rand,trans_randX,trans_randXGR)
  }    
            
            trans_rand$Transc_Class=NA
            
    TranscCompOut[r,'exonCDS']= sum((is.na(trans_rand$exonOL)==F)&(trans_rand$exonOL>0),na.rm = T)/nrow(trans_rand)
   TranscCompOut[r,'UTR'] =  sum((is.na(trans_rand$UTROL)==F)&(trans_rand$UTROL>0),na.rm = T)/nrow(trans_rand)
   TranscCompOut[r,'5UTR'] =  sum((is.na(trans_rand$UTROL)==F)&(trans_rand$UTROL==1),na.rm = T)/nrow(trans_rand)
   TranscCompOut[r,'3UTR'] =   sum((is.na(trans_rand$UTROL)==F)&(trans_rand$UTROL>1),na.rm = T)/nrow(trans_rand)   
   TranscCompOut[r,'intron'] =  sum((is.na(trans_rand$intronOL)==F)&(trans_rand$intronOL>0),na.rm = T)/nrow(trans_rand)
   TranscCompOut[r,'total'] =  rowSums(TranscCompOut[r,c('exonCDS','UTR','intron')],na.rm=T)
      TranscCompOut[r,'total2'] =  rowSums(TranscCompOut[r,c('exonCDS','3UTR','5UTR','intron')],na.rm=T)
      
      
      
      trans_rand[(is.na(trans_rand$exonOL)==F)&(trans_rand$exonOL>0),'TranscRegion']='exon_CDS'
      trans_rand[(is.na(trans_rand$UTROL)==F)&trans_rand$UTROL>0,'TranscRegion']='UTR'
      trans_rand[(is.na(trans_rand$UTROL)==F)&trans_rand$UTROL==1,'TranscRegion']='5UTR'
      trans_rand[(is.na(trans_rand$UTROL)==F)&trans_rand$UTROL>1,'TranscRegion']='3UTR'
      trans_rand[(is.na(trans_rand$intronOL)==F)&trans_rand$intronOL>0,'TranscRegion']='intron'

      Rano_fracUTR[r,1]= sum((is.na(trans_rand$UTROL)==F)&(trans_rand$UTROL>0),na.rm = T)
      Rano_fracUTR[r,2]=x
      
      Rano_fracUTR[r,1]= sum((is.na(trans_rand$UTROL)==F)&(trans_rand$UTROL>0),na.rm = T)
        Rano_fracUTR[r,2]=x
      Rano_frac5UTR[r,1]= sum((is.na(trans_rand$UTROL)==F)&(trans_rand$UTROL==1),na.rm = T)
        Rano_frac5UTR[r,2]=x
      Rano_frac3UTR[r,1]= sum((is.na(trans_rand$UTROL)==F)&(trans_rand$UTROL>1),na.rm = T)
        Rano_frac3UTR[r,2]=x
      Rano_fracExon[r,1]= sum((is.na(trans_rand$exonOL)==F)&(trans_rand$exonOL>0),na.rm = T)
        Rano_fracExon[r,2]=x
      Rano_fracIntron[r,1]= sum((is.na(trans_rand$intronOL)==F)&(trans_rand$intronOL>0),na.rm = T)
        Rano_fracIntron[r,2]=x
          write.table(Rano_fracUTR, file=paste0('./structure/Random/',CLIPtype,'/Rano_fracUTR.txt'), quote=F, sep="\t", row.names=T, col.names=NA)
          write.table(Rano_frac5UTR, file=paste0('./structure/Random/',CLIPtype,'/Rano_frac5UTR.txt'), quote=F, sep="\t", row.names=T, col.names=NA)
          write.table(Rano_frac3UTR, file=paste0('./structure/Random/',CLIPtype,'/Rano_frac3UTR.txt'), quote=F, sep="\t", row.names=T, col.names=NA)
          write.table(Rano_fracExon, file=paste0('./structure/Random/',CLIPtype,'/Rano_fracExon.txt'), quote=F, sep="\t", row.names=T, col.names=NA)
          write.table(Rano_fracIntron, file=paste0('./structure/Random/',CLIPtype,'/Rano_fracIntron.txt'), quote=F, sep="\t", row.names=T, col.names=NA)

print(r)
}   
```

```{r ,echo=F,warning=F,eval=T,include=T,fig.asp=1.1,fig.height=2, fig.width=2.5, figures-side, fig.show="hold"}

trans_rand$Data_Set="Random_RNA_Transcripts"
gg2=ggplot(trans_rand,aes(fill=TranscRegion,x=Data_Set) )+geom_bar(aes(y = (..count..)/sum(..count..)),position="stack",width=.5)+theme_classic()+ggtitle("Transcript Location of Random Transcript locations")+
ylab("Percent of Locations")+xlab("")+
# theme(plot.title = element_text(hjust = 0.5,size = 25),axis.title=element_text(size=20),axis.text=element_text(size=15),legend.text = element_text( size=30))+
 scale_fill_brewer(palette = "Dark2") 
gg2
# ggplotly(gg2)
# ggplotly(gg+gg2)




# p2=rbind(p[,c('TranscRegion','Data_Set')],
#     trans_rand[,c('TranscRegion','Data_Set')])
# gg2=ggplot(p2,aes(fill=TranscRegion,x=Data_Set) )+geom_bar(aes(y = (..count..)/sum(..count..)),position="stack",width=.5)+theme_classic()+ggtitle("Transcript Location of Random Exonic locations")+
# # theme(plot.title = element_text(hjust = 0.5,size = 25),axis.title=element_text(size=20),axis.text=element_text(size=15),legend.text = element_text( size=30))+
#  scale_fill_brewer(palette = "Dark2") 
# ggplotly(gg2)

Rano_fracUTR=as.data.frame(Rano_fracUTR)
Rano_fracUTR$frac=Rano_fracUTR[,1]/Rano_fracUTR[,2]


```






```{r ,echo=F,warning=F,eval=F,include=F}

      UTRRando=as.data.frame(read.delim(paste0(inRandom,'/Random1/',CLIPtype,'/Rano_Exons_fracUTR.txt'),comment.char = "%",row.names = NULL,header = T,sep = "\t",stringsAsFactors = F))
          colnames(UTRRando)=c('Nrepeats','UTR','SampleSize')
      UTRRando$UTRfrac=UTRRando$UTR/UTRRando$SampleSize    
for (rand in 2:12) {
  
      UTRRando2=as.data.frame(read.delim(paste0(inRandom,'/Random',rand,'/',CLIPtype,'/Rano_Exons_fracUTR.txt'),comment.char = "%",header = T,sep = "\t",stringsAsFactors = F))
          colnames(UTRRando2)=c('Nrepeats','UTR','SampleSize')
          UTRRando2$UTRfrac=UTRRando2$UTR/UTRRando2$SampleSize    

      UTRRando=rbind(UTRRando,UTRRando2)
}

ggplot(UTRRando,aes(x=UTRfrac))+geom_density(size= 2) + theme_classic()+ylab("Density")+xlab("locations in UTR (%)")+ggtitle('Fraction of Exonic CLIP locations that fall with in UTR' )+
       geom_vline(xintercept=(nrow(checkout2_LS[is.na(checkout2_LS$UTR)==F,])/nrow(checkout2_LS[is.na(checkout2_LS$TranscRegion)==F,])),colour="red",size=2) +xlim(0,1)

ggplot(UTRRando,aes(x=UTRfrac))+geom_density(size= 2) + theme_classic()+ylab("Density")+xlab("locations in UTR (%)")+ggtitle('Fraction of Exonic CLIP locations that fall with in UTR' )+
       geom_vline(xintercept=(nrow(checkout2_LS[(checkout2_LS$TranscRegion)%in%c('3UTR','5UTR'),])/nrow(checkout2_LS[is.na(checkout2_LS$TranscRegion)==F,])),colour="red",size=2) +xlim(0,1)

######################################################## 
######### 5UTR

 UTR5Rando=as.data.frame(read.delim(paste0(inRandom,'/Random1/',CLIPtype,'/Rano_Exons_frac5UTR.txt'),comment.char = "%",row.names = NULL,header = T,sep = "\t",stringsAsFactors = F))
          colnames(UTR5Rando)=c('Nrepeats','5UTR','SampleSize')
      UTR5Rando$UTR5frac=UTR5Rando$`5UTR`/UTR5Rando$SampleSize    
for (rand in 2:12) {
  
      UTR5Rando2=as.data.frame(read.delim(paste0(inRandom,'/Random',rand,'/',CLIPtype,'/Rano_Exons_frac5UTR.txt'),comment.char = "%",header = T,sep = "\t",stringsAsFactors = F))
          colnames(UTR5Rando2)=c('Nrepeats','5UTR','SampleSize')
          UTR5Rando2$UTR5frac=UTR5Rando2$`5UTR`/UTR5Rando2$SampleSize    

      UTR5Rando=rbind(UTR5Rando,UTR5Rando2)
}

ggplot(UTR5Rando,aes(x=UTR5frac))+geom_density(size= 2,color='chocolate2') + theme_classic()+ylab("Density")+xlab("locations in 5'UTR (%)")+ggtitle('Fraction of Exonic CLIP locations that fall with in UTR' )+
       geom_vline(xintercept=(nrow(checkout2_LS[is.na(checkout2_LS$`5UTR`)==F,])/nrow(checkout2_LS[is.na(checkout2_LS$TranscRegion)==F,])),colour="black",size=2) +xlim(0,1)


######################################################## 
######### 3UTR

 UTR3Rando=as.data.frame(read.delim(paste0(inRandom,'/Random1/',CLIPtype,'/Rano_Exons_frac3UTR.txt'),comment.char = "%",row.names = NULL,header = T,sep = "\t",stringsAsFactors = F))
          colnames(UTR3Rando)=c('Nrepeats','3UTR','SampleSize')
      UTR3Rando$UTR3frac=UTR3Rando$`3UTR`/UTR3Rando$SampleSize    
for (rand in 2:12) {
  
      UTR3Rando2=as.data.frame(read.delim(paste0(inRandom,'/Random',rand,'/',CLIPtype,'/Rano_Exons_frac3UTR.txt'),comment.char = "%",header = T,sep = "\t",stringsAsFactors = F))
          colnames(UTR3Rando2)=c('Nrepeats','3UTR','SampleSize')
          UTR3Rando2$UTR3frac=UTR3Rando2$`3UTR`/UTR3Rando2$SampleSize    

      UTR3Rando=rbind(UTR3Rando,UTR3Rando2)
}

ggplot(UTR3Rando,aes(x=UTR3frac))+geom_density(size= 2,color='springgreen4') + theme_classic()+ylab("Density")+xlab("locations in 3'UTR (%)")+ggtitle("Fraction of Exonic CLIP locations that fall with in 3'UTR" )+
       geom_vline(xintercept=(nrow(checkout2_LS[is.na(checkout2_LS$`3UTR`)==F,])/nrow(checkout2_LS[is.na(checkout2_LS$TranscRegion)==F,])),colour="black",size=2) +xlim(0,1)




######################################################## 
######### CDS

CDSRando=cbind(UTR5Rando,UTR3Rando[,c('3UTR','UTR3frac')])
CDSRando$CDS=CDSRando$SampleSize-(CDSRando$`5UTR`+CDSRando$`3UTR`)
CDSRando$CDSfrac=CDSRando$CDS/CDSRando$SampleSize

ggplot(CDSRando,aes(x=CDSfrac))+geom_density(size= 2,color='purple') + theme_classic()+ylab("Density")+xlab("locations in CDS (%)")+ggtitle('Fraction of Exonic CLIP locations that fall with in CDS' )+
       geom_vline(xintercept=(nrow(checkout2_LS[checkout2_LS$TranscRegion%in%'exons_CDS',])/nrow(checkout2_LS[is.na(checkout2_LS$TranscRegion)==F,])),colour="black",size=2) +xlim(0,1)

```





```{r ,echo=F,warning=F,eval=F,include=F}

canonical=fread("/Users/homanpj/OneDrive - National Institutes of Health/Resources/ref/mm10/Gencode_VM23/fromUCSC/KnownCanonical/KnownCanonical_GencodeM23_GRCm38.txt", header=T, sep="\t",stringsAsFactors = F,data.table=F)
  canonical$transcript=removeVersion(canonical$transcript)
  canonical$protein=removeVersion(canonical$protein)
 
###############
  
mm10all=fread("/Users/homanpj/OneDrive - National Institutes of Health/Resources/ref/mm10/Gencode_VM23/fromGencode/gencode.vM23.basic.annotation.gtf.txt", header=T, sep="\t",stringsAsFactors = F,data.table=F)
    mm10all$transcript_id=removeVersion(mm10all$transcript_id)
    mm10all$gene_id=removeVersion(mm10all$gene_id)
    mm10all$start=as.numeric(as.character(mm10all$start))
    mm10all$end=as.numeric(as.character(mm10all$end))
  
  mm10_transc=mm10all[mm10all$feature%in%c("transcript"),]
  mm10_transc=mm10_transc[mm10_transc$transcript_id%in%canonical$transcript,]
  mm10_transc=mm10_transc[mm10_transc$transcript_type%in%'protein_coding',]
  
  
###############
  
mm10_3UTRall=fread("/Users/homanpj/OneDrive - National Institutes of Health/Resources/ref/mm10/Gencode_VM23/fromUCSC/KnownGene/KnownGene_GRCm38_GencodeV23_3UTR.bed", header=F, sep="\t",stringsAsFactors = F,data.table=F)
  mm10_3UTRall=mm10_3UTRall[grepl("_",mm10_3UTRall$V1)==F,]
  colnames(mm10_3UTRall)=c('chr','start','end','attribute','V5','strand')
  mm10_3UTRall=separate(mm10_3UTRall,attribute,into=c('transcript_id','feature','exon_number','level','chr2','start+1','dir'),remove = T,sep = "_")
  mm10_3UTRall$transcript_id=removeVersion(mm10_3UTRall$transcript_id)
  mm10_3UTRall$start=mm10_3UTRall$start+1
  mm10_3UTRall$exon_number=as.numeric(mm10_3UTRall$exon_number)+1
  mm10_3UTRall$width=mm10_3UTRall$end-mm10_3UTRall$start
          mm10_3UTR=mm10_3UTRall[mm10_3UTRall$transcript_id%in%mm10_transc$transcript_id,]
  

####################
        
checkout2_UTRlength=merge(checkout2,isoout[,c('ID','USEtranscript')],by.x='CLIP_ID',by.y='ID')
checkout2_UTRlength=separate(checkout2_UTRlength,CLIP_ID,c('chr','start'),sep = ":",remove = F)
checkout2_UTRlength=separate(checkout2_UTRlength,start,c('start','end'),sep = "-",remove = T)
checkout2_UTRlength_GR=GRanges(seqnames=checkout2_UTRlength$chr, 
                          ranges=IRanges(start=checkout2_UTRlength$crosslink,end=checkout2_UTRlength$crosslink),
                          strand=checkout2_UTRlength$strand,
                          transcript_id=checkout2_UTRlength$USEtranscript,
                          CLIP_ID=checkout2_UTRlength$CLIP_ID
                          # chrm=checkout2_UTRmm10$seqname
                          )
checkout2_UTRmm10=mm10_3UTRall[mm10_3UTRall$transcript_id%in%checkout2_UTRlength$transcript_id,]


checkout2_UTRmm10_GR=GRanges(seqnames=checkout2_UTRmm10$chr, 
                          ranges=IRanges(start=checkout2_UTRmm10$start,end=checkout2_UTRmm10$end),
                          strand=checkout2_UTRmm10$strand,
                          transcript_id=checkout2_UTRmm10$transcript_id
                          # chrm=checkout2_UTRmm10$seqname
                          )
        gr = checkout2_UTRlength_GR
        gr2 = checkout2_UTRmm10_GR
        xo=as.data.frame(GenomicRanges::findOverlaps(gr,gr2,type = "any",ignore.strand=F))
          
        mm10UTR_CLIP=as.data.frame(gr2[xo$subjectHits])
        mm10UTR_CLIP$Class='CLIP'
                  # qh=as.data.frame(gr[xo$queryHits],row.names = NULL)
                  # colnames(qh)=paste0(colnames(qh),"_mm10")
                  # sh=as.data.frame(gr2[xo$subjectHits],row.names = NULL)
        CLIP_mm10UTR=gr[xo$queryHits] 
        

        
        

  mm10_UTR3UTRlength=mm10_3UTR
  mm10_UTR3UTRlength$Class="ALL 3'UTR"
  mm10_UTR3UTRlength$width=mm10_UTR3UTRlength$end-mm10_UTR3UTRlength$start
  mm10_UTR3UTRlength=mm10_UTR3UTRlength[mm10_UTR3UTRlength$width>5,]
 
      mm10_UTR3UTRlength=mm10_UTR3UTRlength[order(mm10_UTR3UTRlength$exon_number,decreasing = F),]
      
      # dup=unique(mm10_UTR3UTRlength[duplicated(mm10_UTR3UTRlength$transcript_id),'transcript_id'])
      # 
      # for (d in 1:length(dup)) {
      #   dd=mm10_UTR3UTRlength[mm10_UTR3UTRlength$transcript_id%in%dup[d],]
      #   m=dd[which(max(dd$exon_number)),]
      # }
      
      mm10_UTR3UTRlength=mm10_UTR3UTRlength[duplicated(mm10_UTR3UTRlength$transcript_id)==F,]
    
      # mm10_UTR3UTRlength$width=(mm10_UTR3UTRlength$width)/1000
      # mm10UTR_CLIP$width=(mm10UTR_CLIP$width)/1000
      
      
ggplot(mm10_UTR3UTRlength,aes(x=width))+
  geom_density(size= 2,color='blue') + 
  # geom_density(data = mm10UTR_CLIP,size=2,color='red')
  theme_classic()+ylab("Density")+xlab("3'UTR Width (nt)")#+
   # scale_y_continuous(trans='log2')
  #ggtitle('Fraction of Exonic CLIP locations that fall with in CDS' )+
       #xlim(0,1)


ggplot(mm10UTR_CLIP,aes(x=width))+
  geom_density(size= 2,color='blue') + 
  theme_classic()+ylab("Density")+xlab("3'UTR Width (nt)")#+
   # scale_y_continuous(trans='log2')
  #ggtitle('Fraction of Exonic CLIP locations that fall with in CDS' )+
       #xlim(0,1)

p=rbind(mm10_UTR3UTRlength[,c('width','Class')],mm10UTR_CLIP[,c('width','Class')])
  p$logwidth=log2(p$width)

ggplot(p,aes(x=width,line=Class,color=Class))+
  geom_density(size= 2) + 
  theme_classic()+ylab("Density")+xlab("3'UTR Width (nt)")#+
   # scale_y_continuous(trans='log2')+
  #ggtitle('Fraction of Exonic CLIP locations that fall with in CDS' )+
       # xlim(0,125000)

ggplot(p,aes(x=logwidth,line=Class,color=Class))+
  geom_density(size= 2) + 
  theme_classic()+ylab("Density")+xlab("3'UTR Width (Log2(nt))")#+
   # scale_y_continuous(trans='log2')+
  #ggtitle('Fraction of Exonic CLIP locations that fall with in CDS' )+
       # xlim(0,125000)

p=rbind(mm10_UTR3UTRlength[,c('width','Class')],mm10UTR_CLIP[,c('width','Class')])

ggplot(p,aes(x=width,line=Class,color=Class))+
  geom_histogram(size= 2,binwidth = 100) + 
  theme_classic()+ylab("Density")+xlab("3'UTR Width (Log2(nt))")+xlim(10000,125000)





maxbin=10000
binwidth=500

# maxbin=15
# binwidth=1

 # mm10_UTR3UTRlength_hist=(hist(log2(mm10_UTR3UTRlength$width),seq(0,max(mm10_UTR3UTRlength$width+binwidth),binwidth),plot = F))
 mm10_UTR3UTRlength_hist=(hist((mm10_UTR3UTRlength$width),seq(0,max(mm10_UTR3UTRlength$width+binwidth),binwidth),plot = F))
mm10_UTR3UTRlength_hist=as.data.frame(cbind(mm10_UTR3UTRlength_hist$breaks[1:length(mm10_UTR3UTRlength_hist$breaks)-1],mm10_UTR3UTRlength_hist$counts,mm10_UTR3UTRlength_hist$density))
colnames(mm10_UTR3UTRlength_hist)=c("bin","count","density")
mm10_UTR3UTRlength_hist$percent=mm10_UTR3UTRlength_hist$count/sum(mm10_UTR3UTRlength_hist$count)
addition=as.data.frame(matrix(nrow = 1,ncol = 4,NA));colnames(addition)=colnames(mm10_UTR3UTRlength_hist)
addition$bin=maxbin
mm10_UTR3UTRlength_plot=rbind(mm10_UTR3UTRlength_hist[mm10_UTR3UTRlength_hist$bin<maxbin,],addition)
mm10_UTR3UTRlength_plot[mm10_UTR3UTRlength_plot$bin==(maxbin),c('count','density','percent')]=colSums(mm10_UTR3UTRlength_hist[mm10_UTR3UTRlength_hist$bin>=maxbin,c('count','density','percent')])
mm10_UTR3UTRlength_plot$class="ALL 3'UTR"


ggplot(mm10_UTR3UTRlength_plot,aes(x=bin,y=percent,line=class,color=class))+
  geom_col(size= 2) + 
  theme_classic()+ylab("percent")+xlab("3'UTR Width")


 # mm10UTR_CLIP_hist=(hist(log2(mm10UTR_CLIP$width),seq(0,max(mm10UTR_CLIP$width+binwidth),binwidth),plot = F))
 mm10UTR_CLIP_hist=(hist((mm10UTR_CLIP$width),seq(0,max(mm10UTR_CLIP$width+binwidth),binwidth),plot = F))
mm10UTR_CLIP_hist=as.data.frame(cbind(mm10UTR_CLIP_hist$breaks[1:length(mm10UTR_CLIP_hist$breaks)-1],mm10UTR_CLIP_hist$counts,mm10UTR_CLIP_hist$density))
colnames(mm10UTR_CLIP_hist)=c("bin","count","density")
mm10UTR_CLIP_hist$percent=mm10UTR_CLIP_hist$count/sum(mm10UTR_CLIP_hist$count)
# colSums(mm10UTR_CLIP_hist)
addition=as.data.frame(matrix(nrow = 1,ncol = 4,NA));colnames(addition)=colnames(mm10UTR_CLIP_hist)
addition$bin=maxbin
mm10UTR_CLIP_plot=rbind(mm10UTR_CLIP_hist[mm10UTR_CLIP_hist$bin<maxbin,],addition)
# mm10UTR_CLIP_plot[mm10UTR_CLIP_plot$bin==(maxbin+1),'bin']=paste0('>',maxbin)
mm10UTR_CLIP_plot[mm10UTR_CLIP_plot$bin==(maxbin),c('count','density','percent')]=colSums(mm10UTR_CLIP_hist[mm10UTR_CLIP_hist$bin>=maxbin,c('count','density','percent')])
# colSums(mm10UTR_CLIP_plot)
mm10UTR_CLIP_plot$class="CLIP 3'UTR"


ggplot(mm10UTR_CLIP_plot,aes(x=bin,y=percent,line=class,color=class))+
  geom_col(size= 2) + 
  theme_classic()+ylab("percent")+xlab("3'UTR Width")

sum(mm10UTR_CLIP_plot$count)
sum(mm10UTR_CLIP_hist$count)
colSums(mm10UTR_CLIP_hist)


combPlot=rbind(mm10UTR_CLIP_plot,mm10_UTR3UTRlength_plot)
# combPlot$bin=as.character(combPlot$bin)


ggplot(combPlot,aes(x=bin,y=percent,line=class,color=class,fill=class))+
  geom_col(size= 0, position = 'dodge2',width=`binwidth`/1.5) + 
  theme_classic()+ylab("Percent of 3'UTR")+xlab("3'UTR Width")

```

```{r ,echo=F,warning=F,eval=F,include=F}

        #########################    
        ### RANDOM     
        #########################    
# TranscCompOut=as.data.frame(matrix(nrow=100,ncol=7))
# colnames(TranscCompOut)=c('exonCDS','UTR','5UTR','3UTR','intron','total','total2')
# 
# 
# 
# ntimes=2
# 
# for (r in 1:1) {
# 
# randTrans=sample(1:nrow(mm10_trans),( nrow(Peaksdata4_transc)*ntimes), replace=TRUE)
#       trans_rand=mm10_trans[randTrans,]
#       
#             trans_rand$randLoc=NA
#             trans_rand$UTROL=NA
#             trans_rand$exonOL=NA
#             trans_rand$intronOL=NA
#   for (x in 1:(nrow(Peaksdata4_transc)*ntimes)) {
#     trans_rand[x,'randLoc']=sample(trans_rand[x,'start']:trans_rand[x,'end'], 1, replace=TRUE)
#         trans_randX=trans_rand[x,]
# 
#               trans_randXGR=GRanges(seqnames=trans_randX$seqname, 
#                           ranges=IRanges(start=trans_randX$randLoc,end=trans_randX$randLoc),
#                           strand=trans_randX$strand,
#                           exon=trans_randX$exon_number
#                           )
#               
#     UTR_rand=mm10_UTR[mm10_UTR$transcript_id%in%trans_randX$transcript_id,]
#       if (nrow(UTR_rand)==0) {trans_rand[x,'UTROL']=NA}
#       if (nrow(UTR_rand)>0) {
#               UTR_randGR=GRanges(seqnames=UTR_rand$seqname, 
#                           ranges=IRanges(start=UTR_rand$start,end=UTR_rand$end),
#                           strand=UTR_rand$strand,
#                           exon=UTR_rand$exon_number,
#                           transcript_id=UTR_rand$transcript_id,
#                           UTR='UTR'
#                           )
#               
#               gr = trans_randXGR
#               gr2 = UTR_randGR
#               xo=as.data.frame(GenomicRanges::findOverlaps(gr,gr2,type = "any",ignore.strand=F))
#           
#               UTR_Trans_GR=as.data.frame(gr2[xo$subjectHits])
#               trans_UTR_GR=gr[xo$queryHits] 
#               
#               if (nrow(UTR_Trans_GR)==0) {trans_rand[x,'UTROL']=NA}
#               if (nrow(UTR_Trans_GR)>0) {trans_rand[x,'UTROL']=unique(UTR_Trans_GR$exon)}
#             }
#               
#     exon_rand=mm10_exon[mm10_exon$transcript_id%in%trans_randX$transcript_id,]
#       if (nrow(exon_rand)==0) {trans_rand[x,'exonOL']=NA}
#       if (nrow(exon_rand)>0) {    
#               exon_randGR=GRanges(seqnames=exon_rand$seqname, 
#                           ranges=IRanges(start=exon_rand$start,end=exon_rand$end),
#                           strand=exon_rand$strand,
#                           exon=exon_rand$exon_number,
#                           transcript_id=exon_rand$transcript_id,
#                           exons='exons'
#                           )
#               gr = trans_randXGR
#               gr2 = exon_randGR
#               xo=as.data.frame(GenomicRanges::findOverlaps(gr,gr2,type = "any",ignore.strand=F))
#           
#               exon_Trans_GR=as.data.frame(gr2[xo$subjectHits])
#               Trans_exon_GR=gr[xo$queryHits] 
#       if (nrow(exon_Trans_GR)==0) {trans_rand[x,'exonOL']=NA}
#       if (nrow(exon_Trans_GR)>0) {trans_rand[x,'exonOL']=exon_Trans_GR$exon}
#             }
#               
#     introns_rand=mm10_introns[mm10_introns$transcript_id%in%trans_randX$transcript_id,]
#       if (nrow(introns_rand)==0) {trans_rand[x,'intronOL']=NA }
#       if (nrow(introns_rand)>0) {    
#              introns_randGR=GRanges(seqnames=introns_rand$chr, 
#                           ranges=IRanges(start=introns_rand$start,end=introns_rand$end),
#                           strand=introns_rand$strand,
#                           exon=introns_rand$exon_number,
#                           transcript_id=introns_rand$transcript_id,
#                           introns="introns"
#                           )
#               gr = trans_randXGR
#               gr2 = introns_randGR
#               xo=as.data.frame(GenomicRanges::findOverlaps(gr,gr2,type = "any",ignore.strand=F))
#           
#               introns_Trans_GR=as.data.frame(gr2[xo$subjectHits])
#               trans_intron_GR=gr[xo$queryHits]
#              
#               if (nrow(introns_Trans_GR)==0) {trans_rand[x,'intronOL']=NA }
#               if (nrow(introns_Trans_GR)>0) { trans_rand[x,'intronOL']=introns_Trans_GR$exon}
#           }
#               trans_rand$TranscComp=rowSums((trans_rand[,c("intronOL","exonOL","UTROL")]>0),na.rm = T)
#        remove(introns_rand,exon_rand,UTR_rand,trans_randX,trans_randXGR)      
#   }    
#             trans_rand$Transc_Class=NA
#             
#     TranscCompOut[r,'exonCDS']= sum((is.na(trans_rand$exonOL)==F)&(trans_rand$exonOL>0),na.rm = T)/nrow(trans_rand)
#    TranscCompOut[r,'UTR'] =  sum((is.na(trans_rand$UTROL)==F)&(trans_rand$UTROL>0),na.rm = T)/nrow(trans_rand)
#    TranscCompOut[r,'5UTR'] =  sum((is.na(trans_rand$UTROL)==F)&(trans_rand$UTROL==1),na.rm = T)/nrow(trans_rand)
#    TranscCompOut[r,'3UTR'] =   sum((is.na(trans_rand$UTROL)==F)&(trans_rand$UTROL>1),na.rm = T)/nrow(trans_rand)   
#    TranscCompOut[r,'intron'] =  sum((is.na(trans_rand$intronOL)==F)&(trans_rand$intronOL>0),na.rm = T)/nrow(trans_rand)
#    TranscCompOut[r,'total'] =  rowSums(TranscCompOut[r,c('exonCDS','UTR','intron')],na.rm=T)
#       TranscCompOut[r,'total2'] =  rowSums(TranscCompOut[r,c('exonCDS','3UTR','5UTR','intron')],na.rm=T)
#       
#       
#       
#       trans_rand[(is.na(trans_rand$exonOL)==F)&(trans_rand$exonOL>0),'Transc_Class']='exonCDS'
#       trans_rand[(is.na(trans_rand$UTROL)==F)&trans_rand$UTROL>0,'Transc_Class']='UTR'
#       trans_rand[(is.na(trans_rand$UTROL)==F)&trans_rand$UTROL==1,'Transc_Class']='5UTR'
#       trans_rand[(is.na(trans_rand$UTROL)==F)&trans_rand$UTROL>1,'Transc_Class']='3UTR'
#       trans_rand[(is.na(trans_rand$intronOL)==F)&trans_rand$intronOL>0,'Transc_Class']='intron'
# 
# print(r)
# }   
# 
# trans_rand$Transcript="RNA"
# gg=ggplot(trans_rand,aes(fill=Transc_Class,x=Transcript) )+geom_bar(position="stack",width=.5)+theme_classic()+ggtitle("Number of peaks by catagory\n(Determined by unique reads only)")+
# # theme(plot.title = element_text(hjust = 0.5,size = 25),axis.title=element_text(size=20),axis.text=element_text(size=15),legend.text = element_text( size=30))+
#  scale_fill_brewer(palette = "Dark2") 
# ggplotly(gg)
```




## Create output file  

```{r ,echo=F,warning=F,eval=F,include=F}
## Create output file  


checkout2_out=checkout2[,c('CLIP_ID','ID','exon#',"TranscRegion",'sequence','qgrsScore','qgrsdist','qgrsSEQ','DeltaG','RowMax_MeanChuckBPP','upstream_exon_dist')]
colnames(checkout2_out)=c(c('ID','ID_CL','exon#',"TranscRegion",'sequence','GQuad_score','GQuad_distancefrom_CLIP_peak','GQuad_sequence','DeltaG','Maxum_BPP_for_10ntChucnk','exon_dist'))
colnames(checkout2_out)=paste0('Upstream: ',colnames(checkout2_out))
colnames(checkout2_out)=gsub("Upstream: CLIP_ID","ID",colnames(checkout2_out))
colnames(checkout2_out)=gsub("Upstream: ID_CL","ID_CL",colnames(checkout2_out))

checkout2_dn_out=checkout2_dn[,c('CLIP_ID','exon#','sequence','qgrsScore','qgrsdist','qgrsSEQ','DeltaG','RowMax_MeanChuckBPP')]
colnames(checkout2_dn_out)=c(c('ID','exon#','sequence','GQuad_score','GQuad_distancefrom_CLIP_peak','GQuad_sequence','DeltaG','Maxum_BPP_for_10ntChucnk','exon_dist'))
colnames(checkout2_dn_out)=paste0('Downstream: ',colnames(checkout2_dn_out))
colnames(checkout2_dn_out)=gsub("Downstream: CLIP_ID","ID",colnames(checkout2_dn_out))

Peaksdata4_chunkEx_out=Peaksdata4_chunkEx[,c('ID','strand',"Same Strand: Host_gene_name","Same Strand: Host_gene_ensembl_id","Same Strand: Repeat_Type","Same Strand: RNA_type_Classification")]
Peaksdata4_chunkEx_out=merge(Peaksdata4_chunkEx_out ,checkout2_out,by.x='ID',by.y='Upstream: ID')
Peaksdata4_chunkEx_out=merge(Peaksdata4_chunkEx_out ,checkout2_dn_out,by.x='ID',by.y='Downstream: ID')


###########################
### ADD LINEsineDistance Vishal

# clmns=c('Crosslink_ID','SINE_LINE_id','Distance_on_mmRNA')
# Peaksdata4_chunkEx_out=merge(Peaksdata4_chunkEx_out,LISI_detail[,clmns],by.x='ID_CL',by.y='Crosslink_ID',all.x=T)
# colnames(Peaksdata4_chunkEx_out)[colnames(Peaksdata4_chunkEx_out)%in%c('SINE_LINE_id','Distance_on_mmRNA')]=c('SINE_LINE_id_ExonOnly','Distance_including_ExonsOnly')
# 
# Peaksdata4_chunkEx_out=merge(Peaksdata4_chunkEx_out,LISI_detail_transc[,clmns],by.x='ID_CL',by.y='Crosslink_ID',all.x=T)
# colnames(Peaksdata4_chunkEx_out)[colnames(Peaksdata4_chunkEx_out)%in%c('SINE_LINE_id','Distance_on_mmRNA')]=c('SINE_LINE_id_Exons-Introns','Distance_including_Exons-Introns')
########################
########################

########################
Peaksdata4_chunkEx_out=Peaksdata4_chunkEx_out[colnames(Peaksdata4_chunkEx_out)%in%'ID_CL'==F,]  
    write.table(Peaksdata4_chunkEx_out, file=paste0('./structure/chunkwindow/',CLIPtype,'/CLIP_Peaks_MD14_',CLIPtype,'_Gquad-NucCont_identification.txt'), quote=F, sep="\t", row.names=T, col.names=NA)

    
    
########
    ### Intron LINE/SINE dist
    
    Peaksdata4_chunkInall_out=merge(Peaksdata4_chunkInall[,c('ID','strand',"Same Strand: Host_gene_name","Same Strand: Host_gene_ensembl_id","Same Strand: Repeat_Type","Same Strand: RNA_type_Classification")],outDist[,c('CLIP_ID','Transcript','CLIP_LISI_Dist',"ID_LINESINE","repClass_LINESINE","repName_LINESINE")],by.x='ID',by.y='CLIP_ID',all.x=T)

        write.table(Peaksdata4_chunkInall_out, file=paste0('./structure/chunkwindow/',CLIPtype,'/CLIP_Peaks_MD14_Introns_LINESINEdist.txt'), quote=F, sep="\t", row.names=T, col.names=NA)

```
























```{r ,echo=F,warning=F,eval=F,include=F}

# pt= permTest(A=peaks.GR, B=LISI_exn.GR, randomize.function=resampleRegions,
#   evaluate.function=meanDistance,ntimes=100,count.once=TRUE)

##### downstream
checkout2_dn_LS=separate(checkout2_dn,n,into = c('chr','start'),sep=':',remove = F)
checkout2_dn_LS=separate(checkout2_dn_LS,start,into = c('start','end'),sep='-')
checkout2_dn_LS$start=as.numeric(checkout2_dn_LS$start)
checkout2_dn_LS$end=as.numeric(checkout2_dn_LS$end)

peaks_dn.GR = GRanges(seqnames=checkout2_dn_LS$chr, ranges=IRanges(start=checkout2_dn_LS$start,
end=checkout2_dn_LS$end),
strand=checkout2_dn_LS$strand)


pt2 <- overlapPermTest(A=peaks_dn.GR, B=LISI_exn.GR, ntimes=100)
numOverlaps(A=peaks_dn.GR, B=(LISI_exn.GR),count.once = T)
plot(pt2)
LZ2=localZScore(pt=pt2,
                A=peaks_dn.GR,
                B=LISI_exn.GR,
                window=3000,step=10,count.once=TRUE)
plot(LZ2)

pt= permTest(A=peaks_dn.GR, B=LISI_exn.GR, randomize.function=circularRandomizeRegions,
  evaluate.function=meanDistance,ntimes=100,count.once=TRUE)

# pt= permTest(A=peaks_dn.GR, B=LISI_exn.GR, randomize.function=resampleRegions,
#   evaluate.function=meanDistance,ntimes=100,count.once=TRUE)

plot(pt)
# 
# LZ=localZScore(pt=pt2,
#                 A=peaks.GR,
#                 B=LISI_exn.GR, 
#                 window=10000,step=50,count.once=TRUE)
# plot(pt,asp=1)
# plot(LZ)




###### Introns

Peaksdata4_IN=Peaksdata4[Peaksdata4$`Both Strand: RNA_type_Classification`%in%'protein_coding: Intron', ]
Peaksdata4_IN=Peaksdata4
Peaksdata4_IN=separate(Peaksdata4_IN,ID,into = c('chr','start'),sep=':',remove = F)
Peaksdata4_IN=separate(Peaksdata4_IN,start,into = c('start','end'),sep='-')
Peaksdata4_IN$start=as.numeric(Peaksdata4_IN$start)
Peaksdata4_IN$end=as.numeric(Peaksdata4_IN$end)
Peaksdata4_IN=Peaksdata4_IN[Peaksdata4_IN$chr%in%c('BK000964.3', 'GL456221.1')==F,]

peaksIN.GR = GRanges(seqnames=Peaksdata4_IN$chr, ranges=IRanges(start=Peaksdata4_IN$start,
end=Peaksdata4_IN$end),
strand=Peaksdata4_IN$strand)


introns=fread("/Users/homanpj/OneDrive - National Institutes of Health/Resources/ref/mm10/Gencode_VM23/fromUCSC/KnownGene/KnownGene_GRCm38_introns.bed", header=F, sep="\t",stringsAsFactors = F,data.table=F)
  introns=introns[grepl("_",introns$V1)==F,]
  colnames(introns)=c('chr','start','end','attribute','V5','strand')
  introns=separate(introns,attribute,into=c('transcript_id','feature','exon_number','level','chr2','intronnumber','dir'),remove = T,sep = "_")
  introns$transcript_id=removeVersion(introns$transcript_id)
  # introns_canon=introns[introns$transcript_id%in%canonical$transcript_id,]
  introns$start=introns$start+1
  introns$exon_number=as.numeric(introns$exon_number)+1
        introns=introns[introns$transcript_id%in%canonical$transcript,]

introns.GR=GRanges(seqnames=introns$chr, ranges=IRanges(start=introns$start,end=introns$end),
strand=introns$strand,transcript_id=introns$transcript_id,exon_number=introns$exon_number)


        # gr=peaksIN.GR
        gr = introns.GR
        gr2 = LISI.GR
        xo=as.data.frame(GenomicRanges::findOverlaps(gr,gr2,type = "any",ignore.strand=F))
          
        LISI_int.GR=gr2[xo$subjectHits]


        
##################################################
pt2_int <- overlapPermTest(A=peaksIN.GR, B=LISI_int.GR, ntimes=100,count.once=TRUE)

LZ2_int=localZScore(pt=pt2_int,
                A=peaksIN.GR,
                B=LISI_int.GR, 
                window=1000,step=50,count.once=TRUE)
plot(pt2_int)
plot(LZ2_int)

####################
pt_int= permTest(A=peaksIN.GR, 
                 B=LISI_int.GR, 
                 randomize.function=randomizeRegions,evaluate.function=meanDistance,ntimes=100)

LZ_int=localZScore(pt=pt_int,
                A=peaksIN.GR,
                B=LISI_int.GR, 
                window=1000,step=50,count.once=TRUE)
plot(pt_int)
plot(LZ)


```
